// @databricks/sdk-experimental/dist/config/oauth/Client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
/* eslint-disable @typescript-eslint/naming-convention */
const Token_1 = require("../Token");
const fetch_1 = require("../../fetch");
const BasicCredentials_1 = require("../BasicCredentials");
const querystring = __importStar(require("querystring"));
class Client {
    constructor(issuer, options) {
        this.issuer = issuer;
        this.options = options;
        options.useParams = options.useParams ?? false;
        options.useHeader = options.useHeader ?? false;
        options.headers = options.headers ?? new fetch_1.Headers();
    }
    requestResource(url, token, requestOptions) {
        const newHeaders = new fetch_1.Headers(this.options.headers);
        newHeaders.set("Authorization", `Bearer ${token.accessToken}`);
        requestOptions.headers = newHeaders;
        return this.fetch(url.toString(), requestOptions);
    }
    async exchangeToken(params) {
        params.grant_type = "client_credentials";
        const requestOptions = {
            method: "POST",
            headers: this.options.headers,
        };
        if (this.options.useParams) {
            params["client_id"] = this.options.clientId;
            params["client_secret"] = this.options.clientSecret;
        }
        else if (this.options.useHeader) {
            requestOptions.headers = {
                ...requestOptions.headers,
                Authorization: (0, BasicCredentials_1.getBasicAuthHeader)(this.options.clientId, this.options.clientSecret),
            };
        }
        requestOptions.headers = {
            ...requestOptions.headers,
            "Content-Type": "application/x-www-form-urlencoded",
        };
        requestOptions.body = querystring.stringify(params);
        const response = await this.fetch(this.issuer.tokenEndpoint.toString(), requestOptions);
        if (!response.ok) {
            if (response.headers
                .get("content-type")
                ?.includes("application/json")) {
                const json = (await response.json());
                const code = json.errorCode ||
                    json.error_code ||
                    json.error ||
                    "Unknown";
                const summary = (json.errorSummary ||
                    json.error_description ||
                    "Unknown").replace(/\r?\n/g, " ");
                throw new Error(`Failed to retrieve token: ${code} ${summary}`);
            }
            else {
                throw new Error(`Failed to retrieve token: ${response.status} ${response.statusText}`);
            }
        }
        const tokenSet = await response.json();
        if (!isTokenSet(tokenSet)) {
            throw new Error(`Failed to retrieve token: ${JSON.stringify(tokenSet)}`);
        }
        return new Token_1.Token({
            accessToken: tokenSet.access_token,
            expiry: Date.now() + parseInt(tokenSet.expires_in) * 1000,
        });
    }
    fetch(url, options) {
        return (0, fetch_1.fetch)(url, options);
    }
}
exports.Client = Client;
function isTokenSet(tokenSet) {
    return (typeof tokenSet === "object" &&
        tokenSet !== null &&
        "access_token" in tokenSet &&
        "expires_in" in tokenSet);
}
