// @databricks/appkit-ui/dist/react/table/data-table.js
import { formatFieldLabel } from "../lib/format.js";
import { Button } from "../ui/button.js";
import { DropdownMenu, DropdownMenuCheckboxItem, DropdownMenuContent, DropdownMenuTrigger } from "../ui/dropdown-menu.js";
import { Table as Table$1, TableBody, TableCell, TableHead, TableHeader, TableRow } from "../ui/table.js";
import { Input } from "../ui/input.js";
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "../ui/select.js";
import { TableWrapper } from "./table-wrapper.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { ChevronDown } from "lucide-react";
import { flexRender } from "@tanstack/react-table";

//#region src/react/table/data-table.tsx
/**
* Production-ready data table with automatic data fetching and state management
*
* Features:
*  - Automatic column generation from data structure
*  - Integrated with useAnalyticsQuery for data fetching
*  - Built-in loading, error, and empty states
*  - Dynamic filtering, sorting and pagination
*  - Column visibility controls
*  - Responsive design
*  - Supports opinionated mode (auto columns) and full-control mode (`children(table)`)
*
* @param props - Props for the DataTable component
* @param props.queryKey - The query key to fetch the data
* @param props.parameters - The parameters to pass to the query. Required - use `{}` if none.
* @param props.filterColumn - The column to filter by
* @param props.filterPlaceholder - The placeholder for the filter input
* @param props.children - Optional children for full control mode
* @returns - The rendered data table component
*
* @example
* // Opinionated mode
* <DataTable
*  queryKey="users-list"
*  parameters={{ status: "active" }}
*  filterColumn="email"
*  filterPlaceholder="Filter by email..."
* />
* @example
* // full control mode
* <DataTable queryKey="users-list" parameters={{ status: "active" }}>
*   {(table) => (
*     <div>
*       <h2>Custom Table UI</h2>
*       {table.getRowModel().rows.map(row => (
*         <div key={row.id}>{row.original.name}</div>
*       ))}
*     </div>
*   )}
* </DataTable>
*/
function DataTable(props) {
	const { parameters, queryKey, filterColumn, filterPlaceholder, transform, labels, ariaLabel, testId, className, enableRowSelection, onRowSelectionChange, children, pageSize = 10, pageSizeOptions = [
		10,
		25,
		50,
		100
	] } = props;
	const finalLabels = {
		columnsButton: "Columns",
		noResults: "No results found.",
		rowsFound: `\${count} row(s) found`,
		previousButton: "Previous",
		nextButton: "Next",
		rowsPerPage: "Rows per page",
		showing: `Showing \${from} to \${to} of \${total}`,
		...labels
	};
	return /* @__PURE__ */ jsx(TableWrapper, {
		queryKey,
		parameters,
		ariaLabel,
		testId,
		className,
		transformer: transform,
		enableRowSelection,
		onRowSelectionChange,
		pageSize,
		children: (table) => {
			if (children) return children(table);
			const data = table.options.data;
			const defaultFilterColumn = filterColumn || (data && data.length > 0 ? Object.keys(data[0]).find((key) => typeof data[0][key] === "string") : null);
			const totalRows = table.getFilteredRowModel().rows.length;
			const currentPage = table.getState().pagination.pageIndex + 1;
			const currentPageSize = table.getState().pagination.pageSize;
			const fromRow = totalRows === 0 ? 0 : (currentPage - 1) * currentPageSize + 1;
			const toRow = Math.min(currentPage * currentPageSize, totalRows);
			return /* @__PURE__ */ jsxs("div", {
				className: "w-full",
				children: [
					/* @__PURE__ */ jsxs("div", {
						className: "flex items-center py-4 gap-2",
						children: [defaultFilterColumn && /* @__PURE__ */ jsx(Input, {
							placeholder: filterPlaceholder || `Filter by ${formatFieldLabel(defaultFilterColumn)}...`,
							value: table.getColumn(defaultFilterColumn)?.getFilterValue() ?? "",
							onChange: (event) => table.getColumn(defaultFilterColumn)?.setFilterValue(event.target.value),
							className: "max-w-sm"
						}), /* @__PURE__ */ jsxs(DropdownMenu, { children: [/* @__PURE__ */ jsx(DropdownMenuTrigger, {
							asChild: true,
							children: /* @__PURE__ */ jsxs(Button, {
								variant: "outline",
								className: "ml-auto",
								children: [
									finalLabels.columnsButton,
									" ",
									/* @__PURE__ */ jsx(ChevronDown, { className: "ml-2 h-4 w-4" })
								]
							})
						}), /* @__PURE__ */ jsx(DropdownMenuContent, {
							align: "end",
							children: table.getAllColumns().filter((column) => column.getCanHide()).map((column) => {
								return /* @__PURE__ */ jsx(DropdownMenuCheckboxItem, {
									className: "capitalize",
									checked: column.getIsVisible(),
									onCheckedChange: (value) => column.toggleVisibility(!!value),
									children: formatFieldLabel(column.id)
								}, column.id);
							})
						})] })]
					}),
					/* @__PURE__ */ jsx("div", {
						className: "overflow-hidden rounded-md border border-border",
						children: /* @__PURE__ */ jsx("div", {
							className: "max-h-[600px] overflow-y-auto",
							children: /* @__PURE__ */ jsxs(Table$1, { children: [/* @__PURE__ */ jsx(TableHeader, {
								className: "sticky top-0 bg-background z-10",
								children: table.getHeaderGroups().map((headerGroup) => /* @__PURE__ */ jsx(TableRow, { children: headerGroup.headers.map((header) => {
									const isSelectColumn = header.column.id === "select";
									return /* @__PURE__ */ jsx(TableHead, {
										style: {
											width: header.getSize(),
											position: "relative"
										},
										className: isSelectColumn ? "w-[40px] p-0" : void 0,
										children: header.isPlaceholder ? null : flexRender(header.column.columnDef.header, header.getContext())
									}, header.id);
								}) }, headerGroup.id))
							}), /* @__PURE__ */ jsx(TableBody, { children: table.getRowModel().rows?.length ? table.getRowModel().rows.map((row) => /* @__PURE__ */ jsx(TableRow, {
								"data-state": row.getIsSelected() && "selected",
								children: row.getVisibleCells().map((cell) => {
									const isSelectColumn = cell.column.id === "select";
									return /* @__PURE__ */ jsx(TableCell, {
										style: { width: cell.column.getSize() },
										className: isSelectColumn ? "w-[40px] p-0" : void 0,
										children: flexRender(cell.column.columnDef.cell, cell.getContext())
									}, cell.id);
								})
							}, row.id)) : /* @__PURE__ */ jsx(TableRow, { children: /* @__PURE__ */ jsx(TableCell, {
								colSpan: table.getAllColumns().length,
								className: "h-24 text-center",
								children: finalLabels.noResults
							}) }) })] })
						})
					}),
					/* @__PURE__ */ jsxs("div", {
						className: "flex flex-col-reverse sm:flex-row items-start sm:items-center justify-between gap-3 py-4",
						children: [/* @__PURE__ */ jsxs("div", {
							className: "flex flex-wrap items-center gap-3 sm:gap-6",
							children: [/* @__PURE__ */ jsx("div", {
								className: "text-foreground text-sm",
								children: totalRows > 0 ? finalLabels.showing.replace(`\${from}`, fromRow.toString()).replace(`\${to}`, toRow.toString()).replace(`\${total}`, totalRows.toString()) : finalLabels.noResults
							}), /* @__PURE__ */ jsxs("div", {
								className: "flex items-center gap-2",
								children: [/* @__PURE__ */ jsx("span", {
									className: "text-sm text-foreground",
									children: finalLabels.rowsPerPage
								}), /* @__PURE__ */ jsxs(Select, {
									value: currentPageSize.toString(),
									onValueChange: (value) => {
										table.setPageSize(Number(value));
									},
									children: [/* @__PURE__ */ jsx(SelectTrigger, {
										className: "w-[70px]",
										children: /* @__PURE__ */ jsx(SelectValue, {})
									}), /* @__PURE__ */ jsx(SelectContent, { children: pageSizeOptions.map((size) => /* @__PURE__ */ jsx(SelectItem, {
										value: size.toString(),
										children: size
									}, size)) })]
								})]
							})]
						}), /* @__PURE__ */ jsxs("div", {
							className: "space-x-2",
							children: [/* @__PURE__ */ jsx(Button, {
								variant: "outline",
								size: "sm",
								onClick: () => table.previousPage(),
								disabled: !table.getCanPreviousPage(),
								children: finalLabels.previousButton
							}), /* @__PURE__ */ jsx(Button, {
								variant: "outline",
								size: "sm",
								onClick: () => table.nextPage(),
								disabled: !table.getCanNextPage(),
								children: finalLabels.nextButton
							})]
						})]
					})
				]
			});
		}
	});
}

//#endregion
export { DataTable };
