// @databricks/appkit-ui/dist/react/table/table-wrapper.js
import { useAnalyticsQuery } from "../hooks/use-analytics-query.js";
import { SAFE_KEY_REGEX, formatChartValue, formatFieldLabel } from "../lib/format.js";
import { Button } from "../ui/button.js";
import { Checkbox } from "../ui/checkbox.js";
import { EmptyState } from "./empty.js";
import { ErrorState } from "./error.js";
import { LoadingSkeleton } from "./loading.js";
import "../index.js";
import { useEffect, useMemo, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
import { ArrowUpDown } from "lucide-react";
import { getCoreRowModel, getFilteredRowModel, getPaginationRowModel, getSortedRowModel, useReactTable } from "@tanstack/react-table";

//#region src/react/table/table-wrapper.tsx
const CHECKBOX_COLUMN_WIDTH = 40;
/**
* Wrapper component for tables with automatic data fetching and state management
* This component handles:
* - Data fetching via useAnalyticsQuery
* - Loading, error, and empty states with proper UI components
* - Data transformation (optional)
* - Dynamic column generation from data structure
* - TanStack Table instance creation with all features (sorting, filtering, pagination, etc.)
*
* @template TRaw - The raw data type returned by the analytics query
* @template TProcessed - The processed data type after transformation
*
* @param props - Props for the TableWrapper component
* @param props.queryKey - The query key to fetch the data
* @param props.parameters - The parameters to pass to the query
* @param props.transformer - Optional function to transform raw data before creating table
* @param props.children - Render function that receives the TanStack Table instance
* @param props.className - Optional CSS class name for the wrapper
* @param props.ariaLabel - Optional accessibility label
* @param props.testId - Optional test ID for testing
*
* @returns The rendered table with state management
*/
function TableWrapper(props) {
	const { queryKey, parameters, transformer, children, className, ariaLabel, testId, enableRowSelection = false, onRowSelectionChange, pageSize = 10 } = props;
	const [sorting, setSorting] = useState([]);
	const [columnFilters, setColumnFilters] = useState([]);
	const [columnVisibility, setColumnVisibility] = useState({});
	const [rowSelection, setRowSelection] = useState({});
	const { data, loading, error } = useAnalyticsQuery(queryKey, parameters);
	useEffect(() => {
		if (onRowSelectionChange && enableRowSelection) onRowSelectionChange(rowSelection);
	}, [
		rowSelection,
		onRowSelectionChange,
		enableRowSelection
	]);
	const hasData = data && data.length > 0;
	const processedData = hasData ? transformer ? transformer(data) : data : [];
	const table = useReactTable({
		data: processedData,
		columns: useMemo(() => {
			if (!hasData) return [];
			if (!processedData[0] || typeof processedData[0] !== "object") {
				console.warn("Invalid data format for DataTable");
				return [];
			}
			const dataColumns = Object.keys(processedData[0]).filter((key) => SAFE_KEY_REGEX.test(key)).map((key) => {
				const formattedLabel = formatFieldLabel(key);
				return {
					accessorKey: key,
					header: ({ column }) => {
						return /* @__PURE__ */ jsxs(Button, {
							variant: "ghost",
							onClick: () => column.toggleSorting(column.getIsSorted() === "asc"),
							className: "h-8 px-2",
							children: [formattedLabel, /* @__PURE__ */ jsx(ArrowUpDown, { className: "ml-2 h-4 w-4" })]
						});
					},
					cell: ({ row }) => {
						const value = row.getValue(key);
						if (typeof value === "number" || Number.isFinite(Number(value))) return /* @__PURE__ */ jsx("div", {
							className: "text-right font-mono",
							children: formatChartValue(Number(value), key)
						});
						return /* @__PURE__ */ jsx("div", { children: String(value) });
					}
				};
			});
			if (enableRowSelection) return [{
				id: "select",
				maxSize: CHECKBOX_COLUMN_WIDTH,
				minSize: CHECKBOX_COLUMN_WIDTH,
				header: ({ table }) => /* @__PURE__ */ jsx("div", {
					className: "flex items-center justify-center",
					children: /* @__PURE__ */ jsx(Checkbox, {
						checked: table.getIsAllPageRowsSelected() || table.getIsSomePageRowsSelected() && "indeterminate",
						onCheckedChange: (value) => table.toggleAllPageRowsSelected(!!value),
						"aria-label": "Select all"
					})
				}),
				cell: ({ row }) => /* @__PURE__ */ jsx("div", {
					className: "flex items-center justify-center",
					children: /* @__PURE__ */ jsx(Checkbox, {
						checked: row.getIsSelected(),
						onCheckedChange: (value) => row.toggleSelected(!!value),
						"aria-label": "Select row"
					})
				}),
				enableSorting: false,
				enableHiding: false
			}, ...dataColumns];
			return dataColumns;
		}, [
			hasData,
			processedData,
			enableRowSelection
		]),
		onSortingChange: setSorting,
		onColumnFiltersChange: setColumnFilters,
		getCoreRowModel: getCoreRowModel(),
		getPaginationRowModel: getPaginationRowModel(),
		getSortedRowModel: getSortedRowModel(),
		getFilteredRowModel: getFilteredRowModel(),
		onColumnVisibilityChange: setColumnVisibility,
		onRowSelectionChange: setRowSelection,
		state: {
			sorting,
			columnFilters,
			columnVisibility,
			rowSelection
		},
		initialState: { pagination: { pageSize } }
	});
	if (loading) return /* @__PURE__ */ jsx(LoadingSkeleton, {});
	if (error) return /* @__PURE__ */ jsx(ErrorState, { error: typeof error === "string" ? error : "Unknown error" });
	if (!hasData) return /* @__PURE__ */ jsx(EmptyState, {});
	return /* @__PURE__ */ jsx("section", {
		className,
		"aria-label": ariaLabel,
		"data-testid": testId,
		children: children(table)
	});
}

//#endregion
export { TableWrapper };
