// @databricks/appkit/dist/plugins/lakebase/manifest.js
//#region src/plugins/lakebase/manifest.json
var manifest_default = {
	$schema: "https://databricks.github.io/appkit/schemas/plugin-manifest.schema.json",
	name: "lakebase",
	displayName: "Lakebase",
	description: "SQL query execution against Databricks Lakebase Autoscaling",
	hidden: false,
	resources: {
		"required": [{
			"type": "postgres",
			"alias": "Postgres",
			"resourceKey": "postgres",
			"description": "Lakebase Postgres database for persistent storage",
			"permission": "CAN_CONNECT_AND_CREATE",
			"fields": {
				"project": {
					"description": "Full Lakebase Postgres project resource name. Obtain by running `databricks postgres list-projects`, select the desired item from the output array and use its .name value.",
					"examples": ["projects/{project-id}"],
					"discovery": {
						"type": "kind",
						"resourceKind": "postgres_project",
						"select": "name"
					}
				},
				"branch": {
					"description": "Full Lakebase Postgres branch resource name. Obtain by running `databricks postgres list-branches {project-name}`, select the desired item from the output array and use its .name value. Requires the project resource name.",
					"examples": ["projects/{project-id}/branches/{branch-id}"],
					"discovery": {
						"type": "kind",
						"resourceKind": "postgres_branch",
						"select": "name",
						"dependsOn": "project"
					}
				},
				"database": {
					"description": "Full Lakebase Postgres database resource name. Obtain by running `databricks postgres list-databases {branch-name}`, select the desired item from the output array and use its .name value. Requires the branch resource name.",
					"examples": ["projects/{project-id}/branches/{branch-id}/databases/{database-id}"],
					"discovery": {
						"type": "kind",
						"resourceKind": "postgres_database",
						"select": "name",
						"dependsOn": "branch"
					}
				},
				"host": {
					"env": "PGHOST",
					"localOnly": true,
					"resolve": "postgres:host",
					"description": "Postgres host for local development. Auto-injected by the platform at deploy time."
				},
				"databaseName": {
					"env": "PGDATABASE",
					"localOnly": true,
					"resolve": "postgres:databaseName",
					"description": "Postgres database name for local development. Auto-injected by the platform at deploy time."
				},
				"endpointPath": {
					"env": "LAKEBASE_ENDPOINT",
					"bundleIgnore": true,
					"resolve": "postgres:endpointPath",
					"description": "Lakebase endpoint resource name. Auto-injected at runtime via app.yaml valueFrom: postgres. For local development, obtain by running `databricks postgres list-endpoints {branch-name}`, select the desired item from the output array and use its .name value.",
					"examples": ["projects/{project-id}/branches/{branch-id}/endpoints/{endpoint-id}"]
				},
				"port": {
					"env": "PGPORT",
					"localOnly": true,
					"value": "5432",
					"description": "Postgres port. Auto-injected by the platform at deploy time."
				},
				"sslmode": {
					"env": "PGSSLMODE",
					"localOnly": true,
					"value": "require",
					"description": "Postgres SSL mode. Auto-injected by the platform at deploy time."
				}
			}
		}],
		"optional": []
	},
	scaffolding: { "rules": { "should": ["After init, run any database migrations for your chosen ORM before first request", "After init, verify Lakebase connectivity with 'psql $PGHOST -c \"select 1\"'"] } }
};

//#endregion
export { manifest_default as default };
