// @databricks/appkit/dist/registry/manifest-loader.js
import { createLogger } from "../logging/logger.js";
import { ConfigurationError } from "../errors/configuration.js";
import "../errors/index.js";
import { PERMISSIONS_BY_TYPE, ResourceType } from "./types.generated.js";
import "./types.js";

//#region src/registry/manifest-loader.ts
const logger = createLogger("manifest-loader");
function normalizeType(s) {
	const v = Object.values(ResourceType).find((x) => x === s);
	if (v !== void 0) return v;
	throw new ConfigurationError(`Invalid resource type: "${s}". Valid: ${Object.values(ResourceType).join(", ")}`);
}
function normalizePermission(type, s) {
	const allowed = PERMISSIONS_BY_TYPE[type];
	if (allowed.includes(s)) return s;
	throw new ConfigurationError(`Invalid permission "${s}" for type ${type}. Valid: ${allowed.join(", ")}`);
}
function normalizeResource(r) {
	const type = normalizeType(r.type);
	const permission = normalizePermission(type, r.permission);
	return {
		...r,
		type,
		permission,
		required: false
	};
}
/**
* Loads and validates the manifest from a plugin constructor.
* Normalizes string type/permission to strict ResourceType/ResourcePermission.
*
* @param plugin - The plugin constructor class
* @returns The validated, normalized plugin manifest
* @throws {ConfigurationError} If the manifest is missing, invalid, or has invalid resource type/permission
*/
function getPluginManifest(plugin) {
	const pluginName = plugin.manifest?.name || plugin.name || "unknown";
	if (!plugin.manifest) throw new ConfigurationError(`Plugin ${pluginName} is missing a manifest. All plugins must declare a static manifest property.`);
	const raw = plugin.manifest;
	if (!raw.name || typeof raw.name !== "string") throw new ConfigurationError(`Plugin ${pluginName} manifest has missing or invalid 'name' field`);
	if (!raw.displayName || typeof raw.displayName !== "string") throw new ConfigurationError(`Plugin ${raw.name} manifest has missing or invalid 'displayName' field`);
	if (!raw.description || typeof raw.description !== "string") throw new ConfigurationError(`Plugin ${raw.name} manifest has missing or invalid 'description' field`);
	if (!raw.resources) throw new ConfigurationError(`Plugin ${raw.name} manifest is missing 'resources' field`);
	if (!Array.isArray(raw.resources.required)) throw new ConfigurationError(`Plugin ${raw.name} manifest has invalid 'resources.required' field (expected array)`);
	if (raw.resources.optional !== void 0 && !Array.isArray(raw.resources.optional)) throw new ConfigurationError(`Plugin ${raw.name} manifest has invalid 'resources.optional' field (expected array)`);
	const required = raw.resources.required.map((r) => {
		const { required: _, ...rest } = normalizeResource(r);
		return rest;
	});
	const optional = (raw.resources.optional || []).map((r) => {
		const { required: _, ...rest } = normalizeResource(r);
		return rest;
	});
	logger.debug("Loaded manifest for plugin %s: %d required resources, %d optional resources", raw.name, required.length, optional.length);
	return {
		...raw,
		resources: {
			required,
			optional
		}
	};
}
/**
* Gets the resource requirements from a plugin's manifest.
*
* Combines required and optional resources into a single array with the
* `required` flag set appropriately.
*
* @param plugin - The plugin constructor class
* @returns Combined array of required and optional resources
* @throws {ConfigurationError} If the plugin manifest is missing or invalid
*
* @example
* ```typescript
* const resources = getResourceRequirements(AnalyticsPlugin);
* for (const resource of resources) {
*   console.log(`${resource.type}: ${resource.description} (required: ${resource.required})`);
* }
* ```
*/
function getResourceRequirements(plugin) {
	const manifest = getPluginManifest(plugin);
	const required = manifest.resources.required.map((r) => ({
		...r,
		required: true
	}));
	const optional = (manifest.resources.optional || []).map((r) => ({
		...r,
		required: false
	}));
	return [...required, ...optional];
}

//#endregion
export { getPluginManifest, getResourceRequirements };
