// @databricks/appkit/dist/registry/types.generated.js
//#region src/registry/types.generated.ts
/** Resource types from resourceTypeSchema.options */
let ResourceType = /* @__PURE__ */ function(ResourceType) {
	ResourceType["SECRET"] = "secret";
	ResourceType["JOB"] = "job";
	ResourceType["SQL_WAREHOUSE"] = "sql_warehouse";
	ResourceType["SERVING_ENDPOINT"] = "serving_endpoint";
	ResourceType["VOLUME"] = "volume";
	ResourceType["VECTOR_SEARCH_INDEX"] = "vector_search_index";
	ResourceType["UC_FUNCTION"] = "uc_function";
	ResourceType["UC_CONNECTION"] = "uc_connection";
	ResourceType["DATABASE"] = "database";
	ResourceType["POSTGRES"] = "postgres";
	ResourceType["GENIE_SPACE"] = "genie_space";
	ResourceType["EXPERIMENT"] = "experiment";
	ResourceType["APP"] = "app";
	return ResourceType;
}({});
/** Permission hierarchy per resource type (weakest to strongest). Schema enum order. */
const PERMISSION_HIERARCHY_BY_TYPE = {
	[ResourceType.SECRET]: [
		"READ",
		"WRITE",
		"MANAGE"
	],
	[ResourceType.JOB]: [
		"CAN_VIEW",
		"CAN_MANAGE_RUN",
		"CAN_MANAGE"
	],
	[ResourceType.SQL_WAREHOUSE]: ["CAN_USE", "CAN_MANAGE"],
	[ResourceType.SERVING_ENDPOINT]: [
		"CAN_VIEW",
		"CAN_QUERY",
		"CAN_MANAGE"
	],
	[ResourceType.VOLUME]: ["READ_VOLUME", "WRITE_VOLUME"],
	[ResourceType.VECTOR_SEARCH_INDEX]: ["SELECT"],
	[ResourceType.UC_FUNCTION]: ["EXECUTE"],
	[ResourceType.UC_CONNECTION]: ["USE_CONNECTION"],
	[ResourceType.DATABASE]: ["CAN_CONNECT_AND_CREATE"],
	[ResourceType.POSTGRES]: ["CAN_CONNECT_AND_CREATE"],
	[ResourceType.GENIE_SPACE]: [
		"CAN_VIEW",
		"CAN_RUN",
		"CAN_EDIT",
		"CAN_MANAGE"
	],
	[ResourceType.EXPERIMENT]: [
		"CAN_READ",
		"CAN_EDIT",
		"CAN_MANAGE"
	],
	[ResourceType.APP]: ["CAN_USE"]
};
/** Set of valid permissions per type (for validation). */
const PERMISSIONS_BY_TYPE = PERMISSION_HIERARCHY_BY_TYPE;

//#endregion
export { PERMISSIONS_BY_TYPE, PERMISSION_HIERARCHY_BY_TYPE, ResourceType };
