// @databricks/appkit/dist/errors/base.js
//#region src/errors/base.ts
/** Fields that should be redacted in logs/serialization for security */
const SENSITIVE_FIELD_PATTERNS = [
	/token/i,
	/password/i,
	/secret/i,
	/credential/i,
	/auth/i,
	/key$/i,
	/apikey/i
];
/**
* Base error class for all AppKit errors.
* Provides a consistent structure for error handling across the framework.
*
* @example
* ```typescript
* // Catching errors by type
* try {
*   await lakebase.query("...");
* } catch (e) {
*   if (e instanceof AuthenticationError) {
*     // Re-authenticate
*   } else if (e instanceof ConnectionError && e.isRetryable) {
*     // Retry with backoff
*   }
* }
*
* // Logging errors
* console.error(error.toJSON()); // Safe for logging, sensitive values redacted
* ```
*/
var AppKitError = class extends Error {
	/** Optional cause of the error */
	cause;
	/** Additional context for the error */
	context;
	constructor(message, options) {
		super(message);
		this.name = this.constructor.name;
		this.cause = options?.cause;
		this.context = options?.context;
		if (Error.captureStackTrace) Error.captureStackTrace(this, this.constructor);
	}
	/**
	* Convert error to JSON for logging/serialization.
	* Sensitive values in context are automatically redacted.
	*/
	toJSON() {
		return {
			name: this.name,
			code: this.code,
			message: this.message,
			statusCode: this.statusCode,
			isRetryable: this.isRetryable,
			context: this.sanitizeContext(this.context),
			cause: this.cause?.message,
			stack: this.stack
		};
	}
	/**
	* Create a human-readable string representation
	*/
	toString() {
		let str = `${this.name} [${this.code}]: ${this.message}`;
		if (this.cause) str += ` (caused by: ${this.cause.message})`;
		return str;
	}
	/**
	* Sanitize context by redacting sensitive field values
	*/
	sanitizeContext(context) {
		if (!context) return void 0;
		const sanitized = {};
		for (const [key, value] of Object.entries(context)) if (this.isSensitiveField(key)) sanitized[key] = "[REDACTED]";
		else if (typeof value === "object" && value !== null) sanitized[key] = Array.isArray(value) ? `[Array(${value.length})]` : "[Object]";
		else sanitized[key] = value;
		return sanitized;
	}
	/**
	* Check if a field name matches sensitive patterns
	*/
	isSensitiveField(fieldName) {
		return SENSITIVE_FIELD_PATTERNS.some((pattern) => pattern.test(fieldName));
	}
};

//#endregion
export { AppKitError };
