// @databricks/appkit/dist/errors/configuration.js
import { AppKitError } from "./base.js";
import pc from "picocolors";

//#region src/errors/configuration.ts
function authSetupVerbose() {
	return process.env.APPKIT_VERBOSE_AUTH_ERRORS === "1" || process.env.APPKIT_VERBOSE_AUTH_ERRORS === "true";
}
/** Pulls ` $ databricks ...` from SDK text when present. */
function suggestedDatabricksCliCommand(detail) {
	return detail.match(/\$\s*(databricks[^\n]+)/)?.[1]?.trim();
}
/** Makes `console.error` show only the message (no stack, no extra fields). */
function pinUserFacingAuthError(err) {
	Object.defineProperty(err, "stack", {
		value: "",
		configurable: true,
		enumerable: false,
		writable: true
	});
	Object.defineProperty(err, Symbol.for("nodejs.util.inspect.custom"), {
		value: function() {
			return this.message;
		},
		enumerable: false,
		configurable: true
	});
}
/**
* Error thrown when configuration is missing or invalid.
* Use for missing environment variables, invalid settings, or setup issues.
*
* @example
* ```typescript
* throw new ConfigurationError("DATABRICKS_HOST environment variable is required");
* throw new ConfigurationError("Warehouse ID not found", { context: { env: "production" } });
* ```
*/
var ConfigurationError = class ConfigurationError extends AppKitError {
	code = "CONFIGURATION_ERROR";
	statusCode = 500;
	isRetryable = false;
	/**
	* Create a configuration error for missing environment variable
	*/
	static missingEnvVar(varName) {
		return new ConfigurationError(`${varName} environment variable is required`, { context: { envVar: varName } });
	}
	/**
	* Create a configuration error for missing resource
	*/
	static resourceNotFound(resource, hint) {
		return new ConfigurationError(hint ? `${resource} not found. ${hint}` : `${resource} not found`, { context: { resource } });
	}
	/**
	* Create a configuration error for invalid connection config
	*/
	static invalidConnection(service, details) {
		return new ConfigurationError(details ? `${service} connection not configured. ${details}` : `${service} connection not configured`, { context: { service } });
	}
	/**
	* Create a configuration error for missing connection string parameter
	*/
	static missingConnectionParam(param) {
		return new ConfigurationError(`Connection string must include ${param} parameter`, { context: { parameter: param } });
	}
	/**
	* Databricks CLI / token auth failed while creating the workspace client.
	*
	* By default the message is short; key lines use **picocolors** when the
	* terminal supports it (also respects `NO_COLOR`). `console.error` won’t show
	* stacks or `{ code, context, … }`. Set `APPKIT_VERBOSE_AUTH_ERRORS=1` for full
	* `cause`, stack, and the raw SDK message (verbose appendix is unstyled).
	*/
	static databricksAuthenticationSetupFailed(detail, options) {
		const verbose = authSetupVerbose();
		const host = process.env.DATABRICKS_HOST ?? "(not set)";
		const warehouseId = process.env.DATABRICKS_WAREHOUSE_ID;
		const d = detail.trim();
		const cli = suggestedDatabricksCliCommand(d);
		const lines = [
			pc.bold(pc.red("Databricks authentication failed.")),
			"",
			cli ? `${pc.bold("Run this, then try again:")}\n  ${pc.cyan(cli)}` : pc.yellow("Log in with the Databricks CLI (for example, databricks auth login for this workspace), then try again."),
			"",
			pc.dim("Or set DATABRICKS_TOKEN and DATABRICKS_HOST instead of CLI-based auth."),
			"",
			`${pc.green("DATABRICKS_HOST")}: ${host}`
		];
		if (warehouseId) lines.push(`${pc.green("DATABRICKS_WAREHOUSE_ID")}: ${warehouseId}`);
		if (verbose) lines.push("", d);
		const err = new ConfigurationError(lines.join("\n"), { cause: verbose ? options?.cause : void 0 });
		if (!verbose) pinUserFacingAuthError(err);
		return err;
	}
};

//#endregion
export { ConfigurationError };
