// @databricks/appkit/dist/errors/execution.js
import { AppKitError } from "./base.js";

//#region src/errors/execution.ts
/**
* Error thrown when an operation execution fails.
* Use for statement failures, canceled operations, or unexpected states.
*
* @example
* ```typescript
* throw new ExecutionError("Statement failed: syntax error");
* throw new ExecutionError("Statement was canceled");
* ```
*/
var ExecutionError = class ExecutionError extends AppKitError {
	code = "EXECUTION_ERROR";
	statusCode = 500;
	isRetryable = false;
	/**
	* Create an execution error for statement failure
	*/
	static statementFailed(errorMessage) {
		return new ExecutionError(errorMessage ? `Statement failed: ${errorMessage}` : "Statement failed: Unknown error");
	}
	/**
	* Create an execution error for canceled operation
	*/
	static canceled() {
		return new ExecutionError("Statement was canceled");
	}
	/**
	* Create an execution error for closed/expired results
	*/
	static resultsClosed() {
		return new ExecutionError("Statement execution completed but results are no longer available (CLOSED state)");
	}
	/**
	* Create an execution error for unknown state
	*/
	static unknownState(state) {
		return new ExecutionError(`Unknown statement state: ${state}`, { context: { state } });
	}
	/**
	* Create an execution error for missing data
	*/
	static missingData(dataType) {
		return new ExecutionError(`No ${dataType} found in response`, { context: { dataType } });
	}
};

//#endregion
export { ExecutionError };
