// @databricks/appkit/dist/errors/tunnel.js
import { AppKitError } from "./base.js";

//#region src/errors/tunnel.ts
/**
* Error thrown when remote tunnel operations fail.
* Use for tunnel connection issues, message parsing failures, etc.
*
* @example
* ```typescript
* throw new TunnelError("No tunnel connection available");
* throw new TunnelError("Failed to parse WebSocket message", { cause: parseError });
* ```
*/
var TunnelError = class TunnelError extends AppKitError {
	code = "TUNNEL_ERROR";
	statusCode = 502;
	isRetryable = true;
	/**
	* Create a tunnel error for missing tunnel getter
	*/
	static getterNotRegistered() {
		return new TunnelError("Tunnel getter not registered for DevFileReader singleton");
	}
	/**
	* Create a tunnel error for no available connection
	*/
	static noConnection() {
		return new TunnelError("No tunnel connection available for file read");
	}
	/**
	* Create a tunnel error for asset fetch failure
	*/
	static fetchFailed(path, cause) {
		return new TunnelError("Failed to fetch asset", {
			cause,
			context: { path }
		});
	}
	/**
	* Create a tunnel error for message parsing failure
	*/
	static parseError(messageType, cause) {
		return new TunnelError(`Failed to parse ${messageType} message`, {
			cause,
			context: { messageType }
		});
	}
};

//#endregion
export { TunnelError };
