// @databricks/appkit/dist/plugin/interceptors/retry.js
import { createLogger } from "../../logging/logger.js";
import { AppKitError } from "../../errors/base.js";

//#region src/plugin/interceptors/retry.ts
const logger = createLogger("interceptors:retry");
/**
* Determines whether an error is safe to retry.
*
* Priority:
*  1. AppKitError — reads the `isRetryable` boolean property.
*  2. Databricks SDK ApiError (duck-typed) — calls `isRetryable()` method,
*     or falls back to status-code heuristic (5xx / 429 → retryable).
*  3. Unknown errors — treated as retryable to preserve backward compatibility.
*/
function isRetryableError(error) {
	if (error instanceof AppKitError) return error.isRetryable;
	if (error instanceof Error && "statusCode" in error) {
		const record = error;
		if (typeof record.statusCode !== "number") return true;
		if (typeof record.isRetryable === "function") return record.isRetryable();
		return record.statusCode >= 500 || record.statusCode === 429;
	}
	return true;
}
var RetryInterceptor = class {
	attempts;
	initialDelay;
	maxDelay;
	constructor(config) {
		this.attempts = config.attempts ?? 3;
		this.initialDelay = config.initialDelay ?? 1e3;
		this.maxDelay = config.maxDelay ?? 3e4;
	}
	async intercept(fn, context) {
		let lastError;
		for (let attempt = 1; attempt <= this.attempts; attempt++) try {
			const result = await fn();
			if (attempt > 1) logger.event()?.setExecution({ retry_attempts: attempt - 1 });
			return result;
		} catch (error) {
			lastError = error;
			if (attempt === this.attempts) {
				logger.event()?.setExecution({ retry_attempts: attempt - 1 });
				throw error;
			}
			if (context.signal?.aborted) throw error;
			if (!isRetryableError(error)) throw error;
			const delay = this.calculateDelay(attempt);
			await this.sleep(delay);
		}
		throw lastError;
	}
	calculateDelay(attempt) {
		const delay = this.initialDelay * 2 ** (attempt - 1);
		return Math.min(delay, this.maxDelay) * Math.random();
	}
	sleep(ms) {
		return new Promise((resolve) => setTimeout(resolve, ms));
	}
};

//#endregion
export { RetryInterceptor };
