// @databricks/appkit/dist/plugin/interceptors/telemetry.js
import { getCurrentUserId, isInUserContext } from "../../context/execution-context.js";
import { SpanStatusCode } from "../../telemetry/index.js";
import { isDevOboFallback } from "../plugin.js";

//#region src/plugin/interceptors/telemetry.ts
var TelemetryInterceptor = class {
	constructor(telemetry, config) {
		this.telemetry = telemetry;
		this.config = config;
	}
	async intercept(fn, context) {
		const spanName = this.config?.spanName || "plugin.execute";
		if (context.signal?.aborted) throw new Error("Operation aborted before execution");
		return this.telemetry.startActiveSpan(spanName, { attributes: this.config?.attributes }, async (span) => {
			let abortHandler;
			let isAborted = false;
			if (context.signal) {
				abortHandler = () => {
					if (!span.isRecording()) return;
					isAborted = true;
					span.setAttribute("cancelled", true);
					span.setStatus({
						code: SpanStatusCode.ERROR,
						message: "Operation cancelled by client"
					});
					span.end();
				};
				context.signal.addEventListener("abort", abortHandler, { once: true });
			}
			try {
				span.setAttribute("execution.context", isInUserContext() ? "user" : "service");
				span.setAttribute("caller.id", getCurrentUserId());
				if (isDevOboFallback()) span.setAttribute("execution.obo_dev_fallback", true);
				const result = await fn();
				if (!isAborted) span.setStatus({ code: SpanStatusCode.OK });
				return result;
			} catch (error) {
				if (!isAborted) {
					span.recordException(error);
					span.setStatus({ code: SpanStatusCode.ERROR });
				}
				throw error;
			} finally {
				if (abortHandler && context.signal) context.signal.removeEventListener("abort", abortHandler);
				if (!isAborted) span.end();
			}
		});
	}
};

//#endregion
export { TelemetryInterceptor };
