// @databricks/sdk-experimental/dist/apis/apps/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppsSettingsService = exports.AppsService = void 0;
const apps = __importStar(require("./model"));
const retries_1 = __importDefault(require("../../retries/retries"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
const wait_1 = require("../../wait");
class AppsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Apps", method, message);
    }
}
class AppsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Apps", method, message);
    }
}
/**
 * Apps run directly on a customer’s Databricks instance, integrate with their
 * data, use and extend Databricks services, and enable users to interact through
 * single sign-on.
 */
class AppsService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.app;
        const query = {};
        if (request.hasOwnProperty("no_compute")) {
            query["no_compute"] =
                request.no_compute;
        }
        const path = "/api/2.0/apps";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new app.
     */
    async create(createAppRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const app = await this._create(createAppRequest, context);
        return (0, wait_1.asWaiter)(app, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        name: app.name,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Apps.createAndWait: cancelled");
                        throw new AppsError("createAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.compute_status.state;
                    const statusMessage = pollResponse.compute_status.message;
                    switch (status) {
                        case "ACTIVE": {
                            return pollResponse;
                        }
                        case "ERROR":
                        case "STOPPED": {
                            const errorMessage = `failed to reach ACTIVE state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Apps.createAndWait: ${errorMessage}`);
                            throw new AppsError("createAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach ACTIVE state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Apps.createAndWait: retrying: ${errorMessage}`);
                            throw new AppsRetriableError("createAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _createUpdate(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("app")) {
            body["app"] = request.app;
        }
        if (request.hasOwnProperty("update_mask")) {
            body["update_mask"] = request.update_mask;
        }
        const query = {};
        const path = `/api/2.0/apps/${request.app_name}/update`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates an app update and starts the update process. The update process is
     * asynchronous and the status of the update can be checked with the
     * GetAppUpdate method.
     */
    async createUpdate(asyncUpdateAppRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const appUpdate = await this._createUpdate(asyncUpdateAppRequest, context);
        return (0, wait_1.asWaiter)(appUpdate, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.getUpdate({
                        app_name: asyncUpdateAppRequest.app_name,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Apps.createUpdateAndWait: cancelled");
                        throw new AppsError("createUpdateAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.status.state;
                    const statusMessage = pollResponse.status.message;
                    switch (status) {
                        case "SUCCEEDED": {
                            return pollResponse;
                        }
                        case "FAILED": {
                            const errorMessage = `failed to reach SUCCEEDED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Apps.createUpdateAndWait: ${errorMessage}`);
                            throw new AppsError("createUpdateAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach SUCCEEDED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Apps.createUpdateAndWait: retrying: ${errorMessage}`);
                            throw new AppsRetriableError("createUpdateAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/apps/${request.name}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes an app.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _deploy(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.app_deployment;
        const query = {};
        const path = `/api/2.0/apps/${request.app_name}/deployments`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates an app deployment for the app with the supplied name.
     */
    async deploy(createAppDeploymentRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const appDeployment = await this._deploy(createAppDeploymentRequest, context);
        return (0, wait_1.asWaiter)(appDeployment, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.getDeployment({
                        app_name: createAppDeploymentRequest.app_name,
                        deployment_id: appDeployment.deployment_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Apps.deployAndWait: cancelled");
                        throw new AppsError("deployAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.status.state;
                    const statusMessage = pollResponse.status.message;
                    switch (status) {
                        case "SUCCEEDED": {
                            return pollResponse;
                        }
                        case "FAILED": {
                            const errorMessage = `failed to reach SUCCEEDED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Apps.deployAndWait: ${errorMessage}`);
                            throw new AppsError("deployAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach SUCCEEDED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Apps.deployAndWait: retrying: ${errorMessage}`);
                            throw new AppsRetriableError("deployAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/apps/${request.name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieves information for the app with the supplied name.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _getDeployment(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/apps/${request.app_name}/deployments/${request.deployment_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieves information for the app deployment with the supplied name and
     * deployment id.
     */
    async getDeployment(request, context) {
        return await this._getDeployment(request, context);
    }
    async _getPermissionLevels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/apps/${request.app_name}/permissionLevels`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(request, context) {
        return await this._getPermissionLevels(request, context);
    }
    async _getPermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/apps/${request.app_name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of an app. Apps can inherit permissions from their
     * root object.
     */
    async getPermissions(request, context) {
        return await this._getPermissions(request, context);
    }
    async _getUpdate(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/apps/${request.app_name}/update`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the status of an app update.
     */
    async getUpdate(request, context) {
        return await this._getUpdate(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/apps";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists all apps in the workspace.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.apps || response.apps.length === 0) {
                break;
            }
            for (const v of response.apps) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _listDeployments(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/apps/${request.app_name}/deployments`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists all app deployments for the app with the supplied name.
     */
    async *listDeployments(request, context) {
        while (true) {
            const response = await this._listDeployments(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.app_deployments ||
                response.app_deployments.length === 0) {
                break;
            }
            for (const v of response.app_deployments) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _setPermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/apps/${request.app_name}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their root object.
     */
    async setPermissions(request, context) {
        return await this._setPermissions(request, context);
    }
    async _start(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/apps/${request.name}/start`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Start the last active deployment of the app in the workspace.
     */
    async start(startAppRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const app = await this._start(startAppRequest, context);
        return (0, wait_1.asWaiter)(app, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        name: app.name,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Apps.startAndWait: cancelled");
                        throw new AppsError("startAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.compute_status.state;
                    const statusMessage = pollResponse.compute_status.message;
                    switch (status) {
                        case "ACTIVE": {
                            return pollResponse;
                        }
                        case "ERROR":
                        case "STOPPED": {
                            const errorMessage = `failed to reach ACTIVE state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Apps.startAndWait: ${errorMessage}`);
                            throw new AppsError("startAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach ACTIVE state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Apps.startAndWait: retrying: ${errorMessage}`);
                            throw new AppsRetriableError("startAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _stop(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/apps/${request.name}/stop`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Stops the active deployment of the app in the workspace.
     */
    async stop(stopAppRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const app = await this._stop(stopAppRequest, context);
        return (0, wait_1.asWaiter)(app, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        name: app.name,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Apps.stopAndWait: cancelled");
                        throw new AppsError("stopAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.compute_status.state;
                    const statusMessage = pollResponse.compute_status.message;
                    switch (status) {
                        case "STOPPED": {
                            return pollResponse;
                        }
                        case "ERROR": {
                            const errorMessage = `failed to reach STOPPED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Apps.stopAndWait: ${errorMessage}`);
                            throw new AppsError("stopAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach STOPPED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Apps.stopAndWait: retrying: ${errorMessage}`);
                            throw new AppsRetriableError("stopAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.app;
        const query = {};
        const path = `/api/2.0/apps/${request.name}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the app with the supplied name.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/apps/${request.app_name}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on an app. Apps can inherit permissions from their
     * root object.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
}
exports.AppsService = AppsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_createUpdate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "createUpdate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_deploy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "deploy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_getDeployment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "getDeployment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_getUpdate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "getUpdate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], AppsService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_listDeployments", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], AppsService.prototype, "listDeployments", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_start", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "start", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_stop", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "stop", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsService.prototype, "updatePermissions", null);
class AppsSettingsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("AppsSettings", method, message);
    }
}
class AppsSettingsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("AppsSettings", method, message);
    }
}
/**
 * Apps Settings manage the settings for the Apps service on a customer's
 * Databricks instance.
 */
class AppsSettingsService {
    constructor(client) {
        this.client = client;
    }
    async _createCustomTemplate(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.template;
        const query = {};
        const path = "/api/2.0/apps-settings/templates";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a custom template.
     */
    async createCustomTemplate(request, context) {
        return await this._createCustomTemplate(request, context);
    }
    async _deleteCustomTemplate(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/apps-settings/templates/${request.name}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes the custom template with the specified name.
     */
    async deleteCustomTemplate(request, context) {
        return await this._deleteCustomTemplate(request, context);
    }
    async _getCustomTemplate(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/apps-settings/templates/${request.name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the custom template with the specified name.
     */
    async getCustomTemplate(request, context) {
        return await this._getCustomTemplate(request, context);
    }
    async _listCustomTemplates(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/apps-settings/templates";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists all custom templates in the workspace.
     */
    async *listCustomTemplates(request, context) {
        while (true) {
            const response = await this._listCustomTemplates(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.templates || response.templates.length === 0) {
                break;
            }
            for (const v of response.templates) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _updateCustomTemplate(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.template;
        const query = {};
        const path = `/api/2.0/apps-settings/templates/${request.name}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the custom template with the specified name. Note that the
     * template name cannot be updated.
     */
    async updateCustomTemplate(request, context) {
        return await this._updateCustomTemplate(request, context);
    }
}
exports.AppsSettingsService = AppsSettingsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsSettingsService.prototype, "_createCustomTemplate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsSettingsService.prototype, "createCustomTemplate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsSettingsService.prototype, "_deleteCustomTemplate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsSettingsService.prototype, "deleteCustomTemplate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsSettingsService.prototype, "_getCustomTemplate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsSettingsService.prototype, "getCustomTemplate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsSettingsService.prototype, "_listCustomTemplates", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], AppsSettingsService.prototype, "listCustomTemplates", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsSettingsService.prototype, "_updateCustomTemplate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AppsSettingsService.prototype, "updateCustomTemplate", null);
