// @databricks/appkit/dist/plugins/files/defaults.js
//#region src/plugins/files/defaults.ts
/**
* Execution defaults for read-tier operations (list, read, exists, metadata, preview).
* Cache 60s (ttl in seconds)
* Retry 3x with 1s backoff
* Timeout 30s
**/
const FILES_READ_DEFAULTS = {
	cache: {
		enabled: true,
		ttl: 60
	},
	retry: {
		enabled: true,
		initialDelay: 1e3,
		attempts: 3
	},
	timeout: 3e4
};
/**
* Execution defaults for download-tier operations (download, raw).
* No cache
* Retry 3x with 1s backoff
* Timeout 30s (stream start only)
**/
const FILES_DOWNLOAD_DEFAULTS = {
	cache: { enabled: false },
	retry: {
		enabled: true,
		initialDelay: 1e3,
		attempts: 3
	},
	timeout: 3e4
};
/**
* Execution defaults for write-tier operations (upload, mkdir, delete).
* No cache
* No retry
* Timeout 600s.
**/
const FILES_WRITE_DEFAULTS = {
	cache: { enabled: false },
	retry: { enabled: false },
	timeout: 6e5
};
/**
* Default maximum upload size in bytes (5 GB).
* This matches the Databricks Files API v2 per-file limit.
*/
const FILES_MAX_UPLOAD_SIZE = 5 * 1024 * 1024 * 1024;

//#endregion
export { FILES_DOWNLOAD_DEFAULTS, FILES_MAX_UPLOAD_SIZE, FILES_READ_DEFAULTS, FILES_WRITE_DEFAULTS };
