// @databricks/appkit/dist/plugins/files/manifest.js
//#region src/plugins/files/manifest.json
var manifest_default = {
	$schema: "https://databricks.github.io/appkit/schemas/plugin-manifest.schema.json",
	name: "files",
	displayName: "Files Plugin",
	description: "File operations against Databricks Volumes and Unity Catalog",
	resources: {
		"required": [{
			"type": "volume",
			"alias": "Files",
			"resourceKey": "files",
			"description": "Permission to write to volumes",
			"permission": "WRITE_VOLUME",
			"fields": { "path": {
				"env": "DATABRICKS_VOLUME_FILES",
				"description": "Volume path for file storage (e.g. /Volumes/catalog/schema/volume_name)",
				"discovery": {
					"type": "kind",
					"resourceKind": "volume",
					"select": "full_name"
				}
			} }
		}],
		"optional": []
	},
	scaffolding: { "rules": { "must": ["Before init, verify your Unity Catalog volume exists and you have WRITE_VOLUME permission"] } },
	config: { "schema": {
		"type": "object",
		"properties": {
			"timeout": {
				"type": "number",
				"default": 3e4,
				"description": "File operation timeout in milliseconds"
			},
			"maxUploadSize": {
				"type": "number",
				"default": 5368709120,
				"description": "Maximum upload size in bytes. Defaults to 5 GB (Databricks Files API v2 limit)."
			}
		}
	} }
};

//#endregion
export { manifest_default as default };
