// @databricks/appkit/dist/plugins/files/policy.js
//#region src/plugins/files/policy.ts
/** Actions that only read data. */
const READ_ACTIONS = new Set([
	"list",
	"read",
	"download",
	"raw",
	"exists",
	"metadata",
	"preview"
]);
/** Actions that mutate data. */
const WRITE_ACTIONS = new Set([
	"upload",
	"mkdir",
	"delete"
]);
/**
* Thrown when a policy denies an action.
*/
var PolicyDeniedError = class extends Error {
	action;
	volumeKey;
	constructor(action, volumeKey) {
		super(`Policy denied "${action}" on volume "${volumeKey}"`);
		this.name = "PolicyDeniedError";
		this.action = action;
		this.volumeKey = volumeKey;
	}
};
/** Utility namespace with common policy combinators. */
const policy = {
	all(...policies) {
		if (policies.length === 0) throw new Error("policy.all() requires at least one policy");
		return async (action, resource, user) => {
			for (const p of policies) if (!await p(action, resource, user)) return false;
			return true;
		};
	},
	any(...policies) {
		if (policies.length === 0) throw new Error("policy.any() requires at least one policy");
		return async (action, resource, user) => {
			for (const p of policies) if (await p(action, resource, user)) return true;
			return false;
		};
	},
	not(p) {
		return async (action, resource, user) => !await p(action, resource, user);
	},
	publicRead() {
		return (action) => READ_ACTIONS.has(action);
	},
	denyAll() {
		return () => false;
	},
	allowAll() {
		return () => true;
	}
};

//#endregion
export { PolicyDeniedError, READ_ACTIONS, WRITE_ACTIONS, policy };
