// @databricks/appkit/dist/connectors/mcp/client.js
import { createLogger } from "../../logging/logger.js";
import { assertResolvedHostSafe, checkMcpUrl } from "./host-policy.js";

//#region src/connectors/mcp/client.ts
const logger = createLogger("connector:mcp");
/**
* Hard cap on the size of a single MCP response body, including SSE
* frames bundled into one HTTP response. MCP `initialize` / `tools/list`
* / `tools/call` responses are JSON-RPC payloads — single-digit kilobytes
* in normal use. A response anywhere near this size signals either a
* misbehaving server or an attempt to exhaust client memory; we'd rather
* fail loudly than allocate unbounded buffers from a remote.
*/
const MCP_RESPONSE_BODY_LIMIT_BYTES = 1024 * 1024;
/**
* Read a fetch Response body into a string with a hard size cap. Aborts
* and throws if the cumulative bytes read cross {@link
* MCP_RESPONSE_BODY_LIMIT_BYTES}, so a remote server cannot keep
* streaming data past the limit. Returns the empty string when the
* response has no readable body.
*/
/**
* Empty-object fallback used when an MCP server ships a missing or
* malformed `inputSchema`. Matches the shape downstream adapters expect
* for a function tool that takes no arguments.
*/
const EMPTY_TOOL_PARAMETERS = {
	type: "object",
	properties: {}
};
/**
* Coerce a remote MCP server's reported `inputSchema` into the
* JSONSchema7 shape AppKit's adapters expect for a function tool's
* `parameters`. The MCP wire type is `Record<string, unknown>`, so a
* misbehaving (or malicious) server could ship arbitrary JSON. We accept
* only the standard `{ type: "object", properties: {...} }` shape and
* fall back to an empty-parameters schema otherwise — the tool still
* registers, it just can't accept arguments.
*/
function coerceToolParameters(inputSchema) {
	if (!inputSchema || typeof inputSchema !== "object") return EMPTY_TOOL_PARAMETERS;
	const { type, properties } = inputSchema;
	if (type !== "object") return EMPTY_TOOL_PARAMETERS;
	if (properties !== void 0 && (typeof properties !== "object" || properties === null || Array.isArray(properties))) return EMPTY_TOOL_PARAMETERS;
	return inputSchema;
}
async function readResponseTextCapped(response, maxBytes, contextLabel) {
	if (!response.body) return "";
	const reader = response.body.getReader();
	const decoder = new TextDecoder("utf-8");
	let total = 0;
	let out = "";
	try {
		while (true) {
			const { done, value } = await reader.read();
			if (done) break;
			total += value.byteLength;
			if (total > maxBytes) throw new Error(`MCP ${contextLabel}: response body exceeded ${maxBytes} bytes — refusing to allocate unbounded buffer from a remote server.`);
			out += decoder.decode(value, { stream: true });
		}
		out += decoder.decode();
	} finally {
		reader.releaseLock();
	}
	return out;
}
/**
* Lightweight MCP client for Databricks-hosted MCP servers.
*
* Uses raw fetch() with JSON-RPC 2.0 over HTTP — no @modelcontextprotocol/sdk
* or LangChain dependency. Supports the Streamable HTTP transport only
* (POST with JSON-RPC request, single JSON-RPC response). Implements exactly
* four methods: `initialize`, `notifications/initialized`, `tools/list`,
* `tools/call`. No prompts/resources/completion/sampling.
*
* All outbound URLs are gated by an {@link McpHostPolicy}: unallowlisted hosts
* are rejected before the first byte is sent, and workspace credentials are
* only forwarded to the same-origin workspace. See `mcp-host-policy.ts`.
*
* Rationale for hand-rolling JSON-RPC instead of `@modelcontextprotocol/sdk`:
* see the file-level comment at the top of this module.
*/
var AppKitMcpClient = class {
	connections = /* @__PURE__ */ new Map();
	sessionIds = /* @__PURE__ */ new Map();
	requestId = 0;
	closed = false;
	constructor(workspaceHost, authenticate, policy, options = {}) {
		this.workspaceHost = workspaceHost;
		this.authenticate = authenticate;
		this.policy = policy;
		this.options = options;
	}
	/**
	* Connects every endpoint in parallel and returns a structured summary so
	* callers can distinguish "all connected" from "some failed".
	*
	* Returning the result instead of throwing is deliberate: one
	* misconfigured MCP server should not take down the entire agents plugin
	* at boot, and the agents plugin uses the summary to warn at startup with
	* the failed-endpoint names. Errors are also logged here so a caller
	* that ignores the return still gets per-endpoint diagnostics.
	*
	* @returns `connected` lists the endpoint names that initialised
	*   successfully; `failed` carries `{ name, error }` for the rest.
	*/
	async connectAll(endpoints) {
		const results = await Promise.allSettled(endpoints.map((ep) => this.connect(ep)));
		const out = {
			connected: [],
			failed: []
		};
		for (let i = 0; i < results.length; i++) {
			const r = results[i];
			const name = endpoints[i].name;
			if (r.status === "fulfilled") out.connected.push(name);
			else {
				const error = r.reason instanceof Error ? r.reason : new Error(String(r.reason));
				logger.error("Failed to connect MCP server %s: %O", name, error);
				out.failed.push({
					name,
					error
				});
			}
		}
		return out;
	}
	resolveUrl(endpoint) {
		if (endpoint.url.startsWith("http://") || endpoint.url.startsWith("https://")) return endpoint.url;
		return `${this.workspaceHost}${endpoint.url}`;
	}
	async connect(endpoint) {
		const resolvedUrl = this.resolveUrl(endpoint);
		const check = checkMcpUrl(resolvedUrl, this.policy);
		if (!check.ok) throw new Error(`MCP endpoint '${endpoint.name}' refused at connect: ${check.reason}`);
		await assertResolvedHostSafe(check.url.hostname, this.policy, this.options.dnsLookup);
		logger.info("Connecting to MCP server: %s at %s (forwardWorkspaceAuth=%s)", endpoint.name, resolvedUrl, check.forwardWorkspaceAuth);
		const initResponse = await this.sendRpc(resolvedUrl, "initialize", {
			protocolVersion: "2025-03-26",
			capabilities: {},
			clientInfo: {
				name: "appkit-agent",
				version: "0.1.0"
			}
		}, { forwardWorkspaceAuth: check.forwardWorkspaceAuth });
		if (initResponse.sessionId) this.sessionIds.set(endpoint.name, initResponse.sessionId);
		const sessionId = this.sessionIds.get(endpoint.name);
		await this.sendNotification(resolvedUrl, "notifications/initialized", {
			sessionId,
			forwardWorkspaceAuth: check.forwardWorkspaceAuth
		});
		const toolList = (await this.sendRpc(resolvedUrl, "tools/list", {}, {
			sessionId,
			forwardWorkspaceAuth: check.forwardWorkspaceAuth
		})).result?.tools ?? [];
		const tools = /* @__PURE__ */ new Map();
		for (const tool of toolList) tools.set(tool.name, tool);
		this.connections.set(endpoint.name, {
			config: endpoint,
			resolvedUrl,
			forwardWorkspaceAuth: check.forwardWorkspaceAuth,
			tools
		});
		logger.info("Connected to MCP server %s: %d tools available", endpoint.name, tools.size);
	}
	getAllToolDefinitions() {
		const defs = [];
		for (const [serverName, conn] of this.connections) for (const [toolName, schema] of conn.tools) defs.push({
			name: `mcp.${serverName}.${toolName}`,
			description: schema.description ?? toolName,
			parameters: coerceToolParameters(schema.inputSchema)
		});
		return defs;
	}
	/**
	* Whether the named MCP server may receive workspace-scoped auth headers
	* (e.g., an OBO bearer token from an end-user request). Callers should gate
	* auth-forwarding decisions on this to prevent credential exfiltration to
	* non-workspace hosts.
	*/
	canForwardWorkspaceAuth(serverName) {
		return this.connections.get(serverName)?.forwardWorkspaceAuth ?? false;
	}
	async callTool(qualifiedName, args, authHeaders, callerSignal) {
		const parts = qualifiedName.split(".");
		if (parts.length < 3 || parts[0] !== "mcp") throw new Error(`Invalid MCP tool name: ${qualifiedName}`);
		const serverName = parts[1];
		const toolName = parts.slice(2).join(".");
		const conn = this.connections.get(serverName);
		if (!conn) throw new Error(`MCP server not connected: ${serverName}`);
		const sessionId = this.sessionIds.get(serverName);
		const scopedAuthOverride = conn.forwardWorkspaceAuth ? authHeaders : void 0;
		const result = (await this.sendRpc(conn.resolvedUrl, "tools/call", {
			name: toolName,
			arguments: args
		}, {
			authOverride: scopedAuthOverride,
			sessionId,
			forwardWorkspaceAuth: conn.forwardWorkspaceAuth,
			callerSignal
		})).result;
		const textContent = (result.content ?? []).filter((c) => c.type === "text" && typeof c.text === "string");
		if (result.isError) {
			const errText = textContent.map((c) => c.text).join("\n");
			throw new Error(errText || "MCP tool call failed");
		}
		return textContent.map((c) => c.text).join("\n");
	}
	async close() {
		this.closed = true;
		this.connections.clear();
		this.sessionIds.clear();
	}
	async sendRpc(url, method, params, options) {
		if (this.closed) throw new Error("MCP client is closed");
		const request = {
			jsonrpc: "2.0",
			id: ++this.requestId,
			method,
			...params && { params }
		};
		const authHeaders = await this.resolveAuthHeaders(options);
		const headers = {
			"Content-Type": "application/json",
			Accept: "application/json, text/event-stream",
			...authHeaders
		};
		if (options?.sessionId) headers["Mcp-Session-Id"] = options.sessionId;
		const fetchImpl = this.options.fetchImpl ?? fetch;
		const signals = [AbortSignal.timeout(3e4)];
		if (options?.callerSignal) signals.push(options.callerSignal);
		const response = await fetchImpl(url, {
			method: "POST",
			headers,
			body: JSON.stringify(request),
			signal: signals.length > 1 ? AbortSignal.any(signals) : signals[0]
		});
		if (!response.ok) throw new Error(`MCP request to ${method} failed: ${response.status} ${response.statusText}`);
		const contentType = response.headers.get("content-type") ?? "";
		const bodyText = await readResponseTextCapped(response, MCP_RESPONSE_BODY_LIMIT_BYTES, method);
		let json;
		if (contentType.includes("text/event-stream")) {
			const lastData = bodyText.split("\n").filter((line) => line.startsWith("data: ")).map((line) => line.slice(6)).pop();
			if (!lastData) throw new Error(`MCP SSE response for ${method} contained no data`);
			json = JSON.parse(lastData);
		} else {
			if (bodyText.length === 0) throw new Error(`MCP response for ${method} had an empty body`);
			json = JSON.parse(bodyText);
		}
		if (json.error) throw new Error(`MCP error (${json.error.code}): ${json.error.message}`);
		const sid = response.headers.get("mcp-session-id") ?? void 0;
		return {
			result: json.result,
			sessionId: sid
		};
	}
	async sendNotification(url, method, options) {
		if (this.closed) return;
		const authHeaders = await this.resolveAuthHeaders(options);
		const headers = {
			"Content-Type": "application/json",
			Accept: "application/json, text/event-stream",
			...authHeaders
		};
		if (options?.sessionId) headers["Mcp-Session-Id"] = options.sessionId;
		const fetchImpl = this.options.fetchImpl ?? fetch;
		try {
			const response = await fetchImpl(url, {
				method: "POST",
				headers,
				body: JSON.stringify({
					jsonrpc: "2.0",
					method
				}),
				signal: AbortSignal.timeout(3e4)
			});
			if (!response.ok) logger.warn("MCP notification %s to %s returned %d %s — the server may have rejected the request, but per MCP spec notifications are fire-and-forget and the connection is considered established.", method, url, response.status, response.statusText);
		} catch (err) {
			logger.warn("MCP notification %s to %s failed before a response was received: %O", method, url, err);
		}
	}
	/**
	* Return the auth headers to send on an outbound request. Workspace auth
	* (SP or OBO) is only resolved when `forwardWorkspaceAuth` is true; for
	* non-workspace hosts no bearer token is attached.
	*/
	async resolveAuthHeaders(options) {
		if (!options?.forwardWorkspaceAuth) return {};
		if (options.authOverride) return options.authOverride;
		return this.authenticate();
	}
};

//#endregion
export { AppKitMcpClient };
