// @databricks/appkit/dist/logging/logger.js
import { DEFAULT_SAMPLING_CONFIG, shouldSample } from "./sampling.js";
import { WideEvent } from "./wide-event.js";
import { WideEventEmitter } from "./wide-event-emitter.js";
import { AsyncLocalStorage } from "node:async_hooks";
import { format } from "node:util";
import { trace } from "@opentelemetry/api";
import { createDebug } from "obug";

//#region src/logging/logger.ts
const eventStorage = new AsyncLocalStorage();
const eventsByRequest = /* @__PURE__ */ new WeakMap();
const emitter = new WideEventEmitter();
const MAX_REQUEST_ID_LENGTH = 128;
/**
* Sanitize a request ID from user headers
*/
function sanitizeRequestId(id) {
	return id.replace(/[^a-zA-Z0-9_.-]/g, "").slice(0, MAX_REQUEST_ID_LENGTH);
}
/**
* Generate a request ID from the request
*/
function generateRequestId(req) {
	const existingId = req.headers["x-request-id"] || req.headers["x-correlation-id"] || req.headers["x-amzn-trace-id"];
	if (existingId && typeof existingId === "string" && existingId.length > 0) {
		const sanitized = sanitizeRequestId(existingId);
		if (sanitized.length > 0) return sanitized;
	}
	return `req_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
}
/**
* Create a WideEvent for a request
*/
function createEventForRequest(req) {
	const requestId = generateRequestId(req);
	const wideEvent = new WideEvent(requestId);
	const path = (req.path || req.url || req.originalUrl)?.split("?")[0];
	wideEvent.set("method", req.method).set("path", path);
	const rawUserId = req.headers["x-forwarded-user"];
	if (rawUserId && typeof rawUserId === "string" && rawUserId.length > 0) {
		const userId = rawUserId.replace(/[^a-zA-Z0-9_@.-]/g, "").slice(0, 128);
		if (userId.length > 0) wideEvent.setUser({ id: userId });
	}
	const spanContext = trace.getActiveSpan()?.spanContext();
	if (spanContext?.traceId) {
		wideEvent.set("trace_id", spanContext.traceId);
		createDebug("appkit:logger:event", { useColors: true })("WideEvent created: %s %s (reqId: %s, traceId: %s)", req.method, path, requestId.substring(0, 8), spanContext.traceId.substring(0, 8));
	}
	if (wideEvent.data.service) wideEvent.data.service = {
		...wideEvent.data.service,
		name: "appkit"
	};
	return wideEvent;
}
/**
* Setup response lifecycle handlers for WideEvent finalization
*/
function setupResponseHandlers(req, wideEvent) {
	const res = req.res;
	if (!res) return;
	res.once("finish", () => {
		const finalizedData = wideEvent.finalize(res.statusCode || 200);
		if (shouldSample(finalizedData, DEFAULT_SAMPLING_CONFIG)) emitter.emit(finalizedData);
		eventsByRequest.delete(req);
	});
	res.once("close", () => {
		if (!res.writableFinished) eventsByRequest.delete(req);
	});
}
/**
* Get or create a WideEvent for the given request.
* If called within wideEventMiddleware context, returns the event from AsyncLocalStorage.
* Otherwise creates a new event for the request.
*/
function getOrCreateEvent(req) {
	let wideEvent = eventsByRequest.get(req);
	if (!wideEvent) {
		const alsEvent = eventStorage.getStore();
		if (alsEvent) {
			eventsByRequest.set(req, alsEvent);
			return alsEvent;
		}
		wideEvent = createEventForRequest(req);
		eventsByRequest.set(req, wideEvent);
		setupResponseHandlers(req, wideEvent);
	}
	return wideEvent;
}
/**
* Get current WideEvent from AsyncLocalStorage or request
*/
function getCurrentEvent(req) {
	if (req) return getOrCreateEvent(req);
	return eventStorage.getStore();
}
/**
* Check if the first argument is an Express Request
*/
function isRequest(arg) {
	return typeof arg === "object" && arg !== null && "method" in arg && "path" in arg && typeof arg.method === "string";
}
/**
* Create a logger instance for a specific scope
* @param scope - The scope identifier (e.g., "connectors:lakebase")
* @returns Logger instance with debug, info, warn, and error methods
*
* @example
* ```typescript
* const logger = createLogger("connectors:lakebase");
*
* // Regular logging (no request tracking)
* logger.debug("Connection established with pool size: %d", poolSize);
* logger.info("Server started on port %d", port);
*
* // Request-scoped logging (tracks in WideEvent)
* logger.debug(req, "Processing query: %s", queryId);
* logger.error(req, "Query failed: %O", error);
*
* // Get WideEvent - works in route handlers (with req) or interceptors (from context)
* const event = logger.event(req);  // In route handler
* const event = logger.event();     // In interceptor (gets from AsyncLocalStorage)
* event?.setComponent("analytics", "executeQuery");
* ```
*/
function createLogger(scope) {
	const debug = createDebug(`appkit:${scope}`, { useColors: true });
	const prefix = `[appkit:${scope}]`;
	function debugLog(reqOrMessage, ...args) {
		if (isRequest(reqOrMessage)) {
			const req = reqOrMessage;
			const message = args[0];
			const logArgs = args.slice(1);
			const formatted = format(message, ...logArgs);
			debug(message, ...logArgs);
			getOrCreateEvent(req).addLog("debug", formatted);
		} else debug(reqOrMessage, ...args);
	}
	function infoLog(reqOrMessage, ...args) {
		if (isRequest(reqOrMessage)) {
			const req = reqOrMessage;
			const message = args[0];
			const formatted = format(message, ...args.slice(1));
			console.log(prefix, formatted);
			getOrCreateEvent(req).addLog("info", formatted);
		} else console.log(prefix, format(reqOrMessage, ...args));
	}
	function warnLog(reqOrMessage, ...args) {
		if (isRequest(reqOrMessage)) {
			const req = reqOrMessage;
			const message = args[0];
			const formatted = format(message, ...args.slice(1));
			console.warn(prefix, formatted);
			getOrCreateEvent(req).addLog("warn", formatted);
		} else console.warn(prefix, format(reqOrMessage, ...args));
	}
	function errorLog(reqOrMessage, ...args) {
		if (isRequest(reqOrMessage)) {
			const req = reqOrMessage;
			const message = args[0];
			const formatted = format(message, ...args.slice(1));
			console.error(prefix, formatted);
			getOrCreateEvent(req).addLog("error", formatted);
		} else console.error(prefix, format(reqOrMessage, ...args));
	}
	function event(req) {
		return getCurrentEvent(req);
	}
	return {
		debug: debugLog,
		info: infoLog,
		warn: warnLog,
		error: errorLog,
		event
	};
}

//#endregion
export { createLogger };
