// @databricks/appkit/dist/logging/sampling.js
import { shouldExcludePath } from "../utils/path-exclusions.js";

//#region src/logging/sampling.ts
/**
* Get sample rate from environment variable or default to 1.0 (100%)
*/
function getSampleRate() {
	const envRate = process.env.APPKIT_SAMPLE_RATE;
	if (envRate) {
		const parsed = parseFloat(envRate);
		if (!Number.isNaN(parsed) && parsed >= 0 && parsed <= 1) return parsed;
	}
	return 1;
}
/**
* Default sampling configuration
*/
const DEFAULT_SAMPLING_CONFIG = {
	alwaysSampleIf: {
		hasErrors: true,
		statusCodeGte: 400,
		durationGte: 5e3,
		hasCacheInfo: true
	},
	sampleRate: getSampleRate()
};
/**
* Simple hash function for deterministic sampling
*/
function hashString(str) {
	let hash = 0;
	for (let i = 0; i < str.length; i++) {
		const char = str.charCodeAt(i);
		hash = (hash << 5) - hash + char;
		hash = hash & hash;
	}
	return Math.abs(hash);
}
/**
* Determine if a WideEvent should be sampled based on configuration.
* Uses shared path exclusions from utils/path-exclusions.ts.
*/
function shouldSample(event, config = DEFAULT_SAMPLING_CONFIG) {
	if (shouldExcludePath(event.path)) return false;
	if (config.alwaysSampleIf.hasErrors && event.error) return true;
	if (config.alwaysSampleIf.statusCodeGte && event.status_code && event.status_code >= config.alwaysSampleIf.statusCodeGte) return true;
	if (config.alwaysSampleIf.durationGte && event.duration_ms && event.duration_ms >= config.alwaysSampleIf.durationGte) return true;
	if (config.alwaysSampleIf.hasCacheInfo && event.execution?.cache_hit !== void 0) return true;
	if (config.sampleRate >= 1) return true;
	if (config.sampleRate <= 0) return false;
	return hashString(event.request_id) % 100 < config.sampleRate * 100;
}

//#endregion
export { DEFAULT_SAMPLING_CONFIG, shouldSample };
