// @databricks/appkit/dist/logging/wide-event.js
//#region src/logging/wide-event.ts
/**
* WideEvent
* - Represents a single event for a request
* - Fields are camelCase to match OpenTelemetry
*/
var WideEvent = class {
	data;
	startTime;
	constructor(requestId) {
		this.startTime = Date.now();
		this.data = {
			timestamp: (/* @__PURE__ */ new Date()).toISOString(),
			request_id: requestId,
			service: {
				name: "appkit",
				version: process.env.npm_package_version || "unknown",
				region: process.env.REGION,
				deployment_id: process.env.DEPLOYMENT_ID,
				node_env: process.env.NODE_ENV
			},
			logs: [],
			context: {}
		};
	}
	/**
	* Set a value in the event
	* @param key - The key to set
	* @param value - The value to set
	* @returns The event
	*/
	set(key, value) {
		if (typeof value === "object" && value !== null && !Array.isArray(value)) this.data[key] = {
			...this.data[key],
			...value
		};
		else this.data[key] = value;
		return this;
	}
	/**
	* Set the component name and operation.
	* Component can be a plugin, connector, or service.
	* @param name - The name of the component (e.g., "analytics", "sql-warehouse", "cache-manager")
	* @param operation - The operation being performed (e.g., "query", "getOrExecute")
	* @returns The event
	*/
	setComponent(name, operation) {
		this.data.component = {
			name,
			operation
		};
		return this;
	}
	/**
	* Set the user context
	* @param user - The user context
	* @returns The event
	*/
	setUser(user) {
		this.data.user = {
			...this.data.user,
			...user
		};
		return this;
	}
	/**
	* Set the execution context
	* @param execution - The execution context
	* @returns The event
	*/
	setExecution(execution) {
		this.data.execution = {
			...this.data.execution,
			...execution
		};
		return this;
	}
	/**
	* Set the stream context
	* @param stream - The stream context
	* @returns The event
	*/
	setStream(stream) {
		this.data.stream = {
			...this.data.stream,
			...stream
		};
		return this;
	}
	/**
	* Set the error context
	* @param error - The error context
	* @returns The event
	*/
	setError(error) {
		const isAppKitError = "code" in error && "statusCode" in error;
		const errorCause = error.cause;
		this.data.error = {
			type: error.name,
			code: isAppKitError ? error.code : "UNKNOWN_ERROR",
			message: error.message,
			retriable: isAppKitError ? error.isRetryable : false,
			cause: errorCause ? String(errorCause) : void 0
		};
		return this;
	}
	/**
	* Add scoped context to the event
	* @param scope - The scope name (plugin, connector, or service name)
	* @param ctx - Context data to merge
	* @example
	* event.setContext("analytics", { query_key: "apps_list", rows_returned: 100 });
	* event.setContext("sql-warehouse", { warehouse_id: "1234567890" });
	*/
	setContext(scope, ctx) {
		if (!this.data.context) this.data.context = {};
		this.data.context[scope] = {
			...this.data.context[scope],
			...ctx
		};
		return this;
	}
	/**
	* Add a log to the event
	* @param level - The level of the log
	* @param message - The message of the log
	* @param context - The context of the log
	* @returns The event
	*/
	addLog(level, message, context) {
		if (!this.data.logs) this.data.logs = [];
		this.data.logs.push({
			level,
			message,
			timestamp: (/* @__PURE__ */ new Date()).toISOString(),
			context
		});
		if (this.data.logs.length > 50) this.data.logs = this.data.logs.slice(-50);
		return this;
	}
	/**
	* Finalize the event
	* @param statusCode - The status code of the response
	* @returns The event data
	*/
	finalize(statusCode) {
		this.data.status_code = statusCode;
		this.data.duration_ms = this.getDurationMs();
		return this.data;
	}
	/**
	* Get the duration of the event in milliseconds
	* @returns The duration of the event in milliseconds
	*/
	getDurationMs() {
		return this.data.duration_ms || Date.now() - this.startTime;
	}
	/**
	* Convert the event to a JSON object
	* @returns The event data as a JSON object
	*/
	toJSON() {
		return this.data;
	}
};

//#endregion
export { WideEvent };
