// @databricks/appkit/dist/logging/wide-event-emitter.js
import { SeverityNumber, logs } from "@opentelemetry/api-logs";

//#region src/logging/wide-event-emitter.ts
/**
* Emits WideEvents to OpenTelemetry as structured logs
*/
var WideEventEmitter = class {
	logger = logs.getLogger("appkit", "1.0.0");
	/**
	* Emit a WideEvent to OpenTelemetry.
	* Fails silently to avoid crashing the application due to observability issues.
	*/
	emit(event) {
		try {
			const logRecord = {
				timestamp: Date.parse(event.timestamp),
				severityNumber: this.getSeverityNumber(event),
				severityText: this.getSeverityText(event),
				body: this.createLogBody(event),
				attributes: this.createAttributes(event)
			};
			this.logger.emit(logRecord);
		} catch {}
	}
	/**
	* Get OpenTelemetry severity number based on event data
	*/
	getSeverityNumber(event) {
		if (event.error) return SeverityNumber.ERROR;
		if (event.status_code) {
			if (event.status_code >= 500) return SeverityNumber.ERROR;
			if (event.status_code >= 400) return SeverityNumber.WARN;
		}
		if (event.logs) {
			if (event.logs.some((log) => log.level === "error")) return SeverityNumber.ERROR;
			if (event.logs.some((log) => log.level === "warn")) return SeverityNumber.WARN;
		}
		return SeverityNumber.INFO;
	}
	/**
	* Get severity text based on severity number
	*/
	getSeverityText(event) {
		const severityNumber = this.getSeverityNumber(event);
		if (severityNumber >= SeverityNumber.ERROR) return "ERROR";
		if (severityNumber >= SeverityNumber.WARN) return "WARN";
		if (severityNumber >= SeverityNumber.INFO) return "INFO";
		return "DEBUG";
	}
	/**
	* Create log body from event data
	*/
	createLogBody(event) {
		const parts = [];
		if (event.method && event.path) parts.push(`${event.method} ${event.path}`);
		if (event.status_code) parts.push(`→ ${event.status_code}`);
		if (event.duration_ms) parts.push(`(${event.duration_ms}ms)`);
		if (event.component) {
			const componentStr = event.component.operation ? `${event.component.name}.${event.component.operation}` : event.component.name;
			parts.push(`[${componentStr}]`);
		}
		if (event.error) parts.push(`ERROR: ${event.error.message}`);
		return parts.join(" ");
	}
	/**
	* Create OpenTelemetry attributes from event data
	*/
	createAttributes(event) {
		const attributes = {
			request_id: event.request_id,
			trace_id: event.trace_id,
			"http.method": event.method,
			"http.route": event.path,
			"http.status_code": event.status_code,
			"http.request.duration_ms": event.duration_ms,
			"service.name": event.service?.name,
			"service.version": event.service?.version,
			"service.region": event.service?.region,
			"service.deployment_id": event.service?.deployment_id,
			"service.node_env": event.service?.node_env,
			"component.name": event.component?.name,
			"component.operation": event.component?.operation,
			"user.id": event.user?.id,
			"error.type": event.error?.type,
			"error.code": event.error?.code,
			"error.message": event.error?.message,
			"error.retriable": event.error?.retriable,
			"execution.timeout_ms": event.execution?.timeout_ms,
			"execution.retry_attempts": event.execution?.retry_attempts,
			"execution.cache_hit": event.execution?.cache_hit,
			"execution.cache_key": event.execution?.cache_key,
			"execution.cache_deduplication": event.execution?.cache_deduplication,
			"stream.id": event.stream?.stream_id,
			"stream.events_sent": event.stream?.events_sent,
			log_count: event.logs?.length
		};
		if (event.context) {
			for (const [scope, scopeData] of Object.entries(event.context)) for (const [key, value] of Object.entries(scopeData)) if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") attributes[`${scope}.${key}`] = value;
		}
		return Object.fromEntries(Object.entries(attributes).filter(([_, value]) => value !== void 0));
	}
};

//#endregion
export { WideEventEmitter };
