// @databricks/appkit/dist/cache/storage/memory.js
import { inMemoryStorageDefaults } from "./defaults.js";

//#region src/cache/storage/memory.ts
/**
* In-memory cache storage implementation. Uses a least recently used (LRU) eviction policy
* to manage memory usage and ensure efficient cache operations.
*/
var InMemoryStorage = class {
	cache = /* @__PURE__ */ new Map();
	accessOrder = /* @__PURE__ */ new Map();
	accessCounter;
	maxSize;
	constructor(config) {
		this.cache = /* @__PURE__ */ new Map();
		this.accessOrder = /* @__PURE__ */ new Map();
		this.maxSize = config.maxSize ?? inMemoryStorageDefaults.maxSize;
		this.accessCounter = 0;
	}
	/** Get an entry from the cache */
	async get(key) {
		const entry = this.cache.get(key);
		if (!entry) return null;
		this.accessOrder.set(key, ++this.accessCounter);
		return entry;
	}
	/** Set an entry in the cache */
	async set(key, entry) {
		if (this.cache.size >= this.maxSize && !this.cache.has(key)) this.evictLRU();
		this.cache.set(key, entry);
		this.accessOrder.set(key, ++this.accessCounter);
	}
	/** Delete an entry from the cache */
	async delete(key) {
		this.cache.delete(key);
		this.accessOrder.delete(key);
	}
	/** Clean in-memory cache */
	async clear() {
		this.cache.clear();
		this.accessOrder.clear();
		this.accessCounter = 0;
	}
	/** Check if the cache has an entry */
	async has(key) {
		const entry = this.cache.get(key);
		if (!entry) return false;
		if (Date.now() > entry.expiry) {
			this.cache.delete(key);
			this.accessOrder.delete(key);
			return false;
		}
		return true;
	}
	/** Get the size of the cache */
	async size() {
		return this.cache.size;
	}
	/** Check if the cache is persistent */
	isPersistent() {
		return false;
	}
	/** Check the health of the cache */
	async healthCheck() {
		return true;
	}
	/** Close the cache */
	async close() {
		this.cache.clear();
		this.accessOrder.clear();
		this.accessCounter = 0;
	}
	/** Evict the least recently used entry (LRU) */
	evictLRU() {
		let oldestKey = null;
		let oldestAccess = Infinity;
		for (const [key, accessTime] of this.accessOrder) if (accessTime < oldestAccess) {
			oldestAccess = accessTime;
			oldestKey = key;
		}
		if (oldestKey) {
			this.cache.delete(oldestKey);
			this.accessOrder.delete(oldestKey);
		}
	}
};

//#endregion
export { InMemoryStorage };
