// @databricks/appkit/dist/cache/storage/persistent.js
import { createLogger } from "../../logging/logger.js";
import { InitializationError } from "../../errors/initialization.js";
import { ValidationError } from "../../errors/validation.js";
import "../../errors/index.js";
import { lakebaseStorageDefaults } from "./defaults.js";
import { createHash } from "node:crypto";

//#region src/cache/storage/persistent.ts
const logger = createLogger("cache:persistent");
/**
* Persistent cache storage implementation. Uses a least recently used (LRU) eviction policy
* to manage memory usage and ensure efficient cache operations.
*
* @example
* const pool = createLakebasePool({ workspaceClient });
* const persistentStorage = new PersistentStorage(config, pool);
* await persistentStorage.initialize();
* await persistentStorage.get("my-key");
* await persistentStorage.set("my-key", "my-value");
* await persistentStorage.delete("my-key");
* await persistentStorage.clear();
* await persistentStorage.has("my-key");
*
*/
var PersistentStorage = class {
	pool;
	schemaName;
	tableName;
	qualifiedTableName;
	maxBytes;
	maxEntryBytes;
	evictionBatchSize;
	evictionCheckProbability;
	initialized;
	constructor(config, pool) {
		this.pool = pool;
		this.maxBytes = config.maxBytes ?? lakebaseStorageDefaults.maxBytes;
		this.maxEntryBytes = config.maxEntryBytes ?? lakebaseStorageDefaults.maxEntryBytes;
		this.evictionBatchSize = lakebaseStorageDefaults.evictionBatchSize;
		this.evictionCheckProbability = config.evictionCheckProbability ?? lakebaseStorageDefaults.evictionCheckProbability;
		this.schemaName = lakebaseStorageDefaults.schemaName;
		this.tableName = lakebaseStorageDefaults.tableName;
		this.qualifiedTableName = `${this.schemaName}.${this.tableName}`;
		this.initialized = false;
	}
	/** Initialize the persistent storage and run migrations if necessary */
	async initialize() {
		if (this.initialized) return;
		try {
			await this.runMigrations();
			this.initialized = true;
		} catch (error) {
			logger.error("Error in persistent storage initialization: %O", error);
			throw error;
		}
	}
	/**
	* Get a cached value from the persistent storage
	* @param key - Cache key
	* @returns Promise of the cached value or null if not found
	*/
	async get(key) {
		await this.ensureInitialized();
		const keyHash = this.hashKey(key);
		const result = await this.pool.query(`SELECT value, expiry FROM ${this.qualifiedTableName} WHERE key_hash = $1`, [keyHash]);
		if (result.rows.length === 0) return null;
		const entry = result.rows[0];
		this.pool.query(`UPDATE ${this.qualifiedTableName} SET last_accessed = NOW() WHERE key_hash = $1`, [keyHash]).catch(() => {
			logger.debug("Error updating last_accessed time for key: %s", key);
		});
		return {
			value: this.deserializeValue(entry.value),
			expiry: Number(entry.expiry)
		};
	}
	/**
	* Set a value in the persistent storage
	* @param key - Cache key
	* @param entry - Cache entry
	* @returns Promise of the result
	*/
	async set(key, entry) {
		await this.ensureInitialized();
		const keyHash = this.hashKey(key);
		const keyBytes = Buffer.from(key, "utf-8");
		const valueBytes = this.serializeValue(entry.value);
		const byteSize = keyBytes.length + valueBytes.length;
		if (byteSize > this.maxEntryBytes) throw ValidationError.invalidValue("cache entry size", byteSize, `maximum ${this.maxEntryBytes} bytes`);
		if (Math.random() < this.evictionCheckProbability) {
			if (await this.totalBytes() + byteSize > this.maxBytes) await this.evictBySize(byteSize);
		}
		await this.pool.query(`INSERT INTO ${this.qualifiedTableName} (key_hash, key, value, byte_size, expiry, created_at, last_accessed)
      VALUES ($1, $2, $3, $4, $5, NOW(), NOW())
      ON CONFLICT (key_hash)
      DO UPDATE SET value = $3, byte_size = $4, expiry = $5, last_accessed = NOW()
      `, [
			keyHash,
			keyBytes,
			valueBytes,
			byteSize,
			entry.expiry
		]);
	}
	/**
	* Delete a value from the persistent storage
	* @param key - Cache key
	* @returns Promise of the result
	*/
	async delete(key) {
		await this.ensureInitialized();
		const keyHash = this.hashKey(key);
		await this.pool.query(`DELETE FROM ${this.qualifiedTableName} WHERE key_hash = $1`, [keyHash]);
	}
	/** Clear the persistent storage */
	async clear() {
		await this.ensureInitialized();
		await this.pool.query(`TRUNCATE TABLE ${this.qualifiedTableName}`);
	}
	/**
	* Check if a value exists in the persistent storage
	* @param key - Cache key
	* @returns Promise of true if the value exists, false otherwise
	*/
	async has(key) {
		await this.ensureInitialized();
		const keyHash = this.hashKey(key);
		return (await this.pool.query(`SELECT EXISTS(SELECT 1 FROM ${this.qualifiedTableName} WHERE key_hash = $1) as exists`, [keyHash])).rows[0]?.exists ?? false;
	}
	/**
	* Get the size of the persistent storage
	* @returns Promise of the size of the storage
	*/
	async size() {
		await this.ensureInitialized();
		const result = await this.pool.query(`SELECT COUNT(*) as count FROM ${this.qualifiedTableName}`);
		return parseInt(result.rows[0]?.count ?? "0", 10);
	}
	/** Get the total number of bytes in the persistent storage */
	async totalBytes() {
		await this.ensureInitialized();
		const result = await this.pool.query(`SELECT COALESCE(SUM(byte_size), 0) as total FROM ${this.qualifiedTableName}`);
		return parseInt(result.rows[0]?.total ?? "0", 10);
	}
	/**
	* Check if the persistent storage is persistent
	* @returns true if the storage is persistent, false otherwise
	*/
	isPersistent() {
		return true;
	}
	/**
	* Check if the persistent storage is healthy
	* @returns Promise of true if the storage is healthy, false otherwise
	*/
	async healthCheck() {
		try {
			await this.pool.query("SELECT 1");
			return true;
		} catch {
			return false;
		}
	}
	/** Close the persistent storage */
	async close() {
		await this.pool.end();
	}
	/**
	* Cleanup expired entries from the persistent storage
	* @returns Promise of the number of expired entries
	*/
	async cleanupExpired() {
		await this.ensureInitialized();
		const result = await this.pool.query(`WITH deleted as (DELETE FROM ${this.qualifiedTableName} WHERE expiry < $1 RETURNING *) SELECT COUNT(*) as count FROM deleted`, [Date.now()]);
		return parseInt(result.rows[0]?.count ?? "0", 10);
	}
	/** Evict entries from the persistent storage by size */
	async evictBySize(requiredBytes) {
		if (await this.cleanupExpired() > 0) {
			if (await this.totalBytes() + requiredBytes <= this.maxBytes) return;
		}
		await this.pool.query(`DELETE FROM ${this.qualifiedTableName} WHERE key_hash IN
      (SELECT key_hash FROM ${this.qualifiedTableName} ORDER BY last_accessed ASC LIMIT $1)`, [this.evictionBatchSize]);
	}
	/** Ensure the persistent storage is initialized */
	async ensureInitialized() {
		if (!this.initialized) await this.initialize();
	}
	/** Generate a 64-bit hash for the cache key using SHA256 */
	hashKey(key) {
		if (!key) throw ValidationError.missingField("key");
		return createHash("sha256").update(key).digest().readBigInt64BE(0);
	}
	/** Serialize a value to a buffer */
	serializeValue(value) {
		return Buffer.from(JSON.stringify(value), "utf-8");
	}
	/** Deserialize a value from a buffer */
	deserializeValue(buffer) {
		return JSON.parse(buffer.toString("utf-8"));
	}
	/** Run migrations for the persistent storage */
	async runMigrations() {
		const steps = [
			{
				name: "create schema",
				query: `CREATE SCHEMA IF NOT EXISTS ${this.schemaName}`
			},
			{
				name: "create table",
				query: `CREATE TABLE IF NOT EXISTS ${this.qualifiedTableName} (
                id BIGSERIAL PRIMARY KEY,
                key_hash BIGINT NOT NULL,
                key BYTEA NOT NULL,
                value BYTEA NOT NULL,
                byte_size INTEGER NOT NULL,
                expiry BIGINT NOT NULL,
                created_at TIMESTAMP NOT NULL DEFAULT NOW(),
                last_accessed TIMESTAMP NOT NULL DEFAULT NOW()
            )`
			},
			{
				name: "create index (key_hash)",
				query: `CREATE UNIQUE INDEX IF NOT EXISTS idx_${this.tableName}_key_hash ON ${this.qualifiedTableName} (key_hash)`
			},
			{
				name: "create index (expiry)",
				query: `CREATE INDEX IF NOT EXISTS idx_${this.tableName}_expiry ON ${this.qualifiedTableName} (expiry)`
			},
			{
				name: "create index (last_accessed)",
				query: `CREATE INDEX IF NOT EXISTS idx_${this.tableName}_last_accessed ON ${this.qualifiedTableName} (last_accessed)`
			},
			{
				name: "create index (byte_size)",
				query: `CREATE INDEX IF NOT EXISTS idx_${this.tableName}_byte_size ON ${this.qualifiedTableName} (byte_size)`
			}
		];
		for (const step of steps) try {
			await this.pool.query(step.query);
		} catch (error) {
			logger.error("Migration step '%s' failed: %O", step.name, error);
			throw InitializationError.migrationFailed(error);
		}
	}
};

//#endregion
export { PersistentStorage };
