// @databricks/appkit/dist/type-generator/cache.js
import { createLogger } from "../logging/logger.js";
import crypto from "node:crypto";
import fs from "node:fs/promises";
import path from "node:path";

//#region src/type-generator/cache.ts
const logger = createLogger("type-generator:cache");
const CACHE_VERSION = "3";
const CACHE_FILE = ".appkit-types-cache.json";
const CACHE_DIR = path.join(process.cwd(), "node_modules", ".databricks", "appkit");
/**
* Hash the SQL query
* Uses MD5 to hash the SQL query
* @param sql - the SQL query to hash
* @returns - the hash of the SQL query
*/
function hashSQL(sql) {
	return crypto.createHash("md5").update(sql).digest("hex");
}
/**
* Load the cache from the file system
* If the cache is not found, run the query explain
* @returns - the cache
*/
async function loadCache() {
	const cachePath = path.join(CACHE_DIR, CACHE_FILE);
	try {
		await fs.mkdir(CACHE_DIR, { recursive: true });
		const raw = await fs.readFile(cachePath, "utf8");
		const cache = JSON.parse(raw);
		if (cache.version === CACHE_VERSION) return cache;
	} catch (err) {
		if (err.code !== "ENOENT") logger.warn("Cache file is corrupted, flushing cache completely.");
	}
	return {
		version: CACHE_VERSION,
		queries: {}
	};
}
/**
* Save the cache to the file system
* @param cache - cache object to save
*/
async function saveCache(cache) {
	const cachePath = path.join(CACHE_DIR, CACHE_FILE);
	await fs.writeFile(cachePath, JSON.stringify(cache, null, 2), "utf8");
}

//#endregion
export { CACHE_VERSION, hashSQL, loadCache, saveCache };
