// @databricks/appkit/dist/type-generator/index.js
import { createLogger } from "../logging/logger.js";
import { migrateProjectConfig, removeOldGeneratedTypes, resolveProjectRoot } from "./migration.js";
import { generateQueriesFromDescribe } from "./query-registry.js";
import { generateServingTypes as generateServingTypes$1 } from "./serving/generator.js";
import fs from "node:fs/promises";
import path from "node:path";
import dotenv from "dotenv";

//#region src/type-generator/index.ts
dotenv.config();
const logger = createLogger("type-generator");
/**
* Generate type declarations for QueryRegistry
* Create the d.ts file from the plugin routes and query schemas
* @param querySchemas - the list of query schemas
* @returns - the type declarations as a string
*/
function generateTypeDeclarations(querySchemas = []) {
	const queryEntries = querySchemas.map(({ name, type }) => {
		return `    ${name}: ${type.split("\n").map((line, i) => i === 0 ? line : `    ${line}`).join("\n")}`;
	}).join(";\n");
	return `// Auto-generated by AppKit - DO NOT EDIT
// Generated by 'npx @databricks/appkit generate-types' or Vite plugin during build
import "@databricks/appkit-ui/react";
import type { SQLTypeMarker, SQLStringMarker, SQLNumberMarker, SQLBooleanMarker, SQLBinaryMarker, SQLDateMarker, SQLTimestampMarker } from "@databricks/appkit-ui/js";

declare module "@databricks/appkit-ui/react" {
  interface QueryRegistry {${queryEntries ? `\n${queryEntries};\n  ` : ""}}
}
`;
}
/**
* Entry point for generating type declarations from all imported files
* @param options - the options for the generation
* @param options.entryPoint - the entry point file
* @param options.outFile - the output file
* @param options.querySchemaFile - optional path to query schema file (e.g. config/queries/schema.ts)
*/
async function generateFromEntryPoint(options) {
	const { outFile, queryFolder, warehouseId, noCache } = options;
	const projectRoot = resolveProjectRoot(outFile);
	logger.debug("Starting type generation...");
	let queryRegistry = [];
	if (queryFolder) queryRegistry = await generateQueriesFromDescribe(queryFolder, warehouseId, { noCache });
	const failedQueries = queryRegistry.filter((q) => q.type.includes("result: unknown"));
	if (failedQueries.length > 0) {
		const names = failedQueries.map((q) => q.name).join(", ");
		throw new Error([
			`Type generation failed: ${failedQueries.length} ${failedQueries.length === 1 ? "query" : "queries"} could not be described: ${names}.`,
			`DESCRIBE QUERY failed for these queries — see the error codes above for details.`,
			`Common causes: SQL syntax errors, missing tables/views, or warehouse format incompatibilities.`,
			`To debug: run the failing query directly in a SQL editor against warehouse ${warehouseId}.`
		].join("\n"));
	}
	const typeDeclarations = generateTypeDeclarations(queryRegistry);
	await fs.mkdir(path.dirname(outFile), { recursive: true });
	await fs.writeFile(outFile, typeDeclarations, "utf-8");
	await removeOldGeneratedTypes(projectRoot, "appKitTypes.d.ts");
	await migrateProjectConfig(projectRoot);
	logger.debug("Type generation complete!");
}
const generateServingTypes = generateServingTypes$1;
/** Directory name for generated AppKit type declaration files. */
const TYPES_DIR = "appkit-types";
/** Default filename for analytics query type declarations. */
const ANALYTICS_TYPES_FILE = "analytics.d.ts";
/** Default filename for serving endpoint type declarations. */
const SERVING_TYPES_FILE = "serving.d.ts";

//#endregion
export { ANALYTICS_TYPES_FILE, SERVING_TYPES_FILE, TYPES_DIR, generateFromEntryPoint, generateServingTypes };
