// @databricks/appkit/dist/type-generator/migration.js
import { createLogger } from "../logging/logger.js";
import pc from "picocolors";
import fs from "node:fs/promises";
import path from "node:path";
import fs$1 from "node:fs";

//#region src/type-generator/migration.ts
const logger = createLogger("type-generator:migration");
/**
* Derive project root from an outFile path.
* outFile is always `<projectRoot>/shared/appkit-types/<file>` — both the Vite plugins
* and the CLI construct it this way, so going up two levels is safe.
*
* Validates that the resolved root contains a package.json — if not, logs a warning
* so custom outFile paths don't silently operate on the wrong directory.
*/
function resolveProjectRoot(outFile) {
	const root = path.resolve(path.dirname(outFile), "..", "..");
	if (!fs$1.existsSync(path.join(root, "package.json"))) logger.warn("Resolved project root %s has no package.json — migration may target the wrong directory. Check your outFile path: %s", root, outFile);
	return root;
}
/**
* Remove old generated types from client/src/ (pre-shared/ location).
* Best-effort: silently ignores missing files.
*/
async function removeOldGeneratedTypes(projectRoot, filename) {
	const oldFile = path.join(projectRoot, "client", "src", filename);
	try {
		await fs.unlink(oldFile);
		logger.debug("Removed old types at %s", oldFile);
	} catch {}
}
const migratedProjects = /* @__PURE__ */ new Set();
/**
* One-time config migration: update tsconfig and package.json for shared/ types output.
* Idempotent — each sub-migration checks current file state and skips if already migrated.
* Deduplicates per project root so monorepo builds migrate each app independently.
* Opt-out: set `"appkit": { "autoMigrate": false }` in package.json.
*/
async function migrateProjectConfig(projectRoot) {
	const resolved = path.resolve(projectRoot);
	if (migratedProjects.has(resolved)) return;
	migratedProjects.add(resolved);
	if (await isAutoMigrateDisabled(projectRoot)) {
		logger.debug("Auto-migration disabled via package.json appkit.autoMigrate");
		return;
	}
	const results = [];
	results.push(...await migrateTsconfigClient(projectRoot));
	results.push(...await migrateTsconfigServer(projectRoot));
	results.push(...await migratePackageJsonScripts(projectRoot));
	if (results.length > 0) printMigrationSummary(results);
}
async function isAutoMigrateDisabled(projectRoot) {
	try {
		const raw = await fs.readFile(path.join(projectRoot, "package.json"), "utf-8");
		return JSON.parse(raw).appkit?.autoMigrate === false;
	} catch {
		return false;
	}
}
/** Strip JSONC comments (block and line) so JSON.parse can handle tsconfig files. */
function stripJsonComments(text) {
	return text.replace(/"(?:[^"\\]|\\.)*"|\/\*[\s\S]*?\*\/|\/\/.*/g, (match) => match.startsWith("\"") ? match : "");
}
async function migrateTsconfigClient(projectRoot) {
	const results = [];
	const filePath = path.join(projectRoot, "tsconfig.client.json");
	try {
		const raw = await fs.readFile(filePath, "utf-8");
		const parsed = JSON.parse(stripJsonComments(raw));
		if (!Array.isArray(parsed.include)) return results;
		if (parsed.include.includes("shared/appkit-types")) return results;
		parsed.include.push("shared/appkit-types");
		await fs.writeFile(filePath, `${JSON.stringify(parsed, null, 2)}\n`, "utf-8");
		results.push({
			file: "tsconfig.client.json",
			action: "added \"shared/appkit-types\" to include"
		});
	} catch (err) {
		logger.warn("Failed to migrate tsconfig.client.json: %s", err.message);
	}
	return results;
}
async function migrateTsconfigServer(projectRoot) {
	const results = [];
	const filePath = path.join(projectRoot, "tsconfig.server.json");
	try {
		const raw = await fs.readFile(filePath, "utf-8");
		const parsed = JSON.parse(stripJsonComments(raw));
		const opts = parsed.compilerOptions;
		if (!opts || !opts.outDir) return results;
		delete opts.outDir;
		delete opts.declaration;
		delete opts.declarationMap;
		delete opts.sourceMap;
		opts.noEmit = true;
		await fs.writeFile(filePath, `${JSON.stringify(parsed, null, 2)}\n`, "utf-8");
		results.push({
			file: "tsconfig.server.json",
			action: "switched to noEmit mode"
		});
	} catch (err) {
		logger.warn("Failed to migrate tsconfig.server.json: %s", err.message);
	}
	return results;
}
const SCRIPT_MIGRATIONS = {
	"build:server": {
		old: "tsdown -c tsdown.server.config.ts",
		new: "tsc -b tsconfig.server.json && tsdown -c tsdown.server.config.ts"
	},
	typecheck: {
		old: "tsc -p ./tsconfig.server.json --noEmit && tsc -p ./tsconfig.client.json --noEmit",
		new: "tsc -b tsconfig.server.json && tsc -b tsconfig.client.json"
	}
};
async function migratePackageJsonScripts(projectRoot) {
	const results = [];
	const filePath = path.join(projectRoot, "package.json");
	try {
		const raw = await fs.readFile(filePath, "utf-8");
		const parsed = JSON.parse(raw);
		const scripts = parsed.scripts;
		if (!scripts) return results;
		const updated = [];
		for (const [name, { old, new: replacement }] of Object.entries(SCRIPT_MIGRATIONS)) if (scripts[name] === old) {
			scripts[name] = replacement;
			updated.push(name);
		}
		if (updated.length === 0) return results;
		const indent = raw.match(/^\s+/m)?.[0]?.length === 4 ? 4 : 2;
		await fs.writeFile(filePath, `${JSON.stringify(parsed, null, indent)}\n`, "utf-8");
		results.push({
			file: "package.json",
			action: `updated ${updated.join(" and ")} scripts`
		});
	} catch (err) {
		logger.warn("Failed to migrate package.json scripts: %s", err.message);
	}
	return results;
}
function printMigrationSummary(results) {
	const separator = pc.dim("─".repeat(50));
	console.log("");
	console.log(`  ${pc.bold("Typegen Migration")}`);
	console.log(`  ${separator}`);
	for (const { file, action } of results) console.log(`  ${pc.green("✓")} ${file.padEnd(24)} ${pc.dim(action)}`);
	console.log(`  ${separator}`);
	console.log("");
}

//#endregion
export { migrateProjectConfig, removeOldGeneratedTypes, resolveProjectRoot };
