// @databricks/appkit/dist/type-generator/query-registry.js
import { createLogger } from "../logging/logger.js";
import { CACHE_VERSION, hashSQL, loadCache, saveCache } from "./cache.js";
import { Spinner } from "./spinner.js";
import { sqlTypeToHelper, sqlTypeToMarker } from "./types.js";
import { WorkspaceClient } from "@databricks/sdk-experimental";
import pc from "picocolors";
import fs from "node:fs/promises";
import path from "node:path";

//#region src/type-generator/query-registry.ts
const logger = createLogger("type-generator:query-registry");
/**
* Regex breakdown:
*   '(?:[^']|'')*'   — matches a SQL string literal, including escaped '' pairs
*   |                 — alternation: whichever branch matches first at a position wins
*   --[^\n]*          — matches a single-line SQL comment
*
* Because the regex engine scans left-to-right, a `'` is consumed as a string
* literal before any `--` inside it could match as a comment — giving us
* correct single-pass ordering without a manual state machine.
*
* V1: no block-comment support (deferred to next PR).
*/
const PROTECTED_RANGE_RE = /'(?:[^']|'')*'|--[^\n]*/g;
/**
* Numeric-context patterns for positional type inference.
* Hoisted to module scope — safe because matchAll() clones the regex internally.
*/
const NUMERIC_PATTERNS = [
	/\bLIMIT\s+:([a-zA-Z_]\w*)/gi,
	/\bOFFSET\s+:([a-zA-Z_]\w*)/gi,
	/\bTOP\s+:([a-zA-Z_]\w*)/gi,
	/\bFETCH\s+FIRST\s+:([a-zA-Z_]\w*)\s+ROWS/gi,
	/[+\-*/]\s*:([a-zA-Z_]\w*)/g,
	/:([a-zA-Z_]\w*)\s*[+\-*/]/g
];
function getProtectedRanges(sql) {
	const ranges = [];
	for (const m of sql.matchAll(PROTECTED_RANGE_RE)) ranges.push([m.index, m.index + m[0].length]);
	return ranges;
}
function isInsideProtectedRange(offset, ranges) {
	return ranges.some(([start, end]) => offset >= start && offset < end);
}
/**
* Parse a raw API/SDK error into a structured code + message.
* Handles Databricks-style JSON bodies embedded in the message string,
* e.g. `Response from server (Bad Request) {"error_code":"...","message":"..."}`.
*/
function parseError(raw) {
	const jsonMatch = raw.match(/\{[\s\S]*\}/);
	if (jsonMatch) try {
		const parsed = JSON.parse(jsonMatch[0]);
		if (parsed.error_code || parsed.message) return {
			code: parsed.error_code,
			message: parsed.message || raw
		};
	} catch {}
	return { message: raw };
}
/**
* Extract parameters from a SQL query
* @param sql - the SQL query to extract parameters from
* @returns an array of parameter names
*/
function extractParameters(sql, ranges) {
	const protectedRanges = ranges ?? getProtectedRanges(sql);
	const matches = sql.matchAll(/(?<!:):([a-zA-Z_]\w*)/g);
	const params = /* @__PURE__ */ new Set();
	for (const match of matches) if (!isInsideProtectedRange(match.index, protectedRanges)) params.add(match[1]);
	return Array.from(params);
}
const SERVER_INJECTED_PARAMS = ["workspaceId"];
/**
* Generates the TypeScript type literal for query parameters from SQL.
* Shared by both the success and failure paths.
*/
function formatParametersType(sql) {
	const params = extractParameters(sql).filter((p) => !SERVER_INJECTED_PARAMS.includes(p));
	const paramTypes = extractParameterTypes(sql);
	return params.length > 0 ? `{\n      ${params.map((p) => {
		const sqlType = paramTypes[p];
		const markerType = sqlType ? sqlTypeToMarker[sqlType] : "SQLTypeMarker";
		const helper = sqlType ? sqlTypeToHelper[sqlType] : "sql.*()";
		return `/** ${sqlType || "any"} - use ${helper} */\n      ${p}: ${markerType}`;
	}).join(";\n      ")};\n    }` : "Record<string, never>";
}
function convertToQueryType(result, sql, queryName) {
	const columns = (result.result?.data_array || []).map((row) => ({
		name: row[0] || "",
		type_name: row[1]?.toUpperCase() || "STRING",
		comment: row[2] || void 0
	}));
	const paramsType = formatParametersType(sql);
	const resultFields = columns.map((column) => {
		const mappedType = typeMap[normalizeTypeName(column.type_name)] || "unknown";
		const name = /^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(column.name) ? column.name : `"${column.name}"`;
		return `${column.comment ? `/** ${column.comment} */\n      ` : `/** @sqlType ${column.type_name} */\n      `}${name}: ${mappedType}`;
	});
	const hasResults = resultFields.length > 0;
	return {
		type: `{
    name: "${queryName}";
    parameters: ${paramsType};
    result: ${hasResults ? `Array<{
      ${resultFields.join(";\n      ")};
    }>` : "unknown"};
  }`,
		hasResults
	};
}
/**
* Used when DESCRIBE QUERY fails so the query still appears in QueryRegistry.
* Generates a type with unknown result from SQL alone (no warehouse call).
*/
function generateUnknownResultQuery(sql, queryName) {
	return `{
    name: "${queryName}";
    parameters: ${formatParametersType(sql)};
    result: unknown;
  }`;
}
function extractParameterTypes(sql) {
	const paramTypes = {};
	const matches = sql.matchAll(/--\s*@param\s+(\w+)\s+(STRING|NUMERIC|DECIMAL|BIGINT|TINYINT|SMALLINT|INT|FLOAT|DOUBLE|BOOLEAN|DATE|TIMESTAMP_NTZ|TIMESTAMP|BINARY)\b/gi);
	for (const match of matches) {
		const [, paramName, paramType] = match;
		paramTypes[paramName] = paramType.toUpperCase();
	}
	return paramTypes;
}
function defaultForType(sqlType) {
	switch (sqlType?.toUpperCase()) {
		case "NUMERIC":
		case "DECIMAL":
		case "BIGINT":
		case "TINYINT":
		case "SMALLINT":
		case "INT": return "0";
		case "FLOAT":
		case "DOUBLE": return "0.0";
		case "STRING": return "''";
		case "BOOLEAN": return "true";
		case "DATE": return "'2000-01-01'";
		case "TIMESTAMP": return "'2000-01-01T00:00:00Z'";
		case "TIMESTAMP_NTZ": return "'2000-01-01T00:00:00'";
		case "BINARY": return "X'00'";
		default: return "''";
	}
}
/**
* Infer parameter types from positional context in SQL.
* V1 only infers NUMERIC from patterns like LIMIT, OFFSET, TOP,
* FETCH FIRST ... ROWS, and arithmetic operators.
* Parameters inside string literals or SQL comments are ignored.
*/
function inferParameterTypes(sql, ranges) {
	const inferred = {};
	const protectedRanges = ranges ?? getProtectedRanges(sql);
	for (const pattern of NUMERIC_PATTERNS) for (const match of sql.matchAll(pattern)) if (!isInsideProtectedRange(match.index, protectedRanges)) inferred[match[1]] = "NUMERIC";
	return inferred;
}
/**
* Generate query schemas from a folder of SQL files
* It uses DESCRIBE QUERY to get the schema without executing the query
* @param queryFolder - the folder containing the SQL files
* @param warehouseId - the warehouse id to use for schema analysis
* @param options - options for the query generation
* @param options.noCache - if true, skip the cache and regenerate all types
* @returns an array of query schemas
*/
async function generateQueriesFromDescribe(queryFolder, warehouseId, options = {}) {
	const { noCache = false, concurrency: rawConcurrency = 10 } = options;
	const concurrency = typeof rawConcurrency === "number" && Number.isFinite(rawConcurrency) ? Math.max(1, Math.floor(rawConcurrency)) : 10;
	const [allFiles, cache] = await Promise.all([fs.readdir(queryFolder), noCache ? {
		version: CACHE_VERSION,
		queries: {}
	} : loadCache()]);
	const queryFiles = allFiles.filter((file) => file.endsWith(".sql"));
	logger.debug("Found %d SQL queries", queryFiles.length);
	const client = new WorkspaceClient({});
	const spinner = new Spinner();
	const sqlContents = await Promise.all(queryFiles.map((file) => fs.readFile(path.join(queryFolder, file), "utf8")));
	const startTime = performance.now();
	const cachedResults = [];
	const uncachedQueries = [];
	const logEntries = [];
	for (let i = 0; i < queryFiles.length; i++) {
		const file = queryFiles[i];
		const queryName = normalizeQueryName(path.basename(file, ".sql"));
		const sql = sqlContents[i];
		const sqlHash = hashSQL(sql);
		const cached = cache.queries[queryName];
		if (cached && cached.hash === sqlHash && !cached.retry) {
			cachedResults.push({
				index: i,
				schema: {
					name: queryName,
					type: cached.type
				}
			});
			logEntries.push({
				queryName,
				status: "HIT"
			});
		} else {
			const protectedRanges = getProtectedRanges(sql);
			const annotatedTypes = extractParameterTypes(sql);
			const parameterTypes = {
				...inferParameterTypes(sql, protectedRanges),
				...annotatedTypes
			};
			const sqlWithDefaults = sql.replace(/(?<!:):([a-zA-Z_]\w*)/g, (original, paramName, offset) => {
				if (isInsideProtectedRange(offset, protectedRanges)) return original;
				return defaultForType(parameterTypes[paramName]);
			});
			const allParams = extractParameters(sql, protectedRanges);
			for (const param of allParams) {
				if (SERVER_INJECTED_PARAMS.includes(param)) continue;
				if (parameterTypes[param]) continue;
				logger.warn("%s: parameter \":%s\" has no type annotation or inference. Add %s to the query file.", queryFiles[i], param, `-- @param ${param} <TYPE>`);
			}
			const cleanedSql = sqlWithDefaults.trim().replace(/;\s*$/, "");
			uncachedQueries.push({
				index: i,
				queryName,
				sql,
				sqlHash,
				cleanedSql
			});
		}
	}
	const freshResults = [];
	if (uncachedQueries.length > 0) {
		let completed = 0;
		const total = uncachedQueries.length;
		spinner.start(`Describing ${total} ${total === 1 ? "query" : "queries"} (0/${total})`);
		const describeOne = async ({ index, queryName, sql, sqlHash, cleanedSql }) => {
			const result = await client.statementExecution.executeStatement({
				statement: `DESCRIBE QUERY ${cleanedSql}`,
				warehouse_id: warehouseId
			});
			completed++;
			spinner.update(`Describing ${total} ${total === 1 ? "query" : "queries"} (${completed}/${total})`);
			logger.debug("DESCRIBE result for %s: state=%s, rows=%d", queryName, result.status.state, result.result?.data_array?.length ?? 0);
			if (result.status.state === "FAILED") {
				const sqlError = result.status.error?.message || "Query execution failed";
				logger.warn("DESCRIBE failed for %s: %s", queryName, sqlError);
				const type = generateUnknownResultQuery(sql, queryName);
				return {
					status: "fail",
					index,
					schema: {
						name: queryName,
						type
					},
					cacheEntry: {
						hash: sqlHash,
						type,
						retry: true
					},
					error: parseError(sqlError)
				};
			}
			const { type, hasResults } = convertToQueryType(result, sql, queryName);
			return {
				status: "ok",
				index,
				schema: {
					name: queryName,
					type
				},
				cacheEntry: {
					hash: sqlHash,
					type,
					retry: !hasResults
				}
			};
		};
		const processBatchResults = (settled, batchOffset) => {
			for (let i = 0; i < settled.length; i++) {
				const entry = settled[i];
				const { queryName } = uncachedQueries[batchOffset + i];
				if (entry.status === "fulfilled") {
					const res = entry.value;
					freshResults.push({
						index: res.index,
						schema: res.schema
					});
					cache.queries[queryName] = res.cacheEntry;
					logEntries.push({
						queryName,
						status: "MISS",
						failed: res.status === "fail",
						error: res.status === "fail" ? res.error : void 0
					});
				} else {
					const { sql, sqlHash, index } = uncachedQueries[batchOffset + i];
					const reason = entry.reason instanceof Error ? entry.reason.message : String(entry.reason);
					logger.warn("DESCRIBE rejected for %s: %s", queryName, reason);
					const type = generateUnknownResultQuery(sql, queryName);
					freshResults.push({
						index,
						schema: {
							name: queryName,
							type
						}
					});
					cache.queries[queryName] = {
						hash: sqlHash,
						type,
						retry: true
					};
					logEntries.push({
						queryName,
						status: "MISS",
						failed: true,
						error: parseError(reason)
					});
				}
			}
		};
		if (uncachedQueries.length > concurrency) for (let b = 0; b < uncachedQueries.length; b += concurrency) {
			const batch = uncachedQueries.slice(b, b + concurrency);
			processBatchResults(await Promise.allSettled(batch.map(describeOne)), b);
			await saveCache(cache);
		}
		else {
			processBatchResults(await Promise.allSettled(uncachedQueries.map(describeOne)), 0);
			await saveCache(cache);
		}
		spinner.stop("");
	}
	const elapsed = ((performance.now() - startTime) / 1e3).toFixed(2);
	if (logEntries.length > 0) {
		const maxNameLen = Math.max(...logEntries.map((e) => e.queryName.length));
		const separator = pc.dim("─".repeat(50));
		console.log("");
		console.log(`  ${pc.bold("Typegen Queries")} ${pc.dim(`(${logEntries.length})`)}`);
		console.log(`  ${separator}`);
		for (const entry of logEntries) {
			const tag = entry.failed ? pc.bold(pc.red("ERROR")) : entry.status === "HIT" ? `cache ${pc.bold(pc.green("HIT  "))}` : `cache ${pc.bold(pc.yellow("MISS "))}`;
			const rawName = entry.queryName.padEnd(maxNameLen);
			const name = entry.failed ? pc.dim(pc.strikethrough(rawName)) : rawName;
			const errorCode = entry.error?.message.match(/\[([^\]]+)\]/)?.[1];
			const reason = errorCode ? `  ${pc.dim(errorCode)}` : "";
			console.log(`  ${tag}  ${name}${reason}`);
		}
		const newCount = logEntries.filter((e) => e.status === "MISS" && !e.failed).length;
		const cacheCount = logEntries.filter((e) => e.status === "HIT" && !e.failed).length;
		const errorCount = logEntries.filter((e) => e.failed).length;
		console.log(`  ${separator}`);
		const parts = [`${newCount} new`, `${cacheCount} from cache`];
		if (errorCount > 0) parts.push(`${errorCount} ${errorCount === 1 ? "error" : "errors"}`);
		console.log(`  ${parts.join(", ")}. ${pc.dim(`${elapsed}s`)}`);
		console.log("");
	}
	return [...cachedResults, ...freshResults].sort((a, b) => a.index - b.index).map((r) => r.schema);
}
/**
* Normalize query name by removing the .obo extension
* @param queryName - the query name to normalize
* @returns the normalized query name
*/
function normalizeQueryName(fileName) {
	return fileName.replace(/\.obo$/, "");
}
/**
* Normalize SQL type name by removing parameters/generics
* Examples:
*   DECIMAL(38,6) -> DECIMAL
*   ARRAY<STRING> -> ARRAY
*   MAP<STRING,INT> -> MAP
*   STRUCT<name:STRING> -> STRUCT
*   INTERVAL DAY TO SECOND -> INTERVAL
*   GEOGRAPHY(4326) -> GEOGRAPHY
*/
function normalizeTypeName(typeName) {
	return typeName.replace(/\(.*\)$/, "").replace(/<.*>$/, "").split(" ")[0];
}
/** Type Map for Databricks data types to JavaScript types */
const typeMap = {
	STRING: "string",
	BINARY: "string",
	BOOLEAN: "boolean",
	TINYINT: "number",
	SMALLINT: "number",
	INT: "number",
	BIGINT: "number",
	FLOAT: "number",
	DOUBLE: "number",
	DECIMAL: "number",
	DATE: "string",
	TIMESTAMP: "string",
	TIMESTAMP_NTZ: "string",
	INTERVAL: "string",
	ARRAY: "unknown[]",
	MAP: "Record<string, unknown>",
	STRUCT: "Record<string, unknown>",
	OBJECT: "Record<string, unknown>",
	VARIANT: "unknown",
	GEOGRAPHY: "unknown",
	GEOMETRY: "unknown",
	VOID: "null"
};

//#endregion
export { generateQueriesFromDescribe };
