// @databricks/appkit/dist/plugin/dev-reader.js
import { createLogger } from "../logging/logger.js";
import { TunnelError } from "../errors/tunnel.js";
import "../errors/index.js";
import { isRemoteTunnelAllowedByEnv } from "../plugins/server/remote-tunnel/gate.js";
import { randomUUID } from "node:crypto";

//#region src/plugin/dev-reader.ts
const logger = createLogger("plugin:dev-reader");
/**
* This class is used to read files from the local filesystem in dev mode
* through the WebSocket tunnel.
*/
var DevFileReader = class DevFileReader {
	static instance = null;
	getTunnelForRequest = null;
	constructor() {}
	static getInstance() {
		if (!DevFileReader.instance) DevFileReader.instance = new Proxy(new DevFileReader(), {
			get(target, prop, receiver) {
				if (isRemoteTunnelAllowedByEnv()) return Reflect.get(target, prop, receiver);
				const value = Reflect.get(target, prop, receiver);
				if (typeof value === "function") return function noop() {
					logger.debug("Noop: %s (remote server disabled)", String(prop));
					return Promise.resolve("");
				};
				return value;
			},
			set(target, prop, value, receiver) {
				return Reflect.set(target, prop, value, receiver);
			}
		});
		return DevFileReader.instance;
	}
	registerTunnelGetter(getter) {
		this.getTunnelForRequest = getter;
	}
	async readFile(filePath, req) {
		if (!this.getTunnelForRequest) throw TunnelError.getterNotRegistered();
		const tunnel = this.getTunnelForRequest(req);
		if (!tunnel) throw TunnelError.noConnection();
		const { ws, pendingFileReads } = tunnel;
		const requestId = randomUUID();
		return new Promise((resolve, reject) => {
			const timeout = setTimeout(() => {
				pendingFileReads.delete(requestId);
				reject(/* @__PURE__ */ new Error(`File read timeout: ${filePath}`));
			}, 1e4);
			pendingFileReads.set(requestId, {
				resolve,
				reject,
				timeout
			});
			ws.send(JSON.stringify({
				type: "file:read",
				requestId,
				path: filePath
			}));
		});
	}
	async readdir(dirPath, req) {
		if (!this.getTunnelForRequest) throw TunnelError.getterNotRegistered();
		const tunnel = this.getTunnelForRequest(req);
		if (!tunnel) throw TunnelError.noConnection();
		const { ws, pendingFileReads } = tunnel;
		const requestId = randomUUID();
		return new Promise((resolve, reject) => {
			const timeout = setTimeout(() => {
				pendingFileReads.delete(requestId);
				reject(/* @__PURE__ */ new Error(`Directory read timeout: ${dirPath}`));
			}, 1e4);
			pendingFileReads.set(requestId, {
				resolve: (data) => {
					try {
						const files = JSON.parse(data);
						if (!Array.isArray(files)) {
							reject(/* @__PURE__ */ new Error("Invalid directory listing format: expected array, got " + typeof files));
							return;
						}
						if (!files.every((f) => typeof f === "string")) {
							reject(/* @__PURE__ */ new Error("Invalid directory listing format: expected array of strings"));
							return;
						}
						resolve(files);
					} catch (error) {
						reject(/* @__PURE__ */ new Error(`Failed to parse directory listing: ${error.message}`));
					}
				},
				reject,
				timeout
			});
			ws.send(JSON.stringify({
				type: "dir:list",
				requestId,
				path: dirPath
			}));
		});
	}
};

//#endregion
export { DevFileReader };
