// @databricks/sdk-experimental/dist/apis/iamv2/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceIamV2Service = exports.AccountIamV2Service = void 0;
/**
 * These APIs allow you to manage Account Iam V2, Workspace Iam V2, etc.
 */
const config_1 = require("../../config");
const HostType_1 = require("../../config/HostType");
const iamv2 = __importStar(require("./model"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
class AccountIamV2RetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("AccountIamV2", method, message);
    }
}
class AccountIamV2Error extends apiError_1.ApiError {
    constructor(method, message) {
        super("AccountIamV2", method, message);
    }
}
/**
 * These APIs are used to manage identities and the workspace access of these
 * identities in <Databricks>.
 */
class AccountIamV2Service {
    constructor(client) {
        this.client = client;
    }
    async _getWorkspaceAccessDetail(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("view")) {
            query["view"] = request.view;
        }
        const path = `/api/2.0/identity/accounts/${config.accountId}/workspaces/${request.workspace_id}/workspaceAccessDetails/${request.principal_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns the access details for a principal in a workspace. Allows for
     * checking access details for any provisioned principal (user, service
     * principal, or group) in a workspace. * Provisioned principal here refers
     * to one that has been synced into Databricks from the customer's IdP or
     * added explicitly to Databricks via SCIM/UI. Allows for passing in a "view"
     * parameter to control what fields are returned (BASIC by default or FULL).
     */
    async getWorkspaceAccessDetail(request, context) {
        return await this._getWorkspaceAccessDetail(request, context);
    }
    async _resolveGroup(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("external_id")) {
            body["external_id"] = request.external_id;
        }
        const query = {};
        const path = `/api/2.0/identity/accounts/${config.accountId}/groups/resolveByExternalId`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Resolves a group with the given external ID from the customer's IdP. If
     * the group does not exist, it will be created in the account. If the
     * customer is not onboarded onto Automatic Identity Management (AIM), this
     * will return an error.
     */
    async resolveGroup(request, context) {
        return await this._resolveGroup(request, context);
    }
    async _resolveServicePrincipal(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("external_id")) {
            body["external_id"] = request.external_id;
        }
        const query = {};
        const path = `/api/2.0/identity/accounts/${config.accountId}/servicePrincipals/resolveByExternalId`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Resolves an SP with the given external ID from the customer's IdP. If the
     * SP does not exist, it will be created. If the customer is not onboarded
     * onto Automatic Identity Management (AIM), this will return an error.
     */
    async resolveServicePrincipal(request, context) {
        return await this._resolveServicePrincipal(request, context);
    }
    async _resolveUser(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("external_id")) {
            body["external_id"] = request.external_id;
        }
        const query = {};
        const path = `/api/2.0/identity/accounts/${config.accountId}/users/resolveByExternalId`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Resolves a user with the given external ID from the customer's IdP. If the
     * user does not exist, it will be created. If the customer is not onboarded
     * onto Automatic Identity Management (AIM), this will return an error.
     */
    async resolveUser(request, context) {
        return await this._resolveUser(request, context);
    }
}
exports.AccountIamV2Service = AccountIamV2Service;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountIamV2Service.prototype, "_getWorkspaceAccessDetail", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountIamV2Service.prototype, "getWorkspaceAccessDetail", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountIamV2Service.prototype, "_resolveGroup", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountIamV2Service.prototype, "resolveGroup", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountIamV2Service.prototype, "_resolveServicePrincipal", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountIamV2Service.prototype, "resolveServicePrincipal", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountIamV2Service.prototype, "_resolveUser", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountIamV2Service.prototype, "resolveUser", null);
class WorkspaceIamV2RetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("WorkspaceIamV2", method, message);
    }
}
class WorkspaceIamV2Error extends apiError_1.ApiError {
    constructor(method, message) {
        super("WorkspaceIamV2", method, message);
    }
}
/**
 * These APIs are used to manage identities and the workspace access of these
 * identities in <Databricks>.
 */
class WorkspaceIamV2Service {
    constructor(client) {
        this.client = client;
    }
    async _getWorkspaceAccessDetailLocal(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("view")) {
            query["view"] = request.view;
        }
        const path = `/api/2.0/identity/workspaceAccessDetails/${request.principal_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns the access details for a principal in the current workspace.
     * Allows for checking access details for any provisioned principal (user,
     * service principal, or group) in the current workspace. * Provisioned
     * principal here refers to one that has been synced into Databricks from the
     * customer's IdP or added explicitly to Databricks via SCIM/UI. Allows for
     * passing in a "view" parameter to control what fields are returned (BASIC
     * by default or FULL).
     */
    async getWorkspaceAccessDetailLocal(request, context) {
        return await this._getWorkspaceAccessDetailLocal(request, context);
    }
    async _resolveGroupProxy(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("external_id")) {
            body["external_id"] = request.external_id;
        }
        const query = {};
        const path = "/api/2.0/identity/groups/resolveByExternalId";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Resolves a group with the given external ID from the customer's IdP. If
     * the group does not exist, it will be created in the account. If the
     * customer is not onboarded onto Automatic Identity Management (AIM), this
     * will return an error.
     */
    async resolveGroupProxy(request, context) {
        return await this._resolveGroupProxy(request, context);
    }
    async _resolveServicePrincipalProxy(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("external_id")) {
            body["external_id"] = request.external_id;
        }
        const query = {};
        const path = "/api/2.0/identity/servicePrincipals/resolveByExternalId";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Resolves an SP with the given external ID from the customer's IdP. If the
     * SP does not exist, it will be created. If the customer is not onboarded
     * onto Automatic Identity Management (AIM), this will return an error.
     */
    async resolveServicePrincipalProxy(request, context) {
        return await this._resolveServicePrincipalProxy(request, context);
    }
    async _resolveUserProxy(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("external_id")) {
            body["external_id"] = request.external_id;
        }
        const query = {};
        const path = "/api/2.0/identity/users/resolveByExternalId";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Resolves a user with the given external ID from the customer's IdP. If the
     * user does not exist, it will be created. If the customer is not onboarded
     * onto Automatic Identity Management (AIM), this will return an error.
     */
    async resolveUserProxy(request, context) {
        return await this._resolveUserProxy(request, context);
    }
}
exports.WorkspaceIamV2Service = WorkspaceIamV2Service;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceIamV2Service.prototype, "_getWorkspaceAccessDetailLocal", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceIamV2Service.prototype, "getWorkspaceAccessDetailLocal", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceIamV2Service.prototype, "_resolveGroupProxy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceIamV2Service.prototype, "resolveGroupProxy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceIamV2Service.prototype, "_resolveServicePrincipalProxy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceIamV2Service.prototype, "resolveServicePrincipalProxy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceIamV2Service.prototype, "_resolveUserProxy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceIamV2Service.prototype, "resolveUserProxy", null);
