// @databricks/sdk-experimental/dist/apis/compute/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolicyFamiliesService = exports.PolicyFamiliesError = exports.PolicyFamiliesRetriableError = exports.PolicyComplianceForClustersService = exports.PolicyComplianceForClustersError = exports.PolicyComplianceForClustersRetriableError = exports.LibrariesService = exports.LibrariesError = exports.LibrariesRetriableError = exports.InstanceProfilesService = exports.InstanceProfilesError = exports.InstanceProfilesRetriableError = exports.InstancePoolsService = exports.InstancePoolsError = exports.InstancePoolsRetriableError = exports.GlobalInitScriptsService = exports.GlobalInitScriptsError = exports.GlobalInitScriptsRetriableError = exports.CommandExecutionService = exports.CommandExecutionError = exports.CommandExecutionRetriableError = exports.ClustersService = exports.ClustersError = exports.ClustersRetriableError = exports.ClusterPoliciesService = exports.ClusterPoliciesError = exports.ClusterPoliciesRetriableError = void 0;
const compute = __importStar(require("./model"));
const retries_1 = __importDefault(require("../../retries/retries"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
const wait_1 = require("../../wait");
class ClusterPoliciesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("ClusterPolicies", method, message);
    }
}
exports.ClusterPoliciesRetriableError = ClusterPoliciesRetriableError;
class ClusterPoliciesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("ClusterPolicies", method, message);
    }
}
exports.ClusterPoliciesError = ClusterPoliciesError;
/**
 * You can use cluster policies to control users' ability to configure clusters
 * based on a set of rules. These rules specify which attributes or attribute
 * values can be used during cluster creation. Cluster policies have ACLs that
 * limit their use to specific users and groups.
 *
 * With cluster policies, you can: - Auto-install cluster libraries on the next
 * restart by listing them in the policy's "libraries" field (Public Preview). -
 * Limit users to creating clusters with the prescribed settings. - Simplify the
 * user interface, enabling more users to create clusters, by fixing and hiding
 * some fields. - Manage costs by setting limits on attributes that impact the
 * hourly rate.
 *
 * Cluster policy permissions limit which policies a user can select in the
 * Policy drop-down when the user creates a cluster: - A user who has
 * unrestricted cluster create permission can select the Unrestricted policy and
 * create fully-configurable clusters. - A user who has both unrestricted cluster
 * create permission and access to cluster policies can select the Unrestricted
 * policy and policies they have access to. - A user that has access to only
 * cluster policies, can select the policies they have access to.
 *
 * If no policies exist in the workspace, the Policy drop-down doesn't appear.
 * Only admin users can create, edit, and delete policies. Admin users also have
 * access to all policies.
 */
class ClusterPoliciesService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("definition")) {
            body["definition"] = request.definition;
        }
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("libraries")) {
            body["libraries"] = request.libraries;
        }
        if (request.hasOwnProperty("max_clusters_per_user")) {
            body["max_clusters_per_user"] = request.max_clusters_per_user;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("policy_family_definition_overrides")) {
            body["policy_family_definition_overrides"] =
                request.policy_family_definition_overrides;
        }
        if (request.hasOwnProperty("policy_family_id")) {
            body["policy_family_id"] = request.policy_family_id;
        }
        const query = {};
        const path = "/api/2.0/policies/clusters/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new policy with prescribed settings.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("policy_id")) {
            body["policy_id"] = request.policy_id;
        }
        const query = {};
        const path = "/api/2.0/policies/clusters/delete";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a policy for a cluster. Clusters governed by this policy can still
     * run, but cannot be edited.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _edit(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("definition")) {
            body["definition"] = request.definition;
        }
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("libraries")) {
            body["libraries"] = request.libraries;
        }
        if (request.hasOwnProperty("max_clusters_per_user")) {
            body["max_clusters_per_user"] = request.max_clusters_per_user;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("policy_family_definition_overrides")) {
            body["policy_family_definition_overrides"] =
                request.policy_family_definition_overrides;
        }
        if (request.hasOwnProperty("policy_family_id")) {
            body["policy_family_id"] = request.policy_family_id;
        }
        if (request.hasOwnProperty("policy_id")) {
            body["policy_id"] = request.policy_id;
        }
        const query = {};
        const path = "/api/2.0/policies/clusters/edit";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update an existing policy for cluster. This operation may make some
     * clusters governed by the previous policy invalid.
     */
    async edit(request, context) {
        return await this._edit(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("policy_id")) {
            query["policy_id"] =
                request.policy_id;
        }
        const path = "/api/2.0/policies/clusters/get";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a cluster policy entity. Creation and editing is available to admins
     * only.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _getPermissionLevels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/cluster-policies/${request.cluster_policy_id}/permissionLevels`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(request, context) {
        return await this._getPermissionLevels(request, context);
    }
    async _getPermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/cluster-policies/${request.cluster_policy_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of a cluster policy. Cluster policies can inherit
     * permissions from their root object.
     */
    async getPermissions(request, context) {
        return await this._getPermissions(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("sort_column")) {
            query["sort_column"] =
                request.sort_column;
        }
        if (request.hasOwnProperty("sort_order")) {
            query["sort_order"] =
                request.sort_order;
        }
        const path = "/api/2.0/policies/clusters/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns a list of policies accessible by the requesting user.
     */
    async *list(request, context) {
        const response = (await this._list(request, context)).policies;
        for (const v of response || []) {
            yield v;
        }
    }
    async _setPermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/cluster-policies/${request.cluster_policy_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their root object.
     */
    async setPermissions(request, context) {
        return await this._setPermissions(request, context);
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/cluster-policies/${request.cluster_policy_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on a cluster policy. Cluster policies can inherit
     * permissions from their root object.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
}
exports.ClusterPoliciesService = ClusterPoliciesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "_edit", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "edit", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "_getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ClusterPoliciesService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "_setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClusterPoliciesService.prototype, "updatePermissions", null);
class ClustersRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Clusters", method, message);
    }
}
exports.ClustersRetriableError = ClustersRetriableError;
class ClustersError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Clusters", method, message);
    }
}
exports.ClustersError = ClustersError;
/**
 * The Clusters API allows you to create, start, edit, list, terminate, and
 * delete clusters.
 *
 * Databricks maps cluster node instance types to compute units known as DBUs.
 * See the instance type pricing page for a list of the supported instance types
 * and their corresponding DBUs.
 *
 * A Databricks cluster is a set of computation resources and configurations on
 * which you run data engineering, data science, and data analytics workloads,
 * such as production ETL pipelines, streaming analytics, ad-hoc analytics, and
 * machine learning.
 *
 * You run these workloads as a set of commands in a notebook or as an automated
 * job. Databricks makes a distinction between all-purpose clusters and job
 * clusters. You use all-purpose clusters to analyze data collaboratively using
 * interactive notebooks. You use job clusters to run fast and robust automated
 * jobs.
 *
 * You can create an all-purpose cluster using the UI, CLI, or REST API. You can
 * manually terminate and restart an all-purpose cluster. Multiple users can
 * share such clusters to do collaborative interactive analysis.
 *
 * IMPORTANT: Databricks retains cluster configuration information for terminated
 * clusters for 30 days. To keep an all-purpose cluster configuration even after
 * it has been terminated for more than 30 days, an administrator can pin a
 * cluster to the cluster list.
 */
class ClustersService {
    constructor(client) {
        this.client = client;
    }
    async _changeOwner(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        if (request.hasOwnProperty("owner_username")) {
            body["owner_username"] = request.owner_username;
        }
        const query = {};
        const path = "/api/2.1/clusters/change-owner";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Change the owner of the cluster. You must be an admin and the cluster must
     * be terminated to perform this operation. The service principal application
     * ID can be supplied as an argument to `owner_username`.
     */
    async changeOwner(request, context) {
        return await this._changeOwner(request, context);
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("apply_policy_default_values")) {
            body["apply_policy_default_values"] =
                request.apply_policy_default_values;
        }
        if (request.hasOwnProperty("autoscale")) {
            body["autoscale"] = request.autoscale;
        }
        if (request.hasOwnProperty("autotermination_minutes")) {
            body["autotermination_minutes"] = request.autotermination_minutes;
        }
        if (request.hasOwnProperty("aws_attributes")) {
            body["aws_attributes"] = request.aws_attributes;
        }
        if (request.hasOwnProperty("azure_attributes")) {
            body["azure_attributes"] = request.azure_attributes;
        }
        if (request.hasOwnProperty("clone_from")) {
            body["clone_from"] = request.clone_from;
        }
        if (request.hasOwnProperty("cluster_log_conf")) {
            body["cluster_log_conf"] = request.cluster_log_conf;
        }
        if (request.hasOwnProperty("cluster_name")) {
            body["cluster_name"] = request.cluster_name;
        }
        if (request.hasOwnProperty("custom_tags")) {
            body["custom_tags"] = request.custom_tags;
        }
        if (request.hasOwnProperty("data_security_mode")) {
            body["data_security_mode"] = request.data_security_mode;
        }
        if (request.hasOwnProperty("docker_image")) {
            body["docker_image"] = request.docker_image;
        }
        if (request.hasOwnProperty("driver_instance_pool_id")) {
            body["driver_instance_pool_id"] = request.driver_instance_pool_id;
        }
        if (request.hasOwnProperty("driver_node_type_id")) {
            body["driver_node_type_id"] = request.driver_node_type_id;
        }
        if (request.hasOwnProperty("enable_elastic_disk")) {
            body["enable_elastic_disk"] = request.enable_elastic_disk;
        }
        if (request.hasOwnProperty("enable_local_disk_encryption")) {
            body["enable_local_disk_encryption"] =
                request.enable_local_disk_encryption;
        }
        if (request.hasOwnProperty("gcp_attributes")) {
            body["gcp_attributes"] = request.gcp_attributes;
        }
        if (request.hasOwnProperty("init_scripts")) {
            body["init_scripts"] = request.init_scripts;
        }
        if (request.hasOwnProperty("instance_pool_id")) {
            body["instance_pool_id"] = request.instance_pool_id;
        }
        if (request.hasOwnProperty("is_single_node")) {
            body["is_single_node"] = request.is_single_node;
        }
        if (request.hasOwnProperty("kind")) {
            body["kind"] = request.kind;
        }
        if (request.hasOwnProperty("node_type_id")) {
            body["node_type_id"] = request.node_type_id;
        }
        if (request.hasOwnProperty("num_workers")) {
            body["num_workers"] = request.num_workers;
        }
        if (request.hasOwnProperty("policy_id")) {
            body["policy_id"] = request.policy_id;
        }
        if (request.hasOwnProperty("remote_disk_throughput")) {
            body["remote_disk_throughput"] = request.remote_disk_throughput;
        }
        if (request.hasOwnProperty("runtime_engine")) {
            body["runtime_engine"] = request.runtime_engine;
        }
        if (request.hasOwnProperty("single_user_name")) {
            body["single_user_name"] = request.single_user_name;
        }
        if (request.hasOwnProperty("spark_conf")) {
            body["spark_conf"] = request.spark_conf;
        }
        if (request.hasOwnProperty("spark_env_vars")) {
            body["spark_env_vars"] = request.spark_env_vars;
        }
        if (request.hasOwnProperty("spark_version")) {
            body["spark_version"] = request.spark_version;
        }
        if (request.hasOwnProperty("ssh_public_keys")) {
            body["ssh_public_keys"] = request.ssh_public_keys;
        }
        if (request.hasOwnProperty("total_initial_remote_disk_size")) {
            body["total_initial_remote_disk_size"] =
                request.total_initial_remote_disk_size;
        }
        if (request.hasOwnProperty("use_ml_runtime")) {
            body["use_ml_runtime"] = request.use_ml_runtime;
        }
        if (request.hasOwnProperty("workload_type")) {
            body["workload_type"] = request.workload_type;
        }
        const query = {};
        const path = "/api/2.1/clusters/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new Spark cluster. This method will acquire new instances from
     * the cloud provider if necessary. This method is asynchronous; the returned
     * ``cluster_id`` can be used to poll the cluster status. When this method
     * returns, the cluster will be in a ``PENDING`` state. The cluster will be
     * usable once it enters a ``RUNNING`` state. Note: Databricks may not be
     * able to acquire some of the requested nodes, due to cloud provider
     * limitations (account limits, spot price, etc.) or transient network
     * issues.
     *
     * If Databricks acquires at least 85% of the requested on-demand nodes,
     * cluster creation will succeed. Otherwise the cluster will terminate with
     * an informative error message.
     *
     * Rather than authoring the cluster's JSON definition from scratch,
     * Databricks recommends filling out the [create compute UI] and then copying
     * the generated JSON definition from the UI.
     *
     * [create compute UI]: https://docs.databricks.com/compute/configure.html
     */
    async create(createCluster, context) {
        let cancellationToken = context?.cancellationToken;
        const createClusterResponse = await this._create(createCluster, context);
        return (0, wait_1.asWaiter)(createClusterResponse, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        cluster_id: createClusterResponse.cluster_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Clusters.createAndWait: cancelled");
                        throw new ClustersError("createAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state;
                    const statusMessage = pollResponse.state_message;
                    switch (status) {
                        case "RUNNING": {
                            return pollResponse;
                        }
                        case "ERROR":
                        case "TERMINATED": {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.createAndWait: ${errorMessage}`);
                            throw new ClustersError("createAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.createAndWait: retrying: ${errorMessage}`);
                            throw new ClustersRetriableError("createAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _delete(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        const query = {};
        const path = "/api/2.1/clusters/delete";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Terminates the Spark cluster with the specified ID. The cluster is removed
     * asynchronously. Once the termination has completed, the cluster will be in
     * a `TERMINATED` state. If the cluster is already in a `TERMINATING` or
     * `TERMINATED` state, nothing will happen.
     */
    async delete(deleteCluster, context) {
        let cancellationToken = context?.cancellationToken;
        await this._delete(deleteCluster, context);
        return (0, wait_1.asWaiter)(null, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        cluster_id: deleteCluster.cluster_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Clusters.deleteAndWait: cancelled");
                        throw new ClustersError("deleteAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state;
                    const statusMessage = pollResponse.state_message;
                    switch (status) {
                        case "TERMINATED": {
                            return pollResponse;
                        }
                        case "ERROR": {
                            const errorMessage = `failed to reach TERMINATED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.deleteAndWait: ${errorMessage}`);
                            throw new ClustersError("deleteAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach TERMINATED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.deleteAndWait: retrying: ${errorMessage}`);
                            throw new ClustersRetriableError("deleteAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _edit(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("apply_policy_default_values")) {
            body["apply_policy_default_values"] =
                request.apply_policy_default_values;
        }
        if (request.hasOwnProperty("autoscale")) {
            body["autoscale"] = request.autoscale;
        }
        if (request.hasOwnProperty("autotermination_minutes")) {
            body["autotermination_minutes"] = request.autotermination_minutes;
        }
        if (request.hasOwnProperty("aws_attributes")) {
            body["aws_attributes"] = request.aws_attributes;
        }
        if (request.hasOwnProperty("azure_attributes")) {
            body["azure_attributes"] = request.azure_attributes;
        }
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        if (request.hasOwnProperty("cluster_log_conf")) {
            body["cluster_log_conf"] = request.cluster_log_conf;
        }
        if (request.hasOwnProperty("cluster_name")) {
            body["cluster_name"] = request.cluster_name;
        }
        if (request.hasOwnProperty("custom_tags")) {
            body["custom_tags"] = request.custom_tags;
        }
        if (request.hasOwnProperty("data_security_mode")) {
            body["data_security_mode"] = request.data_security_mode;
        }
        if (request.hasOwnProperty("docker_image")) {
            body["docker_image"] = request.docker_image;
        }
        if (request.hasOwnProperty("driver_instance_pool_id")) {
            body["driver_instance_pool_id"] = request.driver_instance_pool_id;
        }
        if (request.hasOwnProperty("driver_node_type_id")) {
            body["driver_node_type_id"] = request.driver_node_type_id;
        }
        if (request.hasOwnProperty("enable_elastic_disk")) {
            body["enable_elastic_disk"] = request.enable_elastic_disk;
        }
        if (request.hasOwnProperty("enable_local_disk_encryption")) {
            body["enable_local_disk_encryption"] =
                request.enable_local_disk_encryption;
        }
        if (request.hasOwnProperty("gcp_attributes")) {
            body["gcp_attributes"] = request.gcp_attributes;
        }
        if (request.hasOwnProperty("init_scripts")) {
            body["init_scripts"] = request.init_scripts;
        }
        if (request.hasOwnProperty("instance_pool_id")) {
            body["instance_pool_id"] = request.instance_pool_id;
        }
        if (request.hasOwnProperty("is_single_node")) {
            body["is_single_node"] = request.is_single_node;
        }
        if (request.hasOwnProperty("kind")) {
            body["kind"] = request.kind;
        }
        if (request.hasOwnProperty("node_type_id")) {
            body["node_type_id"] = request.node_type_id;
        }
        if (request.hasOwnProperty("num_workers")) {
            body["num_workers"] = request.num_workers;
        }
        if (request.hasOwnProperty("policy_id")) {
            body["policy_id"] = request.policy_id;
        }
        if (request.hasOwnProperty("remote_disk_throughput")) {
            body["remote_disk_throughput"] = request.remote_disk_throughput;
        }
        if (request.hasOwnProperty("runtime_engine")) {
            body["runtime_engine"] = request.runtime_engine;
        }
        if (request.hasOwnProperty("single_user_name")) {
            body["single_user_name"] = request.single_user_name;
        }
        if (request.hasOwnProperty("spark_conf")) {
            body["spark_conf"] = request.spark_conf;
        }
        if (request.hasOwnProperty("spark_env_vars")) {
            body["spark_env_vars"] = request.spark_env_vars;
        }
        if (request.hasOwnProperty("spark_version")) {
            body["spark_version"] = request.spark_version;
        }
        if (request.hasOwnProperty("ssh_public_keys")) {
            body["ssh_public_keys"] = request.ssh_public_keys;
        }
        if (request.hasOwnProperty("total_initial_remote_disk_size")) {
            body["total_initial_remote_disk_size"] =
                request.total_initial_remote_disk_size;
        }
        if (request.hasOwnProperty("use_ml_runtime")) {
            body["use_ml_runtime"] = request.use_ml_runtime;
        }
        if (request.hasOwnProperty("workload_type")) {
            body["workload_type"] = request.workload_type;
        }
        const query = {};
        const path = "/api/2.1/clusters/edit";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the configuration of a cluster to match the provided attributes
     * and size. A cluster can be updated if it is in a `RUNNING` or `TERMINATED`
     * state.
     *
     * If a cluster is updated while in a `RUNNING` state, it will be restarted
     * so that the new attributes can take effect.
     *
     * If a cluster is updated while in a `TERMINATED` state, it will remain
     * `TERMINATED`. The next time it is started using the `clusters/start` API,
     * the new attributes will take effect. Any attempt to update a cluster in
     * any other state will be rejected with an `INVALID_STATE` error code.
     *
     * Clusters created by the Databricks Jobs service cannot be edited.
     */
    async edit(editCluster, context) {
        let cancellationToken = context?.cancellationToken;
        await this._edit(editCluster, context);
        return (0, wait_1.asWaiter)(null, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        cluster_id: editCluster.cluster_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Clusters.editAndWait: cancelled");
                        throw new ClustersError("editAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state;
                    const statusMessage = pollResponse.state_message;
                    switch (status) {
                        case "RUNNING": {
                            return pollResponse;
                        }
                        case "ERROR":
                        case "TERMINATED": {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.editAndWait: ${errorMessage}`);
                            throw new ClustersError("editAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.editAndWait: retrying: ${errorMessage}`);
                            throw new ClustersRetriableError("editAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _events(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        if (request.hasOwnProperty("end_time")) {
            body["end_time"] = request.end_time;
        }
        if (request.hasOwnProperty("event_types")) {
            body["event_types"] = request.event_types;
        }
        if (request.hasOwnProperty("limit")) {
            body["limit"] = request.limit;
        }
        if (request.hasOwnProperty("offset")) {
            body["offset"] = request.offset;
        }
        if (request.hasOwnProperty("order")) {
            body["order"] = request.order;
        }
        if (request.hasOwnProperty("page_size")) {
            body["page_size"] = request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            body["page_token"] = request.page_token;
        }
        if (request.hasOwnProperty("start_time")) {
            body["start_time"] = request.start_time;
        }
        const query = {};
        const path = "/api/2.1/clusters/events";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieves a list of events about the activity of a cluster. This API is
     * paginated. If there are more events to read, the response includes all the
     * parameters necessary to request the next page of events.
     */
    async *events(request, context) {
        while (true) {
            const response = await this._events(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.events || response.events.length === 0) {
                break;
            }
            for (const v of response.events) {
                yield v;
            }
            if (!response.next_page) {
                break;
            }
            request = response.next_page;
        }
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("cluster_id")) {
            query["cluster_id"] =
                request.cluster_id;
        }
        const path = "/api/2.1/clusters/get";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieves the information for a cluster given its identifier. Clusters can
     * be described while they are running, or up to 60 days after they are
     * terminated.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _getPermissionLevels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/clusters/${request.cluster_id}/permissionLevels`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(request, context) {
        return await this._getPermissionLevels(request, context);
    }
    async _getPermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/clusters/${request.cluster_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of a cluster. Clusters can inherit permissions from
     * their root object.
     */
    async getPermissions(request, context) {
        return await this._getPermissions(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("filter_by")) {
            query["filter_by"] =
                request.filter_by;
        }
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        if (request.hasOwnProperty("sort_by")) {
            query["sort_by"] = request.sort_by;
        }
        const path = "/api/2.1/clusters/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Return information about all pinned and active clusters, and all clusters
     * terminated within the last 30 days. Clusters terminated prior to this
     * period are not included.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.clusters || response.clusters.length === 0) {
                break;
            }
            for (const v of response.clusters) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _listNodeTypes(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.1/clusters/list-node-types";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns a list of supported Spark node types. These node types can be used
     * to launch a cluster.
     */
    async listNodeTypes(context) {
        return await this._listNodeTypes(context);
    }
    async _listZones(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.1/clusters/list-zones";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns a list of availability zones where clusters can be created in (For
     * example, us-west-2a). These zones can be used to launch a cluster.
     */
    async listZones(context) {
        return await this._listZones(context);
    }
    async _permanentDelete(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        const query = {};
        const path = "/api/2.1/clusters/permanent-delete";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Permanently deletes a Spark cluster. This cluster is terminated and
     * resources are asynchronously removed.
     *
     * In addition, users will no longer see permanently deleted clusters in the
     * cluster list, and API users can no longer perform any action on
     * permanently deleted clusters.
     */
    async permanentDelete(request, context) {
        return await this._permanentDelete(request, context);
    }
    async _pin(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        const query = {};
        const path = "/api/2.1/clusters/pin";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Pinning a cluster ensures that the cluster will always be returned by the
     * ListClusters API. Pinning a cluster that is already pinned will have no
     * effect. This API can only be called by workspace admins.
     */
    async pin(request, context) {
        return await this._pin(request, context);
    }
    async _resize(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("autoscale")) {
            body["autoscale"] = request.autoscale;
        }
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        if (request.hasOwnProperty("num_workers")) {
            body["num_workers"] = request.num_workers;
        }
        const query = {};
        const path = "/api/2.1/clusters/resize";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Resizes a cluster to have a desired number of workers. This will fail
     * unless the cluster is in a `RUNNING` state.
     */
    async resize(resizeCluster, context) {
        let cancellationToken = context?.cancellationToken;
        await this._resize(resizeCluster, context);
        return (0, wait_1.asWaiter)(null, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        cluster_id: resizeCluster.cluster_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Clusters.resizeAndWait: cancelled");
                        throw new ClustersError("resizeAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state;
                    const statusMessage = pollResponse.state_message;
                    switch (status) {
                        case "RUNNING": {
                            return pollResponse;
                        }
                        case "ERROR":
                        case "TERMINATED": {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.resizeAndWait: ${errorMessage}`);
                            throw new ClustersError("resizeAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.resizeAndWait: retrying: ${errorMessage}`);
                            throw new ClustersRetriableError("resizeAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _restart(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        if (request.hasOwnProperty("restart_user")) {
            body["restart_user"] = request.restart_user;
        }
        const query = {};
        const path = "/api/2.1/clusters/restart";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Restarts a Spark cluster with the supplied ID. If the cluster is not
     * currently in a `RUNNING` state, nothing will happen.
     */
    async restart(restartCluster, context) {
        let cancellationToken = context?.cancellationToken;
        await this._restart(restartCluster, context);
        return (0, wait_1.asWaiter)(null, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        cluster_id: restartCluster.cluster_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Clusters.restartAndWait: cancelled");
                        throw new ClustersError("restartAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state;
                    const statusMessage = pollResponse.state_message;
                    switch (status) {
                        case "RUNNING": {
                            return pollResponse;
                        }
                        case "ERROR":
                        case "TERMINATED": {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.restartAndWait: ${errorMessage}`);
                            throw new ClustersError("restartAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.restartAndWait: retrying: ${errorMessage}`);
                            throw new ClustersRetriableError("restartAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _setPermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/clusters/${request.cluster_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their root object.
     */
    async setPermissions(request, context) {
        return await this._setPermissions(request, context);
    }
    async _sparkVersions(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.1/clusters/spark-versions";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns the list of available Spark versions. These versions can be used
     * to launch a cluster.
     */
    async sparkVersions(context) {
        return await this._sparkVersions(context);
    }
    async _start(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        const query = {};
        const path = "/api/2.1/clusters/start";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Starts a terminated Spark cluster with the supplied ID. This works similar
     * to `createCluster` except: - The previous cluster id and attributes are
     * preserved. - The cluster starts with the last specified cluster size. - If
     * the previous cluster was an autoscaling cluster, the current cluster
     * starts with the minimum number of nodes. - If the cluster is not currently
     * in a ``TERMINATED`` state, nothing will happen. - Clusters launched to run
     * a job cannot be started.
     */
    async start(startCluster, context) {
        let cancellationToken = context?.cancellationToken;
        await this._start(startCluster, context);
        return (0, wait_1.asWaiter)(null, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        cluster_id: startCluster.cluster_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Clusters.startAndWait: cancelled");
                        throw new ClustersError("startAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state;
                    const statusMessage = pollResponse.state_message;
                    switch (status) {
                        case "RUNNING": {
                            return pollResponse;
                        }
                        case "ERROR":
                        case "TERMINATED": {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.startAndWait: ${errorMessage}`);
                            throw new ClustersError("startAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.startAndWait: retrying: ${errorMessage}`);
                            throw new ClustersRetriableError("startAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _unpin(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        const query = {};
        const path = "/api/2.1/clusters/unpin";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Unpinning a cluster will allow the cluster to eventually be removed from
     * the ListClusters API. Unpinning a cluster that is not pinned will have no
     * effect. This API can only be called by workspace admins.
     */
    async unpin(request, context) {
        return await this._unpin(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("cluster")) {
            body["cluster"] = request.cluster;
        }
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        if (request.hasOwnProperty("update_mask")) {
            body["update_mask"] = request.update_mask;
        }
        const query = {};
        const path = "/api/2.1/clusters/update";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the configuration of a cluster to match the partial set of
     * attributes and size. Denote which fields to update using the `update_mask`
     * field in the request body. A cluster can be updated if it is in a
     * `RUNNING` or `TERMINATED` state. If a cluster is updated while in a
     * `RUNNING` state, it will be restarted so that the new attributes can take
     * effect. If a cluster is updated while in a `TERMINATED` state, it will
     * remain `TERMINATED`. The updated attributes will take effect the next time
     * the cluster is started using the `clusters/start` API. Attempts to update
     * a cluster in any other state will be rejected with an `INVALID_STATE`
     * error code. Clusters created by the Databricks Jobs service cannot be
     * updated.
     */
    async update(updateCluster, context) {
        let cancellationToken = context?.cancellationToken;
        await this._update(updateCluster, context);
        return (0, wait_1.asWaiter)(null, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        cluster_id: updateCluster.cluster_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Clusters.updateAndWait: cancelled");
                        throw new ClustersError("updateAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state;
                    const statusMessage = pollResponse.state_message;
                    switch (status) {
                        case "RUNNING": {
                            return pollResponse;
                        }
                        case "ERROR":
                        case "TERMINATED": {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.updateAndWait: ${errorMessage}`);
                            throw new ClustersError("updateAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Clusters.updateAndWait: retrying: ${errorMessage}`);
                            throw new ClustersRetriableError("updateAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/clusters/${request.cluster_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on a cluster. Clusters can inherit permissions
     * from their root object.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
}
exports.ClustersService = ClustersService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_changeOwner", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "changeOwner", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_edit", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "edit", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_events", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ClustersService.prototype, "events", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ClustersService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_listNodeTypes", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "listNodeTypes", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_listZones", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "listZones", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_permanentDelete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "permanentDelete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_pin", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "pin", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_resize", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "resize", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_restart", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "restart", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_sparkVersions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "sparkVersions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_start", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "start", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_unpin", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "unpin", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ClustersService.prototype, "updatePermissions", null);
class CommandExecutionRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("CommandExecution", method, message);
    }
}
exports.CommandExecutionRetriableError = CommandExecutionRetriableError;
class CommandExecutionError extends apiError_1.ApiError {
    constructor(method, message) {
        super("CommandExecution", method, message);
    }
}
exports.CommandExecutionError = CommandExecutionError;
/**
 * This API allows execution of Python, Scala, SQL, or R commands on running
 * Databricks Clusters. This API only supports (classic) all-purpose clusters.
 * Serverless compute is not supported.
 */
class CommandExecutionService {
    constructor(client) {
        this.client = client;
    }
    async _cancel(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("clusterId")) {
            body["clusterId"] = request.clusterId;
        }
        if (request.hasOwnProperty("commandId")) {
            body["commandId"] = request.commandId;
        }
        if (request.hasOwnProperty("contextId")) {
            body["contextId"] = request.contextId;
        }
        const query = {};
        const path = "/api/1.2/commands/cancel";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Cancels a currently running command within an execution context.
     *
     * The command ID is obtained from a prior successful call to __execute__.
     */
    async cancel(cancelCommand, context) {
        let cancellationToken = context?.cancellationToken;
        await this._cancel(cancelCommand, context);
        return (0, wait_1.asWaiter)(null, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.commandStatus({
                        clusterId: cancelCommand.clusterId,
                        commandId: cancelCommand.commandId,
                        contextId: cancelCommand.contextId,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("CommandExecution.cancelAndWait: cancelled");
                        throw new CommandExecutionError("cancelAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.status;
                    const statusMessage = pollResponse.results.cause;
                    switch (status) {
                        case "Cancelled": {
                            return pollResponse;
                        }
                        case "Error": {
                            const errorMessage = `failed to reach Cancelled state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`CommandExecution.cancelAndWait: ${errorMessage}`);
                            throw new CommandExecutionError("cancelAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach Cancelled state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`CommandExecution.cancelAndWait: retrying: ${errorMessage}`);
                            throw new CommandExecutionRetriableError("cancelAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _commandStatus(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("clusterId")) {
            query["clusterId"] =
                request.clusterId;
        }
        if (request.hasOwnProperty("commandId")) {
            query["commandId"] =
                request.commandId;
        }
        if (request.hasOwnProperty("contextId")) {
            query["contextId"] =
                request.contextId;
        }
        const path = "/api/1.2/commands/status";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the status of and, if available, the results from a currently
     * executing command.
     *
     * The command ID is obtained from a prior successful call to __execute__.
     */
    async commandStatus(request, context) {
        return await this._commandStatus(request, context);
    }
    async _contextStatus(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("clusterId")) {
            query["clusterId"] =
                request.clusterId;
        }
        if (request.hasOwnProperty("contextId")) {
            query["contextId"] =
                request.contextId;
        }
        const path = "/api/1.2/contexts/status";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the status for an execution context.
     */
    async contextStatus(request, context) {
        return await this._contextStatus(request, context);
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("clusterId")) {
            body["clusterId"] = request.clusterId;
        }
        if (request.hasOwnProperty("language")) {
            body["language"] = request.language;
        }
        const query = {};
        const path = "/api/1.2/contexts/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates an execution context for running cluster commands.
     *
     * If successful, this method returns the ID of the new execution context.
     */
    async create(createContext, context) {
        let cancellationToken = context?.cancellationToken;
        const created = await this._create(createContext, context);
        return (0, wait_1.asWaiter)(created, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.contextStatus({
                        clusterId: createContext.clusterId,
                        contextId: created.id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("CommandExecution.createAndWait: cancelled");
                        throw new CommandExecutionError("createAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.status;
                    const statusMessage = pollResponse;
                    switch (status) {
                        case "Running": {
                            return pollResponse;
                        }
                        case "Error": {
                            const errorMessage = `failed to reach Running state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`CommandExecution.createAndWait: ${errorMessage}`);
                            throw new CommandExecutionError("createAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach Running state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`CommandExecution.createAndWait: retrying: ${errorMessage}`);
                            throw new CommandExecutionRetriableError("createAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _destroy(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("clusterId")) {
            body["clusterId"] = request.clusterId;
        }
        if (request.hasOwnProperty("contextId")) {
            body["contextId"] = request.contextId;
        }
        const query = {};
        const path = "/api/1.2/contexts/destroy";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes an execution context.
     */
    async destroy(request, context) {
        return await this._destroy(request, context);
    }
    async _execute(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("clusterId")) {
            body["clusterId"] = request.clusterId;
        }
        if (request.hasOwnProperty("command")) {
            body["command"] = request.command;
        }
        if (request.hasOwnProperty("contextId")) {
            body["contextId"] = request.contextId;
        }
        if (request.hasOwnProperty("language")) {
            body["language"] = request.language;
        }
        const query = {};
        const path = "/api/1.2/commands/execute";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Runs a cluster command in the given execution context, using the provided
     * language.
     *
     * If successful, it returns an ID for tracking the status of the command's
     * execution.
     */
    async execute(command, context) {
        let cancellationToken = context?.cancellationToken;
        const created = await this._execute(command, context);
        return (0, wait_1.asWaiter)(created, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.commandStatus({
                        clusterId: command.clusterId,
                        commandId: created.id,
                        contextId: command.contextId,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("CommandExecution.executeAndWait: cancelled");
                        throw new CommandExecutionError("executeAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.status;
                    const statusMessage = pollResponse;
                    switch (status) {
                        case "Finished":
                        case "Error": {
                            return pollResponse;
                        }
                        case "Cancelled":
                        case "Cancelling": {
                            const errorMessage = `failed to reach Finished or Error state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`CommandExecution.executeAndWait: ${errorMessage}`);
                            throw new CommandExecutionError("executeAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach Finished or Error state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`CommandExecution.executeAndWait: retrying: ${errorMessage}`);
                            throw new CommandExecutionRetriableError("executeAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
}
exports.CommandExecutionService = CommandExecutionService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CommandExecutionService.prototype, "_cancel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CommandExecutionService.prototype, "cancel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CommandExecutionService.prototype, "_commandStatus", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CommandExecutionService.prototype, "commandStatus", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CommandExecutionService.prototype, "_contextStatus", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CommandExecutionService.prototype, "contextStatus", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CommandExecutionService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CommandExecutionService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CommandExecutionService.prototype, "_destroy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CommandExecutionService.prototype, "destroy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CommandExecutionService.prototype, "_execute", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CommandExecutionService.prototype, "execute", null);
class GlobalInitScriptsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("GlobalInitScripts", method, message);
    }
}
exports.GlobalInitScriptsRetriableError = GlobalInitScriptsRetriableError;
class GlobalInitScriptsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("GlobalInitScripts", method, message);
    }
}
exports.GlobalInitScriptsError = GlobalInitScriptsError;
/**
 * The Global Init Scripts API enables Workspace administrators to configure
 * global initialization scripts for their workspace. These scripts run on every
 * node in every cluster in the workspace.
 *
 * **Important:** Existing clusters must be restarted to pick up any changes made
 * to global init scripts. Global init scripts are run in order. If the init
 * script returns with a bad exit code, the Apache Spark container fails to
 * launch and init scripts with later position are skipped. If enough containers
 * fail, the entire cluster fails with a `GLOBAL_INIT_SCRIPT_FAILURE` error code.
 */
class GlobalInitScriptsService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("enabled")) {
            body["enabled"] = request.enabled;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("position")) {
            body["position"] = request.position;
        }
        if (request.hasOwnProperty("script")) {
            body["script"] = request.script;
        }
        const query = {};
        const path = "/api/2.0/global-init-scripts";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new global init script in this workspace.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/global-init-scripts/${request.script_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a global init script.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/global-init-scripts/${request.script_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets all the details of a script, including its Base64-encoded contents.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/global-init-scripts";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a list of all global init scripts for this workspace. This returns all
     * properties for each script but **not** the script contents. To retrieve
     * the contents of a script, use the [get a global init
     * script](:method:globalinitscripts/get) operation.
     */
    async *list(context) {
        const response = (await this._list(context)).scripts;
        for (const v of response || []) {
            yield v;
        }
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("enabled")) {
            body["enabled"] = request.enabled;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("position")) {
            body["position"] = request.position;
        }
        if (request.hasOwnProperty("script")) {
            body["script"] = request.script;
        }
        const query = {};
        const path = `/api/2.0/global-init-scripts/${request.script_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates a global init script, specifying only the fields to change. All
     * fields are optional. Unspecified fields retain their current value.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.GlobalInitScriptsService = GlobalInitScriptsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GlobalInitScriptsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GlobalInitScriptsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GlobalInitScriptsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GlobalInitScriptsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GlobalInitScriptsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GlobalInitScriptsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], GlobalInitScriptsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Object)
], GlobalInitScriptsService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GlobalInitScriptsService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GlobalInitScriptsService.prototype, "update", null);
class InstancePoolsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("InstancePools", method, message);
    }
}
exports.InstancePoolsRetriableError = InstancePoolsRetriableError;
class InstancePoolsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("InstancePools", method, message);
    }
}
exports.InstancePoolsError = InstancePoolsError;
/**
 * Instance Pools API are used to create, edit, delete and list instance pools by
 * using ready-to-use cloud instances which reduces a cluster start and
 * auto-scaling times.
 *
 * Databricks pools reduce cluster start and auto-scaling times by maintaining a
 * set of idle, ready-to-use instances. When a cluster is attached to a pool,
 * cluster nodes are created using the pool’s idle instances. If the pool has
 * no idle instances, the pool expands by allocating a new instance from the
 * instance provider in order to accommodate the cluster’s request. When a
 * cluster releases an instance, it returns to the pool and is free for another
 * cluster to use. Only clusters attached to a pool can use that pool’s idle
 * instances.
 *
 * You can specify a different pool for the driver node and worker nodes, or use
 * the same pool for both.
 *
 * Databricks does not charge DBUs while instances are idle in the pool. Instance
 * provider billing does apply. See pricing.
 */
class InstancePoolsService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("aws_attributes")) {
            body["aws_attributes"] = request.aws_attributes;
        }
        if (request.hasOwnProperty("azure_attributes")) {
            body["azure_attributes"] = request.azure_attributes;
        }
        if (request.hasOwnProperty("custom_tags")) {
            body["custom_tags"] = request.custom_tags;
        }
        if (request.hasOwnProperty("disk_spec")) {
            body["disk_spec"] = request.disk_spec;
        }
        if (request.hasOwnProperty("enable_elastic_disk")) {
            body["enable_elastic_disk"] = request.enable_elastic_disk;
        }
        if (request.hasOwnProperty("gcp_attributes")) {
            body["gcp_attributes"] = request.gcp_attributes;
        }
        if (request.hasOwnProperty("idle_instance_autotermination_minutes")) {
            body["idle_instance_autotermination_minutes"] =
                request.idle_instance_autotermination_minutes;
        }
        if (request.hasOwnProperty("instance_pool_name")) {
            body["instance_pool_name"] = request.instance_pool_name;
        }
        if (request.hasOwnProperty("max_capacity")) {
            body["max_capacity"] = request.max_capacity;
        }
        if (request.hasOwnProperty("min_idle_instances")) {
            body["min_idle_instances"] = request.min_idle_instances;
        }
        if (request.hasOwnProperty("node_type_id")) {
            body["node_type_id"] = request.node_type_id;
        }
        if (request.hasOwnProperty("preloaded_docker_images")) {
            body["preloaded_docker_images"] = request.preloaded_docker_images;
        }
        if (request.hasOwnProperty("preloaded_spark_versions")) {
            body["preloaded_spark_versions"] = request.preloaded_spark_versions;
        }
        if (request.hasOwnProperty("remote_disk_throughput")) {
            body["remote_disk_throughput"] = request.remote_disk_throughput;
        }
        if (request.hasOwnProperty("total_initial_remote_disk_size")) {
            body["total_initial_remote_disk_size"] =
                request.total_initial_remote_disk_size;
        }
        const query = {};
        const path = "/api/2.0/instance-pools/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new instance pool using idle and ready-to-use cloud instances.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("instance_pool_id")) {
            body["instance_pool_id"] = request.instance_pool_id;
        }
        const query = {};
        const path = "/api/2.0/instance-pools/delete";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes the instance pool permanently. The idle instances in the pool are
     * terminated asynchronously.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _edit(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("custom_tags")) {
            body["custom_tags"] = request.custom_tags;
        }
        if (request.hasOwnProperty("idle_instance_autotermination_minutes")) {
            body["idle_instance_autotermination_minutes"] =
                request.idle_instance_autotermination_minutes;
        }
        if (request.hasOwnProperty("instance_pool_id")) {
            body["instance_pool_id"] = request.instance_pool_id;
        }
        if (request.hasOwnProperty("instance_pool_name")) {
            body["instance_pool_name"] = request.instance_pool_name;
        }
        if (request.hasOwnProperty("max_capacity")) {
            body["max_capacity"] = request.max_capacity;
        }
        if (request.hasOwnProperty("min_idle_instances")) {
            body["min_idle_instances"] = request.min_idle_instances;
        }
        if (request.hasOwnProperty("node_type_id")) {
            body["node_type_id"] = request.node_type_id;
        }
        if (request.hasOwnProperty("remote_disk_throughput")) {
            body["remote_disk_throughput"] = request.remote_disk_throughput;
        }
        if (request.hasOwnProperty("total_initial_remote_disk_size")) {
            body["total_initial_remote_disk_size"] =
                request.total_initial_remote_disk_size;
        }
        const query = {};
        const path = "/api/2.0/instance-pools/edit";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Modifies the configuration of an existing instance pool.
     */
    async edit(request, context) {
        return await this._edit(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("instance_pool_id")) {
            query["instance_pool_id"] =
                request.instance_pool_id;
        }
        const path = "/api/2.0/instance-pools/get";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieve the information for an instance pool based on its identifier.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _getPermissionLevels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/instance-pools/${request.instance_pool_id}/permissionLevels`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(request, context) {
        return await this._getPermissionLevels(request, context);
    }
    async _getPermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/instance-pools/${request.instance_pool_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of an instance pool. Instance pools can inherit
     * permissions from their root object.
     */
    async getPermissions(request, context) {
        return await this._getPermissions(request, context);
    }
    async _list(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/instance-pools/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a list of instance pools with their statistics.
     */
    async *list(context) {
        const response = (await this._list(context)).instance_pools;
        for (const v of response || []) {
            yield v;
        }
    }
    async _setPermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/instance-pools/${request.instance_pool_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their root object.
     */
    async setPermissions(request, context) {
        return await this._setPermissions(request, context);
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/instance-pools/${request.instance_pool_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on an instance pool. Instance pools can inherit
     * permissions from their root object.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
}
exports.InstancePoolsService = InstancePoolsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "_edit", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "edit", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "_getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Object)
], InstancePoolsService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "_setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstancePoolsService.prototype, "updatePermissions", null);
class InstanceProfilesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("InstanceProfiles", method, message);
    }
}
exports.InstanceProfilesRetriableError = InstanceProfilesRetriableError;
class InstanceProfilesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("InstanceProfiles", method, message);
    }
}
exports.InstanceProfilesError = InstanceProfilesError;
/**
 * The Instance Profiles API allows admins to add, list, and remove instance
 * profiles that users can launch clusters with. Regular users can list the
 * instance profiles available to them. See [Secure access to S3 buckets] using
 * instance profiles for more information.
 *
 * [Secure access to S3 buckets]: https://docs.databricks.com/administration-guide/cloud-configurations/aws/instance-profiles.html
 */
class InstanceProfilesService {
    constructor(client) {
        this.client = client;
    }
    async _add(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("iam_role_arn")) {
            body["iam_role_arn"] = request.iam_role_arn;
        }
        if (request.hasOwnProperty("instance_profile_arn")) {
            body["instance_profile_arn"] = request.instance_profile_arn;
        }
        if (request.hasOwnProperty("is_meta_instance_profile")) {
            body["is_meta_instance_profile"] = request.is_meta_instance_profile;
        }
        if (request.hasOwnProperty("skip_validation")) {
            body["skip_validation"] = request.skip_validation;
        }
        const query = {};
        const path = "/api/2.0/instance-profiles/add";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Registers an instance profile in Databricks. In the UI, you can then give
     * users the permission to use this instance profile when launching clusters.
     *
     * This API is only available to admin users.
     */
    async add(request, context) {
        return await this._add(request, context);
    }
    async _edit(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("iam_role_arn")) {
            body["iam_role_arn"] = request.iam_role_arn;
        }
        if (request.hasOwnProperty("instance_profile_arn")) {
            body["instance_profile_arn"] = request.instance_profile_arn;
        }
        if (request.hasOwnProperty("is_meta_instance_profile")) {
            body["is_meta_instance_profile"] = request.is_meta_instance_profile;
        }
        const query = {};
        const path = "/api/2.0/instance-profiles/edit";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * The only supported field to change is the optional IAM role ARN associated
     * with the instance profile. It is required to specify the IAM role ARN if
     * both of the following are true:
     *
     * * Your role name and instance profile name do not match. The name is the
     * part after the last slash in each ARN. * You want to use the instance
     * profile with [Databricks SQL Serverless].
     *
     * To understand where these fields are in the AWS console, see [Enable
     * serverless SQL warehouses].
     *
     * This API is only available to admin users.
     *
     * [Databricks SQL Serverless]: https://docs.databricks.com/sql/admin/serverless.html
     * [Enable serverless SQL warehouses]: https://docs.databricks.com/sql/admin/serverless.html
     */
    async edit(request, context) {
        return await this._edit(request, context);
    }
    async _list(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/instance-profiles/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List the instance profiles that the calling user can use to launch a
     * cluster.
     *
     * This API is available to all users.
     */
    async *list(context) {
        const response = (await this._list(context)).instance_profiles;
        for (const v of response || []) {
            yield v;
        }
    }
    async _remove(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("instance_profile_arn")) {
            body["instance_profile_arn"] = request.instance_profile_arn;
        }
        const query = {};
        const path = "/api/2.0/instance-profiles/remove";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Remove the instance profile with the provided ARN. Existing clusters with
     * this instance profile will continue to function.
     *
     * This API is only accessible to admin users.
     */
    async remove(request, context) {
        return await this._remove(request, context);
    }
}
exports.InstanceProfilesService = InstanceProfilesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstanceProfilesService.prototype, "_add", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstanceProfilesService.prototype, "add", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstanceProfilesService.prototype, "_edit", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstanceProfilesService.prototype, "edit", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], InstanceProfilesService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Object)
], InstanceProfilesService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstanceProfilesService.prototype, "_remove", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], InstanceProfilesService.prototype, "remove", null);
class LibrariesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Libraries", method, message);
    }
}
exports.LibrariesRetriableError = LibrariesRetriableError;
class LibrariesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Libraries", method, message);
    }
}
exports.LibrariesError = LibrariesError;
/**
 * The Libraries API allows you to install and uninstall libraries and get the
 * status of libraries on a cluster.
 *
 * To make third-party or custom code available to notebooks and jobs running on
 * your clusters, you can install a library. Libraries can be written in Python,
 * Java, Scala, and R. You can upload Python, Java, Scala and R libraries and
 * point to external packages in PyPI, Maven, and CRAN repositories.
 *
 * Cluster libraries can be used by all notebooks running on a cluster. You can
 * install a cluster library directly from a public repository such as PyPI or
 * Maven, using a previously installed workspace library, or using an init
 * script.
 *
 * When you uninstall a library from a cluster, the library is removed only when
 * you restart the cluster. Until you restart the cluster, the status of the
 * uninstalled library appears as Uninstall pending restart.
 */
class LibrariesService {
    constructor(client) {
        this.client = client;
    }
    async _allClusterStatuses(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/libraries/all-cluster-statuses";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get the status of all libraries on all clusters. A status is returned for
     * all libraries installed on this cluster via the API or the libraries UI.
     */
    async *allClusterStatuses(context) {
        const response = (await this._allClusterStatuses(context)).statuses;
        for (const v of response || []) {
            yield v;
        }
    }
    async _clusterStatus(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("cluster_id")) {
            query["cluster_id"] =
                request.cluster_id;
        }
        const path = "/api/2.0/libraries/cluster-status";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get the status of libraries on a cluster. A status is returned for all
     * libraries installed on this cluster via the API or the libraries UI. The
     * order of returned libraries is as follows: 1. Libraries set to be
     * installed on this cluster, in the order that the libraries were added to
     * the cluster, are returned first. 2. Libraries that were previously
     * requested to be installed on this cluster or, but are now marked for
     * removal, in no particular order, are returned last.
     */
    async *clusterStatus(request, context) {
        const response = (await this._clusterStatus(request, context))
            .library_statuses;
        for (const v of response || []) {
            yield v;
        }
    }
    async _install(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        if (request.hasOwnProperty("libraries")) {
            body["libraries"] = request.libraries;
        }
        const query = {};
        const path = "/api/2.0/libraries/install";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Add libraries to install on a cluster. The installation is asynchronous;
     * it happens in the background after the completion of this request.
     */
    async install(request, context) {
        return await this._install(request, context);
    }
    async _uninstall(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        if (request.hasOwnProperty("libraries")) {
            body["libraries"] = request.libraries;
        }
        const query = {};
        const path = "/api/2.0/libraries/uninstall";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Set libraries to uninstall from a cluster. The libraries won't be
     * uninstalled until the cluster is restarted. A request to uninstall a
     * library that is not currently installed is ignored.
     */
    async uninstall(request, context) {
        return await this._uninstall(request, context);
    }
}
exports.LibrariesService = LibrariesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], LibrariesService.prototype, "_allClusterStatuses", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Object)
], LibrariesService.prototype, "allClusterStatuses", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LibrariesService.prototype, "_clusterStatus", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], LibrariesService.prototype, "clusterStatus", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LibrariesService.prototype, "_install", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LibrariesService.prototype, "install", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LibrariesService.prototype, "_uninstall", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LibrariesService.prototype, "uninstall", null);
class PolicyComplianceForClustersRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("PolicyComplianceForClusters", method, message);
    }
}
exports.PolicyComplianceForClustersRetriableError = PolicyComplianceForClustersRetriableError;
class PolicyComplianceForClustersError extends apiError_1.ApiError {
    constructor(method, message) {
        super("PolicyComplianceForClusters", method, message);
    }
}
exports.PolicyComplianceForClustersError = PolicyComplianceForClustersError;
/**
 * The policy compliance APIs allow you to view and manage the policy compliance
 * status of clusters in your workspace.
 *
 * A cluster is compliant with its policy if its configuration satisfies all its
 * policy rules. Clusters could be out of compliance if their policy was updated
 * after the cluster was last edited.
 *
 * The get and list compliance APIs allow you to view the policy compliance
 * status of a cluster. The enforce compliance API allows you to update a cluster
 * to be compliant with the current version of its policy.
 */
class PolicyComplianceForClustersService {
    constructor(client) {
        this.client = client;
    }
    async _enforceCompliance(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("cluster_id")) {
            body["cluster_id"] = request.cluster_id;
        }
        if (request.hasOwnProperty("validate_only")) {
            body["validate_only"] = request.validate_only;
        }
        const query = {};
        const path = "/api/2.0/policies/clusters/enforce-compliance";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates a cluster to be compliant with the current version of its policy.
     * A cluster can be updated if it is in a `RUNNING` or `TERMINATED` state.
     *
     * If a cluster is updated while in a `RUNNING` state, it will be restarted
     * so that the new attributes can take effect.
     *
     * If a cluster is updated while in a `TERMINATED` state, it will remain
     * `TERMINATED`. The next time the cluster is started, the new attributes
     * will take effect.
     *
     * Clusters created by the Databricks Jobs, DLT, or Models services cannot be
     * enforced by this API. Instead, use the "Enforce job policy compliance" API
     * to enforce policy compliance on jobs.
     */
    async enforceCompliance(request, context) {
        return await this._enforceCompliance(request, context);
    }
    async _getCompliance(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("cluster_id")) {
            query["cluster_id"] =
                request.cluster_id;
        }
        const path = "/api/2.0/policies/clusters/get-compliance";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns the policy compliance status of a cluster. Clusters could be out
     * of compliance if their policy was updated after the cluster was last
     * edited.
     */
    async getCompliance(request, context) {
        return await this._getCompliance(request, context);
    }
    async _listCompliance(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        if (request.hasOwnProperty("policy_id")) {
            query["policy_id"] =
                request.policy_id;
        }
        const path = "/api/2.0/policies/clusters/list-compliance";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns the policy compliance status of all clusters that use a given
     * policy. Clusters could be out of compliance if their policy was updated
     * after the cluster was last edited.
     */
    async *listCompliance(request, context) {
        while (true) {
            const response = await this._listCompliance(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.clusters || response.clusters.length === 0) {
                break;
            }
            for (const v of response.clusters) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
}
exports.PolicyComplianceForClustersService = PolicyComplianceForClustersService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PolicyComplianceForClustersService.prototype, "_enforceCompliance", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PolicyComplianceForClustersService.prototype, "enforceCompliance", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PolicyComplianceForClustersService.prototype, "_getCompliance", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PolicyComplianceForClustersService.prototype, "getCompliance", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PolicyComplianceForClustersService.prototype, "_listCompliance", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], PolicyComplianceForClustersService.prototype, "listCompliance", null);
class PolicyFamiliesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("PolicyFamilies", method, message);
    }
}
exports.PolicyFamiliesRetriableError = PolicyFamiliesRetriableError;
class PolicyFamiliesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("PolicyFamilies", method, message);
    }
}
exports.PolicyFamiliesError = PolicyFamiliesError;
/**
 * View available policy families. A policy family contains a policy definition
 * providing best practices for configuring clusters for a particular use case.
 *
 * Databricks manages and provides policy families for several common cluster use
 * cases. You cannot create, edit, or delete policy families.
 *
 * Policy families cannot be used directly to create clusters. Instead, you
 * create cluster policies using a policy family. Cluster policies created using
 * a policy family inherit the policy family's policy definition.
 */
class PolicyFamiliesService {
    constructor(client) {
        this.client = client;
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("version")) {
            query["version"] = request.version;
        }
        const path = `/api/2.0/policy-families/${request.policy_family_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieve the information for an policy family based on its identifier and
     * version
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/policy-families";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns the list of policy definition types available to use at their
     * latest version. This API is paginated.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.policy_families ||
                response.policy_families.length === 0) {
                break;
            }
            for (const v of response.policy_families) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
}
exports.PolicyFamiliesService = PolicyFamiliesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PolicyFamiliesService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PolicyFamiliesService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PolicyFamiliesService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], PolicyFamiliesService.prototype, "list", null);
