// @databricks/sdk-experimental/dist/apis/workspace/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceService = exports.WorkspaceError = exports.WorkspaceRetriableError = exports.SecretsService = exports.SecretsError = exports.SecretsRetriableError = exports.ReposService = exports.ReposError = exports.ReposRetriableError = exports.GitCredentialsService = exports.GitCredentialsError = exports.GitCredentialsRetriableError = void 0;
const workspace = __importStar(require("./model"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
class GitCredentialsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("GitCredentials", method, message);
    }
}
exports.GitCredentialsRetriableError = GitCredentialsRetriableError;
class GitCredentialsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("GitCredentials", method, message);
    }
}
exports.GitCredentialsError = GitCredentialsError;
/**
 * Registers personal access token for Databricks to do operations on behalf of
 * the user.
 *
 * See [more info].
 *
 * [more info]: https://docs.databricks.com/repos/get-access-tokens-from-git-provider.html
 */
class GitCredentialsService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("git_email")) {
            body["git_email"] = request.git_email;
        }
        if (request.hasOwnProperty("git_provider")) {
            body["git_provider"] = request.git_provider;
        }
        if (request.hasOwnProperty("git_username")) {
            body["git_username"] = request.git_username;
        }
        if (request.hasOwnProperty("is_default_for_provider")) {
            body["is_default_for_provider"] = request.is_default_for_provider;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("personal_access_token")) {
            body["personal_access_token"] = request.personal_access_token;
        }
        const query = {};
        const path = "/api/2.0/git-credentials";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a Git credential entry for the user. Only one Git credential per
     * user is supported, so any attempts to create credentials if an entry
     * already exists will fail. Use the PATCH endpoint to update existing
     * credentials, or the DELETE endpoint to delete existing credentials.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/git-credentials/${request.credential_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes the specified Git credential.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/git-credentials/${request.credential_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the Git credential with the specified credential ID.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/git-credentials";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists the calling user's Git credentials. One credential per user is
     * supported.
     */
    async *list(context) {
        const response = (await this._list(context)).credentials;
        for (const v of response || []) {
            yield v;
        }
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("git_email")) {
            body["git_email"] = request.git_email;
        }
        if (request.hasOwnProperty("git_provider")) {
            body["git_provider"] = request.git_provider;
        }
        if (request.hasOwnProperty("git_username")) {
            body["git_username"] = request.git_username;
        }
        if (request.hasOwnProperty("is_default_for_provider")) {
            body["is_default_for_provider"] = request.is_default_for_provider;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("personal_access_token")) {
            body["personal_access_token"] = request.personal_access_token;
        }
        const query = {};
        const path = `/api/2.0/git-credentials/${request.credential_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the specified Git credential.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.GitCredentialsService = GitCredentialsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GitCredentialsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GitCredentialsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GitCredentialsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GitCredentialsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GitCredentialsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GitCredentialsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], GitCredentialsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Object)
], GitCredentialsService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GitCredentialsService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GitCredentialsService.prototype, "update", null);
class ReposRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Repos", method, message);
    }
}
exports.ReposRetriableError = ReposRetriableError;
class ReposError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Repos", method, message);
    }
}
exports.ReposError = ReposError;
/**
 * The Repos API allows users to manage their git repos. Users can use the API to
 * access all repos that they have manage permissions on.
 *
 * Databricks Repos is a visual Git client in Databricks. It supports common Git
 * operations such a cloning a repository, committing and pushing, pulling,
 * branch management, and visual comparison of diffs when committing.
 *
 * Within Repos you can develop code in notebooks or other files and follow data
 * science and engineering code development best practices using Git for version
 * control, collaboration, and CI/CD.
 */
class ReposService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("path")) {
            body["path"] = request.path;
        }
        if (request.hasOwnProperty("provider")) {
            body["provider"] = request.provider;
        }
        if (request.hasOwnProperty("sparse_checkout")) {
            body["sparse_checkout"] = request.sparse_checkout;
        }
        if (request.hasOwnProperty("url")) {
            body["url"] = request.url;
        }
        const query = {};
        const path = "/api/2.0/repos";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a repo in the workspace and links it to the remote Git repo
     * specified. Note that repos created programmatically must be linked to a
     * remote Git repo, unlike repos created in the browser.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/repos/${request.repo_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes the specified repo.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/repos/${request.repo_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns the repo with the given repo ID.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _getPermissionLevels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/repos/${request.repo_id}/permissionLevels`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(request, context) {
        return await this._getPermissionLevels(request, context);
    }
    async _getPermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/repos/${request.repo_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of a repo. Repos can inherit permissions from their
     * root object.
     */
    async getPermissions(request, context) {
        return await this._getPermissions(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("next_page_token")) {
            query["next_page_token"] =
                request.next_page_token;
        }
        if (request.hasOwnProperty("path_prefix")) {
            query["path_prefix"] =
                request.path_prefix;
        }
        const path = "/api/2.0/repos";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns repos that the calling user has Manage permissions on. Use
     * `next_page_token` to iterate through additional pages.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.repos || response.repos.length === 0) {
                break;
            }
            for (const v of response.repos) {
                yield v;
            }
            request.next_page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _setPermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/repos/${request.repo_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their root object.
     */
    async setPermissions(request, context) {
        return await this._setPermissions(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("branch")) {
            body["branch"] = request.branch;
        }
        if (request.hasOwnProperty("sparse_checkout")) {
            body["sparse_checkout"] = request.sparse_checkout;
        }
        if (request.hasOwnProperty("tag")) {
            body["tag"] = request.tag;
        }
        const query = {};
        const path = `/api/2.0/repos/${request.repo_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the repo to a different branch or tag, or updates the repo to the
     * latest commit on the same branch.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/repos/${request.repo_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on a repo. Repos can inherit permissions from
     * their root object.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
}
exports.ReposService = ReposService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "_getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ReposService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "_setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ReposService.prototype, "updatePermissions", null);
class SecretsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Secrets", method, message);
    }
}
exports.SecretsRetriableError = SecretsRetriableError;
class SecretsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Secrets", method, message);
    }
}
exports.SecretsError = SecretsError;
/**
 * The Secrets API allows you to manage secrets, secret scopes, and access
 * permissions.
 *
 * Sometimes accessing data requires that you authenticate to external data
 * sources through JDBC. Instead of directly entering your credentials into a
 * notebook, use Databricks secrets to store your credentials and reference them
 * in notebooks and jobs.
 *
 * Administrators, secret creators, and users granted permission can read
 * Databricks secrets. While Databricks makes an effort to redact secret values
 * that might be displayed in notebooks, it is not possible to prevent such users
 * from reading secrets.
 */
class SecretsService {
    constructor(client) {
        this.client = client;
    }
    async _createScope(request, context) {
        const headers = new Headers({ "Content-Type": "application/json" });
        const body = {};
        if (request.hasOwnProperty("backend_azure_keyvault")) {
            body["backend_azure_keyvault"] = request.backend_azure_keyvault;
        }
        if (request.hasOwnProperty("initial_manage_principal")) {
            body["initial_manage_principal"] = request.initial_manage_principal;
        }
        if (request.hasOwnProperty("scope")) {
            body["scope"] = request.scope;
        }
        if (request.hasOwnProperty("scope_backend_type")) {
            body["scope_backend_type"] = request.scope_backend_type;
        }
        const query = {};
        const path = "/api/2.0/secrets/scopes/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new secret scope.
     *
     * The scope name must consist of alphanumeric characters, dashes,
     * underscores, and periods, and may not exceed 128 characters.
     *
     * Example request:
     *
     * .. code::
     *
     * { "scope": "my-simple-databricks-scope", "initial_manage_principal":
     * "users" "scope_backend_type": "databricks|azure_keyvault", # below is only
     * required if scope type is azure_keyvault "backend_azure_keyvault": {
     * "resource_id":
     * "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/xxxx/providers/Microsoft.KeyVault/vaults/xxxx",
     * "tenant_id": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx", "dns_name":
     * "https://xxxx.vault.azure.net/", } }
     *
     * If ``initial_manage_principal`` is specified, the initial ACL applied to
     * the scope is applied to the supplied principal (user or group) with
     * ``MANAGE`` permissions. The only supported principal for this option is
     * the group ``users``, which contains all users in the workspace. If
     * ``initial_manage_principal`` is not specified, the initial ACL with
     * ``MANAGE`` permission applied to the scope is assigned to the API request
     * issuer's user identity.
     *
     * If ``scope_backend_type`` is ``azure_keyvault``, a secret scope is created
     * with secrets from a given Azure KeyVault. The caller must provide the
     * keyvault_resource_id and the tenant_id for the key vault. If
     * ``scope_backend_type`` is ``databricks`` or is unspecified, an empty
     * secret scope is created and stored in Databricks's own storage.
     *
     * Throws ``RESOURCE_ALREADY_EXISTS`` if a scope with the given name already
     * exists. Throws ``RESOURCE_LIMIT_EXCEEDED`` if maximum number of scopes in
     * the workspace is exceeded. Throws ``INVALID_PARAMETER_VALUE`` if the scope
     * name is invalid. Throws ``BAD_REQUEST`` if request violated constraints.
     * Throws ``CUSTOMER_UNAUTHORIZED`` if normal user attempts to create a scope
     * with name reserved for databricks internal usage. Throws
     * ``UNAUTHENTICATED`` if unable to verify user access permission on Azure
     * KeyVault
     */
    async createScope(request, context) {
        return await this._createScope(request, context);
    }
    async _deleteAcl(request, context) {
        const headers = new Headers({ "Content-Type": "application/json" });
        const body = {};
        if (request.hasOwnProperty("principal")) {
            body["principal"] = request.principal;
        }
        if (request.hasOwnProperty("scope")) {
            body["scope"] = request.scope;
        }
        const query = {};
        const path = "/api/2.0/secrets/acls/delete";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes the given ACL on the given scope.
     *
     * Users must have the ``MANAGE`` permission to invoke this API.
     *
     * Example request:
     *
     * .. code::
     *
     * { "scope": "my-secret-scope", "principal": "data-scientists" }
     *
     * Throws ``RESOURCE_DOES_NOT_EXIST`` if no such secret scope, principal, or
     * ACL exists. Throws ``PERMISSION_DENIED`` if the user does not have
     * permission to make this API call. Throws ``INVALID_PARAMETER_VALUE`` if
     * the permission or principal is invalid.
     */
    async deleteAcl(request, context) {
        return await this._deleteAcl(request, context);
    }
    async _deleteScope(request, context) {
        const headers = new Headers({ "Content-Type": "application/json" });
        const body = {};
        if (request.hasOwnProperty("scope")) {
            body["scope"] = request.scope;
        }
        const query = {};
        const path = "/api/2.0/secrets/scopes/delete";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a secret scope.
     *
     * Example request:
     *
     * .. code::
     *
     * { "scope": "my-secret-scope" }
     *
     * Throws ``RESOURCE_DOES_NOT_EXIST`` if the scope does not exist. Throws
     * ``PERMISSION_DENIED`` if the user does not have permission to make this
     * API call. Throws ``BAD_REQUEST`` if system user attempts to delete
     * internal secret scope.
     */
    async deleteScope(request, context) {
        return await this._deleteScope(request, context);
    }
    async _deleteSecret(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("key")) {
            body["key"] = request.key;
        }
        if (request.hasOwnProperty("scope")) {
            body["scope"] = request.scope;
        }
        const query = {};
        const path = "/api/2.0/secrets/delete";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes the secret stored in this secret scope. You must have ``WRITE`` or
     * ``MANAGE`` permission on the Secret Scope.
     *
     * Example request:
     *
     * .. code::
     *
     * { "scope": "my-secret-scope", "key": "my-secret-key" }
     *
     * Throws ``RESOURCE_DOES_NOT_EXIST`` if no such secret scope or secret
     * exists. Throws ``PERMISSION_DENIED`` if the user does not have permission
     * to make this API call. Throws ``BAD_REQUEST`` if system user attempts to
     * delete an internal secret, or request is made against Azure KeyVault
     * backed scope.
     */
    async deleteSecret(request, context) {
        return await this._deleteSecret(request, context);
    }
    async _getAcl(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("principal")) {
            query["principal"] =
                request.principal;
        }
        if (request.hasOwnProperty("scope")) {
            query["scope"] = request.scope;
        }
        const path = "/api/2.0/secrets/acls/get";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Describes the details about the given ACL, such as the group and
     * permission.
     *
     * Users must have the ``MANAGE`` permission to invoke this API.
     *
     * Example response:
     *
     * .. code::
     *
     * { "principal": "data-scientists", "permission": "READ" }
     *
     * Throws ``RESOURCE_DOES_NOT_EXIST`` if no such secret scope exists. Throws
     * ``PERMISSION_DENIED`` if the user does not have permission to make this
     * API call. Throws ``INVALID_PARAMETER_VALUE`` if the permission or
     * principal is invalid.
     */
    async getAcl(request, context) {
        return await this._getAcl(request, context);
    }
    async _getSecret(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("key")) {
            query["key"] = request.key;
        }
        if (request.hasOwnProperty("scope")) {
            query["scope"] = request.scope;
        }
        const path = "/api/2.0/secrets/get";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a secret for a given key and scope. This API can only be called from
     * the DBUtils interface. Users need the READ permission to make this call.
     *
     * Example response:
     *
     * .. code::
     *
     * { "key": "my-string-key", "value": <bytes of the secret value> }
     *
     * Note that the secret value returned is in bytes. The interpretation of the
     * bytes is determined by the caller in DBUtils and the type the data is
     * decoded into.
     *
     * Throws ``RESOURCE_DOES_NOT_EXIST`` if no such secret or secret scope
     * exists. Throws ``PERMISSION_DENIED`` if the user does not have permission
     * to make this API call.
     *
     * Note: This is explicitly an undocumented API. It also doesn't need to be
     * supported for the /preview prefix, because it's not a customer-facing API
     * (i.e. only used for DBUtils SecretUtils to fetch secrets).
     *
     * Throws ``RESOURCE_DOES_NOT_EXIST`` if no such secret scope or secret
     * exists. Throws ``BAD_REQUEST`` if normal user calls get secret outside of
     * a notebook. AKV specific errors: Throws ``INVALID_PARAMETER_VALUE`` if
     * secret name is not alphanumeric or too long. Throws ``PERMISSION_DENIED``
     * if secret manager cannot access AKV with 403 error Throws
     * ``MALFORMED_REQUEST`` if secret manager cannot access AKV with any other
     * 4xx error
     */
    async getSecret(request, context) {
        return await this._getSecret(request, context);
    }
    async _listAcls(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("scope")) {
            query["scope"] = request.scope;
        }
        const path = "/api/2.0/secrets/acls/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists the ACLs set on the given scope.
     *
     * Users must have the ``MANAGE`` permission to invoke this API.
     *
     * Example response:
     *
     * .. code::
     *
     * { "acls": [{ "principal": "admins", "permission": "MANAGE" },{
     * "principal": "data-scientists", "permission": "READ" }] }
     *
     * Throws ``RESOURCE_DOES_NOT_EXIST`` if no such secret scope exists. Throws
     * ``PERMISSION_DENIED`` if the user does not have permission to make this
     * API call.
     */
    async *listAcls(request, context) {
        const response = (await this._listAcls(request, context)).items;
        for (const v of response || []) {
            yield v;
        }
    }
    async _listScopes(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/secrets/scopes/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists all secret scopes available in the workspace.
     *
     * Example response:
     *
     * .. code::
     *
     * { "scopes": [{ "name": "my-databricks-scope", "backend_type": "DATABRICKS"
     * },{ "name": "mount-points", "backend_type": "DATABRICKS" }] }
     *
     * Throws ``PERMISSION_DENIED`` if the user does not have permission to make
     * this API call.
     */
    async *listScopes(context) {
        const response = (await this._listScopes(context)).scopes;
        for (const v of response || []) {
            yield v;
        }
    }
    async _listSecrets(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("scope")) {
            query["scope"] = request.scope;
        }
        const path = "/api/2.0/secrets/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists the secret keys that are stored at this scope. This is a
     * metadata-only operation; secret data cannot be retrieved using this API.
     * Users need the READ permission to make this call.
     *
     * Example response:
     *
     * .. code::
     *
     * { "secrets": [ { "key": "my-string-key"", "last_updated_timestamp":
     * "1520467595000" }, { "key": "my-byte-key", "last_updated_timestamp":
     * "1520467595000" }, ] }
     *
     * The lastUpdatedTimestamp returned is in milliseconds since epoch.
     *
     * Throws ``RESOURCE_DOES_NOT_EXIST`` if no such secret scope exists. Throws
     * ``PERMISSION_DENIED`` if the user does not have permission to make this
     * API call.
     */
    async *listSecrets(request, context) {
        const response = (await this._listSecrets(request, context)).secrets;
        for (const v of response || []) {
            yield v;
        }
    }
    async _putAcl(request, context) {
        const headers = new Headers({ "Content-Type": "application/json" });
        const body = {};
        if (request.hasOwnProperty("permission")) {
            body["permission"] = request.permission;
        }
        if (request.hasOwnProperty("principal")) {
            body["principal"] = request.principal;
        }
        if (request.hasOwnProperty("scope")) {
            body["scope"] = request.scope;
        }
        const query = {};
        const path = "/api/2.0/secrets/acls/put";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates or overwrites the ACL associated with the given principal (user or
     * group) on the specified scope point. In general, a user or group will use
     * the most powerful permission available to them, and permissions are
     * ordered as follows:
     *
     * * ``MANAGE`` - Allowed to change ACLs, and read and write to this secret
     * scope. * ``WRITE`` - Allowed to read and write to this secret scope. *
     * ``READ`` - Allowed to read this secret scope and list what secrets are
     * available.
     *
     * Note that in general, secret values can only be read from within a command
     * on a cluster (for example, through a notebook). There is no API to read
     * the actual secret value material outside of a cluster. However, the user's
     * permission will be applied based on who is executing the command, and they
     * must have at least READ permission.
     *
     * Users must have the ``MANAGE`` permission to invoke this API.
     *
     * Example request:
     *
     * .. code::
     *
     * { "scope": "my-secret-scope", "principal": "data-scientists",
     * "permission": "READ" }
     *
     * The principal is a user or group name corresponding to an existing
     * Databricks principal to be granted or revoked access.
     *
     * Throws ``RESOURCE_DOES_NOT_EXIST`` if no such secret scope exists. Throws
     * ``RESOURCE_ALREADY_EXISTS`` if a permission for the principal already
     * exists. Throws ``INVALID_PARAMETER_VALUE`` if the permission or principal
     * is invalid. Throws ``PERMISSION_DENIED`` if the user does not have
     * permission to make this API call.
     */
    async putAcl(request, context) {
        return await this._putAcl(request, context);
    }
    async _putSecret(request, context) {
        const headers = new Headers({ "Content-Type": "application/json" });
        const body = {};
        if (request.hasOwnProperty("bytes_value")) {
            body["bytes_value"] = request.bytes_value;
        }
        if (request.hasOwnProperty("key")) {
            body["key"] = request.key;
        }
        if (request.hasOwnProperty("scope")) {
            body["scope"] = request.scope;
        }
        if (request.hasOwnProperty("string_value")) {
            body["string_value"] = request.string_value;
        }
        const query = {};
        const path = "/api/2.0/secrets/put";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Inserts a secret under the provided scope with the given name. If a secret
     * already exists with the same name, this command overwrites the existing
     * secret's value. The server encrypts the secret using the secret scope's
     * encryption settings before storing it. You must have ``WRITE`` or
     * ``MANAGE`` permission on the secret scope.
     *
     * The secret key must consist of alphanumeric characters, dashes,
     * underscores, and periods, and cannot exceed 128 characters. The maximum
     * allowed secret value size is 128 KB. The maximum number of secrets in a
     * given scope is 1000.
     *
     * Example request:
     *
     * .. code::
     *
     * { "scope": "my-databricks-scope", "key": "my-string-key", "string_value":
     * "foobar" }
     *
     * The input fields "string_value" or "bytes_value" specify the type of the
     * secret, which will determine the value returned when the secret value is
     * requested. Exactly one must be specified.
     *
     * Throws ``RESOURCE_DOES_NOT_EXIST`` if no such secret scope exists. Throws
     * ``RESOURCE_LIMIT_EXCEEDED`` if maximum number of secrets in scope is
     * exceeded. Throws ``INVALID_PARAMETER_VALUE`` if the request parameters are
     * invalid. Throws ``PERMISSION_DENIED`` if the user does not have permission
     * to make this API call. Throws ``MALFORMED_REQUEST`` if request is
     * incorrectly formatted or conflicting. Throws ``BAD_REQUEST`` if request is
     * made against Azure KeyVault backed scope.
     */
    async putSecret(request, context) {
        return await this._putSecret(request, context);
    }
}
exports.SecretsService = SecretsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "_createScope", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "createScope", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "_deleteAcl", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "deleteAcl", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "_deleteScope", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "deleteScope", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "_deleteSecret", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "deleteSecret", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "_getAcl", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "getAcl", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "_getSecret", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "getSecret", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "_listAcls", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], SecretsService.prototype, "listAcls", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "_listScopes", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Object)
], SecretsService.prototype, "listScopes", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "_listSecrets", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], SecretsService.prototype, "listSecrets", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "_putAcl", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "putAcl", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "_putSecret", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SecretsService.prototype, "putSecret", null);
class WorkspaceRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Workspace", method, message);
    }
}
exports.WorkspaceRetriableError = WorkspaceRetriableError;
class WorkspaceError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Workspace", method, message);
    }
}
exports.WorkspaceError = WorkspaceError;
/**
 * The Workspace API allows you to list, import, export, and delete notebooks and
 * folders.
 *
 * A notebook is a web-based interface to a document that contains runnable code,
 * visualizations, and explanatory text.
 */
class WorkspaceService {
    constructor(client) {
        this.client = client;
    }
    async _delete(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("path")) {
            body["path"] = request.path;
        }
        if (request.hasOwnProperty("recursive")) {
            body["recursive"] = request.recursive;
        }
        const query = {};
        const path = "/api/2.0/workspace/delete";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes an object or a directory (and optionally recursively deletes all
     * objects in the directory). * If `path` does not exist, this call returns
     * an error `RESOURCE_DOES_NOT_EXIST`. * If `path` is a non-empty directory
     * and `recursive` is set to `false`, this call returns an error
     * `DIRECTORY_NOT_EMPTY`.
     *
     * Object deletion cannot be undone and deleting a directory recursively is
     * not atomic.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _export(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("format")) {
            query["format"] = request.format;
        }
        if (request.hasOwnProperty("path")) {
            query["path"] = request.path;
        }
        const path = "/api/2.0/workspace/export";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Exports an object or the contents of an entire directory.
     *
     * If `path` does not exist, this call returns an error
     * `RESOURCE_DOES_NOT_EXIST`.
     *
     * If the exported data would exceed size limit, this call returns
     * `MAX_NOTEBOOK_SIZE_EXCEEDED`. Currently, this API does not support
     * exporting a library.
     */
    async export(request, context) {
        return await this._export(request, context);
    }
    async _getPermissionLevels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/${request.workspace_object_type}/${request.workspace_object_id}/permissionLevels`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(request, context) {
        return await this._getPermissionLevels(request, context);
    }
    async _getPermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/${request.workspace_object_type}/${request.workspace_object_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of a workspace object. Workspace objects can inherit
     * permissions from their parent objects or root object.
     */
    async getPermissions(request, context) {
        return await this._getPermissions(request, context);
    }
    async _getStatus(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("path")) {
            query["path"] = request.path;
        }
        const path = "/api/2.0/workspace/get-status";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the status of an object or a directory. If `path` does not exist,
     * this call returns an error `RESOURCE_DOES_NOT_EXIST`.
     */
    async getStatus(request, context) {
        return await this._getStatus(request, context);
    }
    async _import(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("content")) {
            body["content"] = request.content;
        }
        if (request.hasOwnProperty("format")) {
            body["format"] = request.format;
        }
        if (request.hasOwnProperty("language")) {
            body["language"] = request.language;
        }
        if (request.hasOwnProperty("overwrite")) {
            body["overwrite"] = request.overwrite;
        }
        if (request.hasOwnProperty("path")) {
            body["path"] = request.path;
        }
        const query = {};
        const path = "/api/2.0/workspace/import";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Imports a workspace object (for example, a notebook or file) or the
     * contents of an entire directory. If `path` already exists and `overwrite`
     * is set to `false`, this call returns an error `RESOURCE_ALREADY_EXISTS`.
     * To import a directory, you can use either the `DBC` format or the `SOURCE`
     * format with the `language` field unset. To import a single file as
     * `SOURCE`, you must set the `language` field. Zip files within directories
     * are not supported.
     */
    async import(request, context) {
        return await this._import(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("notebooks_modified_after")) {
            query["notebooks_modified_after"] =
                request.notebooks_modified_after;
        }
        if (request.hasOwnProperty("path")) {
            query["path"] = request.path;
        }
        const path = "/api/2.0/workspace/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists the contents of a directory, or the object if it is not a directory.
     * If the input path does not exist, this call returns an error
     * `RESOURCE_DOES_NOT_EXIST`.
     */
    async *list(request, context) {
        const response = (await this._list(request, context)).objects;
        for (const v of response || []) {
            yield v;
        }
    }
    async _mkdirs(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("path")) {
            body["path"] = request.path;
        }
        const query = {};
        const path = "/api/2.0/workspace/mkdirs";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates the specified directory (and necessary parent directories if they
     * do not exist). If there is an object (not a directory) at any prefix of
     * the input path, this call returns an error `RESOURCE_ALREADY_EXISTS`.
     *
     * Note that if this operation fails it may have succeeded in creating some
     * of the necessary parent directories.
     */
    async mkdirs(request, context) {
        return await this._mkdirs(request, context);
    }
    async _setPermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/${request.workspace_object_type}/${request.workspace_object_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their parent objects or root object.
     */
    async setPermissions(request, context) {
        return await this._setPermissions(request, context);
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/${request.workspace_object_type}/${request.workspace_object_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on a workspace object. Workspace objects can
     * inherit permissions from their parent objects or root object.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
}
exports.WorkspaceService = WorkspaceService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "_export", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "export", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "_getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "_getStatus", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "getStatus", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "_import", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "import", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], WorkspaceService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "_mkdirs", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "mkdirs", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "_setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceService.prototype, "updatePermissions", null);
