// @databricks/sdk-experimental/dist/config/AzureCliCredentials.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const AzureCliCredentials_1 = require("./AzureCliCredentials");
const Config_1 = require("./Config");
const assert = __importStar(require("node:assert"));
const path = __importStar(require("node:path"));
const f = __importStar(require("../fetch"));
describe(__filename, function () {
    this.timeout(20000);
    let envBackup;
    let fetchBackup;
    beforeEach(() => {
        envBackup = process.env;
        process.env = {};
        process.chdir(__dirname);
        process.env.PATH =
            path.join(__dirname, "testdata") + path.delimiter + envBackup.PATH;
        // mock management API
        fetchBackup = f.fetch;
        f.fetch = (url) => {
            if (!url.href.startsWith("https://management.azure.com")) {
                throw new Error(`unknown URL: ${url}`);
            }
            return {
                async json() {
                    return {
                        properties: {
                            workspaceUrl: "x",
                        },
                    };
                },
            };
        };
    });
    afterEach(() => {
        process.env = envBackup;
        f.fetch = fetchBackup;
    });
    it("should only set auth header when no managed resource id is set", async () => {
        const config = new Config_1.Config({
            host: "https://x.azuredatabricks.net",
        });
        const cliCreds = new AzureCliCredentials_1.AzureCliCredentials();
        const visitor = await cliCreds.configure(config);
        assert.ok(visitor);
        const headers = new f.Headers();
        await visitor(headers);
        assert.equal(headers.get("Authorization"), "Bearer ...");
        assert.equal(headers.get("X-Databricks-Azure-SP-Management-Token"), "...");
        assert.equal(headers.get("X-Databricks-Azure-Workspace-Resource-Id"), undefined);
    });
    it("should set auth header and managed resource id when managed resource id is set", async () => {
        const config = new Config_1.Config({
            azureWorkspaceResourceId: "/sub/rg/ws",
        });
        const cliCreds = new AzureCliCredentials_1.AzureCliCredentials();
        const visitor = await cliCreds.configure(config);
        assert.ok(visitor);
        const headers = new f.Headers();
        await visitor(headers);
        assert.equal(headers.get("Authorization"), "Bearer ...");
        assert.equal(headers.get("X-Databricks-Azure-SP-Management-Token"), "...");
        assert.equal(headers.get("X-Databricks-Azure-Workspace-Resource-Id"), "/sub/rg/ws");
    });
    it("should set fallback if subscription fails", async () => {
        process.env["FAIL_IF"] = "--subscription";
        const config = new Config_1.Config({
            azureWorkspaceResourceId: "/sub/rg/ws",
        });
        const cliCreds = new AzureCliCredentials_1.AzureCliCredentials();
        const visitor = await cliCreds.configure(config);
        assert.ok(visitor);
        const headers = new f.Headers();
        await visitor(headers);
        assert.equal(headers.get("Authorization"), "Bearer ...");
        assert.equal(headers.get("X-Databricks-Azure-SP-Management-Token"), "...");
        assert.equal(headers.get("X-Databricks-Azure-Workspace-Resource-Id"), "/sub/rg/ws");
    });
});
