// @databricks/sdk-experimental/dist/config/Azure.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAzureEnvironment = getAzureEnvironment;
exports.azureEnsureWorkspaceUrl = azureEnsureWorkspaceUrl;
exports.getAzureLoginAppId = getAzureLoginAppId;
exports.azureVisitor = azureVisitor;
exports.serviceToServiceVisitor = serviceToServiceVisitor;
exports.isConfigWithAzureClientSecret = isConfigWithAzureClientSecret;
exports.getOAuthClient = getOAuthClient;
const fetch_1 = require("../fetch");
const OidcEndpoints_1 = require("./oauth/OidcEndpoints");
// based on github.com/Azure/go-autorest/autorest/azure/azureEnvironments.go
const azureEnvironments = {
    PUBLIC: {
        name: "AzurePublicCloud",
        serviceManagementEndpoint: "https://management.core.windows.net/",
        resourceManagerEndpoint: "https://management.azure.com/",
        activeDirectoryEndpoint: "https://login.microsoftonline.com/",
    },
    USGOVERNMENT: {
        name: "AzureUSGovernmentCloud",
        serviceManagementEndpoint: "https://management.core.usgovcloudapi.net/",
        resourceManagerEndpoint: "https://management.usgovcloudapi.net/",
        activeDirectoryEndpoint: "https://login.microsoftonline.us/",
    },
    CHINA: {
        name: "AzureChinaCloud",
        serviceManagementEndpoint: "https://management.core.chinacloudapi.cn/",
        resourceManagerEndpoint: "https://management.chinacloudapi.cn/",
        activeDirectoryEndpoint: "https://login.chinacloudapi.cn/",
    },
    GERMAN: {
        name: "AzureGermanCloud",
        serviceManagementEndpoint: "https://management.core.cloudapi.de/",
        resourceManagerEndpoint: "https://management.microsoftazure.de/",
        activeDirectoryEndpoint: "https://login.microsoftonline.de/",
    },
};
function getAzureEnvironment(config) {
    if (!config.azureEnvironment) {
        config.azureEnvironment = "PUBLIC";
    }
    const env = azureEnvironments[config.azureEnvironment.toUpperCase()];
    if (!env) {
        throw new Error(`azure environment not found: ${config.azureEnvironment}`);
    }
    return env;
}
async function azureEnsureWorkspaceUrl(config, tokenSourceFor) {
    if (!config.azureWorkspaceResourceId && config.host) {
        return;
    }
    if (config.isAccountClient()) {
        return;
    }
    const env = getAzureEnvironment(config);
    const tokenSource = tokenSourceFor(env.resourceManagerEndpoint);
    const url = `${env.resourceManagerEndpoint.replace(/\/$/g, "")}${config.azureWorkspaceResourceId}?api-version=2018-04-01`;
    try {
        const token = await tokenSource();
        const response = await (0, fetch_1.fetch)(new URL(url), {
            headers: {
                Authorization: `Bearer ${token.accessToken}`,
            },
        });
        const body = (await response.json());
        if (body.properties?.workspaceUrl) {
            config.host = `https://${body.properties.workspaceUrl}`;
            config.logger.debug(`Discovered workspace url: ${config.host}`);
        }
    }
    catch (error) {
        throw new Error(`cannot resolve workspace: ${error.message}`);
    }
}
// Resource ID of the Azure application we need to log in.
const azureDatabricksLoginAppId = "2ff814a6-3304-4ab8-85cb-cd0e6f879c1d";
function getAzureLoginAppId(config) {
    if (config.azureLoginAppId) {
        return config.azureLoginAppId;
    }
    return azureDatabricksLoginAppId;
}
function azureVisitor(config, inner) {
    return async (headers) => {
        await inner(headers);
        if (config.azureWorkspaceResourceId) {
            headers.set("X-Databricks-Azure-Workspace-Resource-Id", config.azureWorkspaceResourceId);
        }
    };
}
function serviceToServiceVisitor(auth, secondary, secondaryHeader) {
    return async (headers) => {
        const innerToken = await auth();
        const cloudToken = await secondary();
        headers.set("Authorization", `Bearer ${innerToken?.accessToken}`);
        headers.set(secondaryHeader, cloudToken?.accessToken || "");
    };
}
function isConfigWithAzureClientSecret(config) {
    return (!!config.azureClientId &&
        !!config.azureClientSecret &&
        !!config.azureTenantId);
}
function getOAuthClient(config) {
    const env = getAzureEnvironment(config);
    const issuer = new OidcEndpoints_1.OidcEndpoints(config, new URL(`${env.activeDirectoryEndpoint}/${config.azureTenantId}/oauth2/authorize`), new URL(`${env.activeDirectoryEndpoint}/${config.azureTenantId}/oauth2/token`));
    const client = issuer.getClient({
        clientId: config.azureClientId,
        clientSecret: config.azureClientSecret,
        useHeader: true,
    });
    return client;
}
