// @databricks/sdk-experimental/dist/config/AzureMsiCredentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureMsiToken = exports.AzureMsiCredentials = void 0;
/* eslint-disable @typescript-eslint/naming-convention */
const Azure_1 = require("./Azure");
const Config_1 = require("./Config");
const Token_1 = require("./Token");
const fetch_1 = require("../fetch");
// well-known URL for Azure Instance Metadata Service (IMDS)
// https://learn.microsoft.com/en-us/azure-stack/user/instance-metadata-service
const instanceMetadataPrefix = "http://169.254.169.254/metadata";
// timeout to wait for IMDS response
const azureMsiTimeout = 10 * 1000;
class AzureMsiCredentials {
    constructor() {
        this.name = "azure-msi";
    }
    async configure(config) {
        if (!config.isAzure() ||
            !config.azureUseMsi ||
            (config.azureWorkspaceResourceId === "" &&
                !config.isAccountClient())) {
            return;
        }
        const env = (0, Azure_1.getAzureEnvironment)(config);
        if (!config.isAccountClient()) {
            await (0, Azure_1.azureEnsureWorkspaceUrl)(config, (resource) => {
                return this.getTokenSource(config, resource);
            });
        }
        config.logger.debug(`Generating AAD token via Azure MSI`);
        return (0, Azure_1.azureVisitor)(config, (0, Azure_1.serviceToServiceVisitor)((0, Token_1.refreshableTokenSource)(this.getTokenSource(config, (0, Azure_1.getAzureLoginAppId)(config))), (0, Token_1.refreshableTokenSource)(this.getTokenSource(config, env.serviceManagementEndpoint)), "X-Databricks-Azure-SP-Management-Token"));
    }
    getTokenSource(config, resource) {
        return async () => {
            const inner = await new AzureMsiTokenSource(config, resource, config.azureClientId).token();
            return inner;
        };
    }
}
exports.AzureMsiCredentials = AzureMsiCredentials;
class AzureMsiTokenSource {
    constructor(config, resource, clientId) {
        this.config = config;
        this.resource = resource;
        this.clientId = clientId;
    }
    async token() {
        const url = new URL(`${instanceMetadataPrefix}/identity/oauth2/token`);
        url.searchParams.append("api-version", "2018-02-01");
        url.searchParams.append("resource", this.resource);
        if (this.clientId) {
            url.searchParams.append("client_id", this.clientId);
        }
        const abortController = new fetch_1.AbortController();
        setTimeout(() => {
            abortController.abort();
        }, azureMsiTimeout);
        const response = await (0, fetch_1.fetch)(url, {
            headers: {
                Metadata: "true",
            },
            signal: abortController.signal,
        });
        if (!response.ok) {
            throw new Config_1.ConfigError(`azure-msi: failed to get token: ${await response.text()}`, this.config);
        }
        const responseJson = await response.json();
        const inner = AzureMsiToken.fromJson(this.config, responseJson);
        return inner.getOauthToken();
    }
}
class AzureMsiToken {
    constructor(config, tokenType, accessToken, expiresIn, refreshToken) {
        this.config = config;
        this.tokenType = tokenType;
        this.accessToken = accessToken;
        this.expiresIn = expiresIn;
        this.refreshToken = refreshToken;
    }
    static fromJson(config, json) {
        if (typeof json !== "object" || json === null) {
            throw new Config_1.ConfigError("azure-msi: invalid token", config);
        }
        if (!("token_type" in json) ||
            typeof json.token_type !== "string" ||
            !("access_token" in json) ||
            typeof json.access_token !== "string" ||
            ("refresh_token" in json &&
                typeof json.refresh_token !== "string") ||
            !("expires_in" in json) ||
            typeof json.expires_in !== "string") {
            throw new Config_1.ConfigError("azure-msi: invalid token", config);
        }
        return new AzureMsiToken(config, json.token_type, json.access_token, parseInt(json.expires_in), json.refresh_token);
    }
    getOauthToken() {
        return new Token_1.Token({
            accessToken: this.accessToken,
            refreshToken: this.refreshToken,
            expiry: Date.now() + this.expiresIn * 1000,
        });
    }
}
exports.AzureMsiToken = AzureMsiToken;
