// @databricks/sdk-experimental/dist/config/GoogleDefaultCredentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleDefaultCredentials = exports.gcpScopes = void 0;
const Config_1 = require("./Config");
const promises_1 = require("fs/promises");
const google_auth_library_1 = require("google-auth-library");
exports.gcpScopes = [
    "https://www.googleapis.com/auth/cloud-platform",
    "https://www.googleapis.com/auth/compute",
];
class GoogleDefaultCredentials {
    constructor() {
        this.name = "google-credentials";
    }
    async configure(config) {
        if (!config.googleCredentials || !config.isGcp() || !config.host) {
            return;
        }
        const host = config.host;
        let json;
        try {
            json = await this.readCredentials(config.googleCredentials);
        }
        catch (err) {
            throw new Config_1.ConfigError(`could not read GoogleCredentials. Make sure the file exists, or the JSON content is valid: ${err.message}`, config);
        }
        const tokenClient = await new google_auth_library_1.GoogleAuth({
            credentials: json,
        }).getIdTokenClient(host);
        const gcpClient = await new google_auth_library_1.GoogleAuth({
            credentials: json,
            scopes: exports.gcpScopes,
        }).getClient();
        const isAccount = config.isAccountClient();
        return async (headers) => {
            headers.set("authorization", `Bearer ${await tokenClient.idTokenProvider.fetchIdToken(host)}`);
            if (isAccount) {
                const gcpToken = await gcpClient.getAccessToken();
                if (!gcpToken.token) {
                    throw new Config_1.ConfigError("failed to get GCP access token", config);
                }
                headers.set("x-databricks-gcp-sa-access-token", gcpToken.token);
            }
        };
    }
    /**
     * Reads credentials as JSON. Credentials can be either a path to JSON file,
     * or actual JSON string.
     */
    async readCredentials(credentialsOrPath) {
        try {
            await (0, promises_1.stat)(credentialsOrPath);
        }
        catch (_e) {
            return JSON.parse(credentialsOrPath);
        }
        return JSON.parse(await (0, promises_1.readFile)(credentialsOrPath, "utf8"));
    }
}
exports.GoogleDefaultCredentials = GoogleDefaultCredentials;
