// @databricks/sdk-experimental/dist/config/HostType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigType = exports.HostType = void 0;
/**
 * HostType represents the type of Databricks host.
 * This determines what APIs the host supports.
 */
var HostType;
(function (HostType) {
    /** Traditional workspace-only endpoint (e.g., dbc-12345.cloud.databricks.com) */
    HostType["workspaceHost"] = "workspaceHost";
    /** Traditional account-only endpoint (e.g., accounts.cloud.databricks.com) */
    HostType["accountHost"] = "accountHost";
    /** Unified endpoint supporting both workspace and account APIs (e.g., company.databricks.com) */
    HostType["unifiedHost"] = "unifiedHost";
})(HostType || (exports.HostType = HostType = {}));
/**
 * ConfigType represents what the client is configured for.
 * This is independent of what the host supports.
 */
var ConfigType;
(function (ConfigType) {
    /** Client is configured for workspace-level operations */
    ConfigType["workspaceConfig"] = "workspaceConfig";
    /** Client is configured for account-level operations */
    ConfigType["accountConfig"] = "accountConfig";
    /** Client configuration is invalid (e.g., unified host without accountId) */
    ConfigType["invalidConfig"] = "invalidConfig";
})(ConfigType || (exports.ConfigType = ConfigType = {}));
