// @databricks/sdk-experimental/dist/config/KnownConfigLoader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KnownConfigLoader = void 0;
exports.loadConfigFile = loadConfigFile;
exports.flattenIniObject = flattenIniObject;
const node_path_1 = __importDefault(require("node:path"));
const node_os_1 = __importDefault(require("node:os"));
const promises_1 = __importDefault(require("node:fs/promises"));
const Config_1 = require("./Config");
const ini_1 = require("ini");
/**
 * Loads configuration from the Databricks config file which is by default
 * in `~/.databrickscfg`.
 */
class KnownConfigLoader {
    constructor() {
        this.name = "config-file";
    }
    async configure(cfg) {
        // Skip loading config file if some authentication is already explicitly
        // configured directly in the config by a user.
        // See: https://github.com/databricks/databricks-sdk-go/issues/304
        // Also skip if host or azureResourceId is configured, because these do
        // not explcitly specify an auth but they can completely define some auths such as
        // databricks-cli (oauth u2m).
        if (cfg.profile === undefined &&
            ((0, Config_1.isAnyAuthConfigured)(cfg) ||
                cfg.host !== undefined ||
                cfg.azureWorkspaceResourceId !== undefined)) {
            return;
        }
        let configPath, iniFile;
        try {
            ({ path: configPath, iniFile } = await loadConfigFile(cfg.configFile));
        }
        catch (e) {
            if (e instanceof Error) {
                cfg.logger.debug(e.message);
            }
            return;
        }
        const profile = cfg.profile || "DEFAULT";
        const hasExplicitProfile = cfg.profile !== undefined;
        if (!iniFile[profile]) {
            if (!hasExplicitProfile) {
                cfg.logger.debug(`resolve: ${configPath} has no ${profile} profile configured`);
                return;
            }
            throw new Config_1.ConfigError(`resolve: ${configPath} has no ${profile} profile configured`, cfg);
        }
        cfg.logger.info(`loading ${profile} profile from ${configPath}`);
        cfg.attributes.resolveFromStringMap(iniFile[profile]);
    }
}
exports.KnownConfigLoader = KnownConfigLoader;
async function loadConfigFile(configFile) {
    configFile = configFile || "~/.databrickscfg";
    const configPath = node_path_1.default.resolve(configFile.replace(/^~/, process.env.HOME || node_os_1.default.homedir()));
    try {
        await promises_1.default.stat(configPath);
    }
    catch (_e) {
        throw new Error(`${configPath} does not exist`);
    }
    return {
        iniFile: flattenIniObject((0, ini_1.parse)(await promises_1.default.readFile(configPath, { encoding: "utf-8" }))),
        path: configPath,
    };
}
/**
 * The ini library encodes dots as nested objects, which is not what we want.
 * See: https://github.com/npm/ini/issues/22
 *
 * Reduce the nested object back to a flat object by concatenating object keys with a `.`
 */
function flattenIniObject(obj) {
    function _flattenIniObject(obj, topLevel = [], resp = {}) {
        const props = {};
        for (const key in obj) {
            if (typeof obj[key] === "object") {
                topLevel.push(key);
                resp = {
                    ...resp,
                    ..._flattenIniObject(obj[key], topLevel, resp),
                };
            }
            else {
                props[key] = obj[key];
            }
        }
        const topLevelName = topLevel.join(".");
        if (topLevelName !== "" && Object.keys(props).length > 0) {
            resp = { ...resp, [topLevelName]: props };
        }
        topLevel.pop();
        return resp;
    }
    return _flattenIniObject(obj);
}
