// @databricks/sdk-experimental/dist/config/MetadataServiceCredentials.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const http = __importStar(require("http"));
const node_assert_1 = __importDefault(require("node:assert"));
const Config_1 = require("./Config");
const fetch_1 = require("../fetch");
const MetadataServiceCredentials_1 = require("./MetadataServiceCredentials");
describe(__filename, () => {
    let server;
    let address;
    let sockets = {};
    beforeEach(async () => {
        let i = 0;
        server = http.createServer((req, res) => {
            if (req.headers[MetadataServiceCredentials_1.MetadataServiceVersionHeader] !==
                MetadataServiceCredentials_1.MetadataServiceVersion) {
                res.writeHead(400);
                res.end();
                return;
            }
            if (req.headers[MetadataServiceCredentials_1.MetadataServiceHostHeader] !== "https://test.com") {
                res.writeHead(404);
                res.end();
                return;
            }
            const response = {
                access_token: `XXXX-${i++}`,
                expires_on: Math.floor(new Date(Date.now() + 1000).getTime() / 1000),
                token_type: "Bearer",
            };
            // res.writeHead(200);
            res.end(JSON.stringify(response));
        });
        let nextSocketId = 0;
        server.on("connection", function (socket) {
            const socketId = nextSocketId++;
            sockets[socketId] = socket;
            socket.on("close", function () {
                // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
                delete sockets[socketId];
            });
        });
        await new Promise((resolve, reject) => {
            server.listen(0, "127.0.0.1", () => {
                resolve(null);
            });
            server.on("error", (err) => {
                reject(err);
            });
        });
        address = server.address();
    });
    afterEach(async () => {
        await new Promise((resolve) => {
            server.close(() => {
                resolve(null);
            });
            for (const socketId of Object.keys(sockets)) {
                sockets[Number(socketId)].destroy();
            }
            sockets = {};
        });
    });
    it("should authorize using local server", async () => {
        const lmsCredentials = new MetadataServiceCredentials_1.MetadataServiceCredentials();
        const config = new Config_1.Config({
            host: "https://test.com",
            authType: "metadata-service",
            metadataServiceUrl: `http://${address.address}:${address.port}`,
        });
        const visitor = await lmsCredentials.configure(config);
        node_assert_1.default.ok(visitor);
        const headers = new fetch_1.Headers();
        await visitor(headers);
        node_assert_1.default.equal(headers.get("Authorization"), "Bearer XXXX-1");
    });
    it("should check host", async () => {
        const lmsCredentials = new MetadataServiceCredentials_1.MetadataServiceCredentials();
        const config = new Config_1.Config({
            host: "https://test2.com",
            authType: "metadata-service",
            metadataServiceUrl: `http://${address.address}:${address.port}`,
        });
        const visitor = await lmsCredentials.configure(config);
        node_assert_1.default.ok(!visitor);
    });
    it("should refresh credentials", async () => {
        const lmsCredentials = new MetadataServiceCredentials_1.MetadataServiceCredentials();
        const config = new Config_1.Config({
            host: "https://test.com",
            authType: "metadata-service",
            metadataServiceUrl: `http://${address.address}:${address.port}`,
        });
        const visitor = await lmsCredentials.configure(config);
        node_assert_1.default.ok(visitor);
        const headers = new fetch_1.Headers();
        await visitor(headers);
        node_assert_1.default.equal(headers.get("Authorization"), "Bearer XXXX-1");
        // expires immediately since expiery is lower than expiryDelta
        await visitor(headers);
        node_assert_1.default.equal(headers.get("Authorization"), "Bearer XXXX-2");
    });
});
