// @databricks/sdk-experimental/dist/config/UnifiedAuth.test.js
"use strict";
/**
 * These are auto-generated tests for Unified Authentication
 * In case of editing this file, make sure the change is propagated to all Databricks SDK codebases
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_assert_1 = __importDefault(require("node:assert"));
const Config_1 = require("./Config");
const node_path_1 = __importDefault(require("node:path"));
const os = __importStar(require("node:os"));
const f = __importStar(require("../fetch"));
describe(__dirname, function () {
    let envBackup;
    let fetchBackup;
    let cwdBackup = "";
    const debug = false;
    this.timeout(10000);
    beforeEach(() => {
        // mock management API
        fetchBackup = f.fetch;
        f.fetch = (url) => {
            if (!url.href.startsWith("https://management.azure.com")) {
                throw new Error(`unknown URL: ${url}`);
            }
            return {
                async json() {
                    return {
                        properties: {
                            workspaceUrl: "x",
                        },
                    };
                },
            };
        };
        envBackup = process.env;
        process.env = {};
        cwdBackup = process.cwd();
        process.chdir(__dirname);
    });
    afterEach(() => {
        process.env = envBackup;
        process.chdir(cwdBackup);
        f.fetch = fetchBackup;
    });
    it("TestConfig_NoParams", async () => {
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
            // Set configFile to non-existent path to prevent loading from user's ~/.databrickscfg
            configFile: "/nonexistent/.databrickscfg",
        });
        await expectError(config, "default auth: cannot configure default credentials, please check https://docs.databricks.com/en/dev-tools/auth.html#databricks-client-unified-authentication to configure credentials for your preferred authentication method. Config: config_file=/nonexistent/.databrickscfg");
    });
    it("TestConfig_HostEnv", async () => {
        setEnv("DATABRICKS_HOST", "x");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await expectError(config, "default auth: cannot configure default credentials, please check https://docs.databricks.com/en/dev-tools/auth.html#databricks-client-unified-authentication to configure credentials for your preferred authentication method. Config: host=https://x. Env: DATABRICKS_HOST");
    });
    it("TestConfig_TokenEnv", async () => {
        setEnv("DATABRICKS_TOKEN", "x");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await expectError(config, "default auth: cannot configure default credentials, please check https://docs.databricks.com/en/dev-tools/auth.html#databricks-client-unified-authentication to configure credentials for your preferred authentication method. Config: token=***. Env: DATABRICKS_TOKEN");
    });
    it("TestConfig_HostTokenEnv", async () => {
        setEnv("DATABRICKS_HOST", "x");
        setEnv("DATABRICKS_TOKEN", "x");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "pat");
        node_assert_1.default.equal(config.host, "https://x".toLowerCase());
    });
    it("TestConfig_HostParamTokenEnv", async () => {
        setEnv("DATABRICKS_TOKEN", "x");
        const config = new Config_1.Config({
            host: "https://x",
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "pat");
        node_assert_1.default.equal(config.host, "https://x".toLowerCase());
    });
    it("TestConfig_UserPasswordEnv", async () => {
        setEnv("DATABRICKS_PASSWORD", "x");
        setEnv("DATABRICKS_USERNAME", "x");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await expectError(config, "default auth: cannot configure default credentials, please check https://docs.databricks.com/en/dev-tools/auth.html#databricks-client-unified-authentication to configure credentials for your preferred authentication method. Config: username=x, password=***. Env: DATABRICKS_USERNAME, DATABRICKS_PASSWORD");
    });
    it("TestConfig_BasicAuth", async () => {
        setEnv("DATABRICKS_HOST", "x");
        setEnv("DATABRICKS_PASSWORD", "x");
        setEnv("DATABRICKS_USERNAME", "x");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "basic");
        node_assert_1.default.equal(config.host, "https://x".toLowerCase());
    });
    it("TestConfig_AttributePrecedence", async () => {
        setEnv("DATABRICKS_HOST", "x");
        setEnv("DATABRICKS_PASSWORD", "x");
        setEnv("DATABRICKS_USERNAME", "x");
        const config = new Config_1.Config({
            host: "y",
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "basic");
        node_assert_1.default.equal(config.host, "https://y".toLowerCase());
    });
    it("TestConfig_BasicAuth_Mix", async () => {
        setEnv("DATABRICKS_PASSWORD", "x");
        const config = new Config_1.Config({
            host: "y",
            username: "x",
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "basic");
        node_assert_1.default.equal(config.host, "https://y".toLowerCase());
    });
    it("TestConfig_BasicAuth_Attrs", async () => {
        const config = new Config_1.Config({
            host: "y",
            username: "x",
            password: "x",
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "basic");
        node_assert_1.default.equal(config.host, "https://y".toLowerCase());
    });
    it("TestConfig_ConflictingEnvs", async () => {
        setEnv("DATABRICKS_HOST", "x");
        setEnv("DATABRICKS_PASSWORD", "x");
        setEnv("DATABRICKS_TOKEN", "x");
        setEnv("DATABRICKS_USERNAME", "x");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await expectError(config, "validate: more than one authorization method configured: basic and pat. Config: host=x, token=***, username=x, password=***. Env: DATABRICKS_HOST, DATABRICKS_TOKEN, DATABRICKS_USERNAME, DATABRICKS_PASSWORD");
    });
    it("TestConfig_ConflictingEnvs_AuthType", async () => {
        setEnv("DATABRICKS_HOST", "x");
        setEnv("DATABRICKS_PASSWORD", "x");
        setEnv("DATABRICKS_TOKEN", "x");
        setEnv("DATABRICKS_USERNAME", "x");
        const config = new Config_1.Config({
            authType: "basic",
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "basic");
        node_assert_1.default.equal(config.host, "https://x".toLowerCase());
    });
    it("TestConfig_ConfigFile", async () => {
        setEnv("DATABRICKS_CONFIG_FILE", "x");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await expectError(config, "default auth: cannot configure default credentials, please check https://docs.databricks.com/en/dev-tools/auth.html#databricks-client-unified-authentication to configure credentials for your preferred authentication method. Config: config_file=x. Env: DATABRICKS_CONFIG_FILE");
    });
    it("TestConfig_ConfigFileSkipDefaultProfileIfHostSpecified", async () => {
        setEnv("HOME", "testdata");
        const config = new Config_1.Config({
            host: "x",
            logger: debug ? console : undefined,
        });
        await expectError(config, "default auth: cannot configure default credentials, please check https://docs.databricks.com/en/dev-tools/auth.html#databricks-client-unified-authentication to configure credentials for your preferred authentication method. Config: host=https://x");
    });
    it("TestConfig_ConfigFileWithEmptyDefaultProfileSelectDefault", async () => {
        setEnv("HOME", "testdata/empty_default");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await expectError(config, "default auth: cannot configure default credentials, please check https://docs.databricks.com/en/dev-tools/auth.html#databricks-client-unified-authentication to configure credentials for your preferred authentication method");
    });
    it("TestConfig_ConfigFileWithEmptyDefaultProfileSelectAbc", async () => {
        setEnv("DATABRICKS_CONFIG_PROFILE", "abc");
        setEnv("HOME", "testdata/empty_default");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "pat");
        node_assert_1.default.equal(config.host, "https://foo".toLowerCase());
    });
    it("TestConfig_PatFromDatabricksCfg", async () => {
        setEnv("HOME", "testdata");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "pat");
        node_assert_1.default.equal(config.host, "https://dbc-XXXXXXXX-YYYY.cloud.databricks.com".toLowerCase());
    });
    it("TestConfig_PatFromDatabricksCfg_DotProfile", async () => {
        setEnv("DATABRICKS_CONFIG_PROFILE", "pat.with.dot");
        setEnv("HOME", "testdata");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "pat");
        node_assert_1.default.equal(config.host, "https://dbc-XXXXXXXX-YYYY.cloud.databricks.com".toLowerCase());
    });
    it("TestConfig_PatFromDatabricksCfg_NohostProfile", async () => {
        setEnv("DATABRICKS_CONFIG_PROFILE", "nohost");
        setEnv("HOME", "testdata");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await expectError(config, "default auth: cannot configure default credentials, please check https://docs.databricks.com/en/dev-tools/auth.html#databricks-client-unified-authentication to configure credentials for your preferred authentication method. Config: token=***, profile=nohost. Env: DATABRICKS_CONFIG_PROFILE");
    });
    it("TestConfig_ConfigProfileAndToken", async () => {
        setEnv("DATABRICKS_CONFIG_PROFILE", "nohost");
        setEnv("DATABRICKS_TOKEN", "x");
        setEnv("HOME", "testdata");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await expectError(config, "default auth: cannot configure default credentials, please check https://docs.databricks.com/en/dev-tools/auth.html#databricks-client-unified-authentication to configure credentials for your preferred authentication method. Config: token=***, profile=nohost. Env: DATABRICKS_TOKEN, DATABRICKS_CONFIG_PROFILE");
    });
    it("TestConfig_ConfigProfileAndPassword", async () => {
        setEnv("DATABRICKS_CONFIG_PROFILE", "nohost");
        setEnv("DATABRICKS_USERNAME", "x");
        setEnv("HOME", "testdata");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await expectError(config, "validate: more than one authorization method configured: basic and pat. Config: token=***, username=x, profile=nohost. Env: DATABRICKS_USERNAME, DATABRICKS_CONFIG_PROFILE");
    });
    it("TestConfig_AzurePAT", async () => {
        const config = new Config_1.Config({
            host: "https://adb-xxx.y.azuredatabricks.net/",
            token: "y",
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "pat");
        node_assert_1.default.equal(config.host, "https://adb-xxx.y.azuredatabricks.net".toLowerCase());
        node_assert_1.default.ok(config.isAzure());
    });
    it("TestConfig_AzureCliHost", async () => {
        setEnv("HOME", "testdata/azure");
        setEnv("PATH", "testdata:/bin");
        const config = new Config_1.Config({
            host: "x",
            azureWorkspaceResourceId: "/sub/rg/ws",
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "azure-cli");
        node_assert_1.default.equal(config.host, "https://x".toLowerCase());
        node_assert_1.default.ok(config.isAzure());
    });
    it("TestConfig_AzureCliHost_Fail", async () => {
        setEnv("FAIL", "yes");
        setEnv("HOME", "testdata/azure");
        setEnv("PATH", "testdata:/bin");
        const config = new Config_1.Config({
            azureWorkspaceResourceId: "/sub/rg/ws",
            logger: debug ? console : undefined,
        });
        await expectError(config, "default auth: azure-cli: cannot get access token: This is just a failing script.\n. Config: azure_workspace_resource_id=/sub/rg/ws");
    });
    it("TestConfig_AzureCliHost_AzNotInstalled", async () => {
        setEnv("HOME", "testdata/azure");
        setEnv("PATH", "whatever");
        const config = new Config_1.Config({
            azureWorkspaceResourceId: "/sub/rg/ws",
            logger: debug ? console : undefined,
        });
        await expectError(config, "default auth: cannot configure default credentials, please check https://docs.databricks.com/en/dev-tools/auth.html#databricks-client-unified-authentication to configure credentials for your preferred authentication method. Config: azure_workspace_resource_id=/sub/rg/ws");
    });
    it("TestConfig_AzureCliHost_PatConflict_WithConfigFilePresentWithoutDefaultProfile", async () => {
        setEnv("HOME", "testdata/azure");
        setEnv("PATH", "testdata:/bin");
        const config = new Config_1.Config({
            token: "x",
            azureWorkspaceResourceId: "/sub/rg/ws",
            logger: debug ? console : undefined,
        });
        await expectError(config, "validate: more than one authorization method configured: azure and pat. Config: token=***, azure_workspace_resource_id=/sub/rg/ws");
    });
    it("TestConfig_AzureCliHostAndResourceID", async () => {
        setEnv("HOME", "testdata");
        setEnv("PATH", "testdata:/bin");
        const config = new Config_1.Config({
            host: "x",
            azureWorkspaceResourceId: "/sub/rg/ws",
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "azure-cli");
        node_assert_1.default.equal(config.host, "https://x".toLowerCase());
        node_assert_1.default.ok(config.isAzure());
    });
    it("TestConfig_AzureCliHostAndResourceID_ConfigurationPrecedence", async () => {
        setEnv("DATABRICKS_CONFIG_PROFILE", "justhost");
        setEnv("HOME", "testdata/azure");
        setEnv("PATH", "testdata:/bin");
        const config = new Config_1.Config({
            host: "x",
            azureWorkspaceResourceId: "/sub/rg/ws",
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "azure-cli");
        node_assert_1.default.equal(config.host, "https://x".toLowerCase());
        node_assert_1.default.ok(config.isAzure());
    });
    it("TestConfig_AzureAndPasswordConflict", async () => {
        setEnv("DATABRICKS_USERNAME", "x");
        setEnv("HOME", "testdata/azure");
        setEnv("PATH", "testdata:/bin");
        const config = new Config_1.Config({
            host: "x",
            azureWorkspaceResourceId: "/sub/rg/ws",
            logger: debug ? console : undefined,
        });
        await expectError(config, "validate: more than one authorization method configured: azure and basic. Config: host=x, username=x, azure_workspace_resource_id=/sub/rg/ws. Env: DATABRICKS_USERNAME");
    });
    it("TestConfig_CorruptConfig", async () => {
        setEnv("DATABRICKS_CONFIG_PROFILE", "DEFAULT");
        setEnv("HOME", "testdata/corrupt");
        const config = new Config_1.Config({
            logger: debug ? console : undefined,
        });
        await expectError(config, "resolve: testdata/corrupt/.databrickscfg has no DEFAULT profile configured. Config: profile=DEFAULT. Env: DATABRICKS_CONFIG_PROFILE");
    });
    it("TestConfig_AuthTypeFromEnv", async () => {
        setEnv("DATABRICKS_AUTH_TYPE", "basic");
        setEnv("DATABRICKS_PASSWORD", "password");
        setEnv("DATABRICKS_TOKEN", "token");
        setEnv("DATABRICKS_USERNAME", "user");
        const config = new Config_1.Config({
            host: "x",
            logger: debug ? console : undefined,
        });
        await config.authenticate(new f.Headers());
        node_assert_1.default.equal(config.authType, "basic");
        node_assert_1.default.equal(config.host, "https://x".toLowerCase());
    });
    async function expectError(config, errorMessage) {
        let hadError = false;
        try {
            await config.authenticate(new f.Headers());
        }
        catch (error) {
            hadError = true;
            node_assert_1.default.equal(error.message
                .replace(/\r?\n/g, "\n") // normalize line endings between windows and unix
                .replace(__dirname + node_path_1.default.sep, "") // make paths relative
                .replace(/\\/g, "/"), // normalize path separators
            errorMessage);
        }
        node_assert_1.default.ok(hadError, "expected code to throw an error");
    }
    function setEnv(key, value) {
        if (key === "HOME") {
            process.env[key] = __dirname + "/" + value;
        }
        else if (key === "PATH") {
            const origPath = envBackup["PATH"] || "";
            let parts = value.split(":");
            if (parts[1] === "/bin") {
                parts = [parts[0], origPath];
            }
            if (os.platform() === "win32") {
                parts.push("c:\\windows\\system32");
            }
            process.env[key] = parts.join(node_path_1.default.delimiter);
        }
        else {
            process.env[key] = value;
        }
    }
});
