// @databricks/appkit/dist/plugins/genie/genie.js
import { createLogger } from "../../logging/logger.js";
import { getWorkspaceClient } from "../../context/execution-context.js";
import "../../context/index.js";
import { Plugin } from "../../plugin/plugin.js";
import { toPlugin } from "../../plugin/to-plugin.js";
import "../../plugin/index.js";
import "../../logging/index.js";
import { GenieConnector } from "../../connectors/genie/client.js";
import "../../connectors/index.js";
import { buildToolkitEntries } from "../../core/agent/build-toolkit.js";
import { defineTool, executeFromRegistry, toolsFromRegistry } from "../../core/agent/tools/define-tool.js";
import { genieStreamDefaults } from "./defaults.js";
import manifest_default from "./manifest.js";
import { randomUUID } from "node:crypto";
import { z } from "zod";

//#region src/plugins/genie/genie.ts
const logger = createLogger("genie");
var GeniePlugin = class extends Plugin {
	static manifest = manifest_default;
	static description = "AI/BI Genie space integration for natural language data queries";
	genieConnector;
	tools = {};
	constructor(config) {
		super(config);
		this.config = {
			...config,
			spaces: config.spaces ?? this.defaultSpaces()
		};
		this.genieConnector = new GenieConnector({
			timeout: this.config.timeout,
			maxMessages: 200
		});
		const spaces = this.config.spaces ?? {};
		const missingAliases = Object.entries(spaces).filter(([, id]) => !id).map(([alias]) => alias);
		if (missingAliases.length > 0) {
			const plural = missingAliases.length > 1;
			throw new Error(`GeniePlugin: space ${plural ? "aliases" : "alias"} ${missingAliases.map((a) => `"${a}"`).join(", ")} ${plural ? "were" : "was"} configured with a missing Genie Space ID. This usually means an environment variable used to populate the config is unset. Set the env var, or remove the alias from the config.`);
		}
		for (const alias of Object.keys(spaces)) Object.assign(this.tools, this._defineSpaceTools(alias));
	}
	/**
	* Builds the registry entries for a single Genie space alias.
	* One set of tools per configured space, keyed by `${alias}.${method}`.
	*/
	_defineSpaceTools(alias) {
		return {
			[`${alias}.sendMessage`]: defineTool({
				description: `Send a natural language question to the Genie space "${alias}" and get data analysis results`,
				schema: z.object({
					content: z.string().describe("The natural language question to ask"),
					conversationId: z.string().optional().describe("Optional conversation ID to continue an existing conversation")
				}),
				annotations: {
					effect: "read",
					requiresUserContext: true
				},
				execute: async (args, signal) => {
					const events = [];
					for await (const event of this.sendMessage(alias, args.content, args.conversationId, { signal })) events.push(event);
					return events;
				}
			}),
			[`${alias}.getConversation`]: defineTool({
				description: `Retrieve the conversation history from the Genie space "${alias}"`,
				schema: z.object({ conversationId: z.string().describe("The conversation ID to retrieve") }),
				annotations: {
					effect: "read",
					requiresUserContext: true
				},
				autoInheritable: true,
				execute: (args, signal) => this.getConversation(alias, args.conversationId, signal)
			})
		};
	}
	defaultSpaces() {
		const spaceId = process.env.DATABRICKS_GENIE_SPACE_ID;
		return spaceId ? { default: spaceId } : {};
	}
	resolveSpaceId(alias) {
		return this.config.spaces?.[alias] ?? null;
	}
	injectRoutes(router) {
		this.route(router, {
			name: "sendMessage",
			method: "post",
			path: "/:alias/messages",
			handler: async (req, res) => {
				await this.asUser(req)._handleSendMessage(req, res);
			}
		});
		this.route(router, {
			name: "getConversation",
			method: "get",
			path: "/:alias/conversations/:conversationId",
			handler: async (req, res) => {
				await this.asUser(req)._handleGetConversation(req, res);
			}
		});
		this.route(router, {
			name: "getMessage",
			method: "get",
			path: "/:alias/conversations/:conversationId/messages/:messageId",
			handler: async (req, res) => {
				await this.asUser(req)._handleGetMessage(req, res);
			}
		});
	}
	async _handleSendMessage(req, res) {
		const { alias } = req.params;
		const spaceId = this.resolveSpaceId(alias);
		if (!spaceId) {
			res.status(404).json({ error: `Unknown space alias: ${alias}` });
			return;
		}
		const { content, conversationId } = req.body;
		if (!content) {
			res.status(400).json({ error: "content is required" });
			return;
		}
		logger.debug("Sending message to space %s (alias=%s, conversationId=%s)", spaceId, alias, conversationId ?? "new");
		const timeout = this.config.timeout ?? 12e4;
		const requestId = typeof req.query.requestId === "string" && req.query.requestId || randomUUID();
		const streamSettings = {
			...genieStreamDefaults,
			default: {
				...genieStreamDefaults.default,
				timeout
			},
			stream: {
				...genieStreamDefaults.stream,
				streamId: requestId
			}
		};
		const workspaceClient = getWorkspaceClient();
		await this.executeStream(res, (signal) => this.genieConnector.streamSendMessage(workspaceClient, spaceId, content, conversationId, {
			timeout,
			signal
		}), streamSettings);
	}
	async _handleGetConversation(req, res) {
		const { alias, conversationId } = req.params;
		const spaceId = this.resolveSpaceId(alias);
		if (!spaceId) {
			res.status(404).json({ error: `Unknown space alias: ${alias}` });
			return;
		}
		const includeQueryResults = req.query.includeQueryResults !== "false";
		const pageToken = typeof req.query.pageToken === "string" ? req.query.pageToken : void 0;
		const requestId = typeof req.query.requestId === "string" && req.query.requestId || randomUUID();
		logger.debug("Fetching conversation %s from space %s (alias=%s, includeQueryResults=%s, pageToken=%s)", conversationId, spaceId, alias, includeQueryResults, pageToken ?? "none");
		const streamSettings = {
			...genieStreamDefaults,
			stream: {
				...genieStreamDefaults.stream,
				streamId: requestId
			}
		};
		const workspaceClient = getWorkspaceClient();
		await this.executeStream(res, (signal) => this.genieConnector.streamConversation(workspaceClient, spaceId, conversationId, {
			includeQueryResults,
			pageToken,
			signal
		}), streamSettings);
	}
	async _handleGetMessage(req, res) {
		const { alias, conversationId, messageId } = req.params;
		const spaceId = this.resolveSpaceId(alias);
		if (!spaceId) {
			res.status(404).json({ error: `Unknown space alias: ${alias}` });
			return;
		}
		const requestId = typeof req.query.requestId === "string" && req.query.requestId || randomUUID();
		logger.debug("Polling message %s in conversation %s from space %s (alias=%s)", messageId, conversationId, spaceId, alias);
		const timeout = this.config.timeout ?? 12e4;
		const streamSettings = {
			...genieStreamDefaults,
			default: {
				...genieStreamDefaults.default,
				timeout
			},
			stream: {
				...genieStreamDefaults.stream,
				streamId: requestId
			}
		};
		const workspaceClient = getWorkspaceClient();
		await this.executeStream(res, (signal) => this.genieConnector.streamGetMessage(workspaceClient, spaceId, conversationId, messageId, {
			timeout,
			signal
		}), streamSettings);
	}
	async getConversation(alias, conversationId, signal) {
		signal?.throwIfAborted();
		const spaceId = this.resolveSpaceId(alias);
		if (!spaceId) throw new Error(`Unknown space alias: ${alias}`);
		const workspaceClient = getWorkspaceClient();
		return this.genieConnector.getConversation(workspaceClient, spaceId, conversationId);
	}
	/**
	* Send a message and consume events as a stream (message_start, status,
	* message_result, query_result, error).
	*/
	async *sendMessage(alias, content, conversationId, options) {
		options?.signal?.throwIfAborted();
		const spaceId = this.resolveSpaceId(alias);
		if (!spaceId) throw new Error(`Unknown space alias: ${alias}`);
		const workspaceClient = getWorkspaceClient();
		const timeout = options?.timeout ?? this.config.timeout ?? 12e4;
		yield* this.genieConnector.streamSendMessage(workspaceClient, spaceId, content, conversationId, {
			timeout,
			signal: options?.signal
		});
	}
	async shutdown() {
		this.streamManager.abortAll();
	}
	getAgentTools() {
		return toolsFromRegistry(this.tools);
	}
	async executeAgentTool(name, args, signal) {
		return executeFromRegistry(this.tools, name, args, signal);
	}
	toolkit(opts) {
		return buildToolkitEntries(this.name, this.tools, opts);
	}
	exports() {
		return {
			sendMessage: this.sendMessage,
			getConversation: this.getConversation
		};
	}
};
/**
* @internal
*/
const genie = toPlugin(GeniePlugin);

//#endregion
export { GeniePlugin, genie };
