// @databricks/appkit/dist/plugins/genie/manifest.js
//#region src/plugins/genie/manifest.json
var manifest_default = {
	$schema: "https://databricks.github.io/appkit/schemas/plugin-manifest.schema.json",
	name: "genie",
	displayName: "Genie Plugin",
	description: "AI/BI Genie space integration for natural language data queries",
	resources: {
		"required": [{
			"type": "genie_space",
			"alias": "Genie Space",
			"resourceKey": "genie-space",
			"description": "Genie Space for AI-powered data queries. Space IDs configured via plugin config.",
			"permission": "CAN_RUN",
			"fields": {
				"id": {
					"env": "DATABRICKS_GENIE_SPACE_ID",
					"description": "Default Genie Space ID",
					"discovery": {
						"type": "kind",
						"resourceKind": "genie_space"
					}
				},
				"name": { "description": "Genie Space display name" }
			}
		}],
		"optional": []
	},
	scaffolding: { "rules": { "must": ["After init, configure the 'spaces' map in plugin config with alias-to-Space-ID mappings"] } },
	config: { "schema": {
		"type": "object",
		"properties": {
			"spaces": {
				"type": "object",
				"description": "Map of alias names to Genie Space IDs",
				"additionalProperties": { "type": "string" }
			},
			"timeout": {
				"type": "number",
				"default": 12e4,
				"description": "Genie polling timeout in ms. Set to 0 for indefinite."
			}
		},
		"required": ["spaces"]
	} }
};

//#endregion
export { manifest_default as default };
