// @databricks/appkit/dist/plugins/jobs/manifest.js
//#region src/plugins/jobs/manifest.json
var manifest_default = {
	$schema: "https://databricks.github.io/appkit/schemas/plugin-manifest.schema.json",
	name: "jobs",
	displayName: "Jobs Plugin",
	description: "Manage Databricks Lakeflow Jobs.",
	resources: {
		"required": [{
			"type": "job",
			"alias": "Job",
			"resourceKey": "job",
			"description": "A Databricks job to trigger and monitor",
			"permission": "CAN_MANAGE_RUN",
			"fields": { "id": {
				"env": "DATABRICKS_JOB_ID",
				"description": "Numeric Databricks job ID. Find it in the Jobs UI or via `databricks jobs list`."
			} }
		}],
		"optional": []
	},
	config: { "schema": {
		"type": "object",
		"properties": {
			"timeout": {
				"type": "number",
				"default": 6e4,
				"description": "Default timeout for Jobs API calls in milliseconds"
			},
			"pollIntervalMs": {
				"type": "number",
				"default": 5e3,
				"description": "Poll interval for waiting on run completion in milliseconds"
			}
		}
	} }
};

//#endregion
export { manifest_default as default };
