// @databricks/appkit/dist/plugins/jobs/plugin.js
import { createLogger } from "../../logging/logger.js";
import { ExecutionError } from "../../errors/execution.js";
import { ValidationError } from "../../errors/validation.js";
import "../../errors/index.js";
import { getCurrentUserId, getWorkspaceClient } from "../../context/execution-context.js";
import "../../context/index.js";
import { Plugin } from "../../plugin/plugin.js";
import { ResourceType } from "../../registry/types.generated.js";
import "../../registry/index.js";
import { toPlugin } from "../../plugin/to-plugin.js";
import "../../plugin/index.js";
import { JobsConnector } from "../../connectors/jobs/client.js";
import "../../connectors/jobs/index.js";
import { JOBS_READ_DEFAULTS, JOBS_STREAM_DEFAULTS, JOBS_WRITE_DEFAULTS } from "./defaults.js";
import manifest_default from "./manifest.js";
import { mapParams } from "./params.js";
import { toJSONSchema } from "zod";
import { STATUS_CODES } from "node:http";

//#region src/plugins/jobs/plugin.ts
const logger = createLogger("jobs");
const DEFAULT_WAIT_TIMEOUT = 6e5;
const DEFAULT_POLL_INTERVAL = 5e3;
/** Cap on param-key count when a job has no Zod schema. Jobs that need more keys must define a schema. */
const MAX_UNVALIDATED_PARAM_KEYS = 50;
/** Replace upstream error messages with generic descriptions keyed by HTTP status. */
function errorResult(status) {
	return {
		ok: false,
		status,
		message: STATUS_CODES[status] ?? "Request failed"
	};
}
function isTerminalRunState(state) {
	return state === "TERMINATED" || state === "SKIPPED" || state === "INTERNAL_ERROR";
}
/** Exponential backoff (1.5x) with +/- 20% jitter, capped at `max`. */
function nextPollDelay(current, max) {
	const jitter = 1 + (Math.random() * .4 - .2);
	return {
		delay: Math.min(current * jitter, max),
		next: Math.min(current * 1.5, max)
	};
}
function abortableSleep(ms, signal) {
	return new Promise((resolve) => {
		if (signal?.aborted) {
			resolve();
			return;
		}
		const timer = setTimeout(resolve, ms);
		signal?.addEventListener("abort", () => {
			clearTimeout(timer);
			resolve();
		}, { once: true });
	});
}
var JobsPlugin = class JobsPlugin extends Plugin {
	static manifest = manifest_default;
	connector;
	jobIds = {};
	jobConfigs = {};
	jobKeys = [];
	/**
	* Scans process.env for DATABRICKS_JOB_* keys and merges with explicit config.
	* Explicit config wins for per-job overrides; auto-discovered jobs get default `{}` config.
	*/
	static discoverJobs(config) {
		const explicit = config.jobs ?? {};
		const discovered = {};
		const prefix = "DATABRICKS_JOB_";
		for (const key of Object.keys(process.env)) {
			if (!key.startsWith(prefix)) continue;
			if (key === "DATABRICKS_JOB_ID") continue;
			const suffix = key.slice(15);
			if (!suffix || !process.env[key]) continue;
			const jobKey = suffix.toLowerCase();
			if (!(jobKey in explicit)) discovered[jobKey] = {};
		}
		if (process.env.DATABRICKS_JOB_ID && Object.keys(explicit).length === 0 && Object.keys(discovered).length === 0) discovered.default = {};
		return {
			...discovered,
			...explicit
		};
	}
	/**
	* Generates resource requirements dynamically from discovered + configured jobs.
	* Each job key maps to a `DATABRICKS_JOB_{KEY_UPPERCASE}` env var (or `DATABRICKS_JOB_ID` for "default").
	*/
	static getResourceRequirements(config) {
		const jobs = JobsPlugin.discoverJobs(config);
		return Object.keys(jobs).map((key) => ({
			type: ResourceType.JOB,
			alias: `job-${key}`,
			resourceKey: `job-${key}`,
			description: `Databricks Job "${key}"`,
			permission: "CAN_MANAGE_RUN",
			fields: { id: {
				env: key === "default" ? "DATABRICKS_JOB_ID" : `DATABRICKS_JOB_${key.toUpperCase()}`,
				description: `Job ID for "${key}"`
			} },
			required: true
		}));
	}
	constructor(config) {
		super(config);
		this.config = config;
		this.connector = new JobsConnector({ telemetry: config.telemetry });
		const jobs = JobsPlugin.discoverJobs(config);
		this.jobKeys = Object.keys(jobs);
		this.jobConfigs = jobs;
		for (const key of this.jobKeys) {
			const envVar = key === "default" ? "DATABRICKS_JOB_ID" : `DATABRICKS_JOB_${key.toUpperCase()}`;
			const jobIdStr = process.env[envVar];
			if (jobIdStr) {
				const parsed = Number.parseInt(jobIdStr, 10);
				if (!Number.isNaN(parsed)) this.jobIds[key] = parsed;
			}
		}
	}
	async setup() {
		logger.info(`Jobs plugin initialized with ${this.jobKeys.length} job(s): ${this.jobKeys.join(", ")}`);
	}
	get client() {
		return getWorkspaceClient();
	}
	getJobId(jobKey) {
		const id = this.jobIds[jobKey];
		if (!id) {
			const envVar = jobKey === "default" ? "DATABRICKS_JOB_ID" : `DATABRICKS_JOB_${jobKey.toUpperCase()}`;
			throw new Error(`Job "${jobKey}" has no configured job ID. Set ${envVar} env var.`);
		}
		return id;
	}
	_readSettings(cacheKey) {
		return { default: {
			...JOBS_READ_DEFAULTS,
			...this.config.timeout != null && { timeout: this.config.timeout },
			cache: {
				...JOBS_READ_DEFAULTS.cache,
				cacheKey
			}
		} };
	}
	_writeSettings() {
		return { default: {
			...JOBS_WRITE_DEFAULTS,
			...this.config.timeout != null && { timeout: this.config.timeout }
		} };
	}
	/**
	* Validates params against the job's Zod schema (if any) and maps them
	* to SDK request fields based on the task type. Shared by runNow and runAndWait.
	*/
	_validateAndMap(jobKey, params) {
		const jobConfig = this.jobConfigs[jobKey];
		let validated = params;
		if (jobConfig?.params) {
			const result = jobConfig.params.safeParse(params ?? {});
			if (!result.success) throw new ValidationError(`Parameter validation failed for job "${jobKey}": ${result.error.message}`);
			validated = result.data;
		}
		return jobConfig?.taskType && validated ? mapParams(jobConfig.taskType, validated) : validated ?? {};
	}
	/**
	* Creates a JobAPI for a specific configured job key.
	* Each method is scoped to the job's configured ID.
	*/
	createJobAPI(jobKey) {
		const jobId = this.getJobId(jobKey);
		const jobConfig = this.jobConfigs[jobKey];
		const self = this;
		const client = this.client;
		const userKey = getCurrentUserId();
		/**
		* Verify that `runId` belongs to this job's configured `jobId`. Returns
		* null if the run is in scope; otherwise returns a 404 `ExecutionResult`.
		* Prevents cross-job access via the `/:jobKey/runs/:runId` HTTP surface.
		*/
		const verifyRunScope = async (runId) => {
			const result = await self.execute(async (signal) => self.connector.getRun(client, { run_id: runId }, signal), self._readSettings([
				"jobs:getRun",
				jobKey,
				runId
			]), userKey);
			if (!result.ok) return errorResult(result.status);
			if (result.data.job_id !== jobId) return errorResult(404);
			return null;
		};
		return {
			runNow: async (params) => {
				const sdkFields = self._validateAndMap(jobKey, params);
				const result = await self.execute(async (signal) => self.connector.runNow(client, {
					...sdkFields,
					job_id: jobId
				}, signal), self._writeSettings(), userKey);
				return result.ok ? result : errorResult(result.status);
			},
			async *runAndWait(params, signal) {
				const sdkFields = self._validateAndMap(jobKey, params);
				const runResult = await self.execute(async (signal) => self.connector.runNow(client, {
					...sdkFields,
					job_id: jobId
				}, signal), self._writeSettings(), userKey);
				if (!runResult.ok) throw new ExecutionError("Failed to trigger job run");
				const runId = runResult.data.run_id;
				if (!runId) throw new Error("runNow did not return a run_id");
				const basePollInterval = self.config.pollIntervalMs ?? DEFAULT_POLL_INTERVAL;
				const maxPollInterval = basePollInterval * 6;
				const timeout = jobConfig?.waitTimeout ?? DEFAULT_WAIT_TIMEOUT;
				const startTime = Date.now();
				let currentInterval = basePollInterval;
				while (!signal?.aborted) {
					if (Date.now() - startTime > timeout) throw new Error(`Job run ${runId} polling timeout after ${timeout}ms`);
					const runStatusResult = await self.execute(async (signal) => self.connector.getRun(client, { run_id: runId }, signal), { default: {
						...JOBS_READ_DEFAULTS,
						cache: { enabled: false }
					} }, userKey);
					if (!runStatusResult.ok) throw new ExecutionError(`Failed to poll run status for run ${runId}`);
					const run = runStatusResult.data;
					const state = run.state?.life_cycle_state;
					yield {
						status: state,
						timestamp: Date.now(),
						run
					};
					if (isTerminalRunState(state)) return;
					const { delay, next } = nextPollDelay(currentInterval, maxPollInterval);
					currentInterval = next;
					await abortableSleep(delay, signal);
				}
			},
			lastRun: async () => {
				const result = await self.execute(async (signal) => self.connector.listRuns(client, {
					job_id: jobId,
					limit: 1
				}, signal), self._readSettings(["jobs:lastRun", jobKey]), userKey);
				if (!result.ok) return errorResult(result.status);
				return {
					ok: true,
					data: result.data[0]
				};
			},
			listRuns: async (options) => {
				const result = await self.execute(async (signal) => self.connector.listRuns(client, {
					job_id: jobId,
					limit: options?.limit
				}, signal), self._readSettings([
					"jobs:listRuns",
					jobKey,
					options?.limit ?? "default"
				]), userKey);
				return result.ok ? result : errorResult(result.status);
			},
			getRun: async (runId) => {
				const result = await self.execute(async (signal) => self.connector.getRun(client, { run_id: runId }, signal), self._readSettings([
					"jobs:getRun",
					jobKey,
					runId
				]), userKey);
				if (!result.ok) return errorResult(result.status);
				if (result.data.job_id !== jobId) return errorResult(404);
				return result;
			},
			getRunOutput: async (runId) => {
				const scopeError = await verifyRunScope(runId);
				if (scopeError) return scopeError;
				const result = await self.execute(async (signal) => self.connector.getRunOutput(client, { run_id: runId }, signal), self._readSettings([
					"jobs:getRunOutput",
					jobKey,
					runId
				]), userKey);
				return result.ok ? result : errorResult(result.status);
			},
			cancelRun: async (runId) => {
				const scopeError = await verifyRunScope(runId);
				if (scopeError) return scopeError;
				const result = await self.execute(async (signal) => self.connector.cancelRun(client, { run_id: runId }, signal), self._writeSettings(), userKey);
				return result.ok ? result : errorResult(result.status);
			},
			getJob: async () => {
				const result = await self.execute(async (signal) => self.connector.getJob(client, { job_id: jobId }, signal), self._readSettings(["jobs:getJob", jobKey]), userKey);
				return result.ok ? result : errorResult(result.status);
			}
		};
	}
	/**
	* Resolve `:jobKey` from the request. Returns the key and ID,
	* or sends a 404 and returns `{ jobKey: undefined, jobId: undefined }`.
	*/
	_resolveJob(req, res) {
		const jobKey = req.params.jobKey;
		if (!this.jobKeys.includes(jobKey)) {
			const safeKey = jobKey.replace(/[^a-zA-Z0-9_-]/g, "");
			res.status(404).json({
				error: `Unknown job "${safeKey}"`,
				plugin: this.name
			});
			return {
				jobKey: void 0,
				jobId: void 0
			};
		}
		const jobId = this.jobIds[jobKey];
		if (!jobId) {
			res.status(404).json({
				error: `Job "${jobKey}" has no configured job ID`,
				plugin: this.name
			});
			return {
				jobKey: void 0,
				jobId: void 0
			};
		}
		return {
			jobKey,
			jobId
		};
	}
	_sendStatusError(res, status) {
		res.status(status).json({
			error: STATUS_CODES[status] ?? "Unknown Error",
			plugin: this.name
		});
	}
	/**
	* Validate params from an HTTP request body. Eager validation lets streaming
	* requests get a clean 400 instead of a generic SSE error event. Throws
	* ValidationError so handlers can map to a 400 response via their catch block.
	*/
	_parseRunParams(jobKey, rawParams) {
		if (rawParams !== void 0 && (typeof rawParams !== "object" || rawParams === null || Array.isArray(rawParams))) throw new ValidationError("params must be a plain object");
		const jobConfig = this.jobConfigs[jobKey];
		if (jobConfig?.params) {
			if (!jobConfig.params.safeParse(rawParams ?? {}).success) throw new ValidationError("Invalid job parameters");
			return rawParams;
		}
		if (rawParams !== void 0) {
			if (!jobConfig?.taskType) throw new ValidationError("This job does not accept parameters");
			const keyCount = Object.keys(rawParams).length;
			if (keyCount > MAX_UNVALIDATED_PARAM_KEYS) throw new ValidationError(`Too many parameters (${keyCount}). Define a Zod schema to accept more than ${MAX_UNVALIDATED_PARAM_KEYS}.`);
		}
		return rawParams;
	}
	async _handleRun(req, res) {
		const { jobKey } = this._resolveJob(req, res);
		if (!jobKey) return;
		const stream = req.query.stream === "true";
		try {
			const params = this._parseRunParams(jobKey, req.body?.params);
			const api = this.createJobAPI(jobKey);
			if (stream) {
				const streamSettings = { default: JOBS_STREAM_DEFAULTS };
				await this.executeStream(res, (signal) => api.runAndWait(params, signal), streamSettings);
			} else {
				const result = await api.runNow(params);
				if (!result.ok) {
					this._sendStatusError(res, result.status);
					return;
				}
				res.json({ runId: result.data.run_id });
			}
		} catch (error) {
			if (error instanceof ValidationError) {
				if (!res.headersSent) res.status(400).json({
					error: error.message,
					plugin: this.name
				});
				return;
			}
			logger.error("Run failed for job %s: %O", jobKey, error);
			if (!res.headersSent) res.status(500).json({
				error: "Run failed",
				plugin: this.name
			});
		}
	}
	injectRoutes(router) {
		this.route(router, {
			name: "run",
			method: "post",
			path: "/:jobKey/run",
			handler: (req, res) => this._handleRun(req, res)
		});
		this.route(router, {
			name: "runs",
			method: "get",
			path: "/:jobKey/runs",
			handler: async (req, res) => {
				const { jobKey } = this._resolveJob(req, res);
				if (!jobKey) return;
				const limit = Math.max(1, Math.min(Number.parseInt(req.query.limit, 10) || 20, 100));
				try {
					const result = await this.createJobAPI(jobKey).listRuns({ limit });
					if (!result.ok) {
						this._sendStatusError(res, result.status);
						return;
					}
					res.json({ runs: result.data });
				} catch (error) {
					logger.error("List runs failed for job %s: %O", jobKey, error);
					res.status(500).json({
						error: "List runs failed",
						plugin: this.name
					});
				}
			}
		});
		this.route(router, {
			name: "run-detail",
			method: "get",
			path: "/:jobKey/runs/:runId",
			handler: async (req, res) => {
				const { jobKey } = this._resolveJob(req, res);
				if (!jobKey) return;
				const runId = Number.parseInt(req.params.runId, 10);
				if (Number.isNaN(runId) || runId <= 0) {
					res.status(400).json({
						error: "Invalid runId",
						plugin: this.name
					});
					return;
				}
				try {
					const result = await this.createJobAPI(jobKey).getRun(runId);
					if (!result.ok) {
						this._sendStatusError(res, result.status);
						return;
					}
					res.json(result.data);
				} catch (error) {
					logger.error("Get run failed for job %s run %d: %O", jobKey, runId, error);
					res.status(500).json({
						error: "Get run failed",
						plugin: this.name
					});
				}
			}
		});
		this.route(router, {
			name: "status",
			method: "get",
			path: "/:jobKey/status",
			handler: async (req, res) => {
				const { jobKey } = this._resolveJob(req, res);
				if (!jobKey) return;
				try {
					const result = await this.createJobAPI(jobKey).lastRun();
					if (!result.ok) {
						this._sendStatusError(res, result.status);
						return;
					}
					res.json({
						status: result.data?.state?.life_cycle_state ?? null,
						run: result.data ?? null
					});
				} catch (error) {
					logger.error("Status check failed for job %s: %O", jobKey, error);
					res.status(500).json({
						error: "Status check failed",
						plugin: this.name
					});
				}
			}
		});
		this.route(router, {
			name: "cancel-run",
			method: "delete",
			path: "/:jobKey/runs/:runId",
			handler: async (req, res) => {
				const { jobKey } = this._resolveJob(req, res);
				if (!jobKey) return;
				const runId = Number.parseInt(req.params.runId, 10);
				if (Number.isNaN(runId) || runId <= 0) {
					res.status(400).json({
						error: "Invalid runId",
						plugin: this.name
					});
					return;
				}
				try {
					const result = await this.createJobAPI(jobKey).cancelRun(runId);
					if (!result.ok) {
						this._sendStatusError(res, result.status);
						return;
					}
					res.status(204).end();
				} catch (error) {
					logger.error("Cancel run failed for job %s run %d: %O", jobKey, runId, error);
					res.status(500).json({
						error: "Cancel run failed",
						plugin: this.name
					});
				}
			}
		});
	}
	exports() {
		const resolveJob = (jobKey) => {
			if (!this.jobKeys.includes(jobKey)) throw new Error(`Unknown job "${jobKey}". Available jobs: ${this.jobKeys.join(", ")}`);
			return {
				...this.createJobAPI(jobKey),
				asUser: (req) => {
					return this.asUser(req).createJobAPI(jobKey);
				}
			};
		};
		return resolveJob;
	}
	clientConfig() {
		const jobs = {};
		for (const key of this.jobKeys) {
			const config = this.jobConfigs[key];
			jobs[key] = {
				params: config?.params ? toJSONSchema(config.params) : null,
				taskType: config?.taskType ?? null
			};
		}
		return { jobs };
	}
};
/**
* @internal
*/
const jobs = toPlugin(JobsPlugin);

//#endregion
export { jobs };
