// @databricks/appkit-ui/dist/react/file-browser/directory-list.js
import { Button } from "../ui/button.js";
import { Card } from "../ui/card.js";
import { Skeleton } from "../ui/skeleton.js";
import { FileEntry } from "./file-entry.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { AlertCircle, ArrowLeft, FileIcon } from "lucide-react";

//#region src/react/file-browser/directory-list.tsx
/** Card-wrapped directory listing with loading, error, and empty states */
function DirectoryList({ entries, loading, error, onEntryClick, onNavigateToParent, onRetry, isAtRoot, selectedPath, resolveEntryPath, headerContent, hasCurrentPath, formatSize, labels, className, ...props }) {
	return /* @__PURE__ */ jsx("div", {
		"data-slot": "directory-list",
		className,
		...props,
		children: /* @__PURE__ */ jsxs(Card, {
			className: "p-0 overflow-hidden",
			children: [
				!isAtRoot && onNavigateToParent && /* @__PURE__ */ jsxs("button", {
					type: "button",
					onClick: onNavigateToParent,
					className: "flex items-center gap-2 px-4 py-3 w-full text-left hover:bg-muted/50 border-b text-sm text-muted-foreground transition-colors",
					children: [/* @__PURE__ */ jsx(ArrowLeft, { className: "h-4 w-4" }), labels?.backToParent ?? "Back to parent"]
				}),
				headerContent,
				loading && /* @__PURE__ */ jsxs("div", {
					className: "p-4 space-y-3",
					children: [
						/* @__PURE__ */ jsx(Skeleton, { className: "h-10 w-full" }),
						/* @__PURE__ */ jsx(Skeleton, { className: "h-10 w-full" }),
						/* @__PURE__ */ jsx(Skeleton, { className: "h-10 w-full" })
					]
				}),
				error && /* @__PURE__ */ jsxs("div", {
					className: "p-6 text-center",
					children: [
						/* @__PURE__ */ jsx(AlertCircle, { className: "h-8 w-8 text-destructive mx-auto mb-2" }),
						/* @__PURE__ */ jsx("p", {
							className: "text-sm text-destructive",
							children: error
						}),
						onRetry && /* @__PURE__ */ jsx(Button, {
							variant: "outline",
							size: "sm",
							className: "mt-3",
							onClick: onRetry,
							children: labels?.retry ?? "Retry"
						})
					]
				}),
				!loading && !error && entries.length === 0 && /* @__PURE__ */ jsxs("div", {
					className: "p-6 text-center text-muted-foreground",
					children: [/* @__PURE__ */ jsx(FileIcon, { className: "h-8 w-8 mx-auto mb-2 opacity-50" }), /* @__PURE__ */ jsx("p", {
						className: "text-sm",
						children: hasCurrentPath ? labels?.emptyDirectory ?? "This directory is empty." : labels?.noVolumeConfigured ?? "No volume configured. Configure volumes in the files plugin to get started."
					})]
				}),
				!loading && !error && entries.map((entry) => {
					const entryPath = resolveEntryPath(entry);
					return /* @__PURE__ */ jsx(FileEntry, {
						entry,
						entryPath,
						isSelected: selectedPath === entryPath,
						formatSize,
						onClick: () => onEntryClick(entry)
					}, entryPath);
				})
			]
		})
	});
}

//#endregion
export { DirectoryList };
