// @databricks/appkit-ui/dist/react/file-browser/file-preview-panel.js
import { formatFileSize } from "../lib/format.js";
import { Button } from "../ui/button.js";
import { Card } from "../ui/card.js";
import { Skeleton } from "../ui/skeleton.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { AlertCircle, Download, FileIcon, Loader2, Trash2 } from "lucide-react";

//#region src/react/file-browser/file-preview-panel.tsx
/** Preview panel displaying file metadata, image/text preview, and download/delete actions */
function FilePreviewPanel({ selectedFile, preview, previewLoading, onDownload, onDelete, deleting, imagePreviewSrc, formatSize = formatFileSize, labels, className, ...props }) {
	const resolveImageSrc = (filePath) => {
		if (!imagePreviewSrc) return "";
		if (typeof imagePreviewSrc === "string") return imagePreviewSrc;
		return imagePreviewSrc(filePath);
	};
	return /* @__PURE__ */ jsx("div", {
		"data-slot": "file-preview-panel",
		className,
		...props,
		children: /* @__PURE__ */ jsxs(Card, {
			className: "p-6",
			children: [
				!selectedFile && /* @__PURE__ */ jsxs("div", {
					className: "text-center text-muted-foreground py-8",
					children: [/* @__PURE__ */ jsx(FileIcon, { className: "h-10 w-10 mx-auto mb-3 opacity-40" }), /* @__PURE__ */ jsx("p", {
						className: "text-sm",
						children: labels?.selectFilePrompt ?? "Select a file to preview"
					})]
				}),
				selectedFile && previewLoading && /* @__PURE__ */ jsxs("div", {
					className: "space-y-3",
					children: [
						/* @__PURE__ */ jsx(Skeleton, { className: "h-5 w-3/4" }),
						/* @__PURE__ */ jsx(Skeleton, { className: "h-4 w-1/2" }),
						/* @__PURE__ */ jsx(Skeleton, { className: "h-4 w-2/3" }),
						/* @__PURE__ */ jsx(Skeleton, { className: "h-32 w-full mt-4" })
					]
				}),
				selectedFile && !previewLoading && preview && /* @__PURE__ */ jsxs("div", {
					className: "space-y-4",
					children: [
						/* @__PURE__ */ jsxs("div", { children: [/* @__PURE__ */ jsx("h3", {
							className: "font-semibold text-foreground truncate mb-1",
							children: selectedFile.split("/").pop()
						}), /* @__PURE__ */ jsx("p", {
							className: "text-xs text-muted-foreground truncate",
							children: selectedFile
						})] }),
						/* @__PURE__ */ jsxs("div", {
							className: "space-y-2 text-sm",
							children: [
								/* @__PURE__ */ jsxs("div", {
									className: "flex justify-between",
									children: [/* @__PURE__ */ jsx("span", {
										className: "text-muted-foreground",
										children: labels?.size ?? "Size"
									}), /* @__PURE__ */ jsx("span", {
										className: "text-foreground",
										children: formatSize(preview.contentLength)
									})]
								}),
								/* @__PURE__ */ jsxs("div", {
									className: "flex justify-between",
									children: [/* @__PURE__ */ jsx("span", {
										className: "text-muted-foreground",
										children: labels?.type ?? "Type"
									}), /* @__PURE__ */ jsx("span", {
										className: "text-foreground truncate ml-2",
										children: preview.contentType ?? labels?.unknown ?? "Unknown"
									})]
								}),
								preview.lastModified && /* @__PURE__ */ jsxs("div", {
									className: "flex justify-between",
									children: [/* @__PURE__ */ jsx("span", {
										className: "text-muted-foreground",
										children: labels?.modified ?? "Modified"
									}), /* @__PURE__ */ jsx("span", {
										className: "text-foreground",
										children: preview.lastModified
									})]
								})
							]
						}),
						/* @__PURE__ */ jsxs("div", {
							className: "flex gap-2",
							children: [onDownload && /* @__PURE__ */ jsxs(Button, {
								variant: "outline",
								size: "sm",
								className: "flex-1",
								onClick: () => onDownload(selectedFile),
								children: [/* @__PURE__ */ jsx(Download, { className: "h-4 w-4 mr-2" }), labels?.download ?? "Download"]
							}), onDelete && /* @__PURE__ */ jsx(Button, {
								variant: "destructive",
								size: "sm",
								disabled: deleting,
								onClick: () => onDelete(selectedFile),
								children: deleting ? /* @__PURE__ */ jsx(Loader2, { className: "h-4 w-4 animate-spin" }) : /* @__PURE__ */ jsx(Trash2, { className: "h-4 w-4" })
							})]
						}),
						preview.isImage && imagePreviewSrc && /* @__PURE__ */ jsx("div", {
							className: "border rounded overflow-hidden",
							children: /* @__PURE__ */ jsx("img", {
								src: resolveImageSrc(selectedFile),
								alt: selectedFile.split("/").pop() ?? "Preview",
								className: "w-full h-auto"
							})
						}),
						preview.isText && preview.textPreview !== null && /* @__PURE__ */ jsx("div", {
							className: "border rounded",
							children: /* @__PURE__ */ jsx("pre", {
								className: "p-3 text-xs font-mono overflow-auto max-h-80 whitespace-pre-wrap text-foreground bg-muted/30",
								children: preview.textPreview
							})
						}),
						!preview.isText && !preview.isImage && /* @__PURE__ */ jsx("div", {
							className: "text-center py-4 text-sm text-muted-foreground",
							children: labels?.previewNotAvailable ?? "Preview not available for this file type."
						})
					]
				}),
				selectedFile && !previewLoading && !preview && /* @__PURE__ */ jsxs("div", {
					className: "text-center py-8",
					children: [/* @__PURE__ */ jsx(AlertCircle, { className: "h-8 w-8 text-destructive mx-auto mb-2" }), /* @__PURE__ */ jsx("p", {
						className: "text-sm text-destructive",
						children: labels?.previewFailed ?? "Failed to load preview"
					})]
				})
			]
		})
	});
}

//#endregion
export { FilePreviewPanel };
