// @databricks/appkit-ui/dist/react/file-browser/new-folder-input.js
import { cn } from "../lib/utils.js";
import { Button } from "../ui/button.js";
import { useEffect, useRef } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
import { FolderPlus, Loader2, X } from "lucide-react";

//#region src/react/file-browser/new-folder-input.tsx
/** Inline folder-name input with create/cancel actions */
function NewFolderInput({ value, onChange, onCreate, onCancel, creating, autoFocus = true, labels, className, ...props }) {
	const inputRef = useRef(null);
	useEffect(() => {
		if (autoFocus) inputRef.current?.focus();
	}, [autoFocus]);
	return /* @__PURE__ */ jsxs("div", {
		"data-slot": "new-folder-input",
		className: cn("flex items-center gap-2 px-4 py-3 border-b bg-muted/30", className),
		...props,
		children: [
			/* @__PURE__ */ jsx(FolderPlus, { className: "h-5 w-5 text-blue-500 shrink-0" }),
			/* @__PURE__ */ jsx("input", {
				ref: inputRef,
				type: "text",
				value,
				onChange: (e) => onChange(e.target.value),
				onKeyDown: (e) => {
					if (e.key === "Enter") onCreate();
					if (e.key === "Escape") onCancel();
				},
				placeholder: labels?.folderNamePlaceholder ?? "Folder name",
				className: "flex-1 text-sm bg-background border rounded px-2 py-1 outline-none focus:ring-1 focus:ring-ring",
				disabled: creating
			}),
			/* @__PURE__ */ jsx(Button, {
				variant: "ghost",
				size: "sm",
				disabled: creating || !value.trim(),
				onClick: onCreate,
				children: creating ? /* @__PURE__ */ jsx(Loader2, { className: "h-4 w-4 animate-spin" }) : labels?.create ?? "Create"
			}),
			/* @__PURE__ */ jsx("button", {
				type: "button",
				onClick: onCancel,
				className: "text-muted-foreground hover:text-foreground",
				children: /* @__PURE__ */ jsx(X, { className: "h-4 w-4" })
			})
		]
	});
}

//#endregion
export { NewFolderInput };
