// @databricks/appkit/dist/telemetry/noop.js
import { INVALID_SPAN_CONTEXT, createNoopMeter } from "@opentelemetry/api";
import { NOOP_LOGGER } from "@opentelemetry/api-logs";

//#region src/telemetry/noop.ts
var NonRecordingSpan = class {
	_spanContext;
	constructor(spanContext = INVALID_SPAN_CONTEXT) {
		this._spanContext = spanContext;
	}
	spanContext() {
		return this._spanContext;
	}
	setAttribute(_key, _value) {
		return this;
	}
	setAttributes(_attributes) {
		return this;
	}
	addEvent(_name, _attributesOrStartTime, _startTime) {
		return this;
	}
	addLink(_link) {
		return this;
	}
	addLinks(_links) {
		return this;
	}
	setStatus(_status) {
		return this;
	}
	updateName(_name) {
		return this;
	}
	end(_endTime) {}
	isRecording() {
		return false;
	}
	recordException(_exception, _time) {}
};
var NoopTracer = class {
	startSpan(_name, _options, _context) {
		return new NonRecordingSpan(INVALID_SPAN_CONTEXT);
	}
	startActiveSpan(_name, ...args) {
		const fn = args[args.length - 1];
		if (typeof fn !== "function") return;
		return fn(new NonRecordingSpan(INVALID_SPAN_CONTEXT));
	}
};
const NOOP_TRACER = new NoopTracer();
const NOOP_METER = createNoopMeter();

//#endregion
export { NOOP_LOGGER, NOOP_METER, NOOP_TRACER };
