// @databricks/appkit/dist/telemetry/telemetry-provider.js
import { normalizeTelemetryOptions } from "./config.js";
import { NOOP_LOGGER, NOOP_METER, NOOP_TRACER } from "./noop.js";
import { metrics, trace } from "@opentelemetry/api";
import { logs } from "@opentelemetry/api-logs";

//#region src/telemetry/telemetry-provider.ts
/**
* Scoped telemetry instance for specific plugins and other classes.
* Automatically uses the plugin name as the default tracer/meter name.
*/
var TelemetryProvider = class {
	pluginName;
	globalManager;
	config;
	constructor(pluginName, globalManager, telemetryConfig) {
		this.pluginName = pluginName;
		this.globalManager = globalManager;
		this.config = normalizeTelemetryOptions(telemetryConfig);
	}
	/**
	* Gets a tracer for creating spans.
	* @param options - Optional tracer configuration.
	*/
	getTracer(options) {
		if (!this.config.traces) return NOOP_TRACER;
		const tracerName = this.getInstrumentName(options);
		return trace.getTracer(tracerName);
	}
	/**
	* Gets a meter for recording metrics.
	* @param config - Optional meter configuration.
	*/
	getMeter(config) {
		if (!this.config.metrics) return NOOP_METER;
		const meterName = this.getInstrumentName(config);
		return metrics.getMeter(meterName);
	}
	/**
	* Gets a logger for emitting log records.
	* @param config - Optional logger configuration.
	*/
	getLogger(config) {
		if (!this.config.logs) return NOOP_LOGGER;
		const loggerName = this.getInstrumentName(config);
		return logs.getLogger(loggerName);
	}
	/**
	* Emits a log record using the default logger.
	* Convenience method for logging without explicitly getting a logger.
	* @param logRecord - The log record to emit
	*/
	emit(logRecord) {
		this.getLogger().emit(logRecord);
	}
	/**
	* Register OpenTelemetry instrumentations.
	* Can be called at any time, but recommended to call in plugin constructor.
	* @param instrumentations - Array of OpenTelemetry instrumentations to register
	*/
	registerInstrumentations(instrumentations) {
		if (!this.config.traces) return;
		this.globalManager.registerInstrumentations(instrumentations);
	}
	/**
	* Starts an active span and executes a callback function within its context.
	* Uses the plugin's default tracer unless custom tracer options are provided.
	* @param name - The name of the span
	* @param options - Span options including attributes, kind, etc.
	* @param fn - Callback function to execute within the span context
	* @param tracerOptions - Optional tracer configuration
	* @returns Promise resolving to the callback's return value
	*/
	startActiveSpan(name, options, fn, tracerOptions) {
		return this.getTracer(tracerOptions).startActiveSpan(name, options, fn);
	}
	getInstrumentName(options) {
		const prefix = this.pluginName;
		if (!options || !options.name) return this.pluginName;
		return options.includePrefix ? `${prefix}-${options.name}` : options.name;
	}
};

//#endregion
export { TelemetryProvider };
