// @databricks/appkit/dist/telemetry/telemetry-manager.js
import { createLogger } from "../logging/logger.js";
import { TelemetryProvider } from "./telemetry-provider.js";
import { AppKitSampler } from "./trace-sampler.js";
import { getNodeAutoInstrumentations } from "@opentelemetry/auto-instrumentations-node";
import { OTLPLogExporter } from "@opentelemetry/exporter-logs-otlp-proto";
import { OTLPMetricExporter } from "@opentelemetry/exporter-metrics-otlp-proto";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-proto";
import { registerInstrumentations } from "@opentelemetry/instrumentation";
import { detectResources, envDetector, hostDetector, processDetector, resourceFromAttributes } from "@opentelemetry/resources";
import { BatchLogRecordProcessor } from "@opentelemetry/sdk-logs";
import { PeriodicExportingMetricReader } from "@opentelemetry/sdk-metrics";
import { NodeSDK } from "@opentelemetry/sdk-node";
import { ATTR_SERVICE_NAME, ATTR_SERVICE_VERSION } from "@opentelemetry/semantic-conventions";

//#region src/telemetry/telemetry-manager.ts
const logger = createLogger("telemetry");
var TelemetryManager = class TelemetryManager {
	static DEFAULT_EXPORT_INTERVAL_MS = 1e4;
	static DEFAULT_FALLBACK_APP_NAME = "databricks-app";
	static instance;
	sdk;
	/**
	* Create a scoped telemetry provider for a specific plugin.
	* The plugin's name will be used as the default tracer/meter name.
	* @param pluginName - The name of the plugin to create scoped telemetry for
	* @param telemetryConfig - The telemetry configuration for the plugin
	* @returns A scoped telemetry instance for the plugin
	*/
	static getProvider(pluginName, telemetryConfig) {
		return new TelemetryProvider(pluginName, TelemetryManager.getInstance(), telemetryConfig);
	}
	constructor() {}
	static getInstance() {
		if (!TelemetryManager.instance) TelemetryManager.instance = new TelemetryManager();
		return TelemetryManager.instance;
	}
	static initialize(config = {}) {
		TelemetryManager.getInstance()._initialize(config);
	}
	_initialize(config) {
		if (this.sdk) return;
		if (!process.env.OTEL_EXPORTER_OTLP_ENDPOINT) return;
		try {
			this.sdk = new NodeSDK({
				resource: this.createResource(config),
				autoDetectResources: false,
				sampler: new AppKitSampler(),
				traceExporter: new OTLPTraceExporter({ headers: config.headers }),
				metricReaders: [new PeriodicExportingMetricReader({
					exporter: new OTLPMetricExporter({ headers: config.headers }),
					exportIntervalMillis: config.exportIntervalMs || TelemetryManager.DEFAULT_EXPORT_INTERVAL_MS
				})],
				logRecordProcessors: [new BatchLogRecordProcessor(new OTLPLogExporter({ headers: config.headers }))],
				instrumentations: this.getDefaultInstrumentations()
			});
			this.sdk.start();
			this.registerShutdown();
			logger.debug("Initialized successfully");
		} catch (error) {
			logger.error("Failed to initialize: %O", error);
		}
	}
	/**
	* Register OpenTelemetry instrumentations.
	* Can be called at any time, but recommended to call in plugin constructor.
	* @param instrumentations - Array of OpenTelemetry instrumentations to register
	*/
	registerInstrumentations(instrumentations) {
		registerInstrumentations({ instrumentations });
	}
	createResource(config) {
		const serviceName = config.serviceName || process.env.OTEL_SERVICE_NAME || process.env.DATABRICKS_APP_NAME || TelemetryManager.DEFAULT_FALLBACK_APP_NAME;
		const initialResource = resourceFromAttributes({
			[ATTR_SERVICE_NAME]: serviceName,
			[ATTR_SERVICE_VERSION]: config.serviceVersion ?? void 0
		});
		const detectedResource = detectResources({ detectors: [
			envDetector,
			hostDetector,
			processDetector
		] });
		return initialResource.merge(detectedResource);
	}
	getDefaultInstrumentations() {
		return [...getNodeAutoInstrumentations({
			"@opentelemetry/instrumentation-http": { enabled: false },
			"@opentelemetry/instrumentation-express": { enabled: false },
			"@opentelemetry/instrumentation-fs": { enabled: false },
			"@opentelemetry/instrumentation-dns": { enabled: false },
			"@opentelemetry/instrumentation-net": { enabled: false }
		})];
	}
	registerShutdown() {
		const shutdownFn = async () => {
			await TelemetryManager.getInstance().shutdown();
		};
		process.once("SIGTERM", shutdownFn);
		process.once("SIGINT", shutdownFn);
	}
	async shutdown() {
		if (!this.sdk) return;
		try {
			await this.sdk.shutdown();
			this.sdk = void 0;
		} catch (error) {
			logger.error("Error shutting down: %O", error);
		}
	}
};

//#endregion
export { TelemetryManager };
