// @databricks/appkit/dist/connectors/sql-warehouse/client.js
import { createLogger } from "../../logging/logger.js";
import { AppKitError } from "../../errors/base.js";
import { ConnectionError } from "../../errors/connection.js";
import { ExecutionError } from "../../errors/execution.js";
import { ValidationError } from "../../errors/validation.js";
import "../../errors/index.js";
import { TelemetryManager } from "../../telemetry/telemetry-manager.js";
import { SpanKind, SpanStatusCode } from "../../telemetry/index.js";
import { ArrowStreamProcessor } from "../../stream/arrow-stream-processor.js";
import { executeStatementDefaults } from "./defaults.js";
import { Context } from "@databricks/sdk-experimental";

//#region src/connectors/sql-warehouse/client.ts
const logger = createLogger("connectors:sql-warehouse");
var SQLWarehouseConnector = class {
	name = "sql-warehouse";
	config;
	_arrowProcessor = null;
	telemetry;
	telemetryMetrics;
	constructor(config) {
		this.config = config;
		this.telemetry = TelemetryManager.getProvider(this.name, this.config.telemetry);
		this.telemetryMetrics = {
			queryCount: this.telemetry.getMeter().createCounter("query.count", {
				description: "Total number of queries executed",
				unit: "1"
			}),
			queryDuration: this.telemetry.getMeter().createHistogram("query.duration", {
				description: "Duration of queries executed",
				unit: "ms"
			})
		};
	}
	/**
	* Lazily initializes and returns the ArrowStreamProcessor.
	* Only created on first Arrow format query to avoid unnecessary allocation.
	*/
	get arrowProcessor() {
		if (!this._arrowProcessor) this._arrowProcessor = new ArrowStreamProcessor({
			timeout: this.config.timeout || executeStatementDefaults.timeout,
			maxConcurrentDownloads: ArrowStreamProcessor.DEFAULT_MAX_CONCURRENT_DOWNLOADS,
			retries: ArrowStreamProcessor.DEFAULT_RETRIES
		});
		return this._arrowProcessor;
	}
	async executeStatement(workspaceClient, input, signal) {
		const startTime = Date.now();
		let success = false;
		if (signal?.aborted) throw ExecutionError.canceled();
		return this.telemetry.startActiveSpan("sql.query", {
			kind: SpanKind.CLIENT,
			attributes: {
				"db.system": "databricks",
				"db.warehouse_id": input.warehouse_id || "",
				"db.catalog": input.catalog ?? "",
				"db.schema": input.schema ?? "",
				"db.statement": input.statement?.substring(0, 500) || "",
				"db.has_parameters": !!input.parameters
			}
		}, async (span) => {
			let abortHandler;
			let isAborted = false;
			if (signal) {
				abortHandler = () => {
					if (!span.isRecording()) return;
					isAborted = true;
					span.setAttribute("cancelled", true);
					span.setStatus({
						code: SpanStatusCode.ERROR,
						message: "Query cancelled by client"
					});
					span.end();
				};
				signal.addEventListener("abort", abortHandler, { once: true });
			}
			try {
				if (!input.statement) throw ValidationError.missingField("statement");
				if (!input.warehouse_id) throw ValidationError.missingField("warehouse_id");
				const body = {
					statement: input.statement,
					parameters: input.parameters,
					warehouse_id: input.warehouse_id,
					catalog: input.catalog,
					schema: input.schema,
					wait_timeout: input.wait_timeout || executeStatementDefaults.wait_timeout,
					disposition: input.disposition || executeStatementDefaults.disposition,
					format: input.format || executeStatementDefaults.format,
					byte_limit: input.byte_limit,
					row_limit: input.row_limit,
					on_wait_timeout: input.on_wait_timeout || executeStatementDefaults.on_wait_timeout
				};
				span.addEvent("statement.submitting", { "db.warehouse_id": input.warehouse_id });
				const response = await workspaceClient.statementExecution.executeStatement(body, this._createContext(signal));
				if (!response) throw ConnectionError.apiFailure("SQL Warehouse");
				const status = response.status;
				const statementId = response.statement_id;
				span.setAttribute("db.statement_id", statementId);
				span.addEvent("statement.submitted", {
					"db.statement_id": response.statement_id,
					"db.status": status?.state
				});
				let result;
				switch (status?.state) {
					case "RUNNING":
					case "PENDING":
						span.addEvent("statement.polling_started", { "db.status": response.status?.state });
						result = await this._pollForStatementResult(workspaceClient, statementId, this.config.timeout, signal);
						break;
					case "SUCCEEDED":
						result = this._transformDataArray(response);
						break;
					case "FAILED": throw ExecutionError.statementFailed(status.error?.message);
					case "CANCELED": throw ExecutionError.canceled();
					case "CLOSED": throw ExecutionError.resultsClosed();
					default: throw ExecutionError.unknownState(String(status?.state ?? "unknown"));
				}
				const resultData = result.result;
				const rowCount = resultData?.data?.length ?? resultData?.data_array?.length ?? 0;
				if (rowCount > 0) span.setAttribute("db.result.row_count", rowCount);
				const duration = Date.now() - startTime;
				logger.event()?.setContext("sql-warehouse", {
					warehouse_id: input.warehouse_id,
					rows_returned: rowCount,
					query_duration_ms: duration
				});
				success = true;
				if (!isAborted) span.setStatus({ code: SpanStatusCode.OK });
				return result;
			} catch (error) {
				if (!isAborted) {
					span.recordException(error);
					span.setStatus({
						code: SpanStatusCode.ERROR,
						message: error instanceof Error ? error.message : String(error)
					});
					logger.error("Statement execution failed: %s", error instanceof Error ? error.message : String(error));
				}
				if (error instanceof AppKitError) throw error;
				throw ExecutionError.statementFailed(error instanceof Error ? error.message : String(error));
			} finally {
				if (abortHandler && signal) signal.removeEventListener("abort", abortHandler);
				const duration = Date.now() - startTime;
				if (!isAborted) span.end();
				const attributes = {
					"db.warehouse_id": input.warehouse_id,
					"db.catalog": input.catalog ?? "",
					"db.schema": input.schema ?? "",
					"db.statement": input.statement?.substring(0, 500) || "",
					success: success.toString()
				};
				this.telemetryMetrics.queryCount.add(1, attributes);
				this.telemetryMetrics.queryDuration.record(duration, attributes);
			}
		}, {
			name: this.name,
			includePrefix: true
		});
	}
	async _pollForStatementResult(workspaceClient, statementId, timeout = executeStatementDefaults.timeout, signal) {
		return this.telemetry.startActiveSpan("sql.poll", { attributes: {
			"db.statement_id": statementId,
			"db.polling.timeout": timeout
		} }, async (span) => {
			try {
				const startTime = Date.now();
				let delay = 1e3;
				const maxDelayBetweenPolls = 5e3;
				let pollCount = 0;
				while (true) {
					pollCount++;
					span.setAttribute("db.polling.current_attempt", pollCount);
					const elapsedTime = Date.now() - startTime;
					if (elapsedTime > timeout) {
						const error = ExecutionError.statementFailed(`Polling timeout exceeded after ${timeout}ms (elapsed: ${elapsedTime}ms)`);
						span.recordException(error);
						span.setStatus({ code: SpanStatusCode.ERROR });
						throw error;
					}
					if (signal?.aborted) {
						const error = ExecutionError.canceled();
						span.recordException(error);
						span.setStatus({ code: SpanStatusCode.ERROR });
						throw error;
					}
					span.addEvent("polling.attempt", {
						"poll.attempt": pollCount,
						"poll.delay_ms": delay,
						"poll.elapsed_ms": elapsedTime
					});
					const response = await workspaceClient.statementExecution.getStatement({ statement_id: statementId }, this._createContext(signal));
					if (!response) throw ConnectionError.apiFailure("SQL Warehouse");
					const status = response.status;
					span.addEvent("polling.status_check", {
						"db.status": status?.state,
						"poll.attempt": pollCount
					});
					switch (status?.state) {
						case "PENDING":
						case "RUNNING": break;
						case "SUCCEEDED":
							span.setAttribute("db.polling.attempts", pollCount);
							span.setAttribute("db.polling.total_duration_ms", elapsedTime);
							span.addEvent("polling.completed", {
								"poll.attempts": pollCount,
								"poll.duration_ms": elapsedTime
							});
							span.setStatus({ code: SpanStatusCode.OK });
							return this._transformDataArray(response);
						case "FAILED": throw ExecutionError.statementFailed(status.error?.message);
						case "CANCELED": throw ExecutionError.canceled();
						case "CLOSED": throw ExecutionError.resultsClosed();
						default: throw ExecutionError.unknownState(String(status?.state ?? "unknown"));
					}
					await new Promise((resolve) => setTimeout(resolve, delay));
					delay = Math.min(delay * 2, maxDelayBetweenPolls);
				}
			} catch (error) {
				span.recordException(error);
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: error instanceof Error ? error.message : String(error)
				});
				if (error instanceof AppKitError) throw error;
				throw ExecutionError.statementFailed(error instanceof Error ? error.message : String(error));
			} finally {
				span.end();
			}
		}, {
			name: this.name,
			includePrefix: true
		});
	}
	_transformDataArray(response) {
		if (response.manifest?.format === "ARROW_STREAM") return this.updateWithArrowStatus(response);
		if (!response.result?.data_array || !response.manifest?.schema?.columns) return response;
		const columns = response.manifest.schema.columns;
		const transformedData = response.result.data_array.map((row) => {
			const obj = {};
			row.forEach((value, index) => {
				const column = columns[index];
				const columnName = column?.name || `column_${index}`;
				if (column?.type_name === "STRING" && typeof value === "string" && value && (value[0] === "{" || value[0] === "[")) try {
					obj[columnName] = JSON.parse(value);
				} catch {
					obj[columnName] = value;
				}
				else obj[columnName] = value;
			});
			return obj;
		});
		const { data_array: _data_array, ...restResult } = response.result;
		return {
			...response,
			result: {
				...restResult,
				data: transformedData
			}
		};
	}
	updateWithArrowStatus(response) {
		return { result: {
			statement_id: response.statement_id,
			status: {
				state: response.status?.state,
				error: response.status?.error
			}
		} };
	}
	async getArrowData(workspaceClient, jobId, signal) {
		const startTime = Date.now();
		return this.telemetry.startActiveSpan("arrow.getData", {
			kind: SpanKind.CLIENT,
			attributes: {
				"db.system": "databricks",
				"arrow.job_id": jobId
			}
		}, async (span) => {
			try {
				const response = await workspaceClient.statementExecution.getStatement({ statement_id: jobId }, this._createContext(signal));
				const chunks = response.result?.external_links;
				const schema = response.manifest?.schema;
				if (!chunks || !schema) throw ExecutionError.missingData("chunks or schema");
				span.setAttribute("arrow.chunk_count", chunks.length);
				const result = await this.arrowProcessor.processChunks(chunks, schema, signal);
				span.setAttribute("arrow.data_size_bytes", result.data.length);
				span.setStatus({ code: SpanStatusCode.OK });
				const duration = Date.now() - startTime;
				this.telemetryMetrics.queryDuration.record(duration, {
					operation: "arrow.getData",
					status: "success"
				});
				logger.event()?.setContext("sql-warehouse", {
					arrow_data_size_bytes: result.data.length,
					arrow_job_id: jobId
				});
				return result;
			} catch (error) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: error instanceof Error ? error.message : "Unknown error"
				});
				span.recordException(error);
				const duration = Date.now() - startTime;
				this.telemetryMetrics.queryDuration.record(duration, {
					operation: "arrow.getData",
					status: "error"
				});
				logger.error("Failed Arrow job: %s %O", jobId, error);
				if (error instanceof AppKitError) throw error;
				throw ExecutionError.statementFailed(error instanceof Error ? error.message : String(error));
			}
		});
	}
	_createContext(signal) {
		return new Context({ cancellationToken: {
			isCancellationRequested: signal?.aborted ?? false,
			onCancellationRequested: (cb) => {
				signal?.addEventListener("abort", cb, { once: true });
			}
		} });
	}
};

//#endregion
export { SQLWarehouseConnector };
