// @databricks/appkit/dist/connectors/lakebase/index.js
import { createLogger } from "../../logging/logger.js";
import { createLakebasePoolManager } from "./pool-manager.js";
import { RoutingPool } from "./routing-pool.js";
import { RequestedClaimsPermissionSet, createLakebasePool, generateDatabaseCredential, getLakebaseOrmConfig, getLakebasePgConfig, getUsernameWithApiLookup, getWorkspaceClient } from "@databricks/lakebase";

//#region src/connectors/lakebase/index.ts
/**
* Create a Lakebase pool with appkit's logger integration.
* Telemetry automatically uses appkit's OpenTelemetry configuration via global registry.
*
* @param config - Lakebase pool configuration
* @returns PostgreSQL pool with appkit integration
*/
function createLakebasePool$1(config) {
	const logger = createLogger("connectors:lakebase");
	return createLakebasePool({
		...config,
		logger
	});
}

//#endregion
export { RequestedClaimsPermissionSet, createLakebasePool$1 as createLakebasePool, generateDatabaseCredential, getLakebaseOrmConfig, getLakebasePgConfig, getUsernameWithApiLookup, getWorkspaceClient };
