// @databricks/appkit/dist/connectors/lakebase/pool-manager.js
import { createLakebasePool } from "./index.js";

//#region src/connectors/lakebase/pool-manager.ts
/** Interval for removing empty (connectionless) pools from the Map. */
const CLEANUP_INTERVAL_MS = 300 * 1e3;
/**
* Create a pool manager that maintains per-key Lakebase connection pools.
*
* Each pool is created via `createLakebasePool` with the base config merged
* with per-pool overrides (e.g. a user's `workspaceClient` and `user`).
*
* A periodic cleanup removes empty Pool objects (where all connections have
* been closed by pg's built-in `idleTimeoutMillis`) from the internal Map.
*
* @example OBO usage
* ```typescript
* const poolManager = createLakebasePoolManager();
*
* // In a route handler:
* const userPool = poolManager.getPool(userName, {
*   workspaceClient: new WorkspaceClient({ token: userToken, host, authType: "pat" }),
*   user: userName,
* });
* const result = await userPool.query("SELECT * FROM products");
* ```
*/
function createLakebasePoolManager(baseConfig) {
	const entries = /* @__PURE__ */ new Map();
	const cleanupTimer = setInterval(() => {
		for (const [key, entry] of entries) if (entry.pool.totalCount === 0) {
			entry.pool.end().catch(() => {});
			entries.delete(key);
		}
	}, CLEANUP_INTERVAL_MS);
	cleanupTimer.unref();
	return {
		getPool(key, perPoolConfig, tokenFingerprint) {
			const existing = entries.get(key);
			if (existing) {
				if (!(tokenFingerprint && existing.tokenFingerprint && tokenFingerprint !== existing.tokenFingerprint)) return existing.pool;
				existing.pool.end().catch(() => {});
			}
			const pool = createLakebasePool({
				...baseConfig,
				...perPoolConfig
			});
			entries.set(key, {
				pool,
				tokenFingerprint
			});
			return pool;
		},
		hasPool(key) {
			return entries.has(key);
		},
		async closePool(key) {
			const entry = entries.get(key);
			if (entry) {
				await entry.pool.end();
				entries.delete(key);
			}
		},
		async closeAll() {
			clearInterval(cleanupTimer);
			const endPromises = [...entries.values()].map((e) => e.pool.end());
			await Promise.all(endPromises);
			entries.clear();
		},
		get size() {
			return entries.size;
		}
	};
}

//#endregion
export { createLakebasePoolManager };
