// @databricks/appkit/dist/connectors/lakebase/routing-pool.js
import { getUserContext } from "../../context/execution-context.js";

//#region src/connectors/lakebase/routing-pool.ts
/**
* A `pg.Pool`-like wrapper that routes queries to the appropriate pool
* based on the current execution context.
*
* When called inside `runInUserContext()` (set up by `Plugin.asUser(req)`),
* queries route to the per-user pool returned by `resolveUserPool`.
* Otherwise, queries route to the service-principal pool.
*
* This enables OBO (On-Behalf-Of) without custom `asUser()` overrides —
* the base class sets up AsyncLocalStorage context, and the RoutingPool
* reads it transparently.
*/
var RoutingPool = class {
	constructor(spPool, resolveUserPool) {
		this.spPool = spPool;
		this.resolveUserPool = resolveUserPool;
	}
	activePool() {
		const userCtx = getUserContext();
		return userCtx ? this.resolveUserPool(userCtx) : this.spPool;
	}
	query(text, values) {
		return this.activePool().query(text, values);
	}
	connect() {
		return this.activePool().connect();
	}
	async end() {
		await this.spPool.end();
	}
	get totalCount() {
		return this.spPool.totalCount;
	}
	get idleCount() {
		return this.spPool.idleCount;
	}
	get waitingCount() {
		return this.spPool.waitingCount;
	}
};

//#endregion
export { RoutingPool };
