// @databricks/appkit/dist/connectors/genie/client.js
import { createLogger } from "../../logging/logger.js";
import "../../logging/index.js";
import { genieConnectorDefaults } from "./defaults.js";
import { pollWaiter } from "./poll-waiter.js";
import * as SDK from "@databricks/sdk-experimental";

//#region src/connectors/genie/client.ts
const { TimeUnits } = SDK;
const Time = SDK.Time ?? SDK.default.Time;
const logger = createLogger("connectors:genie");
const GenieErrors = {
	SPACE_ACCESS_DENIED: "You don't have access to this Genie Space.",
	TABLE_PERMISSIONS: "You may not have access to the data tables. Please verify your table permissions.",
	REQUEST_FAILED: "Genie request failed",
	QUERY_RESULT_FAILED: "Failed to fetch query result"
};
function mapAttachments(message) {
	return message.attachments?.map((att) => ({
		attachmentId: att.attachment_id,
		query: att.query ? {
			title: att.query.title,
			description: att.query.description,
			query: att.query.query,
			statementId: att.query.statement_id
		} : void 0,
		text: att.text ? { content: att.text.content } : void 0,
		suggestedQuestions: att.suggested_questions?.questions
	})) ?? [];
}
function toMessageResponse(message) {
	return {
		messageId: message.message_id,
		conversationId: message.conversation_id,
		spaceId: message.space_id,
		status: message.status ?? "COMPLETED",
		content: message.content,
		attachments: mapAttachments(message),
		error: message.error?.error
	};
}
function classifyGenieError(error) {
	const message = error instanceof Error ? error.message : String(error);
	if (message.includes("RESOURCE_DOES_NOT_EXIST")) return GenieErrors.SPACE_ACCESS_DENIED;
	if (message.includes("failed to reach COMPLETED state") && message.includes("FAILED")) return GenieErrors.TABLE_PERMISSIONS;
	return message || GenieErrors.REQUEST_FAILED;
}
var GenieConnector = class {
	config;
	constructor(config = {}) {
		this.config = {
			timeout: config.timeout ?? genieConnectorDefaults.timeout,
			maxMessages: config.maxMessages ?? genieConnectorDefaults.maxMessages
		};
	}
	async startMessage(workspaceClient, spaceId, content, conversationId) {
		if (conversationId) {
			const waiter = await workspaceClient.genie.createMessage({
				space_id: spaceId,
				conversation_id: conversationId,
				content
			});
			return {
				messageWaiter: waiter,
				conversationId,
				messageId: waiter.message_id ?? ""
			};
		}
		const start = await workspaceClient.genie.startConversation({
			space_id: spaceId,
			content
		});
		return {
			messageWaiter: start,
			conversationId: start.conversation_id,
			messageId: start.message_id
		};
	}
	async waitForMessage(messageWaiter, options) {
		const timeout = options?.timeout ?? this.config.timeout;
		const waitOptions = timeout > 0 ? { timeout: new Time(timeout, TimeUnits.milliseconds) } : {};
		return messageWaiter.wait(waitOptions);
	}
	async listConversationMessages(workspaceClient, spaceId, conversationId, options) {
		const pageSize = options?.pageSize ?? genieConnectorDefaults.initialPageSize;
		const response = await workspaceClient.genie.listConversationMessages({
			space_id: spaceId,
			conversation_id: conversationId,
			page_size: pageSize,
			...options?.pageToken ? { page_token: options.pageToken } : {}
		});
		return {
			messages: (response.messages ?? []).reverse().map(toMessageResponse),
			nextPageToken: response.next_page_token ?? null
		};
	}
	async getMessageAttachmentQueryResult(workspaceClient, spaceId, conversationId, messageId, attachmentId, _signal) {
		return (await workspaceClient.genie.getMessageAttachmentQueryResult({
			space_id: spaceId,
			conversation_id: conversationId,
			message_id: messageId,
			attachment_id: attachmentId
		})).statement_response;
	}
	async *streamSendMessage(workspaceClient, spaceId, content, conversationId, options) {
		try {
			const { messageWaiter, conversationId: resultConversationId, messageId: resultMessageId } = await this.startMessage(workspaceClient, spaceId, content, conversationId);
			yield {
				type: "message_start",
				conversationId: resultConversationId,
				messageId: resultMessageId,
				spaceId
			};
			const timeout = options?.timeout != null ? options.timeout : this.config.timeout;
			const waitOptions = timeout > 0 ? { timeout: new Time(timeout, TimeUnits.milliseconds) } : {};
			let completedMessage;
			for await (const event of pollWaiter(messageWaiter, waitOptions)) if (event.type === "progress" && event.value.status) yield {
				type: "status",
				status: event.value.status
			};
			else if (event.type === "completed") completedMessage = event.value;
			const messageResponse = toMessageResponse(completedMessage);
			yield {
				type: "message_result",
				message: messageResponse
			};
			yield* this.emitQueryResults(workspaceClient, spaceId, resultConversationId, messageResponse.messageId, messageResponse);
		} catch (error) {
			logger.error("Genie message error (spaceId=%s, conversationId=%s): %O", spaceId, conversationId ?? "new", error);
			yield {
				type: "error",
				error: classifyGenieError(error)
			};
		}
	}
	async *emitQueryResults(workspaceClient, spaceId, conversationId, messageId, messageResponse) {
		const attachments = messageResponse.attachments ?? [];
		for (const att of attachments) {
			if (!att.query?.statementId || !att.attachmentId) continue;
			try {
				const data = await this.getMessageAttachmentQueryResult(workspaceClient, spaceId, conversationId, messageId, att.attachmentId);
				yield {
					type: "query_result",
					attachmentId: att.attachmentId,
					statementId: att.query.statementId,
					data
				};
			} catch (error) {
				logger.error("Failed to fetch query result for attachment %s: %O", att.attachmentId, error);
				yield {
					type: "error",
					error: `${GenieErrors.QUERY_RESULT_FAILED} for attachment ${att.attachmentId}`
				};
			}
		}
	}
	async *streamConversation(workspaceClient, spaceId, conversationId, options) {
		const includeQueryResults = options?.includeQueryResults !== false;
		try {
			const { messages: messageResponses, nextPageToken } = await this.listConversationMessages(workspaceClient, spaceId, conversationId, {
				pageSize: options?.pageSize,
				pageToken: options?.pageToken
			});
			for (const messageResponse of messageResponses) yield {
				type: "message_result",
				message: messageResponse
			};
			yield {
				type: "history_info",
				conversationId,
				spaceId,
				nextPageToken,
				loadedCount: messageResponses.length
			};
			if (includeQueryResults) {
				const queryAttachments = [];
				for (const msg of messageResponses) for (const att of msg.attachments ?? []) if (att.query?.statementId && att.attachmentId) queryAttachments.push({
					messageId: msg.messageId,
					attachmentId: att.attachmentId,
					statementId: att.query.statementId
				});
				const results = await Promise.allSettled(queryAttachments.map(async (att) => {
					const data = await this.getMessageAttachmentQueryResult(workspaceClient, spaceId, conversationId, att.messageId, att.attachmentId);
					return {
						attachmentId: att.attachmentId,
						statementId: att.statementId,
						data
					};
				}));
				for (const result of results) if (result.status === "fulfilled") yield {
					type: "query_result",
					attachmentId: result.value.attachmentId,
					statementId: result.value.statementId,
					data: result.value.data
				};
				else {
					logger.error("Failed to fetch query result: %O", result.reason);
					yield {
						type: "error",
						error: result.reason instanceof Error ? result.reason.message : GenieErrors.QUERY_RESULT_FAILED
					};
				}
			}
		} catch (error) {
			logger.error("Genie getConversation error (spaceId=%s, conversationId=%s): %O", spaceId, conversationId, error);
			yield {
				type: "error",
				error: classifyGenieError(error)
			};
		}
	}
	/**
	* Polls a single message via `getMessage` until it reaches a terminal
	* state (`COMPLETED` or `FAILED`). Yields the same event types as
	* `streamSendMessage` so callers can reuse the same SSE processing logic.
	*/
	async *streamGetMessage(workspaceClient, spaceId, conversationId, messageId, options) {
		const pollInterval = options?.pollInterval ?? 3e3;
		const signal = options?.signal;
		let lastStatus = "";
		try {
			while (true) {
				if (signal?.aborted) return;
				const message = await workspaceClient.genie.getMessage({
					space_id: spaceId,
					conversation_id: conversationId,
					message_id: messageId
				});
				if (message.status && message.status !== lastStatus) {
					lastStatus = message.status;
					yield {
						type: "status",
						status: message.status
					};
				}
				if (message.status === "COMPLETED" || message.status === "FAILED") {
					const messageResponse = toMessageResponse(message);
					yield {
						type: "message_result",
						message: messageResponse
					};
					yield* this.emitQueryResults(workspaceClient, spaceId, conversationId, messageId, messageResponse);
					return;
				}
				await new Promise((resolve) => {
					const timer = setTimeout(resolve, pollInterval);
					signal?.addEventListener("abort", () => {
						clearTimeout(timer);
						resolve();
					}, { once: true });
				});
			}
		} catch (error) {
			if (signal?.aborted) return;
			logger.error("Genie getMessage poll error (spaceId=%s, conversationId=%s, messageId=%s): %O", spaceId, conversationId, messageId, error);
			yield {
				type: "error",
				error: classifyGenieError(error)
			};
		}
	}
	async sendMessage(workspaceClient, spaceId, content, conversationId) {
		const { messageWaiter, conversationId: resultConversationId } = await this.startMessage(workspaceClient, spaceId, content, conversationId);
		return {
			...toMessageResponse(await this.waitForMessage(messageWaiter)),
			conversationId: resultConversationId
		};
	}
	async getConversation(workspaceClient, spaceId, conversationId) {
		const allMessages = [];
		let pageToken;
		do {
			const { messages, nextPageToken } = await this.listConversationMessages(workspaceClient, spaceId, conversationId, {
				pageSize: genieConnectorDefaults.pageSize,
				pageToken
			});
			allMessages.push(...messages);
			pageToken = nextPageToken ?? void 0;
		} while (pageToken && allMessages.length < this.config.maxMessages);
		return {
			conversationId,
			spaceId,
			messages: allMessages.slice(0, this.config.maxMessages)
		};
	}
};

//#endregion
export { GenieConnector };
