// @databricks/appkit/dist/connectors/serving/client.js
import { createLogger } from "../../logging/logger.js";
import { Context } from "@databricks/sdk-experimental";

//#region src/connectors/serving/client.ts
const logger = createLogger("connectors:serving");
/**
* Bridges {@link AbortSignal} to the SDK's {@link CancellationToken} so
* `apiClient.request` can abort the outbound HTTP request (and stop pulling
* the SSE body) when the agent run is cancelled.
*/
function cancellationTokenFromAbortSignal(signal) {
	const listeners = /* @__PURE__ */ new Set();
	const fire = () => {
		for (const cb of listeners) try {
			cb();
		} catch {}
	};
	signal.addEventListener("abort", fire, { passive: true });
	return {
		get isCancellationRequested() {
			return signal.aborted;
		},
		onCancellationRequested(callback) {
			listeners.add(callback);
			if (signal.aborted) callback();
		}
	};
}
/**
* Invokes a serving endpoint using the SDK's high-level query API.
* Returns a typed QueryEndpointResponse.
*/
async function invoke(client, endpointName, body) {
	const { stream: _stream, ...cleanBody } = body;
	logger.debug("Invoking endpoint %s", endpointName);
	return client.servingEndpoints.query({
		name: endpointName,
		...cleanBody
	});
}
/**
* Returns the raw SSE byte stream from a serving endpoint.
* No parsing is performed — bytes are passed through as-is.
*
* Uses the SDK's low-level `apiClient.request({ raw: true })` because
* the high-level `servingEndpoints.query()` returns `Promise<QueryEndpointResponse>`
* and does not support SSE streaming.
*/
async function stream(client, endpointName, body, signal) {
	const { stream: _stream, ...cleanBody } = body;
	logger.debug("Streaming from endpoint %s", endpointName);
	const context = signal ? new Context({ cancellationToken: cancellationTokenFromAbortSignal(signal) }) : void 0;
	const response = await client.apiClient.request({
		path: `/serving-endpoints/${encodeURIComponent(endpointName)}/invocations`,
		method: "POST",
		headers: new Headers({
			"Content-Type": "application/json",
			Accept: "text/event-stream"
		}),
		payload: {
			...cleanBody,
			stream: true
		},
		raw: true
	}, context);
	if (!response.contents) throw new Error("Response body is null — streaming not supported");
	return response.contents;
}

//#endregion
export { invoke, stream };
