// @databricks/sdk-experimental/dist/apis/vectorsearch/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VectorSearchIndexesService = exports.VectorSearchEndpointsService = void 0;
const vectorsearch = __importStar(require("./model"));
const retries_1 = __importDefault(require("../../retries/retries"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
const wait_1 = require("../../wait");
class VectorSearchEndpointsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("VectorSearchEndpoints", method, message);
    }
}
class VectorSearchEndpointsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("VectorSearchEndpoints", method, message);
    }
}
/**
 * **Endpoint**: Represents the compute resources to host vector search indexes.
 */
class VectorSearchEndpointsService {
    constructor(client) {
        this.client = client;
    }
    async _createEndpoint(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("budget_policy_id")) {
            body["budget_policy_id"] = request.budget_policy_id;
        }
        if (request.hasOwnProperty("endpoint_type")) {
            body["endpoint_type"] = request.endpoint_type;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        const query = {};
        const path = "/api/2.0/vector-search/endpoints";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a new endpoint.
     */
    async createEndpoint(createEndpoint, context) {
        let cancellationToken = context?.cancellationToken;
        const endpointInfo = await this._createEndpoint(createEndpoint, context);
        return (0, wait_1.asWaiter)(endpointInfo, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.getEndpoint({
                        endpoint_name: endpointInfo.name,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("VectorSearchEndpoints.createEndpointAndWait: cancelled");
                        throw new VectorSearchEndpointsError("createEndpointAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.endpoint_status.state;
                    const statusMessage = pollResponse.endpoint_status.message;
                    switch (status) {
                        case "ONLINE": {
                            return pollResponse;
                        }
                        case "OFFLINE": {
                            const errorMessage = `failed to reach ONLINE state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`VectorSearchEndpoints.createEndpointAndWait: ${errorMessage}`);
                            throw new VectorSearchEndpointsError("createEndpointAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach ONLINE state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`VectorSearchEndpoints.createEndpointAndWait: retrying: ${errorMessage}`);
                            throw new VectorSearchEndpointsRetriableError("createEndpointAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _deleteEndpoint(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/vector-search/endpoints/${request.endpoint_name}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a vector search endpoint.
     */
    async deleteEndpoint(request, context) {
        return await this._deleteEndpoint(request, context);
    }
    async _getEndpoint(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/vector-search/endpoints/${request.endpoint_name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get details for a single vector search endpoint.
     */
    async getEndpoint(request, context) {
        return await this._getEndpoint(request, context);
    }
    async _listEndpoints(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/vector-search/endpoints";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List all vector search endpoints in the workspace.
     */
    async *listEndpoints(request, context) {
        while (true) {
            const response = await this._listEndpoints(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.endpoints || response.endpoints.length === 0) {
                break;
            }
            for (const v of response.endpoints) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _retrieveUserVisibleMetrics(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("end_time")) {
            body["end_time"] = request.end_time;
        }
        if (request.hasOwnProperty("granularity_in_seconds")) {
            body["granularity_in_seconds"] = request.granularity_in_seconds;
        }
        if (request.hasOwnProperty("metrics")) {
            body["metrics"] = request.metrics;
        }
        if (request.hasOwnProperty("page_token")) {
            body["page_token"] = request.page_token;
        }
        if (request.hasOwnProperty("start_time")) {
            body["start_time"] = request.start_time;
        }
        const query = {};
        const path = `/api/2.0/vector-search/endpoints/${request.name}/metrics`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieve user-visible metrics for an endpoint
     */
    async retrieveUserVisibleMetrics(request, context) {
        return await this._retrieveUserVisibleMetrics(request, context);
    }
    async _updateEndpointBudgetPolicy(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("budget_policy_id")) {
            body["budget_policy_id"] = request.budget_policy_id;
        }
        const query = {};
        const path = `/api/2.0/vector-search/endpoints/${request.endpoint_name}/budget-policy`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update the budget policy of an endpoint
     */
    async updateEndpointBudgetPolicy(request, context) {
        return await this._updateEndpointBudgetPolicy(request, context);
    }
    async _updateEndpointCustomTags(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("custom_tags")) {
            body["custom_tags"] = request.custom_tags;
        }
        const query = {};
        const path = `/api/2.0/vector-search/endpoints/${request.endpoint_name}/tags`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update the custom tags of an endpoint.
     */
    async updateEndpointCustomTags(request, context) {
        return await this._updateEndpointCustomTags(request, context);
    }
}
exports.VectorSearchEndpointsService = VectorSearchEndpointsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchEndpointsService.prototype, "_createEndpoint", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchEndpointsService.prototype, "createEndpoint", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchEndpointsService.prototype, "_deleteEndpoint", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchEndpointsService.prototype, "deleteEndpoint", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchEndpointsService.prototype, "_getEndpoint", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchEndpointsService.prototype, "getEndpoint", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchEndpointsService.prototype, "_listEndpoints", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], VectorSearchEndpointsService.prototype, "listEndpoints", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchEndpointsService.prototype, "_retrieveUserVisibleMetrics", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchEndpointsService.prototype, "retrieveUserVisibleMetrics", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchEndpointsService.prototype, "_updateEndpointBudgetPolicy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchEndpointsService.prototype, "updateEndpointBudgetPolicy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchEndpointsService.prototype, "_updateEndpointCustomTags", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchEndpointsService.prototype, "updateEndpointCustomTags", null);
class VectorSearchIndexesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("VectorSearchIndexes", method, message);
    }
}
class VectorSearchIndexesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("VectorSearchIndexes", method, message);
    }
}
/**
 * **Index**: An efficient representation of your embedding vectors that supports
 * real-time and efficient approximate nearest neighbor (ANN) search queries.
 *
 * There are 2 types of Vector Search indexes: - **Delta Sync Index**: An index
 * that automatically syncs with a source Delta Table, automatically and
 * incrementally updating the index as the underlying data in the Delta Table
 * changes. - **Direct Vector Access Index**: An index that supports direct read
 * and write of vectors and metadata through our REST and SDK APIs. With this
 * model, the user manages index updates.
 */
class VectorSearchIndexesService {
    constructor(client) {
        this.client = client;
    }
    async _createIndex(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("delta_sync_index_spec")) {
            body["delta_sync_index_spec"] = request.delta_sync_index_spec;
        }
        if (request.hasOwnProperty("direct_access_index_spec")) {
            body["direct_access_index_spec"] = request.direct_access_index_spec;
        }
        if (request.hasOwnProperty("endpoint_name")) {
            body["endpoint_name"] = request.endpoint_name;
        }
        if (request.hasOwnProperty("index_type")) {
            body["index_type"] = request.index_type;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("primary_key")) {
            body["primary_key"] = request.primary_key;
        }
        const query = {};
        const path = "/api/2.0/vector-search/indexes";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a new index.
     */
    async createIndex(request, context) {
        return await this._createIndex(request, context);
    }
    async _deleteDataVectorIndex(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("primary_keys")) {
            query["primary_keys"] =
                request.primary_keys;
        }
        const path = `/api/2.0/vector-search/indexes/${request.index_name}/delete-data`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Handles the deletion of data from a specified vector index.
     */
    async deleteDataVectorIndex(request, context) {
        return await this._deleteDataVectorIndex(request, context);
    }
    async _deleteIndex(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/vector-search/indexes/${request.index_name}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete an index.
     */
    async deleteIndex(request, context) {
        return await this._deleteIndex(request, context);
    }
    async _getIndex(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("ensure_reranker_compatible")) {
            query["ensure_reranker_compatible"] =
                request.ensure_reranker_compatible;
        }
        const path = `/api/2.0/vector-search/indexes/${request.index_name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get an index.
     */
    async getIndex(request, context) {
        return await this._getIndex(request, context);
    }
    async _listIndexes(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("endpoint_name")) {
            query["endpoint_name"] =
                request.endpoint_name;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/vector-search/indexes";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List all indexes in the given endpoint.
     */
    async *listIndexes(request, context) {
        while (true) {
            const response = await this._listIndexes(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.vector_indexes ||
                response.vector_indexes.length === 0) {
                break;
            }
            for (const v of response.vector_indexes) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _queryIndex(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("columns")) {
            body["columns"] = request.columns;
        }
        if (request.hasOwnProperty("columns_to_rerank")) {
            body["columns_to_rerank"] = request.columns_to_rerank;
        }
        if (request.hasOwnProperty("filters_json")) {
            body["filters_json"] = request.filters_json;
        }
        if (request.hasOwnProperty("num_results")) {
            body["num_results"] = request.num_results;
        }
        if (request.hasOwnProperty("query_text")) {
            body["query_text"] = request.query_text;
        }
        if (request.hasOwnProperty("query_type")) {
            body["query_type"] = request.query_type;
        }
        if (request.hasOwnProperty("query_vector")) {
            body["query_vector"] = request.query_vector;
        }
        if (request.hasOwnProperty("reranker")) {
            body["reranker"] = request.reranker;
        }
        if (request.hasOwnProperty("score_threshold")) {
            body["score_threshold"] = request.score_threshold;
        }
        const query = {};
        const path = `/api/2.0/vector-search/indexes/${request.index_name}/query`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Query the specified vector index.
     */
    async queryIndex(request, context) {
        return await this._queryIndex(request, context);
    }
    async _queryNextPage(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("endpoint_name")) {
            body["endpoint_name"] = request.endpoint_name;
        }
        if (request.hasOwnProperty("page_token")) {
            body["page_token"] = request.page_token;
        }
        const query = {};
        const path = `/api/2.0/vector-search/indexes/${request.index_name}/query-next-page`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Use `next_page_token` returned from previous `QueryVectorIndex` or
     * `QueryVectorIndexNextPage` request to fetch next page of results.
     */
    async queryNextPage(request, context) {
        return await this._queryNextPage(request, context);
    }
    async _scanIndex(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("last_primary_key")) {
            body["last_primary_key"] = request.last_primary_key;
        }
        if (request.hasOwnProperty("num_results")) {
            body["num_results"] = request.num_results;
        }
        const query = {};
        const path = `/api/2.0/vector-search/indexes/${request.index_name}/scan`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Scan the specified vector index and return the first `num_results` entries
     * after the exclusive `primary_key`.
     */
    async scanIndex(request, context) {
        return await this._scanIndex(request, context);
    }
    async _syncIndex(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/vector-search/indexes/${request.index_name}/sync`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Triggers a synchronization process for a specified vector index.
     */
    async syncIndex(request, context) {
        return await this._syncIndex(request, context);
    }
    async _upsertDataVectorIndex(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("inputs_json")) {
            body["inputs_json"] = request.inputs_json;
        }
        const query = {};
        const path = `/api/2.0/vector-search/indexes/${request.index_name}/upsert-data`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Handles the upserting of data into a specified vector index.
     */
    async upsertDataVectorIndex(request, context) {
        return await this._upsertDataVectorIndex(request, context);
    }
}
exports.VectorSearchIndexesService = VectorSearchIndexesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "_createIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "createIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "_deleteDataVectorIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "deleteDataVectorIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "_deleteIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "deleteIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "_getIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "getIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "_listIndexes", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], VectorSearchIndexesService.prototype, "listIndexes", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "_queryIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "queryIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "_queryNextPage", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "queryNextPage", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "_scanIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "scanIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "_syncIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "syncIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "_upsertDataVectorIndex", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VectorSearchIndexesService.prototype, "upsertDataVectorIndex", null);
