// @databricks/appkit/dist/plugins/server/index.js
import { createLogger } from "../../logging/logger.js";
import { ServerError } from "../../errors/server.js";
import "../../errors/index.js";
import { instrumentations } from "../../telemetry/instrumentations.js";
import "../../telemetry/index.js";
import { TelemetryReporter } from "../../internal-telemetry/reporter.js";
import "../../internal-telemetry/index.js";
import { Plugin } from "../../plugin/plugin.js";
import { toPlugin } from "../../plugin/to-plugin.js";
import "../../plugin/index.js";
import { sanitizeClientConfig } from "./client-config-sanitizer.js";
import manifest_default from "./manifest.js";
import { RemoteTunnelController } from "./remote-tunnel/remote-tunnel-controller.js";
import { getRoutes, printRoutes } from "./utils.js";
import { StaticServer } from "./static-server.js";
import { ViteDevServer } from "./vite-dev-server.js";
import path from "node:path";
import fs from "node:fs";
import dotenv from "dotenv";
import express from "express";
import getPort, { portNumbers } from "get-port";

//#region src/plugins/server/index.ts
dotenv.config({ path: path.resolve(process.cwd(), "./.env") });
const logger = createLogger("server");
/** Dev-only: try `requested` then consecutive ports (see `get-port` `portNumbers`). */
const devListenPortSpan = 100;
/**
* Server plugin for the AppKit.
*
* This plugin is responsible for starting the server and serving the static files.
* It also handles the remote tunneling for development purposes.
*
* The server is started automatically by `createApp` after all plugins are set up
* and the optional `onPluginsReady` callback has run.
*
* @example
* ```ts
* createApp({
*   plugins: [server(), analytics({})],
*   onPluginsReady(appkit) {
*     appkit.server.extend((app) => {
*       app.get("/custom", (_req, res) => res.json({ ok: true }));
*     });
*   },
* });
* ```
*
*/
var ServerPlugin = class ServerPlugin extends Plugin {
	static DEFAULT_CONFIG = {
		host: process.env.FLASK_RUN_HOST || "0.0.0.0",
		port: Number(process.env.DATABRICKS_APP_PORT) || 8e3
	};
	/** Plugin manifest declaring metadata and resource requirements */
	static manifest = manifest_default;
	serverApplication;
	server;
	viteDevServer;
	remoteTunnelController;
	/** Bound listen port after optional dev-time resolution. */
	resolvedListenPort;
	serverExtensions = [];
	rawBodyPaths = /* @__PURE__ */ new Set();
	static phase = "deferred";
	constructor(config) {
		super(config);
		if ("autoStart" in config) throw new ServerError("server({ autoStart }) has been removed. The server is now started automatically by createApp.\n\nRun `npx appkit codemod on-plugins-ready --write` to auto-migrate.");
		this.config = config;
		this.serverApplication = express();
		this.server = null;
		this.serverExtensions = [];
	}
	attachContext(deps = {}) {
		super.attachContext(deps);
		this.telemetry.registerInstrumentations([instrumentations.http, instrumentations.express]);
		this.context?.registerAsRouteTarget(this);
	}
	/** Setup the server plugin. */
	async setup() {}
	/** Get the server configuration. */
	getConfig() {
		const { plugins: _plugins, ...config } = this.config;
		return config;
	}
	/**
	* Start the server.
	*
	* This method starts the server and sets up the frontend.
	* It also sets up the remote tunneling if enabled.
	*
	* @returns The express application.
	*/
	async start() {
		this.serverApplication.use(requestMetricsMiddleware);
		this.serverApplication.use(express.json({
			limit: this.config.bodyLimit ?? "1mb",
			type: (req) => {
				const urlPath = req.url?.split("?")[0];
				if (urlPath && this.rawBodyPaths.has(urlPath)) return false;
				return (req.headers["content-type"] ?? "").includes("json");
			}
		}));
		const { endpoints, pluginConfigs } = await this.extendRoutes();
		for (const extension of this.serverExtensions) extension(this.serverApplication);
		this.remoteTunnelController = new RemoteTunnelController(this.devFileReader);
		this.serverApplication.use(this.remoteTunnelController.middleware);
		await this.setupFrontend(endpoints, pluginConfigs);
		const listenPort = await this.resolveListenPort();
		const server = this.serverApplication.listen(listenPort, this.config.host ?? ServerPlugin.DEFAULT_CONFIG.host, () => this.logStartupInfo());
		this.server = server;
		this.remoteTunnelController.setServer(server);
		process.on("SIGTERM", () => this._gracefulShutdown());
		process.on("SIGINT", () => this._gracefulShutdown());
		if (process.env.NODE_ENV === "development") printRoutes(getRoutes(this.serverApplication._router.stack));
		return this.serverApplication;
	}
	/**
	* Get the low level node.js http server instance.
	*
	* Only use this method if you need to access the server instance for advanced usage like a custom websocket server, etc.
	*
	* @throws {Error} If the server has not started yet.
	* @returns {HTTPServer} The server instance.
	*/
	getServer() {
		if (!this.server) throw ServerError.notStarted();
		return this.server;
	}
	/**
	* Extend the server with custom routes or middleware.
	*
	* Call this inside the `onPluginsReady` callback of `createApp` to register
	* custom Express routes or middleware before the server starts listening.
	*
	* @param fn - A function that receives the express application.
	* @returns The server plugin instance for chaining.
	*/
	extend(fn) {
		this.serverExtensions.push(fn);
		return this;
	}
	/**
	* Register a server extension from another plugin during setup.
	* Unlike extend(), this is designed for internal plugin-to-plugin
	* coordination where extensions are registered before the server starts
	* listening — typically called by PluginContext when flushing buffered routes.
	*/
	addExtension(fn) {
		this.serverExtensions.push(fn);
	}
	/**
	* Setup the routes with the plugins.
	*
	* This method goes through all the plugins and injects the routes into the server application.
	* Returns a map of plugin names to their registered named endpoints,
	* and a map of plugin names to their client-exposed configs.
	*/
	async extendRoutes() {
		const endpoints = {};
		const pluginConfigs = {};
		const plugins = this.context?.getPlugins();
		if (!plugins || plugins.size === 0) return {
			endpoints,
			pluginConfigs
		};
		this.serverApplication.get("/health", (_, res) => {
			res.status(200).json({ status: "ok" });
		});
		this.registerEndpoint("health", "/health");
		for (const plugin of plugins.values()) {
			if (EXCLUDED_PLUGINS.includes(plugin.name)) continue;
			if (plugin?.injectRoutes && typeof plugin.injectRoutes === "function") {
				const router = express.Router();
				plugin.injectRoutes(router);
				const basePath = `/api/${plugin.name}`;
				this.serverApplication.use(basePath, router);
				endpoints[plugin.name] = plugin.getEndpoints();
				if (plugin.getSkipBodyParsingPaths && typeof plugin.getSkipBodyParsingPaths === "function") for (const p of plugin.getSkipBodyParsingPaths()) this.rawBodyPaths.add(p);
			}
			if (typeof plugin.clientConfig === "function") try {
				const raw = plugin.clientConfig();
				if (raw != null) {
					const sanitized = sanitizeClientConfig(plugin.name, raw);
					if (Object.keys(sanitized).length > 0) pluginConfigs[plugin.name] = sanitized;
				}
			} catch (error) {
				logger.error("Plugin '%s' clientConfig() failed, skipping its config: %O", plugin.name, error);
			}
		}
		return {
			endpoints,
			pluginConfigs
		};
	}
	/**
	* Setup frontend serving based on environment:
	* - If staticPath is explicitly provided: use static server
	* - Dev mode (no staticPath): Vite for HMR
	* - Production (no staticPath): Static files auto-detected
	*/
	async setupFrontend(endpoints, pluginConfigs) {
		const isDev = process.env.NODE_ENV === "development";
		if (this.config.staticPath !== void 0) {
			new StaticServer(this.serverApplication, this.config.staticPath, endpoints, pluginConfigs).setup();
			return;
		}
		if (isDev) {
			this.viteDevServer = new ViteDevServer(this.serverApplication, endpoints, pluginConfigs);
			await this.viteDevServer.setup();
			return;
		}
		const staticPath = ServerPlugin.findStaticPath();
		if (staticPath) new StaticServer(this.serverApplication, staticPath, endpoints, pluginConfigs).setup();
	}
	static findStaticPath() {
		const staticPaths = [
			"dist",
			"client/dist",
			"build",
			"public",
			"out"
		];
		const cwd = process.cwd();
		for (const p of staticPaths) {
			const fullPath = path.resolve(cwd, p);
			if (fs.existsSync(path.resolve(fullPath, "index.html"))) {
				logger.debug("Static files: serving from %s", fullPath);
				return fullPath;
			}
		}
	}
	/**
	* In development, prefers {@link ServerConfig.port} / env / default (8000), then
	* scans upward using `get-port`'s `portNumbers()` on the listen host until one binds.
	* In non-development, uses config / env / default only (no fallback).
	*/
	async resolveListenPort() {
		const requested = this.config.port ?? ServerPlugin.DEFAULT_CONFIG.port;
		if (process.env.NODE_ENV !== "development") {
			this.resolvedListenPort = requested;
			return requested;
		}
		const port = await getPort({
			host: this.config.host ?? ServerPlugin.DEFAULT_CONFIG.host,
			port: portNumbers(requested, Math.min(requested + devListenPortSpan - 1, 65535))
		});
		this.resolvedListenPort = port;
		if (port !== requested) logger.info("Port %d was busy, picking %d", requested, port);
		return port;
	}
	logStartupInfo() {
		const isDev = process.env.NODE_ENV === "development";
		const hasExplicitStaticPath = this.config.staticPath !== void 0;
		const port = this.resolvedListenPort ?? this.config.port ?? ServerPlugin.DEFAULT_CONFIG.port;
		const host = this.config.host ?? ServerPlugin.DEFAULT_CONFIG.host;
		logger.info("Server running on http://%s:%d", host, port);
		if (hasExplicitStaticPath) logger.info("Mode: static (%s)", this.config.staticPath);
		else if (isDev) logger.info("Mode: development (Vite HMR)");
		else logger.info("Mode: production (static)");
		const remoteServerController = this.remoteTunnelController;
		if (!remoteServerController) logger.debug("Remote tunnel: disabled (controller not initialized)");
		else logger.debug("Remote tunnel: %s; %s", remoteServerController.isAllowedByEnv() ? "allowed" : "blocked", remoteServerController.isActive() ? "active" : "inactive");
	}
	async _gracefulShutdown() {
		logger.info("Starting graceful shutdown...");
		if (this.viteDevServer) await this.viteDevServer.close();
		if (this.remoteTunnelController) this.remoteTunnelController.cleanup();
		TelemetryReporter.getInstance()?.stop();
		const shutdownPlugins = this.context?.getPlugins();
		if (shutdownPlugins) {
			for (const plugin of shutdownPlugins.values()) if (plugin.abortActiveOperations) try {
				plugin.abortActiveOperations();
			} catch (err) {
				logger.error("Error aborting operations for plugin %s: %O", plugin.name, err);
			}
		}
		if (this.server) {
			this.server.close(() => {
				logger.debug("Server closed gracefully");
				process.exit(0);
			});
			setTimeout(() => {
				logger.debug("Force shutdown after timeout");
				process.exit(1);
			}, 15e3);
		} else process.exit(0);
	}
	/**
	* Returns the public exports for the server plugin.
	* Exposes server management methods.
	*/
	exports() {
		const self = this;
		return {
			extend(fn) {
				self.extend(fn);
				return this;
			},
			getServer: this.getServer,
			getConfig: this.getConfig,
			start() {
				throw new ServerError("server.start() has been removed. Use the onPluginsReady callback instead:\n\n  createApp({\n    plugins: [server(), ...],\n    onPluginsReady(appkit) {\n      appkit.server.extend(...);\n    },\n  });\n\nRun `npx appkit codemod on-plugins-ready --write` to auto-migrate.");
			}
		};
	}
};
const EXCLUDED_PLUGINS = [ServerPlugin.manifest.name];
/** @internal Exported for unit tests. */
function requestMetricsMiddleware(req, res, next) {
	const startMs = Date.now();
	res.on("finish", () => {
		const reporter = TelemetryReporter.getInstance();
		if (!reporter) return;
		const routePath = req.route?.path;
		if (!routePath) return;
		const template = `${req.baseUrl ?? ""}${routePath}`;
		reporter.recordRequest(req.method, template, res.statusCode, Date.now() - startMs);
	});
	next();
}
/**
* @internal
*/
const server = toPlugin(ServerPlugin);

//#endregion
export { server };
