// @databricks/appkit/dist/plugins/server/utils.js
import crypto from "node:crypto";
import pc from "picocolors";
import path from "node:path";
import fs from "node:fs";

//#region src/plugins/server/utils.ts
function parseCookies(req) {
	const cookieHeader = req.headers.cookie;
	if (!cookieHeader) return {};
	if (cookieHeader.indexOf(";") === -1) {
		const eqIndex = cookieHeader.indexOf("=");
		if (eqIndex === -1) return {};
		return { [cookieHeader.slice(0, eqIndex).trim()]: cookieHeader.slice(eqIndex + 1) };
	}
	const cookies = {};
	const parts = cookieHeader.split(";");
	for (let i = 0; i < parts.length; i++) {
		const eqIndex = parts[i].indexOf("=");
		if (eqIndex !== -1) {
			const key = parts[i].slice(0, eqIndex).trim();
			cookies[key] = parts[i].slice(eqIndex + 1);
		}
	}
	return cookies;
}
function generateTunnelIdFromEmail(email) {
	if (!email) return void 0;
	return crypto.createHash("sha256").update(email).digest("base64url").slice(0, 8);
}
function getRoutes(stack, basePath = "") {
	const routes = [];
	stack.forEach((layer) => {
		if (layer.route) {
			const path = basePath + layer.route.path;
			const methods = Object.keys(layer.route.methods).map((m) => m.toUpperCase());
			routes.push({
				path,
				methods
			});
		} else if (layer.name === "router" && layer.handle.stack) {
			const nestedBase = basePath + layer.regexp.source.replace("^\\", "").replace("\\/?(?=\\/|$)", "").replace(/\\\//g, "/").replace(/\$$/, "") || "";
			routes.push(...getRoutes(layer.handle.stack, nestedBase));
		}
	});
	return routes;
}
const METHOD_COLORS = {
	GET: pc.green,
	POST: pc.blue,
	PUT: pc.yellow,
	PATCH: pc.yellow,
	DELETE: pc.red,
	HEAD: pc.magenta,
	OPTIONS: pc.magenta
};
function printRoutes(routes) {
	if (routes.length === 0) return;
	const rows = routes.flatMap((r) => r.methods.map((m) => ({
		method: m,
		path: r.path
	}))).sort((a, b) => a.method.localeCompare(b.method) || a.path.localeCompare(b.path));
	const maxMethodLen = Math.max(...rows.map((r) => r.method.length));
	const separator = pc.dim("─".repeat(50));
	const colorizeParams = (p) => p.replace(/(:[a-zA-Z_]\w*)/g, (match) => pc.cyan(match));
	console.log("");
	console.log(`  ${pc.bold("Registered Routes")} ${pc.dim(`(${rows.length})`)}`);
	console.log(`  ${separator}`);
	for (const { method, path } of rows) {
		const methodStr = (METHOD_COLORS[method] || pc.white)(pc.bold(method.padEnd(maxMethodLen)));
		console.log(`  ${methodStr}  ${colorizeParams(path)}`);
	}
	console.log(`  ${separator}`);
	console.log("");
}
function getQueries(configFolder) {
	const queriesFolder = path.join(configFolder, "queries");
	if (!fs.existsSync(queriesFolder)) return {};
	return Object.fromEntries(fs.readdirSync(queriesFolder).filter((f) => path.extname(f) === ".sql").map((f) => [path.basename(f, ".sql"), path.basename(f, ".sql")]));
}
const APPKIT_CONFIG_SCRIPT_ID = "__appkit__";
const EMPTY_RUNTIME_CONFIG_JSON = JSON.stringify({
	appName: "",
	queries: {},
	endpoints: {},
	plugins: {}
});
const JSON_SCRIPT_ESCAPE_MAP = {
	"<": "\\u003c",
	">": "\\u003e",
	"&": "\\u0026",
	"\u2028": "\\u2028",
	"\u2029": "\\u2029"
};
function getRuntimeConfig(endpoints = {}, pluginConfigs = {}) {
	const configFolder = path.join(process.cwd(), "config");
	return {
		appName: process.env.DATABRICKS_APP_NAME || "",
		queries: getQueries(configFolder),
		endpoints,
		plugins: pluginConfigs
	};
}
function getConfigScript(endpoints = {}, pluginConfigs = {}) {
	return `
    <script id="${APPKIT_CONFIG_SCRIPT_ID}" type="application/json">
      ${serializeRuntimeConfig(getRuntimeConfig(endpoints, pluginConfigs))}
    <\/script>
    <script>
      window.__appkit__ = JSON.parse(
        document.getElementById("${APPKIT_CONFIG_SCRIPT_ID}")?.textContent ||
          '${EMPTY_RUNTIME_CONFIG_JSON}',
      );
    <\/script>
  `;
}
function serializeRuntimeConfig(config) {
	return JSON.stringify(config).replace(/[<>&\u2028\u2029]/g, (char) => JSON_SCRIPT_ESCAPE_MAP[char] ?? char);
}

//#endregion
export { generateTunnelIdFromEmail, getConfigScript, getRoutes, parseCookies, printRoutes };
