// @databricks/appkit/dist/plugins/server/vite-dev-server.js
import { createLogger } from "../../logging/logger.js";
import { ServerError } from "../../errors/server.js";
import "../../errors/index.js";
import { mergeConfigDedup } from "../../utils/vite-config-merge.js";
import { BaseServer } from "./base-server.js";
import { appKitServingTypesPlugin } from "../../type-generator/serving/vite-plugin.js";
import { appKitTypesPlugin } from "../../type-generator/vite-plugin.js";
import { reactSourceLocPlugin } from "./react-source-loc-vite-plugin.js";
import path from "node:path";
import fs from "node:fs";

//#region src/plugins/server/vite-dev-server.ts
const logger = createLogger("server:vite");
/**
* Vite dev server for the AppKit.
*
* This class is responsible for serving the Vite dev server for the development server.
* It also handles the index.html file for the development server.
*
* @example
* ```ts
* const viteDevServer = new ViteDevServer(app, endpoints);
* await viteDevServer.setup();
* ```
*/
var ViteDevServer = class extends BaseServer {
	vite;
	constructor(app, endpoints = {}, pluginConfigs = {}) {
		super(app, endpoints, pluginConfigs);
		this.vite = null;
	}
	/**
	* Setup the Vite dev server.
	*
	* This method sets up the Vite dev server and the index.html file for the development server.
	*
	* @returns
	*/
	async setup() {
		const { createServer: createViteServer, loadConfigFromFile, mergeConfig } = await import("vite");
		const react = await import("@vitejs/plugin-react");
		const clientRoot = this.findClientRoot();
		const projectRoot = process.cwd();
		const userConfig = (await loadConfigFromFile({
			mode: "development",
			command: "serve"
		}, void 0, clientRoot))?.config ?? {};
		const viteClientPort = process.env.VITE_CLIENT_PORT;
		this.vite = await createViteServer(mergeConfigDedup(userConfig, {
			configFile: false,
			root: clientRoot,
			server: {
				middlewareMode: true,
				...viteClientPort ? { hmr: { clientPort: viteClientPort } } : {},
				watch: {
					useFsEvents: true,
					ignored: ["**/node_modules/**", "!**/node_modules/@databricks/**"]
				}
			},
			plugins: [
				react.default(),
				reactSourceLocPlugin({ projectRoot }),
				appKitTypesPlugin(),
				appKitServingTypesPlugin()
			],
			appType: "custom"
		}, mergeConfig));
		this.app.use(this.vite.middlewares);
		this.app.use("*", async (req, res, next) => {
			if (req.originalUrl.startsWith("/api") || req.originalUrl.startsWith("/query")) return next();
			const vite = this.vite;
			this.validateVite(vite);
			try {
				const indexPath = path.resolve(clientRoot, "index.html");
				let html = fs.readFileSync(indexPath, "utf-8");
				html = html.replace("<body>", `<body>${this.getConfigScript()}`);
				html = await vite.transformIndexHtml(req.originalUrl, html);
				res.status(200).set({ "Content-Type": "text/html" }).end(html);
			} catch (e) {
				vite.ssrFixStacktrace(e);
				next(e);
			}
		});
	}
	/** Close the Vite dev server. */
	async close() {
		await this.vite?.close();
	}
	/** Find the client root. */
	findClientRoot() {
		const cwd = process.cwd();
		const candidates = [
			"client",
			"src",
			"app",
			"frontend",
			"."
		];
		for (const dir of candidates) {
			const fullPath = path.resolve(cwd, dir);
			const hasViteConfig = fs.existsSync(path.join(fullPath, "vite.config.ts")) || fs.existsSync(path.join(fullPath, "vite.config.js"));
			const hasIndexHtml = fs.existsSync(path.join(fullPath, "index.html"));
			if (hasViteConfig && hasIndexHtml) {
				logger.debug("Vite dev server: using client root %s", fullPath);
				return fullPath;
			}
		}
		throw ServerError.clientDirectoryNotFound(candidates);
	}
	validateVite(vite) {
		if (!vite) throw ServerError.viteNotInitialized();
	}
};

//#endregion
export { ViteDevServer };
