// @databricks/sdk-experimental/dist/apis/files/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilesService = exports.FilesError = exports.FilesRetriableError = exports.DbfsService = exports.DbfsError = exports.DbfsRetriableError = void 0;
const files = __importStar(require("./model"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
class DbfsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Dbfs", method, message);
    }
}
exports.DbfsRetriableError = DbfsRetriableError;
class DbfsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Dbfs", method, message);
    }
}
exports.DbfsError = DbfsError;
/**
 * DBFS API makes it simple to interact with various data sources without having
 * to include a users credentials every time to read a file.
 */
class DbfsService {
    constructor(client) {
        this.client = client;
    }
    async _addBlock(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("data")) {
            body["data"] = request.data;
        }
        if (request.hasOwnProperty("handle")) {
            body["handle"] = request.handle;
        }
        const query = {};
        const path = "/api/2.0/dbfs/add-block";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Appends a block of data to the stream specified by the input handle. If
     * the handle does not exist, this call will throw an exception with
     * ``RESOURCE_DOES_NOT_EXIST``.
     *
     * If the block of data exceeds 1 MB, this call will throw an exception with
     * ``MAX_BLOCK_SIZE_EXCEEDED``.
     */
    async addBlock(request, context) {
        return await this._addBlock(request, context);
    }
    async _close(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("handle")) {
            body["handle"] = request.handle;
        }
        const query = {};
        const path = "/api/2.0/dbfs/close";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Closes the stream specified by the input handle. If the handle does not
     * exist, this call throws an exception with ``RESOURCE_DOES_NOT_EXIST``.
     */
    async close(request, context) {
        return await this._close(request, context);
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("overwrite")) {
            body["overwrite"] = request.overwrite;
        }
        if (request.hasOwnProperty("path")) {
            body["path"] = request.path;
        }
        const query = {};
        const path = "/api/2.0/dbfs/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Opens a stream to write to a file and returns a handle to this stream.
     * There is a 10 minute idle timeout on this handle. If a file or directory
     * already exists on the given path and __overwrite__ is set to false, this
     * call will throw an exception with ``RESOURCE_ALREADY_EXISTS``.
     *
     * A typical workflow for file upload would be:
     *
     * 1. Issue a ``create`` call and get a handle. 2. Issue one or more
     * ``add-block`` calls with the handle you have. 3. Issue a ``close`` call
     * with the handle you have.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("path")) {
            body["path"] = request.path;
        }
        if (request.hasOwnProperty("recursive")) {
            body["recursive"] = request.recursive;
        }
        const query = {};
        const path = "/api/2.0/dbfs/delete";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete the file or directory (optionally recursively delete all files in
     * the directory). This call throws an exception with `IO_ERROR` if the path
     * is a non-empty directory and `recursive` is set to `false` or on other
     * similar errors.
     *
     * When you delete a large number of files, the delete operation is done in
     * increments. The call returns a response after approximately 45 seconds
     * with an error message (503 Service Unavailable) asking you to re-invoke
     * the delete operation until the directory structure is fully deleted.
     *
     * For operations that delete more than 10K files, we discourage using the
     * DBFS REST API, but advise you to perform such operations in the context of
     * a cluster, using the [File system utility
     * (dbutils.fs)](/dev-tools/databricks-utils.html#dbutils-fs). `dbutils.fs`
     * covers the functional scope of the DBFS REST API, but from notebooks.
     * Running such operations using notebooks provides better control and
     * manageability, such as selective deletes, and the possibility to automate
     * periodic delete jobs.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _getStatus(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("path")) {
            query["path"] = request.path;
        }
        const path = "/api/2.0/dbfs/get-status";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the file information for a file or directory. If the file or
     * directory does not exist, this call throws an exception with
     * `RESOURCE_DOES_NOT_EXIST`.
     */
    async getStatus(request, context) {
        return await this._getStatus(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("path")) {
            query["path"] = request.path;
        }
        const path = "/api/2.0/dbfs/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List the contents of a directory, or details of the file. If the file or
     * directory does not exist, this call throws an exception with
     * `RESOURCE_DOES_NOT_EXIST`.
     *
     * When calling list on a large directory, the list operation will time out
     * after approximately 60 seconds. We strongly recommend using list only on
     * directories containing less than 10K files and discourage using the DBFS
     * REST API for operations that list more than 10K files. Instead, we
     * recommend that you perform such operations in the context of a cluster,
     * using the [File system utility
     * (dbutils.fs)](/dev-tools/databricks-utils.html#dbutils-fs), which provides
     * the same functionality without timing out.
     */
    async *list(request, context) {
        const response = (await this._list(request, context)).files;
        for (const v of response || []) {
            yield v;
        }
    }
    async _mkdirs(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("path")) {
            body["path"] = request.path;
        }
        const query = {};
        const path = "/api/2.0/dbfs/mkdirs";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates the given directory and necessary parent directories if they do
     * not exist. If a file (not a directory) exists at any prefix of the input
     * path, this call throws an exception with `RESOURCE_ALREADY_EXISTS`.
     * **Note**: If this operation fails, it might have succeeded in creating
     * some of the necessary parent directories.
     */
    async mkdirs(request, context) {
        return await this._mkdirs(request, context);
    }
    async _move(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("destination_path")) {
            body["destination_path"] = request.destination_path;
        }
        if (request.hasOwnProperty("source_path")) {
            body["source_path"] = request.source_path;
        }
        const query = {};
        const path = "/api/2.0/dbfs/move";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Moves a file from one location to another location within DBFS. If the
     * source file does not exist, this call throws an exception with
     * `RESOURCE_DOES_NOT_EXIST`. If a file already exists in the destination
     * path, this call throws an exception with `RESOURCE_ALREADY_EXISTS`. If the
     * given source path is a directory, this call always recursively moves all
     * files.
     */
    async move(request, context) {
        return await this._move(request, context);
    }
    async _put(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("contents")) {
            body["contents"] = request.contents;
        }
        if (request.hasOwnProperty("overwrite")) {
            body["overwrite"] = request.overwrite;
        }
        if (request.hasOwnProperty("path")) {
            body["path"] = request.path;
        }
        const query = {};
        const path = "/api/2.0/dbfs/put";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Uploads a file through the use of multipart form post. It is mainly used
     * for streaming uploads, but can also be used as a convenient single call
     * for data upload.
     *
     * Alternatively you can pass contents as base64 string.
     *
     * The amount of data that can be passed (when not streaming) using the
     * __contents__ parameter is limited to 1 MB. `MAX_BLOCK_SIZE_EXCEEDED` will
     * be thrown if this limit is exceeded.
     *
     * If you want to upload large files, use the streaming upload. For details,
     * see :method:dbfs/create, :method:dbfs/addBlock, :method:dbfs/close.
     */
    async put(request, context) {
        return await this._put(request, context);
    }
    async _read(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("length")) {
            query["length"] = request.length;
        }
        if (request.hasOwnProperty("offset")) {
            query["offset"] = request.offset;
        }
        if (request.hasOwnProperty("path")) {
            query["path"] = request.path;
        }
        const path = "/api/2.0/dbfs/read";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns the contents of a file. If the file does not exist, this call
     * throws an exception with `RESOURCE_DOES_NOT_EXIST`. If the path is a
     * directory, the read length is negative, or if the offset is negative, this
     * call throws an exception with `INVALID_PARAMETER_VALUE`. If the read
     * length exceeds 1 MB, this call throws an exception with
     * `MAX_READ_SIZE_EXCEEDED`.
     *
     * If `offset + length` exceeds the number of bytes in a file, it reads the
     * contents until the end of file.
     */
    async read(request, context) {
        return await this._read(request, context);
    }
}
exports.DbfsService = DbfsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "_addBlock", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "addBlock", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "_close", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "close", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "_getStatus", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "getStatus", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], DbfsService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "_mkdirs", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "mkdirs", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "_move", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "move", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "_put", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "put", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "_read", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbfsService.prototype, "read", null);
class FilesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Files", method, message);
    }
}
exports.FilesRetriableError = FilesRetriableError;
class FilesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Files", method, message);
    }
}
exports.FilesError = FilesError;
/**
 * The Files API is a standard HTTP API that allows you to read, write, list, and
 * delete files and directories by referring to their URI. The API makes working
 * with file content as raw bytes easier and more efficient.
 *
 * The API supports [Unity Catalog volumes], where files and directories to
 * operate on are specified using their volume URI path, which follows the format
 * /Volumes/&lt;catalog_name&gt;/&lt;schema_name&gt;/&lt;volume_name&gt;/&lt;path_to_file&gt;.
 *
 * The Files API has two distinct endpoints, one for working with files
 * (`/fs/files`) and another one for working with directories
 * (`/fs/directories`). Both endpoints use the standard HTTP methods GET, HEAD,
 * PUT, and DELETE to manage files and directories specified using their URI
 * path. The path is always absolute.
 *
 * Some Files API client features are currently experimental. To enable them, set
 * `enable_experimental_files_api_client = True` in your configuration profile or
 * use the environment variable
 * `DATABRICKS_ENABLE_EXPERIMENTAL_FILES_API_CLIENT=True`.
 *
 * Use of Files API may incur Databricks data transfer charges.
 *
 * [Unity Catalog volumes]: https://docs.databricks.com/en/connect/unity-catalog/volumes.html
 */
class FilesService {
    constructor(client) {
        this.client = client;
    }
    async _createDirectory(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/fs/directories${request.directory_path}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates an empty directory. If necessary, also creates any parent
     * directories of the new, empty directory (like the shell command `mkdir
     * -p`). If called on an existing directory, returns a success response; this
     * method is idempotent (it will succeed if the directory already exists).
     */
    async createDirectory(request, context) {
        return await this._createDirectory(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/fs/files${request.file_path}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a file. If the request is successful, there is no response body.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _deleteDirectory(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/fs/directories${request.directory_path}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes an empty directory.
     *
     * To delete a non-empty directory, first delete all of its contents. This
     * can be done by listing the directory contents and deleting each file and
     * subdirectory recursively.
     */
    async deleteDirectory(request, context) {
        return await this._deleteDirectory(request, context);
    }
    async _download(request, context) {
        const headers = new Headers({ Accept: "application/octet-stream" });
        const body = undefined;
        const query = {};
        const responseHeaders = [
            "content-length",
            "content-type",
            "last-modified",
        ];
        const path = `/api/2.0/fs/files${request.file_path}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            responseHeaders,
            raw: true,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Downloads a file. The file contents are the response body. This is a
     * standard HTTP file download, not a JSON RPC. It supports the Range and
     * If-Unmodified-Since HTTP headers.
     */
    async download(request, context) {
        return await this._download(request, context);
    }
    async _getDirectoryMetadata(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/fs/directories${request.directory_path}`;
        return (await this.client.request({
            path,
            method: "HEAD",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get the metadata of a directory. The response HTTP headers contain the
     * metadata. There is no response body.
     *
     * This method is useful to check if a directory exists and the caller has
     * access to it.
     *
     * If you wish to ensure the directory exists, you can instead use `PUT`,
     * which will create the directory if it does not exist, and is idempotent
     * (it will succeed if the directory already exists).
     */
    async getDirectoryMetadata(request, context) {
        return await this._getDirectoryMetadata(request, context);
    }
    async _getMetadata(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const responseHeaders = [
            "content-length",
            "content-type",
            "last-modified",
        ];
        const path = `/api/2.0/fs/files${request.file_path}`;
        return (await this.client.request({
            path,
            method: "HEAD",
            headers,
            responseHeaders,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get the metadata of a file. The response HTTP headers contain the
     * metadata. There is no response body.
     */
    async getMetadata(request, context) {
        return await this._getMetadata(request, context);
    }
    async _listDirectoryContents(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/fs/directories${request.directory_path}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns the contents of a directory. If there is no directory at the
     * specified path, the API returns a HTTP 404 error.
     */
    async *listDirectoryContents(request, context) {
        while (true) {
            const response = await this._listDirectoryContents(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.contents || response.contents.length === 0) {
                break;
            }
            for (const v of response.contents) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _upload(request, context) {
        const headers = new Headers({
            "Content-Type": "application/octet-stream",
        });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("overwrite")) {
            query["overwrite"] =
                request.overwrite;
        }
        const path = `/api/2.0/fs/files${request.file_path}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Uploads a file of up to 5 GiB. The file contents should be sent as the
     * request body as raw bytes (an octet stream); do not encode or otherwise
     * modify the bytes before sending. The contents of the resulting file will
     * be exactly the bytes sent in the request body. If the request is
     * successful, there is no response body.
     */
    async upload(request, context) {
        return await this._upload(request, context);
    }
}
exports.FilesService = FilesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "_createDirectory", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "createDirectory", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "_deleteDirectory", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "deleteDirectory", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "_download", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "download", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "_getDirectoryMetadata", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "getDirectoryMetadata", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "_getMetadata", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "getMetadata", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "_listDirectoryContents", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], FilesService.prototype, "listDirectoryContents", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "_upload", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FilesService.prototype, "upload", null);
