// @databricks/appkit-ui/dist/react/hooks/use-chart-data.js
import { useAnalyticsQuery } from "./use-analytics-query.js";
import { useMemo } from "react";

//#region src/react/hooks/use-chart-data.ts
/** Threshold for auto-selecting Arrow format (row count hint) */
const ARROW_THRESHOLD = 500;
/**
* Resolves the data format based on hints and preferences
*/
function resolveFormat(format, parameters) {
	if (format === "json") return "JSON_ARRAY";
	if (format === "arrow") return "ARROW_STREAM";
	if (format === "auto") {
		if (parameters?._preferArrow === true) return "ARROW_STREAM";
		if (parameters?._preferJson === true) return "JSON_ARRAY";
		const limit = parameters?.limit;
		if (typeof limit === "number" && limit > ARROW_THRESHOLD) return "ARROW_STREAM";
		if (parameters?.startDate && parameters?.endDate) return "ARROW_STREAM";
		return "JSON_ARRAY";
	}
	return "JSON_ARRAY";
}
/**
* Hook for fetching chart data in either JSON or Arrow format.
* Automatically selects the best format based on query hints.
*
* @example
* ```tsx
* // Auto-select format
* const { data, isArrow, loading } = useChartData({
*   queryKey: "spend_data",
*   parameters: { limit: 1000 }
* });
*
* // Force Arrow format
* const { data } = useChartData({
*   queryKey: "big_query",
*   format: "arrow"
* });
* ```
*/
function useChartData(options) {
	const { queryKey, parameters, format = "auto", transformer } = options;
	const resolvedFormat = useMemo(() => resolveFormat(format, parameters), [format, parameters]);
	const isArrowFormat = resolvedFormat === "ARROW_STREAM";
	const { data: rawData, loading, error } = useAnalyticsQuery(queryKey, parameters, {
		autoStart: true,
		format: resolvedFormat
	});
	const processedData = useMemo(() => {
		if (!rawData) return null;
		if (transformer) try {
			return transformer(rawData);
		} catch (err) {
			console.error("[useChartData] Transformer error:", err);
			return rawData;
		}
		return rawData;
	}, [rawData, transformer]);
	const isEmpty = useMemo(() => {
		if (!processedData) return true;
		if (typeof processedData === "object" && "numRows" in processedData && typeof processedData.numRows === "number") return processedData.numRows === 0;
		if (Array.isArray(processedData)) return processedData.length === 0;
		return true;
	}, [processedData]);
	return {
		data: processedData,
		isArrow: useMemo(() => {
			if (!processedData) return isArrowFormat;
			return typeof processedData === "object" && processedData !== null && "schema" in processedData && "numRows" in processedData && typeof processedData.getChild === "function";
		}, [processedData, isArrowFormat]),
		loading,
		error,
		isEmpty
	};
}

//#endregion
export { useChartData };
