// @databricks/appkit-ui/dist/react/hooks/use-serving-stream.js
import { connectSSE } from "../../js/sse/connect-sse.js";
import "../../js/index.js";
import { usePluginClientConfig } from "./use-plugin-config.js";
import { useCallback, useEffect, useMemo, useRef, useState } from "react";

//#region src/react/hooks/use-serving-stream.ts
/**
* Hook for streaming invocation of a serving endpoint via SSE.
* Calls `POST /api/serving/stream` (default) or `POST /api/serving/{alias}/stream` (named).
* Accumulates parsed chunks in state.
*
* When the type generator has populated `ServingEndpointRegistry`, the chunk type
* is automatically inferred from the endpoint's OpenAPI schema.
*/
function useServingStream(body, options = {}) {
	const { alias, autoStart = false, onComplete } = options;
	const config = usePluginClientConfig("serving");
	const aliasError = useMemo(() => {
		if (!alias || !config.aliases) return null;
		const aliasStr = String(alias);
		if (!config.aliases.includes(aliasStr)) return `Unknown serving alias "${aliasStr}". Available: ${config.aliases.join(", ")}`;
		return null;
	}, [alias, config.aliases]);
	const [chunks, setChunks] = useState([]);
	const [streaming, setStreaming] = useState(false);
	const [error, setError] = useState(aliasError);
	const abortControllerRef = useRef(null);
	const chunksRef = useRef([]);
	const onCompleteRef = useRef(onComplete);
	onCompleteRef.current = onComplete;
	const urlSuffix = alias ? `/api/serving/${encodeURIComponent(String(alias))}/stream` : "/api/serving/stream";
	const reset = useCallback(() => {
		abortControllerRef.current?.abort();
		abortControllerRef.current = null;
		chunksRef.current = [];
		setChunks([]);
		setStreaming(false);
		setError(null);
	}, []);
	const bodyJson = JSON.stringify(body);
	const stream = useCallback((overrideBody) => {
		if (aliasError) {
			setError(aliasError);
			return;
		}
		abortControllerRef.current?.abort();
		setStreaming(true);
		setError(null);
		setChunks([]);
		chunksRef.current = [];
		const abortController = new AbortController();
		abortControllerRef.current = abortController;
		connectSSE({
			url: urlSuffix,
			payload: overrideBody ? JSON.stringify(overrideBody) : bodyJson,
			signal: abortController.signal,
			onMessage: async (message) => {
				if (abortController.signal.aborted) return;
				try {
					const parsed = JSON.parse(message.data);
					chunksRef.current = [...chunksRef.current, parsed];
					setChunks(chunksRef.current);
				} catch {}
			},
			onError: (err) => {
				if (abortController.signal.aborted) return;
				setStreaming(false);
				setError(err instanceof Error ? err.message : "Streaming failed");
			}
		}).then(() => {
			if (abortController.signal.aborted) return;
			setStreaming(false);
			onCompleteRef.current?.(chunksRef.current);
		}).catch(() => {
			if (abortController.signal.aborted) return;
			setStreaming(false);
			setError("Connection error");
		});
	}, [
		urlSuffix,
		bodyJson,
		aliasError
	]);
	useEffect(() => {
		if (autoStart) stream();
		return () => {
			abortControllerRef.current?.abort();
		};
	}, [stream, autoStart]);
	return {
		stream,
		chunks,
		streaming,
		error,
		reset
	};
}

//#endregion
export { useServingStream };
