// @databricks/appkit-ui/dist/react/hooks/use-analytics-query.js
import { ArrowClient } from "../../js/arrow/arrow-client.js";
import { connectSSE } from "../../js/sse/connect-sse.js";
import "../../js/index.js";
import { useQueryHMR } from "./use-query-hmr.js";
import { useCallback, useEffect, useMemo, useRef, useState } from "react";

//#region src/react/hooks/use-analytics-query.ts
/**
* Shallow structural equality for analytics query parameter objects.
*
* Analytics query parameters are produced by the `sql.*` builders and are
* always plain objects keyed to primitive values (string | number | boolean
* | null | undefined), so shallow equality is sufficient and substantially
* cheaper than a full deep-equal.
*/
function shallowEqualParams(a, b) {
	if (Object.is(a, b)) return true;
	if (a === null || b === null || typeof a !== "object" || typeof b !== "object") return false;
	const aKeys = Object.keys(a);
	const bKeys = Object.keys(b);
	if (aKeys.length !== bKeys.length) return false;
	for (const key of aKeys) {
		if (!Object.hasOwn(b, key)) return false;
		if (!Object.is(a[key], b[key])) return false;
	}
	return true;
}
/**
* Stabilize a value's identity across renders when it is structurally equal
* to the previous value. Used to make object-literal parameters safe to pass
* directly to `useAnalyticsQuery` without forcing every consumer to wrap
* params in `useMemo`.
*/
function useStableParams(value) {
	const ref = useRef(value);
	if (!shallowEqualParams(ref.current, value)) ref.current = value;
	return ref.current;
}
function getDevMode() {
	const dev = new URL(window.location.href).searchParams.get("dev");
	return dev ? `?dev=${dev}` : "";
}
function getArrowStreamUrl(id) {
	return `/api/analytics/arrow-result/${id}`;
}
/**
* Subscribe to an analytics query over SSE and returns its latest result.
* Integration hook between client and analytics plugin.
*
* The return type is automatically inferred based on the format:
* - `format: "JSON_ARRAY"` (default): Returns typed array from QueryRegistry
* - `format: "ARROW_STREAM"`: Returns TypedArrowTable with row type preserved
*
* Note: User context execution is determined by query file naming:
* - `queryKey.obo.sql`: Executes as user (OBO = on-behalf-of / user delegation)
* - `queryKey.sql`: Executes as service principal
*
* @param queryKey - Analytics query identifier
* @param parameters - Query parameters (type-safe based on QueryRegistry)
* @param options - Analytics query settings including format
* @returns Query result state with format-appropriate data type
*
* @example JSON format (default)
* ```typescript
* const { data } = useAnalyticsQuery("spend_data", params);
* // data: Array<{ group_key: string; cost_usd: number; ... }> | null
* ```
*
* @example Arrow format
* ```typescript
* const { data } = useAnalyticsQuery("spend_data", params, { format: "ARROW_STREAM" });
* // data: TypedArrowTable<{ group_key: string; cost_usd: number; ... }> | null
* ```
*/
function useAnalyticsQuery(queryKey, parameters, options = {}) {
	const format = options?.format ?? "JSON_ARRAY";
	const maxParametersSize = options?.maxParametersSize ?? 100 * 1024;
	const autoStart = options?.autoStart ?? true;
	const devMode = getDevMode();
	const urlSuffix = `/api/analytics/query/${encodeURIComponent(queryKey)}${devMode}`;
	const [data, setData] = useState(null);
	const [loading, setLoading] = useState(false);
	const [error, setError] = useState(null);
	const abortControllerRef = useRef(null);
	if (!queryKey || queryKey.trim().length === 0) throw new Error("useAnalyticsQuery: 'queryKey' must be a non-empty string.");
	const stableParameters = useStableParams(parameters);
	const payload = useMemo(() => {
		try {
			const serialized = JSON.stringify({
				parameters: stableParameters,
				format
			});
			if (new Blob([serialized]).size > maxParametersSize) throw new Error("useAnalyticsQuery: Parameters size exceeds the maximum allowed size");
			return serialized;
		} catch (error) {
			console.error("useAnalyticsQuery: Failed to serialize parameters", error);
			return null;
		}
	}, [
		stableParameters,
		format,
		maxParametersSize
	]);
	const start = useCallback(() => {
		if (payload === null) {
			setError("Failed to serialize query parameters");
			return;
		}
		if (abortControllerRef.current) abortControllerRef.current.abort();
		setLoading(true);
		setError(null);
		setData(null);
		const abortController = new AbortController();
		abortControllerRef.current = abortController;
		connectSSE({
			url: urlSuffix,
			payload,
			signal: abortController.signal,
			onMessage: async (message) => {
				try {
					const parsed = JSON.parse(message.data);
					if (parsed.type === "result") {
						setLoading(false);
						setData(parsed.data);
						return;
					}
					if (parsed.type === "arrow") try {
						const arrowData = await ArrowClient.fetchArrow(getArrowStreamUrl(parsed.statement_id));
						const table = await ArrowClient.processArrowBuffer(arrowData);
						setLoading(false);
						setData(table);
						return;
					} catch (error) {
						console.error("[useAnalyticsQuery] Failed to fetch Arrow data", error);
						setLoading(false);
						setError("Unable to load data, please try again");
						return;
					}
					if (parsed.type === "error" || parsed.error || parsed.code) {
						const errorMsg = parsed.error || parsed.message || "Unable to execute query";
						setLoading(false);
						setError(errorMsg);
						if (parsed.code) console.error(`[useAnalyticsQuery] Code: ${parsed.code}, Message: ${errorMsg}`);
						return;
					}
				} catch (error) {
					console.warn("[useAnalyticsQuery] Malformed message received", error);
				}
			},
			onError: (error) => {
				if (abortController.signal.aborted) return;
				setLoading(false);
				let userMessage = "Unable to load data, please try again";
				if (error instanceof Error) {
					if (error.name === "AbortError") userMessage = "Request timed out, please try again";
					else if (error.message.includes("Failed to fetch")) userMessage = "Network error. Please check your connection.";
					console.error("[useAnalyticsQuery] Error", {
						queryKey,
						error: error.message,
						stack: error.stack
					});
				}
				setError(userMessage);
			}
		});
	}, [
		queryKey,
		payload,
		urlSuffix
	]);
	useEffect(() => {
		if (autoStart) start();
		return () => {
			abortControllerRef.current?.abort();
		};
	}, [start, autoStart]);
	useQueryHMR(queryKey, start);
	return {
		data,
		loading,
		error
	};
}

//#endregion
export { useAnalyticsQuery };
