// @databricks/appkit-ui/dist/react/hooks/use-agent-chat.js
import { connectSSE } from "../../js/sse/connect-sse.js";
import "../../js/index.js";
import { useCallback, useEffect, useRef, useState } from "react";

//#region src/react/hooks/use-agent-chat.ts
/**
* React hook for chatting with an agent registered via the `agents()`
* plugin. Wraps {@link connectSSE} (which owns the buffer cap, abort
* composition, retry/backoff, and frame parsing) with the small amount
* of stateful glue every chat UI needs: accumulated assistant text,
* thread id, streaming flag, and an event callback.
*
* The hook is intentionally lower-level than a full chat component —
* it owns one stream at a time, not a multi-turn message history. The
* caller composes its own messages array (typically a `useState`) and
* appends to it via the `onEvent` callback for tool calls and via the
* `content` field for assistant text.
*
* @example
* ```tsx
* function Chat({ agent }: { agent: string }) {
*   const [messages, setMessages] = useState<Message[]>([]);
*   const { content, threadId, isStreaming, send, reset } = useAgentChat({
*     agent,
*     onEvent(ev) {
*       if (ev.type === "response.output_item.added" && ev.item?.type === "function_call") {
*         setMessages((m) => [...m, { role: "tool", name: ev.item?.name, args: ev.item?.arguments }]);
*       }
*     },
*   });
*   // `content` reflects the latest assistant turn; reset() between conversations.
*   // ...
* }
* ```
*/
function useAgentChat({ agent, endpoint = "/api/agents/chat", onEvent }) {
	const [content, setContent] = useState("");
	const [events, setEvents] = useState([]);
	const [threadId, setThreadId] = useState(null);
	const [isStreaming, setIsStreaming] = useState(false);
	const [error, setError] = useState(null);
	const threadIdRef = useRef(null);
	const contentRef = useRef("");
	const onEventRef = useRef(onEvent);
	onEventRef.current = onEvent;
	const abortControllerRef = useRef(null);
	const reset = useCallback(() => {
		abortControllerRef.current?.abort();
		abortControllerRef.current = null;
		threadIdRef.current = null;
		contentRef.current = "";
		setContent("");
		setEvents([]);
		setThreadId(null);
		setIsStreaming(false);
		setError(null);
	}, []);
	const send = useCallback(async (message) => {
		abortControllerRef.current?.abort();
		const controller = new AbortController();
		abortControllerRef.current = controller;
		contentRef.current = "";
		setContent("");
		setEvents([]);
		setError(null);
		setIsStreaming(true);
		const payload = {
			message,
			agent,
			...threadIdRef.current ? { threadId: threadIdRef.current } : {}
		};
		try {
			await connectSSE({
				url: endpoint,
				payload,
				signal: controller.signal,
				maxRetries: 0,
				onMessage: async ({ data }) => {
					if (controller.signal.aborted) return;
					if (!data || data === "[DONE]") return;
					let event;
					try {
						event = JSON.parse(data);
					} catch {
						return;
					}
					if (!event.type) return;
					try {
						onEventRef.current?.(event);
					} catch {}
					setEvents((prev) => [...prev, event]);
					if (event.type === "appkit.metadata") {
						const tid = event.data?.threadId;
						if (typeof tid === "string") {
							threadIdRef.current = tid;
							setThreadId(tid);
						}
					} else if (event.type === "response.output_text.delta" && typeof event.delta === "string") {
						contentRef.current += event.delta;
						setContent(contentRef.current);
					}
				},
				onError: (err) => {
					if (controller.signal.aborted) return;
					setError(err instanceof Error ? err.message : "Chat stream failed");
				}
			});
		} catch (err) {
			if (!controller.signal.aborted) setError(err instanceof Error ? err.message : "Chat stream failed");
		} finally {
			if (abortControllerRef.current === controller) abortControllerRef.current = null;
			setIsStreaming(false);
		}
	}, [agent, endpoint]);
	useEffect(() => {
		return () => {
			abortControllerRef.current?.abort();
		};
	}, []);
	return {
		content,
		events,
		threadId,
		isStreaming,
		error,
		send,
		reset
	};
}

//#endregion
export { useAgentChat };
