// @databricks/appkit-ui/dist/react/ui/accordion.js
import { cn } from "../lib/utils.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { ChevronDownIcon } from "lucide-react";
import * as AccordionPrimitive from "@radix-ui/react-accordion";

//#region src/react/ui/accordion.tsx
/** Collapsible content sections organized in a vertical stack */
function Accordion({ ...props }) {
	return /* @__PURE__ */ jsx(AccordionPrimitive.Root, {
		"data-slot": "accordion",
		...props
	});
}
/** Individual collapsible section within an accordion */
function AccordionItem({ className, ...props }) {
	return /* @__PURE__ */ jsx(AccordionPrimitive.Item, {
		"data-slot": "accordion-item",
		className: cn("border-b last:border-b-0", className),
		...props
	});
}
/** Clickable button that triggers accordion content visibility */
function AccordionTrigger({ className, children, ...props }) {
	return /* @__PURE__ */ jsx(AccordionPrimitive.Header, {
		className: "flex",
		children: /* @__PURE__ */ jsxs(AccordionPrimitive.Trigger, {
			"data-slot": "accordion-trigger",
			className: cn("focus-visible:border-ring focus-visible:ring-ring/50 flex flex-1 items-start justify-between gap-4 rounded-md py-4 text-left text-sm font-medium transition-all outline-none hover:underline focus-visible:ring-[3px] disabled:pointer-events-none disabled:opacity-50 [&[data-state=open]>svg]:rotate-180", className),
			...props,
			children: [children, /* @__PURE__ */ jsx(ChevronDownIcon, { className: "text-muted-foreground pointer-events-none size-4 shrink-0 translate-y-0.5 transition-transform duration-200" })]
		})
	});
}
/** Content area that expands and collapses within an accordion item */
function AccordionContent({ className, children, ...props }) {
	return /* @__PURE__ */ jsx(AccordionPrimitive.Content, {
		"data-slot": "accordion-content",
		className: "data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down overflow-hidden text-sm",
		...props,
		children: /* @__PURE__ */ jsx("div", {
			className: cn("pt-0 pb-4", className),
			children
		})
	});
}

//#endregion
export { Accordion, AccordionContent, AccordionItem, AccordionTrigger };
