// @databricks/appkit-ui/dist/react/ui/alert-dialog.js
import { cn } from "../lib/utils.js";
import { buttonVariants } from "./button.js";
import { useResolvedPortalContainer } from "../portal-container-context.js";
import { jsx, jsxs } from "react/jsx-runtime";
import * as AlertDialogPrimitive from "@radix-ui/react-alert-dialog";

//#region src/react/ui/alert-dialog.tsx
/** Modal dialog that interrupts the user with critical information requiring immediate action */
function AlertDialog({ ...props }) {
	return /* @__PURE__ */ jsx(AlertDialogPrimitive.Root, {
		"data-slot": "alert-dialog",
		...props
	});
}
/** Button that triggers the alert dialog to open */
function AlertDialogTrigger({ ...props }) {
	return /* @__PURE__ */ jsx(AlertDialogPrimitive.Trigger, {
		"data-slot": "alert-dialog-trigger",
		...props
	});
}
/** Portal container for rendering alert dialog content outside the DOM hierarchy */
function AlertDialogPortal({ container, ...props }) {
	return /* @__PURE__ */ jsx(AlertDialogPrimitive.Portal, {
		"data-slot": "alert-dialog-portal",
		container: useResolvedPortalContainer(container),
		...props
	});
}
/** Background overlay that dims content behind the alert dialog */
function AlertDialogOverlay({ className, ...props }) {
	return /* @__PURE__ */ jsx(AlertDialogPrimitive.Overlay, {
		"data-slot": "alert-dialog-overlay",
		className: cn("data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 fixed inset-0 z-50 bg-black/50", className),
		...props
	});
}
/** Main content container for the alert dialog */
function AlertDialogContent({ className, ...props }) {
	return /* @__PURE__ */ jsxs(AlertDialogPortal, { children: [/* @__PURE__ */ jsx(AlertDialogOverlay, {}), /* @__PURE__ */ jsx(AlertDialogPrimitive.Content, {
		"data-slot": "alert-dialog-content",
		className: cn("bg-background data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 fixed top-[50%] left-[50%] z-50 grid w-full max-w-[calc(100%-2rem)] translate-x-[-50%] translate-y-[-50%] gap-4 rounded-lg border p-6 shadow-lg duration-200 sm:max-w-lg", className),
		...props
	})] });
}
/** Header section containing title and description */
function AlertDialogHeader({ className, ...props }) {
	return /* @__PURE__ */ jsx("div", {
		"data-slot": "alert-dialog-header",
		className: cn("flex flex-col gap-2 text-center sm:text-left", className),
		...props
	});
}
/** Footer section containing action buttons */
function AlertDialogFooter({ className, ...props }) {
	return /* @__PURE__ */ jsx("div", {
		"data-slot": "alert-dialog-footer",
		className: cn("flex flex-col-reverse gap-2 sm:flex-row sm:justify-end", className),
		...props
	});
}
/** Title heading for the alert dialog */
function AlertDialogTitle({ className, ...props }) {
	return /* @__PURE__ */ jsx(AlertDialogPrimitive.Title, {
		"data-slot": "alert-dialog-title",
		className: cn("text-lg font-semibold", className),
		...props
	});
}
/** Descriptive text explaining the alert */
function AlertDialogDescription({ className, ...props }) {
	return /* @__PURE__ */ jsx(AlertDialogPrimitive.Description, {
		"data-slot": "alert-dialog-description",
		className: cn("text-muted-foreground text-sm", className),
		...props
	});
}
/** Primary action button that confirms the alert */
function AlertDialogAction({ className, ...props }) {
	return /* @__PURE__ */ jsx(AlertDialogPrimitive.Action, {
		className: cn(buttonVariants(), className),
		...props
	});
}
/** Cancel button that dismisses the alert dialog */
function AlertDialogCancel({ className, ...props }) {
	return /* @__PURE__ */ jsx(AlertDialogPrimitive.Cancel, {
		className: cn(buttonVariants({ variant: "outline" }), className),
		...props
	});
}

//#endregion
export { AlertDialog, AlertDialogAction, AlertDialogCancel, AlertDialogContent, AlertDialogDescription, AlertDialogFooter, AlertDialogHeader, AlertDialogOverlay, AlertDialogPortal, AlertDialogTitle, AlertDialogTrigger };
