// @databricks/appkit-ui/dist/react/ui/pagination.js
import { cn } from "../lib/utils.js";
import { buttonVariants } from "./button.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { ChevronLeftIcon, ChevronRightIcon, MoreHorizontalIcon } from "lucide-react";

//#region src/react/ui/pagination.tsx
/** Navigation component for paginated content */
function Pagination({ className, ...props }) {
	return /* @__PURE__ */ jsx("nav", {
		role: "navigation",
		"aria-label": "pagination",
		"data-slot": "pagination",
		className: cn("mx-auto flex w-full justify-center", className),
		...props
	});
}
/** Container for pagination items */
function PaginationContent({ className, ...props }) {
	return /* @__PURE__ */ jsx("ul", {
		"data-slot": "pagination-content",
		className: cn("flex flex-row items-center gap-1", className),
		...props
	});
}
/** Individual pagination item wrapper */
function PaginationItem({ ...props }) {
	return /* @__PURE__ */ jsx("li", {
		"data-slot": "pagination-item",
		...props
	});
}
/** Clickable link for navigating to a specific page */
function PaginationLink({ className, isActive, size = "icon", ...props }) {
	return /* @__PURE__ */ jsx("a", {
		"aria-current": isActive ? "page" : void 0,
		"data-slot": "pagination-link",
		"data-active": isActive,
		className: cn(buttonVariants({
			variant: isActive ? "outline" : "ghost",
			size
		}), className),
		...props
	});
}
/** Button for navigating to the previous page */
function PaginationPrevious({ className, ...props }) {
	return /* @__PURE__ */ jsxs(PaginationLink, {
		"aria-label": "Go to previous page",
		size: "default",
		className: cn("gap-1 px-2.5 sm:pl-2.5", className),
		...props,
		children: [/* @__PURE__ */ jsx(ChevronLeftIcon, {}), /* @__PURE__ */ jsx("span", {
			className: "hidden sm:block",
			children: "Previous"
		})]
	});
}
/** Button for navigating to the next page */
function PaginationNext({ className, ...props }) {
	return /* @__PURE__ */ jsxs(PaginationLink, {
		"aria-label": "Go to next page",
		size: "default",
		className: cn("gap-1 px-2.5 sm:pr-2.5", className),
		...props,
		children: [/* @__PURE__ */ jsx("span", {
			className: "hidden sm:block",
			children: "Next"
		}), /* @__PURE__ */ jsx(ChevronRightIcon, {})]
	});
}
/** Ellipsis indicator for skipped page numbers */
function PaginationEllipsis({ className, ...props }) {
	return /* @__PURE__ */ jsxs("span", {
		"aria-hidden": true,
		"data-slot": "pagination-ellipsis",
		className: cn("flex size-9 items-center justify-center", className),
		...props,
		children: [/* @__PURE__ */ jsx(MoreHorizontalIcon, { className: "size-4" }), /* @__PURE__ */ jsx("span", {
			className: "sr-only",
			children: "More pages"
		})]
	});
}

//#endregion
export { Pagination, PaginationContent, PaginationEllipsis, PaginationItem, PaginationLink, PaginationNext, PaginationPrevious };
