// @databricks/sdk-experimental/dist/AccountClient.js
"use strict";
/* eslint-disable @typescript-eslint/no-unused-vars */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountClient = void 0;
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
const Config_1 = require("./config/Config");
const api_client_1 = require("./api-client");
const billing = __importStar(require("./apis/billing"));
const catalog = __importStar(require("./apis/catalog"));
const iam = __importStar(require("./apis/iam"));
const iamv2 = __importStar(require("./apis/iamv2"));
const oauth2 = __importStar(require("./apis/oauth2"));
const provisioning = __importStar(require("./apis/provisioning"));
const settings = __importStar(require("./apis/settings"));
const settingsv2 = __importStar(require("./apis/settingsv2"));
class AccountClient {
    constructor(config, options = {}) {
        if (!(config instanceof Config_1.Config)) {
            config = new Config_1.Config(config);
        }
        this.config = config;
        this.apiClient = new api_client_1.ApiClient(this.config, options);
        this.accountAccessControl = new iam.AccountAccessControlService(this.apiClient);
        this.billableUsage = new billing.BillableUsageService(this.apiClient);
        this.budgetPolicy = new billing.BudgetPolicyService(this.apiClient);
        this.credentials = new provisioning.CredentialsService(this.apiClient);
        this.customAppIntegration = new oauth2.CustomAppIntegrationService(this.apiClient);
        this.encryptionKeys = new provisioning.EncryptionKeysService(this.apiClient);
        this.accountFederationPolicy =
            new oauth2.AccountFederationPolicyService(this.apiClient);
        this.accountGroupsV2 = new iam.AccountGroupsV2Service(this.apiClient);
        this.accountIpAccessLists = new settings.AccountIpAccessListsService(this.apiClient);
        this.logDelivery = new billing.LogDeliveryService(this.apiClient);
        this.accountMetastoreAssignments =
            new catalog.AccountMetastoreAssignmentsService(this.apiClient);
        this.accountMetastores = new catalog.AccountMetastoresService(this.apiClient);
        this.networkConnectivity = new settings.NetworkConnectivityService(this.apiClient);
        this.networkPolicies = new settings.NetworkPoliciesService(this.apiClient);
        this.networks = new provisioning.NetworksService(this.apiClient);
        this.oAuthPublishedApps = new oauth2.OAuthPublishedAppsService(this.apiClient);
        this.privateAccess = new provisioning.PrivateAccessService(this.apiClient);
        this.publishedAppIntegration =
            new oauth2.PublishedAppIntegrationService(this.apiClient);
        this.servicePrincipalFederationPolicy =
            new oauth2.ServicePrincipalFederationPolicyService(this.apiClient);
        this.servicePrincipalSecrets =
            new oauth2.ServicePrincipalSecretsService(this.apiClient);
        this.accountServicePrincipalsV2 =
            new iam.AccountServicePrincipalsV2Service(this.apiClient);
        this.accountSettings = new settings.AccountSettingsService(this.apiClient);
        this.accountSettingsV2 = new settingsv2.AccountSettingsV2Service(this.apiClient);
        this.storage = new provisioning.StorageService(this.apiClient);
        this.accountStorageCredentials =
            new catalog.AccountStorageCredentialsService(this.apiClient);
        this.usageDashboards = new billing.UsageDashboardsService(this.apiClient);
        this.accountUsersV2 = new iam.AccountUsersV2Service(this.apiClient);
        this.vpcEndpoints = new provisioning.VpcEndpointsService(this.apiClient);
        this.workspaceAssignment = new iam.WorkspaceAssignmentService(this.apiClient);
        this.workspaceNetworkConfiguration =
            new settings.WorkspaceNetworkConfigurationService(this.apiClient);
        this.workspaces = new provisioning.WorkspacesService(this.apiClient);
        this.accountIamV2 = new iamv2.AccountIamV2Service(this.apiClient);
        this.budgets = new billing.BudgetsService(this.apiClient);
    }
}
exports.AccountClient = AccountClient;
