// @databricks/sdk-experimental/dist/api-client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiClient = void 0;
const https = __importStar(require("node:https"));
const logging_1 = require("./logging");
const context_1 = require("./context");
const context_2 = require("./context");
const HostType_1 = require("./config/HostType");
const retries_1 = __importStar(require("./retries/retries"));
const Time_1 = __importStar(require("./retries/Time"));
const apierr_1 = require("./apierr");
const fetch_1 = require("./fetch");
const fetch_2 = require("./fetch");
// eslint-disable-next-line @typescript-eslint/no-require-imports
const sdkVersion = require("../package.json").version;
function logAndReturnError(url, request, response, error, context) {
    context?.logger?.error(url.toString(), {
        request,
        response,
        error: logging_1.Utils.liftAllErrorProps(error),
    });
    return error;
}
class ApiClient {
    constructor(config, options = {}) {
        this.config = config;
        this._agent = options.agent;
        this.product = options.product || "unknown";
        this.productVersion = options.productVersion || "0.0.0";
        this.userAgentExtra = options.userAgentExtra || {};
    }
    get host() {
        return this.config.getHost();
    }
    get accountId() {
        return this.config.ensureResolved().then(() => {
            return this.config.accountId;
        });
    }
    async getAgent() {
        if (!this._agent) {
            await this.config.ensureResolved();
            this._agent = new https.Agent({
                keepAlive: true,
                keepAliveMsecs: 15000,
                rejectUnauthorized: this.config.skipVerify === false,
                timeout: (this.config.httpTimeoutSeconds || 5) * 1000,
            });
        }
        return this._agent;
    }
    userAgent() {
        const pairs = [
            `${this.product}/${this.productVersion}`,
            `databricks-sdk-js/${sdkVersion}`,
            `nodejs/${process.version.slice(1)}`,
            `os/${process.platform}`,
            `auth/${this.config.authType}`,
        ];
        for (const [key, value] of Object.entries(this.userAgentExtra)) {
            pairs.push(`${key}/${value}`);
        }
        return pairs.join(" ");
    }
    searchParams(query) {
        // Flatten the payload object into a list of key-value pairs to support nested payload objects
        const flatten = (obj, parentKey = "", res = []) => {
            for (const key of Object.keys(obj)) {
                const value = obj[key];
                const propName = parentKey ? `${parentKey}.${key}` : key;
                if (Array.isArray(value)) {
                    // We assume here that arrays don't contain non-primitive items
                    for (const item of value) {
                        if (item !== null && item !== undefined) {
                            res.push([propName, item.toString()]);
                        }
                    }
                }
                else if (typeof value === "object" && value !== null) {
                    // Recursively flatten nested objects
                    flatten(value, propName, res);
                }
                else if (value !== null && value !== undefined) {
                    res.push([propName, value.toString()]);
                }
            }
            return res;
        };
        return new URLSearchParams(flatten(query)).toString();
    }
    async request(options, context) {
        const { path, method, query = {}, headers, responseHeaders = [], payload, raw = false, } = options;
        headers.set("User-Agent", this.userAgent());
        await this.config.authenticate(headers);
        // Add X-Databricks-Org-Id header for workspace clients on unified hosts
        if (this.config.hostType() === HostType_1.HostType.unifiedHost &&
            this.config.workspaceId) {
            headers.set("X-Databricks-Org-Id", this.config.workspaceId);
        }
        // create a copy of the URL, so that we can modify it
        const url = new URL((await this.host).toString());
        url.pathname = path;
        const requestOptions = {
            method,
            headers,
            agent: await this.getAgent(),
        };
        // Add query parameters to URL for all methods
        if (query && Object.keys(query).length > 0) {
            url.search = this.searchParams(query);
        }
        // Add payload as request body for POST/PUT methods
        if (payload) {
            if (method === "POST" || method === "PUT") {
                if (payload instanceof fetch_2.ReadableStream) {
                    requestOptions.body = payload;
                }
                else {
                    requestOptions.body = JSON.stringify(payload);
                }
            }
        }
        const response = await (0, retries_1.default)({
            timeout: new Time_1.default(this.config.retryTimeoutSeconds || 300, Time_1.TimeUnits.seconds),
            fn: async () => {
                let response;
                try {
                    const controller = new fetch_1.AbortController();
                    if (context?.cancellationToken?.onCancellationRequested) {
                        context?.cancellationToken?.onCancellationRequested(() => controller.abort());
                    }
                    if (context?.cancellationToken?.isCancellationRequested &&
                        !controller.signal.aborted) {
                        controller.abort();
                    }
                    response = await (0, fetch_1.fetch)(url.toString(), {
                        signal: controller.signal,
                        ...requestOptions,
                    });
                }
                catch (e) {
                    const err = e.code && e.code === "ENOTFOUND"
                        ? new apierr_1.HttpError(`Can't connect to ${url.toString()}`, 500)
                        : e;
                    throw logAndReturnError(url, requestOptions, "", err, context);
                }
                if (!response.ok) {
                    const err = (0, apierr_1.parseErrorFromResponse)(response.status, response.statusText, await response.text());
                    if (err.isRetryable()) {
                        throw new retries_1.RetriableError();
                    }
                    else {
                        throw logAndReturnError(url, requestOptions, response, err, context);
                    }
                }
                return {
                    body: raw ? response.body : await response.text(),
                    responseHeaders: response.headers,
                };
            },
        });
        const responseHeadersMap = {};
        response.responseHeaders.forEach((value, key) => {
            if (responseHeaders.includes(key.toLowerCase())) {
                responseHeadersMap[key.toLowerCase()] = value;
            }
        });
        let bodyJson;
        if (typeof response.body === "string") {
            const responseText = response.body;
            try {
                bodyJson =
                    responseText.length === 0 ? {} : JSON.parse(responseText);
            }
            catch (e) {
                logAndReturnError(url, options, responseText, e, context);
                throw new Error(`Can't parse reponse as JSON: ${responseText}`);
            }
        }
        else {
            bodyJson = {
                contents: response.body,
            };
        }
        const responseJson = Object.assign({}, bodyJson, responseHeadersMap);
        context?.logger?.debug(url.toString(), {
            request: requestOptions,
            response: responseJson,
        });
        return responseJson;
    }
}
exports.ApiClient = ApiClient;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_2.Context]),
    __metadata("design:returntype", Promise)
], ApiClient.prototype, "request", null);
