// @databricks/appkit/dist/stream/buffers.js
import { ValidationError } from "../errors/validation.js";
import "../errors/index.js";

//#region src/stream/buffers.ts
var RingBuffer = class {
	buffer;
	capacity;
	writeIndex;
	size;
	keyExtractor;
	keyIndex;
	constructor(capacity, keyExtractor) {
		if (capacity <= 0) throw ValidationError.invalidValue("capacity", capacity, "greater than 0");
		this.capacity = capacity;
		this.buffer = new Array(capacity).fill(null);
		this.writeIndex = 0;
		this.size = 0;
		this.keyExtractor = keyExtractor;
		this.keyIndex = /* @__PURE__ */ new Map();
	}
	add(item) {
		const key = this.keyExtractor(item);
		const existingIndex = this.keyIndex.get(key);
		if (existingIndex !== void 0) {
			this.buffer[existingIndex] = item;
			return;
		}
		const evicted = this.buffer[this.writeIndex];
		if (evicted !== null) {
			const evictedKey = this.keyExtractor(evicted);
			this.keyIndex.delete(evictedKey);
		}
		this.buffer[this.writeIndex] = item;
		this.keyIndex.set(key, this.writeIndex);
		this.writeIndex = (this.writeIndex + 1) % this.capacity;
		this.size = Math.min(this.size + 1, this.capacity);
	}
	get(key) {
		const index = this.keyIndex.get(key);
		if (index === void 0) return null;
		return this.buffer[index];
	}
	has(key) {
		return this.keyIndex.has(key);
	}
	remove(key) {
		const index = this.keyIndex.get(key);
		if (index === void 0) return;
		this.buffer[index] = null;
		this.keyIndex.delete(key);
		this.size = Math.max(this.size - 1, 0);
	}
	getAll() {
		const result = [];
		for (let i = 0; i < this.size; i++) {
			const index = (this.writeIndex - this.size + i + this.capacity) % this.capacity;
			const item = this.buffer[index];
			if (item !== null) result.push(item);
		}
		return result;
	}
	getSize() {
		return this.size;
	}
	clear() {
		this.buffer = new Array(this.capacity).fill(null);
		this.keyIndex.clear();
		this.writeIndex = 0;
		this.size = 0;
	}
};
var EventRingBuffer = class {
	buffer;
	constructor(capacity = 100) {
		this.buffer = new RingBuffer(capacity, (event) => event.id);
	}
	add(event) {
		this.buffer.add(event);
	}
	has(eventId) {
		return this.buffer.has(eventId);
	}
	getEventsSince(lastEventId) {
		const allEvents = this.buffer.getAll();
		const result = [];
		let foundLastEvent = false;
		for (const event of allEvents) if (foundLastEvent) result.push(event);
		else if (event.id === lastEventId) foundLastEvent = true;
		return result;
	}
	clear() {
		this.buffer.clear();
	}
};

//#endregion
export { EventRingBuffer, RingBuffer };
