// @databricks/appkit/dist/stream/sse-writer.js
import { streamDefaults } from "./defaults.js";
import { SSEErrorCode, SSEWarningCode } from "./types.js";
import { StreamValidator } from "./validator.js";

//#region src/stream/sse-writer.ts
var SSEWriter = class {
	setupHeaders(res) {
		res.setHeader("Content-Type", "text/event-stream; charset=utf-8");
		res.setHeader("Cache-Control", "no-cache, no-transform");
		res.setHeader("X-Accel-Buffering", "no");
		res.flushHeaders?.();
	}
	writeEvent(res, eventId, event) {
		if (res.writableEnded) return;
		const eventType = StreamValidator.sanitizeEventType(event.type);
		const eventData = JSON.stringify(event);
		res.write(`id: ${eventId}\n`);
		res.write(`event: ${eventType}\n`);
		res.write(`data: ${eventData}\n\n`);
	}
	writeError(res, eventId, error, code = SSEErrorCode.INTERNAL_ERROR) {
		if (res.writableEnded) return;
		const errorData = {
			error,
			code
		};
		res.write(`id: ${eventId}\n`);
		res.write(`event: error\n`);
		res.write(`data: ${JSON.stringify(errorData)}\n\n`);
	}
	writeBufferedEvent(res, event) {
		if (res.writableEnded) return;
		res.write(`id: ${event.id}\n`);
		res.write(`event: ${event.type}\n`);
		res.write(`data: ${event.data}\n\n`);
	}
	writeBufferOverflowWarning(res, lastEventId) {
		if (res.writableEnded) return;
		try {
			res.write(`event: warning\n`);
			res.write(`data: ${JSON.stringify({
				warning: "Buffer overflow detected - some events were lost",
				code: SSEWarningCode.BUFFER_OVERFLOW_RESTART,
				lastEventId
			})}\n\n`);
		} catch (_error) {}
	}
	startHeartbeat(res, signal, interval) {
		const heartbeatInterval = interval ?? streamDefaults.heartbeatInterval;
		return setInterval(() => {
			if (!signal.aborted && !res.writableEnded) try {
				res.write(`: heartbeat\n\n`);
			} catch (_error) {}
		}, heartbeatInterval);
	}
};

//#endregion
export { SSEWriter };
