// @databricks/appkit/dist/stream/stream-manager.js
import { createLogger } from "../logging/logger.js";
import { EventRingBuffer } from "./buffers.js";
import { streamDefaults } from "./defaults.js";
import { SSEErrorCode } from "./types.js";
import { StreamValidator } from "./validator.js";
import { SSEWriter } from "./sse-writer.js";
import { StreamRegistry } from "./stream-registry.js";
import { randomUUID } from "node:crypto";
import { context } from "@opentelemetry/api";

//#region src/stream/stream-manager.ts
const logger = createLogger("stream");
var StreamManager = class {
	activeOperations;
	streamRegistry;
	sseWriter;
	maxEventSize;
	bufferTTL;
	constructor(options) {
		this.streamRegistry = new StreamRegistry(options?.maxActiveStreams ?? streamDefaults.maxActiveStreams);
		this.sseWriter = new SSEWriter();
		this.maxEventSize = options?.maxEventSize ?? streamDefaults.maxEventSize;
		this.bufferTTL = options?.bufferTTL ?? streamDefaults.bufferTTL;
		this.activeOperations = /* @__PURE__ */ new Set();
	}
	async stream(res, handler, options, ownerKey) {
		const { streamId } = options || {};
		if (res.writableEnded || res.destroyed) return;
		this.sseWriter.setupHeaders(res);
		if (streamId && StreamValidator.validateStreamId(streamId)) {
			const existingStream = this.streamRegistry.get(streamId);
			if (existingStream) {
				if (existingStream.ownerKey !== ownerKey) {
					this.sseWriter.writeError(res, randomUUID(), "Stream not found or access denied", SSEErrorCode.STREAM_FORBIDDEN);
					res.end();
					return;
				}
				return this._attachToExistingStream(res, existingStream, options);
			}
		}
		return this._createNewStream(res, handler, options, ownerKey);
	}
	abortAll() {
		this.activeOperations.forEach((operation) => {
			if (operation.heartbeat) clearInterval(operation.heartbeat);
			operation.controller.abort(new DOMException("Server shutdown", "AbortError"));
		});
		this.activeOperations.clear();
		this.streamRegistry.clear();
	}
	getActiveCount() {
		return this.activeOperations.size;
	}
	async _attachToExistingStream(res, streamEntry, options) {
		const lastEventId = res.req?.headers["last-event-id"];
		if (StreamValidator.validateEventId(lastEventId)) {
			const validEventId = lastEventId;
			if (streamEntry.eventBuffer.has(validEventId)) {
				const missedEvents = streamEntry.eventBuffer.getEventsSince(validEventId);
				for (const event of missedEvents) {
					if (options?.userSignal?.aborted) break;
					this.sseWriter.writeBufferedEvent(res, event);
				}
			} else this.sseWriter.writeBufferOverflowWarning(res, validEventId);
		}
		streamEntry.clients.add(res);
		streamEntry.lastAccess = Date.now();
		const combinedSignal = this._combineSignals(streamEntry.abortController.signal, options?.userSignal);
		const heartbeat = this.sseWriter.startHeartbeat(res, combinedSignal);
		const streamOperation = {
			controller: streamEntry.abortController,
			type: "stream",
			heartbeat
		};
		this.activeOperations.add(streamOperation);
		res.on("close", () => {
			clearInterval(heartbeat);
			streamEntry.clients.delete(res);
			this.activeOperations.delete(streamOperation);
			if (streamEntry.clients.size === 0 && !streamEntry.isCompleted) streamEntry.abortController.abort(new DOMException("All clients disconnected", "AbortError"));
			if (streamEntry.isCompleted && streamEntry.clients.size === 0) setTimeout(() => {
				if (streamEntry.clients.size === 0) this.streamRegistry.remove(streamEntry.streamId);
			}, this.bufferTTL);
		});
		if (streamEntry.isCompleted) {
			res.end();
			this.activeOperations.delete(streamOperation);
			clearInterval(heartbeat);
		}
	}
	async _createNewStream(res, handler, options, ownerKey) {
		const streamId = options?.streamId ?? randomUUID();
		if (res.writableEnded || res.destroyed) return;
		const abortController = new AbortController();
		const eventBuffer = new EventRingBuffer(options?.bufferSize ?? streamDefaults.bufferSize);
		const combinedSignal = this._combineSignals(abortController.signal, options?.userSignal);
		const heartbeat = this.sseWriter.startHeartbeat(res, combinedSignal);
		const traceContext = context.active();
		if (res.writableEnded || res.destroyed) {
			clearInterval(heartbeat);
			return;
		}
		const streamEntry = {
			streamId,
			ownerKey,
			generator: handler(combinedSignal),
			eventBuffer,
			clients: new Set([res]),
			isCompleted: false,
			lastAccess: Date.now(),
			abortController,
			traceContext
		};
		this.streamRegistry.add(streamEntry);
		const streamOperation = {
			controller: abortController,
			type: "stream",
			heartbeat
		};
		this.activeOperations.add(streamOperation);
		res.on("close", () => {
			clearInterval(heartbeat);
			this.activeOperations.delete(streamOperation);
			streamEntry.clients.delete(res);
			if (streamEntry.clients.size === 0 && !streamEntry.isCompleted) abortController.abort(new DOMException("Client disconnected", "AbortError"));
		});
		await this._processGeneratorInBackground(streamEntry);
		clearInterval(heartbeat);
		this.activeOperations.delete(streamOperation);
	}
	async _processGeneratorInBackground(streamEntry) {
		return context.with(streamEntry.traceContext, async () => {
			try {
				for await (const event of streamEntry.generator) {
					if (streamEntry.abortController.signal.aborted) break;
					const eventId = randomUUID();
					const eventData = JSON.stringify(event);
					if (eventData.length > this.maxEventSize) {
						const errorMsg = `Event exceeds max size of ${this.maxEventSize} bytes`;
						const errorCode = SSEErrorCode.INVALID_REQUEST;
						this._broadcastErrorToClients(streamEntry, eventId, errorMsg, errorCode);
						continue;
					}
					streamEntry.eventBuffer.add({
						id: eventId,
						type: event.type,
						data: eventData,
						timestamp: Date.now()
					});
					this._broadcastEventsToClients(streamEntry, eventId, event);
					streamEntry.lastAccess = Date.now();
				}
				streamEntry.isCompleted = true;
				this._closeAllClients(streamEntry);
				this._cleanupStream(streamEntry);
			} catch (error) {
				const errorMsg = error instanceof Error ? error.message : "Internal server error";
				const errorEventId = randomUUID();
				const errorCode = this._categorizeError(error);
				if (errorCode === SSEErrorCode.STREAM_ABORTED) logger.info("Stream aborted by client (code=%s)", errorCode);
				else logger.error("Stream execution failed: %s (code=%s)", errorMsg, errorCode);
				streamEntry.eventBuffer.add({
					id: errorEventId,
					type: "error",
					data: JSON.stringify({
						error: errorMsg,
						code: errorCode
					}),
					timestamp: Date.now()
				});
				this._broadcastErrorToClients(streamEntry, errorEventId, errorMsg, errorCode, true);
				streamEntry.isCompleted = true;
			}
		});
	}
	_combineSignals(internalSignal, userSignal) {
		if (!userSignal) return internalSignal || new AbortController().signal;
		const signals = [internalSignal, userSignal].filter(Boolean);
		const controller = new AbortController();
		signals.forEach((signal) => {
			if (signal?.aborted) {
				controller.abort(signal.reason);
				return;
			}
			signal?.addEventListener("abort", () => {
				controller.abort(signal.reason);
			}, { once: true });
		});
		return controller.signal;
	}
	_broadcastEventsToClients(streamEntry, eventId, event) {
		for (const client of streamEntry.clients) if (!client.writableEnded) this.sseWriter.writeEvent(client, eventId, event);
	}
	_broadcastErrorToClients(streamEntry, eventId, errorMessage, errorCode, closeClients = false) {
		for (const client of streamEntry.clients) if (!client.writableEnded) {
			this.sseWriter.writeError(client, eventId, errorMessage, errorCode);
			if (closeClients) client.end();
		}
	}
	_closeAllClients(streamEntry) {
		for (const client of streamEntry.clients) if (!client.writableEnded) client.end();
	}
	_cleanupStream(streamEntry) {
		if (streamEntry.clients.size === 0) setTimeout(() => {
			if (streamEntry.clients.size === 0) this.streamRegistry.remove(streamEntry.streamId);
		}, this.bufferTTL);
	}
	_categorizeError(error) {
		if (error instanceof Error) {
			const message = error.message.toLowerCase();
			if (message.includes("timeout") || message.includes("timed out")) return SSEErrorCode.TIMEOUT;
			if (message.includes("unavailable") || message.includes("econnrefused")) return SSEErrorCode.TEMPORARY_UNAVAILABLE;
			if (error.name === "AbortError") return SSEErrorCode.STREAM_ABORTED;
			if ("statusCode" in error && typeof error.statusCode === "number") return SSEErrorCode.UPSTREAM_ERROR;
		}
		return SSEErrorCode.INTERNAL_ERROR;
	}
};

//#endregion
export { StreamManager };
