// @databricks/appkit/dist/stream/stream-registry.js
import { RingBuffer } from "./buffers.js";
import { SSEErrorCode } from "./types.js";

//#region src/stream/stream-registry.ts
var StreamRegistry = class {
	streams;
	constructor(maxActiveStreams) {
		this.streams = new RingBuffer(maxActiveStreams, (entry) => entry.streamId);
	}
	add(entry) {
		if (this.streams.getSize() >= this.streams.capacity) this._evictOldestStream(entry.streamId);
		this.streams.add(entry);
	}
	get(streamId) {
		return this.streams.get(streamId);
	}
	has(streamId) {
		return this.streams.has(streamId);
	}
	remove(streamId) {
		this.streams.remove(streamId);
	}
	size() {
		return this.streams.getSize();
	}
	clear() {
		const allStreams = this.streams.getAll();
		for (const stream of allStreams) stream.abortController.abort("Server shutdown");
		this.streams.clear();
	}
	_evictOldestStream(excludeStreamId) {
		const allStreams = this.streams.getAll();
		let oldestStream = null;
		let oldestAccess = Infinity;
		for (const stream of allStreams) if (stream.streamId !== excludeStreamId && stream.lastAccess < oldestAccess) {
			oldestStream = stream;
			oldestAccess = stream.lastAccess;
		}
		if (oldestStream) {
			for (const client of oldestStream.clients) if (!client.writableEnded) try {
				client.write(`event: error\n`);
				client.write(`data: ${JSON.stringify({
					error: "Stream evicted",
					code: SSEErrorCode.STREAM_EVICTED
				})}\n\n`);
			} catch (_error) {}
			oldestStream.abortController.abort("Stream evicted");
			this.streams.remove(oldestStream.streamId);
		}
	}
};

//#endregion
export { StreamRegistry };
