// @databricks/sdk-experimental/dist/apis/serving/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServingEndpointsDataPlaneService = exports.ServingEndpointsDataPlaneError = exports.ServingEndpointsDataPlaneRetriableError = exports.ServingEndpointsService = exports.ServingEndpointsError = exports.ServingEndpointsRetriableError = void 0;
const serving = __importStar(require("./model"));
const retries_1 = __importDefault(require("../../retries/retries"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
const wait_1 = require("../../wait");
class ServingEndpointsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("ServingEndpoints", method, message);
    }
}
exports.ServingEndpointsRetriableError = ServingEndpointsRetriableError;
class ServingEndpointsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("ServingEndpoints", method, message);
    }
}
exports.ServingEndpointsError = ServingEndpointsError;
/**
 * The Serving Endpoints API allows you to create, update, and delete model
 * serving endpoints.
 *
 * You can use a serving endpoint to serve models from the Databricks Model
 * Registry or from Unity Catalog. Endpoints expose the underlying models as
 * scalable REST API endpoints using serverless compute. This means the endpoints
 * and associated compute resources are fully managed by Databricks and will not
 * appear in your cloud account. A serving endpoint can consist of one or more
 * MLflow models from the Databricks Model Registry, called served entities. A
 * serving endpoint can have at most ten served entities. You can configure
 * traffic settings to define how requests should be routed to your served
 * entities behind an endpoint. Additionally, you can configure the scale of
 * resources that should be applied to each served entity.
 */
class ServingEndpointsService {
    constructor(client) {
        this.client = client;
    }
    async _buildLogs(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/serving-endpoints/${request.name}/served-models/${request.served_model_name}/build-logs`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieves the build logs associated with the provided served model.
     */
    async buildLogs(request, context) {
        return await this._buildLogs(request, context);
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("ai_gateway")) {
            body["ai_gateway"] = request.ai_gateway;
        }
        if (request.hasOwnProperty("budget_policy_id")) {
            body["budget_policy_id"] = request.budget_policy_id;
        }
        if (request.hasOwnProperty("config")) {
            body["config"] = request.config;
        }
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("email_notifications")) {
            body["email_notifications"] = request.email_notifications;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("rate_limits")) {
            body["rate_limits"] = request.rate_limits;
        }
        if (request.hasOwnProperty("route_optimized")) {
            body["route_optimized"] = request.route_optimized;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        const query = {};
        const path = "/api/2.0/serving-endpoints";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a new serving endpoint.
     */
    async create(createServingEndpoint, context) {
        let cancellationToken = context?.cancellationToken;
        const servingEndpointDetailed = await this._create(createServingEndpoint, context);
        return (0, wait_1.asWaiter)(servingEndpointDetailed, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        name: servingEndpointDetailed.name,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("ServingEndpoints.createAndWait: cancelled");
                        throw new ServingEndpointsError("createAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state.config_update;
                    const statusMessage = pollResponse;
                    switch (status) {
                        case "NOT_UPDATING": {
                            return pollResponse;
                        }
                        case "UPDATE_FAILED":
                        case "UPDATE_CANCELED": {
                            const errorMessage = `failed to reach NOT_UPDATING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`ServingEndpoints.createAndWait: ${errorMessage}`);
                            throw new ServingEndpointsError("createAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach NOT_UPDATING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`ServingEndpoints.createAndWait: retrying: ${errorMessage}`);
                            throw new ServingEndpointsRetriableError("createAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _createProvisionedThroughputEndpoint(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("ai_gateway")) {
            body["ai_gateway"] = request.ai_gateway;
        }
        if (request.hasOwnProperty("budget_policy_id")) {
            body["budget_policy_id"] = request.budget_policy_id;
        }
        if (request.hasOwnProperty("config")) {
            body["config"] = request.config;
        }
        if (request.hasOwnProperty("email_notifications")) {
            body["email_notifications"] = request.email_notifications;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        const query = {};
        const path = "/api/2.0/serving-endpoints/pt";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a new PT serving endpoint.
     */
    async createProvisionedThroughputEndpoint(createPtEndpointRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const servingEndpointDetailed = await this._createProvisionedThroughputEndpoint(createPtEndpointRequest, context);
        return (0, wait_1.asWaiter)(servingEndpointDetailed, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        name: servingEndpointDetailed.name,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("ServingEndpoints.createProvisionedThroughputEndpointAndWait: cancelled");
                        throw new ServingEndpointsError("createProvisionedThroughputEndpointAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state.config_update;
                    const statusMessage = pollResponse;
                    switch (status) {
                        case "NOT_UPDATING": {
                            return pollResponse;
                        }
                        case "UPDATE_FAILED":
                        case "UPDATE_CANCELED": {
                            const errorMessage = `failed to reach NOT_UPDATING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`ServingEndpoints.createProvisionedThroughputEndpointAndWait: ${errorMessage}`);
                            throw new ServingEndpointsError("createProvisionedThroughputEndpointAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach NOT_UPDATING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`ServingEndpoints.createProvisionedThroughputEndpointAndWait: retrying: ${errorMessage}`);
                            throw new ServingEndpointsRetriableError("createProvisionedThroughputEndpointAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _delete(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/serving-endpoints/${request.name}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a serving endpoint.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _exportMetrics(request, context) {
        const headers = new Headers({ Accept: "text/plain" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/serving-endpoints/${request.name}/metrics`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: true,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieves the metrics associated with the provided serving endpoint in
     * either Prometheus or OpenMetrics exposition format.
     */
    async exportMetrics(request, context) {
        return await this._exportMetrics(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/serving-endpoints/${request.name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieves the details for a single serving endpoint.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _getOpenApi(request, context) {
        const headers = new Headers({ Accept: "text/plain" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/serving-endpoints/${request.name}/openapi`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: true,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get the query schema of the serving endpoint in OpenAPI format. The schema
     * contains information for the supported paths, input and output format and
     * datatypes.
     */
    async getOpenApi(request, context) {
        return await this._getOpenApi(request, context);
    }
    async _getPermissionLevels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/serving-endpoints/${request.serving_endpoint_id}/permissionLevels`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(request, context) {
        return await this._getPermissionLevels(request, context);
    }
    async _getPermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/serving-endpoints/${request.serving_endpoint_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of a serving endpoint. Serving endpoints can inherit
     * permissions from their root object.
     */
    async getPermissions(request, context) {
        return await this._getPermissions(request, context);
    }
    async _httpRequest(request, context) {
        const headers = new Headers({
            "Accept": "text/plain",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("connection_name")) {
            body["connection_name"] = request.connection_name;
        }
        if (request.hasOwnProperty("headers")) {
            body["headers"] = request.headers;
        }
        if (request.hasOwnProperty("json")) {
            body["json"] = request.json;
        }
        if (request.hasOwnProperty("method")) {
            body["method"] = request.method;
        }
        if (request.hasOwnProperty("params")) {
            body["params"] = request.params;
        }
        if (request.hasOwnProperty("path")) {
            body["path"] = request.path;
        }
        const query = {};
        const path = "/api/2.0/external-function";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: true,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Make external services call using the credentials stored in UC Connection.
     */
    async httpRequest(request, context) {
        return await this._httpRequest(request, context);
    }
    async _list(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/serving-endpoints";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get all serving endpoints.
     */
    async *list(context) {
        const response = (await this._list(context)).endpoints;
        for (const v of response || []) {
            yield v;
        }
    }
    async _logs(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/serving-endpoints/${request.name}/served-models/${request.served_model_name}/logs`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieves the service logs associated with the provided served model.
     */
    async logs(request, context) {
        return await this._logs(request, context);
    }
    async _patch(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("add_tags")) {
            body["add_tags"] = request.add_tags;
        }
        if (request.hasOwnProperty("delete_tags")) {
            body["delete_tags"] = request.delete_tags;
        }
        const query = {};
        const path = `/api/2.0/serving-endpoints/${request.name}/tags`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Used to batch add and delete tags from a serving endpoint with a single
     * API call.
     */
    async patch(request, context) {
        return await this._patch(request, context);
    }
    async _put(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("rate_limits")) {
            body["rate_limits"] = request.rate_limits;
        }
        const query = {};
        const path = `/api/2.0/serving-endpoints/${request.name}/rate-limits`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deprecated: Please use AI Gateway to manage rate limits instead.
     */
    async put(request, context) {
        return await this._put(request, context);
    }
    async _putAiGateway(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("fallback_config")) {
            body["fallback_config"] = request.fallback_config;
        }
        if (request.hasOwnProperty("guardrails")) {
            body["guardrails"] = request.guardrails;
        }
        if (request.hasOwnProperty("inference_table_config")) {
            body["inference_table_config"] = request.inference_table_config;
        }
        if (request.hasOwnProperty("rate_limits")) {
            body["rate_limits"] = request.rate_limits;
        }
        if (request.hasOwnProperty("usage_tracking_config")) {
            body["usage_tracking_config"] = request.usage_tracking_config;
        }
        const query = {};
        const path = `/api/2.0/serving-endpoints/${request.name}/ai-gateway`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Used to update the AI Gateway of a serving endpoint. NOTE: External model,
     * provisioned throughput, and pay-per-token endpoints are fully supported;
     * agent endpoints currently only support inference tables.
     */
    async putAiGateway(request, context) {
        return await this._putAiGateway(request, context);
    }
    async _query(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("client_request_id")) {
            body["client_request_id"] = request.client_request_id;
        }
        if (request.hasOwnProperty("dataframe_records")) {
            body["dataframe_records"] = request.dataframe_records;
        }
        if (request.hasOwnProperty("dataframe_split")) {
            body["dataframe_split"] = request.dataframe_split;
        }
        if (request.hasOwnProperty("extra_params")) {
            body["extra_params"] = request.extra_params;
        }
        if (request.hasOwnProperty("input")) {
            body["input"] = request.input;
        }
        if (request.hasOwnProperty("inputs")) {
            body["inputs"] = request.inputs;
        }
        if (request.hasOwnProperty("instances")) {
            body["instances"] = request.instances;
        }
        if (request.hasOwnProperty("max_tokens")) {
            body["max_tokens"] = request.max_tokens;
        }
        if (request.hasOwnProperty("messages")) {
            body["messages"] = request.messages;
        }
        if (request.hasOwnProperty("n")) {
            body["n"] = request.n;
        }
        if (request.hasOwnProperty("prompt")) {
            body["prompt"] = request.prompt;
        }
        if (request.hasOwnProperty("stop")) {
            body["stop"] = request.stop;
        }
        if (request.hasOwnProperty("stream")) {
            body["stream"] = request.stream;
        }
        if (request.hasOwnProperty("temperature")) {
            body["temperature"] = request.temperature;
        }
        if (request.hasOwnProperty("usage_context")) {
            body["usage_context"] = request.usage_context;
        }
        const query = {};
        const responseHeaders = ["served-model-name"];
        const path = `/serving-endpoints/${request.name}/invocations`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            responseHeaders,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Query a serving endpoint
     */
    async query(request, context) {
        return await this._query(request, context);
    }
    async _setPermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/serving-endpoints/${request.serving_endpoint_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their root object.
     */
    async setPermissions(request, context) {
        return await this._setPermissions(request, context);
    }
    async _updateConfig(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("auto_capture_config")) {
            body["auto_capture_config"] = request.auto_capture_config;
        }
        if (request.hasOwnProperty("served_entities")) {
            body["served_entities"] = request.served_entities;
        }
        if (request.hasOwnProperty("served_models")) {
            body["served_models"] = request.served_models;
        }
        if (request.hasOwnProperty("traffic_config")) {
            body["traffic_config"] = request.traffic_config;
        }
        const query = {};
        const path = `/api/2.0/serving-endpoints/${request.name}/config`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates any combination of the serving endpoint's served entities, the
     * compute configuration of those served entities, and the endpoint's traffic
     * config. An endpoint that already has an update in progress can not be
     * updated until the current update completes or fails.
     */
    async updateConfig(endpointCoreConfigInput, context) {
        let cancellationToken = context?.cancellationToken;
        const servingEndpointDetailed = await this._updateConfig(endpointCoreConfigInput, context);
        return (0, wait_1.asWaiter)(servingEndpointDetailed, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        name: servingEndpointDetailed.name,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("ServingEndpoints.updateConfigAndWait: cancelled");
                        throw new ServingEndpointsError("updateConfigAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state.config_update;
                    const statusMessage = pollResponse;
                    switch (status) {
                        case "NOT_UPDATING": {
                            return pollResponse;
                        }
                        case "UPDATE_FAILED":
                        case "UPDATE_CANCELED": {
                            const errorMessage = `failed to reach NOT_UPDATING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`ServingEndpoints.updateConfigAndWait: ${errorMessage}`);
                            throw new ServingEndpointsError("updateConfigAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach NOT_UPDATING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`ServingEndpoints.updateConfigAndWait: retrying: ${errorMessage}`);
                            throw new ServingEndpointsRetriableError("updateConfigAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _updateNotifications(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("email_notifications")) {
            body["email_notifications"] = request.email_notifications;
        }
        const query = {};
        const path = `/api/2.0/serving-endpoints/${request.name}/notifications`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the email and webhook notification settings for an endpoint.
     */
    async updateNotifications(request, context) {
        return await this._updateNotifications(request, context);
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/serving-endpoints/${request.serving_endpoint_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on a serving endpoint. Serving endpoints can
     * inherit permissions from their root object.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
    async _updateProvisionedThroughputEndpointConfig(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("config")) {
            body["config"] = request.config;
        }
        const query = {};
        const path = `/api/2.0/serving-endpoints/pt/${request.name}/config`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates any combination of the pt endpoint's served entities, the compute
     * configuration of those served entities, and the endpoint's traffic config.
     * Updates are instantaneous and endpoint should be updated instantly
     */
    async updateProvisionedThroughputEndpointConfig(updateProvisionedThroughputEndpointConfigRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const servingEndpointDetailed = await this._updateProvisionedThroughputEndpointConfig(updateProvisionedThroughputEndpointConfigRequest, context);
        return (0, wait_1.asWaiter)(servingEndpointDetailed, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        name: servingEndpointDetailed.name,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("ServingEndpoints.updateProvisionedThroughputEndpointConfigAndWait: cancelled");
                        throw new ServingEndpointsError("updateProvisionedThroughputEndpointConfigAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state.config_update;
                    const statusMessage = pollResponse;
                    switch (status) {
                        case "NOT_UPDATING": {
                            return pollResponse;
                        }
                        case "UPDATE_FAILED":
                        case "UPDATE_CANCELED": {
                            const errorMessage = `failed to reach NOT_UPDATING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`ServingEndpoints.updateProvisionedThroughputEndpointConfigAndWait: ${errorMessage}`);
                            throw new ServingEndpointsError("updateProvisionedThroughputEndpointConfigAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach NOT_UPDATING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`ServingEndpoints.updateProvisionedThroughputEndpointConfigAndWait: retrying: ${errorMessage}`);
                            throw new ServingEndpointsRetriableError("updateProvisionedThroughputEndpointConfigAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
}
exports.ServingEndpointsService = ServingEndpointsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_buildLogs", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "buildLogs", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_createProvisionedThroughputEndpoint", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "createProvisionedThroughputEndpoint", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_exportMetrics", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "exportMetrics", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_getOpenApi", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "getOpenApi", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_httpRequest", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "httpRequest", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Object)
], ServingEndpointsService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_logs", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "logs", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_put", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "put", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_putAiGateway", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "putAiGateway", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_query", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "query", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_updateConfig", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "updateConfig", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_updateNotifications", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "updateNotifications", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "_updateProvisionedThroughputEndpointConfig", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsService.prototype, "updateProvisionedThroughputEndpointConfig", null);
class ServingEndpointsDataPlaneRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("ServingEndpointsDataPlane", method, message);
    }
}
exports.ServingEndpointsDataPlaneRetriableError = ServingEndpointsDataPlaneRetriableError;
class ServingEndpointsDataPlaneError extends apiError_1.ApiError {
    constructor(method, message) {
        super("ServingEndpointsDataPlane", method, message);
    }
}
exports.ServingEndpointsDataPlaneError = ServingEndpointsDataPlaneError;
/**
 * Serving endpoints DataPlane provides a set of operations to interact with data
 * plane endpoints for Serving endpoints service.
 */
class ServingEndpointsDataPlaneService {
    constructor(client) {
        this.client = client;
    }
    async _query(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("client_request_id")) {
            body["client_request_id"] = request.client_request_id;
        }
        if (request.hasOwnProperty("dataframe_records")) {
            body["dataframe_records"] = request.dataframe_records;
        }
        if (request.hasOwnProperty("dataframe_split")) {
            body["dataframe_split"] = request.dataframe_split;
        }
        if (request.hasOwnProperty("extra_params")) {
            body["extra_params"] = request.extra_params;
        }
        if (request.hasOwnProperty("input")) {
            body["input"] = request.input;
        }
        if (request.hasOwnProperty("inputs")) {
            body["inputs"] = request.inputs;
        }
        if (request.hasOwnProperty("instances")) {
            body["instances"] = request.instances;
        }
        if (request.hasOwnProperty("max_tokens")) {
            body["max_tokens"] = request.max_tokens;
        }
        if (request.hasOwnProperty("messages")) {
            body["messages"] = request.messages;
        }
        if (request.hasOwnProperty("n")) {
            body["n"] = request.n;
        }
        if (request.hasOwnProperty("prompt")) {
            body["prompt"] = request.prompt;
        }
        if (request.hasOwnProperty("stop")) {
            body["stop"] = request.stop;
        }
        if (request.hasOwnProperty("stream")) {
            body["stream"] = request.stream;
        }
        if (request.hasOwnProperty("temperature")) {
            body["temperature"] = request.temperature;
        }
        if (request.hasOwnProperty("usage_context")) {
            body["usage_context"] = request.usage_context;
        }
        const query = {};
        const responseHeaders = ["served-model-name"];
        const path = `/serving-endpoints/${request.name}/invocations`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            responseHeaders,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Query a serving endpoint
     */
    async query(request, context) {
        return await this._query(request, context);
    }
}
exports.ServingEndpointsDataPlaneService = ServingEndpointsDataPlaneService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsDataPlaneService.prototype, "_query", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServingEndpointsDataPlaneService.prototype, "query", null);
