// @databricks/appkit-ui/dist/js/config.js
//#region src/js/config.ts
const APPKIT_CONFIG_SCRIPT_ID = "__appkit__";
const EMPTY_CONFIG = Object.freeze({
	appName: "",
	queries: Object.freeze({}),
	endpoints: Object.freeze({}),
	plugins: Object.freeze({})
});
function normalizeClientConfig(config) {
	if (!config || typeof config !== "object" || Array.isArray(config)) return EMPTY_CONFIG;
	const normalized = config;
	return {
		appName: normalized.appName ?? EMPTY_CONFIG.appName,
		queries: normalized.queries ?? EMPTY_CONFIG.queries,
		endpoints: normalized.endpoints ?? EMPTY_CONFIG.endpoints,
		plugins: normalized.plugins ?? EMPTY_CONFIG.plugins
	};
}
function readClientConfigFromDom() {
	if (typeof document === "undefined") return;
	const configScript = document.getElementById(APPKIT_CONFIG_SCRIPT_ID);
	if (!configScript?.textContent) return;
	try {
		return normalizeClientConfig(JSON.parse(configScript.textContent));
	} catch (error) {
		console.warn("[appkit] Failed to parse config from DOM:", error);
		return;
	}
}
let _cache;
/**
* @internal Reset the module-scoped config cache. Test utility only.
*/
function _resetConfigCache() {
	_cache = void 0;
}
function getClientConfig() {
	if (typeof window === "undefined") return EMPTY_CONFIG;
	if (!_cache) _cache = readClientConfigFromDom() ?? normalizeClientConfig(window.__appkit__);
	return _cache;
}
const EMPTY_PLUGIN_CONFIG = Object.freeze({});
function getPluginClientConfig(pluginName) {
	return getClientConfig().plugins[pluginName] ?? EMPTY_PLUGIN_CONFIG;
}

//#endregion
export { _resetConfigCache, getClientConfig, getPluginClientConfig };
