// @databricks/sdk-experimental/dist/api-client.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
require(".");
const node_assert_1 = __importDefault(require("node:assert"));
const http = __importStar(require("node:http"));
const https = __importStar(require("node:https"));
const api_client_1 = require("./api-client");
const _1 = require(".");
// eslint-disable-next-line @typescript-eslint/no-require-imports
const sdkVersion = require("../package.json").version;
// Store original fetch for restoration
const originalFetch = global.fetch;
describe(__filename, () => {
    beforeEach(() => {
        delete process.env.DATABRICKS_CONFIG_FILE;
    });
    afterEach(() => {
        // Restore original fetch after each test
        global.fetch = originalFetch;
    });
    it("should create proper user agent", () => {
        const ua = new api_client_1.ApiClient(new _1.Config({
            authType: "pat",
        }), {
            product: "unit",
            productVersion: "3.4.5",
        }).userAgent();
        node_assert_1.default.equal(ua, `unit/3.4.5 databricks-sdk-js/${sdkVersion} nodejs/${process.version.slice(1)} os/${process.platform} auth/pat`);
    });
    it("should correctly flatten and encode search parameters", () => {
        const client = new api_client_1.ApiClient(new _1.Config({ authType: "pat" }));
        const payload = {
            key1: "value1",
            key2: {
                nestedKey1: "nestedValue1",
                nestedKey2: ["arrayValue1", 2],
            },
            key3: false,
        };
        const expectedParams = "key1=value1&key2.nestedKey1=nestedValue1&key2.nestedKey2=arrayValue1&key2.nestedKey2=2&key3=false";
        const searchParams = client.searchParams(payload);
        node_assert_1.default.equal(searchParams, expectedParams);
    });
    it("should handle empty values in search parameters", () => {
        const client = new api_client_1.ApiClient(new _1.Config({ authType: "pat" }));
        const payload = {
            key1: "value1",
            key2: {
                nestedKey1: null,
                nestedKey2: undefined,
                nestedKey3: "nestedValue3",
            },
            key3: null,
        };
        const expectedParams = "key1=value1&key2.nestedKey3=nestedValue3";
        const searchParams = client.searchParams(payload);
        node_assert_1.default.equal(searchParams, expectedParams);
    });
    it("should handle large integers in query parameters", () => {
        const client = new api_client_1.ApiClient(new _1.Config({ authType: "pat" }));
        const payload = {
            job_id: 58726373873523, // Large integer like in the bug report
            page_token: "token123",
        };
        const searchParams = client.searchParams(payload);
        node_assert_1.default.ok(searchParams.includes("job_id=58726373873523"));
        node_assert_1.default.ok(searchParams.includes("page_token=token123"));
    });
    it("should use http.Agent for http:// host", async () => {
        const config = new _1.Config({
            host: "http://127.0.0.1:8004",
            token: "test-token",
            loaders: [],
        });
        const client = new api_client_1.ApiClient(config);
        const agent = await client.getAgent();
        node_assert_1.default.ok(agent instanceof http.Agent, "agent should be an instance of http.Agent");
        node_assert_1.default.ok(!(agent instanceof https.Agent), "agent should NOT be an instance of https.Agent");
    });
    it("should use https.Agent for https:// host", async () => {
        const config = new _1.Config({
            host: "https://my-workspace.cloud.databricks.com",
            token: "test-token",
            loaders: [],
        });
        const client = new api_client_1.ApiClient(config);
        const agent = await client.getAgent();
        node_assert_1.default.ok(agent instanceof https.Agent, "agent should be an instance of https.Agent");
    });
});
