// @databricks/sdk-experimental/dist/fetch.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbortController = exports.ReadableStream = exports.Headers = void 0;
exports.fetch = fetch;
exports.readableToWeb = readableToWeb;
/* eslint-disable @typescript-eslint/naming-convention */
/**
 * This file contains a subset of the `fetch` API that is compatible with
 * Node.js. It is not a complete implementation of the `fetch` API.
 *
 * We just implement enough to make the SDK work.
 */
const node_https_1 = __importDefault(require("node:https"));
const node_http_1 = __importDefault(require("node:http"));
const node_assert_1 = __importDefault(require("node:assert"));
const node_stream_1 = require("node:stream");
const web_1 = require("stream/web");
exports.Headers = getGlobal().Headers;
class ReadableStream extends web_1.ReadableStream {
    async *[Symbol.asyncIterator]() {
        const reader = this.getReader();
        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done || value === undefined) {
                    break;
                }
                yield value;
            }
        }
        finally {
            reader.releaseLock();
        }
    }
    async [Symbol.asyncDispose]() {
        return Promise.resolve();
    }
}
exports.ReadableStream = ReadableStream;
class AbortError extends Error {
    constructor(message, type = "aborted") {
        super(message);
        this.name = "AbortError";
        this[_a] = "AbortError";
        this.type = type;
    }
}
_a = Symbol.toStringTag;
class AbortControllerPolyfill {
    constructor() {
        this.aborted = false;
        this.listeners = [];
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const self = this;
        this.signal = {
            get aborted() {
                return self.aborted;
            },
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            addEventListener: (type = "abort", listener) => {
                self.listeners.push(listener);
            },
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            removeEventListener: (type = "abort", listener) => {
                self.listeners = self.listeners.filter((l) => l !== listener);
            },
        };
        this.abort = this.abort.bind(this);
    }
    abort() {
        this.aborted = true;
        this.listeners.forEach((listener) => listener.call(this.signal));
    }
}
function toHeaders(headers) {
    const result = new exports.Headers();
    if (Array.isArray(headers)) {
        headers.forEach(([name, value]) => {
            result.append(name, value);
        });
    }
    else if (headers instanceof exports.Headers) {
        return headers;
    }
    else {
        for (const [name, value] of Object.entries(headers)) {
            result.append(name, value);
        }
    }
    return result;
}
class Response {
    constructor(nodeResponse) {
        this.nodeResponse = nodeResponse;
        this.headers = new exports.Headers();
        for (const [key, value] of Object.entries(nodeResponse.headers)) {
            this.headers.set(key, value);
        }
        this.ok =
            nodeResponse.statusCode === undefined
                ? true
                : nodeResponse.statusCode < 400;
        this.status = nodeResponse.statusCode ?? 200;
        this.statusText = nodeResponse.statusMessage ?? "";
        this.url = nodeResponse.url ?? "";
    }
    get body() {
        return readableToWeb(this.nodeResponse);
    }
    async text() {
        let body = "";
        this.nodeResponse.setEncoding("utf8");
        this.nodeResponse.on("data", (chunk) => {
            body += chunk;
        });
        await new Promise((resolve, reject) => {
            this.nodeResponse.on("end", resolve);
            this.nodeResponse.on("error", reject);
        });
        return body;
    }
    async json() {
        const text = await this.text();
        return JSON.parse(text);
    }
}
function fetch(input, init) {
    const signal = init?.signal;
    const method = init?.method ?? "GET";
    const headers = toHeaders(init?.headers ?? new exports.Headers());
    return new Promise((resolve, reject) => {
        if (signal?.aborted) {
            return reject(new AbortError("The operation was aborted."));
        }
        const url = input instanceof URL ? input : new URL(input);
        if (init?.body) {
            (0, node_assert_1.default)(method === "POST" || method === "PUT");
            if (typeof init.body === "string") {
                headers.set("Content-Length", Buffer.byteLength(init.body).toString());
            }
        }
        const protocol = url.protocol === "https:" ? node_https_1.default : node_http_1.default;
        const outgoingHeaders = {};
        headers.forEach((value, key) => {
            outgoingHeaders[key] = value;
        });
        // Databricks extension sets it based on the http.proxyStrictSSL vscode setting
        const proxyStrictSSL = process.env.DATABRICKS_SDK_PROXY_STRICT_SSL === "false"
            ? false
            : true;
        const req = protocol
            .request(url, {
            method,
            headers: outgoingHeaders,
            agent: init?.agent,
            hostname: url.hostname,
            port: url.port,
            path: url.pathname + url.search,
            rejectUnauthorized: proxyStrictSSL,
        }, (res) => {
            // redirect
            if ((res.statusCode === 301 || res.statusCode === 302) &&
                res.headers.location) {
                return fetch(res.headers.location, init);
            }
            res.on("end", () => {
                if (signal) {
                    signal.removeEventListener("abort", abort);
                }
                if (signal?.aborted) {
                    return;
                }
            });
            res.on("error", () => {
                if (signal) {
                    signal.removeEventListener("abort", abort);
                }
            });
            resolve(new Response(res));
        })
            .on("error", (err) => {
            reject(err);
        });
        if (signal) {
            signal.addEventListener("abort", abort);
        }
        function abort() {
            const err = new AbortError("The operation was aborted.");
            try {
                req.destroy();
            }
            catch (_e) {
                // pass
            }
            reject(err);
        }
        new Promise((resolve, reject) => {
            if (init?.body) {
                if (init.body instanceof web_1.ReadableStream ||
                    init.body instanceof node_stream_1.Readable) {
                    const stream = init.body instanceof web_1.ReadableStream
                        ? webToReadable(init.body)
                        : init.body;
                    stream.pipe(req, { end: true });
                    stream.on("error", reject);
                    stream.on("end", resolve);
                }
                else if (typeof init.body === "string") {
                    req.write(init.body, (err) => {
                        if (err) {
                            reject(err);
                        }
                        else {
                            req.end(() => {
                                resolve();
                            });
                        }
                    });
                }
                else {
                    reject(new Error("Unsupported body type"));
                }
            }
            else {
                req.end(() => {
                    resolve();
                });
            }
        }).catch((err) => {
            reject(err);
        });
    });
}
const abortControllerResolved = getGlobal()?.AbortController || AbortControllerPolyfill;
exports.AbortController = abortControllerResolved;
function getGlobal() {
    if (typeof window !== "undefined") {
        // eslint-disable-next-line no-undef
        return window;
    }
    if (typeof global !== "undefined") {
        return global;
    }
}
// Helper function to convert a Node.js Readable stream to a Web ReadableStream.
function readableToWeb(readable) {
    return new ReadableStream({
        start(controller) {
            readable.on("data", (chunk) => {
                controller.enqueue(chunk);
            });
            readable.on("end", () => {
                controller.close();
            });
            readable.on("error", (err) => {
                controller.error(err);
            });
        },
    });
}
function webToReadable(stream) {
    const reader = stream.getReader();
    return new node_stream_1.Readable({
        async read() {
            const { done, value } = await reader.read();
            if (done) {
                this.push(null);
            }
            else {
                this.push(value);
            }
        },
    });
}
