// @databricks/sdk-experimental/dist/WorkspaceClient.js
"use strict";
/* eslint-disable @typescript-eslint/no-unused-vars */
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceClient = void 0;
const Config_1 = require("./config/Config");
const api_client_1 = require("./api-client");
const agentbricks = __importStar(require("./apis/agentbricks"));
const apps = __importStar(require("./apis/apps"));
const catalog = __importStar(require("./apis/catalog"));
const cleanrooms = __importStar(require("./apis/cleanrooms"));
const compute = __importStar(require("./apis/compute"));
const dashboards = __importStar(require("./apis/dashboards"));
const database = __importStar(require("./apis/database"));
const dataquality = __importStar(require("./apis/dataquality"));
const files = __importStar(require("./apis/files"));
const iam = __importStar(require("./apis/iam"));
const iamv2 = __importStar(require("./apis/iamv2"));
const jobs = __importStar(require("./apis/jobs"));
const marketplace = __importStar(require("./apis/marketplace"));
const ml = __importStar(require("./apis/ml"));
const oauth2 = __importStar(require("./apis/oauth2"));
const pipelines = __importStar(require("./apis/pipelines"));
const qualitymonitorv2 = __importStar(require("./apis/qualitymonitorv2"));
const serving = __importStar(require("./apis/serving"));
const settings = __importStar(require("./apis/settings"));
const settingsv2 = __importStar(require("./apis/settingsv2"));
const sharing = __importStar(require("./apis/sharing"));
const sql = __importStar(require("./apis/sql"));
const tags = __importStar(require("./apis/tags"));
const vectorsearch = __importStar(require("./apis/vectorsearch"));
const workspace = __importStar(require("./apis/workspace"));
const compute_1 = require("./mixins/compute");
class WorkspaceClient {
    constructor(config, options = {}) {
        if (!(config instanceof Config_1.Config)) {
            config = new Config_1.Config(config);
        }
        this.config = config;
        this.apiClient = new api_client_1.ApiClient(this.config, options);
        this.accessControl = new iam.AccessControlService(this.apiClient);
        this.accountAccessControlProxy =
            new iam.AccountAccessControlProxyService(this.apiClient);
        this.agentBricks = new agentbricks.AgentBricksService(this.apiClient);
        this.alerts = new sql.AlertsService(this.apiClient);
        this.alertsLegacy = new sql.AlertsLegacyService(this.apiClient);
        this.alertsV2 = new sql.AlertsV2Service(this.apiClient);
        this.apps = new apps.AppsService(this.apiClient);
        this.appsSettings = new apps.AppsSettingsService(this.apiClient);
        this.artifactAllowlists = new catalog.ArtifactAllowlistsService(this.apiClient);
        this.catalogs = new catalog.CatalogsService(this.apiClient);
        this.cleanRoomAssetRevisions =
            new cleanrooms.CleanRoomAssetRevisionsService(this.apiClient);
        this.cleanRoomAssets = new cleanrooms.CleanRoomAssetsService(this.apiClient);
        this.cleanRoomAutoApprovalRules =
            new cleanrooms.CleanRoomAutoApprovalRulesService(this.apiClient);
        this.cleanRoomTaskRuns = new cleanrooms.CleanRoomTaskRunsService(this.apiClient);
        this.cleanRooms = new cleanrooms.CleanRoomsService(this.apiClient);
        this.clusterPolicies = new compute.ClusterPoliciesService(this.apiClient);
        this.clusters = new compute_1.ClustersExt(this.apiClient);
        this.commandExecution = new compute.CommandExecutionService(this.apiClient);
        this.connections = new catalog.ConnectionsService(this.apiClient);
        this.consumerFulfillments = new marketplace.ConsumerFulfillmentsService(this.apiClient);
        this.consumerInstallations =
            new marketplace.ConsumerInstallationsService(this.apiClient);
        this.consumerListings = new marketplace.ConsumerListingsService(this.apiClient);
        this.consumerPersonalizationRequests =
            new marketplace.ConsumerPersonalizationRequestsService(this.apiClient);
        this.consumerProviders = new marketplace.ConsumerProvidersService(this.apiClient);
        this.credentials = new catalog.CredentialsService(this.apiClient);
        this.credentialsManager = new settings.CredentialsManagerService(this.apiClient);
        this.currentUser = new iam.CurrentUserService(this.apiClient);
        this.dashboardWidgets = new sql.DashboardWidgetsService(this.apiClient);
        this.dashboards = new sql.DashboardsService(this.apiClient);
        this.dataQuality = new dataquality.DataQualityService(this.apiClient);
        this.dataSources = new sql.DataSourcesService(this.apiClient);
        this.database = new database.DatabaseService(this.apiClient);
        this.dbfs = new files.DbfsService(this.apiClient);
        this.dbsqlPermissions = new sql.DbsqlPermissionsService(this.apiClient);
        this.entityTagAssignments = new catalog.EntityTagAssignmentsService(this.apiClient);
        this.experiments = new ml.ExperimentsService(this.apiClient);
        this.externalLineage = new catalog.ExternalLineageService(this.apiClient);
        this.externalLocations = new catalog.ExternalLocationsService(this.apiClient);
        this.externalMetadata = new catalog.ExternalMetadataService(this.apiClient);
        this.featureEngineering = new ml.FeatureEngineeringService(this.apiClient);
        this.featureStore = new ml.FeatureStoreService(this.apiClient);
        this.files = new files.FilesService(this.apiClient);
        this.functions = new catalog.FunctionsService(this.apiClient);
        this.genie = new dashboards.GenieService(this.apiClient);
        this.gitCredentials = new workspace.GitCredentialsService(this.apiClient);
        this.globalInitScripts = new compute.GlobalInitScriptsService(this.apiClient);
        this.grants = new catalog.GrantsService(this.apiClient);
        this.groupsV2 = new iam.GroupsV2Service(this.apiClient);
        this.instancePools = new compute.InstancePoolsService(this.apiClient);
        this.instanceProfiles = new compute.InstanceProfilesService(this.apiClient);
        this.ipAccessLists = new settings.IpAccessListsService(this.apiClient);
        this.jobs = new jobs.JobsService(this.apiClient);
        this.lakeview = new dashboards.LakeviewService(this.apiClient);
        this.lakeviewEmbedded = new dashboards.LakeviewEmbeddedService(this.apiClient);
        this.libraries = new compute.LibrariesService(this.apiClient);
        this.materializedFeatures = new ml.MaterializedFeaturesService(this.apiClient);
        this.metastores = new catalog.MetastoresService(this.apiClient);
        this.modelRegistry = new ml.ModelRegistryService(this.apiClient);
        this.modelVersions = new catalog.ModelVersionsService(this.apiClient);
        this.notificationDestinations =
            new settings.NotificationDestinationsService(this.apiClient);
        this.onlineTables = new catalog.OnlineTablesService(this.apiClient);
        this.permissionMigration = new iam.PermissionMigrationService(this.apiClient);
        this.permissions = new iam.PermissionsService(this.apiClient);
        this.pipelines = new pipelines.PipelinesService(this.apiClient);
        this.policies = new catalog.PoliciesService(this.apiClient);
        this.policyComplianceForClusters =
            new compute.PolicyComplianceForClustersService(this.apiClient);
        this.policyComplianceForJobs = new jobs.PolicyComplianceForJobsService(this.apiClient);
        this.policyFamilies = new compute.PolicyFamiliesService(this.apiClient);
        this.providerExchangeFilters =
            new marketplace.ProviderExchangeFiltersService(this.apiClient);
        this.providerExchanges = new marketplace.ProviderExchangesService(this.apiClient);
        this.providerFiles = new marketplace.ProviderFilesService(this.apiClient);
        this.providerListings = new marketplace.ProviderListingsService(this.apiClient);
        this.providerPersonalizationRequests =
            new marketplace.ProviderPersonalizationRequestsService(this.apiClient);
        this.providerProviderAnalyticsDashboards =
            new marketplace.ProviderProviderAnalyticsDashboardsService(this.apiClient);
        this.providerProviders = new marketplace.ProviderProvidersService(this.apiClient);
        this.providers = new sharing.ProvidersService(this.apiClient);
        this.qualityMonitorV2 = new qualitymonitorv2.QualityMonitorV2Service(this.apiClient);
        this.qualityMonitors = new catalog.QualityMonitorsService(this.apiClient);
        this.queries = new sql.QueriesService(this.apiClient);
        this.queriesLegacy = new sql.QueriesLegacyService(this.apiClient);
        this.queryHistory = new sql.QueryHistoryService(this.apiClient);
        this.queryVisualizations = new sql.QueryVisualizationsService(this.apiClient);
        this.queryVisualizationsLegacy =
            new sql.QueryVisualizationsLegacyService(this.apiClient);
        this.recipientActivation = new sharing.RecipientActivationService(this.apiClient);
        this.recipientFederationPolicies =
            new sharing.RecipientFederationPoliciesService(this.apiClient);
        this.recipients = new sharing.RecipientsService(this.apiClient);
        this.redashConfig = new sql.RedashConfigService(this.apiClient);
        this.registeredModels = new catalog.RegisteredModelsService(this.apiClient);
        this.repos = new workspace.ReposService(this.apiClient);
        this.resourceQuotas = new catalog.ResourceQuotasService(this.apiClient);
        this.rfa = new catalog.RfaService(this.apiClient);
        this.schemas = new catalog.SchemasService(this.apiClient);
        this.secrets = new workspace.SecretsService(this.apiClient);
        this.servicePrincipalSecretsProxy =
            new oauth2.ServicePrincipalSecretsProxyService(this.apiClient);
        this.servicePrincipalsV2 = new iam.ServicePrincipalsV2Service(this.apiClient);
        this.servingEndpoints = new serving.ServingEndpointsService(this.apiClient);
        this.settings = new settings.SettingsService(this.apiClient);
        this.shares = new sharing.SharesService(this.apiClient);
        this.statementExecution = new sql.StatementExecutionService(this.apiClient);
        this.storageCredentials = new catalog.StorageCredentialsService(this.apiClient);
        this.systemSchemas = new catalog.SystemSchemasService(this.apiClient);
        this.tableConstraints = new catalog.TableConstraintsService(this.apiClient);
        this.tables = new catalog.TablesService(this.apiClient);
        this.tagPolicies = new tags.TagPoliciesService(this.apiClient);
        this.temporaryPathCredentials =
            new catalog.TemporaryPathCredentialsService(this.apiClient);
        this.temporaryTableCredentials =
            new catalog.TemporaryTableCredentialsService(this.apiClient);
        this.tokenManagement = new settings.TokenManagementService(this.apiClient);
        this.tokens = new settings.TokensService(this.apiClient);
        this.usersV2 = new iam.UsersV2Service(this.apiClient);
        this.vectorSearchEndpoints =
            new vectorsearch.VectorSearchEndpointsService(this.apiClient);
        this.vectorSearchIndexes = new vectorsearch.VectorSearchIndexesService(this.apiClient);
        this.volumes = new catalog.VolumesService(this.apiClient);
        this.warehouses = new sql.WarehousesService(this.apiClient);
        this.workspace = new workspace.WorkspaceService(this.apiClient);
        this.workspaceBindings = new catalog.WorkspaceBindingsService(this.apiClient);
        this.workspaceConf = new settings.WorkspaceConfService(this.apiClient);
        this.workspaceSettingsV2 = new settingsv2.WorkspaceSettingsV2Service(this.apiClient);
        this.forecasting = new ml.ForecastingService(this.apiClient);
        this.workspaceIamV2 = new iamv2.WorkspaceIamV2Service(this.apiClient);
    }
}
exports.WorkspaceClient = WorkspaceClient;
