// @databricks/sdk-experimental/dist/apis/pipelines/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PipelinesService = exports.PipelinesError = exports.PipelinesRetriableError = void 0;
const pipelines = __importStar(require("./model"));
const retries_1 = __importDefault(require("../../retries/retries"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
const wait_1 = require("../../wait");
class PipelinesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Pipelines", method, message);
    }
}
exports.PipelinesRetriableError = PipelinesRetriableError;
class PipelinesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Pipelines", method, message);
    }
}
exports.PipelinesError = PipelinesError;
/**
 * The Lakeflow Spark Declarative Pipelines API allows you to create, edit,
 * delete, start, and view details about pipelines.
 *
 * Spark Declarative Pipelines is a framework for building reliable,
 * maintainable, and testable data processing pipelines. You define the
 * transformations to perform on your data, and Spark Declarative Pipelines
 * manages task orchestration, cluster management, monitoring, data quality, and
 * error handling.
 *
 * Instead of defining your data pipelines using a series of separate Apache
 * Spark tasks, Spark Declarative Pipelines manages how your data is transformed
 * based on a target schema you define for each processing step. You can also
 * enforce data quality with Spark Declarative Pipelines expectations.
 * Expectations allow you to define expected data quality and specify how to
 * handle records that fail those expectations.
 */
class PipelinesService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("allow_duplicate_names")) {
            body["allow_duplicate_names"] = request.allow_duplicate_names;
        }
        if (request.hasOwnProperty("budget_policy_id")) {
            body["budget_policy_id"] = request.budget_policy_id;
        }
        if (request.hasOwnProperty("catalog")) {
            body["catalog"] = request.catalog;
        }
        if (request.hasOwnProperty("channel")) {
            body["channel"] = request.channel;
        }
        if (request.hasOwnProperty("clusters")) {
            body["clusters"] = request.clusters;
        }
        if (request.hasOwnProperty("configuration")) {
            body["configuration"] = request.configuration;
        }
        if (request.hasOwnProperty("continuous")) {
            body["continuous"] = request.continuous;
        }
        if (request.hasOwnProperty("deployment")) {
            body["deployment"] = request.deployment;
        }
        if (request.hasOwnProperty("development")) {
            body["development"] = request.development;
        }
        if (request.hasOwnProperty("dry_run")) {
            body["dry_run"] = request.dry_run;
        }
        if (request.hasOwnProperty("edition")) {
            body["edition"] = request.edition;
        }
        if (request.hasOwnProperty("environment")) {
            body["environment"] = request.environment;
        }
        if (request.hasOwnProperty("event_log")) {
            body["event_log"] = request.event_log;
        }
        if (request.hasOwnProperty("filters")) {
            body["filters"] = request.filters;
        }
        if (request.hasOwnProperty("gateway_definition")) {
            body["gateway_definition"] = request.gateway_definition;
        }
        if (request.hasOwnProperty("id")) {
            body["id"] = request.id;
        }
        if (request.hasOwnProperty("ingestion_definition")) {
            body["ingestion_definition"] = request.ingestion_definition;
        }
        if (request.hasOwnProperty("libraries")) {
            body["libraries"] = request.libraries;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("notifications")) {
            body["notifications"] = request.notifications;
        }
        if (request.hasOwnProperty("photon")) {
            body["photon"] = request.photon;
        }
        if (request.hasOwnProperty("restart_window")) {
            body["restart_window"] = request.restart_window;
        }
        if (request.hasOwnProperty("root_path")) {
            body["root_path"] = request.root_path;
        }
        if (request.hasOwnProperty("run_as")) {
            body["run_as"] = request.run_as;
        }
        if (request.hasOwnProperty("schema")) {
            body["schema"] = request.schema;
        }
        if (request.hasOwnProperty("serverless")) {
            body["serverless"] = request.serverless;
        }
        if (request.hasOwnProperty("storage")) {
            body["storage"] = request.storage;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        if (request.hasOwnProperty("target")) {
            body["target"] = request.target;
        }
        if (request.hasOwnProperty("trigger")) {
            body["trigger"] = request.trigger;
        }
        if (request.hasOwnProperty("usage_policy_id")) {
            body["usage_policy_id"] = request.usage_policy_id;
        }
        const query = {};
        const path = "/api/2.0/pipelines";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new data processing pipeline based on the requested
     * configuration. If successful, this method returns the ID of the new
     * pipeline.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/pipelines/${request.pipeline_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a pipeline. Deleting a pipeline is a permanent action that stops
     * and removes the pipeline and its tables. You cannot undo this action.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/pipelines/${request.pipeline_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a pipeline.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _getPermissionLevels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/pipelines/${request.pipeline_id}/permissionLevels`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(request, context) {
        return await this._getPermissionLevels(request, context);
    }
    async _getPermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/pipelines/${request.pipeline_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of a pipeline. Pipelines can inherit permissions from
     * their root object.
     */
    async getPermissions(request, context) {
        return await this._getPermissions(request, context);
    }
    async _getUpdate(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/pipelines/${request.pipeline_id}/updates/${request.update_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets an update from an active pipeline.
     */
    async getUpdate(request, context) {
        return await this._getUpdate(request, context);
    }
    async _listPipelineEvents(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("filter")) {
            query["filter"] = request.filter;
        }
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("order_by")) {
            query["order_by"] =
                request.order_by;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/pipelines/${request.pipeline_id}/events`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieves events for a pipeline.
     */
    async *listPipelineEvents(request, context) {
        while (true) {
            const response = await this._listPipelineEvents(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.events || response.events.length === 0) {
                break;
            }
            for (const v of response.events) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
            // page_token field is mutually exclusive with all fields in the request except max_results.
            // An error is returned if any fields other than max_results are set when page_token is set.
            request = {
                pipeline_id: request.pipeline_id,
                max_results: request.max_results,
                page_token: response.next_page_token,
            };
            if (!request.max_results) {
                delete request.max_results;
            }
        }
    }
    async _listPipelines(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("filter")) {
            query["filter"] = request.filter;
        }
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("order_by")) {
            query["order_by"] =
                request.order_by;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/pipelines";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists pipelines defined in the Spark Declarative Pipelines system.
     */
    async *listPipelines(request, context) {
        while (true) {
            const response = await this._listPipelines(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.statuses || response.statuses.length === 0) {
                break;
            }
            for (const v of response.statuses) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _listUpdates(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        if (request.hasOwnProperty("until_update_id")) {
            query["until_update_id"] =
                request.until_update_id;
        }
        const path = `/api/2.0/pipelines/${request.pipeline_id}/updates`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List updates for an active pipeline.
     */
    async listUpdates(request, context) {
        return await this._listUpdates(request, context);
    }
    async _setPermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/pipelines/${request.pipeline_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their root object.
     */
    async setPermissions(request, context) {
        return await this._setPermissions(request, context);
    }
    async _startUpdate(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("cause")) {
            body["cause"] = request.cause;
        }
        if (request.hasOwnProperty("full_refresh")) {
            body["full_refresh"] = request.full_refresh;
        }
        if (request.hasOwnProperty("full_refresh_selection")) {
            body["full_refresh_selection"] = request.full_refresh_selection;
        }
        if (request.hasOwnProperty("refresh_selection")) {
            body["refresh_selection"] = request.refresh_selection;
        }
        if (request.hasOwnProperty("validate_only")) {
            body["validate_only"] = request.validate_only;
        }
        const query = {};
        const path = `/api/2.0/pipelines/${request.pipeline_id}/updates`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Starts a new update for the pipeline. If there is already an active update
     * for the pipeline, the request will fail and the active update will remain
     * running.
     */
    async startUpdate(request, context) {
        return await this._startUpdate(request, context);
    }
    async _stop(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/pipelines/${request.pipeline_id}/stop`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Stops the pipeline by canceling the active update. If there is no active
     * update for the pipeline, this request is a no-op.
     */
    async stop(stopRequest, context) {
        let cancellationToken = context?.cancellationToken;
        await this._stop(stopRequest, context);
        return (0, wait_1.asWaiter)(null, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        pipeline_id: stopRequest.pipeline_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Pipelines.stopAndWait: cancelled");
                        throw new PipelinesError("stopAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state;
                    const statusMessage = pollResponse.cause;
                    switch (status) {
                        case "IDLE": {
                            return pollResponse;
                        }
                        case "FAILED": {
                            const errorMessage = `failed to reach IDLE state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Pipelines.stopAndWait: ${errorMessage}`);
                            throw new PipelinesError("stopAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach IDLE state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Pipelines.stopAndWait: retrying: ${errorMessage}`);
                            throw new PipelinesRetriableError("stopAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("allow_duplicate_names")) {
            body["allow_duplicate_names"] = request.allow_duplicate_names;
        }
        if (request.hasOwnProperty("budget_policy_id")) {
            body["budget_policy_id"] = request.budget_policy_id;
        }
        if (request.hasOwnProperty("catalog")) {
            body["catalog"] = request.catalog;
        }
        if (request.hasOwnProperty("channel")) {
            body["channel"] = request.channel;
        }
        if (request.hasOwnProperty("clusters")) {
            body["clusters"] = request.clusters;
        }
        if (request.hasOwnProperty("configuration")) {
            body["configuration"] = request.configuration;
        }
        if (request.hasOwnProperty("continuous")) {
            body["continuous"] = request.continuous;
        }
        if (request.hasOwnProperty("deployment")) {
            body["deployment"] = request.deployment;
        }
        if (request.hasOwnProperty("development")) {
            body["development"] = request.development;
        }
        if (request.hasOwnProperty("edition")) {
            body["edition"] = request.edition;
        }
        if (request.hasOwnProperty("environment")) {
            body["environment"] = request.environment;
        }
        if (request.hasOwnProperty("event_log")) {
            body["event_log"] = request.event_log;
        }
        if (request.hasOwnProperty("expected_last_modified")) {
            body["expected_last_modified"] = request.expected_last_modified;
        }
        if (request.hasOwnProperty("filters")) {
            body["filters"] = request.filters;
        }
        if (request.hasOwnProperty("gateway_definition")) {
            body["gateway_definition"] = request.gateway_definition;
        }
        if (request.hasOwnProperty("id")) {
            body["id"] = request.id;
        }
        if (request.hasOwnProperty("ingestion_definition")) {
            body["ingestion_definition"] = request.ingestion_definition;
        }
        if (request.hasOwnProperty("libraries")) {
            body["libraries"] = request.libraries;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("notifications")) {
            body["notifications"] = request.notifications;
        }
        if (request.hasOwnProperty("photon")) {
            body["photon"] = request.photon;
        }
        if (request.hasOwnProperty("restart_window")) {
            body["restart_window"] = request.restart_window;
        }
        if (request.hasOwnProperty("root_path")) {
            body["root_path"] = request.root_path;
        }
        if (request.hasOwnProperty("run_as")) {
            body["run_as"] = request.run_as;
        }
        if (request.hasOwnProperty("schema")) {
            body["schema"] = request.schema;
        }
        if (request.hasOwnProperty("serverless")) {
            body["serverless"] = request.serverless;
        }
        if (request.hasOwnProperty("storage")) {
            body["storage"] = request.storage;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        if (request.hasOwnProperty("target")) {
            body["target"] = request.target;
        }
        if (request.hasOwnProperty("trigger")) {
            body["trigger"] = request.trigger;
        }
        if (request.hasOwnProperty("usage_policy_id")) {
            body["usage_policy_id"] = request.usage_policy_id;
        }
        const query = {};
        const path = `/api/2.0/pipelines/${request.pipeline_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates a pipeline with the supplied configuration.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/pipelines/${request.pipeline_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on a pipeline. Pipelines can inherit permissions
     * from their root object.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
}
exports.PipelinesService = PipelinesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_getUpdate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "getUpdate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_listPipelineEvents", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], PipelinesService.prototype, "listPipelineEvents", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_listPipelines", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], PipelinesService.prototype, "listPipelines", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_listUpdates", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "listUpdates", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_startUpdate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "startUpdate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_stop", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "stop", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PipelinesService.prototype, "updatePermissions", null);
