// @databricks/sdk-experimental/dist/apis/sql/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WarehousesService = exports.WarehousesError = exports.WarehousesRetriableError = exports.StatementExecutionService = exports.StatementExecutionError = exports.StatementExecutionRetriableError = exports.RedashConfigService = exports.RedashConfigError = exports.RedashConfigRetriableError = exports.QueryVisualizationsLegacyService = exports.QueryVisualizationsLegacyError = exports.QueryVisualizationsLegacyRetriableError = exports.QueryVisualizationsService = exports.QueryVisualizationsError = exports.QueryVisualizationsRetriableError = exports.QueryHistoryService = exports.QueryHistoryError = exports.QueryHistoryRetriableError = exports.QueriesLegacyService = exports.QueriesLegacyError = exports.QueriesLegacyRetriableError = exports.QueriesService = exports.QueriesError = exports.QueriesRetriableError = exports.DbsqlPermissionsService = exports.DbsqlPermissionsError = exports.DbsqlPermissionsRetriableError = exports.DataSourcesService = exports.DataSourcesError = exports.DataSourcesRetriableError = exports.DashboardsService = exports.DashboardsError = exports.DashboardsRetriableError = exports.DashboardWidgetsService = exports.DashboardWidgetsError = exports.DashboardWidgetsRetriableError = exports.AlertsV2Service = exports.AlertsV2Error = exports.AlertsV2RetriableError = exports.AlertsLegacyService = exports.AlertsLegacyError = exports.AlertsLegacyRetriableError = exports.AlertsService = exports.AlertsError = exports.AlertsRetriableError = void 0;
const sql = __importStar(require("./model"));
const retries_1 = __importDefault(require("../../retries/retries"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
const wait_1 = require("../../wait");
class AlertsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Alerts", method, message);
    }
}
exports.AlertsRetriableError = AlertsRetriableError;
class AlertsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Alerts", method, message);
    }
}
exports.AlertsError = AlertsError;
/**
 * The alerts API can be used to perform CRUD operations on alerts. An alert is a
 * Databricks SQL object that periodically runs a query, evaluates a condition of
 * its result, and notifies one or more users and/or notification destinations if
 * the condition was met. Alerts can be scheduled using the `sql_task` type of
 * the Jobs API, e.g. :method:jobs/create.
 */
class AlertsService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("alert")) {
            body["alert"] = request.alert;
        }
        if (request.hasOwnProperty("auto_resolve_display_name")) {
            body["auto_resolve_display_name"] =
                request.auto_resolve_display_name;
        }
        const query = {};
        const path = "/api/2.0/sql/alerts";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates an alert.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/sql/alerts/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Moves an alert to the trash. Trashed alerts immediately disappear from
     * searches and list views, and can no longer trigger. You can restore a
     * trashed alert through the UI. A trashed alert is permanently deleted after
     * 30 days.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/sql/alerts/${request.id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets an alert.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/sql/alerts";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a list of alerts accessible to the user, ordered by creation time.
     * **Warning:** Calling this API concurrently 10 or more times could result
     * in throttling, service degradation, or a temporary ban.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.results || response.results.length === 0) {
                break;
            }
            for (const v of response.results) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("alert")) {
            body["alert"] = request.alert;
        }
        if (request.hasOwnProperty("auto_resolve_display_name")) {
            body["auto_resolve_display_name"] =
                request.auto_resolve_display_name;
        }
        if (request.hasOwnProperty("update_mask")) {
            body["update_mask"] = request.update_mask;
        }
        const query = {};
        const path = `/api/2.0/sql/alerts/${request.id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates an alert.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.AlertsService = AlertsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], AlertsService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsService.prototype, "update", null);
class AlertsLegacyRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("AlertsLegacy", method, message);
    }
}
exports.AlertsLegacyRetriableError = AlertsLegacyRetriableError;
class AlertsLegacyError extends apiError_1.ApiError {
    constructor(method, message) {
        super("AlertsLegacy", method, message);
    }
}
exports.AlertsLegacyError = AlertsLegacyError;
/**
 * The alerts API can be used to perform CRUD operations on alerts. An alert is a
 * Databricks SQL object that periodically runs a query, evaluates a condition of
 * its result, and notifies one or more users and/or notification destinations if
 * the condition was met. Alerts can be scheduled using the `sql_task` type of
 * the Jobs API, e.g. :method:jobs/create.
 *
 * **Note**: A new version of the Databricks SQL API is now available. Please see
 * the latest version. [Learn more]
 *
 * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
 */
class AlertsLegacyService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("options")) {
            body["options"] = request.options;
        }
        if (request.hasOwnProperty("parent")) {
            body["parent"] = request.parent;
        }
        if (request.hasOwnProperty("query_id")) {
            body["query_id"] = request.query_id;
        }
        if (request.hasOwnProperty("rearm")) {
            body["rearm"] = request.rearm;
        }
        const query = {};
        const path = "/api/2.0/preview/sql/alerts";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates an alert. An alert is a Databricks SQL object that periodically
     * runs a query, evaluates a condition of its result, and notifies users or
     * notification destinations if the condition was met.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:alerts/create instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/sql/alerts/${request.alert_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes an alert. Deleted alerts are no longer accessible and cannot be
     * restored. **Note**: Unlike queries and dashboards, alerts cannot be moved
     * to the trash.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:alerts/delete instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/sql/alerts/${request.alert_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets an alert.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:alerts/get instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/preview/sql/alerts";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a list of alerts.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:alerts/list instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async list(context) {
        return await this._list(context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("options")) {
            body["options"] = request.options;
        }
        if (request.hasOwnProperty("query_id")) {
            body["query_id"] = request.query_id;
        }
        if (request.hasOwnProperty("rearm")) {
            body["rearm"] = request.rearm;
        }
        const query = {};
        const path = `/api/2.0/preview/sql/alerts/${request.alert_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates an alert.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:alerts/update instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.AlertsLegacyService = AlertsLegacyService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsLegacyService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsLegacyService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsLegacyService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsLegacyService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsLegacyService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsLegacyService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsLegacyService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsLegacyService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsLegacyService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsLegacyService.prototype, "update", null);
class AlertsV2RetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("AlertsV2", method, message);
    }
}
exports.AlertsV2RetriableError = AlertsV2RetriableError;
class AlertsV2Error extends apiError_1.ApiError {
    constructor(method, message) {
        super("AlertsV2", method, message);
    }
}
exports.AlertsV2Error = AlertsV2Error;
/**
 * New version of SQL Alerts
 */
class AlertsV2Service {
    constructor(client) {
        this.client = client;
    }
    async _createAlert(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.alert;
        const query = {};
        const path = "/api/2.0/alerts";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create Alert
     */
    async createAlert(request, context) {
        return await this._createAlert(request, context);
    }
    async _getAlert(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/alerts/${request.id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets an alert.
     */
    async getAlert(request, context) {
        return await this._getAlert(request, context);
    }
    async _listAlerts(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/alerts";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a list of alerts accessible to the user, ordered by creation time.
     */
    async *listAlerts(request, context) {
        while (true) {
            const response = await this._listAlerts(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.alerts || response.alerts.length === 0) {
                break;
            }
            for (const v of response.alerts) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _trashAlert(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/alerts/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Moves an alert to the trash. Trashed alerts immediately disappear from
     * list views, and can no longer trigger. You can restore a trashed alert
     * through the UI. A trashed alert is permanently deleted after 30 days.
     */
    async trashAlert(request, context) {
        return await this._trashAlert(request, context);
    }
    async _updateAlert(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.alert;
        const query = {};
        if (request.hasOwnProperty("update_mask")) {
            query["update_mask"] =
                request.update_mask;
        }
        const path = `/api/2.0/alerts/${request.id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update alert
     */
    async updateAlert(request, context) {
        return await this._updateAlert(request, context);
    }
}
exports.AlertsV2Service = AlertsV2Service;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsV2Service.prototype, "_createAlert", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsV2Service.prototype, "createAlert", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsV2Service.prototype, "_getAlert", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsV2Service.prototype, "getAlert", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsV2Service.prototype, "_listAlerts", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], AlertsV2Service.prototype, "listAlerts", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsV2Service.prototype, "_trashAlert", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsV2Service.prototype, "trashAlert", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsV2Service.prototype, "_updateAlert", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AlertsV2Service.prototype, "updateAlert", null);
class DashboardWidgetsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("DashboardWidgets", method, message);
    }
}
exports.DashboardWidgetsRetriableError = DashboardWidgetsRetriableError;
class DashboardWidgetsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("DashboardWidgets", method, message);
    }
}
exports.DashboardWidgetsError = DashboardWidgetsError;
/**
 * This is an evolving API that facilitates the addition and removal of widgets
 * from existing dashboards within the Databricks Workspace. Data structures may
 * change over time.
 */
class DashboardWidgetsService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("dashboard_id")) {
            body["dashboard_id"] = request.dashboard_id;
        }
        if (request.hasOwnProperty("options")) {
            body["options"] = request.options;
        }
        if (request.hasOwnProperty("text")) {
            body["text"] = request.text;
        }
        if (request.hasOwnProperty("visualization_id")) {
            body["visualization_id"] = request.visualization_id;
        }
        if (request.hasOwnProperty("width")) {
            body["width"] = request.width;
        }
        const query = {};
        const path = "/api/2.0/preview/sql/widgets";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Adds a widget to a dashboard
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/sql/widgets/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Removes a widget from a dashboard
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("dashboard_id")) {
            body["dashboard_id"] = request.dashboard_id;
        }
        if (request.hasOwnProperty("options")) {
            body["options"] = request.options;
        }
        if (request.hasOwnProperty("text")) {
            body["text"] = request.text;
        }
        if (request.hasOwnProperty("visualization_id")) {
            body["visualization_id"] = request.visualization_id;
        }
        if (request.hasOwnProperty("width")) {
            body["width"] = request.width;
        }
        const query = {};
        const path = `/api/2.0/preview/sql/widgets/${request.id}`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates an existing widget
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.DashboardWidgetsService = DashboardWidgetsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardWidgetsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardWidgetsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardWidgetsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardWidgetsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardWidgetsService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardWidgetsService.prototype, "update", null);
class DashboardsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Dashboards", method, message);
    }
}
exports.DashboardsRetriableError = DashboardsRetriableError;
class DashboardsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Dashboards", method, message);
    }
}
exports.DashboardsError = DashboardsError;
/**
 * In general, there is little need to modify dashboards using the API. However,
 * it can be useful to use dashboard objects to look-up a collection of related
 * query IDs. The API can also be used to duplicate multiple dashboards at once
 * since you can get a dashboard definition with a GET request and then POST it
 * to create a new one. Dashboards can be scheduled using the `sql_task` type of
 * the Jobs API, e.g. :method:jobs/create.
 */
class DashboardsService {
    constructor(client) {
        this.client = client;
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/sql/dashboards/${request.dashboard_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Moves a dashboard to the trash. Trashed dashboards do not appear in list
     * views or searches, and cannot be shared.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/sql/dashboards/${request.dashboard_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns a JSON representation of a dashboard object, including its
     * visualization and query objects.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("order")) {
            query["order"] = request.order;
        }
        if (request.hasOwnProperty("page")) {
            query["page"] = request.page;
        }
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("q")) {
            query["q"] = request.q;
        }
        const path = "/api/2.0/preview/sql/dashboards";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Fetch a paginated list of dashboard objects.
     *
     * **Warning**: Calling this API concurrently 10 or more times could result
     * in throttling, service degradation, or a temporary ban.
     */
    async *list(request, context) {
        request.page = 1; // start iterating from the first page
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.results || response.results.length === 0) {
                break;
            }
            for (const v of response.results) {
                yield v;
            }
            // paginate by increments of 1
            request.page += 1;
        }
    }
    async _restore(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/sql/dashboards/trash/${request.dashboard_id}`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * A restored dashboard appears in list views and searches and can be shared.
     */
    async restore(request, context) {
        return await this._restore(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("run_as_role")) {
            body["run_as_role"] = request.run_as_role;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        const query = {};
        const path = `/api/2.0/preview/sql/dashboards/${request.dashboard_id}`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Modify this dashboard definition. This operation only affects attributes
     * of the dashboard object. It does not add, modify, or remove widgets.
     *
     * **Note**: You cannot undo this operation.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.DashboardsService = DashboardsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], DashboardsService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardsService.prototype, "_restore", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardsService.prototype, "restore", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardsService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DashboardsService.prototype, "update", null);
class DataSourcesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("DataSources", method, message);
    }
}
exports.DataSourcesRetriableError = DataSourcesRetriableError;
class DataSourcesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("DataSources", method, message);
    }
}
exports.DataSourcesError = DataSourcesError;
/**
 * This API is provided to assist you in making new query objects. When creating
 * a query object, you may optionally specify a `data_source_id` for the SQL
 * warehouse against which it will run. If you don't already know the
 * `data_source_id` for your desired SQL warehouse, this API will help you find
 * it.
 *
 * This API does not support searches. It returns the full list of SQL warehouses
 * in your workspace. We advise you to use any text editor, REST client, or
 * `grep` to search the response from this API for the name of your SQL warehouse
 * as it appears in Databricks SQL.
 *
 * **Note**: A new version of the Databricks SQL API is now available. [Learn
 * more]
 *
 * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
 */
class DataSourcesService {
    constructor(client) {
        this.client = client;
    }
    async _list(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/preview/sql/data_sources";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieves a full list of SQL warehouses available in this workspace. All
     * fields that appear in this API response are enumerated for clarity.
     * However, you need only a SQL warehouse's `id` to create new queries
     * against it.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:warehouses/list instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async list(context) {
        return await this._list(context);
    }
}
exports.DataSourcesService = DataSourcesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], DataSourcesService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], DataSourcesService.prototype, "list", null);
class DbsqlPermissionsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("DbsqlPermissions", method, message);
    }
}
exports.DbsqlPermissionsRetriableError = DbsqlPermissionsRetriableError;
class DbsqlPermissionsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("DbsqlPermissions", method, message);
    }
}
exports.DbsqlPermissionsError = DbsqlPermissionsError;
/**
 * The SQL Permissions API is similar to the endpoints of the
 * :method:permissions/set. However, this exposes only one endpoint, which gets
 * the Access Control List for a given object. You cannot modify any permissions
 * using this API.
 *
 * There are three levels of permission:
 *
 * - `CAN_VIEW`: Allows read-only access
 *
 * - `CAN_RUN`: Allows read access and run access (superset of `CAN_VIEW`)
 *
 * - `CAN_MANAGE`: Allows all actions: read, run, edit, delete, modify
 * permissions (superset of `CAN_RUN`)
 *
 * **Note**: A new version of the Databricks SQL API is now available. [Learn
 * more]
 *
 * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
 */
class DbsqlPermissionsService {
    constructor(client) {
        this.client = client;
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/sql/permissions/${request.objectType}/${request.objectId}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a JSON representation of the access control list (ACL) for a
     * specified object.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:workspace/getpermissions instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _set(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/preview/sql/permissions/${request.objectType}/${request.objectId}`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets the access control list (ACL) for a specified object. This operation
     * will complete rewrite the ACL.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:workspace/setpermissions instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async set(request, context) {
        return await this._set(request, context);
    }
    async _transferOwnership(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("new_owner")) {
            body["new_owner"] = request.new_owner;
        }
        const query = {};
        const path = `/api/2.0/preview/sql/permissions/${request.objectType}/${request.objectId}/transfer`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Transfers ownership of a dashboard, query, or alert to an active user.
     * Requires an admin API key.
     *
     * **Note**: A new version of the Databricks SQL API is now available. For
     * queries and alerts, please use :method:queries/update and
     * :method:alerts/update respectively instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async transferOwnership(request, context) {
        return await this._transferOwnership(request, context);
    }
}
exports.DbsqlPermissionsService = DbsqlPermissionsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbsqlPermissionsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbsqlPermissionsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbsqlPermissionsService.prototype, "_set", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbsqlPermissionsService.prototype, "set", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbsqlPermissionsService.prototype, "_transferOwnership", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DbsqlPermissionsService.prototype, "transferOwnership", null);
class QueriesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Queries", method, message);
    }
}
exports.QueriesRetriableError = QueriesRetriableError;
class QueriesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Queries", method, message);
    }
}
exports.QueriesError = QueriesError;
/**
 * The queries API can be used to perform CRUD operations on queries. A query is
 * a Databricks SQL object that includes the target SQL warehouse, query text,
 * name, description, tags, and parameters. Queries can be scheduled using the
 * `sql_task` type of the Jobs API, e.g. :method:jobs/create.
 */
class QueriesService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("auto_resolve_display_name")) {
            body["auto_resolve_display_name"] =
                request.auto_resolve_display_name;
        }
        if (request.hasOwnProperty("query")) {
            body["query"] = request.query;
        }
        const query = {};
        const path = "/api/2.0/sql/queries";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a query.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/sql/queries/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Moves a query to the trash. Trashed queries immediately disappear from
     * searches and list views, and cannot be used for alerts. You can restore a
     * trashed query through the UI. A trashed query is permanently deleted after
     * 30 days.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/sql/queries/${request.id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a query.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/sql/queries";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a list of queries accessible to the user, ordered by creation time.
     * **Warning:** Calling this API concurrently 10 or more times could result
     * in throttling, service degradation, or a temporary ban.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.results || response.results.length === 0) {
                break;
            }
            for (const v of response.results) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _listVisualizations(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/sql/queries/${request.id}/visualizations`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a list of visualizations on a query.
     */
    async *listVisualizations(request, context) {
        while (true) {
            const response = await this._listVisualizations(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.results || response.results.length === 0) {
                break;
            }
            for (const v of response.results) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("auto_resolve_display_name")) {
            body["auto_resolve_display_name"] =
                request.auto_resolve_display_name;
        }
        if (request.hasOwnProperty("query")) {
            body["query"] = request.query;
        }
        if (request.hasOwnProperty("update_mask")) {
            body["update_mask"] = request.update_mask;
        }
        const query = {};
        const path = `/api/2.0/sql/queries/${request.id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates a query.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.QueriesService = QueriesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], QueriesService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesService.prototype, "_listVisualizations", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], QueriesService.prototype, "listVisualizations", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesService.prototype, "update", null);
class QueriesLegacyRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("QueriesLegacy", method, message);
    }
}
exports.QueriesLegacyRetriableError = QueriesLegacyRetriableError;
class QueriesLegacyError extends apiError_1.ApiError {
    constructor(method, message) {
        super("QueriesLegacy", method, message);
    }
}
exports.QueriesLegacyError = QueriesLegacyError;
/**
 * These endpoints are used for CRUD operations on query definitions. Query
 * definitions include the target SQL warehouse, query text, name, description,
 * tags, parameters, and visualizations. Queries can be scheduled using the
 * `sql_task` type of the Jobs API, e.g. :method:jobs/create.
 *
 * **Note**: A new version of the Databricks SQL API is now available. Please see
 * the latest version. [Learn more]
 *
 * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
 */
class QueriesLegacyService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("data_source_id")) {
            body["data_source_id"] = request.data_source_id;
        }
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("options")) {
            body["options"] = request.options;
        }
        if (request.hasOwnProperty("parent")) {
            body["parent"] = request.parent;
        }
        if (request.hasOwnProperty("query")) {
            body["query"] = request.query;
        }
        if (request.hasOwnProperty("run_as_role")) {
            body["run_as_role"] = request.run_as_role;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        const query = {};
        const path = "/api/2.0/preview/sql/queries";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new query definition. Queries created with this endpoint belong
     * to the authenticated user making the request.
     *
     * The `data_source_id` field specifies the ID of the SQL warehouse to run
     * this query against. You can use the Data Sources API to see a complete
     * list of available SQL warehouses. Or you can copy the `data_source_id`
     * from an existing query.
     *
     * **Note**: You cannot add a visualization until you create the query.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:queries/create instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/sql/queries/${request.query_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Moves a query to the trash. Trashed queries immediately disappear from
     * searches and list views, and they cannot be used for alerts. The trash is
     * deleted after 30 days.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:queries/delete instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/sql/queries/${request.query_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieve a query object definition along with contextual permissions
     * information about the currently authenticated user.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:queries/get instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("order")) {
            query["order"] = request.order;
        }
        if (request.hasOwnProperty("page")) {
            query["page"] = request.page;
        }
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("q")) {
            query["q"] = request.q;
        }
        const path = "/api/2.0/preview/sql/queries";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a list of queries. Optionally, this list can be filtered by a search
     * term.
     *
     * **Warning**: Calling this API concurrently 10 or more times could result
     * in throttling, service degradation, or a temporary ban.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:queries/list instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async *list(request, context) {
        request.page = 1; // start iterating from the first page
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.results || response.results.length === 0) {
                break;
            }
            for (const v of response.results) {
                yield v;
            }
            // paginate by increments of 1
            request.page += 1;
        }
    }
    async _restore(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/sql/queries/trash/${request.query_id}`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Restore a query that has been moved to the trash. A restored query appears
     * in list views and searches. You can use restored queries for alerts.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * see the latest version. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async restore(request, context) {
        return await this._restore(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("data_source_id")) {
            body["data_source_id"] = request.data_source_id;
        }
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("options")) {
            body["options"] = request.options;
        }
        if (request.hasOwnProperty("query")) {
            body["query"] = request.query;
        }
        if (request.hasOwnProperty("run_as_role")) {
            body["run_as_role"] = request.run_as_role;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        const query = {};
        const path = `/api/2.0/preview/sql/queries/${request.query_id}`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Modify this query definition.
     *
     * **Note**: You cannot undo this operation.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:queries/update instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.QueriesLegacyService = QueriesLegacyService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesLegacyService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesLegacyService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesLegacyService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesLegacyService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesLegacyService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesLegacyService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesLegacyService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], QueriesLegacyService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesLegacyService.prototype, "_restore", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesLegacyService.prototype, "restore", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesLegacyService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueriesLegacyService.prototype, "update", null);
class QueryHistoryRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("QueryHistory", method, message);
    }
}
exports.QueryHistoryRetriableError = QueryHistoryRetriableError;
class QueryHistoryError extends apiError_1.ApiError {
    constructor(method, message) {
        super("QueryHistory", method, message);
    }
}
exports.QueryHistoryError = QueryHistoryError;
/**
 * A service responsible for storing and retrieving the list of queries run
 * against SQL endpoints and serverless compute.
 */
class QueryHistoryService {
    constructor(client) {
        this.client = client;
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("filter_by")) {
            query["filter_by"] =
                request.filter_by;
        }
        if (request.hasOwnProperty("include_metrics")) {
            query["include_metrics"] =
                request.include_metrics;
        }
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/sql/history/queries";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List the history of queries through SQL warehouses, and serverless
     * compute.
     *
     * You can filter by user ID, warehouse ID, status, and time range. Most
     * recently started queries are returned first (up to max_results in
     * request). The pagination token returned in response can be used to list
     * subsequent query statuses.
     */
    async list(request, context) {
        return await this._list(request, context);
    }
}
exports.QueryHistoryService = QueryHistoryService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryHistoryService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryHistoryService.prototype, "list", null);
class QueryVisualizationsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("QueryVisualizations", method, message);
    }
}
exports.QueryVisualizationsRetriableError = QueryVisualizationsRetriableError;
class QueryVisualizationsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("QueryVisualizations", method, message);
    }
}
exports.QueryVisualizationsError = QueryVisualizationsError;
/**
 * This is an evolving API that facilitates the addition and removal of
 * visualizations from existing queries in the Databricks Workspace. Data
 * structures can change over time.
 */
class QueryVisualizationsService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("visualization")) {
            body["visualization"] = request.visualization;
        }
        const query = {};
        const path = "/api/2.0/sql/visualizations";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Adds a visualization to a query.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/sql/visualizations/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Removes a visualization.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("update_mask")) {
            body["update_mask"] = request.update_mask;
        }
        if (request.hasOwnProperty("visualization")) {
            body["visualization"] = request.visualization;
        }
        const query = {};
        const path = `/api/2.0/sql/visualizations/${request.id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates a visualization.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.QueryVisualizationsService = QueryVisualizationsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryVisualizationsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryVisualizationsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryVisualizationsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryVisualizationsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryVisualizationsService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryVisualizationsService.prototype, "update", null);
class QueryVisualizationsLegacyRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("QueryVisualizationsLegacy", method, message);
    }
}
exports.QueryVisualizationsLegacyRetriableError = QueryVisualizationsLegacyRetriableError;
class QueryVisualizationsLegacyError extends apiError_1.ApiError {
    constructor(method, message) {
        super("QueryVisualizationsLegacy", method, message);
    }
}
exports.QueryVisualizationsLegacyError = QueryVisualizationsLegacyError;
/**
 * This is an evolving API that facilitates the addition and removal of
 * vizualisations from existing queries within the Databricks Workspace. Data
 * structures may change over time.
 *
 * **Note**: A new version of the Databricks SQL API is now available. Please see
 * the latest version. [Learn more]
 *
 * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
 */
class QueryVisualizationsLegacyService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("options")) {
            body["options"] = request.options;
        }
        if (request.hasOwnProperty("query_id")) {
            body["query_id"] = request.query_id;
        }
        if (request.hasOwnProperty("type")) {
            body["type"] = request.type;
        }
        const query = {};
        const path = "/api/2.0/preview/sql/visualizations";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates visualization in the query.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:queryvisualizations/create instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/sql/visualizations/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Removes a visualization from the query.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:queryvisualizations/delete instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("created_at")) {
            body["created_at"] = request.created_at;
        }
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("options")) {
            body["options"] = request.options;
        }
        if (request.hasOwnProperty("query")) {
            body["query"] = request.query;
        }
        if (request.hasOwnProperty("type")) {
            body["type"] = request.type;
        }
        if (request.hasOwnProperty("updated_at")) {
            body["updated_at"] = request.updated_at;
        }
        const query = {};
        const path = `/api/2.0/preview/sql/visualizations/${request.id}`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates visualization in the query.
     *
     * **Note**: A new version of the Databricks SQL API is now available. Please
     * use :method:queryvisualizations/update instead. [Learn more]
     *
     * [Learn more]: https://docs.databricks.com/en/sql/dbsql-api-latest.html
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.QueryVisualizationsLegacyService = QueryVisualizationsLegacyService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryVisualizationsLegacyService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryVisualizationsLegacyService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryVisualizationsLegacyService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryVisualizationsLegacyService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryVisualizationsLegacyService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], QueryVisualizationsLegacyService.prototype, "update", null);
class RedashConfigRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("RedashConfig", method, message);
    }
}
exports.RedashConfigRetriableError = RedashConfigRetriableError;
class RedashConfigError extends apiError_1.ApiError {
    constructor(method, message) {
        super("RedashConfig", method, message);
    }
}
exports.RedashConfigError = RedashConfigError;
/**
 * Redash V2 service for workspace configurations (internal)
 */
class RedashConfigService {
    constructor(client) {
        this.client = client;
    }
    async _getConfig(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/redash-v2/config";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Read workspace configuration for Redash-v2.
     */
    async getConfig(context) {
        return await this._getConfig(context);
    }
}
exports.RedashConfigService = RedashConfigService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], RedashConfigService.prototype, "_getConfig", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], RedashConfigService.prototype, "getConfig", null);
class StatementExecutionRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("StatementExecution", method, message);
    }
}
exports.StatementExecutionRetriableError = StatementExecutionRetriableError;
class StatementExecutionError extends apiError_1.ApiError {
    constructor(method, message) {
        super("StatementExecution", method, message);
    }
}
exports.StatementExecutionError = StatementExecutionError;
/**
 * The Databricks SQL Statement Execution API can be used to execute SQL
 * statements on a SQL warehouse and fetch the result.
 *
 * **Getting started**
 *
 * We suggest beginning with the [Databricks SQL Statement Execution API
 * tutorial].
 *
 * **Overview of statement execution and result fetching**
 *
 * Statement execution begins by issuing a
 * :method:statementexecution/executeStatement request with a valid SQL statement
 * and warehouse ID, along with optional parameters such as the data catalog and
 * output format. If no other parameters are specified, the server will wait for
 * up to 10s before returning a response. If the statement has completed within
 * this timespan, the response will include the result data as a JSON array and
 * metadata. Otherwise, if no result is available after the 10s timeout expired,
 * the response will provide the statement ID that can be used to poll for
 * results by using a :method:statementexecution/getStatement request.
 *
 * You can specify whether the call should behave synchronously, asynchronously
 * or start synchronously with a fallback to asynchronous execution. This is
 * controlled with the `wait_timeout` and `on_wait_timeout` settings. If
 * `wait_timeout` is set between 5-50 seconds (default: 10s), the call waits for
 * results up to the specified timeout; when set to `0s`, the call is
 * asynchronous and responds immediately with a statement ID. The
 * `on_wait_timeout` setting specifies what should happen when the timeout is
 * reached while the statement execution has not yet finished. This can be set to
 * either `CONTINUE`, to fallback to asynchronous mode, or it can be set to
 * `CANCEL`, which cancels the statement.
 *
 * In summary: - **Synchronous mode** (`wait_timeout=30s` and
 * `on_wait_timeout=CANCEL`): The call waits up to 30 seconds; if the statement
 * execution finishes within this time, the result data is returned directly in
 * the response. If the execution takes longer than 30 seconds, the execution is
 * canceled and the call returns with a `CANCELED` state. - **Asynchronous mode**
 * (`wait_timeout=0s` and `on_wait_timeout` is ignored): The call doesn't wait
 * for the statement to finish but returns directly with a statement ID. The
 * status of the statement execution can be polled by issuing
 * :method:statementexecution/getStatement with the statement ID. Once the
 * execution has succeeded, this call also returns the result and metadata in the
 * response. - **[Default] Hybrid mode** (`wait_timeout=10s` and
 * `on_wait_timeout=CONTINUE`): The call waits for up to 10 seconds; if the
 * statement execution finishes within this time, the result data is returned
 * directly in the response. If the execution takes longer than 10 seconds, a
 * statement ID is returned. The statement ID can be used to fetch status and
 * results in the same way as in the asynchronous mode.
 *
 * Depending on the size, the result can be split into multiple chunks. If the
 * statement execution is successful, the statement response contains a manifest
 * and the first chunk of the result. The manifest contains schema information
 * and provides metadata for each chunk in the result. Result chunks can be
 * retrieved by index with :method:statementexecution/getStatementResultChunkN
 * which may be called in any order and in parallel. For sequential fetching,
 * each chunk, apart from the last, also contains a `next_chunk_index` and
 * `next_chunk_internal_link` that point to the next chunk.
 *
 * A statement can be canceled with :method:statementexecution/cancelExecution.
 *
 * **Fetching result data: format and disposition**
 *
 * To specify the format of the result data, use the `format` field, which can be
 * set to one of the following options: `JSON_ARRAY` (JSON), `ARROW_STREAM`
 * ([Apache Arrow Columnar]), or `CSV`.
 *
 * There are two ways to receive statement results, controlled by the
 * `disposition` setting, which can be either `INLINE` or `EXTERNAL_LINKS`:
 *
 * - `INLINE`: In this mode, the result data is directly included in the
 * response. It's best suited for smaller results. This mode can only be used
 * with the `JSON_ARRAY` format.
 *
 * - `EXTERNAL_LINKS`: In this mode, the response provides links that can be used
 * to download the result data in chunks separately. This approach is ideal for
 * larger results and offers higher throughput. This mode can be used with all
 * the formats: `JSON_ARRAY`, `ARROW_STREAM`, and `CSV`.
 *
 * By default, the API uses `format=JSON_ARRAY` and `disposition=INLINE`.
 *
 * **Limits and limitations**
 *
 * Note: The byte limit for INLINE disposition is based on internal storage
 * metrics and will not exactly match the byte count of the actual payload.
 *
 * - Statements with `disposition=INLINE` are limited to 25 MiB and will fail
 * when this limit is exceeded. - Statements with `disposition=EXTERNAL_LINKS`
 * are limited to 100 GiB. Result sets larger than this limit will be truncated.
 * Truncation is indicated by the `truncated` field in the result manifest. - The
 * maximum query text size is 16 MiB. - Cancelation might silently fail. A
 * successful response from a cancel request indicates that the cancel request
 * was successfully received and sent to the processing engine. However, an
 * outstanding statement might have already completed execution when the cancel
 * request arrives. Polling for status until a terminal state is reached is a
 * reliable way to determine the final state. - Wait timeouts are approximate,
 * occur server-side, and cannot account for things such as caller delays and
 * network latency from caller to service. - To guarantee that the statement is
 * kept alive, you must poll at least once every 15 minutes. - The results are
 * only available for one hour after success; polling does not extend this. - The
 * SQL Execution API must be used for the entire lifecycle of the statement. For
 * example, you cannot use the Jobs API to execute the command, and then the SQL
 * Execution API to cancel it.
 *
 * [Apache Arrow Columnar]: https://arrow.apache.org/overview/
 * [Databricks SQL Statement Execution API tutorial]: https://docs.databricks.com/sql/api/sql-execution-tutorial.html
 */
class StatementExecutionService {
    constructor(client) {
        this.client = client;
    }
    async _cancelExecution(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/sql/statements/${request.statement_id}/cancel`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Requests that an executing statement be canceled. Callers must poll for
     * status to see the terminal state. Cancel response is empty; receiving
     * response indicates successful receipt.
     */
    async cancelExecution(request, context) {
        return await this._cancelExecution(request, context);
    }
    async _executeStatement(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("byte_limit")) {
            body["byte_limit"] = request.byte_limit;
        }
        if (request.hasOwnProperty("catalog")) {
            body["catalog"] = request.catalog;
        }
        if (request.hasOwnProperty("disposition")) {
            body["disposition"] = request.disposition;
        }
        if (request.hasOwnProperty("format")) {
            body["format"] = request.format;
        }
        if (request.hasOwnProperty("on_wait_timeout")) {
            body["on_wait_timeout"] = request.on_wait_timeout;
        }
        if (request.hasOwnProperty("parameters")) {
            body["parameters"] = request.parameters;
        }
        if (request.hasOwnProperty("row_limit")) {
            body["row_limit"] = request.row_limit;
        }
        if (request.hasOwnProperty("schema")) {
            body["schema"] = request.schema;
        }
        if (request.hasOwnProperty("statement")) {
            body["statement"] = request.statement;
        }
        if (request.hasOwnProperty("wait_timeout")) {
            body["wait_timeout"] = request.wait_timeout;
        }
        if (request.hasOwnProperty("warehouse_id")) {
            body["warehouse_id"] = request.warehouse_id;
        }
        const query = {};
        const path = "/api/2.0/sql/statements";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Execute a SQL statement and optionally await its results for a specified
     * time.
     *
     * **Use case: small result sets with INLINE + JSON_ARRAY**
     *
     * For flows that generate small and predictable result sets (<= 25 MiB),
     * `INLINE` responses of `JSON_ARRAY` result data are typically the simplest
     * way to execute and fetch result data.
     *
     * **Use case: large result sets with EXTERNAL_LINKS**
     *
     * Using `EXTERNAL_LINKS` to fetch result data allows you to fetch large
     * result sets efficiently. The main differences from using `INLINE`
     * disposition are that the result data is accessed with URLs, and that there
     * are 3 supported formats: `JSON_ARRAY`, `ARROW_STREAM` and `CSV` compared
     * to only `JSON_ARRAY` with `INLINE`.
     *
     * ** URLs**
     *
     * External links point to data stored within your workspace's internal
     * storage, in the form of a URL. The URLs are valid for only a short period,
     * <= 15 minutes. Alongside each `external_link` is an expiration field
     * indicating the time at which the URL is no longer valid. In
     * `EXTERNAL_LINKS` mode, chunks can be resolved and fetched multiple times
     * and in parallel.
     *
     * ----
     *
     * ### **Warning: Databricks strongly recommends that you protect the URLs
     * that are returned by the `EXTERNAL_LINKS` disposition.**
     *
     * When you use the `EXTERNAL_LINKS` disposition, a short-lived, URL is
     * generated, which can be used to download the results directly from . As a
     * short-lived is embedded in this URL, you should protect the URL.
     *
     * Because URLs are already generated with embedded temporary s, you must not
     * set an `Authorization` header in the download requests.
     *
     * The `EXTERNAL_LINKS` disposition can be disabled upon request by creating
     * a support case.
     *
     * See also [Security best practices].
     *
     * ----
     *
     * StatementResponse contains `statement_id` and `status`; other fields might
     * be absent or present depending on context. If the SQL warehouse fails to
     * execute the provided statement, a 200 response is returned with
     * `status.state` set to `FAILED` (in contrast to a failure when accepting
     * the request, which results in a non-200 response). Details of the error
     * can be found at `status.error` in case of execution failures.
     *
     * [Security best practices]: https://docs.databricks.com/sql/admin/sql-execution-tutorial.html#security-best-practices
     */
    async executeStatement(request, context) {
        return await this._executeStatement(request, context);
    }
    async _getStatement(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/sql/statements/${request.statement_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * This request can be used to poll for the statement's status.
     * StatementResponse contains `statement_id` and `status`; other fields might
     * be absent or present depending on context. When the `status.state` field
     * is `SUCCEEDED` it will also return the result manifest and the first chunk
     * of the result data. When the statement is in the terminal states
     * `CANCELED`, `CLOSED` or `FAILED`, it returns HTTP 200 with the state set.
     * After at least 12 hours in terminal state, the statement is removed from
     * the warehouse and further calls will receive an HTTP 404 response.
     *
     * **NOTE** This call currently might take up to 5 seconds to get the latest
     * status and result.
     */
    async getStatement(request, context) {
        return await this._getStatement(request, context);
    }
    async _getStatementResultChunkN(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/sql/statements/${request.statement_id}/result/chunks/${request.chunk_index}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * After the statement execution has `SUCCEEDED`, this request can be used to
     * fetch any chunk by index. Whereas the first chunk with `chunk_index=0` is
     * typically fetched with :method:statementexecution/executeStatement or
     * :method:statementexecution/getStatement, this request can be used to fetch
     * subsequent chunks. The response structure is identical to the nested
     * `result` element described in the :method:statementexecution/getStatement
     * request, and similarly includes the `next_chunk_index` and
     * `next_chunk_internal_link` fields for simple iteration through the result
     * set. Depending on `disposition`, the response returns chunks of data
     * either inline, or as links.
     */
    async getStatementResultChunkN(request, context) {
        return await this._getStatementResultChunkN(request, context);
    }
}
exports.StatementExecutionService = StatementExecutionService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StatementExecutionService.prototype, "_cancelExecution", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StatementExecutionService.prototype, "cancelExecution", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StatementExecutionService.prototype, "_executeStatement", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StatementExecutionService.prototype, "executeStatement", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StatementExecutionService.prototype, "_getStatement", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StatementExecutionService.prototype, "getStatement", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StatementExecutionService.prototype, "_getStatementResultChunkN", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StatementExecutionService.prototype, "getStatementResultChunkN", null);
class WarehousesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Warehouses", method, message);
    }
}
exports.WarehousesRetriableError = WarehousesRetriableError;
class WarehousesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Warehouses", method, message);
    }
}
exports.WarehousesError = WarehousesError;
/**
 * A SQL warehouse is a compute resource that lets you run SQL commands on data
 * objects within Databricks SQL. Compute resources are infrastructure resources
 * that provide processing capabilities in the cloud.
 */
class WarehousesService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("auto_stop_mins")) {
            body["auto_stop_mins"] = request.auto_stop_mins;
        }
        if (request.hasOwnProperty("channel")) {
            body["channel"] = request.channel;
        }
        if (request.hasOwnProperty("cluster_size")) {
            body["cluster_size"] = request.cluster_size;
        }
        if (request.hasOwnProperty("creator_name")) {
            body["creator_name"] = request.creator_name;
        }
        if (request.hasOwnProperty("enable_photon")) {
            body["enable_photon"] = request.enable_photon;
        }
        if (request.hasOwnProperty("enable_serverless_compute")) {
            body["enable_serverless_compute"] =
                request.enable_serverless_compute;
        }
        if (request.hasOwnProperty("instance_profile_arn")) {
            body["instance_profile_arn"] = request.instance_profile_arn;
        }
        if (request.hasOwnProperty("max_num_clusters")) {
            body["max_num_clusters"] = request.max_num_clusters;
        }
        if (request.hasOwnProperty("min_num_clusters")) {
            body["min_num_clusters"] = request.min_num_clusters;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("spot_instance_policy")) {
            body["spot_instance_policy"] = request.spot_instance_policy;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        if (request.hasOwnProperty("warehouse_type")) {
            body["warehouse_type"] = request.warehouse_type;
        }
        const query = {};
        const path = "/api/2.0/sql/warehouses";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new SQL warehouse.
     */
    async create(createWarehouseRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const createWarehouseResponse = await this._create(createWarehouseRequest, context);
        return (0, wait_1.asWaiter)(createWarehouseResponse, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        id: createWarehouseResponse.id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Warehouses.createAndWait: cancelled");
                        throw new WarehousesError("createAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state;
                    const statusMessage = pollResponse.health.summary;
                    switch (status) {
                        case "RUNNING": {
                            return pollResponse;
                        }
                        case "STOPPED":
                        case "DELETED": {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Warehouses.createAndWait: ${errorMessage}`);
                            throw new WarehousesError("createAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Warehouses.createAndWait: retrying: ${errorMessage}`);
                            throw new WarehousesRetriableError("createAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/sql/warehouses/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a SQL warehouse.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _edit(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("auto_stop_mins")) {
            body["auto_stop_mins"] = request.auto_stop_mins;
        }
        if (request.hasOwnProperty("channel")) {
            body["channel"] = request.channel;
        }
        if (request.hasOwnProperty("cluster_size")) {
            body["cluster_size"] = request.cluster_size;
        }
        if (request.hasOwnProperty("creator_name")) {
            body["creator_name"] = request.creator_name;
        }
        if (request.hasOwnProperty("enable_photon")) {
            body["enable_photon"] = request.enable_photon;
        }
        if (request.hasOwnProperty("enable_serverless_compute")) {
            body["enable_serverless_compute"] =
                request.enable_serverless_compute;
        }
        if (request.hasOwnProperty("instance_profile_arn")) {
            body["instance_profile_arn"] = request.instance_profile_arn;
        }
        if (request.hasOwnProperty("max_num_clusters")) {
            body["max_num_clusters"] = request.max_num_clusters;
        }
        if (request.hasOwnProperty("min_num_clusters")) {
            body["min_num_clusters"] = request.min_num_clusters;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("spot_instance_policy")) {
            body["spot_instance_policy"] = request.spot_instance_policy;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        if (request.hasOwnProperty("warehouse_type")) {
            body["warehouse_type"] = request.warehouse_type;
        }
        const query = {};
        const path = `/api/2.0/sql/warehouses/${request.id}/edit`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the configuration for a SQL warehouse.
     */
    async edit(editWarehouseRequest, context) {
        let cancellationToken = context?.cancellationToken;
        await this._edit(editWarehouseRequest, context);
        return (0, wait_1.asWaiter)(null, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        id: editWarehouseRequest.id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Warehouses.editAndWait: cancelled");
                        throw new WarehousesError("editAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state;
                    const statusMessage = pollResponse.health.summary;
                    switch (status) {
                        case "RUNNING": {
                            return pollResponse;
                        }
                        case "STOPPED":
                        case "DELETED": {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Warehouses.editAndWait: ${errorMessage}`);
                            throw new WarehousesError("editAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Warehouses.editAndWait: retrying: ${errorMessage}`);
                            throw new WarehousesRetriableError("editAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/sql/warehouses/${request.id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the information for a single SQL warehouse.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _getPermissionLevels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/warehouses/${request.warehouse_id}/permissionLevels`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(request, context) {
        return await this._getPermissionLevels(request, context);
    }
    async _getPermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/warehouses/${request.warehouse_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of a SQL warehouse. SQL warehouses can inherit
     * permissions from their root object.
     */
    async getPermissions(request, context) {
        return await this._getPermissions(request, context);
    }
    async _getWorkspaceWarehouseConfig(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/sql/config/warehouses";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the workspace level configuration that is shared by all SQL
     * warehouses in a workspace.
     */
    async getWorkspaceWarehouseConfig(context) {
        return await this._getWorkspaceWarehouseConfig(context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        if (request.hasOwnProperty("run_as_user_id")) {
            query["run_as_user_id"] =
                request.run_as_user_id;
        }
        const path = "/api/2.0/sql/warehouses";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists all SQL warehouses that a user has access to.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.warehouses || response.warehouses.length === 0) {
                break;
            }
            for (const v of response.warehouses) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _setPermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/warehouses/${request.warehouse_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their root object.
     */
    async setPermissions(request, context) {
        return await this._setPermissions(request, context);
    }
    async _setWorkspaceWarehouseConfig(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("channel")) {
            body["channel"] = request.channel;
        }
        if (request.hasOwnProperty("config_param")) {
            body["config_param"] = request.config_param;
        }
        if (request.hasOwnProperty("data_access_config")) {
            body["data_access_config"] = request.data_access_config;
        }
        if (request.hasOwnProperty("enable_serverless_compute")) {
            body["enable_serverless_compute"] =
                request.enable_serverless_compute;
        }
        if (request.hasOwnProperty("enabled_warehouse_types")) {
            body["enabled_warehouse_types"] = request.enabled_warehouse_types;
        }
        if (request.hasOwnProperty("global_param")) {
            body["global_param"] = request.global_param;
        }
        if (request.hasOwnProperty("google_service_account")) {
            body["google_service_account"] = request.google_service_account;
        }
        if (request.hasOwnProperty("instance_profile_arn")) {
            body["instance_profile_arn"] = request.instance_profile_arn;
        }
        if (request.hasOwnProperty("security_policy")) {
            body["security_policy"] = request.security_policy;
        }
        if (request.hasOwnProperty("sql_configuration_parameters")) {
            body["sql_configuration_parameters"] =
                request.sql_configuration_parameters;
        }
        const query = {};
        const path = "/api/2.0/sql/config/warehouses";
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets the workspace level configuration that is shared by all SQL
     * warehouses in a workspace.
     */
    async setWorkspaceWarehouseConfig(request, context) {
        return await this._setWorkspaceWarehouseConfig(request, context);
    }
    async _start(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/sql/warehouses/${request.id}/start`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Starts a SQL warehouse.
     */
    async start(startRequest, context) {
        let cancellationToken = context?.cancellationToken;
        await this._start(startRequest, context);
        return (0, wait_1.asWaiter)(null, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        id: startRequest.id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Warehouses.startAndWait: cancelled");
                        throw new WarehousesError("startAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state;
                    const statusMessage = pollResponse.health.summary;
                    switch (status) {
                        case "RUNNING": {
                            return pollResponse;
                        }
                        case "STOPPED":
                        case "DELETED": {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Warehouses.startAndWait: ${errorMessage}`);
                            throw new WarehousesError("startAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Warehouses.startAndWait: retrying: ${errorMessage}`);
                            throw new WarehousesRetriableError("startAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _stop(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/sql/warehouses/${request.id}/stop`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Stops a SQL warehouse.
     */
    async stop(stopRequest, context) {
        let cancellationToken = context?.cancellationToken;
        await this._stop(stopRequest, context);
        return (0, wait_1.asWaiter)(null, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        id: stopRequest.id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Warehouses.stopAndWait: cancelled");
                        throw new WarehousesError("stopAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state;
                    const statusMessage = pollResponse.health.summary;
                    switch (status) {
                        case "STOPPED": {
                            return pollResponse;
                        }
                        default: {
                            const errorMessage = `failed to reach STOPPED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Warehouses.stopAndWait: retrying: ${errorMessage}`);
                            throw new WarehousesRetriableError("stopAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/warehouses/${request.warehouse_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on a SQL warehouse. SQL warehouses can inherit
     * permissions from their root object.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
}
exports.WarehousesService = WarehousesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "_edit", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "edit", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "_getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "_getWorkspaceWarehouseConfig", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "getWorkspaceWarehouseConfig", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], WarehousesService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "_setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "_setWorkspaceWarehouseConfig", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "setWorkspaceWarehouseConfig", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "_start", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "start", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "_stop", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "stop", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WarehousesService.prototype, "updatePermissions", null);
