// @databricks/sdk-experimental/dist/apis/provisioning/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspacesService = exports.WorkspacesError = exports.WorkspacesRetriableError = exports.VpcEndpointsService = exports.VpcEndpointsError = exports.VpcEndpointsRetriableError = exports.StorageService = exports.StorageError = exports.StorageRetriableError = exports.PrivateAccessService = exports.PrivateAccessError = exports.PrivateAccessRetriableError = exports.NetworksService = exports.NetworksError = exports.NetworksRetriableError = exports.EncryptionKeysService = exports.EncryptionKeysError = exports.EncryptionKeysRetriableError = exports.CredentialsService = exports.CredentialsError = exports.CredentialsRetriableError = void 0;
/**
 * These APIs allow you to manage Credentials, Encryption Keys, Networks, Private Access, Storage, Vpc Endpoints, Workspaces, etc.
 */
const config_1 = require("../../config");
const HostType_1 = require("../../config/HostType");
const provisioning = __importStar(require("./model"));
const retries_1 = __importDefault(require("../../retries/retries"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
const wait_1 = require("../../wait");
class CredentialsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Credentials", method, message);
    }
}
exports.CredentialsRetriableError = CredentialsRetriableError;
class CredentialsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Credentials", method, message);
    }
}
exports.CredentialsError = CredentialsError;
/**
 * These APIs manage credential configurations for this workspace. Databricks
 * needs access to a cross-account service IAM role in your AWS account so that
 * Databricks can deploy clusters in the appropriate VPC for the new workspace. A
 * credential configuration encapsulates this role information, and its ID is
 * used when creating a new workspace.
 */
class CredentialsService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("aws_credentials")) {
            body["aws_credentials"] = request.aws_credentials;
        }
        if (request.hasOwnProperty("credentials_name")) {
            body["credentials_name"] = request.credentials_name;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/credentials`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a Databricks credential configuration that represents cloud
     * cross-account credentials for a specified account. Databricks uses this to
     * set up network infrastructure properly to host Databricks clusters. For
     * your AWS IAM role, you need to trust the External ID (the Databricks
     * Account API account ID) in the returned credential object, and configure
     * the required access policy.
     *
     * Save the response's `credentials_id` field, which is the ID for your new
     * credential configuration object.
     *
     * For information about how to create a new workspace with this API, see
     * [Create a new workspace using the Account API]
     *
     * [Create a new workspace using the Account API]: http://docs.databricks.com/administration-guide/account-api/new-workspace.html
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/credentials/${request.credentials_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a Databricks credential configuration object for an account, both
     * specified by ID. You cannot delete a credential that is associated with
     * any workspace.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/credentials/${request.credentials_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a Databricks credential configuration object for an account, both
     * specified by ID.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/credentials`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List Databricks credential configuration objects for an account, specified
     * by ID.
     */
    async list(context) {
        return await this._list(context);
    }
}
exports.CredentialsService = CredentialsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CredentialsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CredentialsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CredentialsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CredentialsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CredentialsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CredentialsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], CredentialsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], CredentialsService.prototype, "list", null);
class EncryptionKeysRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("EncryptionKeys", method, message);
    }
}
exports.EncryptionKeysRetriableError = EncryptionKeysRetriableError;
class EncryptionKeysError extends apiError_1.ApiError {
    constructor(method, message) {
        super("EncryptionKeys", method, message);
    }
}
exports.EncryptionKeysError = EncryptionKeysError;
/**
 * These APIs manage encryption key configurations for this workspace (optional).
 * A key configuration encapsulates the AWS KMS key information and some
 * information about how the key configuration can be used. There are two
 * possible uses for key configurations:
 *
 * * Managed services: A key configuration can be used to encrypt a workspace's
 * notebook and secret data in the control plane, as well as Databricks SQL
 * queries and query history. * Storage: A key configuration can be used to
 * encrypt a workspace's DBFS and EBS data in the data plane.
 *
 * In both of these cases, the key configuration's ID is used when creating a new
 * workspace. This Preview feature is available if your account is on the E2
 * version of the platform. Updating a running workspace with workspace storage
 * encryption requires that the workspace is on the E2 version of the platform.
 * If you have an older workspace, it might not be on the E2 version of the
 * platform. If you are not sure, contact your Databricks representative.
 */
class EncryptionKeysService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("aws_key_info")) {
            body["aws_key_info"] = request.aws_key_info;
        }
        if (request.hasOwnProperty("gcp_key_info")) {
            body["gcp_key_info"] = request.gcp_key_info;
        }
        if (request.hasOwnProperty("use_cases")) {
            body["use_cases"] = request.use_cases;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/customer-managed-keys`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a customer-managed key configuration object for an account,
     * specified by ID. This operation uploads a reference to a customer-managed
     * key to Databricks. If the key is assigned as a workspace's
     * customer-managed key for managed services, Databricks uses the key to
     * encrypt the workspaces notebooks and secrets in the control plane, in
     * addition to Databricks SQL queries and query history. If it is specified
     * as a workspace's customer-managed key for workspace storage, the key
     * encrypts the workspace's root S3 bucket (which contains the workspace's
     * root DBFS and system data) and, optionally, cluster EBS volume data.
     *
     * **Important**: Customer-managed keys are supported only for some
     * deployment types, subscription types, and AWS regions that currently
     * support creation of Databricks workspaces.
     *
     * This operation is available only if your account is on the E2 version of
     * the platform or on a select custom plan that allows multiple workspaces
     * per account.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/customer-managed-keys/${request.customer_managed_key_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a customer-managed key configuration object for an account. You
     * cannot delete a configuration that is associated with a running workspace.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/customer-managed-keys/${request.customer_managed_key_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a customer-managed key configuration object for an account, specified
     * by ID. This operation uploads a reference to a customer-managed key to
     * Databricks. If assigned as a workspace's customer-managed key for managed
     * services, Databricks uses the key to encrypt the workspaces notebooks and
     * secrets in the control plane, in addition to Databricks SQL queries and
     * query history. If it is specified as a workspace's customer-managed key
     * for storage, the key encrypts the workspace's root S3 bucket (which
     * contains the workspace's root DBFS and system data) and, optionally,
     * cluster EBS volume data.
     *
     * **Important**: Customer-managed keys are supported only for some
     * deployment types, subscription types, and AWS regions.
     *
     * This operation is available only if your account is on the E2 version of
     * the platform.",
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/customer-managed-keys`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists Databricks customer-managed key configurations for an account.
     */
    async list(context) {
        return await this._list(context);
    }
}
exports.EncryptionKeysService = EncryptionKeysService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], EncryptionKeysService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], EncryptionKeysService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], EncryptionKeysService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], EncryptionKeysService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], EncryptionKeysService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], EncryptionKeysService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], EncryptionKeysService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], EncryptionKeysService.prototype, "list", null);
class NetworksRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Networks", method, message);
    }
}
exports.NetworksRetriableError = NetworksRetriableError;
class NetworksError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Networks", method, message);
    }
}
exports.NetworksError = NetworksError;
/**
 * These APIs manage network configurations for customer-managed VPCs (optional).
 * Its ID is used when creating a new workspace if you use customer-managed VPCs.
 */
class NetworksService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("gcp_network_info")) {
            body["gcp_network_info"] = request.gcp_network_info;
        }
        if (request.hasOwnProperty("network_name")) {
            body["network_name"] = request.network_name;
        }
        if (request.hasOwnProperty("security_group_ids")) {
            body["security_group_ids"] = request.security_group_ids;
        }
        if (request.hasOwnProperty("subnet_ids")) {
            body["subnet_ids"] = request.subnet_ids;
        }
        if (request.hasOwnProperty("vpc_endpoints")) {
            body["vpc_endpoints"] = request.vpc_endpoints;
        }
        if (request.hasOwnProperty("vpc_id")) {
            body["vpc_id"] = request.vpc_id;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/networks`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a Databricks network configuration that represents an VPC and its
     * resources. The VPC will be used for new Databricks clusters. This requires
     * a pre-existing VPC and subnets.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/networks/${request.network_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a Databricks network configuration, which represents a cloud VPC
     * and its resources. You cannot delete a network that is associated with a
     * workspace.
     *
     * This operation is available only if your account is on the E2 version of
     * the platform.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/networks/${request.network_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a Databricks network configuration, which represents a cloud VPC and
     * its resources.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/networks`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists Databricks network configurations for an account.
     */
    async list(context) {
        return await this._list(context);
    }
}
exports.NetworksService = NetworksService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], NetworksService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], NetworksService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], NetworksService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], NetworksService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], NetworksService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], NetworksService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], NetworksService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], NetworksService.prototype, "list", null);
class PrivateAccessRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("PrivateAccess", method, message);
    }
}
exports.PrivateAccessRetriableError = PrivateAccessRetriableError;
class PrivateAccessError extends apiError_1.ApiError {
    constructor(method, message) {
        super("PrivateAccess", method, message);
    }
}
exports.PrivateAccessError = PrivateAccessError;
/**
 * These APIs manage private access settings for this account.
 */
class PrivateAccessService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("allowed_vpc_endpoint_ids")) {
            body["allowed_vpc_endpoint_ids"] = request.allowed_vpc_endpoint_ids;
        }
        if (request.hasOwnProperty("private_access_level")) {
            body["private_access_level"] = request.private_access_level;
        }
        if (request.hasOwnProperty("private_access_settings_name")) {
            body["private_access_settings_name"] =
                request.private_access_settings_name;
        }
        if (request.hasOwnProperty("public_access_enabled")) {
            body["public_access_enabled"] = request.public_access_enabled;
        }
        if (request.hasOwnProperty("region")) {
            body["region"] = request.region;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/private-access-settings`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a private access settings configuration, which represents network
     * access restrictions for workspace resources. Private access settings
     * configure whether workspaces can be accessed from the public internet or
     * only from private endpoints.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/private-access-settings/${request.private_access_settings_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a Databricks private access settings configuration, both specified
     * by ID.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/private-access-settings/${request.private_access_settings_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a Databricks private access settings configuration, both specified by
     * ID.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/private-access-settings`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists Databricks private access settings for an account.
     */
    async list(context) {
        return await this._list(context);
    }
    async _replace(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.customer_facing_private_access_settings;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/private-access-settings/${request.private_access_settings_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates an existing private access settings object, which specifies how
     * your workspace is accessed over AWS PrivateLink. To use AWS PrivateLink, a
     * workspace must have a private access settings object referenced by ID in
     * the workspace's private_access_settings_id property. This operation
     * completely overwrites your existing private access settings object
     * attached to your workspaces. All workspaces attached to the private access
     * settings are affected by any change. If public_access_enabled,
     * private_access_level, or allowed_vpc_endpoint_ids are updated, effects of
     * these changes might take several minutes to propagate to the workspace
     * API. You can share one private access settings object with multiple
     * workspaces in a single account. However, private access settings are
     * specific to AWS regions, so only workspaces in the same AWS region can use
     * a given private access settings object. Before configuring PrivateLink,
     * read the Databricks article about PrivateLink.
     */
    async replace(request, context) {
        return await this._replace(request, context);
    }
}
exports.PrivateAccessService = PrivateAccessService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PrivateAccessService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PrivateAccessService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PrivateAccessService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PrivateAccessService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PrivateAccessService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PrivateAccessService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], PrivateAccessService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], PrivateAccessService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PrivateAccessService.prototype, "_replace", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PrivateAccessService.prototype, "replace", null);
class StorageRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Storage", method, message);
    }
}
exports.StorageRetriableError = StorageRetriableError;
class StorageError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Storage", method, message);
    }
}
exports.StorageError = StorageError;
/**
 * These APIs manage storage configurations for this workspace. A root storage S3
 * bucket in your account is required to store objects like cluster logs,
 * notebook revisions, and job results. You can also use the root storage S3
 * bucket for storage of non-production DBFS data. A storage configuration
 * encapsulates this bucket information, and its ID is used when creating a new
 * workspace.
 */
class StorageService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("role_arn")) {
            body["role_arn"] = request.role_arn;
        }
        if (request.hasOwnProperty("root_bucket_info")) {
            body["root_bucket_info"] = request.root_bucket_info;
        }
        if (request.hasOwnProperty("storage_configuration_name")) {
            body["storage_configuration_name"] =
                request.storage_configuration_name;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/storage-configurations`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a Databricks storage configuration for an account.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/storage-configurations/${request.storage_configuration_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a Databricks storage configuration. You cannot delete a storage
     * configuration that is associated with any workspace.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/storage-configurations/${request.storage_configuration_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a Databricks storage configuration for an account, both specified by
     * ID.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/storage-configurations`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists Databricks storage configurations for an account, specified by ID.
     */
    async list(context) {
        return await this._list(context);
    }
}
exports.StorageService = StorageService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StorageService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StorageService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StorageService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StorageService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StorageService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], StorageService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], StorageService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], StorageService.prototype, "list", null);
class VpcEndpointsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("VpcEndpoints", method, message);
    }
}
exports.VpcEndpointsRetriableError = VpcEndpointsRetriableError;
class VpcEndpointsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("VpcEndpoints", method, message);
    }
}
exports.VpcEndpointsError = VpcEndpointsError;
/**
 * These APIs manage VPC endpoint configurations for this account.
 */
class VpcEndpointsService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("aws_vpc_endpoint_id")) {
            body["aws_vpc_endpoint_id"] = request.aws_vpc_endpoint_id;
        }
        if (request.hasOwnProperty("gcp_vpc_endpoint_info")) {
            body["gcp_vpc_endpoint_info"] = request.gcp_vpc_endpoint_info;
        }
        if (request.hasOwnProperty("region")) {
            body["region"] = request.region;
        }
        if (request.hasOwnProperty("vpc_endpoint_name")) {
            body["vpc_endpoint_name"] = request.vpc_endpoint_name;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/vpc-endpoints`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a VPC endpoint configuration, which represents a [VPC endpoint]
     * object in AWS used to communicate privately with Databricks over [AWS
     * PrivateLink].
     *
     * After you create the VPC endpoint configuration, the Databricks [endpoint
     * service] automatically accepts the VPC endpoint.
     *
     * Before configuring PrivateLink, read the [Databricks article about
     * PrivateLink].
     *
     * [AWS PrivateLink]: https://aws.amazon.com/privatelink
     * [Databricks article about PrivateLink]: https://docs.databricks.com/administration-guide/cloud-configurations/aws/privatelink.html
     * [VPC endpoint]: https://docs.aws.amazon.com/vpc/latest/privatelink/vpc-endpoints.html
     * [endpoint service]: https://docs.aws.amazon.com/vpc/latest/privatelink/privatelink-share-your-services.html
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/vpc-endpoints/${request.vpc_endpoint_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a Databricks VPC endpoint configuration. You cannot delete a VPC
     * endpoint configuration that is associated with any workspace.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/vpc-endpoints/${request.vpc_endpoint_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a VPC endpoint configuration, which represents a [VPC endpoint]
     * object in AWS used to communicate privately with Databricks over [AWS
     * PrivateLink].
     *
     * [AWS PrivateLink]: https://aws.amazon.com/privatelink
     * [VPC endpoint]: https://docs.aws.amazon.com/vpc/latest/privatelink/concepts.html
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/vpc-endpoints`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists Databricks VPC endpoint configurations for an account.
     */
    async list(context) {
        return await this._list(context);
    }
}
exports.VpcEndpointsService = VpcEndpointsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VpcEndpointsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VpcEndpointsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VpcEndpointsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VpcEndpointsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VpcEndpointsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], VpcEndpointsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], VpcEndpointsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], VpcEndpointsService.prototype, "list", null);
class WorkspacesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Workspaces", method, message);
    }
}
exports.WorkspacesRetriableError = WorkspacesRetriableError;
class WorkspacesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Workspaces", method, message);
    }
}
exports.WorkspacesError = WorkspacesError;
/**
 * These APIs manage workspaces for this account. A Databricks workspace is an
 * environment for accessing all of your Databricks assets. The workspace
 * organizes objects (notebooks, libraries, and experiments) into folders, and
 * provides access to data and computational resources such as clusters and jobs.
 *
 * These endpoints are available if your account is on the E2 version of the
 * platform or on a select custom plan that allows multiple workspaces per
 * account.
 */
class WorkspacesService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("aws_region")) {
            body["aws_region"] = request.aws_region;
        }
        if (request.hasOwnProperty("cloud")) {
            body["cloud"] = request.cloud;
        }
        if (request.hasOwnProperty("cloud_resource_container")) {
            body["cloud_resource_container"] = request.cloud_resource_container;
        }
        if (request.hasOwnProperty("compute_mode")) {
            body["compute_mode"] = request.compute_mode;
        }
        if (request.hasOwnProperty("credentials_id")) {
            body["credentials_id"] = request.credentials_id;
        }
        if (request.hasOwnProperty("custom_tags")) {
            body["custom_tags"] = request.custom_tags;
        }
        if (request.hasOwnProperty("deployment_name")) {
            body["deployment_name"] = request.deployment_name;
        }
        if (request.hasOwnProperty("gcp_managed_network_config")) {
            body["gcp_managed_network_config"] =
                request.gcp_managed_network_config;
        }
        if (request.hasOwnProperty("gke_config")) {
            body["gke_config"] = request.gke_config;
        }
        if (request.hasOwnProperty("location")) {
            body["location"] = request.location;
        }
        if (request.hasOwnProperty("managed_services_customer_managed_key_id")) {
            body["managed_services_customer_managed_key_id"] =
                request.managed_services_customer_managed_key_id;
        }
        if (request.hasOwnProperty("network_connectivity_config_id")) {
            body["network_connectivity_config_id"] =
                request.network_connectivity_config_id;
        }
        if (request.hasOwnProperty("network_id")) {
            body["network_id"] = request.network_id;
        }
        if (request.hasOwnProperty("pricing_tier")) {
            body["pricing_tier"] = request.pricing_tier;
        }
        if (request.hasOwnProperty("private_access_settings_id")) {
            body["private_access_settings_id"] =
                request.private_access_settings_id;
        }
        if (request.hasOwnProperty("storage_configuration_id")) {
            body["storage_configuration_id"] = request.storage_configuration_id;
        }
        if (request.hasOwnProperty("storage_customer_managed_key_id")) {
            body["storage_customer_managed_key_id"] =
                request.storage_customer_managed_key_id;
        }
        if (request.hasOwnProperty("workspace_name")) {
            body["workspace_name"] = request.workspace_name;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/workspaces`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new workspace using a credential configuration and a storage
     * configuration, an optional network configuration (if using a
     * customer-managed VPC), an optional managed services key configuration (if
     * using customer-managed keys for managed services), and an optional storage
     * key configuration (if using customer-managed keys for storage). The key
     * configurations used for managed services and storage encryption can be the
     * same or different.
     *
     * Important: This operation is asynchronous. A response with HTTP status
     * code 200 means the request has been accepted and is in progress, but does
     * not mean that the workspace deployed successfully and is running. The
     * initial workspace status is typically PROVISIONING. Use the workspace ID
     * (workspace_id) field in the response to identify the new workspace and
     * make repeated GET requests with the workspace ID and check its status. The
     * workspace becomes available when the status changes to RUNNING.
     *
     * You can share one customer-managed VPC with multiple workspaces in a
     * single account. It is not required to create a new VPC for each workspace.
     * However, you cannot reuse subnets or Security Groups between workspaces.
     * If you plan to share one VPC with multiple workspaces, make sure you size
     * your VPC and subnets accordingly. Because a Databricks Account API network
     * configuration encapsulates this information, you cannot reuse a Databricks
     * Account API network configuration across workspaces.
     *
     * For information about how to create a new workspace with this API
     * including error handling, see [Create a new workspace using the Account
     * API].
     *
     * Important: Customer-managed VPCs, PrivateLink, and customer-managed keys
     * are supported on a limited set of deployment and subscription types. If
     * you have questions about availability, contact your Databricks
     * representative.
     *
     * This operation is available only if your account is on the E2 version of
     * the platform or on a select custom plan that allows multiple workspaces
     * per account.
     *
     * [Create a new workspace using the Account API]: http://docs.databricks.com/administration-guide/account-api/new-workspace.html
     */
    async create(createWorkspaceRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const workspace = await this._create(createWorkspaceRequest, context);
        return (0, wait_1.asWaiter)(workspace, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        workspace_id: workspace.workspace_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Workspaces.createAndWait: cancelled");
                        throw new WorkspacesError("createAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.workspace_status;
                    const statusMessage = pollResponse.workspace_status_message;
                    switch (status) {
                        case "RUNNING": {
                            return pollResponse;
                        }
                        case "BANNED":
                        case "FAILED": {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Workspaces.createAndWait: ${errorMessage}`);
                            throw new WorkspacesError("createAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Workspaces.createAndWait: retrying: ${errorMessage}`);
                            throw new WorkspacesRetriableError("createAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/workspaces/${request.workspace_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a Databricks workspace, both specified by ID.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/workspaces/${request.workspace_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets information including status for a Databricks workspace, specified by
     * ID. In the response, the `workspace_status` field indicates the current
     * status. After initial workspace creation (which is asynchronous), make
     * repeated `GET` requests with the workspace ID and check its status. The
     * workspace becomes available when the status changes to `RUNNING`. For
     * information about how to create a new workspace with this API **including
     * error handling**, see [Create a new workspace using the Account API].
     *
     * [Create a new workspace using the Account API]: http://docs.databricks.com/administration-guide/account-api/new-workspace.html
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/workspaces`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists Databricks workspaces for an account.
     */
    async list(context) {
        return await this._list(context);
    }
    async _update(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.customer_facing_workspace;
        const query = {};
        if (request.hasOwnProperty("update_mask")) {
            query["update_mask"] =
                request.update_mask;
        }
        const path = `/api/2.0/accounts/${config.accountId}/workspaces/${request.workspace_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates a workspace.
     */
    async update(updateWorkspaceRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const workspace = await this._update(updateWorkspaceRequest, context);
        return (0, wait_1.asWaiter)(workspace, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        workspace_id: workspace.workspace_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Workspaces.updateAndWait: cancelled");
                        throw new WorkspacesError("updateAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.workspace_status;
                    const statusMessage = pollResponse.workspace_status_message;
                    switch (status) {
                        case "RUNNING": {
                            return pollResponse;
                        }
                        case "BANNED":
                        case "FAILED": {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Workspaces.updateAndWait: ${errorMessage}`);
                            throw new WorkspacesError("updateAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach RUNNING state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Workspaces.updateAndWait: retrying: ${errorMessage}`);
                            throw new WorkspacesRetriableError("updateAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
}
exports.WorkspacesService = WorkspacesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspacesService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspacesService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspacesService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspacesService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspacesService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspacesService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspacesService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspacesService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspacesService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspacesService.prototype, "update", null);
