// @databricks/sdk-experimental/dist/apis/sharing/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SharesService = exports.SharesError = exports.SharesRetriableError = exports.RecipientsService = exports.RecipientsError = exports.RecipientsRetriableError = exports.RecipientFederationPoliciesService = exports.RecipientFederationPoliciesError = exports.RecipientFederationPoliciesRetriableError = exports.RecipientActivationService = exports.RecipientActivationError = exports.RecipientActivationRetriableError = exports.ProvidersService = exports.ProvidersError = exports.ProvidersRetriableError = void 0;
const sharing = __importStar(require("./model"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
class ProvidersRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Providers", method, message);
    }
}
exports.ProvidersRetriableError = ProvidersRetriableError;
class ProvidersError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Providers", method, message);
    }
}
exports.ProvidersError = ProvidersError;
/**
 * A data provider is an object representing the organization in the real world
 * who shares the data. A provider contains shares which further contain the
 * shared data.
 */
class ProvidersService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("authentication_type")) {
            body["authentication_type"] = request.authentication_type;
        }
        if (request.hasOwnProperty("comment")) {
            body["comment"] = request.comment;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("recipient_profile_str")) {
            body["recipient_profile_str"] = request.recipient_profile_str;
        }
        const query = {};
        const path = "/api/2.1/unity-catalog/providers";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new authentication provider minimally based on a name and
     * authentication type. The caller must be an admin on the metastore.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.1/unity-catalog/providers/${request.name}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes an authentication provider, if the caller is a metastore admin or
     * is the owner of the provider.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.1/unity-catalog/providers/${request.name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a specific authentication provider. The caller must supply the name
     * of the provider, and must either be a metastore admin or the owner of the
     * provider.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("data_provider_global_metastore_id")) {
            query["data_provider_global_metastore_id"] =
                request.data_provider_global_metastore_id;
        }
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.1/unity-catalog/providers";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets an array of available authentication providers. The caller must
     * either be a metastore admin or the owner of the providers. Providers not
     * owned by the caller are not included in the response. There is no
     * guarantee of a specific ordering of the elements in the array.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.providers || response.providers.length === 0) {
                break;
            }
            for (const v of response.providers) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _listProviderShareAssets(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("function_max_results")) {
            query["function_max_results"] =
                request.function_max_results;
        }
        if (request.hasOwnProperty("notebook_max_results")) {
            query["notebook_max_results"] =
                request.notebook_max_results;
        }
        if (request.hasOwnProperty("table_max_results")) {
            query["table_max_results"] =
                request.table_max_results;
        }
        if (request.hasOwnProperty("volume_max_results")) {
            query["volume_max_results"] =
                request.volume_max_results;
        }
        const path = `/api/2.1/data-sharing/providers/${request.provider_name}/shares/${request.share_name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get arrays of assets associated with a specified provider's share. The
     * caller is the recipient of the share.
     */
    async listProviderShareAssets(request, context) {
        return await this._listProviderShareAssets(request, context);
    }
    async _listShares(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.1/unity-catalog/providers/${request.name}/shares`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets an array of a specified provider's shares within the metastore where:
     *
     * * the caller is a metastore admin, or * the caller is the owner.
     */
    async *listShares(request, context) {
        while (true) {
            const response = await this._listShares(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.shares || response.shares.length === 0) {
                break;
            }
            for (const v of response.shares) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("comment")) {
            body["comment"] = request.comment;
        }
        if (request.hasOwnProperty("new_name")) {
            body["new_name"] = request.new_name;
        }
        if (request.hasOwnProperty("owner")) {
            body["owner"] = request.owner;
        }
        if (request.hasOwnProperty("recipient_profile_str")) {
            body["recipient_profile_str"] = request.recipient_profile_str;
        }
        const query = {};
        const path = `/api/2.1/unity-catalog/providers/${request.name}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the information for an authentication provider, if the caller is a
     * metastore admin or is the owner of the provider. If the update changes the
     * provider name, the caller must be both a metastore admin and the owner of
     * the provider.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.ProvidersService = ProvidersService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ProvidersService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ProvidersService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ProvidersService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ProvidersService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ProvidersService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ProvidersService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ProvidersService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ProvidersService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ProvidersService.prototype, "_listProviderShareAssets", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ProvidersService.prototype, "listProviderShareAssets", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ProvidersService.prototype, "_listShares", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ProvidersService.prototype, "listShares", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ProvidersService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ProvidersService.prototype, "update", null);
class RecipientActivationRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("RecipientActivation", method, message);
    }
}
exports.RecipientActivationRetriableError = RecipientActivationRetriableError;
class RecipientActivationError extends apiError_1.ApiError {
    constructor(method, message) {
        super("RecipientActivation", method, message);
    }
}
exports.RecipientActivationError = RecipientActivationError;
/**
 * The Recipient Activation API is only applicable in the open sharing model
 * where the recipient object has the authentication type of `TOKEN`. The data
 * recipient follows the activation link shared by the data provider to download
 * the credential file that includes the access token. The recipient will then
 * use the credential file to establish a secure connection with the provider to
 * receive the shared data.
 *
 * Note that you can download the credential file only once. Recipients should
 * treat the downloaded credential as a secret and must not share it outside of
 * their organization.
 */
class RecipientActivationService {
    constructor(client) {
        this.client = client;
    }
    async _getActivationUrlInfo(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.1/unity-catalog/public/data_sharing_activation_info/${request.activation_url}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets an activation URL for a share.
     */
    async getActivationUrlInfo(request, context) {
        return await this._getActivationUrlInfo(request, context);
    }
    async _retrieveToken(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.1/unity-catalog/public/data_sharing_activation/${request.activation_url}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieve access token with an activation url. This is a public API without
     * any authentication.
     */
    async retrieveToken(request, context) {
        return await this._retrieveToken(request, context);
    }
}
exports.RecipientActivationService = RecipientActivationService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientActivationService.prototype, "_getActivationUrlInfo", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientActivationService.prototype, "getActivationUrlInfo", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientActivationService.prototype, "_retrieveToken", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientActivationService.prototype, "retrieveToken", null);
class RecipientFederationPoliciesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("RecipientFederationPolicies", method, message);
    }
}
exports.RecipientFederationPoliciesRetriableError = RecipientFederationPoliciesRetriableError;
class RecipientFederationPoliciesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("RecipientFederationPolicies", method, message);
    }
}
exports.RecipientFederationPoliciesError = RecipientFederationPoliciesError;
/**
 * The Recipient Federation Policies APIs are only applicable in the open sharing
 * model where the recipient object has the authentication type of
 * `OIDC_RECIPIENT`, enabling data sharing from Databricks to non-Databricks
 * recipients. OIDC Token Federation enables secure, secret-less authentication
 * for accessing Delta Sharing servers. Users and applications authenticate using
 * short-lived OIDC tokens issued by their own Identity Provider (IdP), such as
 * Azure Entra ID or Okta, without the need for managing static credentials or
 * client secrets. A federation policy defines how non-Databricks recipients
 * authenticate using OIDC tokens. It validates the OIDC claims in federated
 * tokens and is set at the recipient level. The caller must be the owner of the
 * recipient to create or manage a federation policy. Federation policies support
 * the following scenarios: - User-to-Machine (U2M) flow: A user accesses Delta
 * Shares using their own identity, such as connecting through PowerBI Delta
 * Sharing Client. - Machine-to-Machine (M2M) flow: An application accesses Delta
 * Shares using its own identity, typically for automation tasks like nightly
 * jobs through Python Delta Sharing Client. OIDC Token Federation enables
 * fine-grained access control, supports Multi-Factor Authentication (MFA), and
 * enhances security by minimizing the risk of credential leakage through the use
 * of short-lived, expiring tokens. It is designed for strong identity
 * governance, secure cross-platform data sharing, and reduced operational
 * overhead for credential management.
 *
 * For more information, see
 * https://www.databricks.com/blog/announcing-oidc-token-federation-enhanced-delta-sharing-security
 * and https://docs.databricks.com/en/delta-sharing/create-recipient-oidc-fed
 */
class RecipientFederationPoliciesService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.policy;
        const query = {};
        const path = `/api/2.0/data-sharing/recipients/${request.recipient_name}/federation-policies`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a federation policy for an OIDC_FEDERATION recipient for sharing
     * data from Databricks to non-Databricks recipients. The caller must be the
     * owner of the recipient. When sharing data from Databricks to
     * non-Databricks clients, you can define a federation policy to authenticate
     * non-Databricks recipients. The federation policy validates OIDC claims in
     * federated tokens and is defined at the recipient level. This enables
     * secretless sharing clients to authenticate using OIDC tokens.
     *
     * Supported scenarios for federation policies: 1. **User-to-Machine (U2M)
     * flow** (e.g., PowerBI): A user accesses a resource using their own
     * identity. 2. **Machine-to-Machine (M2M) flow** (e.g., OAuth App): An OAuth
     * App accesses a resource using its own identity, typically for tasks like
     * running nightly jobs.
     *
     * For an overview, refer to: - Blog post: Overview of feature:
     * https://www.databricks.com/blog/announcing-oidc-token-federation-enhanced-delta-sharing-security
     *
     * For detailed configuration guides based on your use case: - Creating a
     * Federation Policy as a provider:
     * https://docs.databricks.com/en/delta-sharing/create-recipient-oidc-fed -
     * Configuration and usage for Machine-to-Machine (M2M) applications (e.g.,
     * Python Delta Sharing Client):
     * https://docs.databricks.com/aws/en/delta-sharing/sharing-over-oidc-m2m -
     * Configuration and usage for User-to-Machine (U2M) applications (e.g.,
     * PowerBI):
     * https://docs.databricks.com/aws/en/delta-sharing/sharing-over-oidc-u2m
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/data-sharing/recipients/${request.recipient_name}/federation-policies/${request.name}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes an existing federation policy for an OIDC_FEDERATION recipient.
     * The caller must be the owner of the recipient.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _getFederationPolicy(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/data-sharing/recipients/${request.recipient_name}/federation-policies/${request.name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Reads an existing federation policy for an OIDC_FEDERATION recipient for
     * sharing data from Databricks to non-Databricks recipients. The caller must
     * have read access to the recipient.
     */
    async getFederationPolicy(request, context) {
        return await this._getFederationPolicy(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/data-sharing/recipients/${request.recipient_name}/federation-policies`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists federation policies for an OIDC_FEDERATION recipient for sharing
     * data from Databricks to non-Databricks recipients. The caller must have
     * read access to the recipient.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.policies || response.policies.length === 0) {
                break;
            }
            for (const v of response.policies) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
}
exports.RecipientFederationPoliciesService = RecipientFederationPoliciesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientFederationPoliciesService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientFederationPoliciesService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientFederationPoliciesService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientFederationPoliciesService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientFederationPoliciesService.prototype, "_getFederationPolicy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientFederationPoliciesService.prototype, "getFederationPolicy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientFederationPoliciesService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], RecipientFederationPoliciesService.prototype, "list", null);
class RecipientsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Recipients", method, message);
    }
}
exports.RecipientsRetriableError = RecipientsRetriableError;
class RecipientsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Recipients", method, message);
    }
}
exports.RecipientsError = RecipientsError;
/**
 * A recipient is an object you create using :method:recipients/create to
 * represent an organization which you want to allow access shares. The way how
 * sharing works differs depending on whether or not your recipient has access to
 * a Databricks workspace that is enabled for Unity Catalog:
 *
 * - For recipients with access to a Databricks workspace that is enabled for
 * Unity Catalog, you can create a recipient object along with a unique sharing
 * identifier you get from the recipient. The sharing identifier is the key
 * identifier that enables the secure connection. This sharing mode is called
 * **Databricks-to-Databricks sharing**.
 *
 * - For recipients without access to a Databricks workspace that is enabled for
 * Unity Catalog, when you create a recipient object, Databricks generates an
 * activation link you can send to the recipient. The recipient follows the
 * activation link to download the credential file, and then uses the credential
 * file to establish a secure connection to receive the shared data. This sharing
 * mode is called **open sharing**.
 */
class RecipientsService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("authentication_type")) {
            body["authentication_type"] = request.authentication_type;
        }
        if (request.hasOwnProperty("comment")) {
            body["comment"] = request.comment;
        }
        if (request.hasOwnProperty("data_recipient_global_metastore_id")) {
            body["data_recipient_global_metastore_id"] =
                request.data_recipient_global_metastore_id;
        }
        if (request.hasOwnProperty("expiration_time")) {
            body["expiration_time"] = request.expiration_time;
        }
        if (request.hasOwnProperty("ip_access_list")) {
            body["ip_access_list"] = request.ip_access_list;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("owner")) {
            body["owner"] = request.owner;
        }
        if (request.hasOwnProperty("properties_kvpairs")) {
            body["properties_kvpairs"] = request.properties_kvpairs;
        }
        if (request.hasOwnProperty("sharing_code")) {
            body["sharing_code"] = request.sharing_code;
        }
        const query = {};
        const path = "/api/2.1/unity-catalog/recipients";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new recipient with the delta sharing authentication type in the
     * metastore. The caller must be a metastore admin or have the
     * **CREATE_RECIPIENT** privilege on the metastore.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.1/unity-catalog/recipients/${request.name}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes the specified recipient from the metastore. The caller must be the
     * owner of the recipient.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.1/unity-catalog/recipients/${request.name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a share recipient from the metastore if:
     *
     * * the caller is the owner of the share recipient, or: * is a metastore
     * admin
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("data_recipient_global_metastore_id")) {
            query["data_recipient_global_metastore_id"] =
                request.data_recipient_global_metastore_id;
        }
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.1/unity-catalog/recipients";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets an array of all share recipients within the current metastore where:
     *
     * * the caller is a metastore admin, or * the caller is the owner. There is
     * no guarantee of a specific ordering of the elements in the array.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.recipients || response.recipients.length === 0) {
                break;
            }
            for (const v of response.recipients) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _rotateToken(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("existing_token_expire_in_seconds")) {
            body["existing_token_expire_in_seconds"] =
                request.existing_token_expire_in_seconds;
        }
        const query = {};
        const path = `/api/2.1/unity-catalog/recipients/${request.name}/rotate-token`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Refreshes the specified recipient's delta sharing authentication token
     * with the provided token info. The caller must be the owner of the
     * recipient.
     */
    async rotateToken(request, context) {
        return await this._rotateToken(request, context);
    }
    async _sharePermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.1/unity-catalog/recipients/${request.name}/share-permissions`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the share permissions for the specified Recipient. The caller must
     * have the USE_RECIPIENT privilege on the metastore or be the owner of the
     * Recipient.
     */
    async sharePermissions(request, context) {
        return await this._sharePermissions(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("comment")) {
            body["comment"] = request.comment;
        }
        if (request.hasOwnProperty("expiration_time")) {
            body["expiration_time"] = request.expiration_time;
        }
        if (request.hasOwnProperty("ip_access_list")) {
            body["ip_access_list"] = request.ip_access_list;
        }
        if (request.hasOwnProperty("new_name")) {
            body["new_name"] = request.new_name;
        }
        if (request.hasOwnProperty("owner")) {
            body["owner"] = request.owner;
        }
        if (request.hasOwnProperty("properties_kvpairs")) {
            body["properties_kvpairs"] = request.properties_kvpairs;
        }
        const query = {};
        const path = `/api/2.1/unity-catalog/recipients/${request.name}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates an existing recipient in the metastore. The caller must be a
     * metastore admin or the owner of the recipient. If the recipient name will
     * be updated, the user must be both a metastore admin and the owner of the
     * recipient.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.RecipientsService = RecipientsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], RecipientsService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientsService.prototype, "_rotateToken", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientsService.prototype, "rotateToken", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientsService.prototype, "_sharePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientsService.prototype, "sharePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientsService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], RecipientsService.prototype, "update", null);
class SharesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Shares", method, message);
    }
}
exports.SharesRetriableError = SharesRetriableError;
class SharesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Shares", method, message);
    }
}
exports.SharesError = SharesError;
/**
 * A share is a container instantiated with :method:shares/create. Once created
 * you can iteratively register a collection of existing data assets defined
 * within the metastore using :method:shares/update. You can register data assets
 * under their original name, qualified by their original schema, or provide
 * alternate exposed names.
 */
class SharesService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("comment")) {
            body["comment"] = request.comment;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("storage_root")) {
            body["storage_root"] = request.storage_root;
        }
        const query = {};
        const path = "/api/2.1/unity-catalog/shares";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new share for data objects. Data objects can be added after
     * creation with **update**. The caller must be a metastore admin or have the
     * **CREATE_SHARE** privilege on the metastore.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.1/unity-catalog/shares/${request.name}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a data object share from the metastore. The caller must be an
     * owner of the share.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("include_shared_data")) {
            query["include_shared_data"] =
                request.include_shared_data;
        }
        const path = `/api/2.1/unity-catalog/shares/${request.name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a data object share from the metastore. The caller must have the
     * USE_SHARE privilege on the metastore or be the owner of the share.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _listShares(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.1/unity-catalog/shares";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets an array of data object shares from the metastore. If the caller has
     * the USE_SHARE privilege on the metastore, all shares are returned.
     * Otherwise, only shares owned by the caller are returned. There is no
     * guarantee of a specific ordering of the elements in the array.
     */
    async *listShares(request, context) {
        while (true) {
            const response = await this._listShares(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.shares || response.shares.length === 0) {
                break;
            }
            for (const v of response.shares) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _sharePermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.1/unity-catalog/shares/${request.name}/permissions`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions for a data share from the metastore. The caller must
     * have the USE_SHARE privilege on the metastore or be the owner of the
     * share.
     */
    async sharePermissions(request, context) {
        return await this._sharePermissions(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("comment")) {
            body["comment"] = request.comment;
        }
        if (request.hasOwnProperty("new_name")) {
            body["new_name"] = request.new_name;
        }
        if (request.hasOwnProperty("owner")) {
            body["owner"] = request.owner;
        }
        if (request.hasOwnProperty("storage_root")) {
            body["storage_root"] = request.storage_root;
        }
        if (request.hasOwnProperty("updates")) {
            body["updates"] = request.updates;
        }
        const query = {};
        const path = `/api/2.1/unity-catalog/shares/${request.name}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the share with the changes and data objects in the request. The
     * caller must be the owner of the share or a metastore admin.
     *
     * When the caller is a metastore admin, only the __owner__ field can be
     * updated.
     *
     * In the case the share name is changed, **updateShare** requires that the
     * caller is the owner of the share and has the CREATE_SHARE privilege.
     *
     * If there are notebook files in the share, the __storage_root__ field
     * cannot be updated.
     *
     * For each table that is added through this method, the share owner must
     * also have **SELECT** privilege on the table. This privilege must be
     * maintained indefinitely for recipients to be able to access the table.
     * Typically, you should use a group as the share owner.
     *
     * Table removals through **update** do not require additional privileges.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("changes")) {
            body["changes"] = request.changes;
        }
        if (request.hasOwnProperty("omit_permissions_list")) {
            body["omit_permissions_list"] = request.omit_permissions_list;
        }
        const query = {};
        const path = `/api/2.1/unity-catalog/shares/${request.name}/permissions`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions for a data share in the metastore. The caller must
     * have both the USE_SHARE and SET_SHARE_PERMISSION privileges on the
     * metastore, or be the owner of the share.
     *
     * For new recipient grants, the user must also be the owner of the
     * recipients. recipient revocations do not require additional privileges.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
}
exports.SharesService = SharesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SharesService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SharesService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SharesService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SharesService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SharesService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SharesService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SharesService.prototype, "_listShares", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], SharesService.prototype, "listShares", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SharesService.prototype, "_sharePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SharesService.prototype, "sharePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SharesService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SharesService.prototype, "update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SharesService.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], SharesService.prototype, "updatePermissions", null);
