// @databricks/appkit/dist/type-generator/serving/cache.js
import { createLogger } from "../../logging/logger.js";
import crypto from "node:crypto";
import fs from "node:fs/promises";
import path from "node:path";

//#region src/type-generator/serving/cache.ts
const logger = createLogger("type-generator:serving:cache");
const CACHE_VERSION = "1";
const CACHE_FILE = ".appkit-serving-types-cache.json";
const CACHE_DIR = path.join(process.cwd(), "node_modules", ".databricks", "appkit");
function hashSchema(schemaJson) {
	return crypto.createHash("sha256").update(schemaJson).digest("hex");
}
async function loadServingCache() {
	const cachePath = path.join(CACHE_DIR, CACHE_FILE);
	try {
		await fs.mkdir(CACHE_DIR, { recursive: true });
		const raw = await fs.readFile(cachePath, "utf8");
		const cache = JSON.parse(raw);
		if (cache.version === CACHE_VERSION) return cache;
		logger.debug("Cache version mismatch, starting fresh");
	} catch (err) {
		if (err.code !== "ENOENT") logger.warn("Cache file is corrupted, flushing cache completely.");
	}
	return {
		version: CACHE_VERSION,
		endpoints: {}
	};
}
async function saveServingCache(cache) {
	const cachePath = path.join(CACHE_DIR, CACHE_FILE);
	await fs.mkdir(CACHE_DIR, { recursive: true });
	await fs.writeFile(cachePath, JSON.stringify(cache, null, 2), "utf8");
}

//#endregion
export { CACHE_VERSION, hashSchema, loadServingCache, saveServingCache };
