// @databricks/appkit/dist/type-generator/serving/converter.js
//#region src/type-generator/serving/converter.ts
/**
* Converts an OpenAPI schema to a TypeScript type string.
*/
function schemaToTypeString(schema, indent = 0) {
	const pad = "  ".repeat(indent);
	if (schema.oneOf) return schema.oneOf.map((s) => schemaToTypeString(s, indent)).join(" | ");
	if (schema.enum) return schema.enum.map((v) => JSON.stringify(v)).join(" | ");
	switch (schema.type) {
		case "string": return "string";
		case "integer":
		case "number": return "number";
		case "boolean": return "boolean";
		case "array": {
			if (!schema.items) return "unknown[]";
			const itemType = schemaToTypeString(schema.items, indent);
			if (itemType.includes(" | ") && !itemType.startsWith("{")) return `(${itemType})[]`;
			return `${itemType}[]`;
		}
		case "object": {
			if (!schema.properties) return "Record<string, unknown>";
			const required = new Set(schema.required ?? []);
			return `{\n${Object.entries(schema.properties).map(([key, prop]) => {
				const optional = !required.has(key) ? "?" : "";
				const nullable = prop.nullable ? " | null" : "";
				const typeStr = schemaToTypeString(prop, indent + 1);
				return `${pad}  ${prop.format && (prop.type === "number" || prop.type === "integer") ? `/** @openapi ${prop.format}${prop.nullable ? ", nullable" : ""} */\n${pad}  ` : prop.nullable && prop.type === "integer" ? `/** @openapi integer, nullable */\n${pad}  ` : ""}${key}${optional}: ${typeStr}${nullable};`;
			}).join("\n")}\n${pad}}`;
		}
		default: return "unknown";
	}
}
/**
* Extracts the top-level property keys from the request schema.
* Strips the `stream` property (plugin-controlled).
*/
function extractRequestKeys(operation) {
	const schema = operation.requestBody?.content?.["application/json"]?.schema;
	if (!schema?.properties) return [];
	return Object.keys(schema.properties).filter((k) => k !== "stream");
}
/**
* Extracts and converts the request schema from an OpenAPI path operation.
* Strips the `stream` property from the request type.
*/
function convertRequestSchema(operation) {
	const schema = operation.requestBody?.content?.["application/json"]?.schema;
	if (!schema || !schema.properties) return "Record<string, unknown>";
	const { stream: _stream, ...filteredProps } = schema.properties;
	const filteredRequired = (schema.required ?? []).filter((r) => r !== "stream");
	return schemaToTypeString({
		...schema,
		properties: filteredProps,
		required: filteredRequired.length > 0 ? filteredRequired : void 0
	});
}
/**
* Extracts and converts the response schema from an OpenAPI path operation.
*/
function convertResponseSchema(operation) {
	const schema = (operation.responses?.["200"])?.content?.["application/json"]?.schema;
	if (!schema) return "unknown";
	return schemaToTypeString(schema);
}
/**
* Derives a streaming chunk type from the response schema.
* Returns null if the response doesn't follow OpenAI-compatible format.
*
* OpenAI-compatible heuristic: response has `choices` array where items
* have a `message` object property.
*/
function deriveChunkType(operation) {
	const schema = (operation.responses?.["200"])?.content?.["application/json"]?.schema;
	if (!schema?.properties) return null;
	const choicesProp = schema.properties.choices;
	if (!choicesProp || choicesProp.type !== "array" || !choicesProp.items) return null;
	const choiceItemProps = choicesProp.items.properties;
	if (!choiceItemProps?.message) return null;
	const messageSchema = choiceItemProps.message;
	const chunkProperties = {};
	for (const [key, prop] of Object.entries(schema.properties)) {
		if (key === "usage") continue;
		if (key === "choices") {
			const chunkChoiceProps = {};
			for (const [ck, cp] of Object.entries(choiceItemProps)) if (ck === "message") chunkChoiceProps.delta = { ...messageSchema };
			else if (ck === "finish_reason") chunkChoiceProps[ck] = {
				...cp,
				nullable: true
			};
			else chunkChoiceProps[ck] = cp;
			chunkProperties[key] = {
				type: "array",
				items: {
					type: "object",
					properties: chunkChoiceProps
				}
			};
		} else chunkProperties[key] = prop;
	}
	return schemaToTypeString({
		type: "object",
		properties: chunkProperties
	});
}

//#endregion
export { convertRequestSchema, convertResponseSchema, deriveChunkType, extractRequestKeys };
