// @databricks/appkit/dist/type-generator/serving/generator.js
import { createLogger } from "../../logging/logger.js";
import { migrateProjectConfig, removeOldGeneratedTypes, resolveProjectRoot } from "../migration.js";
import { CACHE_VERSION, hashSchema, loadServingCache, saveServingCache } from "./cache.js";
import { convertRequestSchema, convertResponseSchema, deriveChunkType, extractRequestKeys } from "./converter.js";
import { fetchOpenApiSchema } from "./fetcher.js";
import { extractServingEndpoints, findServerFile } from "./server-file-extractor.js";
import { WorkspaceClient } from "@databricks/sdk-experimental";
import pc from "picocolors";
import fs from "node:fs/promises";
import path from "node:path";

//#region src/type-generator/serving/generator.ts
const logger = createLogger("type-generator:serving");
const GENERIC_REQUEST = "Record<string, unknown>";
const GENERIC_RESPONSE = "unknown";
const GENERIC_CHUNK = "unknown";
/**
* Generates TypeScript type declarations for serving endpoints
* by fetching their OpenAPI schemas and converting to TypeScript.
*
* Endpoint discovery order (when `endpoints` is not provided):
* 1. AST extraction from server file (server/index.ts or server/server.ts)
* 2. DATABRICKS_SERVING_ENDPOINT_NAME env var (single default endpoint)
*/
async function generateServingTypes(options) {
	const { outFile, noCache } = options;
	const projectRoot = resolveProjectRoot(outFile);
	const endpoints = options.endpoints ?? resolveEndpointsFromServerFile() ?? resolveDefaultEndpoints();
	if (Object.keys(endpoints).length === 0) {
		logger.debug("No serving endpoints configured, skipping type generation");
		return;
	}
	const startTime = performance.now();
	const cache = noCache ? {
		version: CACHE_VERSION,
		endpoints: {}
	} : await loadServingCache();
	let client;
	let updated = false;
	const registryEntries = [];
	const logEntries = [];
	for (const [alias, config] of Object.entries(endpoints)) {
		client ??= new WorkspaceClient({});
		const result = await processEndpoint(alias, config, client, cache);
		if (result.cacheUpdated) updated = true;
		registryEntries.push(result.entry);
		logEntries.push(result.log);
	}
	printLogTable(logEntries, startTime);
	const output = generateTypeDeclarations(registryEntries);
	await fs.mkdir(path.dirname(outFile), { recursive: true });
	await fs.writeFile(outFile, output, "utf-8");
	await removeOldGeneratedTypes(projectRoot, "appKitServingTypes.d.ts");
	await migrateProjectConfig(projectRoot);
	if (registryEntries.length === 0) logger.debug("Wrote empty serving types to %s (no endpoints resolved)", outFile);
	else logger.debug("Wrote serving types to %s", outFile);
	if (updated) await saveServingCache(cache);
}
function genericEntry(alias) {
	return buildRegistryEntry(alias, GENERIC_REQUEST, GENERIC_RESPONSE, GENERIC_CHUNK);
}
async function processEndpoint(alias, config, client, cache) {
	const endpointName = process.env[config.env];
	if (!endpointName) return {
		entry: genericEntry(alias),
		log: {
			alias,
			status: "MISS",
			error: `env ${config.env} not set`
		},
		cacheUpdated: false
	};
	const result = await fetchOpenApiSchema(client, endpointName, config.servedModel);
	if (!result) return {
		entry: genericEntry(alias),
		log: {
			alias,
			status: "MISS",
			error: "schema fetch failed"
		},
		cacheUpdated: false
	};
	const { spec, pathKey } = result;
	const hash = hashSchema(JSON.stringify(spec));
	const cached = cache.endpoints[alias];
	if (cached && cached.hash === hash) return {
		entry: buildRegistryEntry(alias, cached.requestType, cached.responseType, cached.chunkType),
		log: {
			alias,
			status: "HIT"
		},
		cacheUpdated: false
	};
	const operation = spec.paths[pathKey]?.post;
	if (!operation) return {
		entry: genericEntry(alias),
		log: {
			alias,
			status: "MISS",
			error: "no POST operation"
		},
		cacheUpdated: false
	};
	try {
		const requestType = convertRequestSchema(operation);
		const responseType = convertResponseSchema(operation);
		const chunkType = deriveChunkType(operation);
		const requestKeys = extractRequestKeys(operation);
		cache.endpoints[alias] = {
			hash,
			requestType,
			responseType,
			chunkType,
			requestKeys
		};
		return {
			entry: buildRegistryEntry(alias, requestType, responseType, chunkType),
			log: {
				alias,
				status: "MISS"
			},
			cacheUpdated: true
		};
	} catch (convErr) {
		logger.warn("Schema conversion failed for '%s': %s", alias, convErr.message);
		return {
			entry: genericEntry(alias),
			log: {
				alias,
				status: "MISS",
				error: "schema conversion failed"
			},
			cacheUpdated: false
		};
	}
}
function printLogTable(logEntries, startTime) {
	if (logEntries.length === 0) return;
	const maxNameLen = Math.max(...logEntries.map((e) => e.alias.length));
	const separator = pc.dim("─".repeat(50));
	console.log("");
	console.log(`  ${pc.bold("Typegen Serving")} ${pc.dim(`(${logEntries.length})`)}`);
	console.log(`  ${separator}`);
	for (const entry of logEntries) {
		const tag = entry.status === "HIT" ? `cache ${pc.bold(pc.green("HIT  "))}` : `cache ${pc.bold(pc.yellow("MISS "))}`;
		const rawName = entry.alias.padEnd(maxNameLen);
		const reason = entry.error ? `  ${pc.dim(entry.error)}` : "";
		console.log(`  ${tag}  ${rawName}${reason}`);
	}
	const elapsed = ((performance.now() - startTime) / 1e3).toFixed(2);
	const newCount = logEntries.filter((e) => e.status === "MISS").length;
	const cacheCount = logEntries.filter((e) => e.status === "HIT").length;
	console.log(`  ${separator}`);
	console.log(`  ${newCount} new, ${cacheCount} from cache. ${pc.dim(`${elapsed}s`)}`);
	console.log("");
}
function resolveEndpointsFromServerFile() {
	try {
		const serverFile = findServerFile(process.cwd());
		if (!serverFile) return void 0;
		return extractServingEndpoints(serverFile) ?? void 0;
	} catch (error) {
		logger.debug("Failed to extract endpoints from server file: %s", error.message);
		return;
	}
}
function resolveDefaultEndpoints() {
	if (process.env.DATABRICKS_SERVING_ENDPOINT_NAME) return { default: { env: "DATABRICKS_SERVING_ENDPOINT_NAME" } };
	return {};
}
function buildRegistryEntry(alias, requestType, responseType, chunkType) {
	const indent = "      ";
	const chunkEntry = chunkType ? chunkType : "unknown";
	return `    ${alias}: {
${indent}request: ${indentType(requestType, indent)};
${indent}response: ${indentType(responseType, indent)};
${indent}chunk: ${indentType(chunkEntry, indent)};
    };`;
}
function indentType(typeStr, baseIndent) {
	if (!typeStr.includes("\n")) return typeStr;
	return typeStr.split("\n").map((line, i) => i === 0 ? line : `${baseIndent}${line}`).join("\n");
}
function generateTypeDeclarations(entries) {
	return `// Auto-generated by AppKit - DO NOT EDIT
// Generated from serving endpoint OpenAPI schemas
import "@databricks/appkit";
import "@databricks/appkit-ui/react";

declare module "@databricks/appkit" {
  interface ServingEndpointRegistry {
${entries.join("\n")}
  }
}

declare module "@databricks/appkit-ui/react" {
  interface ServingEndpointRegistry {
${entries.join("\n")}
  }
}
`;
}

//#endregion
export { generateServingTypes };
