// @databricks/appkit/dist/type-generator/serving/server-file-extractor.js
import { createLogger } from "../../logging/logger.js";
import path from "node:path";
import fs from "node:fs";
import { Lang, parse } from "@ast-grep/napi";

//#region src/type-generator/serving/server-file-extractor.ts
const logger = createLogger("type-generator:serving:extractor");
/**
* Candidate paths for the server entry file, relative to the project root.
* Checked in order; the first that exists is used.
* Same convention as plugin sync (sync.ts SERVER_FILE_CANDIDATES).
*/
const SERVER_FILE_CANDIDATES = ["server/index.ts", "server/server.ts"];
/**
* Find the server entry file by checking candidate paths in order.
*
* @param basePath - Project root directory to search from
* @returns Absolute path to the server file, or null if none found
*/
function findServerFile(basePath) {
	for (const candidate of SERVER_FILE_CANDIDATES) {
		const fullPath = path.join(basePath, candidate);
		if (fs.existsSync(fullPath)) return fullPath;
	}
	return null;
}
/**
* Extract serving endpoint config from a server file by AST-parsing it.
* Looks for `serving({ endpoints: { alias: { env: "..." }, ... } })` calls
* and extracts the endpoint alias names and their environment variable mappings.
*
* @param serverFilePath - Absolute path to the server entry file
* @returns Extracted endpoint config, or null if not found or not extractable
*/
function extractServingEndpoints(serverFilePath) {
	let content;
	try {
		content = fs.readFileSync(serverFilePath, "utf-8");
	} catch {
		logger.debug("Could not read server file: %s", serverFilePath);
		return null;
	}
	const servingCall = findServingCall(parse(serverFilePath.endsWith(".tsx") ? Lang.Tsx : Lang.TypeScript, content).root());
	if (!servingCall) {
		logger.debug("No serving() call found in %s", serverFilePath);
		return null;
	}
	const args = servingCall.field("arguments");
	if (!args) return null;
	const configArg = args.children().find((child) => child.kind() === "object");
	if (!configArg) return null;
	const endpointsPair = findProperty(configArg, "endpoints");
	if (!endpointsPair) return null;
	const endpointsValue = getPropertyValue(endpointsPair);
	if (!endpointsValue || endpointsValue.kind() !== "object") {
		logger.debug("serving() endpoints is not an inline object literal in %s. Pass endpoints explicitly via appKitServingTypesPlugin({ endpoints }) in vite.config.ts.", serverFilePath);
		return null;
	}
	const endpoints = {};
	const pairs = endpointsValue.children().filter((child) => child.kind() === "pair");
	for (const pair of pairs) {
		const entry = extractEndpointEntry(pair);
		if (entry) endpoints[entry.alias] = entry.config;
	}
	if (Object.keys(endpoints).length === 0) return null;
	logger.debug("Extracted %d endpoint(s) from %s: %s", Object.keys(endpoints).length, serverFilePath, Object.keys(endpoints).join(", "));
	return endpoints;
}
/**
* Find the serving() call expression in the AST.
* Looks for call expressions where the callee identifier is "serving".
*/
function findServingCall(root) {
	const callExpressions = root.findAll({ rule: { kind: "call_expression" } });
	for (const call of callExpressions) {
		const callee = call.children()[0];
		if (callee?.kind() === "identifier" && callee.text() === "serving") return call;
	}
	return null;
}
/**
* Find a property (pair node) with the given key name in an object expression.
*/
function findProperty(objectNode, propertyName) {
	const pairs = objectNode.children().filter((child) => child.kind() === "pair");
	for (const pair of pairs) {
		const key = pair.children()[0];
		if (!key) continue;
		if ((key.kind() === "property_identifier" ? key.text() : key.kind() === "string" ? key.text().replace(/^['"]|['"]$/g, "") : null) === propertyName) return pair;
	}
	return null;
}
/**
* Get the value node from a pair (property: value).
* The value is typically the last meaningful child after the colon.
*/
function getPropertyValue(pairNode) {
	const children = pairNode.children();
	return children.length >= 3 ? children[children.length - 1] : null;
}
/**
* Extract a single endpoint entry from a pair node like:
* `demo: { env: "DATABRICKS_SERVING_ENDPOINT_NAME", servedModel: "my-model" }`
*/
function extractEndpointEntry(pair) {
	const children = pair.children();
	if (children.length < 3) return null;
	const keyNode = children[0];
	const alias = keyNode.kind() === "property_identifier" ? keyNode.text() : keyNode.kind() === "string" ? keyNode.text().replace(/^['"]|['"]$/g, "") : null;
	if (!alias) return null;
	const valueNode = children[children.length - 1];
	if (valueNode.kind() !== "object") return null;
	const envPair = findProperty(valueNode, "env");
	if (!envPair) return null;
	const envValue = getPropertyValue(envPair);
	if (!envValue || envValue.kind() !== "string") return null;
	const config = { env: envValue.text().replace(/^['"]|['"]$/g, "") };
	const servedModelPair = findProperty(valueNode, "servedModel");
	if (servedModelPair) {
		const servedModelValue = getPropertyValue(servedModelPair);
		if (servedModelValue?.kind() === "string") config.servedModel = servedModelValue.text().replace(/^['"]|['"]$/g, "");
	}
	return {
		alias,
		config
	};
}

//#endregion
export { extractServingEndpoints, findServerFile };
