// @databricks/appkit/dist/type-generator/serving/vite-plugin.js
import { createLogger } from "../../logging/logger.js";
import { extractServingEndpoints, findServerFile } from "./server-file-extractor.js";
import { SERVING_TYPES_FILE, TYPES_DIR, generateServingTypes } from "../index.js";
import path from "node:path";

//#region src/type-generator/serving/vite-plugin.ts
const logger = createLogger("type-generator:serving:vite-plugin");
/**
* Vite plugin to generate TypeScript types for AppKit serving endpoints.
* Fetches OpenAPI schemas from Databricks and generates a .d.ts with
* ServingEndpointRegistry module augmentation.
*
* Endpoint discovery order:
* 1. Explicit `endpoints` option (override)
* 2. AST extraction from server file (server/index.ts or server/server.ts)
* 3. DATABRICKS_SERVING_ENDPOINT_NAME env var (single default endpoint)
*/
function appKitServingTypesPlugin(options) {
	let outFile;
	let projectRoot;
	async function generate() {
		try {
			let endpoints = options?.endpoints;
			if (!endpoints) {
				const serverFile = findServerFile(projectRoot);
				if (serverFile) endpoints = extractServingEndpoints(serverFile) ?? void 0;
			}
			await generateServingTypes({
				outFile,
				endpoints,
				noCache: false
			});
		} catch (error) {
			if (process.env.NODE_ENV === "production") throw error;
			logger.error("Error generating serving types: %O", error);
		}
	}
	return {
		name: "appkit-serving-types",
		apply() {
			if (options?.endpoints && Object.keys(options.endpoints).length > 0) return true;
			if (process.env.DATABRICKS_SERVING_ENDPOINT_NAME) return true;
			if (findServerFile(process.cwd())) return true;
			if (findServerFile(path.resolve(process.cwd(), ".."))) return true;
			logger.debug("No serving endpoints configured. Skipping type generation.");
			return false;
		},
		configResolved(config) {
			projectRoot = path.resolve(config.root, "..");
			outFile = path.resolve(projectRoot, options?.outFile ?? `shared/${TYPES_DIR}/${SERVING_TYPES_FILE}`);
		},
		async buildStart() {
			await generate();
		}
	};
}

//#endregion
export { appKitServingTypesPlugin };
