// @databricks/appkit-ui/dist/react/charts/base.js
import { normalizeChartData, normalizeHeatmapData } from "./normalize.js";
import { buildCartesianOption, buildHeatmapOption, buildHorizontalBarOption, buildPieOption, buildRadarOption } from "./options.js";
import { useThemeColors } from "./theme.js";
import ReactECharts from "echarts-for-react";
import { useCallback, useMemo, useRef } from "react";
import { jsx } from "react/jsx-runtime";

//#region src/react/charts/base.tsx
/**
* Determines the appropriate color palette for a chart type.
* - Heatmaps use sequential (low → high intensity)
* - All other charts use categorical (distinct categories)
*/
function getDefaultPalette(chartType) {
	switch (chartType) {
		case "heatmap": return "sequential";
		default: return "categorical";
	}
}
/**
* Base chart component that handles both Arrow and JSON data.
* Renders using ECharts for consistent output across both formats.
*/
function BaseChart({ data, chartType, xKey, yKey, orientation, height = 300, title, showLegend = true, colorPalette, colors: customColors, showSymbol = false, smooth = true, stacked = false, symbolSize = 8, showArea = true, innerRadius = 0, showLabels = true, labelPosition = "outside", yAxisKey, min, max, options: customOptions, className }) {
	const themeColors = useThemeColors(colorPalette ?? getDefaultPalette(chartType));
	const colors = customColors ?? themeColors;
	const echartsInstanceRef = useRef(null);
	const chartRefCallback = useCallback((node) => {
		if (echartsInstanceRef.current && !echartsInstanceRef.current.isDisposed()) echartsInstanceRef.current.dispose();
		if (node) echartsInstanceRef.current = node.getEchartsInstance();
		else {
			if (echartsInstanceRef.current && !echartsInstanceRef.current.isDisposed()) echartsInstanceRef.current.dispose();
			echartsInstanceRef.current = null;
		}
	}, []);
	const normalized = useMemo(() => chartType === "heatmap" ? normalizeHeatmapData(data, xKey, yAxisKey, yKey) : normalizeChartData(data, xKey, yKey, orientation), [
		data,
		xKey,
		yKey,
		yAxisKey,
		orientation,
		chartType
	]);
	const option = useMemo(() => {
		const { xData, yFields, xField, chartType: detectedChartType } = normalized;
		if (xData.length === 0) return null;
		const isHeatmap = chartType === "heatmap";
		const baseCtx = {
			xData,
			yDataMap: "yDataMap" in normalized ? normalized.yDataMap : {},
			yFields,
			colors,
			title,
			showLegend,
			xField
		};
		const isPie = chartType === "pie" || chartType === "donut";
		const isRadar = chartType === "radar";
		const isHorizontal = !isPie && !isRadar && !isHeatmap && (orientation === "horizontal" || detectedChartType === "categorical" && !orientation && chartType === "bar");
		const isTimeSeries = detectedChartType === "timeseries" && !isHorizontal && !isRadar && !isHeatmap;
		let opt;
		if (isHeatmap && "yAxisData" in normalized && "heatmapData" in normalized) {
			const heatmapNorm = normalized;
			opt = buildHeatmapOption({
				...baseCtx,
				yAxisData: heatmapNorm.yAxisData,
				heatmapData: heatmapNorm.heatmapData,
				min: min ?? heatmapNorm.min,
				max: max ?? heatmapNorm.max,
				showLabels
			});
		} else if (isRadar) opt = buildRadarOption(baseCtx, showArea);
		else if (isPie) opt = buildPieOption(baseCtx, chartType, innerRadius, showLabels, labelPosition);
		else if (isHorizontal) opt = buildHorizontalBarOption(baseCtx, stacked);
		else opt = buildCartesianOption({
			...baseCtx,
			chartType,
			isTimeSeries,
			stacked,
			smooth,
			showSymbol,
			symbolSize
		});
		return customOptions ? {
			...opt,
			...customOptions
		} : opt;
	}, [
		normalized,
		colors,
		title,
		showLegend,
		chartType,
		orientation,
		innerRadius,
		showLabels,
		labelPosition,
		stacked,
		smooth,
		showSymbol,
		symbolSize,
		showArea,
		min,
		max,
		customOptions
	]);
	if (!option) return /* @__PURE__ */ jsx("div", {
		className: "flex items-center justify-center h-full text-muted-foreground",
		children: "No data"
	});
	return /* @__PURE__ */ jsx(ReactECharts, {
		ref: chartRefCallback,
		option,
		style: { height },
		className,
		opts: { renderer: "canvas" },
		notMerge: false,
		lazyUpdate: true
	});
}

//#endregion
export { BaseChart };
