// @databricks/appkit-ui/dist/react/charts/constants.js
import { DATE_FIELD_PATTERNS, METADATA_DATE_PATTERNS, NAME_FIELD_PATTERNS } from "../../js/constants.js";

//#region src/react/charts/constants.ts
/** CSS variable names for categorical chart colors (distinct categories) */
const CHART_COLOR_VARS_CATEGORICAL = [
	"--chart-cat-1",
	"--chart-cat-2",
	"--chart-cat-3",
	"--chart-cat-4",
	"--chart-cat-5",
	"--chart-cat-6",
	"--chart-cat-7",
	"--chart-cat-8"
];
/** CSS variable names for sequential chart colors (low → high) */
const CHART_COLOR_VARS_SEQUENTIAL = [
	"--chart-seq-1",
	"--chart-seq-2",
	"--chart-seq-3",
	"--chart-seq-4",
	"--chart-seq-5",
	"--chart-seq-6",
	"--chart-seq-7",
	"--chart-seq-8"
];
/** CSS variable names for diverging chart colors (negative ↔ positive) */
const CHART_COLOR_VARS_DIVERGING = [
	"--chart-div-1",
	"--chart-div-2",
	"--chart-div-3",
	"--chart-div-4",
	"--chart-div-5",
	"--chart-div-6",
	"--chart-div-7",
	"--chart-div-8"
];
/** Legacy: CSS variable names for chart colors (aliases to categorical) */
const CHART_COLOR_VARS = [
	"--chart-1",
	"--chart-2",
	"--chart-3",
	"--chart-4",
	"--chart-5",
	"--chart-6",
	"--chart-7",
	"--chart-8"
];
/** Fallback categorical colors */
const FALLBACK_COLORS_CATEGORICAL = [
	"hsla(221, 83%, 53%, 1)",
	"hsla(160, 60%, 45%, 1)",
	"hsla(291, 47%, 51%, 1)",
	"hsla(35, 92%, 55%, 1)",
	"hsla(349, 72%, 52%, 1)",
	"hsla(189, 75%, 42%, 1)",
	"hsla(271, 55%, 60%, 1)",
	"hsla(142, 55%, 45%, 1)"
];
/** Fallback sequential colors (light → dark blue) */
const FALLBACK_COLORS_SEQUENTIAL = [
	"hsla(221, 70%, 94%, 1)",
	"hsla(221, 72%, 85%, 1)",
	"hsla(221, 74%, 74%, 1)",
	"hsla(221, 76%, 63%, 1)",
	"hsla(221, 78%, 52%, 1)",
	"hsla(221, 80%, 42%, 1)",
	"hsla(221, 82%, 32%, 1)",
	"hsla(221, 84%, 24%, 1)"
];
/** Fallback diverging colors (blue → red) */
const FALLBACK_COLORS_DIVERGING = [
	"hsla(221, 80%, 35%, 1)",
	"hsla(221, 70%, 50%, 1)",
	"hsla(221, 55%, 65%, 1)",
	"hsla(221, 35%, 82%, 1)",
	"hsla(10, 35%, 82%, 1)",
	"hsla(10, 60%, 65%, 1)",
	"hsla(10, 72%, 50%, 1)",
	"hsla(10, 80%, 40%, 1)"
];

//#endregion
export { CHART_COLOR_VARS, CHART_COLOR_VARS_CATEGORICAL, CHART_COLOR_VARS_DIVERGING, CHART_COLOR_VARS_SEQUENTIAL, FALLBACK_COLORS_CATEGORICAL, FALLBACK_COLORS_DIVERGING, FALLBACK_COLORS_SEQUENTIAL };
