// @databricks/appkit-ui/dist/react/charts/create-chart.js
import { BaseChart } from "./base.js";
import { ChartWrapper } from "./wrapper.js";
import { jsx } from "react/jsx-runtime";

//#region src/react/charts/create-chart.tsx
/**
* Factory function to create chart components.
* Eliminates boilerplate by generating components with the same pattern.
*
* @param chartType - The ECharts chart type
* @param displayName - Component display name for React DevTools
* @returns A typed chart component
*
* @example
* ```tsx
* export const BarChart = createChart<BarChartProps>("bar", "BarChart");
* export const LineChart = createChart<LineChartProps>("line", "LineChart");
* ```
*/
function createChart(chartType, displayName) {
	const Component = (props) => {
		const { queryKey, parameters, format, transformer, data, height = 300, className, ariaLabel, testId, ...chartProps } = props;
		return /* @__PURE__ */ jsx(ChartWrapper, {
			...data !== void 0 ? {
				data,
				height,
				className,
				ariaLabel,
				testId
			} : {
				queryKey,
				parameters,
				format,
				transformer,
				height,
				className,
				ariaLabel,
				testId: testId ?? `${chartType}-chart-${queryKey}`
			},
			children: (chartData) => /* @__PURE__ */ jsx(BaseChart, {
				data: chartData,
				chartType,
				height,
				className,
				...chartProps
			})
		});
	};
	Component.displayName = displayName;
	return Component;
}

//#endregion
export { createChart };
