// @databricks/appkit-ui/dist/react/charts/normalize.js
import { DATE_FIELD_PATTERNS, NAME_FIELD_PATTERNS } from "../../js/constants.js";
import { ArrowClient } from "../../js/arrow/arrow-client.js";
import "../../js/index.js";
import { isArrowTable } from "./types.js";
import { sortNumericAscending, sortTimeSeriesAscending, toChartArray } from "./utils.js";

//#region src/react/charts/normalize.ts
/**
* Checks if a value looks like an ISO date string
*/
function isDateString(value) {
	if (typeof value !== "string") return false;
	return /^\d{4}-\d{2}-\d{2}(T|$)/.test(value);
}
/**
* Checks if a value is numeric (number or numeric string)
*/
function isNumericValue(value) {
	if (typeof value === "number") return true;
	if (typeof value === "bigint") return true;
	if (typeof value === "string") {
		const trimmed = value.trim();
		if (trimmed === "" || isDateString(trimmed)) return false;
		const parsed = Number(trimmed);
		return !Number.isNaN(parsed) && Number.isFinite(parsed);
	}
	return false;
}
/**
* Checks if a value looks like a category/label (non-numeric string)
*/
function isCategoryValue(value) {
	if (typeof value !== "string") return false;
	const trimmed = value.trim();
	if (trimmed === "") return false;
	if (/^\d{4}-\d{2}-\d{2}/.test(trimmed)) return false;
	const parsed = Number(trimmed);
	return Number.isNaN(parsed) || !Number.isFinite(parsed);
}
/**
* Detects fields from JSON data for charting
*/
function detectFieldsFromJson(data, orientation) {
	if (!data || data.length === 0) return {
		xField: "x",
		yFields: ["y"],
		chartType: "categorical"
	};
	const firstRow = data[0];
	const keys = Object.keys(firstRow);
	const dateFields = keys.filter((key) => {
		const value = firstRow[key];
		const keyMatchesDatePattern = DATE_FIELD_PATTERNS.some((p) => key.toLowerCase().includes(p));
		const valueIsDateString = isDateString(value);
		return keyMatchesDatePattern || valueIsDateString;
	});
	let nameFields = keys.filter((key) => {
		const value = firstRow[key];
		return isCategoryValue(value) && !isDateString(value) && NAME_FIELD_PATTERNS.some((p) => key.toLowerCase().includes(p));
	});
	if (nameFields.length === 0) nameFields = keys.filter((key) => {
		const value = firstRow[key];
		return isCategoryValue(value) && !isDateString(value) && !dateFields.includes(key) && !key.toLowerCase().endsWith("_id");
	});
	const numericFields = keys.filter((key) => {
		const value = firstRow[key];
		return isNumericValue(value) && !dateFields.includes(key);
	});
	if (orientation === "horizontal" || nameFields.length > 0 && dateFields.length === 0) {
		const xField = nameFields[0] || dateFields[0] || keys[0];
		return {
			xField,
			yFields: numericFields.length > 0 ? numericFields : keys.filter((k) => k !== xField),
			chartType: "categorical"
		};
	}
	const xField = dateFields[0] || nameFields[0] || keys[0];
	return {
		xField,
		yFields: numericFields.length > 0 ? numericFields : keys.filter((k) => k !== xField),
		chartType: dateFields.length > 0 ? "timeseries" : "categorical"
	};
}
/**
* Converts a JSON value to a chart-compatible value.
*/
function jsonValueToChartValue(value, isYValue, isDateField) {
	if (value === null || value === void 0) return isYValue ? 0 : "";
	if (typeof value === "number") return value;
	if (typeof value === "bigint") return Number(value);
	if (typeof value === "string") {
		if (isDateField && isDateString(value)) {
			const timestamp = new Date(value).getTime();
			if (!Number.isNaN(timestamp)) return timestamp;
		}
		if (isYValue) {
			const trimmed = value.trim();
			const parsed = Number(trimmed);
			if (!Number.isNaN(parsed) && Number.isFinite(parsed)) return parsed;
		}
		return value;
	}
	return String(value);
}
/**
* Extracts chart data from JSON array
*/
function extractFromJson(data, xField, yFields) {
	const xData = [];
	const yDataMap = {};
	for (const field of yFields) yDataMap[field] = [];
	const xIsDateField = data.length > 0 && isDateString(data[0][xField]);
	for (const row of data) {
		xData.push(jsonValueToChartValue(row[xField], false, xIsDateField));
		for (const field of yFields) yDataMap[field].push(jsonValueToChartValue(row[field], true, false));
	}
	return {
		xData,
		yDataMap
	};
}
/**
* Normalizes chart data from either Arrow or JSON format.
* Converts BigInt and Date values to chart-compatible types.
*/
function normalizeChartData(data, xKey, yKey, orientation) {
	if (isArrowTable(data)) {
		const table = data;
		const detected = ArrowClient.detectFieldsFromArrow(table, orientation);
		const resolvedXKey = xKey ?? detected.xField;
		const resolvedYKeys = yKey ? Array.isArray(yKey) ? yKey : [yKey] : detected.yFields;
		const { xData: rawXData, yDataMap: rawYDataMap } = ArrowClient.extractChartData(table, resolvedXKey, resolvedYKeys);
		let xData = toChartArray(rawXData);
		let yDataMap = {};
		for (const key of resolvedYKeys) yDataMap[key] = toChartArray(rawYDataMap[key] ?? []);
		if (detected.chartType === "timeseries") ({xData, yDataMap} = sortTimeSeriesAscending(xData, yDataMap, resolvedYKeys));
		else if (xData.length > 0 && xData.every(isNumericValue)) ({xData, yDataMap} = sortNumericAscending(xData, yDataMap, resolvedYKeys));
		return {
			xData,
			yDataMap,
			xField: resolvedXKey,
			yFields: resolvedYKeys,
			chartType: detected.chartType
		};
	}
	const jsonData = data;
	const detected = detectFieldsFromJson(jsonData, orientation);
	const resolvedXKey = xKey ?? detected.xField;
	const resolvedYKeys = yKey ? Array.isArray(yKey) ? yKey : [yKey] : detected.yFields;
	const { xData: rawXData, yDataMap: rawYDataMap } = extractFromJson(jsonData, resolvedXKey, resolvedYKeys);
	let xData = toChartArray(rawXData);
	let yDataMap = {};
	for (const key of resolvedYKeys) yDataMap[key] = toChartArray(rawYDataMap[key] ?? []);
	if (detected.chartType === "timeseries") ({xData, yDataMap} = sortTimeSeriesAscending(xData, yDataMap, resolvedYKeys));
	else if (xData.length > 0 && xData.every(isNumericValue)) ({xData, yDataMap} = sortNumericAscending(xData, yDataMap, resolvedYKeys));
	return {
		xData,
		yDataMap,
		xField: resolvedXKey,
		yFields: resolvedYKeys,
		chartType: detected.chartType
	};
}
/**
* Normalizes data specifically for heatmap charts.
* Expects data in format: `{ xKey: string, yAxisKey: string, valueKey: number }`
*
* @param data - Raw data (Arrow Table or JSON array)
* @param xKey - Field key for X-axis (columns)
* @param yAxisKey - Field key for Y-axis (rows)
* @param valueKey - Field key for the cell values
*/
function normalizeHeatmapData(data, xKey, yAxisKey, valueKey) {
	const jsonData = isArrowTable(data) ? extractJsonFromArrow(data) : data;
	if (jsonData.length === 0) return {
		xData: [],
		xField: xKey ?? "x",
		yFields: [],
		chartType: "categorical",
		yAxisData: [],
		heatmapData: [],
		min: 0,
		max: 0
	};
	const keys = Object.keys(jsonData[0]);
	const resolvedXKey = xKey ?? keys[0];
	const resolvedYAxisKey = yAxisKey ?? keys[1];
	const resolvedValueKey = valueKey ? Array.isArray(valueKey) ? valueKey[0] : valueKey : keys[2];
	const xSet = /* @__PURE__ */ new Set();
	const ySet = /* @__PURE__ */ new Set();
	for (const row of jsonData) {
		const xVal = jsonValueToChartValue(row[resolvedXKey], false, false);
		const yVal = jsonValueToChartValue(row[resolvedYAxisKey], false, false);
		xSet.add(xVal);
		ySet.add(yVal);
	}
	const xData = Array.from(xSet);
	const yAxisData = Array.from(ySet);
	const xIndexMap = /* @__PURE__ */ new Map();
	const yIndexMap = /* @__PURE__ */ new Map();
	xData.forEach((v, i) => {
		xIndexMap.set(v, i);
	});
	yAxisData.forEach((v, i) => {
		yIndexMap.set(v, i);
	});
	const heatmapData = [];
	let min = Number.POSITIVE_INFINITY;
	let max = Number.NEGATIVE_INFINITY;
	for (const row of jsonData) {
		const xVal = jsonValueToChartValue(row[resolvedXKey], false, false);
		const yVal = jsonValueToChartValue(row[resolvedYAxisKey], false, false);
		const value = jsonValueToChartValue(row[resolvedValueKey], true, false);
		const xIdx = xIndexMap.get(xVal);
		const yIdx = yIndexMap.get(yVal);
		const numValue = typeof value === "number" ? value : 0;
		if (xIdx !== void 0 && yIdx !== void 0) {
			heatmapData.push([
				xIdx,
				yIdx,
				numValue
			]);
			min = Math.min(min, numValue);
			max = Math.max(max, numValue);
		}
	}
	if (heatmapData.length === 0) {
		min = 0;
		max = 0;
	}
	return {
		xData,
		xField: resolvedXKey,
		yFields: [resolvedValueKey],
		chartType: "categorical",
		yAxisData,
		heatmapData,
		min,
		max
	};
}
/**
* Helper to extract JSON array from Arrow table for heatmap processing.
*/
function extractJsonFromArrow(table) {
	const result = [];
	const fields = table.schema.fields.map((f) => f.name);
	for (let i = 0; i < table.numRows; i++) {
		const row = {};
		for (const field of fields) row[field] = table.getChild(field)?.get(i);
		result.push(row);
	}
	return result;
}

//#endregion
export { normalizeChartData, normalizeHeatmapData };
