// @databricks/appkit-ui/dist/react/charts/theme.js
import { CHART_COLOR_VARS_CATEGORICAL, CHART_COLOR_VARS_DIVERGING, CHART_COLOR_VARS_SEQUENTIAL, FALLBACK_COLORS_CATEGORICAL, FALLBACK_COLORS_DIVERGING, FALLBACK_COLORS_SEQUENTIAL } from "./constants.js";
import { useEffect, useState } from "react";

//#region src/react/charts/theme.ts
const PALETTE_CONFIG = {
	categorical: {
		vars: CHART_COLOR_VARS_CATEGORICAL,
		fallback: FALLBACK_COLORS_CATEGORICAL
	},
	sequential: {
		vars: CHART_COLOR_VARS_SEQUENTIAL,
		fallback: FALLBACK_COLORS_SEQUENTIAL
	},
	diverging: {
		vars: CHART_COLOR_VARS_DIVERGING,
		fallback: FALLBACK_COLORS_DIVERGING
	}
};
/**
* Cache for computed theme colors to avoid repeated CSS variable lookups.
* Cache is cleared when theme change events fire (MutationObserver/matchMedia).
*/
const colorCache = /* @__PURE__ */ new Map();
/**
* Clears the theme color cache.
* Called when theme change events fire, or for testing when mocks change.
* @internal
*/
function resetThemeColorCache() {
	colorCache.clear();
}
/**
* Gets theme colors with module-level caching.
* Avoids repeated CSS variable lookups for the same palette within a theme.
*/
function getThemeColors(palette = "categorical") {
	const config = PALETTE_CONFIG[palette];
	if (typeof window === "undefined") return config.fallback;
	const cached = colorCache.get(palette);
	if (cached) return cached;
	const styles = getComputedStyle(document.documentElement);
	const colors = [];
	for (const varName of config.vars) {
		const value = styles.getPropertyValue(varName).trim();
		if (value) colors.push(value);
	}
	const result = colors.length > 0 ? colors : config.fallback;
	colorCache.set(palette, result);
	return result;
}
/**
* Hook to get theme colors with automatic updates on theme change.
* Re-resolves CSS variables when color scheme or theme attributes change.
*
* @param palette - Color palette type: "categorical" (default), "sequential", or "diverging"
*/
function useThemeColors(palette = "categorical") {
	const [colors, setColors] = useState(() => typeof window === "undefined" ? PALETTE_CONFIG[palette].fallback : getThemeColors(palette));
	useEffect(() => {
		const updateColors = () => {
			resetThemeColorCache();
			setColors(getThemeColors(palette));
		};
		const mediaQuery = window.matchMedia("(prefers-color-scheme: dark)");
		mediaQuery.addEventListener("change", updateColors);
		const observer = new MutationObserver(updateColors);
		observer.observe(document.documentElement, {
			attributes: true,
			attributeFilter: [
				"class",
				"data-theme",
				"data-mode"
			]
		});
		return () => {
			mediaQuery.removeEventListener("change", updateColors);
			observer.disconnect();
		};
	}, [palette]);
	return colors;
}
/**
* Hook to get all three color palettes at once.
* Useful when a component needs access to multiple palette types.
*/
function useAllThemeColors() {
	return {
		categorical: useThemeColors("categorical"),
		sequential: useThemeColors("sequential"),
		diverging: useThemeColors("diverging")
	};
}

//#endregion
export { useAllThemeColors, useThemeColors };
