// @databricks/appkit-ui/dist/react/charts/utils.js
//#region src/react/charts/utils.ts
/**
* Converts a value to a chart-compatible type.
* Handles BigInt conversion (Arrow can return BigInt64Array values).
* Handles Date objects by converting to timestamps.
*/
function toChartValue(value) {
	if (value === null || value === void 0) return 0;
	if (typeof value === "bigint") return Number(value);
	if (value instanceof Date) return value.getTime();
	if (typeof value === "string" || typeof value === "number") return value;
	return String(value);
}
/**
* Converts an array of values to chart-compatible types.
*/
function toChartArray(data) {
	if (data.length === 0) return [];
	return data.map(toChartValue);
}
/**
* Formats a field name into a human-readable label.
* Handles camelCase, snake_case, acronyms, and ALL_CAPS.
* E.g., "totalSpend" -> "Total Spend", "user_name" -> "User Name",
*       "userID" -> "User Id", "TOTAL_SPEND" -> "Total Spend"
*/
function formatLabel(field) {
	return field.replace(/([A-Z]+)([A-Z][a-z])/g, "$1 $2").replace(/([a-z])([A-Z])/g, "$1 $2").replace(/_/g, " ").replace(/\s+/g, " ").toLowerCase().replace(/\b\w/g, (l) => l.toUpperCase()).trim();
}
/**
* Truncates a label to a maximum length with ellipsis.
*/
function truncateLabel(value, maxLength = 15) {
	return value.length > maxLength ? `${value.slice(0, maxLength)}...` : value;
}
/**
* Creates time-series data pairs for ECharts.
*/
function createTimeSeriesData(xData, yData) {
	const len = xData.length;
	const result = new Array(len);
	for (let i = 0; i < len; i++) result[i] = [xData[i], yData[i]];
	return result;
}
/**
* Sorts numeric x-data in ascending order, reordering y-data to match.
* Also coerces numeric string x-values to numbers.
*/
function sortNumericAscending(xData, yDataMap, yFields) {
	if (xData.length <= 1) return {
		xData,
		yDataMap
	};
	const indices = xData.map((_, i) => i);
	indices.sort((a, b) => Number(xData[a]) - Number(xData[b]));
	const sortedXData = indices.map((i) => Number(xData[i]));
	const sortedYDataMap = {};
	for (const key of yFields) {
		const original = yDataMap[key];
		sortedYDataMap[key] = indices.map((i) => original[i]);
	}
	return {
		xData: sortedXData,
		yDataMap: sortedYDataMap
	};
}
/**
* Sorts time-series data in ascending chronological order.
*/
function sortTimeSeriesAscending(xData, yDataMap, yFields) {
	if (xData.length <= 1) return {
		xData,
		yDataMap
	};
	const first = xData[0];
	const last = xData[xData.length - 1];
	if (typeof first === "number" && typeof last === "number" && first > last) {
		const indices = xData.map((_, i) => i);
		indices.sort((a, b) => xData[a] - xData[b]);
		const sortedXData = indices.map((i) => xData[i]);
		const sortedYDataMap = {};
		for (const key of yFields) {
			const original = yDataMap[key];
			sortedYDataMap[key] = indices.map((i) => original[i]);
		}
		return {
			xData: sortedXData,
			yDataMap: sortedYDataMap
		};
	}
	return {
		xData,
		yDataMap
	};
}

//#endregion
export { createTimeSeriesData, formatLabel, sortNumericAscending, sortTimeSeriesAscending, toChartArray, toChartValue, truncateLabel };
