// @databricks/appkit-ui/dist/react/charts/wrapper.js
import { isArrowTable } from "./types.js";
import { useChartData } from "../hooks/use-chart-data.js";
import { ChartErrorBoundary } from "./chart-error-boundary.js";
import { EmptyState } from "./empty.js";
import { ErrorState } from "./error.js";
import { LoadingSkeleton } from "./loading.js";
import { jsx } from "react/jsx-runtime";

//#region src/react/charts/wrapper.tsx
function QueryModeContent({ queryKey, parameters, format, transformer, height, className, ariaLabel, testId, children }) {
	const { data, loading, error, isEmpty } = useChartData({
		queryKey,
		parameters,
		format,
		transformer
	});
	if (loading) return /* @__PURE__ */ jsx(LoadingSkeleton, { height: height ?? 300 });
	if (error) return /* @__PURE__ */ jsx(ErrorState, { error });
	if (isEmpty || !data) return /* @__PURE__ */ jsx(EmptyState, {});
	return /* @__PURE__ */ jsx(ChartErrorBoundary, {
		fallback: /* @__PURE__ */ jsx(ErrorState, { error: "Failed to render chart" }),
		children: /* @__PURE__ */ jsx("div", {
			className,
			style: { height },
			"aria-label": ariaLabel,
			"data-testid": testId,
			role: "img",
			children: children(data)
		})
	});
}
function DataModeContent({ data, height, className, ariaLabel, testId, children }) {
	if (isArrowTable(data) ? data.numRows === 0 : !Array.isArray(data) || data.length === 0) return /* @__PURE__ */ jsx(EmptyState, {});
	return /* @__PURE__ */ jsx(ChartErrorBoundary, {
		fallback: /* @__PURE__ */ jsx(ErrorState, { error: "Failed to render chart" }),
		children: /* @__PURE__ */ jsx("div", {
			className,
			style: { height },
			"aria-label": ariaLabel,
			"data-testid": testId,
			role: "img",
			children: children(data)
		})
	});
}
/**
* Wrapper component for charts.
* Handles data fetching (query mode) or direct data injection (data mode).
*
* @example Query mode - fetches data from analytics endpoint
* ```tsx
* <ChartWrapper
*   queryKey="spend_data"
*   parameters={{ limit: 100 }}
*   format="auto"
* >
*   {(data) => <MyChart data={data} />}
* </ChartWrapper>
* ```
*
* @example Data mode - uses provided data directly
* ```tsx
* <ChartWrapper data={myArrowTable}>
*   {(data) => <MyChart data={data} />}
* </ChartWrapper>
* ```
*/
function ChartWrapper(props) {
	const { height = 300, className, ariaLabel, testId, children } = props;
	if ("data" in props && props.data !== void 0) return /* @__PURE__ */ jsx(DataModeContent, {
		data: props.data,
		height,
		className,
		ariaLabel,
		testId,
		children
	});
	if ("queryKey" in props && props.queryKey !== void 0) return /* @__PURE__ */ jsx(QueryModeContent, {
		queryKey: props.queryKey,
		parameters: props.parameters,
		format: props.format,
		transformer: props.transformer,
		height,
		className,
		ariaLabel,
		testId,
		children
	});
	return /* @__PURE__ */ jsx(ErrorState, { error: "Chart requires either 'queryKey' or 'data' prop" });
}

//#endregion
export { ChartWrapper };
