// @databricks/sdk-experimental/dist/apis/agentbricks/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentBricksService = void 0;
const agentbricks = __importStar(require("./model"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
class AgentBricksRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("AgentBricks", method, message);
    }
}
class AgentBricksError extends apiError_1.ApiError {
    constructor(method, message) {
        super("AgentBricks", method, message);
    }
}
/**
 * The Custom LLMs service manages state and powers the UI for the Custom LLM
 * product.
 */
class AgentBricksService {
    constructor(client) {
        this.client = client;
    }
    async _cancelOptimize(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/custom-llms/${request.id}/optimize/cancel`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Cancel a Custom LLM Optimization Run.
     */
    async cancelOptimize(request, context) {
        return await this._cancelOptimize(request, context);
    }
    async _createCustomLlm(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("agent_artifact_path")) {
            body["agent_artifact_path"] = request.agent_artifact_path;
        }
        if (request.hasOwnProperty("datasets")) {
            body["datasets"] = request.datasets;
        }
        if (request.hasOwnProperty("guidelines")) {
            body["guidelines"] = request.guidelines;
        }
        if (request.hasOwnProperty("instructions")) {
            body["instructions"] = request.instructions;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        const query = {};
        const path = "/api/2.0/custom-llms";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a Custom LLM.
     */
    async createCustomLlm(request, context) {
        return await this._createCustomLlm(request, context);
    }
    async _deleteCustomLlm(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/custom-llms/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a Custom LLM.
     */
    async deleteCustomLlm(request, context) {
        return await this._deleteCustomLlm(request, context);
    }
    async _getCustomLlm(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/custom-llms/${request.id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a Custom LLM.
     */
    async getCustomLlm(request, context) {
        return await this._getCustomLlm(request, context);
    }
    async _startOptimize(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/custom-llms/${request.id}/optimize`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Start a Custom LLM Optimization Run.
     */
    async startOptimize(request, context) {
        return await this._startOptimize(request, context);
    }
    async _updateCustomLlm(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("custom_llm")) {
            body["custom_llm"] = request.custom_llm;
        }
        if (request.hasOwnProperty("update_mask")) {
            body["update_mask"] = request.update_mask;
        }
        const query = {};
        const path = `/api/2.0/custom-llms/${request.id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update a Custom LLM.
     */
    async updateCustomLlm(request, context) {
        return await this._updateCustomLlm(request, context);
    }
}
exports.AgentBricksService = AgentBricksService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AgentBricksService.prototype, "_cancelOptimize", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AgentBricksService.prototype, "cancelOptimize", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AgentBricksService.prototype, "_createCustomLlm", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AgentBricksService.prototype, "createCustomLlm", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AgentBricksService.prototype, "_deleteCustomLlm", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AgentBricksService.prototype, "deleteCustomLlm", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AgentBricksService.prototype, "_getCustomLlm", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AgentBricksService.prototype, "getCustomLlm", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AgentBricksService.prototype, "_startOptimize", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AgentBricksService.prototype, "startOptimize", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AgentBricksService.prototype, "_updateCustomLlm", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AgentBricksService.prototype, "updateCustomLlm", null);
