// @databricks/appkit/dist/internal-telemetry/reporter.js
import { buildAppkitPayload } from "./appkit-log.js";

//#region src/internal-telemetry/reporter.ts
const DEFAULT_HEARTBEAT_INTERVAL_MS = 300 * 1e3;
const DEFAULT_METRICS_FLUSH_INTERVAL_MS = 60 * 1e3;
function envIntervalMs(name, fallback) {
	const raw = process.env[name];
	if (!raw) return fallback;
	const n = Number(raw);
	return Number.isFinite(n) && n > 0 ? n : fallback;
}
var TelemetryReporter = class TelemetryReporter {
	static #instance = null;
	#workspaceIdPromise;
	#client;
	#appId;
	#appkitVersion;
	#heartbeatIntervalMs;
	#metricsFlushIntervalMs;
	#heartbeatTimer = null;
	#metricsTimer = null;
	#buckets = /* @__PURE__ */ new Map();
	constructor(opts) {
		this.#workspaceIdPromise = Promise.resolve(opts.workspaceId);
		this.#workspaceIdPromise.catch(() => {});
		this.#client = opts.client;
		this.#appId = opts.appId;
		this.#appkitVersion = opts.appkitVersion;
		this.#heartbeatIntervalMs = opts.heartbeatIntervalMs ?? envIntervalMs("APPKIT_TELEMETRY_HEARTBEAT_INTERVAL_MS", DEFAULT_HEARTBEAT_INTERVAL_MS);
		this.#metricsFlushIntervalMs = opts.metricsFlushIntervalMs ?? envIntervalMs("APPKIT_TELEMETRY_METRICS_FLUSH_INTERVAL_MS", DEFAULT_METRICS_FLUSH_INTERVAL_MS);
	}
	static initialize(opts) {
		TelemetryReporter.#instance?.stop();
		TelemetryReporter.#instance = new TelemetryReporter(opts);
		return TelemetryReporter.#instance;
	}
	static getInstance() {
		return TelemetryReporter.#instance;
	}
	/** @internal Test-only reset. */
	static _reset() {
		TelemetryReporter.#instance?.stop();
		TelemetryReporter.#instance = null;
	}
	start() {
		if (this.#heartbeatTimer || this.#metricsTimer) return;
		this.#heartbeatTimer = setInterval(() => {
			this.sendHeartbeat().catch(() => {});
		}, this.#heartbeatIntervalMs);
		this.#heartbeatTimer.unref?.();
		this.#metricsTimer = setInterval(() => {
			this.flushRequestMetrics().catch(() => {});
		}, this.#metricsFlushIntervalMs);
		this.#metricsTimer.unref?.();
	}
	stop() {
		if (this.#heartbeatTimer) clearInterval(this.#heartbeatTimer);
		if (this.#metricsTimer) clearInterval(this.#metricsTimer);
		this.#heartbeatTimer = null;
		this.#metricsTimer = null;
	}
	recordRequest(method, routeTemplate, statusCode, latencyMs) {
		if (!routeTemplate) return;
		const key = `${method.toUpperCase()} ${routeTemplate}`;
		const bucket = this.#buckets.get(key) ?? {
			count: 0,
			latencyMsTotal: 0,
			http4xx: 0,
			http5xx: 0
		};
		bucket.count += 1;
		bucket.latencyMsTotal += Math.max(0, latencyMs);
		if (statusCode >= 400 && statusCode < 500) bucket.http4xx += 1;
		if (statusCode >= 500 && statusCode < 600) bucket.http5xx += 1;
		this.#buckets.set(key, bucket);
	}
	async sendStartup() {
		await this.#send([this.#wrap({
			event_name: "APP_STARTUP",
			app_startup_event: {}
		})]);
	}
	async sendHeartbeat() {
		await this.#send([this.#wrap({
			event_name: "HEARTBEAT",
			heartbeat_event: {}
		})]);
	}
	async flushRequestMetrics() {
		if (this.#buckets.size === 0) return;
		const drained = this.#buckets;
		this.#buckets = /* @__PURE__ */ new Map();
		const logs = [];
		for (const [endpoint, bucket] of drained) {
			const event = {
				endpoint,
				request_count: bucket.count,
				request_latency_ms_avg: Math.round(bucket.latencyMsTotal / bucket.count),
				response_count_http4xx: bucket.http4xx,
				response_count_http5xx: bucket.http5xx
			};
			logs.push(this.#wrap({
				event_name: "REQUEST_METRICS",
				request_metrics_event: event
			}));
		}
		await this.#send(logs);
	}
	#wrap(partial) {
		return {
			...partial,
			app_id: this.#appId,
			appkit_version: this.#appkitVersion
		};
	}
	async #send(logs) {
		if (logs.length === 0) return;
		const workspaceId = await this.#workspaceIdPromise;
		await this.#client.apiClient.request({
			path: "/telemetry-ext",
			method: "POST",
			query: { o: workspaceId },
			headers: new Headers(),
			payload: buildAppkitPayload(logs),
			raw: false
		});
	}
};

//#endregion
export { TelemetryReporter };
