// @databricks/sdk-experimental/dist/mixins/compute.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
const compute_1 = require("./compute");
/**
 * Unit tests for ClustersExt selector logic
 *
 * Note: These tests focus on the filtering and sorting logic.
 * Full integration tests are in compute.integ.ts
 */
describe("ClustersExt", () => {
    // Helper to create a minimal ClustersExt instance for testing
    function createTestClustersExt() {
        // Create a minimal instance with mocked dependencies
        const instance = Object.create(compute_1.ClustersExt.prototype);
        return instance;
    }
    describe("SemVer extraction from Databricks Runtime keys", () => {
        it("should extract version from standard runtime key", () => {
            const ext = createTestClustersExt();
            const version = ext._extractSemanticVersion("13.3.x-scala2.12");
            assert.ok(version);
            assert.strictEqual(version.major, 13);
            assert.strictEqual(version.minor, 3);
            assert.strictEqual(version.patch, 0);
        });
        it("should extract version from ML runtime key", () => {
            const ext = createTestClustersExt();
            const version = ext._extractSemanticVersion("14.1.x-cpu-ml-scala2.12");
            assert.ok(version);
            assert.strictEqual(version.major, 14);
            assert.strictEqual(version.minor, 1);
            assert.strictEqual(version.patch, 0);
        });
        it("should extract version from GPU runtime key", () => {
            const ext = createTestClustersExt();
            const version = ext._extractSemanticVersion("13.3.x-gpu-ml-scala2.12");
            assert.ok(version);
            assert.strictEqual(version.major, 13);
            assert.strictEqual(version.minor, 3);
            assert.strictEqual(version.patch, 0);
        });
        it("should extract version from Photon runtime key", () => {
            const ext = createTestClustersExt();
            const version = ext._extractSemanticVersion("13.3.x-photon-scala2.12");
            assert.ok(version);
            assert.strictEqual(version.major, 13);
            assert.strictEqual(version.minor, 3);
            assert.strictEqual(version.patch, 0);
        });
        it("should extract version with explicit patch", () => {
            const ext = createTestClustersExt();
            const version = ext._extractSemanticVersion("13.3.2-scala2.12");
            assert.ok(version);
            assert.strictEqual(version.major, 13);
            assert.strictEqual(version.minor, 3);
            assert.strictEqual(version.patch, 2);
        });
        it("should return null for invalid version string", () => {
            const ext = createTestClustersExt();
            const version = ext._extractSemanticVersion("invalid-version");
            assert.strictEqual(version, null);
        });
        it("should return null for empty string", () => {
            const ext = createTestClustersExt();
            const version = ext._extractSemanticVersion("");
            assert.strictEqual(version, null);
        });
    });
    describe("SemVer sorting logic", () => {
        it("should sort versions newest first", () => {
            const ext = createTestClustersExt();
            const versions = [
                { key: "13.3.x-scala2.12", name: "13.3" },
                { key: "14.0.x-scala2.12", name: "14.0" },
                { key: "13.2.x-scala2.12", name: "13.2" },
                { key: "14.1.x-scala2.12", name: "14.1" },
            ];
            const sorted = ext._sortVersionsBySemanticVersion(versions);
            assert.strictEqual(sorted.length, 4);
            assert.strictEqual(sorted[0].key, "14.1.x-scala2.12");
            assert.strictEqual(sorted[1].key, "14.0.x-scala2.12");
            assert.strictEqual(sorted[2].key, "13.3.x-scala2.12");
            assert.strictEqual(sorted[3].key, "13.2.x-scala2.12");
        });
        it("should sort by minor version when major is same", () => {
            const ext = createTestClustersExt();
            const versions = [
                { key: "13.1.x-scala2.12", name: "13.1" },
                { key: "13.3.x-scala2.12", name: "13.3" },
                { key: "13.2.x-scala2.12", name: "13.2" },
            ];
            const sorted = ext._sortVersionsBySemanticVersion(versions);
            assert.strictEqual(sorted[0].key, "13.3.x-scala2.12");
            assert.strictEqual(sorted[1].key, "13.2.x-scala2.12");
            assert.strictEqual(sorted[2].key, "13.1.x-scala2.12");
        });
        it("should sort by patch version when major and minor are same", () => {
            const ext = createTestClustersExt();
            const versions = [
                { key: "13.3.1-scala2.12", name: "13.3.1" },
                { key: "13.3.3-scala2.12", name: "13.3.3" },
                { key: "13.3.2-scala2.12", name: "13.3.2" },
            ];
            const sorted = ext._sortVersionsBySemanticVersion(versions);
            assert.strictEqual(sorted[0].key, "13.3.3-scala2.12");
            assert.strictEqual(sorted[1].key, "13.3.2-scala2.12");
            assert.strictEqual(sorted[2].key, "13.3.1-scala2.12");
        });
        it("should handle versions with 'x' in patch position", () => {
            const ext = createTestClustersExt();
            const versions = [
                { key: "13.3.2-scala2.12", name: "13.3.2" },
                { key: "13.3.x-scala2.12", name: "13.3.x" },
                { key: "13.3.1-scala2.12", name: "13.3.1" },
            ];
            const sorted = ext._sortVersionsBySemanticVersion(versions);
            // 13.3.x should be treated as 13.3.0
            assert.strictEqual(sorted[0].key, "13.3.2-scala2.12");
            assert.strictEqual(sorted[1].key, "13.3.1-scala2.12");
            assert.strictEqual(sorted[2].key, "13.3.x-scala2.12");
        });
        it("should preserve order for unparseable versions", () => {
            const ext = createTestClustersExt();
            const versions = [
                { key: "invalid1", name: "Invalid 1" },
                { key: "13.3.x-scala2.12", name: "13.3" },
                { key: "invalid2", name: "Invalid 2" },
            ];
            const sorted = ext._sortVersionsBySemanticVersion(versions);
            // Valid version should be first, invalid ones maintain relative order
            assert.strictEqual(sorted[0].key, "13.3.x-scala2.12");
        });
        it("should handle empty version list", () => {
            const ext = createTestClustersExt();
            const versions = [];
            const sorted = ext._sortVersionsBySemanticVersion(versions);
            assert.strictEqual(sorted.length, 0);
        });
        it("should handle single version", () => {
            const ext = createTestClustersExt();
            const versions = [
                { key: "13.3.x-scala2.12", name: "13.3" },
            ];
            const sorted = ext._sortVersionsBySemanticVersion(versions);
            assert.strictEqual(sorted.length, 1);
            assert.strictEqual(sorted[0].key, "13.3.x-scala2.12");
        });
        it("should handle ML and standard runtimes with same version", () => {
            const ext = createTestClustersExt();
            const versions = [
                { key: "13.3.x-cpu-ml-scala2.12", name: "13.3 ML" },
                { key: "13.3.x-scala2.12", name: "13.3" },
            ];
            const sorted = ext._sortVersionsBySemanticVersion(versions);
            // Both should have same sort order (stable sort maintains input order)
            assert.strictEqual(sorted.length, 2);
            // Both are version 13.3.0, so order should be maintained
        });
    });
    describe("SemVer parsing and version selection logic", () => {
        it("should filter versions by Scala version", () => {
            const versions = [
                { key: "13.3.x-scala2.12", name: "13.3 LTS" },
                { key: "13.3.x-scala2.11", name: "13.3 LTS Scala 2.11" },
            ];
            const scalaFiltered = versions.filter((v) => v.key.includes("-scala2.12"));
            assert.strictEqual(scalaFiltered.length, 1);
            assert.strictEqual(scalaFiltered[0].key, "13.3.x-scala2.12");
        });
        it("should filter versions by LTS", () => {
            const versions = [
                { key: "13.3.x-scala2.12", name: "13.3 LTS" },
                { key: "14.0.x-scala2.12", name: "14.0" },
            ];
            const ltsFiltered = versions.filter((v) => v.name.includes("LTS") || v.key.includes("-esr-"));
            assert.strictEqual(ltsFiltered.length, 1);
            assert.strictEqual(ltsFiltered[0].key, "13.3.x-scala2.12");
        });
        it("should filter versions by ML runtime", () => {
            const versions = [
                { key: "13.3.x-cpu-ml-scala2.12", name: "13.3 ML" },
                { key: "13.3.x-scala2.12", name: "13.3" },
            ];
            const mlFiltered = versions.filter((v) => v.key.includes("-ml-"));
            assert.strictEqual(mlFiltered.length, 1);
            assert.strictEqual(mlFiltered[0].key, "13.3.x-cpu-ml-scala2.12");
        });
        it("should filter versions by GPU support", () => {
            const versions = [
                { key: "13.3.x-gpu-ml-scala2.12", name: "13.3 GPU ML" },
                { key: "13.3.x-cpu-ml-scala2.12", name: "13.3 ML" },
            ];
            const gpuFiltered = versions.filter((v) => v.key.includes("-gpu-"));
            assert.strictEqual(gpuFiltered.length, 1);
            assert.strictEqual(gpuFiltered[0].key, "13.3.x-gpu-ml-scala2.12");
        });
        it("should exclude beta versions", () => {
            const versions = [
                { key: "14.0.x-scala2.12", name: "14.0 Beta" },
                { key: "13.3.x-scala2.12", name: "13.3 LTS" },
            ];
            const nonBetaFiltered = versions.filter((v) => !v.name.includes("Beta"));
            assert.strictEqual(nonBetaFiltered.length, 1);
            assert.strictEqual(nonBetaFiltered[0].key, "13.3.x-scala2.12");
        });
    });
    describe("Node type filtering logic", () => {
        it("should filter by memory requirement", () => {
            const nodeTypes = [
                { node_type_id: "m5.large", memory_mb: 8192, num_cores: 2 }, // 8 GB
                { node_type_id: "m5.xlarge", memory_mb: 16384, num_cores: 4 }, // 16 GB
                { node_type_id: "m5.2xlarge", memory_mb: 32768, num_cores: 8 }, // 32 GB
            ];
            const minMemoryGb = 16;
            const filtered = nodeTypes.filter((nt) => (nt.memory_mb || 0) / 1024 >= minMemoryGb);
            assert.strictEqual(filtered.length, 2);
            const ids = filtered.map((n) => n.node_type_id);
            assert.ok(ids.includes("m5.xlarge"));
            assert.ok(ids.includes("m5.2xlarge"));
        });
        it("should filter by core count", () => {
            const nodeTypes = [
                { node_type_id: "m5.large", memory_mb: 8192, num_cores: 2 },
                { node_type_id: "m5.xlarge", memory_mb: 16384, num_cores: 4 },
                { node_type_id: "m5.2xlarge", memory_mb: 32768, num_cores: 8 },
            ];
            const minCores = 4;
            const filtered = nodeTypes.filter((nt) => (nt.num_cores || 0) >= minCores);
            assert.strictEqual(filtered.length, 2);
            const ids = filtered.map((n) => n.node_type_id);
            assert.ok(ids.includes("m5.xlarge"));
            assert.ok(ids.includes("m5.2xlarge"));
        });
        it("should skip deprecated nodes", () => {
            const nodeTypes = [
                {
                    node_type_id: "m4.large",
                    memory_mb: 8192,
                    is_deprecated: true,
                },
                {
                    node_type_id: "m5.large",
                    memory_mb: 8192,
                    is_deprecated: false,
                },
            ];
            const filtered = nodeTypes.filter((nt) => !nt.is_deprecated);
            assert.strictEqual(filtered.length, 1);
            assert.strictEqual(filtered[0].node_type_id, "m5.large");
        });
        it("should filter by GPU count", () => {
            const nodeTypes = [
                { node_type_id: "p3.2xlarge", num_gpus: 1 },
                { node_type_id: "p3.8xlarge", num_gpus: 4 },
                { node_type_id: "m5.xlarge", num_gpus: 0 },
            ];
            const minGpus = 1;
            const filtered = nodeTypes.filter((nt) => (nt.num_gpus || 0) >= minGpus);
            assert.strictEqual(filtered.length, 2);
            const ids = filtered.map((n) => n.node_type_id);
            assert.ok(ids.includes("p3.2xlarge"));
            assert.ok(ids.includes("p3.8xlarge"));
        });
        it("should exclude GPU nodes when minGpus is 0", () => {
            const nodeTypes = [
                { node_type_id: "p3.2xlarge", num_gpus: 1 },
                { node_type_id: "m5.xlarge", num_gpus: 0 },
            ];
            const minGpus = 0;
            const filtered = nodeTypes.filter((nt) => !((nt.num_gpus || 0) > 0 && minGpus === 0));
            assert.strictEqual(filtered.length, 1);
            assert.strictEqual(filtered[0].node_type_id, "m5.xlarge");
        });
    });
});
