// @databricks/sdk-experimental/dist/config/AzureMsiCredentials.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const Config_1 = require("./Config");
const f = __importStar(require("../fetch"));
const AzureMsiCredentials_1 = require("./AzureMsiCredentials");
const assert = __importStar(require("node:assert"));
describe(__filename, () => {
    let envBackup;
    let fetchBackup;
    beforeEach(() => {
        envBackup = process.env;
        process.env = {};
        // mock fetch
        fetchBackup = f.fetch;
    });
    afterEach(() => {
        process.env = envBackup;
        f.fetch = fetchBackup;
    });
    it("should work for the happy path", async () => {
        const config = new Config_1.Config({
            azureUseMsi: true,
            azureWorkspaceResourceId: "/a/b/c",
        });
        const auth = new AzureMsiCredentials_1.AzureMsiCredentials();
        const responses = {
            "GET http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https%3A%2F%2Fmanagement.azure.com%2F": {
                access_token: "access_token1",
                client_id: "client_id1",
                expires_in: "86400",
                expires_on: "1704880572",
                ext_expires_in: "86399",
                not_before: "1704793872",
                resource: "https://management.azure.com/",
                token_type: "Bearer",
            },
            "GET https://management.azure.com/a/b/c?api-version=2018-04-01": {
                properties: {
                    workspaceUrl: "https://x",
                },
            },
            "GET http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=2ff814a6-3304-4ab8-85cb-cd0e6f879c1d": {
                access_token: "access_token2",
                client_id: "client_id1",
                expires_in: "86400",
                expires_on: "1704880573",
                ext_expires_in: "86399",
                not_before: "1704793873",
                resource: "2ff814a6-3304-4ab8-85cb-cd0e6f879c1d",
                token_type: "Bearer",
            },
            "GET http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https%3A%2F%2Fmanagement.core.windows.net%2F": {
                access_token: "access_token3",
                client_id: "client_id1",
                expires_in: "86400",
                expires_on: "1704880573",
                ext_expires_in: "86399",
                not_before: "1704793873",
                resource: "https://management.core.windows.net/",
                token_type: "Bearer",
            },
        };
        // mock MSI
        f.fetch = (url) => {
            const response = responses[`GET ${url.href}`];
            if (!response) {
                throw new Error(`unknown URL: ${url}`);
            }
            return {
                ok: true,
                async json() {
                    return response;
                },
            };
        };
        const visitor = await auth.configure(config);
        assert.ok(visitor);
        const headers = new f.Headers();
        await visitor(headers);
        assert.equal(headers.get("authorization"), "Bearer access_token2");
        assert.equal(headers.get("x-databricks-azure-sp-management-token"), "access_token3");
        assert.equal(headers.get("x-databricks-azure-workspace-resource-id"), "/a/b/c");
    });
    it("should fail if token is not found", async () => {
        const config = new Config_1.Config({
            azureUseMsi: true,
            azureClientId: "abc",
            azureWorkspaceResourceId: "/a/b/c",
        });
        const auth = new AzureMsiCredentials_1.AzureMsiCredentials();
        // mock MSI
        f.fetch = () => {
            return {
                status: 400,
                statusText: "Bad Request",
                ok: false,
                async text() {
                    return '{"error":"invalid_request","error_description":"Identity not found"}';
                },
            };
        };
        let hadError = false;
        try {
            await auth.configure(config);
        }
        catch (e) {
            assert.ok(e.message.includes("azure-msi: failed to get token"));
            hadError = true;
        }
        assert.ok(hadError);
    });
});
