// @databricks/sdk-experimental/dist/config/AzureCliCredentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureCliCredentials = void 0;
const Token_1 = require("./Token");
const Config_1 = require("./Config");
const execUtils_1 = require("./execUtils");
const Azure_1 = require("./Azure");
/**
 * Authenticate using Azure CLI
 */
class AzureCliCredentials {
    constructor() {
        this.name = "azure-cli";
    }
    async configure(config) {
        if (!config.isAzure()) {
            return;
        }
        const appId = (0, Azure_1.getAzureLoginAppId)(config);
        const ts = this.getTokenSource(config, appId);
        try {
            await ts();
        }
        catch (error) {
            if (error instanceof Config_1.ConfigError) {
                if (error.message.indexOf("Can't find 'az' command") >= 0) {
                    config.logger.debug("Most likely Azure CLI is not installed. " +
                        "See https://docs.microsoft.com/en-us/cli/azure/?view=azure-cli-latest for details");
                    return;
                }
                throw error;
            }
            throw error;
        }
        await (0, Azure_1.azureEnsureWorkspaceUrl)(config, (resource) => {
            return this.getTokenSource(config, resource);
        });
        return this.getVisitor(config, ts);
    }
    async getVisitor(config, innerTokenSource) {
        const env = (0, Azure_1.getAzureEnvironment)(config);
        const managementTs = (0, Token_1.refreshableTokenSource)(this.getTokenSource(config, env.serviceManagementEndpoint));
        try {
            await managementTs();
        }
        catch (error) {
            config.logger.debug(`Not including service management token in headers: ${error}`);
            return (0, Azure_1.azureVisitor)(config, (0, Token_1.refreshableTokenVisitor)(innerTokenSource));
        }
        return (0, Azure_1.azureVisitor)(config, (0, Azure_1.serviceToServiceVisitor)((0, Token_1.refreshableTokenSource)(innerTokenSource), managementTs, "X-Databricks-Azure-SP-Management-Token"));
    }
    getTokenSource(config, resource) {
        return async () => {
            const subscription = this.getSubscription(config);
            if (subscription !== undefined) {
                try {
                    const token = this.fetchToken(config, resource, subscription);
                    return await token();
                }
                catch (_e) {
                    config.logger.warn(`Failed to get token for subscription. Using resource only token.`);
                }
            }
            else {
                config.logger.warn(`azure_workspace_resource_id field not provided. ` +
                    `It is recommended to specify this field in the Databricks configuration to avoid authentication errors.`);
            }
            const token = this.fetchToken(config, resource);
            return await token();
        };
    }
    fetchToken(config, resource, subscription) {
        return async () => {
            let stdout = "";
            try {
                const args = [
                    "account",
                    "get-access-token",
                    "--resource",
                    resource,
                    "--output",
                    "json",
                ];
                if (subscription !== undefined) {
                    args.push("--subscription", subscription);
                }
                ({ stdout } = await (0, execUtils_1.execFileWithShell)("az", args));
            }
            catch (e) {
                if (e instanceof execUtils_1.FileNotFoundException) {
                    throw new Config_1.ConfigError("azure-cli: Can't find 'az' command. Please install Azure CLI: https://docs.microsoft.com/en-us/cli/azure/install-azure-cli", config);
                }
                if ((0, execUtils_1.isExecFileException)(e)) {
                    throw new Config_1.ConfigError(`azure-cli: cannot get access token: ${e.stderr}`, config);
                }
                throw new Config_1.ConfigError(`azure-cli: cannot get access token: ${e + ""}`, config);
            }
            let token;
            let azureToken;
            try {
                azureToken = JSON.parse(stdout);
            }
            catch (e) {
                throw new Config_1.ConfigError(`azure-cli: cannot parse access token: ${e.message}`, config);
            }
            if (!isAzureToken(azureToken)) {
                throw new Config_1.ConfigError(`azure-cli: cannot parse access token: ${stdout}`, config);
            }
            token = new Token_1.Token({
                accessToken: azureToken.accessToken,
                expiry: new Date(azureToken.expiresOn).getTime(),
            });
            config.logger.info(`Refreshed OAuth token for ${resource} from Azure CLI, which expires on ${azureToken.expiresOn}`);
            return token;
        };
    }
    getSubscription(config) {
        const resourceId = config.azureWorkspaceResourceId;
        if (resourceId === undefined) {
            return undefined;
        }
        const components = resourceId.split("/");
        if (components.length < 3) {
            config.logger.warn(`Invalid azure workspace resource ID`);
            return undefined;
        }
        return components[2];
    }
}
exports.AzureCliCredentials = AzureCliCredentials;
function isAzureToken(token) {
    return (typeof token === "object" &&
        token !== null &&
        "accessToken" in token &&
        "expiresOn" in token);
}
