// @databricks/sdk-experimental/dist/config/Config.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mocha_1 = require("mocha");
const node_assert_1 = __importDefault(require("node:assert"));
const Config_1 = require("./Config");
const HostType_1 = require("./HostType");
const f = __importStar(require("../fetch"));
(0, mocha_1.describe)(__filename, () => {
    let envBackup;
    let fetchBackup;
    (0, mocha_1.beforeEach)(() => {
        envBackup = process.env;
        process.env = {};
        fetchBackup = f.fetch;
    });
    (0, mocha_1.afterEach)(() => {
        process.env = envBackup;
        f.fetch = fetchBackup;
    });
    (0, mocha_1.describe)("hostType()", () => {
        (0, mocha_1.it)("should return UnifiedHost when experimentalIsUnifiedHost is true", () => {
            const config = new Config_1.Config({
                host: "https://company.databricks.com",
                experimentalIsUnifiedHost: true,
            });
            node_assert_1.default.strictEqual(config.hostType(), HostType_1.HostType.unifiedHost);
        });
        (0, mocha_1.it)("should return AccountHost for accounts hosts", () => {
            const config = new Config_1.Config({
                host: "https://accounts.cloud.databricks.com",
            });
            node_assert_1.default.strictEqual(config.hostType(), HostType_1.HostType.accountHost);
        });
        (0, mocha_1.it)("should return WorkspaceHost for traditional workspace hosts", () => {
            const config = new Config_1.Config({
                host: "https://dbc-12345.cloud.databricks.com",
            });
            node_assert_1.default.strictEqual(config.hostType(), HostType_1.HostType.workspaceHost);
        });
        (0, mocha_1.it)("should return WorkspaceHost for Azure workspace hosts", () => {
            const config = new Config_1.Config({
                host: "https://adb-123.45.azuredatabricks.net",
            });
            node_assert_1.default.strictEqual(config.hostType(), HostType_1.HostType.workspaceHost);
        });
        (0, mocha_1.it)("should return WorkspaceHost for GCP workspace hosts", () => {
            const config = new Config_1.Config({
                host: "https://1234567890.7.gcp.databricks.com",
            });
            node_assert_1.default.strictEqual(config.hostType(), HostType_1.HostType.workspaceHost);
        });
        (0, mocha_1.it)("should prioritize experimentalIsUnifiedHost flag over hostname pattern", () => {
            const config = new Config_1.Config({
                host: "https://accounts.cloud.databricks.com",
                experimentalIsUnifiedHost: true,
            });
            node_assert_1.default.strictEqual(config.hostType(), HostType_1.HostType.unifiedHost);
        });
    });
    (0, mocha_1.describe)("configType()", () => {
        (0, mocha_1.it)("should return AccountConfig for account hosts", () => {
            const config = new Config_1.Config({
                host: "https://accounts.cloud.databricks.com",
                accountId: "12345",
            });
            node_assert_1.default.strictEqual(config.configType(), HostType_1.ConfigType.accountConfig);
        });
        (0, mocha_1.it)("should return WorkspaceConfig for workspace hosts", () => {
            const config = new Config_1.Config({
                host: "https://dbc-12345.cloud.databricks.com",
            });
            node_assert_1.default.strictEqual(config.configType(), HostType_1.ConfigType.workspaceConfig);
        });
        (0, mocha_1.it)("should return InvalidConfig for unified host without accountId", () => {
            const config = new Config_1.Config({
                host: "https://company.databricks.com",
                experimentalIsUnifiedHost: true,
            });
            node_assert_1.default.strictEqual(config.configType(), HostType_1.ConfigType.invalidConfig);
        });
        (0, mocha_1.it)("should return AccountConfig for unified host with accountId only", () => {
            const config = new Config_1.Config({
                host: "https://company.databricks.com",
                experimentalIsUnifiedHost: true,
                accountId: "12345",
            });
            node_assert_1.default.strictEqual(config.configType(), HostType_1.ConfigType.accountConfig);
        });
        (0, mocha_1.it)("should return WorkspaceConfig for unified host with accountId and workspaceId", () => {
            const config = new Config_1.Config({
                host: "https://company.databricks.com",
                experimentalIsUnifiedHost: true,
                accountId: "12345",
                workspaceId: "67890",
            });
            node_assert_1.default.strictEqual(config.configType(), HostType_1.ConfigType.workspaceConfig);
        });
    });
    (0, mocha_1.describe)("getOidcEndpoints() - Unified Host", () => {
        (0, mocha_1.it)("should use discovery endpoint with accountId for unified hosts", async () => {
            f.fetch = (url) => {
                if (url ===
                    "https://company.databricks.com/oidc/accounts/12345/.well-known/oauth-authorization-server") {
                    return Promise.resolve({
                        status: 200,
                        json: async () => ({
                            authorization_endpoint: "https://company.databricks.com/oidc/v1/authorize",
                            token_endpoint: "https://company.databricks.com/oidc/v1/token",
                        }),
                    });
                }
                throw new Error(`Unexpected URL: ${url}`);
            };
            const config = new Config_1.Config({
                host: "https://company.databricks.com",
                experimentalIsUnifiedHost: true,
                accountId: "12345",
            });
            const endpoints = await config.getOidcEndpoints();
            node_assert_1.default.ok(endpoints);
            node_assert_1.default.strictEqual(endpoints.authorizationEndpoint.href, "https://company.databricks.com/oidc/v1/authorize");
            node_assert_1.default.strictEqual(endpoints.tokenEndpoint.href, "https://company.databricks.com/oidc/v1/token");
        });
        (0, mocha_1.it)("should return undefined if discovery endpoint returns non-200", async () => {
            f.fetch = () => {
                return Promise.resolve({
                    status: 404,
                });
            };
            const config = new Config_1.Config({
                host: "https://company.databricks.com",
                experimentalIsUnifiedHost: true,
                accountId: "12345",
            });
            const endpoints = await config.getOidcEndpoints();
            node_assert_1.default.strictEqual(endpoints, undefined);
        });
    });
    (0, mocha_1.describe)("getOidcEndpoints() - Account Host", () => {
        (0, mocha_1.it)("should use fixed URL pattern for account hosts", async () => {
            const config = new Config_1.Config({
                host: "https://accounts.cloud.databricks.com",
                accountId: "12345",
            });
            const endpoints = await config.getOidcEndpoints();
            node_assert_1.default.ok(endpoints);
            node_assert_1.default.strictEqual(endpoints.authorizationEndpoint.href, "https://accounts.cloud.databricks.com/oidc/accounts/12345/v1/authorize");
            node_assert_1.default.strictEqual(endpoints.tokenEndpoint.href, "https://accounts.cloud.databricks.com/oidc/accounts/12345/v1/token");
        });
    });
    (0, mocha_1.describe)("getOidcEndpoints() - Workspace Host", () => {
        (0, mocha_1.it)("should use standard discovery endpoint for workspace hosts", async () => {
            f.fetch = (url) => {
                if (url ===
                    "https://dbc-12345.cloud.databricks.com/oidc/.well-known/oauth-authorization-server") {
                    return Promise.resolve({
                        status: 200,
                        json: async () => ({
                            authorization_endpoint: "https://dbc-12345.cloud.databricks.com/oidc/v1/authorize",
                            token_endpoint: "https://dbc-12345.cloud.databricks.com/oidc/v1/token",
                        }),
                    });
                }
                throw new Error(`Unexpected URL: ${url}`);
            };
            const config = new Config_1.Config({
                host: "https://dbc-12345.cloud.databricks.com",
            });
            const endpoints = await config.getOidcEndpoints();
            node_assert_1.default.ok(endpoints);
            node_assert_1.default.strictEqual(endpoints.authorizationEndpoint.href, "https://dbc-12345.cloud.databricks.com/oidc/v1/authorize");
            node_assert_1.default.strictEqual(endpoints.tokenEndpoint.href, "https://dbc-12345.cloud.databricks.com/oidc/v1/token");
        });
    });
    (0, mocha_1.describe)("experimentalIsUnifiedHost attribute", () => {
        (0, mocha_1.it)("should be settable from environment variable", async () => {
            process.env.DATABRICKS_EXPERIMENTAL_IS_UNIFIED_HOST = "true";
            process.env.DATABRICKS_HOST = "https://company.databricks.com";
            const config = new Config_1.Config({});
            await config.ensureResolved();
            // Environment variables are loaded as strings
            node_assert_1.default.strictEqual(config.experimentalIsUnifiedHost, "true");
        });
        (0, mocha_1.it)("should be settable from constructor", () => {
            const config = new Config_1.Config({
                experimentalIsUnifiedHost: true,
            });
            node_assert_1.default.strictEqual(config.experimentalIsUnifiedHost, true);
        });
    });
    (0, mocha_1.describe)("workspaceId attribute", () => {
        (0, mocha_1.it)("should be settable from environment variable", async () => {
            process.env.DATABRICKS_WORKSPACE_ID = "test-workspace-id";
            process.env.DATABRICKS_HOST = "https://company.databricks.com";
            const config = new Config_1.Config({});
            await config.ensureResolved();
            node_assert_1.default.strictEqual(config.workspaceId, "test-workspace-id");
        });
        (0, mocha_1.it)("should be settable from constructor", () => {
            const config = new Config_1.Config({
                workspaceId: "workspace-123",
            });
            node_assert_1.default.strictEqual(config.workspaceId, "workspace-123");
        });
    });
});
