// @databricks/sdk-experimental/dist/config/Config.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.ConfigError = void 0;
exports.isConfigWithHost = isConfigWithHost;
exports.isAnyAuthConfigured = isAnyAuthConfigured;
require("reflect-metadata");
const logging_1 = require("../logging");
const ConfigAttributes_1 = require("./ConfigAttributes");
const DefaultCredentials_1 = require("./DefaultCredentials");
const KnownConfigLoader_1 = require("./KnownConfigLoader");
const fetch_1 = require("../fetch");
const OidcEndpoints_1 = require("./oauth/OidcEndpoints");
const HostType_1 = require("./HostType");
class ConfigError extends Error {
    constructor(baseMessage, config) {
        let msg = baseMessage;
        const debugString = config.attributes.debugString();
        if (debugString) {
            msg += `. ${debugString}`;
        }
        super(msg);
        this.baseMessage = baseMessage;
        this.config = config;
    }
}
exports.ConfigError = ConfigError;
function isConfigWithHost(config) {
    return !!config.host;
}
class Config {
    constructor(config) {
        this.config = config;
        this.resolved = false;
        this.attributes = (0, ConfigAttributes_1.getAttributesFromDecorators)(Object.getPrototypeOf(this), this);
        for (const [key, value] of Object.entries(config)) {
            this[key] = value;
        }
        this.logger =
            config.logger || logging_1.NamedLogger.getOrCreate(logging_1.ExposedLoggers.SDK);
        this.env = config.env || process.env;
    }
    async getHost() {
        await this.ensureResolved();
        if (!this.host) {
            throw new ConfigError("host is not set", this);
        }
        return new URL(this.host);
    }
    setAttribute(name, value) {
        this[name] = value;
    }
    /**
     * Authenticate adds special headers to HTTP request to authorize it to work with Databricks REST API
     */
    async authenticate(headers) {
        await this.ensureResolved();
        await this.configureCredentialProvider();
        if (!this.auth) {
            throw new ConfigError("failed to configure credential provider", this);
        }
        await this.auth(headers);
    }
    /**
     * isAzure returns true if client is configured for Azure Databricks
     */
    isAzure() {
        return ((!!this.host &&
            !!this.host.match(/(\.databricks\.azure\.us|\.databricks\.azure\.cn|\.azuredatabricks\.net)$/)) ||
            !!this.azureWorkspaceResourceId);
    }
    /**
     * isGcp returns true if client is configured for GCP
     */
    isGcp() {
        return !!this.host && this.host.endsWith(".gcp.databricks.com");
    }
    /**
     * isAws returns true if client is configured for AWS
     */
    isAws() {
        return !!this.host && !this.isAzure() && !this.isGcp();
    }
    /**
     * isAccountClient returns true if client is configured for Accounts API
     *
     * @deprecated This method is limited and does not properly handle unified hosts.
     * Use hostType() and configType() instead for more accurate host detection.
     *
     * Note: On unified hosts, this will return false even when the config is for
     * account-level operations, leading to incorrect behavior.
     */
    isAccountClient() {
        return !!this.host && this.host.startsWith("https://accounts.");
    }
    /**
     * Returns the type of host this config is pointing to.
     * - unifiedHost: when experimentalIsUnifiedHost flag is set
     * - accountHost: hosts starting with "https://accounts."
     * - workspaceHost: all other hosts
     */
    hostType() {
        if (this.experimentalIsUnifiedHost) {
            return HostType_1.HostType.unifiedHost;
        }
        if (this.host && this.host.startsWith("https://accounts.")) {
            return HostType_1.HostType.accountHost;
        }
        return HostType_1.HostType.workspaceHost;
    }
    /**
     * Returns the type of client configuration based on host type and config attributes.
     * - accountConfig: For account-level operations
     * - workspaceConfig: For workspace-level operations
     * - invalidConfig: Misconfigured (e.g., unified host without accountId)
     */
    configType() {
        const hType = this.hostType();
        switch (hType) {
            case HostType_1.HostType.accountHost:
                return HostType_1.ConfigType.accountConfig;
            case HostType_1.HostType.workspaceHost:
                return HostType_1.ConfigType.workspaceConfig;
            case HostType_1.HostType.unifiedHost:
                // Unified hosts require accountId
                if (!this.accountId) {
                    return HostType_1.ConfigType.invalidConfig;
                }
                // If workspaceId is set, it's a workspace config
                // Otherwise, it's an account config
                return this.workspaceId
                    ? HostType_1.ConfigType.workspaceConfig
                    : HostType_1.ConfigType.accountConfig;
            default:
                return HostType_1.ConfigType.invalidConfig;
        }
    }
    async ensureResolved() {
        if (this.resolved) {
            return;
        }
        const loaders = this.loaders ?? [
            new ConfigAttributes_1.EnvironmentLoader(),
            new KnownConfigLoader_1.KnownConfigLoader(),
        ];
        for (const loader of loaders) {
            this.logger.info(`Loading config via ${loader.name}`);
            await loader.configure(this);
        }
        await this.attributes.validate();
        this.fixHost();
        this.resolved = true;
    }
    fixHost() {
        if (!this.host) {
            return;
        }
        let host = this.host;
        if (!host.startsWith("http")) {
            host = `https://${host}`;
        }
        this.host = `https://${new URL(host).hostname}`;
    }
    async configureCredentialProvider() {
        if (this.auth) {
            return;
        }
        if (!this.credentials) {
            this.credentials = new DefaultCredentials_1.DefaultCredentials();
        }
        try {
            this.auth = await this.credentials.configure(this);
        }
        catch (e) {
            if (e instanceof ConfigError) {
                throw new ConfigError(`${this.credentials.name} auth: ${e.baseMessage}`, this);
            }
            throw e;
        }
        this.authType = this.credentials.name;
    }
    async getOidcEndpoints() {
        if (!this.host) {
            return;
        }
        const hType = this.hostType();
        // Account hosts: Use fixed URL pattern
        if (hType === HostType_1.HostType.accountHost && this.accountId) {
            const prefix = `${this.host}/oidc/accounts/${this.accountId}`;
            return new OidcEndpoints_1.OidcEndpoints(this, new URL(`${prefix}/v1/authorize`), new URL(`${prefix}/v1/token`));
        }
        // Unified hosts: Use discovery endpoint with account ID
        if (hType === HostType_1.HostType.unifiedHost && this.accountId) {
            const oidcEndpoint = `${this.host}/oidc/accounts/${this.accountId}/.well-known/oauth-authorization-server`;
            return this.fetchOidcDiscovery(oidcEndpoint);
        }
        // Workspace hosts: Use standard discovery endpoint
        if (hType === HostType_1.HostType.workspaceHost) {
            const oidcEndpoint = `${this.host}/oidc/.well-known/oauth-authorization-server`;
            return this.fetchOidcDiscovery(oidcEndpoint);
        }
        return undefined;
    }
    /**
     * Helper method to fetch OIDC endpoints from discovery URL
     */
    async fetchOidcDiscovery(oidcEndpoint) {
        const response = await this.fetch(oidcEndpoint, {});
        if (response.status !== 200) {
            return;
        }
        const json = (await response.json());
        if (!json ||
            typeof json.authorization_endpoint !== "string" ||
            typeof json.token_endpoint !== "string") {
            return;
        }
        return new OidcEndpoints_1.OidcEndpoints(this, new URL(json.authorization_endpoint), new URL(json.token_endpoint));
    }
    fetch(url, options) {
        return (0, fetch_1.fetch)(url, options);
    }
}
exports.Config = Config;
__decorate([
    (0, ConfigAttributes_1.attribute)({ env: "DATABRICKS_HOST" }),
    __metadata("design:type", String)
], Config.prototype, "host", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_METADATA_SERVICE_URL",
        sensitive: true,
    }),
    __metadata("design:type", String)
], Config.prototype, "metadataServiceUrl", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({ env: "DATABRICKS_ACCOUNT_ID" }),
    __metadata("design:type", String)
], Config.prototype, "accountId", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({ env: "DATABRICKS_WORKSPACE_ID" }),
    __metadata("design:type", String)
], Config.prototype, "workspaceId", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({ env: "DATABRICKS_EXPERIMENTAL_IS_UNIFIED_HOST" }),
    __metadata("design:type", Boolean)
], Config.prototype, "experimentalIsUnifiedHost", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_TOKEN",
        auth: "pat",
        sensitive: true,
    }),
    __metadata("design:type", String)
], Config.prototype, "token", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_USERNAME",
        auth: "basic",
    }),
    __metadata("design:type", String)
], Config.prototype, "username", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_PASSWORD",
        auth: "basic",
        sensitive: true,
    }),
    __metadata("design:type", String)
], Config.prototype, "password", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({ env: "DATABRICKS_CONFIG_PROFILE" }),
    __metadata("design:type", String)
], Config.prototype, "profile", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({ env: "DATABRICKS_CONFIG_FILE" }),
    __metadata("design:type", String)
], Config.prototype, "configFile", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_GOOGLE_SERVICE_ACCOUNT",
        auth: "google",
    }),
    __metadata("design:type", String)
], Config.prototype, "googleServiceAccount", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "GOOGLE_CREDENTIALS",
        auth: "google",
        sensitive: true,
    }),
    __metadata("design:type", String)
], Config.prototype, "googleCredentials", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_AZURE_RESOURCE_ID",
        auth: "azure",
    }),
    __metadata("design:type", String)
], Config.prototype, "azureWorkspaceResourceId", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "ARM_USE_MSI",
        auth: "azure",
    }),
    __metadata("design:type", Boolean)
], Config.prototype, "azureUseMsi", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "ARM_CLIENT_SECRET",
        auth: "azure",
        sensitive: true,
    }),
    __metadata("design:type", String)
], Config.prototype, "azureClientSecret", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "ARM_CLIENT_ID",
        auth: "azure",
    }),
    __metadata("design:type", String)
], Config.prototype, "azureClientId", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "ARM_TENANT_ID",
        auth: "azure",
    }),
    __metadata("design:type", String)
], Config.prototype, "azureTenantId", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "ARM_ENVIRONMENT",
    }),
    __metadata("design:type", String)
], Config.prototype, "azureEnvironment", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_AZURE_LOGIN_APP_ID",
        auth: "azure",
    }),
    __metadata("design:type", String)
], Config.prototype, "azureLoginAppId", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_CLIENT_ID",
        auth: "oauth",
    }),
    __metadata("design:type", String)
], Config.prototype, "clientId", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_CLIENT_SECRET",
        auth: "oauth",
        sensitive: true,
    }),
    __metadata("design:type", String)
], Config.prototype, "clientSecret", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_CLI_PATH",
    }),
    __metadata("design:type", String)
], Config.prototype, "databricksCliPath", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_AUTH_TYPE",
    }),
    __metadata("design:type", String)
], Config.prototype, "authType", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_CLUSTER_ID",
    }),
    __metadata("design:type", String)
], Config.prototype, "clusterId", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_SERVERLESS_COMPUTE_ID",
    }),
    __metadata("design:type", String)
], Config.prototype, "serverlessComputeId", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({
        env: "DATABRICKS_WAREHOUSE_ID",
    }),
    __metadata("design:type", String)
], Config.prototype, "warehouseId", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({}),
    __metadata("design:type", Boolean)
], Config.prototype, "skipVerify", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({}),
    __metadata("design:type", Number)
], Config.prototype, "httpTimeoutSeconds", void 0);
__decorate([
    (0, ConfigAttributes_1.attribute)({}),
    __metadata("design:type", Number)
], Config.prototype, "retryTimeoutSeconds", void 0);
function isAnyAuthConfigured(cfg) {
    for (const attr of cfg.attributes) {
        if (attr.auth === undefined) {
            // If auth is undefined, it means that the attribute is not
            // related to a specific auth method. We assume that this means no
            // auth is configured.
            continue;
        }
        if (!attr.isUndefined()) {
            // This means that the attribute is related to a specific auth
            // and it is configured (through a previous loader in the chain).
            return true;
        }
    }
    // If the auth type is set, we can skip profile resolution.
    // For example, to force "azure-cli", only the host and the auth type will be set.
    return cfg.authType !== undefined;
}
