// @databricks/sdk-experimental/dist/config/DatabricksCliCredentials.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const DatabricksCliCredentials_1 = require("./DatabricksCliCredentials");
const Config_1 = require("./Config");
const assert = __importStar(require("node:assert"));
const execUtils = __importStar(require("./execUtils"));
describe(__filename, function () {
    this.timeout(10000);
    let execFileBackup;
    let capturedArgs = [];
    beforeEach(() => {
        execFileBackup = execUtils.execFile;
        capturedArgs = [];
        // Mock execFile to capture arguments
        execUtils.execFile = async (cmd, args) => {
            if (args[0] === "version") {
                return { stdout: "0.200.0", stderr: "" };
            }
            if (args[0] === "auth" && args[1] === "token") {
                capturedArgs = args;
                return {
                    stdout: JSON.stringify({
                        access_token: "test-token",
                        expiry: new Date(Date.now() + 3600000).toISOString(),
                    }),
                    stderr: "",
                };
            }
            throw new Error(`Unexpected command: ${cmd} ${args.join(" ")}`);
        };
    });
    afterEach(() => {
        execUtils.execFile = execFileBackup;
    });
    it("should build CLI command for workspace host", async function () {
        const config = new Config_1.Config({
            host: "https://dbc-12345.cloud.databricks.com",
        });
        await config.ensureResolved();
        const creds = new DatabricksCliCredentials_1.DatabricksCliCredentials();
        await creds.configure(config);
        assert.deepStrictEqual(capturedArgs, [
            "auth",
            "token",
            "--host",
            "https://dbc-12345.cloud.databricks.com",
        ]);
    });
    it("should build CLI command for account host", async function () {
        const config = new Config_1.Config({
            host: "https://accounts.cloud.databricks.com",
            accountId: "test-account-id",
        });
        await config.ensureResolved();
        const creds = new DatabricksCliCredentials_1.DatabricksCliCredentials();
        await creds.configure(config);
        assert.deepStrictEqual(capturedArgs, [
            "auth",
            "token",
            "--host",
            "https://accounts.cloud.databricks.com",
            "--account-id",
            "test-account-id",
        ]);
    });
    it("should build CLI command for unified host with accountId and workspaceId", async function () {
        const config = new Config_1.Config({
            host: "https://company.databricks.com",
            accountId: "test-account-id",
            workspaceId: "test-workspace-id",
            experimentalIsUnifiedHost: true,
        });
        await config.ensureResolved();
        const creds = new DatabricksCliCredentials_1.DatabricksCliCredentials();
        await creds.configure(config);
        assert.deepStrictEqual(capturedArgs, [
            "auth",
            "token",
            "--host",
            "https://company.databricks.com",
            "--experimental-is-unified-host",
            "--account-id",
            "test-account-id",
        ]);
    });
    it("should build CLI command for unified host with accountId only", async function () {
        const config = new Config_1.Config({
            host: "https://company.databricks.com",
            accountId: "test-account-id",
            experimentalIsUnifiedHost: true,
        });
        await config.ensureResolved();
        const creds = new DatabricksCliCredentials_1.DatabricksCliCredentials();
        await creds.configure(config);
        assert.deepStrictEqual(capturedArgs, [
            "auth",
            "token",
            "--host",
            "https://company.databricks.com",
            "--experimental-is-unified-host",
            "--account-id",
            "test-account-id",
        ]);
    });
    it("should build CLI command for unified host without accountId", async function () {
        const config = new Config_1.Config({
            host: "https://company.databricks.com",
            experimentalIsUnifiedHost: true,
        });
        await config.ensureResolved();
        const creds = new DatabricksCliCredentials_1.DatabricksCliCredentials();
        await creds.configure(config);
        assert.deepStrictEqual(capturedArgs, [
            "auth",
            "token",
            "--host",
            "https://company.databricks.com",
            "--experimental-is-unified-host",
        ]);
    });
    it("should not add unified host flag when experimentalIsUnifiedHost is false", async function () {
        const config = new Config_1.Config({
            host: "https://dbc-12345.cloud.databricks.com",
            accountId: "test-account-id",
            experimentalIsUnifiedHost: false,
        });
        await config.ensureResolved();
        const creds = new DatabricksCliCredentials_1.DatabricksCliCredentials();
        await creds.configure(config);
        // Should be treated as workspace host, no account-id since it's not an account host
        assert.deepStrictEqual(capturedArgs, [
            "auth",
            "token",
            "--host",
            "https://dbc-12345.cloud.databricks.com",
        ]);
    });
});
