// @databricks/sdk-experimental/dist/config/GoogleCredentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleCredentials = void 0;
const Config_1 = require("./Config");
const GoogleDefaultCredentials_1 = require("./GoogleDefaultCredentials");
const google_auth_library_1 = require("google-auth-library");
class GoogleCredentials {
    constructor() {
        this.name = "google-id";
    }
    async configure(config) {
        if (!config.googleServiceAccount || !config.isGcp() || !config.host) {
            return;
        }
        const host = config.host;
        // Acquire source credentials:
        const auth = new google_auth_library_1.GoogleAuth();
        const client = await auth.getClient();
        // Create the impersonated credential.
        const targetCredentials = new google_auth_library_1.Impersonated({
            sourceClient: client,
            targetPrincipal: config.googleServiceAccount,
            targetScopes: [],
        });
        // Set the impersonated credential, target audience and token options.
        const gcpImpersonatedCredentials = new google_auth_library_1.Impersonated({
            sourceClient: targetCredentials,
            targetPrincipal: config.googleServiceAccount,
            targetScopes: GoogleDefaultCredentials_1.gcpScopes,
        });
        const isAccount = config.isAccountClient();
        return async (headers) => {
            const idCreds = targetCredentials.fetchIdToken(host, {
                includeEmail: true,
            });
            headers.set("authorization", `Bearer ${idCreds}`);
            if (isAccount) {
                const gcpToken = await gcpImpersonatedCredentials.getAccessToken();
                if (!gcpToken.token) {
                    throw new Config_1.ConfigError("failed to get GCP access token", config);
                }
                headers.set("x-databricks-gcp-sa-access-token", gcpToken.token);
            }
        };
    }
}
exports.GoogleCredentials = GoogleCredentials;
