// @databricks/sdk-experimental/dist/config/Token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Token = void 0;
exports.refreshableTokenSource = refreshableTokenSource;
exports.refreshableTokenVisitor = refreshableTokenVisitor;
// Most oauth libraries refresh a token 10 seconds before it expires.
// Azure Databricks rejects tokens that expire in 30 seconds or less.
// We combine these and refresh the token 40 seconds before it expires.
const expiryDelta = 40 * 1000;
class Token {
    constructor(options) {
        this.refreshToken = options.refreshToken;
        this.accessToken = options.accessToken;
        this.expiry = options.expiry || 0;
    }
    isValid() {
        return this.expiry === 0 || this.expiry > Date.now() + expiryDelta;
    }
}
exports.Token = Token;
function refreshableTokenSource(source) {
    let token;
    return async () => {
        if (!token || !token.isValid()) {
            token = await source();
        }
        return token;
    };
}
function refreshableTokenVisitor(source) {
    const ts = refreshableTokenSource(source);
    return async (headers) => {
        const token = await ts();
        headers.set("Authorization", `Bearer ${token.accessToken}`);
    };
}
