// @databricks/sdk-experimental/dist/apis/iam/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceAssignmentService = exports.WorkspaceAssignmentError = exports.WorkspaceAssignmentRetriableError = exports.UsersV2Service = exports.UsersV2Error = exports.UsersV2RetriableError = exports.ServicePrincipalsV2Service = exports.ServicePrincipalsV2Error = exports.ServicePrincipalsV2RetriableError = exports.PermissionsService = exports.PermissionsError = exports.PermissionsRetriableError = exports.PermissionMigrationService = exports.PermissionMigrationError = exports.PermissionMigrationRetriableError = exports.GroupsV2Service = exports.GroupsV2Error = exports.GroupsV2RetriableError = exports.CurrentUserService = exports.CurrentUserError = exports.CurrentUserRetriableError = exports.AccountUsersV2Service = exports.AccountUsersV2Error = exports.AccountUsersV2RetriableError = exports.AccountServicePrincipalsV2Service = exports.AccountServicePrincipalsV2Error = exports.AccountServicePrincipalsV2RetriableError = exports.AccountGroupsV2Service = exports.AccountGroupsV2Error = exports.AccountGroupsV2RetriableError = exports.AccountAccessControlProxyService = exports.AccountAccessControlProxyError = exports.AccountAccessControlProxyRetriableError = exports.AccountAccessControlService = exports.AccountAccessControlError = exports.AccountAccessControlRetriableError = exports.AccessControlService = exports.AccessControlError = exports.AccessControlRetriableError = void 0;
/**
 * These APIs allow you to manage Access Control, Account Access Control, Account Access Control Proxy, Account Groups V2, Account Service Principals V2, Account Users V2, Current User, Groups V2, Permission Migration, Permissions, Service Principals V2, Users V2, Workspace Assignment, etc.
 */
const config_1 = require("../../config");
const HostType_1 = require("../../config/HostType");
const iam = __importStar(require("./model"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
class AccessControlRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("AccessControl", method, message);
    }
}
exports.AccessControlRetriableError = AccessControlRetriableError;
class AccessControlError extends apiError_1.ApiError {
    constructor(method, message) {
        super("AccessControl", method, message);
    }
}
exports.AccessControlError = AccessControlError;
/**
 * Rule based Access Control for Databricks Resources.
 */
class AccessControlService {
    constructor(client) {
        this.client = client;
    }
    async _checkPolicy(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("actor")) {
            query["actor"] = request.actor;
        }
        if (request.hasOwnProperty("authz_identity")) {
            query["authz_identity"] =
                request.authz_identity;
        }
        if (request.hasOwnProperty("consistency_token")) {
            query["consistency_token"] =
                request.consistency_token;
        }
        if (request.hasOwnProperty("permission")) {
            query["permission"] =
                request.permission;
        }
        if (request.hasOwnProperty("resource")) {
            query["resource"] =
                request.resource;
        }
        if (request.hasOwnProperty("resource_info")) {
            query["resource_info"] =
                request.resource_info;
        }
        const path = "/api/2.0/access-control/check-policy-v2";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Check access policy to a resource.
     */
    async checkPolicy(request, context) {
        return await this._checkPolicy(request, context);
    }
}
exports.AccessControlService = AccessControlService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccessControlService.prototype, "_checkPolicy", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccessControlService.prototype, "checkPolicy", null);
class AccountAccessControlRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("AccountAccessControl", method, message);
    }
}
exports.AccountAccessControlRetriableError = AccountAccessControlRetriableError;
class AccountAccessControlError extends apiError_1.ApiError {
    constructor(method, message) {
        super("AccountAccessControl", method, message);
    }
}
exports.AccountAccessControlError = AccountAccessControlError;
/**
 * These APIs manage access rules on resources in an account. Currently, only
 * grant rules are supported. A grant rule specifies a role assigned to a set of
 * principals. A list of rules attached to a resource is called a rule set.
 */
class AccountAccessControlService {
    constructor(client) {
        this.client = client;
    }
    async _getAssignableRolesForResource(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("resource")) {
            query["resource"] =
                request.resource;
        }
        const path = `/api/2.0/preview/accounts/${config.accountId}/access-control/assignable-roles`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets all the roles that can be granted on an account level resource. A
     * role is grantable if the rule set on the resource can contain an access
     * rule of the role.
     */
    async getAssignableRolesForResource(request, context) {
        return await this._getAssignableRolesForResource(request, context);
    }
    async _getRuleSet(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("etag")) {
            query["etag"] = request.etag;
        }
        if (request.hasOwnProperty("name")) {
            query["name"] = request.name;
        }
        const path = `/api/2.0/preview/accounts/${config.accountId}/access-control/rule-sets`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a rule set by its name. A rule set is always attached to a resource
     * and contains a list of access rules on the said resource. Currently only a
     * default rule set for each resource is supported.
     */
    async getRuleSet(request, context) {
        return await this._getRuleSet(request, context);
    }
    async _updateRuleSet(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("rule_set")) {
            body["rule_set"] = request.rule_set;
        }
        const query = {};
        const path = `/api/2.0/preview/accounts/${config.accountId}/access-control/rule-sets`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Replace the rules of a rule set. First, use get to read the current
     * version of the rule set before modifying it. This pattern helps prevent
     * conflicts between concurrent updates.
     */
    async updateRuleSet(request, context) {
        return await this._updateRuleSet(request, context);
    }
}
exports.AccountAccessControlService = AccountAccessControlService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountAccessControlService.prototype, "_getAssignableRolesForResource", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountAccessControlService.prototype, "getAssignableRolesForResource", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountAccessControlService.prototype, "_getRuleSet", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountAccessControlService.prototype, "getRuleSet", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountAccessControlService.prototype, "_updateRuleSet", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountAccessControlService.prototype, "updateRuleSet", null);
class AccountAccessControlProxyRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("AccountAccessControlProxy", method, message);
    }
}
exports.AccountAccessControlProxyRetriableError = AccountAccessControlProxyRetriableError;
class AccountAccessControlProxyError extends apiError_1.ApiError {
    constructor(method, message) {
        super("AccountAccessControlProxy", method, message);
    }
}
exports.AccountAccessControlProxyError = AccountAccessControlProxyError;
/**
 * These APIs manage access rules on resources in an account. Currently, only
 * grant rules are supported. A grant rule specifies a role assigned to a set of
 * principals. A list of rules attached to a resource is called a rule set. A
 * workspace must belong to an account for these APIs to work
 */
class AccountAccessControlProxyService {
    constructor(client) {
        this.client = client;
    }
    async _getAssignableRolesForResource(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("resource")) {
            query["resource"] =
                request.resource;
        }
        const path = "/api/2.0/preview/accounts/access-control/assignable-roles";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets all the roles that can be granted on an account level resource. A
     * role is grantable if the rule set on the resource can contain an access
     * rule of the role.
     */
    async getAssignableRolesForResource(request, context) {
        return await this._getAssignableRolesForResource(request, context);
    }
    async _getRuleSet(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("etag")) {
            query["etag"] = request.etag;
        }
        if (request.hasOwnProperty("name")) {
            query["name"] = request.name;
        }
        const path = "/api/2.0/preview/accounts/access-control/rule-sets";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a rule set by its name. A rule set is always attached to a resource
     * and contains a list of access rules on the said resource. Currently only a
     * default rule set for each resource is supported.
     */
    async getRuleSet(request, context) {
        return await this._getRuleSet(request, context);
    }
    async _updateRuleSet(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("rule_set")) {
            body["rule_set"] = request.rule_set;
        }
        const query = {};
        const path = "/api/2.0/preview/accounts/access-control/rule-sets";
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Replace the rules of a rule set. First, use get to read the current
     * version of the rule set before modifying it. This pattern helps prevent
     * conflicts between concurrent updates.
     */
    async updateRuleSet(request, context) {
        return await this._updateRuleSet(request, context);
    }
}
exports.AccountAccessControlProxyService = AccountAccessControlProxyService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountAccessControlProxyService.prototype, "_getAssignableRolesForResource", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountAccessControlProxyService.prototype, "getAssignableRolesForResource", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountAccessControlProxyService.prototype, "_getRuleSet", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountAccessControlProxyService.prototype, "getRuleSet", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountAccessControlProxyService.prototype, "_updateRuleSet", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountAccessControlProxyService.prototype, "updateRuleSet", null);
class AccountGroupsV2RetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("AccountGroupsV2", method, message);
    }
}
exports.AccountGroupsV2RetriableError = AccountGroupsV2RetriableError;
class AccountGroupsV2Error extends apiError_1.ApiError {
    constructor(method, message) {
        super("AccountGroupsV2", method, message);
    }
}
exports.AccountGroupsV2Error = AccountGroupsV2Error;
/**
 * Groups simplify identity management, making it easier to assign access to
 * Databricks account, data, and other securable objects.
 *
 * It is best practice to assign access to workspaces and access-control policies
 * in Unity Catalog to groups, instead of to users individually. All Databricks
 * account identities can be assigned as members of groups, and members inherit
 * permissions that are assigned to their group.
 */
class AccountGroupsV2Service {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("displayName")) {
            body["displayName"] = request.displayName;
        }
        if (request.hasOwnProperty("externalId")) {
            body["externalId"] = request.externalId;
        }
        if (request.hasOwnProperty("id")) {
            body["id"] = request.id;
        }
        if (request.hasOwnProperty("members")) {
            body["members"] = request.members;
        }
        if (request.hasOwnProperty("meta")) {
            body["meta"] = request.meta;
        }
        if (request.hasOwnProperty("roles")) {
            body["roles"] = request.roles;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/Groups`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a group in the Databricks account with a unique name, using the
     * supplied group details.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/Groups/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a group from the Databricks account.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/Groups/${request.id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the information for a specific group in the Databricks account.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("attributes")) {
            query["attributes"] =
                request.attributes;
        }
        if (request.hasOwnProperty("count")) {
            query["count"] = request.count;
        }
        if (request.hasOwnProperty("excludedAttributes")) {
            query["excludedAttributes"] =
                request.excludedAttributes;
        }
        if (request.hasOwnProperty("filter")) {
            query["filter"] = request.filter;
        }
        if (request.hasOwnProperty("sortBy")) {
            query["sortBy"] = request.sortBy;
        }
        if (request.hasOwnProperty("sortOrder")) {
            query["sortOrder"] =
                request.sortOrder;
        }
        if (request.hasOwnProperty("startIndex")) {
            query["startIndex"] =
                request.startIndex;
        }
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/Groups`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets all details of the groups associated with the Databricks account. As
     * of 08/22/2025, this endpoint will no longer return members. Instead,
     * members should be retrieved by iterating through `Get group details`.
     * Existing accounts that rely on this attribute will not be impacted and
     * will continue receiving member data as before.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.Resources || response.Resources.length === 0) {
                break;
            }
            for (const v of response.Resources) {
                yield v;
            }
        }
    }
    async _patch(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ "Content-Type": "application/json" });
        const body = {};
        if (request.hasOwnProperty("Operations")) {
            body["Operations"] = request.Operations;
        }
        if (request.hasOwnProperty("schemas")) {
            body["schemas"] = request.schemas;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/Groups/${request.id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Partially updates the details of a group.
     */
    async patch(request, context) {
        return await this._patch(request, context);
    }
    async _update(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("displayName")) {
            body["displayName"] = request.displayName;
        }
        if (request.hasOwnProperty("externalId")) {
            body["externalId"] = request.externalId;
        }
        if (request.hasOwnProperty("members")) {
            body["members"] = request.members;
        }
        if (request.hasOwnProperty("meta")) {
            body["meta"] = request.meta;
        }
        if (request.hasOwnProperty("roles")) {
            body["roles"] = request.roles;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/Groups/${request.id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the details of a group by replacing the entire group entity.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.AccountGroupsV2Service = AccountGroupsV2Service;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountGroupsV2Service.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountGroupsV2Service.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountGroupsV2Service.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountGroupsV2Service.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountGroupsV2Service.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountGroupsV2Service.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountGroupsV2Service.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], AccountGroupsV2Service.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountGroupsV2Service.prototype, "_patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountGroupsV2Service.prototype, "patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountGroupsV2Service.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountGroupsV2Service.prototype, "update", null);
class AccountServicePrincipalsV2RetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("AccountServicePrincipalsV2", method, message);
    }
}
exports.AccountServicePrincipalsV2RetriableError = AccountServicePrincipalsV2RetriableError;
class AccountServicePrincipalsV2Error extends apiError_1.ApiError {
    constructor(method, message) {
        super("AccountServicePrincipalsV2", method, message);
    }
}
exports.AccountServicePrincipalsV2Error = AccountServicePrincipalsV2Error;
/**
 * Identities for use with jobs, automated tools, and systems such as scripts,
 * apps, and CI/CD platforms. Databricks recommends creating service principals
 * to run production jobs or modify production data. If all processes that act on
 * production data run with service principals, interactive users do not need any
 * write, delete, or modify privileges in production. This eliminates the risk of
 * a user overwriting production data by accident.
 */
class AccountServicePrincipalsV2Service {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("active")) {
            body["active"] = request.active;
        }
        if (request.hasOwnProperty("applicationId")) {
            body["applicationId"] = request.applicationId;
        }
        if (request.hasOwnProperty("displayName")) {
            body["displayName"] = request.displayName;
        }
        if (request.hasOwnProperty("externalId")) {
            body["externalId"] = request.externalId;
        }
        if (request.hasOwnProperty("id")) {
            body["id"] = request.id;
        }
        if (request.hasOwnProperty("roles")) {
            body["roles"] = request.roles;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/ServicePrincipals`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new service principal in the Databricks account.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/ServicePrincipals/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a single service principal in the Databricks account.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/ServicePrincipals/${request.id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the details for a single service principal define in the Databricks
     * account.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("attributes")) {
            query["attributes"] =
                request.attributes;
        }
        if (request.hasOwnProperty("count")) {
            query["count"] = request.count;
        }
        if (request.hasOwnProperty("excludedAttributes")) {
            query["excludedAttributes"] =
                request.excludedAttributes;
        }
        if (request.hasOwnProperty("filter")) {
            query["filter"] = request.filter;
        }
        if (request.hasOwnProperty("sortBy")) {
            query["sortBy"] = request.sortBy;
        }
        if (request.hasOwnProperty("sortOrder")) {
            query["sortOrder"] =
                request.sortOrder;
        }
        if (request.hasOwnProperty("startIndex")) {
            query["startIndex"] =
                request.startIndex;
        }
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/ServicePrincipals`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the set of service principals associated with a Databricks account.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.Resources || response.Resources.length === 0) {
                break;
            }
            for (const v of response.Resources) {
                yield v;
            }
        }
    }
    async _patch(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("Operations")) {
            body["Operations"] = request.Operations;
        }
        if (request.hasOwnProperty("schemas")) {
            body["schemas"] = request.schemas;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/ServicePrincipals/${request.id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Partially updates the details of a single service principal in the
     * Databricks account.
     */
    async patch(request, context) {
        return await this._patch(request, context);
    }
    async _update(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("active")) {
            body["active"] = request.active;
        }
        if (request.hasOwnProperty("applicationId")) {
            body["applicationId"] = request.applicationId;
        }
        if (request.hasOwnProperty("displayName")) {
            body["displayName"] = request.displayName;
        }
        if (request.hasOwnProperty("externalId")) {
            body["externalId"] = request.externalId;
        }
        if (request.hasOwnProperty("roles")) {
            body["roles"] = request.roles;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/ServicePrincipals/${request.id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the details of a single service principal.
     *
     * This action replaces the existing service principal with the same name.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.AccountServicePrincipalsV2Service = AccountServicePrincipalsV2Service;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountServicePrincipalsV2Service.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountServicePrincipalsV2Service.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountServicePrincipalsV2Service.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountServicePrincipalsV2Service.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountServicePrincipalsV2Service.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountServicePrincipalsV2Service.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountServicePrincipalsV2Service.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], AccountServicePrincipalsV2Service.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountServicePrincipalsV2Service.prototype, "_patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountServicePrincipalsV2Service.prototype, "patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountServicePrincipalsV2Service.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountServicePrincipalsV2Service.prototype, "update", null);
class AccountUsersV2RetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("AccountUsersV2", method, message);
    }
}
exports.AccountUsersV2RetriableError = AccountUsersV2RetriableError;
class AccountUsersV2Error extends apiError_1.ApiError {
    constructor(method, message) {
        super("AccountUsersV2", method, message);
    }
}
exports.AccountUsersV2Error = AccountUsersV2Error;
/**
 * User identities recognized by Databricks and represented by email addresses.
 *
 * Databricks recommends using SCIM provisioning to sync users and groups
 * automatically from your identity provider to your Databricks account. SCIM
 * streamlines onboarding a new employee or team by using your identity provider
 * to create users and groups in Databricks account and give them the proper
 * level of access. When a user leaves your organization or no longer needs
 * access to Databricks account, admins can terminate the user in your identity
 * provider and that user’s account will also be removed from Databricks
 * account. This ensures a consistent offboarding process and prevents
 * unauthorized users from accessing sensitive data.
 */
class AccountUsersV2Service {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("active")) {
            body["active"] = request.active;
        }
        if (request.hasOwnProperty("displayName")) {
            body["displayName"] = request.displayName;
        }
        if (request.hasOwnProperty("emails")) {
            body["emails"] = request.emails;
        }
        if (request.hasOwnProperty("externalId")) {
            body["externalId"] = request.externalId;
        }
        if (request.hasOwnProperty("id")) {
            body["id"] = request.id;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("roles")) {
            body["roles"] = request.roles;
        }
        if (request.hasOwnProperty("userName")) {
            body["userName"] = request.userName;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/Users`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new user in the Databricks account. This new user will also be
     * added to the Databricks account.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/Users/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a user. Deleting a user from a Databricks account also removes
     * objects associated with the user.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("attributes")) {
            query["attributes"] =
                request.attributes;
        }
        if (request.hasOwnProperty("count")) {
            query["count"] = request.count;
        }
        if (request.hasOwnProperty("excludedAttributes")) {
            query["excludedAttributes"] =
                request.excludedAttributes;
        }
        if (request.hasOwnProperty("filter")) {
            query["filter"] = request.filter;
        }
        if (request.hasOwnProperty("sortBy")) {
            query["sortBy"] = request.sortBy;
        }
        if (request.hasOwnProperty("sortOrder")) {
            query["sortOrder"] =
                request.sortOrder;
        }
        if (request.hasOwnProperty("startIndex")) {
            query["startIndex"] =
                request.startIndex;
        }
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/Users/${request.id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets information for a specific user in Databricks account.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("attributes")) {
            query["attributes"] =
                request.attributes;
        }
        if (request.hasOwnProperty("count")) {
            query["count"] = request.count;
        }
        if (request.hasOwnProperty("excludedAttributes")) {
            query["excludedAttributes"] =
                request.excludedAttributes;
        }
        if (request.hasOwnProperty("filter")) {
            query["filter"] = request.filter;
        }
        if (request.hasOwnProperty("sortBy")) {
            query["sortBy"] = request.sortBy;
        }
        if (request.hasOwnProperty("sortOrder")) {
            query["sortOrder"] =
                request.sortOrder;
        }
        if (request.hasOwnProperty("startIndex")) {
            query["startIndex"] =
                request.startIndex;
        }
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/Users`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets details for all the users associated with a Databricks account.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.Resources || response.Resources.length === 0) {
                break;
            }
            for (const v of response.Resources) {
                yield v;
            }
        }
    }
    async _patch(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("Operations")) {
            body["Operations"] = request.Operations;
        }
        if (request.hasOwnProperty("schemas")) {
            body["schemas"] = request.schemas;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/Users/${request.id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Partially updates a user resource by applying the supplied operations on
     * specific user attributes.
     */
    async patch(request, context) {
        return await this._patch(request, context);
    }
    async _update(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("active")) {
            body["active"] = request.active;
        }
        if (request.hasOwnProperty("displayName")) {
            body["displayName"] = request.displayName;
        }
        if (request.hasOwnProperty("emails")) {
            body["emails"] = request.emails;
        }
        if (request.hasOwnProperty("externalId")) {
            body["externalId"] = request.externalId;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("roles")) {
            body["roles"] = request.roles;
        }
        if (request.hasOwnProperty("userName")) {
            body["userName"] = request.userName;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/scim/v2/Users/${request.id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Replaces a user's information with the data supplied in request.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.AccountUsersV2Service = AccountUsersV2Service;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountUsersV2Service.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountUsersV2Service.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountUsersV2Service.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountUsersV2Service.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountUsersV2Service.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountUsersV2Service.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountUsersV2Service.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], AccountUsersV2Service.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountUsersV2Service.prototype, "_patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountUsersV2Service.prototype, "patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountUsersV2Service.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountUsersV2Service.prototype, "update", null);
class CurrentUserRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("CurrentUser", method, message);
    }
}
exports.CurrentUserRetriableError = CurrentUserRetriableError;
class CurrentUserError extends apiError_1.ApiError {
    constructor(method, message) {
        super("CurrentUser", method, message);
    }
}
exports.CurrentUserError = CurrentUserError;
/**
 * This API allows retrieving information about currently authenticated user or
 * service principal.
 */
class CurrentUserService {
    constructor(client) {
        this.client = client;
    }
    async _me(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/preview/scim/v2/Me";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get details about the current method caller's identity.
     */
    async me(context) {
        return await this._me(context);
    }
}
exports.CurrentUserService = CurrentUserService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], CurrentUserService.prototype, "_me", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], CurrentUserService.prototype, "me", null);
class GroupsV2RetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("GroupsV2", method, message);
    }
}
exports.GroupsV2RetriableError = GroupsV2RetriableError;
class GroupsV2Error extends apiError_1.ApiError {
    constructor(method, message) {
        super("GroupsV2", method, message);
    }
}
exports.GroupsV2Error = GroupsV2Error;
/**
 * Groups simplify identity management, making it easier to assign access to
 * Databricks workspace, data, and other securable objects.
 *
 * It is best practice to assign access to workspaces and access-control policies
 * in Unity Catalog to groups, instead of to users individually. All Databricks
 * workspace identities can be assigned as members of groups, and members inherit
 * permissions that are assigned to their group.
 */
class GroupsV2Service {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("displayName")) {
            body["displayName"] = request.displayName;
        }
        if (request.hasOwnProperty("entitlements")) {
            body["entitlements"] = request.entitlements;
        }
        if (request.hasOwnProperty("externalId")) {
            body["externalId"] = request.externalId;
        }
        if (request.hasOwnProperty("groups")) {
            body["groups"] = request.groups;
        }
        if (request.hasOwnProperty("id")) {
            body["id"] = request.id;
        }
        if (request.hasOwnProperty("members")) {
            body["members"] = request.members;
        }
        if (request.hasOwnProperty("meta")) {
            body["meta"] = request.meta;
        }
        if (request.hasOwnProperty("roles")) {
            body["roles"] = request.roles;
        }
        if (request.hasOwnProperty("schemas")) {
            body["schemas"] = request.schemas;
        }
        const query = {};
        const path = "/api/2.0/preview/scim/v2/Groups";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a group in the Databricks workspace with a unique name, using the
     * supplied group details.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/scim/v2/Groups/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a group from the Databricks workspace.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/scim/v2/Groups/${request.id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the information for a specific group in the Databricks workspace.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("attributes")) {
            query["attributes"] =
                request.attributes;
        }
        if (request.hasOwnProperty("count")) {
            query["count"] = request.count;
        }
        if (request.hasOwnProperty("excludedAttributes")) {
            query["excludedAttributes"] =
                request.excludedAttributes;
        }
        if (request.hasOwnProperty("filter")) {
            query["filter"] = request.filter;
        }
        if (request.hasOwnProperty("sortBy")) {
            query["sortBy"] = request.sortBy;
        }
        if (request.hasOwnProperty("sortOrder")) {
            query["sortOrder"] =
                request.sortOrder;
        }
        if (request.hasOwnProperty("startIndex")) {
            query["startIndex"] =
                request.startIndex;
        }
        const path = "/api/2.0/preview/scim/v2/Groups";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets all details of the groups associated with the Databricks workspace.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.Resources || response.Resources.length === 0) {
                break;
            }
            for (const v of response.Resources) {
                yield v;
            }
        }
    }
    async _patch(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("Operations")) {
            body["Operations"] = request.Operations;
        }
        if (request.hasOwnProperty("schemas")) {
            body["schemas"] = request.schemas;
        }
        const query = {};
        const path = `/api/2.0/preview/scim/v2/Groups/${request.id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Partially updates the details of a group.
     */
    async patch(request, context) {
        return await this._patch(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("displayName")) {
            body["displayName"] = request.displayName;
        }
        if (request.hasOwnProperty("entitlements")) {
            body["entitlements"] = request.entitlements;
        }
        if (request.hasOwnProperty("externalId")) {
            body["externalId"] = request.externalId;
        }
        if (request.hasOwnProperty("groups")) {
            body["groups"] = request.groups;
        }
        if (request.hasOwnProperty("members")) {
            body["members"] = request.members;
        }
        if (request.hasOwnProperty("meta")) {
            body["meta"] = request.meta;
        }
        if (request.hasOwnProperty("roles")) {
            body["roles"] = request.roles;
        }
        if (request.hasOwnProperty("schemas")) {
            body["schemas"] = request.schemas;
        }
        const query = {};
        const path = `/api/2.0/preview/scim/v2/Groups/${request.id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the details of a group by replacing the entire group entity.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.GroupsV2Service = GroupsV2Service;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GroupsV2Service.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GroupsV2Service.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GroupsV2Service.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GroupsV2Service.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GroupsV2Service.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GroupsV2Service.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GroupsV2Service.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], GroupsV2Service.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GroupsV2Service.prototype, "_patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GroupsV2Service.prototype, "patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GroupsV2Service.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GroupsV2Service.prototype, "update", null);
class PermissionMigrationRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("PermissionMigration", method, message);
    }
}
exports.PermissionMigrationRetriableError = PermissionMigrationRetriableError;
class PermissionMigrationError extends apiError_1.ApiError {
    constructor(method, message) {
        super("PermissionMigration", method, message);
    }
}
exports.PermissionMigrationError = PermissionMigrationError;
/**
 * APIs for migrating acl permissions, used only by the ucx tool:
 * https://github.com/databrickslabs/ucx
 */
class PermissionMigrationService {
    constructor(client) {
        this.client = client;
    }
    async _migratePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("from_workspace_group_name")) {
            body["from_workspace_group_name"] =
                request.from_workspace_group_name;
        }
        if (request.hasOwnProperty("size")) {
            body["size"] = request.size;
        }
        if (request.hasOwnProperty("to_account_group_name")) {
            body["to_account_group_name"] = request.to_account_group_name;
        }
        if (request.hasOwnProperty("workspace_id")) {
            body["workspace_id"] = request.workspace_id;
        }
        const query = {};
        const path = "/api/2.0/permissionmigration";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Migrate Permissions.
     */
    async migratePermissions(request, context) {
        return await this._migratePermissions(request, context);
    }
}
exports.PermissionMigrationService = PermissionMigrationService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PermissionMigrationService.prototype, "_migratePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PermissionMigrationService.prototype, "migratePermissions", null);
class PermissionsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Permissions", method, message);
    }
}
exports.PermissionsRetriableError = PermissionsRetriableError;
class PermissionsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Permissions", method, message);
    }
}
exports.PermissionsError = PermissionsError;
/**
 * Permissions API are used to create read, write, edit, update and manage access
 * for various users on different objects and endpoints. * **[Apps
 * permissions](:service:apps)** — Manage which users can manage or use apps. *
 * **[Cluster permissions](:service:clusters)** — Manage which users can
 * manage, restart, or attach to clusters. * **[Cluster policy
 * permissions](:service:clusterpolicies)** — Manage which users can use
 * cluster policies. * **[Delta Live Tables pipeline
 * permissions](:service:pipelines)** — Manage which users can view, manage,
 * run, cancel, or own a Delta Live Tables pipeline. * **[Job
 * permissions](:service:jobs)** — Manage which users can view, manage,
 * trigger, cancel, or own a job. * **[MLflow experiment
 * permissions](:service:experiments)** — Manage which users can read, edit, or
 * manage MLflow experiments. * **[MLflow registered model
 * permissions](:service:modelregistry)** — Manage which users can read, edit,
 * or manage MLflow registered models. * **[Instance Pool
 * permissions](:service:instancepools)** — Manage which users can manage or
 * attach to pools. * **[Repo permissions](repos)** — Manage which users can
 * read, run, edit, or manage a repo. * **[Serving endpoint
 * permissions](:service:servingendpoints)** — Manage which users can view,
 * query, or manage a serving endpoint. * **[SQL warehouse
 * permissions](:service:warehouses)** — Manage which users can use or manage
 * SQL warehouses. * **[Token permissions](:service:tokenmanagement)** — Manage
 * which users can create or use tokens. * **[Workspace object
 * permissions](:service:workspace)** — Manage which users can read, run, edit,
 * or manage alerts, dbsql-dashboards, directories, files, notebooks and queries.
 * For the mapping of the required permissions for specific actions or abilities
 * and other important information, see [Access Control]. Note that to manage
 * access control on service principals, use **[Account Access Control
 * Proxy](:service:accountaccesscontrolproxy)**.
 *
 * [Access Control]: https://docs.databricks.com/security/auth-authz/access-control/index.html
 */
class PermissionsService {
    constructor(client) {
        this.client = client;
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/${request.request_object_type}/${request.request_object_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of an object. Objects can inherit permissions from
     * their parent objects or root object.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _getPermissionLevels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/${request.request_object_type}/${request.request_object_id}/permissionLevels`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(request, context) {
        return await this._getPermissionLevels(request, context);
    }
    async _set(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/${request.request_object_type}/${request.request_object_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their parent objects or root object.
     */
    async set(request, context) {
        return await this._set(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/${request.request_object_type}/${request.request_object_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on an object. Objects can inherit permissions from
     * their parent objects or root object.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.PermissionsService = PermissionsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PermissionsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PermissionsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PermissionsService.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PermissionsService.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PermissionsService.prototype, "_set", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PermissionsService.prototype, "set", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PermissionsService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PermissionsService.prototype, "update", null);
class ServicePrincipalsV2RetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("ServicePrincipalsV2", method, message);
    }
}
exports.ServicePrincipalsV2RetriableError = ServicePrincipalsV2RetriableError;
class ServicePrincipalsV2Error extends apiError_1.ApiError {
    constructor(method, message) {
        super("ServicePrincipalsV2", method, message);
    }
}
exports.ServicePrincipalsV2Error = ServicePrincipalsV2Error;
/**
 * Identities for use with jobs, automated tools, and systems such as scripts,
 * apps, and CI/CD platforms. Databricks recommends creating service principals
 * to run production jobs or modify production data. If all processes that act on
 * production data run with service principals, interactive users do not need any
 * write, delete, or modify privileges in production. This eliminates the risk of
 * a user overwriting production data by accident.
 */
class ServicePrincipalsV2Service {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("active")) {
            body["active"] = request.active;
        }
        if (request.hasOwnProperty("applicationId")) {
            body["applicationId"] = request.applicationId;
        }
        if (request.hasOwnProperty("displayName")) {
            body["displayName"] = request.displayName;
        }
        if (request.hasOwnProperty("entitlements")) {
            body["entitlements"] = request.entitlements;
        }
        if (request.hasOwnProperty("externalId")) {
            body["externalId"] = request.externalId;
        }
        if (request.hasOwnProperty("groups")) {
            body["groups"] = request.groups;
        }
        if (request.hasOwnProperty("id")) {
            body["id"] = request.id;
        }
        if (request.hasOwnProperty("roles")) {
            body["roles"] = request.roles;
        }
        if (request.hasOwnProperty("schemas")) {
            body["schemas"] = request.schemas;
        }
        const query = {};
        const path = "/api/2.0/preview/scim/v2/ServicePrincipals";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new service principal in the Databricks workspace.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/scim/v2/ServicePrincipals/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a single service principal in the Databricks workspace.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/scim/v2/ServicePrincipals/${request.id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the details for a single service principal define in the Databricks
     * workspace.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("attributes")) {
            query["attributes"] =
                request.attributes;
        }
        if (request.hasOwnProperty("count")) {
            query["count"] = request.count;
        }
        if (request.hasOwnProperty("excludedAttributes")) {
            query["excludedAttributes"] =
                request.excludedAttributes;
        }
        if (request.hasOwnProperty("filter")) {
            query["filter"] = request.filter;
        }
        if (request.hasOwnProperty("sortBy")) {
            query["sortBy"] = request.sortBy;
        }
        if (request.hasOwnProperty("sortOrder")) {
            query["sortOrder"] =
                request.sortOrder;
        }
        if (request.hasOwnProperty("startIndex")) {
            query["startIndex"] =
                request.startIndex;
        }
        const path = "/api/2.0/preview/scim/v2/ServicePrincipals";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the set of service principals associated with a Databricks workspace.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.Resources || response.Resources.length === 0) {
                break;
            }
            for (const v of response.Resources) {
                yield v;
            }
        }
    }
    async _patch(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("Operations")) {
            body["Operations"] = request.Operations;
        }
        if (request.hasOwnProperty("schemas")) {
            body["schemas"] = request.schemas;
        }
        const query = {};
        const path = `/api/2.0/preview/scim/v2/ServicePrincipals/${request.id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Partially updates the details of a single service principal in the
     * Databricks workspace.
     */
    async patch(request, context) {
        return await this._patch(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("active")) {
            body["active"] = request.active;
        }
        if (request.hasOwnProperty("applicationId")) {
            body["applicationId"] = request.applicationId;
        }
        if (request.hasOwnProperty("displayName")) {
            body["displayName"] = request.displayName;
        }
        if (request.hasOwnProperty("entitlements")) {
            body["entitlements"] = request.entitlements;
        }
        if (request.hasOwnProperty("externalId")) {
            body["externalId"] = request.externalId;
        }
        if (request.hasOwnProperty("groups")) {
            body["groups"] = request.groups;
        }
        if (request.hasOwnProperty("roles")) {
            body["roles"] = request.roles;
        }
        if (request.hasOwnProperty("schemas")) {
            body["schemas"] = request.schemas;
        }
        const query = {};
        const path = `/api/2.0/preview/scim/v2/ServicePrincipals/${request.id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the details of a single service principal.
     *
     * This action replaces the existing service principal with the same name.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.ServicePrincipalsV2Service = ServicePrincipalsV2Service;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalsV2Service.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalsV2Service.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalsV2Service.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalsV2Service.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalsV2Service.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalsV2Service.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalsV2Service.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ServicePrincipalsV2Service.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalsV2Service.prototype, "_patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalsV2Service.prototype, "patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalsV2Service.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalsV2Service.prototype, "update", null);
class UsersV2RetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("UsersV2", method, message);
    }
}
exports.UsersV2RetriableError = UsersV2RetriableError;
class UsersV2Error extends apiError_1.ApiError {
    constructor(method, message) {
        super("UsersV2", method, message);
    }
}
exports.UsersV2Error = UsersV2Error;
/**
 * User identities recognized by Databricks and represented by email addresses.
 *
 * Databricks recommends using SCIM provisioning to sync users and groups
 * automatically from your identity provider to your Databricks workspace. SCIM
 * streamlines onboarding a new employee or team by using your identity provider
 * to create users and groups in Databricks workspace and give them the proper
 * level of access. When a user leaves your organization or no longer needs
 * access to Databricks workspace, admins can terminate the user in your identity
 * provider and that user’s account will also be removed from Databricks
 * workspace. This ensures a consistent offboarding process and prevents
 * unauthorized users from accessing sensitive data.
 */
class UsersV2Service {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("active")) {
            body["active"] = request.active;
        }
        if (request.hasOwnProperty("displayName")) {
            body["displayName"] = request.displayName;
        }
        if (request.hasOwnProperty("emails")) {
            body["emails"] = request.emails;
        }
        if (request.hasOwnProperty("entitlements")) {
            body["entitlements"] = request.entitlements;
        }
        if (request.hasOwnProperty("externalId")) {
            body["externalId"] = request.externalId;
        }
        if (request.hasOwnProperty("groups")) {
            body["groups"] = request.groups;
        }
        if (request.hasOwnProperty("id")) {
            body["id"] = request.id;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("roles")) {
            body["roles"] = request.roles;
        }
        if (request.hasOwnProperty("schemas")) {
            body["schemas"] = request.schemas;
        }
        if (request.hasOwnProperty("userName")) {
            body["userName"] = request.userName;
        }
        const query = {};
        const path = "/api/2.0/preview/scim/v2/Users";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new user in the Databricks workspace. This new user will also be
     * added to the Databricks account.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/preview/scim/v2/Users/${request.id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a user. Deleting a user from a Databricks workspace also removes
     * objects associated with the user.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("attributes")) {
            query["attributes"] =
                request.attributes;
        }
        if (request.hasOwnProperty("count")) {
            query["count"] = request.count;
        }
        if (request.hasOwnProperty("excludedAttributes")) {
            query["excludedAttributes"] =
                request.excludedAttributes;
        }
        if (request.hasOwnProperty("filter")) {
            query["filter"] = request.filter;
        }
        if (request.hasOwnProperty("sortBy")) {
            query["sortBy"] = request.sortBy;
        }
        if (request.hasOwnProperty("sortOrder")) {
            query["sortOrder"] =
                request.sortOrder;
        }
        if (request.hasOwnProperty("startIndex")) {
            query["startIndex"] =
                request.startIndex;
        }
        const path = `/api/2.0/preview/scim/v2/Users/${request.id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets information for a specific user in Databricks workspace.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _getPermissionLevels(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/permissions/authorization/passwords/permissionLevels";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(context) {
        return await this._getPermissionLevels(context);
    }
    async _getPermissions(context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = "/api/2.0/permissions/authorization/passwords";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of all passwords. Passwords can inherit permissions
     * from their root object.
     */
    async getPermissions(context) {
        return await this._getPermissions(context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("attributes")) {
            query["attributes"] =
                request.attributes;
        }
        if (request.hasOwnProperty("count")) {
            query["count"] = request.count;
        }
        if (request.hasOwnProperty("excludedAttributes")) {
            query["excludedAttributes"] =
                request.excludedAttributes;
        }
        if (request.hasOwnProperty("filter")) {
            query["filter"] = request.filter;
        }
        if (request.hasOwnProperty("sortBy")) {
            query["sortBy"] = request.sortBy;
        }
        if (request.hasOwnProperty("sortOrder")) {
            query["sortOrder"] =
                request.sortOrder;
        }
        if (request.hasOwnProperty("startIndex")) {
            query["startIndex"] =
                request.startIndex;
        }
        const path = "/api/2.0/preview/scim/v2/Users";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets details for all the users associated with a Databricks workspace.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.Resources || response.Resources.length === 0) {
                break;
            }
            for (const v of response.Resources) {
                yield v;
            }
        }
    }
    async _patch(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("Operations")) {
            body["Operations"] = request.Operations;
        }
        if (request.hasOwnProperty("schemas")) {
            body["schemas"] = request.schemas;
        }
        const query = {};
        const path = `/api/2.0/preview/scim/v2/Users/${request.id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Partially updates a user resource by applying the supplied operations on
     * specific user attributes.
     */
    async patch(request, context) {
        return await this._patch(request, context);
    }
    async _setPermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = "/api/2.0/permissions/authorization/passwords";
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their root object.
     */
    async setPermissions(request, context) {
        return await this._setPermissions(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("active")) {
            body["active"] = request.active;
        }
        if (request.hasOwnProperty("displayName")) {
            body["displayName"] = request.displayName;
        }
        if (request.hasOwnProperty("emails")) {
            body["emails"] = request.emails;
        }
        if (request.hasOwnProperty("entitlements")) {
            body["entitlements"] = request.entitlements;
        }
        if (request.hasOwnProperty("externalId")) {
            body["externalId"] = request.externalId;
        }
        if (request.hasOwnProperty("groups")) {
            body["groups"] = request.groups;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("roles")) {
            body["roles"] = request.roles;
        }
        if (request.hasOwnProperty("schemas")) {
            body["schemas"] = request.schemas;
        }
        if (request.hasOwnProperty("userName")) {
            body["userName"] = request.userName;
        }
        const query = {};
        const path = `/api/2.0/preview/scim/v2/Users/${request.id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Replaces a user's information with the data supplied in request.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = "/api/2.0/permissions/authorization/passwords";
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on all passwords. Passwords can inherit
     * permissions from their root object.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
}
exports.UsersV2Service = UsersV2Service;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "_getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(0, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], UsersV2Service.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "_patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "patch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "_setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], UsersV2Service.prototype, "updatePermissions", null);
class WorkspaceAssignmentRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("WorkspaceAssignment", method, message);
    }
}
exports.WorkspaceAssignmentRetriableError = WorkspaceAssignmentRetriableError;
class WorkspaceAssignmentError extends apiError_1.ApiError {
    constructor(method, message) {
        super("WorkspaceAssignment", method, message);
    }
}
exports.WorkspaceAssignmentError = WorkspaceAssignmentError;
/**
 * The Workspace Permission Assignment API allows you to manage workspace
 * permissions for principals in your account.
 */
class WorkspaceAssignmentService {
    constructor(client) {
        this.client = client;
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/workspaces/${request.workspace_id}/permissionassignments/principals/${request.principal_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes the workspace permissions assignment in a given account and
     * workspace for the specified principal.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/workspaces/${request.workspace_id}/permissionassignments/permissions`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get an array of workspace permissions for the specified account and
     * workspace.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/workspaces/${request.workspace_id}/permissionassignments`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get the permission assignments for the specified Databricks account and
     * Databricks workspace.
     */
    async *list(request, context) {
        const response = (await this._list(request, context))
            .permission_assignments;
        for (const v of response || []) {
            yield v;
        }
    }
    async _update(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("permissions")) {
            body["permissions"] = request.permissions;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/workspaces/${request.workspace_id}/permissionassignments/principals/${request.principal_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates or updates the workspace permissions assignment in a given account
     * and workspace for the specified principal.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.WorkspaceAssignmentService = WorkspaceAssignmentService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceAssignmentService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceAssignmentService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceAssignmentService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceAssignmentService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceAssignmentService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], WorkspaceAssignmentService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceAssignmentService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceAssignmentService.prototype, "update", null);
