// @databricks/sdk-experimental/dist/apis/jobs/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolicyComplianceForJobsService = exports.PolicyComplianceForJobsError = exports.PolicyComplianceForJobsRetriableError = exports.JobsService = exports.JobsError = exports.JobsRetriableError = void 0;
const jobs = __importStar(require("./model"));
const retries_1 = __importDefault(require("../../retries/retries"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
const wait_1 = require("../../wait");
class JobsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Jobs", method, message);
    }
}
exports.JobsRetriableError = JobsRetriableError;
class JobsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Jobs", method, message);
    }
}
exports.JobsError = JobsError;
/**
 * The Jobs API allows you to create, edit, and delete jobs.
 *
 * You can use a Databricks job to run a data processing or data analysis task in
 * a Databricks cluster with scalable resources. Your job can consist of a single
 * task or can be a large, multi-task workflow with complex dependencies.
 * Databricks manages the task orchestration, cluster management, monitoring, and
 * error reporting for all of your jobs. You can run your jobs immediately or
 * periodically through an easy-to-use scheduling system. You can implement job
 * tasks using notebooks, JARS, Spark Declarative Pipelines, or Python, Scala,
 * Spark submit, and Java applications.
 *
 * You should never hard code secrets or store them in plain text. Use the
 * [Secrets CLI] to manage secrets in the [Databricks CLI]. Use the [Secrets
 * utility] to reference secrets in notebooks and jobs.
 *
 * [Databricks CLI]: https://docs.databricks.com/dev-tools/cli/index.html
 * [Secrets CLI]: https://docs.databricks.com/dev-tools/cli/secrets-cli.html
 * [Secrets utility]: https://docs.databricks.com/dev-tools/databricks-utils.html#dbutils-secrets
 */
class JobsService {
    constructor(client) {
        this.client = client;
    }
    async _cancelAllRuns(request, context) {
        const headers = new Headers({ "Content-Type": "application/json" });
        const body = {};
        if (request.hasOwnProperty("all_queued_runs")) {
            body["all_queued_runs"] = request.all_queued_runs;
        }
        if (request.hasOwnProperty("job_id")) {
            body["job_id"] = request.job_id;
        }
        const query = {};
        const path = "/api/2.2/jobs/runs/cancel-all";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Cancels all active runs of a job. The runs are canceled asynchronously, so
     * it doesn't prevent new runs from being started.
     */
    async cancelAllRuns(request, context) {
        return await this._cancelAllRuns(request, context);
    }
    async _cancelRun(request, context) {
        const headers = new Headers({ "Content-Type": "application/json" });
        const body = {};
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        const query = {};
        const path = "/api/2.2/jobs/runs/cancel";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Cancels a job run or a task run. The run is canceled asynchronously, so it
     * may still be running when this request completes.
     */
    async cancelRun(cancelRun, context) {
        let cancellationToken = context?.cancellationToken;
        await this._cancelRun(cancelRun, context);
        return (0, wait_1.asWaiter)(null, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.getRun({
                        run_id: cancelRun.run_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Jobs.cancelRunAndWait: cancelled");
                        throw new JobsError("cancelRunAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state.life_cycle_state;
                    const statusMessage = pollResponse.state.state_message;
                    switch (status) {
                        case "TERMINATED":
                        case "SKIPPED": {
                            return pollResponse;
                        }
                        case "INTERNAL_ERROR": {
                            const errorMessage = `failed to reach TERMINATED or SKIPPED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Jobs.cancelRunAndWait: ${errorMessage}`);
                            throw new JobsError("cancelRunAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach TERMINATED or SKIPPED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Jobs.cancelRunAndWait: retrying: ${errorMessage}`);
                            throw new JobsRetriableError("cancelRunAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        if (request.hasOwnProperty("budget_policy_id")) {
            body["budget_policy_id"] = request.budget_policy_id;
        }
        if (request.hasOwnProperty("continuous")) {
            body["continuous"] = request.continuous;
        }
        if (request.hasOwnProperty("deployment")) {
            body["deployment"] = request.deployment;
        }
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("edit_mode")) {
            body["edit_mode"] = request.edit_mode;
        }
        if (request.hasOwnProperty("email_notifications")) {
            body["email_notifications"] = request.email_notifications;
        }
        if (request.hasOwnProperty("environments")) {
            body["environments"] = request.environments;
        }
        if (request.hasOwnProperty("format")) {
            body["format"] = request.format;
        }
        if (request.hasOwnProperty("git_source")) {
            body["git_source"] = request.git_source;
        }
        if (request.hasOwnProperty("health")) {
            body["health"] = request.health;
        }
        if (request.hasOwnProperty("job_clusters")) {
            body["job_clusters"] = request.job_clusters;
        }
        if (request.hasOwnProperty("max_concurrent_runs")) {
            body["max_concurrent_runs"] = request.max_concurrent_runs;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("notification_settings")) {
            body["notification_settings"] = request.notification_settings;
        }
        if (request.hasOwnProperty("parameters")) {
            body["parameters"] = request.parameters;
        }
        if (request.hasOwnProperty("performance_target")) {
            body["performance_target"] = request.performance_target;
        }
        if (request.hasOwnProperty("queue")) {
            body["queue"] = request.queue;
        }
        if (request.hasOwnProperty("run_as")) {
            body["run_as"] = request.run_as;
        }
        if (request.hasOwnProperty("schedule")) {
            body["schedule"] = request.schedule;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        if (request.hasOwnProperty("tasks")) {
            body["tasks"] = request.tasks;
        }
        if (request.hasOwnProperty("timeout_seconds")) {
            body["timeout_seconds"] = request.timeout_seconds;
        }
        if (request.hasOwnProperty("trigger")) {
            body["trigger"] = request.trigger;
        }
        if (request.hasOwnProperty("usage_policy_id")) {
            body["usage_policy_id"] = request.usage_policy_id;
        }
        if (request.hasOwnProperty("webhook_notifications")) {
            body["webhook_notifications"] = request.webhook_notifications;
        }
        const query = {};
        const path = "/api/2.2/jobs/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a new job.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ "Content-Type": "application/json" });
        const body = {};
        if (request.hasOwnProperty("job_id")) {
            body["job_id"] = request.job_id;
        }
        const query = {};
        const path = "/api/2.2/jobs/delete";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a job.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _deleteRun(request, context) {
        const headers = new Headers({ "Content-Type": "application/json" });
        const body = {};
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        const query = {};
        const path = "/api/2.2/jobs/runs/delete";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a non-active run. Returns an error if the run is active.
     */
    async deleteRun(request, context) {
        return await this._deleteRun(request, context);
    }
    async _exportRun(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("run_id")) {
            query["run_id"] = request.run_id;
        }
        if (request.hasOwnProperty("views_to_export")) {
            query["views_to_export"] =
                request.views_to_export;
        }
        const path = "/api/2.2/jobs/runs/export";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Export and retrieve the job run task.
     */
    async exportRun(request, context) {
        return await this._exportRun(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("job_id")) {
            query["job_id"] = request.job_id;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.2/jobs/get";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieves the details for a single job.
     *
     * Large arrays in the results will be paginated when they exceed 100
     * elements. A request for a single job will return all properties for that
     * job, and the first 100 elements of array properties (`tasks`,
     * `job_clusters`, `environments` and `parameters`). Use the
     * `next_page_token` field to check for more results and pass its value as
     * the `page_token` in subsequent requests. If any array properties have more
     * than 100 elements, additional results will be returned on subsequent
     * requests. Arrays without additional results will be empty on later pages.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _getPermissionLevels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/jobs/${request.job_id}/permissionLevels`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(request, context) {
        return await this._getPermissionLevels(request, context);
    }
    async _getPermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/jobs/${request.job_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of a job. Jobs can inherit permissions from their
     * root object.
     */
    async getPermissions(request, context) {
        return await this._getPermissions(request, context);
    }
    async _getRun(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("include_history")) {
            query["include_history"] =
                request.include_history;
        }
        if (request.hasOwnProperty("include_resolved_values")) {
            query["include_resolved_values"] =
                request.include_resolved_values;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        if (request.hasOwnProperty("run_id")) {
            query["run_id"] = request.run_id;
        }
        const path = "/api/2.2/jobs/runs/get";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieves the metadata of a run.
     *
     * Large arrays in the results will be paginated when they exceed 100
     * elements. A request for a single run will return all properties for that
     * run, and the first 100 elements of array properties (`tasks`,
     * `job_clusters`, `job_parameters` and `repair_history`). Use the
     * next_page_token field to check for more results and pass its value as the
     * page_token in subsequent requests. If any array properties have more than
     * 100 elements, additional results will be returned on subsequent requests.
     * Arrays without additional results will be empty on later pages.
     */
    async getRun(request, context) {
        return await this._getRun(request, context);
    }
    async _getRunOutput(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("run_id")) {
            query["run_id"] = request.run_id;
        }
        const path = "/api/2.2/jobs/runs/get-output";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieve the output and metadata of a single task run. When a notebook
     * task returns a value through the `dbutils.notebook.exit()` call, you can
     * use this endpoint to retrieve that value. Databricks restricts this API to
     * returning the first 5 MB of the output. To return a larger result, you can
     * store job results in a cloud storage service.
     *
     * This endpoint validates that the __run_id__ parameter is valid and returns
     * an HTTP status code 400 if the __run_id__ parameter is invalid. Runs are
     * automatically removed after 60 days. If you to want to reference them
     * beyond 60 days, you must save old run results before they expire.
     */
    async getRunOutput(request, context) {
        return await this._getRunOutput(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("expand_tasks")) {
            query["expand_tasks"] =
                request.expand_tasks;
        }
        if (request.hasOwnProperty("limit")) {
            query["limit"] = request.limit;
        }
        if (request.hasOwnProperty("name")) {
            query["name"] = request.name;
        }
        if (request.hasOwnProperty("offset")) {
            query["offset"] = request.offset;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.2/jobs/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Retrieves a list of jobs.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.jobs || response.jobs.length === 0) {
                break;
            }
            for (const v of response.jobs) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _listRuns(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("active_only")) {
            query["active_only"] =
                request.active_only;
        }
        if (request.hasOwnProperty("completed_only")) {
            query["completed_only"] =
                request.completed_only;
        }
        if (request.hasOwnProperty("expand_tasks")) {
            query["expand_tasks"] =
                request.expand_tasks;
        }
        if (request.hasOwnProperty("job_id")) {
            query["job_id"] = request.job_id;
        }
        if (request.hasOwnProperty("limit")) {
            query["limit"] = request.limit;
        }
        if (request.hasOwnProperty("offset")) {
            query["offset"] = request.offset;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        if (request.hasOwnProperty("run_type")) {
            query["run_type"] =
                request.run_type;
        }
        if (request.hasOwnProperty("start_time_from")) {
            query["start_time_from"] =
                request.start_time_from;
        }
        if (request.hasOwnProperty("start_time_to")) {
            query["start_time_to"] =
                request.start_time_to;
        }
        const path = "/api/2.2/jobs/runs/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List runs in descending order by start time.
     */
    async *listRuns(request, context) {
        while (true) {
            const response = await this._listRuns(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.runs || response.runs.length === 0) {
                break;
            }
            for (const v of response.runs) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _repairRun(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("dbt_commands")) {
            body["dbt_commands"] = request.dbt_commands;
        }
        if (request.hasOwnProperty("jar_params")) {
            body["jar_params"] = request.jar_params;
        }
        if (request.hasOwnProperty("job_parameters")) {
            body["job_parameters"] = request.job_parameters;
        }
        if (request.hasOwnProperty("latest_repair_id")) {
            body["latest_repair_id"] = request.latest_repair_id;
        }
        if (request.hasOwnProperty("notebook_params")) {
            body["notebook_params"] = request.notebook_params;
        }
        if (request.hasOwnProperty("performance_target")) {
            body["performance_target"] = request.performance_target;
        }
        if (request.hasOwnProperty("pipeline_params")) {
            body["pipeline_params"] = request.pipeline_params;
        }
        if (request.hasOwnProperty("python_named_params")) {
            body["python_named_params"] = request.python_named_params;
        }
        if (request.hasOwnProperty("python_params")) {
            body["python_params"] = request.python_params;
        }
        if (request.hasOwnProperty("rerun_all_failed_tasks")) {
            body["rerun_all_failed_tasks"] = request.rerun_all_failed_tasks;
        }
        if (request.hasOwnProperty("rerun_dependent_tasks")) {
            body["rerun_dependent_tasks"] = request.rerun_dependent_tasks;
        }
        if (request.hasOwnProperty("rerun_tasks")) {
            body["rerun_tasks"] = request.rerun_tasks;
        }
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        if (request.hasOwnProperty("spark_submit_params")) {
            body["spark_submit_params"] = request.spark_submit_params;
        }
        if (request.hasOwnProperty("sql_params")) {
            body["sql_params"] = request.sql_params;
        }
        const query = {};
        const path = "/api/2.2/jobs/runs/repair";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Re-run one or more tasks. Tasks are re-run as part of the original job
     * run. They use the current job and task settings, and can be viewed in the
     * history for the original job run.
     */
    async repairRun(repairRun, context) {
        let cancellationToken = context?.cancellationToken;
        const repairRunResponse = await this._repairRun(repairRun, context);
        return (0, wait_1.asWaiter)(repairRunResponse, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.getRun({
                        run_id: repairRun.run_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Jobs.repairRunAndWait: cancelled");
                        throw new JobsError("repairRunAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state.life_cycle_state;
                    const statusMessage = pollResponse.state.state_message;
                    switch (status) {
                        case "TERMINATED":
                        case "SKIPPED": {
                            return pollResponse;
                        }
                        case "INTERNAL_ERROR": {
                            const errorMessage = `failed to reach TERMINATED or SKIPPED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Jobs.repairRunAndWait: ${errorMessage}`);
                            throw new JobsError("repairRunAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach TERMINATED or SKIPPED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Jobs.repairRunAndWait: retrying: ${errorMessage}`);
                            throw new JobsRetriableError("repairRunAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _reset(request, context) {
        const headers = new Headers({ "Content-Type": "application/json" });
        const body = {};
        if (request.hasOwnProperty("job_id")) {
            body["job_id"] = request.job_id;
        }
        if (request.hasOwnProperty("new_settings")) {
            body["new_settings"] = request.new_settings;
        }
        const query = {};
        const path = "/api/2.2/jobs/reset";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Overwrite all settings for the given job. Use the [_Update_
     * endpoint](:method:jobs/update) to update job settings partially.
     */
    async reset(request, context) {
        return await this._reset(request, context);
    }
    async _runNow(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("dbt_commands")) {
            body["dbt_commands"] = request.dbt_commands;
        }
        if (request.hasOwnProperty("idempotency_token")) {
            body["idempotency_token"] = request.idempotency_token;
        }
        if (request.hasOwnProperty("jar_params")) {
            body["jar_params"] = request.jar_params;
        }
        if (request.hasOwnProperty("job_id")) {
            body["job_id"] = request.job_id;
        }
        if (request.hasOwnProperty("job_parameters")) {
            body["job_parameters"] = request.job_parameters;
        }
        if (request.hasOwnProperty("notebook_params")) {
            body["notebook_params"] = request.notebook_params;
        }
        if (request.hasOwnProperty("only")) {
            body["only"] = request.only;
        }
        if (request.hasOwnProperty("performance_target")) {
            body["performance_target"] = request.performance_target;
        }
        if (request.hasOwnProperty("pipeline_params")) {
            body["pipeline_params"] = request.pipeline_params;
        }
        if (request.hasOwnProperty("python_named_params")) {
            body["python_named_params"] = request.python_named_params;
        }
        if (request.hasOwnProperty("python_params")) {
            body["python_params"] = request.python_params;
        }
        if (request.hasOwnProperty("queue")) {
            body["queue"] = request.queue;
        }
        if (request.hasOwnProperty("spark_submit_params")) {
            body["spark_submit_params"] = request.spark_submit_params;
        }
        if (request.hasOwnProperty("sql_params")) {
            body["sql_params"] = request.sql_params;
        }
        const query = {};
        const path = "/api/2.2/jobs/run-now";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Run a job and return the `run_id` of the triggered run.
     */
    async runNow(runNow, context) {
        let cancellationToken = context?.cancellationToken;
        const runNowResponse = await this._runNow(runNow, context);
        return (0, wait_1.asWaiter)(runNowResponse, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.getRun({
                        run_id: runNowResponse.run_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Jobs.runNowAndWait: cancelled");
                        throw new JobsError("runNowAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state.life_cycle_state;
                    const statusMessage = pollResponse.state.state_message;
                    switch (status) {
                        case "TERMINATED":
                        case "SKIPPED": {
                            return pollResponse;
                        }
                        case "INTERNAL_ERROR": {
                            const errorMessage = `failed to reach TERMINATED or SKIPPED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Jobs.runNowAndWait: ${errorMessage}`);
                            throw new JobsError("runNowAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach TERMINATED or SKIPPED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Jobs.runNowAndWait: retrying: ${errorMessage}`);
                            throw new JobsRetriableError("runNowAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _setPermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/jobs/${request.job_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their root object.
     */
    async setPermissions(request, context) {
        return await this._setPermissions(request, context);
    }
    async _submit(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        if (request.hasOwnProperty("budget_policy_id")) {
            body["budget_policy_id"] = request.budget_policy_id;
        }
        if (request.hasOwnProperty("email_notifications")) {
            body["email_notifications"] = request.email_notifications;
        }
        if (request.hasOwnProperty("environments")) {
            body["environments"] = request.environments;
        }
        if (request.hasOwnProperty("git_source")) {
            body["git_source"] = request.git_source;
        }
        if (request.hasOwnProperty("health")) {
            body["health"] = request.health;
        }
        if (request.hasOwnProperty("idempotency_token")) {
            body["idempotency_token"] = request.idempotency_token;
        }
        if (request.hasOwnProperty("notification_settings")) {
            body["notification_settings"] = request.notification_settings;
        }
        if (request.hasOwnProperty("queue")) {
            body["queue"] = request.queue;
        }
        if (request.hasOwnProperty("run_as")) {
            body["run_as"] = request.run_as;
        }
        if (request.hasOwnProperty("run_name")) {
            body["run_name"] = request.run_name;
        }
        if (request.hasOwnProperty("tasks")) {
            body["tasks"] = request.tasks;
        }
        if (request.hasOwnProperty("timeout_seconds")) {
            body["timeout_seconds"] = request.timeout_seconds;
        }
        if (request.hasOwnProperty("usage_policy_id")) {
            body["usage_policy_id"] = request.usage_policy_id;
        }
        if (request.hasOwnProperty("webhook_notifications")) {
            body["webhook_notifications"] = request.webhook_notifications;
        }
        const query = {};
        const path = "/api/2.2/jobs/runs/submit";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Submit a one-time run. This endpoint allows you to submit a workload
     * directly without creating a job. Runs submitted using this endpoint
     * don’t display in the UI. Use the `jobs/runs/get` API to check the run
     * state after the job is submitted.
     */
    async submit(submitRun, context) {
        let cancellationToken = context?.cancellationToken;
        const submitRunResponse = await this._submit(submitRun, context);
        return (0, wait_1.asWaiter)(submitRunResponse, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.getRun({
                        run_id: submitRunResponse.run_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Jobs.submitAndWait: cancelled");
                        throw new JobsError("submitAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state.life_cycle_state;
                    const statusMessage = pollResponse.state.state_message;
                    switch (status) {
                        case "TERMINATED":
                        case "SKIPPED": {
                            return pollResponse;
                        }
                        case "INTERNAL_ERROR": {
                            const errorMessage = `failed to reach TERMINATED or SKIPPED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Jobs.submitAndWait: ${errorMessage}`);
                            throw new JobsError("submitAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach TERMINATED or SKIPPED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Jobs.submitAndWait: retrying: ${errorMessage}`);
                            throw new JobsRetriableError("submitAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _update(request, context) {
        const headers = new Headers({ "Content-Type": "application/json" });
        const body = {};
        if (request.hasOwnProperty("fields_to_remove")) {
            body["fields_to_remove"] = request.fields_to_remove;
        }
        if (request.hasOwnProperty("job_id")) {
            body["job_id"] = request.job_id;
        }
        if (request.hasOwnProperty("new_settings")) {
            body["new_settings"] = request.new_settings;
        }
        const query = {};
        const path = "/api/2.2/jobs/update";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Add, update, or remove specific settings of an existing job. Use the
     * [_Reset_ endpoint](:method:jobs/reset) to overwrite all job settings.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/jobs/${request.job_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on a job. Jobs can inherit permissions from their
     * root object.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
}
exports.JobsService = JobsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_cancelAllRuns", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "cancelAllRuns", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_cancelRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "cancelRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_deleteRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "deleteRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_exportRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "exportRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_getRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "getRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_getRunOutput", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "getRunOutput", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], JobsService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_listRuns", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], JobsService.prototype, "listRuns", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_repairRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "repairRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_reset", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "reset", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_runNow", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "runNow", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_submit", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "submit", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], JobsService.prototype, "updatePermissions", null);
class PolicyComplianceForJobsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("PolicyComplianceForJobs", method, message);
    }
}
exports.PolicyComplianceForJobsRetriableError = PolicyComplianceForJobsRetriableError;
class PolicyComplianceForJobsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("PolicyComplianceForJobs", method, message);
    }
}
exports.PolicyComplianceForJobsError = PolicyComplianceForJobsError;
/**
 * The compliance APIs allow you to view and manage the policy compliance status
 * of jobs in your workspace. This API currently only supports compliance
 * controls for cluster policies.
 *
 * A job is in compliance if its cluster configurations satisfy the rules of all
 * their respective cluster policies. A job could be out of compliance if a
 * cluster policy it uses was updated after the job was last edited. The job is
 * considered out of compliance if any of its clusters no longer comply with
 * their updated policies.
 *
 * The get and list compliance APIs allow you to view the policy compliance
 * status of a job. The enforce compliance API allows you to update a job so that
 * it becomes compliant with all of its policies.
 */
class PolicyComplianceForJobsService {
    constructor(client) {
        this.client = client;
    }
    async _enforceCompliance(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("job_id")) {
            body["job_id"] = request.job_id;
        }
        if (request.hasOwnProperty("validate_only")) {
            body["validate_only"] = request.validate_only;
        }
        const query = {};
        const path = "/api/2.0/policies/jobs/enforce-compliance";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates a job so the job clusters that are created when running the job
     * (specified in `new_cluster`) are compliant with the current versions of
     * their respective cluster policies. All-purpose clusters used in the job
     * will not be updated.
     */
    async enforceCompliance(request, context) {
        return await this._enforceCompliance(request, context);
    }
    async _getCompliance(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("job_id")) {
            query["job_id"] = request.job_id;
        }
        const path = "/api/2.0/policies/jobs/get-compliance";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns the policy compliance status of a job. Jobs could be out of
     * compliance if a cluster policy they use was updated after the job was last
     * edited and some of its job clusters no longer comply with their updated
     * policies.
     */
    async getCompliance(request, context) {
        return await this._getCompliance(request, context);
    }
    async _listCompliance(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        if (request.hasOwnProperty("policy_id")) {
            query["policy_id"] =
                request.policy_id;
        }
        const path = "/api/2.0/policies/jobs/list-compliance";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Returns the policy compliance status of all jobs that use a given policy.
     * Jobs could be out of compliance if a cluster policy they use was updated
     * after the job was last edited and its job clusters no longer comply with
     * the updated policy.
     */
    async *listCompliance(request, context) {
        while (true) {
            const response = await this._listCompliance(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.jobs || response.jobs.length === 0) {
                break;
            }
            for (const v of response.jobs) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
}
exports.PolicyComplianceForJobsService = PolicyComplianceForJobsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PolicyComplianceForJobsService.prototype, "_enforceCompliance", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PolicyComplianceForJobsService.prototype, "enforceCompliance", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PolicyComplianceForJobsService.prototype, "_getCompliance", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PolicyComplianceForJobsService.prototype, "getCompliance", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PolicyComplianceForJobsService.prototype, "_listCompliance", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], PolicyComplianceForJobsService.prototype, "listCompliance", null);
