// @databricks/appkit/dist/plugins/analytics/manifest.js
//#region src/plugins/analytics/manifest.json
var manifest_default = {
	$schema: "https://databricks.github.io/appkit/schemas/plugin-manifest.schema.json",
	name: "analytics",
	displayName: "Analytics Plugin",
	description: "SQL query execution against Databricks SQL Warehouses",
	resources: {
		"required": [{
			"type": "sql_warehouse",
			"alias": "SQL Warehouse",
			"resourceKey": "sql-warehouse",
			"description": "SQL Warehouse for executing analytics queries",
			"permission": "CAN_USE",
			"fields": { "id": {
				"env": "DATABRICKS_WAREHOUSE_ID",
				"description": "SQL Warehouse ID",
				"discovery": {
					"type": "kind",
					"resourceKind": "warehouse"
				}
			} }
		}],
		"optional": []
	},
	scaffolding: { "rules": {
		"must": ["Before init, ensure the SQL Warehouse passed via --set analytics.sql-warehouse.id is running"],
		"should": ["After init, ensure config/queries/ has at least one .sql file before running npm run typegen"]
	} },
	config: { "schema": {
		"type": "object",
		"properties": { "timeout": {
			"type": "number",
			"default": 3e4,
			"description": "Query execution timeout in milliseconds"
		} }
	} }
};

//#endregion
export { manifest_default as default };
