// @databricks/appkit/dist/plugins/analytics/query.js
import { isSQLTypeMarker, sql } from "../../shared/src/sql/helpers.js";
import "../../shared/src/index.js";
import { ValidationError } from "../../errors/validation.js";
import "../../errors/index.js";
import { getWorkspaceId } from "../../context/execution-context.js";
import "../../context/index.js";
import { createHash } from "node:crypto";

//#region src/plugins/analytics/query.ts
var QueryProcessor = class {
	async processQueryParams(query, parameters) {
		const processed = { ...parameters };
		const paramMatches = query.matchAll(/:([a-zA-Z_]\w*)/g);
		if (new Set(Array.from(paramMatches, (m) => m[1])).has("workspaceId") && !processed.workspaceId) {
			const workspaceId = await getWorkspaceId();
			if (workspaceId) processed.workspaceId = sql.string(workspaceId);
		}
		return processed;
	}
	hashQuery(query) {
		return createHash("md5").update(query).digest("hex");
	}
	convertToSQLParameters(query, parameters) {
		const sqlParameters = [];
		if (parameters) {
			const queryParamMatches = query.matchAll(/:([a-zA-Z_]\w*)/g);
			const queryParams = new Set(Array.from(queryParamMatches, (m) => m[1]));
			for (const key of Object.keys(parameters)) if (!queryParams.has(key)) {
				const validParams = Array.from(queryParams).join(", ") || "none";
				throw ValidationError.invalidValue(key, parameters[key], `a parameter defined in the query (valid: ${validParams})`);
			}
			for (const [key, value] of Object.entries(parameters)) {
				const parameter = this._createParameter(key, value);
				if (parameter) sqlParameters.push(parameter);
			}
		}
		return {
			statement: query,
			parameters: sqlParameters
		};
	}
	_createParameter(key, value) {
		if (value === null || value === void 0) return null;
		if (!isSQLTypeMarker(value)) throw ValidationError.invalidValue(key, value, "SQL type (use sql.string(), sql.number(), sql.date(), sql.timestamp(), or sql.boolean())");
		return {
			name: key,
			value: value.value,
			type: value.__sql_type
		};
	}
};

//#endregion
export { QueryProcessor };
