// @databricks/appkit/dist/app/index.js
import { createLogger } from "../logging/logger.js";
import fs from "node:fs/promises";
import path from "node:path";

//#region src/app/index.ts
const logger = createLogger("app");
var AppManager = class {
	queriesDir = path.resolve(process.cwd(), "config/queries");
	/**
	* Validates that a file path is within the queries directory
	*/
	validatePath(fileName) {
		const queryFilePath = path.join(this.queriesDir, fileName);
		const resolvedPath = path.resolve(queryFilePath);
		const resolvedQueriesDir = path.resolve(this.queriesDir);
		if (!resolvedPath.startsWith(resolvedQueriesDir)) {
			logger.error("Invalid query path: path traversal detected");
			return null;
		}
		return resolvedPath;
	}
	/**
	* Creates a filesystem adapter based on dev mode or production mode
	*/
	createFsAdapter(req, devFileReader) {
		if (req?.query?.dev !== void 0 && devFileReader && req) return {
			readdir: async (dirPath) => {
				const relativePath = path.relative(process.cwd(), dirPath);
				return devFileReader.readdir(relativePath, req);
			},
			readFile: async (filePath) => {
				const relativePath = path.relative(process.cwd(), filePath);
				return devFileReader.readFile(relativePath, req);
			}
		};
		return {
			readdir: (dirPath) => fs.readdir(dirPath),
			readFile: (filePath) => fs.readFile(filePath, "utf8")
		};
	}
	/**
	* Retrieves a query file by key from the queries directory
	* In dev mode with a request context, reads from local filesystem via WebSocket
	* @param queryKey - The query file name (without extension)
	* @param req - Optional request object to detect dev mode
	* @param devFileReader - Optional DevFileReader instance to read files from local filesystem
	* @returns The query content and execution mode (as user or as service principal)
	*/
	async getAppQuery(queryKey, req, devFileReader) {
		if (!queryKey || !/^[a-zA-Z0-9_-]+$/.test(queryKey)) {
			logger.error("Invalid query key format: %s. Only alphanumeric characters, underscores, and hyphens are allowed.", queryKey);
			return null;
		}
		const fsAdapter = this.createFsAdapter(req, devFileReader);
		const oboFileName = `${queryKey}.obo.sql`;
		const defaultFileName = `${queryKey}.sql`;
		let files;
		try {
			files = await fsAdapter.readdir(this.queriesDir);
		} catch (error) {
			logger.error(`Failed to read queries directory: ${error.message}`);
			return null;
		}
		let queryFileName = null;
		let isAsUser = false;
		if (files.includes(oboFileName)) {
			queryFileName = oboFileName;
			isAsUser = true;
			if (files.includes(defaultFileName)) logger.warn(`Both ${oboFileName} and ${defaultFileName} found for query ${queryKey}. Using ${oboFileName}.`);
		} else if (files.includes(defaultFileName)) {
			queryFileName = defaultFileName;
			isAsUser = false;
		}
		if (!queryFileName) {
			logger.error(`Query file not found: ${queryKey}`);
			return null;
		}
		const resolvedPath = this.validatePath(queryFileName);
		if (!resolvedPath) return null;
		try {
			return {
				query: await fsAdapter.readFile(resolvedPath),
				isAsUser
			};
		} catch (error) {
			logger.error(`Failed to read query file: ${error.message}`);
			return null;
		}
	}
};

//#endregion
export { AppManager };
