// @databricks/appkit-ui/dist/react/genie/genie-query-transform.js
//#region src/react/genie/genie-query-transform.ts
const NUMERIC_SQL_TYPES = new Set([
	"DECIMAL",
	"INT",
	"INTEGER",
	"BIGINT",
	"LONG",
	"SMALLINT",
	"TINYINT",
	"FLOAT",
	"DOUBLE",
	"SHORT",
	"BYTE"
]);
const DATE_SQL_TYPES = new Set([
	"DATE",
	"TIMESTAMP",
	"TIMESTAMP_NTZ"
]);
/**
* Classify a SQL type_name into a high-level category.
*/
function classifySqlType(typeName) {
	const upper = typeName.toUpperCase();
	if (NUMERIC_SQL_TYPES.has(upper)) return "numeric";
	if (DATE_SQL_TYPES.has(upper)) return "date";
	return "string";
}
/**
* Parse a single cell value based on its column category.
*/
function parseValue(raw, category) {
	if (raw == null) return null;
	if (category === "numeric") {
		const n = Number(raw);
		return Number.isNaN(n) ? null : n;
	}
	return raw;
}
/**
* Transform a Genie statement_response into chart-ready rows + column metadata.
*
* Expects `data` to have the shape:
* ```
* {
*   manifest: { schema: { columns: [{ name, type_name }, ...] } },
*   result: { data_array: [["val", ...], ...] }
* }
* ```
*
* Returns `null` when the data is empty or malformed.
*/
function transformGenieData(data) {
	if (!data) return null;
	const rawColumns = data.manifest?.schema?.columns;
	if (!rawColumns || rawColumns.length === 0) return null;
	const dataArray = data.result?.data_array;
	if (!dataArray || dataArray.length === 0) return null;
	const columns = rawColumns.map((col) => ({
		name: col.name,
		typeName: col.type_name,
		category: classifySqlType(col.type_name)
	}));
	return {
		rows: dataArray.map((row) => {
			const record = {};
			for (let i = 0; i < columns.length; i++) {
				const col = columns[i];
				record[col.name] = parseValue(row[i] ?? null, col.category);
			}
			return record;
		}),
		columns
	};
}

//#endregion
export { transformGenieData };
