// @databricks/appkit-ui/dist/react/genie/genie-chart-inference.js
//#region src/react/genie/genie-chart-inference.ts
const INFERENCE_CONFIG = {
	minRows: 2,
	pieMaxCategories: 7,
	barMaxCategories: 100,
	groupedBarMaxCategories: 50
};
function classifyColumns(rows, columns) {
	if (rows.length < INFERENCE_CONFIG.minRows || columns.length < 2) return null;
	const dateCols = columns.filter((c) => c.category === "date");
	const numericCols = columns.filter((c) => c.category === "numeric");
	const stringCols = columns.filter((c) => c.category === "string");
	if (numericCols.length === 0) return null;
	return {
		dateCols,
		numericCols,
		stringCols
	};
}
function countUnique(rows, key) {
	const seen = /* @__PURE__ */ new Set();
	for (const row of rows) seen.add(row[key]);
	return seen.size;
}
function hasNegativeValues(rows, key) {
	for (const row of rows) if (Number(row[key]) < 0) return true;
	return false;
}
/**
* Infer the best chart type for the given Genie query result.
* Returns `null` when the data is not suitable for charting.
*/
function inferChartType(rows, columns) {
	const classified = classifyColumns(rows, columns);
	if (!classified) return null;
	const { dateCols, numericCols, stringCols } = classified;
	if (dateCols.length > 0 && numericCols.length >= 1) return {
		chartType: "line",
		xKey: dateCols[0].name,
		yKey: numericCols.length === 1 ? numericCols[0].name : numericCols.map((c) => c.name)
	};
	if (stringCols.length > 0 && numericCols.length >= 1) {
		const xKey = stringCols[0].name;
		const uniqueCategories = countUnique(rows, xKey);
		if (numericCols.length === 1) {
			const yKey = numericCols[0].name;
			if (uniqueCategories <= INFERENCE_CONFIG.pieMaxCategories && !hasNegativeValues(rows, yKey)) return {
				chartType: "pie",
				xKey,
				yKey
			};
			if (uniqueCategories <= INFERENCE_CONFIG.barMaxCategories) return {
				chartType: "bar",
				xKey,
				yKey
			};
			return {
				chartType: "line",
				xKey,
				yKey
			};
		}
		const yKey = numericCols.map((c) => c.name);
		if (uniqueCategories <= INFERENCE_CONFIG.groupedBarMaxCategories) return {
			chartType: "bar",
			xKey,
			yKey
		};
		return {
			chartType: "line",
			xKey,
			yKey
		};
	}
	if (numericCols.length >= 2 && stringCols.length === 0 && dateCols.length === 0) return {
		chartType: "scatter",
		xKey: numericCols[0].name,
		yKey: numericCols[1].name
	};
	return null;
}
/**
* Return the chart types that are compatible with the given data shape.
* Used to populate the chart-type selector dropdown.
*/
function getCompatibleChartTypes(rows, columns) {
	const classified = classifyColumns(rows, columns);
	if (!classified) return [];
	const { dateCols, numericCols, stringCols } = classified;
	if (dateCols.length > 0 && numericCols.length >= 1) return [
		"line",
		"bar",
		"area"
	];
	if (stringCols.length > 0 && numericCols.length >= 1) {
		const xKey = stringCols[0].name;
		const uniqueCategories = countUnique(rows, xKey);
		if (numericCols.length === 1) {
			const yKey = numericCols[0].name;
			if (uniqueCategories <= INFERENCE_CONFIG.pieMaxCategories && !hasNegativeValues(rows, yKey)) return [
				"pie",
				"donut",
				"bar",
				"line",
				"area"
			];
			return [
				"bar",
				"line",
				"area"
			];
		}
		return [
			"bar",
			"line",
			"area"
		];
	}
	if (numericCols.length >= 2 && stringCols.length === 0 && dateCols.length === 0) return [
		"scatter",
		"line",
		"area"
	];
	return [];
}

//#endregion
export { getCompatibleChartTypes, inferChartType };
