// @databricks/appkit-ui/dist/react/genie/genie-chat.js
import { cn } from "../lib/utils.js";
import { Button } from "../ui/button.js";
import { GenieChatInput } from "./genie-chat-input.js";
import { GenieChatMessageList } from "./genie-chat-message-list.js";
import { useGenieChat } from "./use-genie-chat.js";
import { jsx, jsxs } from "react/jsx-runtime";

//#region src/react/genie/genie-chat.tsx
/** Full-featured chat interface for a single Databricks AI/BI Genie space. Handles message streaming, conversation history, and auto-reconnection via SSE. */
function GenieChat({ alias, basePath, placeholder, className }) {
	const { messages, status, error, sendMessage, reset, hasPreviousPage, fetchPreviousPage } = useGenieChat({
		alias,
		basePath
	});
	return /* @__PURE__ */ jsxs("div", {
		className: cn("flex flex-col h-full overflow-hidden", className),
		children: [
			messages.length > 0 && /* @__PURE__ */ jsx("div", {
				className: "shrink-0 flex justify-end px-4 pt-3 pb-1",
				children: /* @__PURE__ */ jsx(Button, {
					variant: "ghost",
					size: "sm",
					onClick: reset,
					className: "text-xs text-muted-foreground",
					children: "New conversation"
				})
			}),
			/* @__PURE__ */ jsx(GenieChatMessageList, {
				messages,
				status,
				hasPreviousPage,
				onFetchPreviousPage: fetchPreviousPage
			}),
			error && /* @__PURE__ */ jsx("div", {
				className: "shrink-0 px-4 py-2 text-sm text-destructive bg-destructive/10 border-t",
				children: error
			}),
			/* @__PURE__ */ jsx(GenieChatInput, {
				onSend: sendMessage,
				disabled: status === "streaming" || status === "loading-history",
				placeholder
			})
		]
	});
}

//#endregion
export { GenieChat };
