// @databricks/appkit-ui/dist/react/genie/genie-chat-message.js
import { cn } from "../lib/utils.js";
import { Card } from "../ui/card.js";
import { Avatar, AvatarFallback } from "../ui/avatar.js";
import { GenieQueryVisualization } from "./genie-query-visualization.js";
import { useMemo } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
import DOMPurify from "dompurify";
import { marked } from "marked";

//#region src/react/genie/genie-chat-message.tsx
/**
* Using `marked` instead of `react-markdown` because `react-markdown` depends on
* `micromark-util-symbol` which has broken ESM exports with `rolldown-vite`.
* Output is sanitized with DOMPurify before being passed to `dangerouslySetInnerHTML`.
*/
marked.setOptions({
	breaks: true,
	gfm: true
});
const markdownStyles = cn("text-sm break-words", "[&_p]:my-1 [&_ul]:my-1 [&_ol]:my-1 [&_li]:my-0", "[&_pre]:bg-background/50 [&_pre]:p-2 [&_pre]:rounded [&_pre]:text-xs [&_pre]:overflow-x-auto", "[&_code]:text-xs [&_code]:bg-background/50 [&_code]:px-1 [&_code]:rounded", "[&_table]:text-xs [&_table]:block [&_table]:overflow-x-auto [&_table]:max-w-full", "[&_th]:px-2 [&_th]:py-1 [&_td]:px-2 [&_td]:py-1", "[&_table]:border-collapse [&_th]:border [&_td]:border", "[&_th]:border-border [&_td]:border-border", "[&_a]:underline");
function isQueryAttachment(att) {
	return !!(att.query?.title || att.query?.query);
}
/** Renders a single Genie message bubble with optional expandable SQL query attachments. */
function GenieChatMessage({ message, className }) {
	const isUser = message.role === "user";
	const queryAttachments = message.attachments.filter(isQueryAttachment);
	const html = useMemo(() => message.content ? DOMPurify.sanitize(marked.parse(message.content)) : "", [message.content]);
	return /* @__PURE__ */ jsxs("div", {
		className: cn("flex gap-3", isUser ? "flex-row-reverse" : "flex-row", className),
		children: [/* @__PURE__ */ jsx(Avatar, {
			className: "h-8 w-8 shrink-0 mt-1",
			children: /* @__PURE__ */ jsx(AvatarFallback, {
				className: cn("text-xs font-medium", isUser ? "bg-primary text-primary-foreground" : "bg-muted"),
				children: isUser ? "You" : "AI"
			})
		}), /* @__PURE__ */ jsxs("div", {
			className: "flex flex-col gap-2 max-w-[80%] min-w-0 overflow-hidden",
			children: [/* @__PURE__ */ jsxs(Card, {
				className: cn("w-full px-4 py-3 overflow-hidden", isUser ? "bg-primary text-primary-foreground [&_*::selection]:bg-primary-foreground/30 [&::selection]:bg-primary-foreground/30" : "bg-muted"),
				children: [html && /* @__PURE__ */ jsx("div", {
					className: markdownStyles,
					dangerouslySetInnerHTML: { __html: html }
				}), message.error && /* @__PURE__ */ jsx("p", {
					className: "text-sm text-destructive mt-1",
					children: message.error
				})]
			}), queryAttachments.length > 0 && /* @__PURE__ */ jsx("div", {
				className: "flex flex-col gap-2 w-full min-w-0",
				children: queryAttachments.map((att) => {
					const key = att.attachmentId ?? "query";
					const queryResult = att.attachmentId ? message.queryResults.get(att.attachmentId) : void 0;
					return /* @__PURE__ */ jsxs("div", {
						className: "flex flex-col gap-2",
						children: [/* @__PURE__ */ jsx(Card, {
							className: "px-4 py-3 text-xs overflow-hidden shadow-none",
							children: /* @__PURE__ */ jsxs("details", { children: [/* @__PURE__ */ jsx("summary", {
								className: "cursor-pointer select-none font-medium",
								children: att.query?.title ?? "SQL Query"
							}), /* @__PURE__ */ jsxs("div", {
								className: "mt-2 flex flex-col gap-1",
								children: [att.query?.description && /* @__PURE__ */ jsx("span", {
									className: "text-muted-foreground",
									children: att.query.description
								}), att.query?.query && /* @__PURE__ */ jsx("pre", {
									className: "mt-1 p-2 rounded bg-background text-[11px] whitespace-pre-wrap break-all",
									children: att.query.query
								})]
							})] })
						}), queryResult != null && /* @__PURE__ */ jsx(Card, {
							className: "w-full px-4 py-3 overflow-hidden",
							children: /* @__PURE__ */ jsx(GenieQueryVisualization, { data: queryResult })
						})]
					}, key);
				})
			})]
		})]
	});
}

//#endregion
export { GenieChatMessage };
