// @databricks/appkit-ui/dist/react/genie/genie-query-visualization.js
import { BaseChart } from "../charts/base.js";
import { ChartErrorBoundary } from "../charts/chart-error-boundary.js";
import { cn } from "../lib/utils.js";
import { Button } from "../ui/button.js";
import { getCompatibleChartTypes, inferChartType } from "./genie-chart-inference.js";
import { DropdownMenu, DropdownMenuContent, DropdownMenuLabel, DropdownMenuRadioGroup, DropdownMenuRadioItem, DropdownMenuTrigger } from "../ui/dropdown-menu.js";
import { Table, TableBody, TableCell, TableHead, TableHeader, TableRow } from "../ui/table.js";
import { Tabs, TabsContent, TabsList, TabsTrigger } from "../ui/tabs.js";
import { transformGenieData } from "./genie-query-transform.js";
import { useMemo, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
import { BarChart3Icon, ChevronDownIcon } from "lucide-react";

//#region src/react/genie/genie-query-visualization.tsx
const TABLE_ROW_LIMIT = 50;
const CHART_HEIGHT = 250;
const CHART_TYPE_LABELS = {
	bar: "Bar",
	line: "Line",
	area: "Area",
	pie: "Pie",
	donut: "Donut",
	scatter: "Scatter",
	radar: "Radar",
	heatmap: "Heatmap"
};
/**
* Renders a chart + data table for a Genie query result.
*
* - When a chart type can be inferred: shows Tabs with "Chart" (default) and "Table"
* - When no chart fits: shows only the data table
* - When data is empty/malformed: renders nothing
*/
function GenieQueryVisualization({ data, className }) {
	const transformed = useMemo(() => transformGenieData(data), [data]);
	const { inference, compatibleTypes } = useMemo(() => {
		if (!transformed) return {
			inference: null,
			compatibleTypes: []
		};
		const { rows, columns } = transformed;
		return {
			inference: inferChartType(rows, columns),
			compatibleTypes: getCompatibleChartTypes(rows, columns)
		};
	}, [transformed]);
	const [chartTypeOverride, setChartTypeOverride] = useState(null);
	if (!transformed || transformed.rows.length === 0) return null;
	const { rows, columns } = transformed;
	const truncated = rows.length > TABLE_ROW_LIMIT;
	const displayRows = truncated ? rows.slice(0, TABLE_ROW_LIMIT) : rows;
	const activeChartType = chartTypeOverride && compatibleTypes.includes(chartTypeOverride) ? chartTypeOverride : inference?.chartType ?? null;
	const dataTable = /* @__PURE__ */ jsxs("div", {
		className: "overflow-auto max-h-[300px]",
		children: [/* @__PURE__ */ jsxs(Table, { children: [/* @__PURE__ */ jsx(TableHeader, { children: /* @__PURE__ */ jsx(TableRow, { children: columns.map((col) => /* @__PURE__ */ jsx(TableHead, { children: col.name }, col.name)) }) }), /* @__PURE__ */ jsx(TableBody, { children: displayRows.map((row, i) => /* @__PURE__ */ jsx(TableRow, { children: columns.map((col) => /* @__PURE__ */ jsx(TableCell, { children: row[col.name] != null ? String(row[col.name]) : "" }, col.name)) }, i)) })] }), truncated && /* @__PURE__ */ jsxs("p", {
			className: "text-xs text-muted-foreground px-2 py-1",
			children: [
				"Showing ",
				TABLE_ROW_LIMIT,
				" of ",
				rows.length,
				" rows"
			]
		})]
	});
	if (!inference || !activeChartType) return /* @__PURE__ */ jsx("div", {
		className: cn("min-w-0", className),
		children: dataTable
	});
	return /* @__PURE__ */ jsxs(Tabs, {
		defaultValue: "chart",
		className: cn("min-w-0", className),
		children: [/* @__PURE__ */ jsxs("div", {
			className: "flex items-center justify-between",
			children: [/* @__PURE__ */ jsxs(TabsList, { children: [/* @__PURE__ */ jsx(TabsTrigger, {
				value: "chart",
				children: "Chart"
			}), /* @__PURE__ */ jsx(TabsTrigger, {
				value: "table",
				children: "Table"
			})] }), compatibleTypes.length > 1 && /* @__PURE__ */ jsxs(DropdownMenu, { children: [/* @__PURE__ */ jsx(DropdownMenuTrigger, {
				asChild: true,
				children: /* @__PURE__ */ jsxs(Button, {
					variant: "ghost",
					size: "icon-sm",
					"aria-label": "Change chart type",
					className: "gap-0.5",
					children: [/* @__PURE__ */ jsx(BarChart3Icon, { className: "size-3.5" }), /* @__PURE__ */ jsx(ChevronDownIcon, { className: "size-3" })]
				})
			}), /* @__PURE__ */ jsxs(DropdownMenuContent, {
				align: "end",
				children: [/* @__PURE__ */ jsx(DropdownMenuLabel, { children: "Chart type" }), /* @__PURE__ */ jsx(DropdownMenuRadioGroup, {
					value: activeChartType,
					onValueChange: (v) => setChartTypeOverride(v),
					children: compatibleTypes.map((type) => /* @__PURE__ */ jsx(DropdownMenuRadioItem, {
						value: type,
						children: CHART_TYPE_LABELS[type]
					}, type))
				})]
			})] })]
		}), /* @__PURE__ */ jsxs("div", {
			className: "grid min-w-0 [&>*]:col-start-1 [&>*]:row-start-1 [&>*]:min-w-0",
			children: [/* @__PURE__ */ jsx(TabsContent, {
				value: "chart",
				forceMount: true,
				className: "data-[state=inactive]:invisible",
				children: /* @__PURE__ */ jsx(ChartErrorBoundary, {
					fallback: dataTable,
					children: /* @__PURE__ */ jsx(BaseChart, {
						data: rows,
						chartType: activeChartType,
						xKey: inference.xKey,
						yKey: inference.yKey,
						height: CHART_HEIGHT,
						showLegend: Array.isArray(inference.yKey)
					})
				})
			}), /* @__PURE__ */ jsx(TabsContent, {
				value: "table",
				forceMount: true,
				className: "data-[state=inactive]:invisible",
				children: dataTable
			})]
		})]
	});
}

//#endregion
export { GenieQueryVisualization };
