// @databricks/appkit-ui/dist/react/genie/genie-chat-input.js
import { cn } from "../lib/utils.js";
import { Button } from "../ui/button.js";
import { useRef, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";

//#region src/react/genie/genie-chat-input.tsx
/** Auto-expanding textarea input with a send button for chat messages. Submits on Enter (Shift+Enter for newline). */
function GenieChatInput({ onSend, disabled = false, placeholder = "Ask a question...", className }) {
	const [value, setValue] = useState("");
	const textareaRef = useRef(null);
	const handleSubmit = () => {
		const trimmed = value.trim();
		if (!trimmed || disabled) return;
		onSend(trimmed);
		setValue("");
		if (textareaRef.current) textareaRef.current.style.height = "auto";
	};
	const handleKeyDown = (e) => {
		if (e.key === "Enter" && !e.shiftKey) {
			e.preventDefault();
			handleSubmit();
		}
	};
	const MAX_HEIGHT = 200;
	const handleInput = () => {
		const textarea = textareaRef.current;
		if (textarea) {
			textarea.style.height = "auto";
			const clamped = Math.min(textarea.scrollHeight, MAX_HEIGHT);
			textarea.style.height = `${clamped}px`;
			textarea.style.overflowY = textarea.scrollHeight > MAX_HEIGHT ? "auto" : "hidden";
		}
	};
	return /* @__PURE__ */ jsxs("div", {
		className: cn("flex gap-2 p-4 border-t shrink-0", className),
		children: [/* @__PURE__ */ jsx("textarea", {
			ref: textareaRef,
			value,
			onChange: (e) => setValue(e.target.value),
			onKeyDown: handleKeyDown,
			onInput: handleInput,
			placeholder,
			disabled,
			rows: 1,
			className: cn("flex-1 resize-none overflow-hidden rounded-md border border-input bg-background px-3 py-2", "text-sm placeholder:text-muted-foreground", "focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring", "disabled:cursor-not-allowed disabled:opacity-50")
		}), /* @__PURE__ */ jsx(Button, {
			onClick: handleSubmit,
			disabled: disabled || !value.trim(),
			size: "default",
			className: "self-end",
			children: "Send"
		})]
	});
}

//#endregion
export { GenieChatInput };
