// @databricks/appkit-ui/dist/react/genie/genie-chat-message-list.js
import { cn } from "../lib/utils.js";
import { Skeleton } from "../ui/skeleton.js";
import { ScrollArea } from "../ui/scroll-area.js";
import { Spinner } from "../ui/spinner.js";
import { GenieChatMessage } from "./genie-chat-message.js";
import { TERMINAL_STATUSES } from "./types.js";
import { useEffect, useLayoutEffect, useRef } from "react";
import { jsx, jsxs } from "react/jsx-runtime";

//#region src/react/genie/genie-chat-message-list.tsx
const STATUS_LABELS = {
	ASKING_AI: "Asking AI...",
	EXECUTING_QUERY: "Executing query...",
	FILTERING_RESULTS: "Filtering results...",
	COMPLETED: "Done"
};
function formatStatus(status) {
	return STATUS_LABELS[status] ?? status.replace(/_/g, " ").toLowerCase();
}
function getViewport(scrollRef) {
	return scrollRef.current?.querySelector("[data-slot=\"scroll-area-viewport\"]");
}
/**
* Manages scroll position: scrolls to bottom on append/initial load,
* preserves position when older messages are prepended.
*/
function useScrollManagement(scrollRef, messages, status) {
	const prevFirstMessageIdRef = useRef(null);
	const prevScrollHeightRef = useRef(0);
	const prevMessageCountRef = useRef(0);
	useEffect(() => {
		const viewport = getViewport(scrollRef);
		if (!viewport) return;
		const observer = new ResizeObserver(() => {
			prevScrollHeightRef.current = viewport.scrollHeight;
		});
		observer.observe(viewport);
		return () => observer.disconnect();
	}, [scrollRef]);
	useLayoutEffect(() => {
		const viewport = getViewport(scrollRef);
		if (!viewport) return;
		const count = messages.length;
		const countChanged = count !== prevMessageCountRef.current;
		prevMessageCountRef.current = count;
		if (!countChanged) {
			prevScrollHeightRef.current = viewport.scrollHeight;
			return;
		}
		const firstMessageId = messages[0]?.id ?? null;
		if (prevFirstMessageIdRef.current !== null && firstMessageId !== prevFirstMessageIdRef.current && prevScrollHeightRef.current > 0) {
			const delta = viewport.scrollHeight - prevScrollHeightRef.current;
			viewport.scrollTop += delta;
		} else viewport.scrollTop = viewport.scrollHeight;
		prevFirstMessageIdRef.current = firstMessageId;
		prevScrollHeightRef.current = viewport.scrollHeight;
	}, [messages.length, status]);
}
/**
* Observes a sentinel element at the top of the scroll area and triggers
* `onFetchPreviousPage` when the user scrolls to the top (only if content overflows).
* Returns a ref to attach to the sentinel element.
*/
function useLoadOlderOnScroll(scrollRef, shouldObserve, onFetchPreviousPage) {
	const sentinelRef = useRef(null);
	const onFetchPreviousPageRef = useRef(onFetchPreviousPage);
	onFetchPreviousPageRef.current = onFetchPreviousPage;
	useEffect(() => {
		const sentinel = sentinelRef.current;
		const viewport = getViewport(scrollRef);
		if (!sentinel || !viewport || !shouldObserve) return;
		let armed = false;
		const frameId = requestAnimationFrame(() => {
			armed = true;
		});
		const observer = new IntersectionObserver((entries) => {
			if (!armed) return;
			const isScrollable = viewport.scrollHeight > viewport.clientHeight;
			if (entries[0]?.isIntersecting && isScrollable) onFetchPreviousPageRef.current?.();
		}, {
			root: viewport,
			threshold: 0
		});
		observer.observe(sentinel);
		return () => {
			cancelAnimationFrame(frameId);
			observer.disconnect();
		};
	}, [scrollRef, shouldObserve]);
	return sentinelRef;
}
/** Scrollable message list that renders Genie chat messages with auto-scroll, skeleton loaders, and a streaming indicator. */
function GenieChatMessageList({ messages, status, className, hasPreviousPage = false, onFetchPreviousPage }) {
	const scrollRef = useRef(null);
	const sentinelRef = useLoadOlderOnScroll(scrollRef, hasPreviousPage && status !== "loading-older", onFetchPreviousPage);
	useScrollManagement(scrollRef, messages, status);
	const lastMessage = messages[messages.length - 1];
	const showStreamingIndicator = status === "streaming" && lastMessage?.role === "assistant" && !lastMessage.content && !TERMINAL_STATUSES.has(lastMessage.status);
	return /* @__PURE__ */ jsx(ScrollArea, {
		ref: scrollRef,
		className: cn("flex-1 min-h-0 p-4 [&_[data-slot=scroll-area-viewport]>div]:!block", className),
		children: /* @__PURE__ */ jsxs("div", {
			className: "flex flex-col gap-4 min-w-0",
			children: [
				hasPreviousPage && /* @__PURE__ */ jsx("div", {
					ref: sentinelRef,
					className: "h-px"
				}),
				status === "loading-older" && /* @__PURE__ */ jsxs("div", {
					className: "flex items-center justify-center gap-2 py-2",
					children: [/* @__PURE__ */ jsx(Spinner, { className: "h-3 w-3" }), /* @__PURE__ */ jsx("span", {
						className: "text-sm text-muted-foreground",
						children: "Loading older messages..."
					})]
				}),
				status === "loading-history" && messages.length === 0 && /* @__PURE__ */ jsxs("div", {
					className: "flex flex-col gap-4",
					children: [
						/* @__PURE__ */ jsx(Skeleton, { className: "h-12 w-3/4" }),
						/* @__PURE__ */ jsx(Skeleton, { className: "h-20 w-4/5 self-start" }),
						/* @__PURE__ */ jsx(Skeleton, { className: "h-12 w-2/3 self-end" })
					]
				}),
				messages.filter((msg) => msg.role !== "assistant" || msg.content || msg.id !== "" && TERMINAL_STATUSES.has(msg.status)).map((msg) => /* @__PURE__ */ jsx(GenieChatMessage, { message: msg }, msg.id)),
				showStreamingIndicator && /* @__PURE__ */ jsxs("div", {
					className: "flex items-center gap-2 text-sm text-muted-foreground px-11",
					children: [/* @__PURE__ */ jsx(Spinner, { className: "h-3 w-3" }), /* @__PURE__ */ jsx("span", { children: formatStatus(lastMessage.status) })]
				}),
				messages.length === 0 && status === "idle" && /* @__PURE__ */ jsx("div", {
					className: "flex items-center justify-center h-full text-muted-foreground text-sm py-12",
					children: "Start a conversation by typing a question below."
				})
			]
		})
	});
}

//#endregion
export { GenieChatMessageList };
