// @databricks/appkit/dist/plugins/serving/manifest.js
//#region src/plugins/serving/manifest.json
var manifest_default = {
	$schema: "https://databricks.github.io/appkit/schemas/plugin-manifest.schema.json",
	name: "serving",
	displayName: "Model Serving Plugin",
	description: "Authenticated proxy to Databricks Model Serving endpoints",
	resources: {
		"required": [{
			"type": "serving_endpoint",
			"alias": "Serving Endpoint",
			"resourceKey": "serving-endpoint",
			"description": "Model Serving endpoint for inference",
			"permission": "CAN_QUERY",
			"fields": { "name": {
				"env": "DATABRICKS_SERVING_ENDPOINT_NAME",
				"description": "Serving endpoint name"
			} }
		}],
		"optional": []
	},
	config: { "schema": {
		"type": "object",
		"properties": {
			"endpoints": {
				"type": "object",
				"description": "Map of alias names to endpoint configurations",
				"additionalProperties": {
					"type": "object",
					"properties": {
						"env": {
							"type": "string",
							"description": "Environment variable holding the endpoint name"
						},
						"servedModel": {
							"type": "string",
							"description": "Target a specific served model (bypasses traffic routing)"
						}
					},
					"required": ["env"]
				}
			},
			"timeout": {
				"type": "number",
				"default": 12e4,
				"description": "Request timeout in ms. Default: 120000 (2 min)"
			}
		}
	} }
};

//#endregion
export { manifest_default as default };
