// @databricks/appkit/dist/plugins/serving/serving.js
import { createLogger } from "../../logging/logger.js";
import { getWorkspaceClient } from "../../context/execution-context.js";
import "../../context/index.js";
import { Plugin } from "../../plugin/plugin.js";
import { ResourceType } from "../../registry/types.generated.js";
import "../../registry/index.js";
import { toPlugin } from "../../plugin/to-plugin.js";
import "../../plugin/index.js";
import "../../logging/index.js";
import { invoke, stream } from "../../connectors/serving/client.js";
import { servingInvokeDefaults } from "./defaults.js";
import manifest_default from "./manifest.js";
import { filterRequestBody, loadEndpointSchemas } from "./schema-filter.js";
import path from "node:path";
import { Readable } from "node:stream";
import { pipeline } from "node:stream/promises";

//#region src/plugins/serving/serving.ts
const logger = createLogger("serving");
var EndpointNotFoundError = class extends Error {
	constructor(alias) {
		super(`Unknown endpoint alias: ${alias}`);
	}
};
var EndpointNotConfiguredError = class extends Error {
	constructor(alias, envVar) {
		super(`Endpoint '${alias}' is not configured: env var '${envVar}' is not set`);
	}
};
var ServingPlugin = class extends Plugin {
	static manifest = manifest_default;
	static description = "Authenticated proxy to Databricks Model Serving endpoints";
	endpoints;
	isNamedMode;
	schemaAllowlists = /* @__PURE__ */ new Map();
	constructor(config) {
		super(config);
		this.config = config;
		if (config.endpoints) {
			this.endpoints = config.endpoints;
			this.isNamedMode = true;
		} else {
			this.endpoints = { default: { env: "DATABRICKS_SERVING_ENDPOINT_NAME" } };
			this.isNamedMode = false;
		}
	}
	async setup() {
		this.schemaAllowlists = await loadEndpointSchemas(path.join(process.cwd(), "node_modules", ".databricks", "appkit", ".appkit-serving-types-cache.json"));
		if (this.schemaAllowlists.size > 0) logger.debug("Loaded schema allowlists for %d endpoint(s)", this.schemaAllowlists.size);
	}
	static getResourceRequirements(config) {
		const endpoints = config.endpoints ?? { default: { env: "DATABRICKS_SERVING_ENDPOINT_NAME" } };
		return Object.entries(endpoints).map(([alias, endpointConfig]) => ({
			type: ResourceType.SERVING_ENDPOINT,
			alias: `serving-${alias}`,
			resourceKey: `serving-${alias}`,
			description: `Model Serving endpoint for "${alias}" inference`,
			permission: "CAN_QUERY",
			fields: { name: {
				env: endpointConfig.env,
				description: `Serving endpoint name for "${alias}"`
			} },
			required: true
		}));
	}
	resolveAndFilter(alias, body) {
		const config = this.endpoints[alias];
		if (!config) throw new EndpointNotFoundError(alias);
		const name = process.env[config.env];
		if (!name) throw new EndpointNotConfiguredError(alias, config.env);
		return {
			endpoint: { name },
			filteredBody: filterRequestBody(body, this.schemaAllowlists, alias, this.config.filterMode)
		};
	}
	injectRoutes(router) {
		if (this.isNamedMode) {
			this.route(router, {
				name: "invoke",
				method: "post",
				path: "/:alias/invoke",
				handler: async (req, res) => {
					await this.asUser(req)._handleInvoke(req, res);
				}
			});
			this.route(router, {
				name: "stream",
				method: "post",
				path: "/:alias/stream",
				handler: async (req, res) => {
					await this.asUser(req)._handleStream(req, res);
				}
			});
		} else {
			const invokeHandler = async (req, res) => {
				req.params.alias ??= "default";
				await this.asUser(req)._handleInvoke(req, res);
			};
			const streamHandler = async (req, res) => {
				req.params.alias ??= "default";
				await this.asUser(req)._handleStream(req, res);
			};
			this.route(router, {
				name: "invoke",
				method: "post",
				path: "/invoke",
				handler: invokeHandler
			});
			this.route(router, {
				name: "invoke-named",
				method: "post",
				path: "/:alias/invoke",
				handler: invokeHandler
			});
			this.route(router, {
				name: "stream",
				method: "post",
				path: "/stream",
				handler: streamHandler
			});
			this.route(router, {
				name: "stream-named",
				method: "post",
				path: "/:alias/stream",
				handler: streamHandler
			});
		}
	}
	async _handleInvoke(req, res) {
		const { alias } = req.params;
		const rawBody = req.body;
		try {
			const result = await this.invoke(alias, rawBody);
			if (!result.ok) {
				res.status(result.status).json({ error: result.message });
				return;
			}
			res.json(result.data);
		} catch (err) {
			const message = err instanceof Error ? err.message : "Invocation failed";
			if (err instanceof EndpointNotFoundError) res.status(404).json({ error: message });
			else if (err instanceof EndpointNotConfiguredError || message.startsWith("Unknown request parameters:")) res.status(400).json({ error: message });
			else res.status(502).json({ error: message });
		}
	}
	async _handleStream(req, res) {
		const { alias } = req.params;
		const rawBody = req.body;
		let endpoint;
		let filteredBody;
		try {
			({endpoint, filteredBody} = this.resolveAndFilter(alias, rawBody));
		} catch (err) {
			const message = err instanceof Error ? err.message : "Invalid request";
			const status = err instanceof EndpointNotFoundError ? 404 : 400;
			res.status(status).json({ error: message });
			return;
		}
		const timeout = this.config.timeout ?? 12e4;
		const workspaceClient = getWorkspaceClient();
		let rawStream;
		try {
			rawStream = await stream(workspaceClient, endpoint.name, filteredBody);
		} catch (err) {
			const message = err instanceof Error ? err.message : "Streaming request failed";
			res.status(502).json({ error: message });
			return;
		}
		res.setHeader("Content-Type", "text/event-stream; charset=utf-8");
		res.setHeader("Cache-Control", "no-cache, no-transform");
		res.setHeader("X-Accel-Buffering", "no");
		res.flushHeaders();
		const nodeStream = Readable.fromWeb(rawStream);
		const abortController = new AbortController();
		const timeoutId = setTimeout(() => abortController.abort(), timeout);
		req.on("close", () => abortController.abort());
		try {
			await pipeline(nodeStream, res, { signal: abortController.signal });
		} catch (err) {
			if (err instanceof Error && err.name !== "AbortError") logger.warn("Stream pipe error: %s", err.message);
		} finally {
			clearTimeout(timeoutId);
		}
	}
	async invoke(alias, body) {
		const { endpoint, filteredBody } = this.resolveAndFilter(alias, body);
		const workspaceClient = getWorkspaceClient();
		const timeout = this.config.timeout ?? 12e4;
		return this.execute(() => invoke(workspaceClient, endpoint.name, filteredBody), { default: {
			...servingInvokeDefaults,
			timeout
		} });
	}
	clientConfig() {
		return {
			isNamedMode: this.isNamedMode,
			aliases: Object.keys(this.endpoints)
		};
	}
	async shutdown() {
		this.streamManager.abortAll();
	}
	createEndpointAPI(alias) {
		return { invoke: (body) => this.invoke(alias, body) };
	}
	exports() {
		const resolveEndpoint = (alias) => {
			const resolved = alias ?? "default";
			return {
				...this.createEndpointAPI(resolved),
				asUser: (req) => {
					return this.asUser(req).createEndpointAPI(resolved);
				}
			};
		};
		return resolveEndpoint;
	}
};
/**
* @internal
*/
const serving = toPlugin(ServingPlugin);

//#endregion
export { ServingPlugin, serving };
