// @databricks/sdk-experimental/dist/apis/dataquality/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataQualityService = void 0;
const dataquality = __importStar(require("./model"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
class DataQualityRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("DataQuality", method, message);
    }
}
class DataQualityError extends apiError_1.ApiError {
    constructor(method, message) {
        super("DataQuality", method, message);
    }
}
/**
 * Manage the data quality of Unity Catalog objects (currently support `schema`
 * and `table`)
 */
class DataQualityService {
    constructor(client) {
        this.client = client;
    }
    async _cancelRefresh(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = undefined;
        const query = {};
        const path = `/api/data-quality/v1/monitors/${request.object_type}/${request.object_id}/refreshes/${request.refresh_id}/cancel`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Cancels a data quality monitor refresh. Currently only supported for the
     * `table` `object_type`. The call must be made in the same workspace as
     * where the monitor was created.
     *
     * The caller must have either of the following sets of permissions: 1.
     * **MANAGE** and **USE_CATALOG** on the table's parent catalog. 2.
     * **USE_CATALOG** on the table's parent catalog, and **MANAGE** and
     * **USE_SCHEMA** on the table's parent schema. 3. **USE_CATALOG** on the
     * table's parent catalog, **USE_SCHEMA** on the table's parent schema, and
     * **MANAGE** on the table.
     */
    async cancelRefresh(request, context) {
        return await this._cancelRefresh(request, context);
    }
    async _createMonitor(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.monitor;
        const query = {};
        const path = "/api/data-quality/v1/monitors";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a data quality monitor on a Unity Catalog object. The caller must
     * provide either `anomaly_detection_config` for a schema monitor or
     * `data_profiling_config` for a table monitor.
     *
     * For the `table` `object_type`, the caller must have either of the
     * following sets of permissions: 1. **MANAGE** and **USE_CATALOG** on the
     * table's parent catalog, **USE_SCHEMA** on the table's parent schema, and
     * **SELECT** on the table 2. **USE_CATALOG** on the table's parent catalog,
     * **MANAGE** and **USE_SCHEMA** on the table's parent schema, and **SELECT**
     * on the table. 3. **USE_CATALOG** on the table's parent catalog,
     * **USE_SCHEMA** on the table's parent schema, and **MANAGE** and **SELECT**
     * on the table.
     *
     * Workspace assets, such as the dashboard, will be created in the workspace
     * where this call was made.
     *
     * For the `schema` `object_type`, the caller must have either of the
     * following sets of permissions: 1. **MANAGE** and **USE_CATALOG** on the
     * schema's parent catalog. 2. **USE_CATALOG** on the schema's parent
     * catalog, and **MANAGE** and **USE_SCHEMA** on the schema.
     */
    async createMonitor(request, context) {
        return await this._createMonitor(request, context);
    }
    async _createRefresh(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.refresh;
        const query = {};
        const path = `/api/data-quality/v1/monitors/${request.object_type}/${request.object_id}/refreshes`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a refresh. Currently only supported for the `table` `object_type`.
     * The call must be made in the same workspace as where the monitor was
     * created.
     *
     * The caller must have either of the following sets of permissions: 1.
     * **MANAGE** and **USE_CATALOG** on the table's parent catalog. 2.
     * **USE_CATALOG** on the table's parent catalog, and **MANAGE** and
     * **USE_SCHEMA** on the table's parent schema. 3. **USE_CATALOG** on the
     * table's parent catalog, **USE_SCHEMA** on the table's parent schema, and
     * **MANAGE** on the table.
     */
    async createRefresh(request, context) {
        return await this._createRefresh(request, context);
    }
    async _deleteMonitor(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/data-quality/v1/monitors/${request.object_type}/${request.object_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a data quality monitor on Unity Catalog object.
     *
     * For the `table` `object_type`, the caller must have either of the
     * following sets of permissions: **MANAGE** and **USE_CATALOG** on the
     * table's parent catalog. **USE_CATALOG** on the table's parent catalog, and
     * **MANAGE** and **USE_SCHEMA** on the table's parent schema.
     * **USE_CATALOG** on the table's parent catalog, **USE_SCHEMA** on the
     * table's parent schema, and **MANAGE** on the table.
     *
     * Note that the metric tables and dashboard will not be deleted as part of
     * this call; those assets must be manually cleaned up (if desired).
     *
     * For the `schema` `object_type`, the caller must have either of the
     * following sets of permissions: 1. **MANAGE** and **USE_CATALOG** on the
     * schema's parent catalog. 2. **USE_CATALOG** on the schema's parent
     * catalog, and **MANAGE** and **USE_SCHEMA** on the schema.
     */
    async deleteMonitor(request, context) {
        return await this._deleteMonitor(request, context);
    }
    async _deleteRefresh(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/data-quality/v1/monitors/${request.object_type}/${request.object_id}/refreshes/${request.refresh_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * (Unimplemented) Delete a refresh
     */
    async deleteRefresh(request, context) {
        return await this._deleteRefresh(request, context);
    }
    async _getMonitor(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/data-quality/v1/monitors/${request.object_type}/${request.object_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Read a data quality monitor on a Unity Catalog object.
     *
     * For the `table` `object_type`, the caller must have either of the
     * following sets of permissions: 1. **MANAGE** and **USE_CATALOG** on the
     * table's parent catalog. 2. **USE_CATALOG** on the table's parent catalog,
     * and **MANAGE** and **USE_SCHEMA** on the table's parent schema. 3.
     * **USE_CATALOG** on the table's parent catalog, **USE_SCHEMA** on the
     * table's parent schema, and **SELECT** on the table.
     *
     * For the `schema` `object_type`, the caller must have either of the
     * following sets of permissions: 1. **MANAGE** and **USE_CATALOG** on the
     * schema's parent catalog. 2. **USE_CATALOG** on the schema's parent
     * catalog, and **USE_SCHEMA** on the schema.
     *
     * The returned information includes configuration values on the entity and
     * parent entity as well as information on assets created by the monitor.
     * Some information (e.g. dashboard) may be filtered out if the caller is in
     * a different workspace than where the monitor was created.
     */
    async getMonitor(request, context) {
        return await this._getMonitor(request, context);
    }
    async _getRefresh(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/data-quality/v1/monitors/${request.object_type}/${request.object_id}/refreshes/${request.refresh_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get data quality monitor refresh. The call must be made in the same
     * workspace as where the monitor was created.
     *
     * For the `table` `object_type`, the caller must have either of the
     * following sets of permissions: 1. **MANAGE** and **USE_CATALOG** on the
     * table's parent catalog. 2. **USE_CATALOG** on the table's parent catalog,
     * and **MANAGE** and **USE_SCHEMA** on the table's parent schema. 3.
     * **USE_CATALOG** on the table's parent catalog, **USE_SCHEMA** on the
     * table's parent schema, and **SELECT** on the table.
     *
     * For the `schema` `object_type`, the caller must have either of the
     * following sets of permissions: 1. **MANAGE** and **USE_CATALOG** on the
     * schema's parent catalog. 2. **USE_CATALOG** on the schema's parent
     * catalog, and **USE_SCHEMA** on the schema.
     */
    async getRefresh(request, context) {
        return await this._getRefresh(request, context);
    }
    async _listMonitor(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/data-quality/v1/monitors";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * (Unimplemented) List data quality monitors.
     */
    async *listMonitor(request, context) {
        while (true) {
            const response = await this._listMonitor(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.monitors || response.monitors.length === 0) {
                break;
            }
            for (const v of response.monitors) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _listRefresh(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/data-quality/v1/monitors/${request.object_type}/${request.object_id}/refreshes`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List data quality monitor refreshes. The call must be made in the same
     * workspace as where the monitor was created.
     *
     * For the `table` `object_type`, the caller must have either of the
     * following sets of permissions: 1. **MANAGE** and **USE_CATALOG** on the
     * table's parent catalog. 2. **USE_CATALOG** on the table's parent catalog,
     * and **MANAGE** and **USE_SCHEMA** on the table's parent schema. 3.
     * **USE_CATALOG** on the table's parent catalog, **USE_SCHEMA** on the
     * table's parent schema, and **SELECT** on the table.
     *
     * For the `schema` `object_type`, the caller must have either of the
     * following sets of permissions: 1. **MANAGE** and **USE_CATALOG** on the
     * schema's parent catalog. 2. **USE_CATALOG** on the schema's parent
     * catalog, and **USE_SCHEMA** on the schema.
     */
    async *listRefresh(request, context) {
        while (true) {
            const response = await this._listRefresh(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.refreshes || response.refreshes.length === 0) {
                break;
            }
            for (const v of response.refreshes) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _updateMonitor(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.monitor;
        const query = {};
        if (request.hasOwnProperty("update_mask")) {
            query["update_mask"] =
                request.update_mask;
        }
        const path = `/api/data-quality/v1/monitors/${request.object_type}/${request.object_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update a data quality monitor on Unity Catalog object.
     *
     * For the `table` `object_type`, the caller must have either of the
     * following sets of permissions: 1. **MANAGE** and **USE_CATALOG** on the
     * table's parent catalog. 2. **USE_CATALOG** on the table's parent catalog,
     * and **MANAGE** and **USE_SCHEMA** on the table's parent schema. 3.
     * **USE_CATALOG** on the table's parent catalog, **USE_SCHEMA** on the
     * table's parent schema, and **MANAGE** on the table.
     *
     * For the `schema` `object_type`, the caller must have either of the
     * following sets of permissions: 1. **MANAGE** and **USE_CATALOG** on the
     * schema's parent catalog. 2. **USE_CATALOG** on the schema's parent
     * catalog, and **MANAGE** and **USE_SCHEMA** on the schema.
     */
    async updateMonitor(request, context) {
        return await this._updateMonitor(request, context);
    }
    async _updateRefresh(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.refresh;
        const query = {};
        if (request.hasOwnProperty("update_mask")) {
            query["update_mask"] =
                request.update_mask;
        }
        const path = `/api/data-quality/v1/monitors/${request.object_type}/${request.object_id}/refreshes/${request.refresh_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * (Unimplemented) Update a refresh
     */
    async updateRefresh(request, context) {
        return await this._updateRefresh(request, context);
    }
}
exports.DataQualityService = DataQualityService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "_cancelRefresh", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "cancelRefresh", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "_createMonitor", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "createMonitor", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "_createRefresh", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "createRefresh", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "_deleteMonitor", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "deleteMonitor", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "_deleteRefresh", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "deleteRefresh", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "_getMonitor", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "getMonitor", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "_getRefresh", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "getRefresh", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "_listMonitor", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], DataQualityService.prototype, "listMonitor", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "_listRefresh", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], DataQualityService.prototype, "listRefresh", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "_updateMonitor", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "updateMonitor", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "_updateRefresh", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], DataQualityService.prototype, "updateRefresh", null);
