// @databricks/appkit/dist/core/appkit.js
import { createLogger } from "../logging/logger.js";
import { version } from "../appkit/package.js";
import { ServiceContext } from "../context/service-context.js";
import { TelemetryManager } from "../telemetry/telemetry-manager.js";
import "../telemetry/index.js";
import { CacheManager } from "../cache/index.js";
import "../context/index.js";
import { isInternalTelemetryEnabled } from "../internal-telemetry/config.js";
import { TelemetryReporter } from "../internal-telemetry/reporter.js";
import "../internal-telemetry/index.js";
import { isPlainObject } from "../plugin/plugin.js";
import { ResourceType } from "../registry/types.generated.js";
import { ResourceRegistry } from "../registry/resource-registry.js";
import "../registry/index.js";
import { PluginContext, isToolProvider } from "./plugin-context.js";

//#region src/core/appkit.ts
const logger = createLogger("appkit");
var AppKit = class AppKit {
	#pluginInstances = {};
	#setupPromises = [];
	#context;
	constructor(config) {
		const { plugins, ...globalConfig } = config;
		this.#context = new PluginContext();
		const pluginEntries = Object.entries(plugins);
		const corePlugins = pluginEntries.filter(([_, p]) => {
			return (p?.plugin?.phase ?? "normal") === "core";
		});
		const normalPlugins = pluginEntries.filter(([_, p]) => (p?.plugin?.phase ?? "normal") === "normal");
		const deferredPlugins = pluginEntries.filter(([_, p]) => (p?.plugin?.phase ?? "normal") === "deferred");
		for (const [name, pluginData] of corePlugins) if (pluginData) this.createAndRegisterPlugin(globalConfig, name, pluginData, { context: this.#context });
		for (const [name, pluginData] of normalPlugins) if (pluginData) this.createAndRegisterPlugin(globalConfig, name, pluginData, { context: this.#context });
		for (const [name, pluginData] of deferredPlugins) if (pluginData) this.createAndRegisterPlugin(globalConfig, name, pluginData, { context: this.#context });
	}
	createAndRegisterPlugin(config, name, pluginData, extraData) {
		const { plugin: Plugin, config: pluginConfig } = pluginData;
		const baseConfig = {
			...config,
			...Plugin.DEFAULT_CONFIG,
			...pluginConfig,
			name,
			...extraData
		};
		const pluginInstance = new Plugin(baseConfig);
		if (typeof pluginInstance.attachContext === "function") pluginInstance.attachContext({
			context: this.#context,
			telemetryConfig: baseConfig.telemetry
		});
		this.#pluginInstances[name] = pluginInstance;
		this.#context.registerPlugin(name, pluginInstance);
		if (isToolProvider(pluginInstance)) this.#context.registerToolProvider(name, pluginInstance);
		this.#setupPromises.push(pluginInstance.setup());
		const self = this;
		Object.defineProperty(this, name, {
			get() {
				const plugin = self.#pluginInstances[name];
				return self.wrapWithAsUser(plugin);
			},
			enumerable: true
		});
	}
	/**
	* Binds all function properties in an exports object to the given context.
	* Recurses into plain objects to handle nested APIs (e.g., volume APIs).
	*/
	bindExportMethods(exports, context) {
		for (const key in exports) {
			if (!Object.hasOwn(exports, key)) continue;
			const val = exports[key];
			if (typeof val === "function") exports[key] = val.bind(context);
			else if (isPlainObject(val)) this.bindExportMethods(val, context);
		}
	}
	/**
	* Wraps a plugin's exports with an `asUser` method that returns
	* a user-scoped version of the exports.
	*
	* When `exports()` returns a callable (function), it is returned as-is
	* since the plugin manages its own `asUser` per-call (e.g. files plugin).
	* When it returns a plain object, the standard `asUser` wrapper is added.
	*
	* The OBO-side wrapping lives inside `Plugin.asUser` — calling
	* `plugin.asUser(req).exports()` returns exports whose functions already
	* run inside the user's AsyncLocalStorage scope. AppKit only adapts the
	* shape; it does not own the user-context concept.
	*/
	wrapWithAsUser(plugin) {
		const pluginExports = plugin.exports?.() ?? {};
		if (typeof pluginExports === "function") return pluginExports;
		const objExports = pluginExports;
		this.bindExportMethods(objExports, plugin);
		if (typeof plugin.asUser !== "function") return objExports;
		return {
			...objExports,
			asUser: (req) => plugin.asUser(req).exports()
		};
	}
	static async _createApp(config = {}) {
		TelemetryManager.initialize(config?.telemetry);
		await CacheManager.getInstance(config?.cache);
		const rawPlugins = config.plugins;
		const registry = new ResourceRegistry();
		registry.collectResources(rawPlugins);
		const needsWarehouse = registry.getRequired().some((r) => r.type === ResourceType.SQL_WAREHOUSE);
		await ServiceContext.initialize({ warehouseId: needsWarehouse }, config?.client);
		registry.enforceValidation();
		const instance = new AppKit({ plugins: AppKit.preparePlugins(rawPlugins) });
		await Promise.all(instance.#setupPromises);
		await instance.#context.emitLifecycle("setup:complete");
		const handle = instance;
		if (config.onPluginsReady) {
			logger.debug("Running onPluginsReady hook");
			await config.onPluginsReady(handle);
			logger.debug("onPluginsReady hook completed");
		}
		if (isInternalTelemetryEnabled(config)) AppKit.bootstrapInternalTelemetry();
		const serverPlugin = instance.#pluginInstances.server;
		if (serverPlugin && typeof serverPlugin.start === "function") await serverPlugin.start();
		return handle;
	}
	static bootstrapInternalTelemetry() {
		const serviceCtx = ServiceContext.get();
		const reporter = TelemetryReporter.initialize({
			workspaceId: serviceCtx.workspaceId,
			client: serviceCtx.client,
			appId: process.env.DATABRICKS_CLIENT_ID || "",
			appkitVersion: version
		});
		reporter.start();
		reporter.sendStartup().catch(() => {});
	}
	static preparePlugins(plugins) {
		const result = {};
		for (const currentPlugin of plugins) result[currentPlugin.name] = {
			plugin: currentPlugin.plugin,
			config: currentPlugin.config
		};
		return result;
	}
};
/**
* Bootstraps AppKit with the provided configuration.
*
* Initializes telemetry, cache, and service context, then registers plugins
* in phase order (core, normal, deferred) and awaits their setup.
* If a `onPluginsReady` callback is provided it runs after plugin setup but
* before the server starts, giving you access to the full appkit handle
* for registering custom routes or performing async setup.
* The returned object maps each plugin name to its `exports()` API,
* with an `asUser(req)` method for user-scoped execution.
*
* @returns A `PluginMap` keyed by plugin name with typed exports
*
* @example Minimal server
* ```ts
* import { createApp, server } from "@databricks/appkit";
*
* await createApp({
*   plugins: [server()],
* });
* ```
*
* @example Server with custom routes via onPluginsReady
* ```ts
* import { createApp, server, analytics } from "@databricks/appkit";
*
* await createApp({
*   plugins: [server(), analytics({})],
*   onPluginsReady(appkit) {
*     appkit.server.extend((app) => {
*       app.get("/custom", (_req, res) => res.json({ ok: true }));
*     });
*   },
* });
* ```
*/
async function createApp(config = {}) {
	return AppKit._createApp(config);
}

//#endregion
export { createApp };
