// @databricks/appkit/dist/plugins/server/remote-tunnel/remote-tunnel-controller.js
import { createLogger } from "../../../logging/logger.js";
import { hasDevQuery, isLocalDev, isRemoteTunnelAllowedByEnv, isRemoteTunnelAssetRequest } from "./gate.js";

//#region src/plugins/server/remote-tunnel/remote-tunnel-controller.ts
const logger = createLogger("server:remote-tunnel:controller");
/**
* Controller for the remote tunnel
*
* - Reads files from the dev file reader
* - Manages the remote tunnel manager
* - Sets up the web socket
* - Cleans up the remote tunnel
*/
var RemoteTunnelController = class {
	devFileReader;
	server;
	manager;
	initPromise;
	wsReady;
	constructor(devFileReader) {
		this.devFileReader = devFileReader;
		this.manager = null;
		this.initPromise = null;
		this.wsReady = false;
	}
	/**
	* Set the server instance
	* @param server
	*/
	setServer(server) {
		this.server = server;
		this.maybeSetupWebSocket();
	}
	/** Check if the remote tunnel is active */
	isActive() {
		return this.manager != null;
	}
	/** Check if the remote tunnel is allowed by the environment */
	isAllowedByEnv() {
		return isRemoteTunnelAllowedByEnv();
	}
	/**
	* Middleware for the remote tunnel
	* - Hard blocks in local dev
	* - Blocks when not allowed by env
	* - Handles dev query and asset requests
	* @param req - the request
	* @param res - the response
	* @param next - the next function
	* @returns the next function
	*/
	middleware = async (req, res, next) => {
		if (isLocalDev()) return next();
		if (!this.isAllowedByEnv()) return next();
		const wantsDev = hasDevQuery(req);
		const wantsRemoteAssets = isRemoteTunnelAssetRequest(req);
		if (!wantsDev && !wantsRemoteAssets) return next();
		const manager = await this.getOrInitManager();
		if (!manager) return next();
		if (wantsDev) return manager.devModeMiddleware()(req, res, next);
		try {
			await manager.assetMiddleware()(req, res);
		} catch (error) {
			next(error);
		}
	};
	/** Cleanup the remote tunnel */
	cleanup() {
		try {
			this.manager?.cleanup();
		} finally {
			this.manager = null;
			this.initPromise = null;
			this.wsReady = false;
		}
	}
	/**
	* Get or initialize the remote tunnel manager
	* - If the manager is already initialized, return it
	* - If the manager is not initialized, initialize it
	* - If the manager is not allowed by the environment, return null
	* @returns the remote tunnel manager
	*/
	async getOrInitManager() {
		if (this.manager) return this.manager;
		if (this.initPromise) return await this.initPromise;
		this.initPromise = (async () => {
			if (isLocalDev() || !isRemoteTunnelAllowedByEnv()) return null;
			const remoteTunnelManager = new (await (import("./remote-tunnel-manager.js"))).RemoteTunnelManager(this.devFileReader);
			this.manager = remoteTunnelManager;
			this.maybeSetupWebSocket();
			logger.debug("RemoteTunnel: initialized (on-demand)");
			return remoteTunnelManager;
		})();
		return this.initPromise;
	}
	/**
	* Setup the web socket
	* - If the server is not set, return
	* - If the manager is not set, return
	* - If the web socket is already setup, return
	* - Setup the web socket
	*/
	maybeSetupWebSocket() {
		if (!this.server) return;
		if (!this.manager) return;
		if (this.wsReady) return;
		this.manager.setServer(this.server);
		this.manager.setupWebSocket();
		this.wsReady = true;
		logger.debug("RemoteTunnel: web socket setup complete");
	}
};

//#endregion
export { RemoteTunnelController };
