// @databricks/sdk-experimental/dist/apis/ml/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelRegistryService = exports.ModelRegistryError = exports.ModelRegistryRetriableError = exports.MaterializedFeaturesService = exports.MaterializedFeaturesError = exports.MaterializedFeaturesRetriableError = exports.ForecastingService = exports.ForecastingError = exports.ForecastingRetriableError = exports.FeatureStoreService = exports.FeatureStoreError = exports.FeatureStoreRetriableError = exports.FeatureEngineeringService = exports.FeatureEngineeringError = exports.FeatureEngineeringRetriableError = exports.ExperimentsService = exports.ExperimentsError = exports.ExperimentsRetriableError = void 0;
const ml = __importStar(require("./model"));
const retries_1 = __importDefault(require("../../retries/retries"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
const wait_1 = require("../../wait");
class ExperimentsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Experiments", method, message);
    }
}
exports.ExperimentsRetriableError = ExperimentsRetriableError;
class ExperimentsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Experiments", method, message);
    }
}
exports.ExperimentsError = ExperimentsError;
/**
 * Experiments are the primary unit of organization in MLflow; all MLflow runs
 * belong to an experiment. Each experiment lets you visualize, search, and
 * compare runs, as well as download run artifacts or metadata for analysis in
 * other tools. Experiments are maintained in a Databricks hosted MLflow tracking
 * server.
 *
 * Experiments are located in the workspace file tree. You manage experiments
 * using the same tools you use to manage other workspace objects such as
 * folders, notebooks, and libraries.
 */
class ExperimentsService {
    constructor(client) {
        this.client = client;
    }
    async _createExperiment(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("artifact_location")) {
            body["artifact_location"] = request.artifact_location;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        const query = {};
        const path = "/api/2.0/mlflow/experiments/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates an experiment with a name. Returns the ID of the newly created
     * experiment. Validates that another experiment with the same name does not
     * already exist and fails if another experiment with the same name already
     * exists.
     *
     * Throws `RESOURCE_ALREADY_EXISTS` if an experiment with the given name
     * exists.
     */
    async createExperiment(request, context) {
        return await this._createExperiment(request, context);
    }
    async _createLoggedModel(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("experiment_id")) {
            body["experiment_id"] = request.experiment_id;
        }
        if (request.hasOwnProperty("model_type")) {
            body["model_type"] = request.model_type;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("params")) {
            body["params"] = request.params;
        }
        if (request.hasOwnProperty("source_run_id")) {
            body["source_run_id"] = request.source_run_id;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        const query = {};
        const path = "/api/2.0/mlflow/logged-models";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a logged model.
     */
    async createLoggedModel(request, context) {
        return await this._createLoggedModel(request, context);
    }
    async _createRun(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("experiment_id")) {
            body["experiment_id"] = request.experiment_id;
        }
        if (request.hasOwnProperty("run_name")) {
            body["run_name"] = request.run_name;
        }
        if (request.hasOwnProperty("start_time")) {
            body["start_time"] = request.start_time;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        if (request.hasOwnProperty("user_id")) {
            body["user_id"] = request.user_id;
        }
        const query = {};
        const path = "/api/2.0/mlflow/runs/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new run within an experiment. A run is usually a single
     * execution of a machine learning or data ETL pipeline. MLflow uses runs to
     * track the `mlflowParam`, `mlflowMetric`, and `mlflowRunTag` associated
     * with a single execution.
     */
    async createRun(request, context) {
        return await this._createRun(request, context);
    }
    async _deleteExperiment(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("experiment_id")) {
            body["experiment_id"] = request.experiment_id;
        }
        const query = {};
        const path = "/api/2.0/mlflow/experiments/delete";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Marks an experiment and associated metadata, runs, metrics, params, and
     * tags for deletion. If the experiment uses FileStore, artifacts associated
     * with the experiment are also deleted.
     */
    async deleteExperiment(request, context) {
        return await this._deleteExperiment(request, context);
    }
    async _deleteLoggedModel(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/mlflow/logged-models/${request.model_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a logged model.
     */
    async deleteLoggedModel(request, context) {
        return await this._deleteLoggedModel(request, context);
    }
    async _deleteLoggedModelTag(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/mlflow/logged-models/${request.model_id}/tags/${request.tag_key}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a tag on a logged model.
     */
    async deleteLoggedModelTag(request, context) {
        return await this._deleteLoggedModelTag(request, context);
    }
    async _deleteRun(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        const query = {};
        const path = "/api/2.0/mlflow/runs/delete";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Marks a run for deletion.
     */
    async deleteRun(request, context) {
        return await this._deleteRun(request, context);
    }
    async _deleteRuns(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("experiment_id")) {
            body["experiment_id"] = request.experiment_id;
        }
        if (request.hasOwnProperty("max_runs")) {
            body["max_runs"] = request.max_runs;
        }
        if (request.hasOwnProperty("max_timestamp_millis")) {
            body["max_timestamp_millis"] = request.max_timestamp_millis;
        }
        const query = {};
        const path = "/api/2.0/mlflow/databricks/runs/delete-runs";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Bulk delete runs in an experiment that were created prior to or at the
     * specified timestamp. Deletes at most max_runs per request. To call this
     * API from a Databricks Notebook in Python, you can use the client code
     * snippet on
     */
    async deleteRuns(request, context) {
        return await this._deleteRuns(request, context);
    }
    async _deleteTag(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("key")) {
            body["key"] = request.key;
        }
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        const query = {};
        const path = "/api/2.0/mlflow/runs/delete-tag";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a tag on a run. Tags are run metadata that can be updated during a
     * run and after a run completes.
     */
    async deleteTag(request, context) {
        return await this._deleteTag(request, context);
    }
    async _finalizeLoggedModel(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("status")) {
            body["status"] = request.status;
        }
        const query = {};
        const path = `/api/2.0/mlflow/logged-models/${request.model_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Finalize a logged model.
     */
    async finalizeLoggedModel(request, context) {
        return await this._finalizeLoggedModel(request, context);
    }
    async _getByName(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("experiment_name")) {
            query["experiment_name"] =
                request.experiment_name;
        }
        const path = "/api/2.0/mlflow/experiments/get-by-name";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets metadata for an experiment.
     *
     * This endpoint will return deleted experiments, but prefers the active
     * experiment if an active and deleted experiment share the same name. If
     * multiple deleted experiments share the same name, the API will return one
     * of them.
     *
     * Throws `RESOURCE_DOES_NOT_EXIST` if no experiment with the specified name
     * exists.
     */
    async getByName(request, context) {
        return await this._getByName(request, context);
    }
    async _getExperiment(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("experiment_id")) {
            query["experiment_id"] =
                request.experiment_id;
        }
        const path = "/api/2.0/mlflow/experiments/get";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets metadata for an experiment. This method works on deleted experiments.
     */
    async getExperiment(request, context) {
        return await this._getExperiment(request, context);
    }
    async _getHistory(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("metric_key")) {
            query["metric_key"] =
                request.metric_key;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        if (request.hasOwnProperty("run_id")) {
            query["run_id"] = request.run_id;
        }
        if (request.hasOwnProperty("run_uuid")) {
            query["run_uuid"] =
                request.run_uuid;
        }
        const path = "/api/2.0/mlflow/metrics/get-history";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a list of all values for the specified metric for a given run.
     */
    async *getHistory(request, context) {
        while (true) {
            const response = await this._getHistory(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.metrics || response.metrics.length === 0) {
                break;
            }
            for (const v of response.metrics) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _getLoggedModel(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/mlflow/logged-models/${request.model_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a logged model.
     */
    async getLoggedModel(request, context) {
        return await this._getLoggedModel(request, context);
    }
    async _getPermissionLevels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/experiments/${request.experiment_id}/permissionLevels`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(request, context) {
        return await this._getPermissionLevels(request, context);
    }
    async _getPermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/experiments/${request.experiment_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of an experiment. Experiments can inherit permissions
     * from their root object.
     */
    async getPermissions(request, context) {
        return await this._getPermissions(request, context);
    }
    async _getRun(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("run_id")) {
            query["run_id"] = request.run_id;
        }
        if (request.hasOwnProperty("run_uuid")) {
            query["run_uuid"] =
                request.run_uuid;
        }
        const path = "/api/2.0/mlflow/runs/get";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the metadata, metrics, params, and tags for a run. In the case where
     * multiple metrics with the same key are logged for a run, return only the
     * value with the latest timestamp.
     *
     * If there are multiple values with the latest timestamp, return the maximum
     * of these values.
     */
    async getRun(request, context) {
        return await this._getRun(request, context);
    }
    async _listArtifacts(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        if (request.hasOwnProperty("path")) {
            query["path"] = request.path;
        }
        if (request.hasOwnProperty("run_id")) {
            query["run_id"] = request.run_id;
        }
        if (request.hasOwnProperty("run_uuid")) {
            query["run_uuid"] =
                request.run_uuid;
        }
        const path = "/api/2.0/mlflow/artifacts/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List artifacts for a run. Takes an optional `artifact_path` prefix which
     * if specified, the response contains only artifacts with the specified
     * prefix. A maximum of 1000 artifacts will be retrieved for UC Volumes.
     * Please call `/api/2.0/fs/directories{directory_path}` for listing
     * artifacts in UC Volumes, which supports pagination. See [List directory
     * contents | Files API](/api/workspace/files/listdirectorycontents).
     */
    async *listArtifacts(request, context) {
        while (true) {
            const response = await this._listArtifacts(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.files || response.files.length === 0) {
                break;
            }
            for (const v of response.files) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _listExperiments(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        if (request.hasOwnProperty("view_type")) {
            query["view_type"] =
                request.view_type;
        }
        const path = "/api/2.0/mlflow/experiments/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a list of all experiments.
     */
    async *listExperiments(request, context) {
        while (true) {
            const response = await this._listExperiments(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.experiments || response.experiments.length === 0) {
                break;
            }
            for (const v of response.experiments) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _logBatch(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("metrics")) {
            body["metrics"] = request.metrics;
        }
        if (request.hasOwnProperty("params")) {
            body["params"] = request.params;
        }
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        const query = {};
        const path = "/api/2.0/mlflow/runs/log-batch";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Logs a batch of metrics, params, and tags for a run. If any data failed to
     * be persisted, the server will respond with an error (non-200 status code).
     *
     * In case of error (due to internal server error or an invalid request),
     * partial data may be written.
     *
     * You can write metrics, params, and tags in interleaving fashion, but
     * within a given entity type are guaranteed to follow the order specified in
     * the request body.
     *
     * The overwrite behavior for metrics, params, and tags is as follows:
     *
     * * Metrics: metric values are never overwritten. Logging a metric (key,
     * value, timestamp) appends to the set of values for the metric with the
     * provided key.
     *
     * * Tags: tag values can be overwritten by successive writes to the same tag
     * key. That is, if multiple tag values with the same key are provided in the
     * same API request, the last-provided tag value is written. Logging the same
     * tag (key, value) is permitted. Specifically, logging a tag is idempotent.
     *
     * * Parameters: once written, param values cannot be changed (attempting to
     * overwrite a param value will result in an error). However, logging the
     * same param (key, value) is permitted. Specifically, logging a param is
     * idempotent.
     *
     * Request Limits ------------------------------- A single JSON-serialized
     * API request may be up to 1 MB in size and contain:
     *
     * * No more than 1000 metrics, params, and tags in total
     *
     * * Up to 1000 metrics
     *
     * * Up to 100 params
     *
     * * Up to 100 tags
     *
     * For example, a valid request might contain 900 metrics, 50 params, and 50
     * tags, but logging 900 metrics, 50 params, and 51 tags is invalid.
     *
     * The following limits also apply to metric, param, and tag keys and values:
     *
     * * Metric keys, param keys, and tag keys can be up to 250 characters in
     * length
     *
     * * Parameter and tag values can be up to 250 characters in length
     */
    async logBatch(request, context) {
        return await this._logBatch(request, context);
    }
    async _logInputs(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("datasets")) {
            body["datasets"] = request.datasets;
        }
        if (request.hasOwnProperty("models")) {
            body["models"] = request.models;
        }
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        const query = {};
        const path = "/api/2.0/mlflow/runs/log-inputs";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Logs inputs, such as datasets and models, to an MLflow Run.
     */
    async logInputs(request, context) {
        return await this._logInputs(request, context);
    }
    async _logLoggedModelParams(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("params")) {
            body["params"] = request.params;
        }
        const query = {};
        const path = `/api/2.0/mlflow/logged-models/${request.model_id}/params`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Logs params for a logged model. A param is a key-value pair (string key,
     * string value). Examples include hyperparameters used for ML model
     * training. A param can be logged only once for a logged model, and
     * attempting to overwrite an existing param with a different value will
     * result in an error
     */
    async logLoggedModelParams(request, context) {
        return await this._logLoggedModelParams(request, context);
    }
    async _logMetric(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("dataset_digest")) {
            body["dataset_digest"] = request.dataset_digest;
        }
        if (request.hasOwnProperty("dataset_name")) {
            body["dataset_name"] = request.dataset_name;
        }
        if (request.hasOwnProperty("key")) {
            body["key"] = request.key;
        }
        if (request.hasOwnProperty("model_id")) {
            body["model_id"] = request.model_id;
        }
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        if (request.hasOwnProperty("run_uuid")) {
            body["run_uuid"] = request.run_uuid;
        }
        if (request.hasOwnProperty("step")) {
            body["step"] = request.step;
        }
        if (request.hasOwnProperty("timestamp")) {
            body["timestamp"] = request.timestamp;
        }
        if (request.hasOwnProperty("value")) {
            body["value"] = request.value;
        }
        const query = {};
        const path = "/api/2.0/mlflow/runs/log-metric";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Log a metric for a run. A metric is a key-value pair (string key, float
     * value) with an associated timestamp. Examples include the various metrics
     * that represent ML model accuracy. A metric can be logged multiple times.
     */
    async logMetric(request, context) {
        return await this._logMetric(request, context);
    }
    async _logModel(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("model_json")) {
            body["model_json"] = request.model_json;
        }
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        const query = {};
        const path = "/api/2.0/mlflow/runs/log-model";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * **Note:** the [Create a logged
     * model](/api/workspace/experiments/createloggedmodel) API replaces this
     * endpoint.
     *
     * Log a model to an MLflow Run.
     */
    async logModel(request, context) {
        return await this._logModel(request, context);
    }
    async _logOutputs(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("models")) {
            body["models"] = request.models;
        }
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        const query = {};
        const path = "/api/2.0/mlflow/runs/outputs";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Logs outputs, such as models, from an MLflow Run.
     */
    async logOutputs(request, context) {
        return await this._logOutputs(request, context);
    }
    async _logParam(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("key")) {
            body["key"] = request.key;
        }
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        if (request.hasOwnProperty("run_uuid")) {
            body["run_uuid"] = request.run_uuid;
        }
        if (request.hasOwnProperty("value")) {
            body["value"] = request.value;
        }
        const query = {};
        const path = "/api/2.0/mlflow/runs/log-parameter";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Logs a param used for a run. A param is a key-value pair (string key,
     * string value). Examples include hyperparameters used for ML model training
     * and constant dates and values used in an ETL pipeline. A param can be
     * logged only once for a run.
     */
    async logParam(request, context) {
        return await this._logParam(request, context);
    }
    async _restoreExperiment(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("experiment_id")) {
            body["experiment_id"] = request.experiment_id;
        }
        const query = {};
        const path = "/api/2.0/mlflow/experiments/restore";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Restore an experiment marked for deletion. This also restores associated
     * metadata, runs, metrics, params, and tags. If experiment uses FileStore,
     * underlying artifacts associated with experiment are also restored.
     *
     * Throws `RESOURCE_DOES_NOT_EXIST` if experiment was never created or was
     * permanently deleted.
     */
    async restoreExperiment(request, context) {
        return await this._restoreExperiment(request, context);
    }
    async _restoreRun(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        const query = {};
        const path = "/api/2.0/mlflow/runs/restore";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Restores a deleted run. This also restores associated metadata, runs,
     * metrics, params, and tags.
     *
     * Throws `RESOURCE_DOES_NOT_EXIST` if the run was never created or was
     * permanently deleted.
     */
    async restoreRun(request, context) {
        return await this._restoreRun(request, context);
    }
    async _restoreRuns(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("experiment_id")) {
            body["experiment_id"] = request.experiment_id;
        }
        if (request.hasOwnProperty("max_runs")) {
            body["max_runs"] = request.max_runs;
        }
        if (request.hasOwnProperty("min_timestamp_millis")) {
            body["min_timestamp_millis"] = request.min_timestamp_millis;
        }
        const query = {};
        const path = "/api/2.0/mlflow/databricks/runs/restore-runs";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Bulk restore runs in an experiment that were deleted no earlier than the
     * specified timestamp. Restores at most max_runs per request. To call this
     * API from a Databricks Notebook in Python, you can use the client code
     * snippet on
     */
    async restoreRuns(request, context) {
        return await this._restoreRuns(request, context);
    }
    async _searchExperiments(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("filter")) {
            body["filter"] = request.filter;
        }
        if (request.hasOwnProperty("max_results")) {
            body["max_results"] = request.max_results;
        }
        if (request.hasOwnProperty("order_by")) {
            body["order_by"] = request.order_by;
        }
        if (request.hasOwnProperty("page_token")) {
            body["page_token"] = request.page_token;
        }
        if (request.hasOwnProperty("view_type")) {
            body["view_type"] = request.view_type;
        }
        const query = {};
        const path = "/api/2.0/mlflow/experiments/search";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Searches for experiments that satisfy specified search criteria.
     */
    async *searchExperiments(request, context) {
        while (true) {
            const response = await this._searchExperiments(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.experiments || response.experiments.length === 0) {
                break;
            }
            for (const v of response.experiments) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _searchLoggedModels(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("datasets")) {
            body["datasets"] = request.datasets;
        }
        if (request.hasOwnProperty("experiment_ids")) {
            body["experiment_ids"] = request.experiment_ids;
        }
        if (request.hasOwnProperty("filter")) {
            body["filter"] = request.filter;
        }
        if (request.hasOwnProperty("max_results")) {
            body["max_results"] = request.max_results;
        }
        if (request.hasOwnProperty("order_by")) {
            body["order_by"] = request.order_by;
        }
        if (request.hasOwnProperty("page_token")) {
            body["page_token"] = request.page_token;
        }
        const query = {};
        const path = "/api/2.0/mlflow/logged-models/search";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Search for Logged Models that satisfy specified search criteria.
     */
    async searchLoggedModels(request, context) {
        return await this._searchLoggedModels(request, context);
    }
    async _searchRuns(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("experiment_ids")) {
            body["experiment_ids"] = request.experiment_ids;
        }
        if (request.hasOwnProperty("filter")) {
            body["filter"] = request.filter;
        }
        if (request.hasOwnProperty("max_results")) {
            body["max_results"] = request.max_results;
        }
        if (request.hasOwnProperty("order_by")) {
            body["order_by"] = request.order_by;
        }
        if (request.hasOwnProperty("page_token")) {
            body["page_token"] = request.page_token;
        }
        if (request.hasOwnProperty("run_view_type")) {
            body["run_view_type"] = request.run_view_type;
        }
        const query = {};
        const path = "/api/2.0/mlflow/runs/search";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Searches for runs that satisfy expressions.
     *
     * Search expressions can use `mlflowMetric` and `mlflowParam` keys.
     */
    async *searchRuns(request, context) {
        while (true) {
            const response = await this._searchRuns(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.runs || response.runs.length === 0) {
                break;
            }
            for (const v of response.runs) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _setExperimentTag(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("experiment_id")) {
            body["experiment_id"] = request.experiment_id;
        }
        if (request.hasOwnProperty("key")) {
            body["key"] = request.key;
        }
        if (request.hasOwnProperty("value")) {
            body["value"] = request.value;
        }
        const query = {};
        const path = "/api/2.0/mlflow/experiments/set-experiment-tag";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets a tag on an experiment. Experiment tags are metadata that can be
     * updated.
     */
    async setExperimentTag(request, context) {
        return await this._setExperimentTag(request, context);
    }
    async _setLoggedModelTags(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        const query = {};
        const path = `/api/2.0/mlflow/logged-models/${request.model_id}/tags`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Set tags for a logged model.
     */
    async setLoggedModelTags(request, context) {
        return await this._setLoggedModelTags(request, context);
    }
    async _setPermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/experiments/${request.experiment_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their root object.
     */
    async setPermissions(request, context) {
        return await this._setPermissions(request, context);
    }
    async _setTag(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("key")) {
            body["key"] = request.key;
        }
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        if (request.hasOwnProperty("run_uuid")) {
            body["run_uuid"] = request.run_uuid;
        }
        if (request.hasOwnProperty("value")) {
            body["value"] = request.value;
        }
        const query = {};
        const path = "/api/2.0/mlflow/runs/set-tag";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets a tag on a run. Tags are run metadata that can be updated during a
     * run and after a run completes.
     */
    async setTag(request, context) {
        return await this._setTag(request, context);
    }
    async _updateExperiment(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("experiment_id")) {
            body["experiment_id"] = request.experiment_id;
        }
        if (request.hasOwnProperty("new_name")) {
            body["new_name"] = request.new_name;
        }
        const query = {};
        const path = "/api/2.0/mlflow/experiments/update";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates experiment metadata.
     */
    async updateExperiment(request, context) {
        return await this._updateExperiment(request, context);
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/experiments/${request.experiment_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on an experiment. Experiments can inherit
     * permissions from their root object.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
    async _updateRun(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("end_time")) {
            body["end_time"] = request.end_time;
        }
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        if (request.hasOwnProperty("run_name")) {
            body["run_name"] = request.run_name;
        }
        if (request.hasOwnProperty("run_uuid")) {
            body["run_uuid"] = request.run_uuid;
        }
        if (request.hasOwnProperty("status")) {
            body["status"] = request.status;
        }
        const query = {};
        const path = "/api/2.0/mlflow/runs/update";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates run metadata.
     */
    async updateRun(request, context) {
        return await this._updateRun(request, context);
    }
}
exports.ExperimentsService = ExperimentsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_createExperiment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "createExperiment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_createLoggedModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "createLoggedModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_createRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "createRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_deleteExperiment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "deleteExperiment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_deleteLoggedModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "deleteLoggedModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_deleteLoggedModelTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "deleteLoggedModelTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_deleteRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "deleteRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_deleteRuns", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "deleteRuns", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_deleteTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "deleteTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_finalizeLoggedModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "finalizeLoggedModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_getByName", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "getByName", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_getExperiment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "getExperiment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_getHistory", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ExperimentsService.prototype, "getHistory", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_getLoggedModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "getLoggedModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_getRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "getRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_listArtifacts", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ExperimentsService.prototype, "listArtifacts", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_listExperiments", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ExperimentsService.prototype, "listExperiments", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_logBatch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "logBatch", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_logInputs", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "logInputs", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_logLoggedModelParams", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "logLoggedModelParams", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_logMetric", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "logMetric", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_logModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "logModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_logOutputs", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "logOutputs", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_logParam", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "logParam", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_restoreExperiment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "restoreExperiment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_restoreRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "restoreRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_restoreRuns", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "restoreRuns", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_searchExperiments", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ExperimentsService.prototype, "searchExperiments", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_searchLoggedModels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "searchLoggedModels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_searchRuns", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ExperimentsService.prototype, "searchRuns", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_setExperimentTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "setExperimentTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_setLoggedModelTags", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "setLoggedModelTags", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_setTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "setTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_updateExperiment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "updateExperiment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "_updateRun", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ExperimentsService.prototype, "updateRun", null);
class FeatureEngineeringRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("FeatureEngineering", method, message);
    }
}
exports.FeatureEngineeringRetriableError = FeatureEngineeringRetriableError;
class FeatureEngineeringError extends apiError_1.ApiError {
    constructor(method, message) {
        super("FeatureEngineering", method, message);
    }
}
exports.FeatureEngineeringError = FeatureEngineeringError;
/**
 * [description]
 */
class FeatureEngineeringService {
    constructor(client) {
        this.client = client;
    }
    async _batchCreateMaterializedFeatures(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("requests")) {
            body["requests"] = request.requests;
        }
        const query = {};
        const path = "/api/2.0/feature-engineering/materialized-features:batchCreate";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Batch create materialized features.
     */
    async batchCreateMaterializedFeatures(request, context) {
        return await this._batchCreateMaterializedFeatures(request, context);
    }
    async _createFeature(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.feature;
        const query = {};
        const path = "/api/2.0/feature-engineering/features";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a Feature.
     */
    async createFeature(request, context) {
        return await this._createFeature(request, context);
    }
    async _createMaterializedFeature(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.materialized_feature;
        const query = {};
        const path = "/api/2.0/feature-engineering/materialized-features";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a materialized feature.
     */
    async createMaterializedFeature(request, context) {
        return await this._createMaterializedFeature(request, context);
    }
    async _deleteFeature(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/feature-engineering/features/${request.full_name}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a Feature.
     */
    async deleteFeature(request, context) {
        return await this._deleteFeature(request, context);
    }
    async _deleteMaterializedFeature(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/feature-engineering/materialized-features/${request.materialized_feature_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a materialized feature.
     */
    async deleteMaterializedFeature(request, context) {
        return await this._deleteMaterializedFeature(request, context);
    }
    async _getFeature(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/feature-engineering/features/${request.full_name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a Feature.
     */
    async getFeature(request, context) {
        return await this._getFeature(request, context);
    }
    async _getMaterializedFeature(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/feature-engineering/materialized-features/${request.materialized_feature_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a materialized feature.
     */
    async getMaterializedFeature(request, context) {
        return await this._getMaterializedFeature(request, context);
    }
    async _listFeatures(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/feature-engineering/features";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List Features.
     */
    async *listFeatures(request, context) {
        while (true) {
            const response = await this._listFeatures(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.features || response.features.length === 0) {
                break;
            }
            for (const v of response.features) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _listMaterializedFeatures(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("feature_name")) {
            query["feature_name"] =
                request.feature_name;
        }
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/feature-engineering/materialized-features";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List materialized features.
     */
    async *listMaterializedFeatures(request, context) {
        while (true) {
            const response = await this._listMaterializedFeatures(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.materialized_features ||
                response.materialized_features.length === 0) {
                break;
            }
            for (const v of response.materialized_features) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _updateFeature(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.feature;
        const query = {};
        if (request.hasOwnProperty("update_mask")) {
            query["update_mask"] =
                request.update_mask;
        }
        const path = `/api/2.0/feature-engineering/features/${request.full_name}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update a Feature.
     */
    async updateFeature(request, context) {
        return await this._updateFeature(request, context);
    }
    async _updateMaterializedFeature(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.materialized_feature;
        const query = {};
        if (request.hasOwnProperty("update_mask")) {
            query["update_mask"] =
                request.update_mask;
        }
        const path = `/api/2.0/feature-engineering/materialized-features/${request.materialized_feature_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update a materialized feature (pause/resume).
     */
    async updateMaterializedFeature(request, context) {
        return await this._updateMaterializedFeature(request, context);
    }
}
exports.FeatureEngineeringService = FeatureEngineeringService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "_batchCreateMaterializedFeatures", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "batchCreateMaterializedFeatures", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "_createFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "createFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "_createMaterializedFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "createMaterializedFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "_deleteFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "deleteFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "_deleteMaterializedFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "deleteMaterializedFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "_getFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "getFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "_getMaterializedFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "getMaterializedFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "_listFeatures", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], FeatureEngineeringService.prototype, "listFeatures", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "_listMaterializedFeatures", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], FeatureEngineeringService.prototype, "listMaterializedFeatures", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "_updateFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "updateFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "_updateMaterializedFeature", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureEngineeringService.prototype, "updateMaterializedFeature", null);
class FeatureStoreRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("FeatureStore", method, message);
    }
}
exports.FeatureStoreRetriableError = FeatureStoreRetriableError;
class FeatureStoreError extends apiError_1.ApiError {
    constructor(method, message) {
        super("FeatureStore", method, message);
    }
}
exports.FeatureStoreError = FeatureStoreError;
/**
 * A feature store is a centralized repository that enables data scientists to
 * find and share features. Using a feature store also ensures that the code used
 * to compute feature values is the same during model training and when the model
 * is used for inference.
 *
 * An online store is a low-latency database used for feature lookup during
 * real-time model inference or serve feature for real-time applications.
 */
class FeatureStoreService {
    constructor(client) {
        this.client = client;
    }
    async _createOnlineStore(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.online_store;
        const query = {};
        const path = "/api/2.0/feature-store/online-stores";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create an Online Feature Store.
     */
    async createOnlineStore(request, context) {
        return await this._createOnlineStore(request, context);
    }
    async _deleteOnlineStore(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/feature-store/online-stores/${request.name}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete an Online Feature Store.
     */
    async deleteOnlineStore(request, context) {
        return await this._deleteOnlineStore(request, context);
    }
    async _getOnlineStore(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/feature-store/online-stores/${request.name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get an Online Feature Store.
     */
    async getOnlineStore(request, context) {
        return await this._getOnlineStore(request, context);
    }
    async _listOnlineStores(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/feature-store/online-stores";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List Online Feature Stores.
     */
    async *listOnlineStores(request, context) {
        while (true) {
            const response = await this._listOnlineStores(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.online_stores ||
                response.online_stores.length === 0) {
                break;
            }
            for (const v of response.online_stores) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _publishTable(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("publish_spec")) {
            body["publish_spec"] = request.publish_spec;
        }
        const query = {};
        const path = `/api/2.0/feature-store/tables/${request.source_table_name}/publish`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Publish features.
     */
    async publishTable(request, context) {
        return await this._publishTable(request, context);
    }
    async _updateOnlineStore(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.online_store;
        const query = {};
        if (request.hasOwnProperty("update_mask")) {
            query["update_mask"] =
                request.update_mask;
        }
        const path = `/api/2.0/feature-store/online-stores/${request.name}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update an Online Feature Store.
     */
    async updateOnlineStore(request, context) {
        return await this._updateOnlineStore(request, context);
    }
}
exports.FeatureStoreService = FeatureStoreService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureStoreService.prototype, "_createOnlineStore", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureStoreService.prototype, "createOnlineStore", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureStoreService.prototype, "_deleteOnlineStore", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureStoreService.prototype, "deleteOnlineStore", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureStoreService.prototype, "_getOnlineStore", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureStoreService.prototype, "getOnlineStore", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureStoreService.prototype, "_listOnlineStores", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], FeatureStoreService.prototype, "listOnlineStores", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureStoreService.prototype, "_publishTable", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureStoreService.prototype, "publishTable", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureStoreService.prototype, "_updateOnlineStore", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], FeatureStoreService.prototype, "updateOnlineStore", null);
class ForecastingRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Forecasting", method, message);
    }
}
exports.ForecastingRetriableError = ForecastingRetriableError;
class ForecastingError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Forecasting", method, message);
    }
}
exports.ForecastingError = ForecastingError;
/**
 * The Forecasting API allows you to create and get serverless forecasting
 * experiments
 */
class ForecastingService {
    constructor(client) {
        this.client = client;
    }
    async _createExperiment(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("custom_weights_column")) {
            body["custom_weights_column"] = request.custom_weights_column;
        }
        if (request.hasOwnProperty("experiment_path")) {
            body["experiment_path"] = request.experiment_path;
        }
        if (request.hasOwnProperty("forecast_granularity")) {
            body["forecast_granularity"] = request.forecast_granularity;
        }
        if (request.hasOwnProperty("forecast_horizon")) {
            body["forecast_horizon"] = request.forecast_horizon;
        }
        if (request.hasOwnProperty("future_feature_data_path")) {
            body["future_feature_data_path"] = request.future_feature_data_path;
        }
        if (request.hasOwnProperty("holiday_regions")) {
            body["holiday_regions"] = request.holiday_regions;
        }
        if (request.hasOwnProperty("include_features")) {
            body["include_features"] = request.include_features;
        }
        if (request.hasOwnProperty("max_runtime")) {
            body["max_runtime"] = request.max_runtime;
        }
        if (request.hasOwnProperty("prediction_data_path")) {
            body["prediction_data_path"] = request.prediction_data_path;
        }
        if (request.hasOwnProperty("primary_metric")) {
            body["primary_metric"] = request.primary_metric;
        }
        if (request.hasOwnProperty("register_to")) {
            body["register_to"] = request.register_to;
        }
        if (request.hasOwnProperty("split_column")) {
            body["split_column"] = request.split_column;
        }
        if (request.hasOwnProperty("target_column")) {
            body["target_column"] = request.target_column;
        }
        if (request.hasOwnProperty("time_column")) {
            body["time_column"] = request.time_column;
        }
        if (request.hasOwnProperty("timeseries_identifier_columns")) {
            body["timeseries_identifier_columns"] =
                request.timeseries_identifier_columns;
        }
        if (request.hasOwnProperty("train_data_path")) {
            body["train_data_path"] = request.train_data_path;
        }
        if (request.hasOwnProperty("training_frameworks")) {
            body["training_frameworks"] = request.training_frameworks;
        }
        const query = {};
        const path = "/api/2.0/automl/create-forecasting-experiment";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a serverless forecasting experiment. Returns the experiment ID.
     */
    async createExperiment(createForecastingExperimentRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const createForecastingExperimentResponse = await this._createExperiment(createForecastingExperimentRequest, context);
        return (0, wait_1.asWaiter)(createForecastingExperimentResponse, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.getExperiment({
                        experiment_id: createForecastingExperimentResponse.experiment_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Forecasting.createExperimentAndWait: cancelled");
                        throw new ForecastingError("createExperimentAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.state;
                    const statusMessage = pollResponse;
                    switch (status) {
                        case "SUCCEEDED": {
                            return pollResponse;
                        }
                        case "FAILED":
                        case "CANCELLED": {
                            const errorMessage = `failed to reach SUCCEEDED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Forecasting.createExperimentAndWait: ${errorMessage}`);
                            throw new ForecastingError("createExperimentAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach SUCCEEDED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Forecasting.createExperimentAndWait: retrying: ${errorMessage}`);
                            throw new ForecastingRetriableError("createExperimentAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _getExperiment(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/automl/get-forecasting-experiment/${request.experiment_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Public RPC to get forecasting experiment
     */
    async getExperiment(request, context) {
        return await this._getExperiment(request, context);
    }
}
exports.ForecastingService = ForecastingService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ForecastingService.prototype, "_createExperiment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ForecastingService.prototype, "createExperiment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ForecastingService.prototype, "_getExperiment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ForecastingService.prototype, "getExperiment", null);
class MaterializedFeaturesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("MaterializedFeatures", method, message);
    }
}
exports.MaterializedFeaturesRetriableError = MaterializedFeaturesRetriableError;
class MaterializedFeaturesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("MaterializedFeatures", method, message);
    }
}
exports.MaterializedFeaturesError = MaterializedFeaturesError;
/**
 * Materialized Features are columns in tables and views that can be directly
 * used as features to train and serve ML models.
 */
class MaterializedFeaturesService {
    constructor(client) {
        this.client = client;
    }
    async _createFeatureTag(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.feature_tag;
        const query = {};
        const path = `/api/2.0/feature-store/feature-tables/${request.table_name}/features/${request.feature_name}/tags`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a FeatureTag.
     */
    async createFeatureTag(request, context) {
        return await this._createFeatureTag(request, context);
    }
    async _deleteFeatureTag(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/feature-store/feature-tables/${request.table_name}/features/${request.feature_name}/tags/${request.key}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a FeatureTag.
     */
    async deleteFeatureTag(request, context) {
        return await this._deleteFeatureTag(request, context);
    }
    async _getFeatureLineage(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/feature-store/feature-tables/${request.table_name}/features/${request.feature_name}/lineage`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get Feature Lineage.
     */
    async getFeatureLineage(request, context) {
        return await this._getFeatureLineage(request, context);
    }
    async _getFeatureTag(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/feature-store/feature-tables/${request.table_name}/features/${request.feature_name}/tags/${request.key}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a FeatureTag.
     */
    async getFeatureTag(request, context) {
        return await this._getFeatureTag(request, context);
    }
    async _listFeatureTags(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/feature-store/feature-tables/${request.table_name}/features/${request.feature_name}/tags`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists FeatureTags.
     */
    async *listFeatureTags(request, context) {
        while (true) {
            const response = await this._listFeatureTags(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.feature_tags || response.feature_tags.length === 0) {
                break;
            }
            for (const v of response.feature_tags) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _updateFeatureTag(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.feature_tag;
        const query = {};
        if (request.hasOwnProperty("update_mask")) {
            query["update_mask"] =
                request.update_mask;
        }
        const path = `/api/2.0/feature-store/feature-tables/${request.table_name}/features/${request.feature_name}/tags/${request.key}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates a FeatureTag.
     */
    async updateFeatureTag(request, context) {
        return await this._updateFeatureTag(request, context);
    }
}
exports.MaterializedFeaturesService = MaterializedFeaturesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], MaterializedFeaturesService.prototype, "_createFeatureTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], MaterializedFeaturesService.prototype, "createFeatureTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], MaterializedFeaturesService.prototype, "_deleteFeatureTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], MaterializedFeaturesService.prototype, "deleteFeatureTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], MaterializedFeaturesService.prototype, "_getFeatureLineage", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], MaterializedFeaturesService.prototype, "getFeatureLineage", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], MaterializedFeaturesService.prototype, "_getFeatureTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], MaterializedFeaturesService.prototype, "getFeatureTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], MaterializedFeaturesService.prototype, "_listFeatureTags", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], MaterializedFeaturesService.prototype, "listFeatureTags", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], MaterializedFeaturesService.prototype, "_updateFeatureTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], MaterializedFeaturesService.prototype, "updateFeatureTag", null);
class ModelRegistryRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("ModelRegistry", method, message);
    }
}
exports.ModelRegistryRetriableError = ModelRegistryRetriableError;
class ModelRegistryError extends apiError_1.ApiError {
    constructor(method, message) {
        super("ModelRegistry", method, message);
    }
}
exports.ModelRegistryError = ModelRegistryError;
/**
 * Note: This API reference documents APIs for the Workspace Model Registry.
 * Databricks recommends using [Models in Unity
 * Catalog](/api/workspace/registeredmodels) instead. Models in Unity Catalog
 * provides centralized model governance, cross-workspace access, lineage, and
 * deployment. Workspace Model Registry will be deprecated in the future.
 *
 * The Workspace Model Registry is a centralized model repository and a UI and
 * set of APIs that enable you to manage the full lifecycle of MLflow Models.
 */
class ModelRegistryService {
    constructor(client) {
        this.client = client;
    }
    async _approveTransitionRequest(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("archive_existing_versions")) {
            body["archive_existing_versions"] =
                request.archive_existing_versions;
        }
        if (request.hasOwnProperty("comment")) {
            body["comment"] = request.comment;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("stage")) {
            body["stage"] = request.stage;
        }
        if (request.hasOwnProperty("version")) {
            body["version"] = request.version;
        }
        const query = {};
        const path = "/api/2.0/mlflow/transition-requests/approve";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Approves a model version stage transition request.
     */
    async approveTransitionRequest(request, context) {
        return await this._approveTransitionRequest(request, context);
    }
    async _createComment(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("comment")) {
            body["comment"] = request.comment;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("version")) {
            body["version"] = request.version;
        }
        const query = {};
        const path = "/api/2.0/mlflow/comments/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Posts a comment on a model version. A comment can be submitted either by a
     * user or programmatically to display relevant information about the model.
     * For example, test results or deployment errors.
     */
    async createComment(request, context) {
        return await this._createComment(request, context);
    }
    async _createModel(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        const query = {};
        const path = "/api/2.0/mlflow/registered-models/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a new registered model with the name specified in the request
     * body. Throws `RESOURCE_ALREADY_EXISTS` if a registered model with the
     * given name exists.
     */
    async createModel(request, context) {
        return await this._createModel(request, context);
    }
    async _createModelVersion(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("run_id")) {
            body["run_id"] = request.run_id;
        }
        if (request.hasOwnProperty("run_link")) {
            body["run_link"] = request.run_link;
        }
        if (request.hasOwnProperty("source")) {
            body["source"] = request.source;
        }
        if (request.hasOwnProperty("tags")) {
            body["tags"] = request.tags;
        }
        const query = {};
        const path = "/api/2.0/mlflow/model-versions/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a model version.
     */
    async createModelVersion(request, context) {
        return await this._createModelVersion(request, context);
    }
    async _createTransitionRequest(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("comment")) {
            body["comment"] = request.comment;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("stage")) {
            body["stage"] = request.stage;
        }
        if (request.hasOwnProperty("version")) {
            body["version"] = request.version;
        }
        const query = {};
        const path = "/api/2.0/mlflow/transition-requests/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Creates a model version stage transition request.
     */
    async createTransitionRequest(request, context) {
        return await this._createTransitionRequest(request, context);
    }
    async _createWebhook(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("events")) {
            body["events"] = request.events;
        }
        if (request.hasOwnProperty("http_url_spec")) {
            body["http_url_spec"] = request.http_url_spec;
        }
        if (request.hasOwnProperty("job_spec")) {
            body["job_spec"] = request.job_spec;
        }
        if (request.hasOwnProperty("model_name")) {
            body["model_name"] = request.model_name;
        }
        if (request.hasOwnProperty("status")) {
            body["status"] = request.status;
        }
        const query = {};
        const path = "/api/2.0/mlflow/registry-webhooks/create";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * **NOTE:** This endpoint is in Public Preview. Creates a registry webhook.
     */
    async createWebhook(request, context) {
        return await this._createWebhook(request, context);
    }
    async _deleteComment(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("id")) {
            query["id"] = request.id;
        }
        const path = "/api/2.0/mlflow/comments/delete";
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a comment on a model version.
     */
    async deleteComment(request, context) {
        return await this._deleteComment(request, context);
    }
    async _deleteModel(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("name")) {
            query["name"] = request.name;
        }
        const path = "/api/2.0/mlflow/registered-models/delete";
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a registered model.
     */
    async deleteModel(request, context) {
        return await this._deleteModel(request, context);
    }
    async _deleteModelTag(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("key")) {
            query["key"] = request.key;
        }
        if (request.hasOwnProperty("name")) {
            query["name"] = request.name;
        }
        const path = "/api/2.0/mlflow/registered-models/delete-tag";
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes the tag for a registered model.
     */
    async deleteModelTag(request, context) {
        return await this._deleteModelTag(request, context);
    }
    async _deleteModelVersion(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("name")) {
            query["name"] = request.name;
        }
        if (request.hasOwnProperty("version")) {
            query["version"] = request.version;
        }
        const path = "/api/2.0/mlflow/model-versions/delete";
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a model version.
     */
    async deleteModelVersion(request, context) {
        return await this._deleteModelVersion(request, context);
    }
    async _deleteModelVersionTag(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("key")) {
            query["key"] = request.key;
        }
        if (request.hasOwnProperty("name")) {
            query["name"] = request.name;
        }
        if (request.hasOwnProperty("version")) {
            query["version"] = request.version;
        }
        const path = "/api/2.0/mlflow/model-versions/delete-tag";
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Deletes a model version tag.
     */
    async deleteModelVersionTag(request, context) {
        return await this._deleteModelVersionTag(request, context);
    }
    async _deleteTransitionRequest(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("comment")) {
            query["comment"] = request.comment;
        }
        if (request.hasOwnProperty("creator")) {
            query["creator"] = request.creator;
        }
        if (request.hasOwnProperty("name")) {
            query["name"] = request.name;
        }
        if (request.hasOwnProperty("stage")) {
            query["stage"] = request.stage;
        }
        if (request.hasOwnProperty("version")) {
            query["version"] = request.version;
        }
        const path = "/api/2.0/mlflow/transition-requests/delete";
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Cancels a model version stage transition request.
     */
    async deleteTransitionRequest(request, context) {
        return await this._deleteTransitionRequest(request, context);
    }
    async _deleteWebhook(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("id")) {
            query["id"] = request.id;
        }
        const path = "/api/2.0/mlflow/registry-webhooks/delete";
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * **NOTE:** This endpoint is in Public Preview. Deletes a registry webhook.
     */
    async deleteWebhook(request, context) {
        return await this._deleteWebhook(request, context);
    }
    async _getLatestVersions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("stages")) {
            body["stages"] = request.stages;
        }
        const query = {};
        const path = "/api/2.0/mlflow/registered-models/get-latest-versions";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the latest version of a registered model.
     */
    async *getLatestVersions(request, context) {
        const response = (await this._getLatestVersions(request, context))
            .model_versions;
        for (const v of response || []) {
            yield v;
        }
    }
    async _getModel(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("name")) {
            query["name"] = request.name;
        }
        const path = "/api/2.0/mlflow/databricks/registered-models/get";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get the details of a model. This is a Databricks workspace version of the
     * [MLflow endpoint] that also returns the model's Databricks workspace ID
     * and the permission level of the requesting user on the model.
     *
     * [MLflow endpoint]: https://www.mlflow.org/docs/latest/rest-api.html#get-registeredmodel
     */
    async getModel(request, context) {
        return await this._getModel(request, context);
    }
    async _getModelVersion(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("name")) {
            query["name"] = request.name;
        }
        if (request.hasOwnProperty("version")) {
            query["version"] = request.version;
        }
        const path = "/api/2.0/mlflow/model-versions/get";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a model version.
     */
    async getModelVersion(request, context) {
        return await this._getModelVersion(request, context);
    }
    async _getModelVersionDownloadUri(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("name")) {
            query["name"] = request.name;
        }
        if (request.hasOwnProperty("version")) {
            query["version"] = request.version;
        }
        const path = "/api/2.0/mlflow/model-versions/get-download-uri";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a URI to download the model version.
     */
    async getModelVersionDownloadUri(request, context) {
        return await this._getModelVersionDownloadUri(request, context);
    }
    async _getPermissionLevels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/registered-models/${request.registered_model_id}/permissionLevels`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permission levels that a user can have on an object.
     */
    async getPermissionLevels(request, context) {
        return await this._getPermissionLevels(request, context);
    }
    async _getPermissions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/permissions/registered-models/${request.registered_model_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the permissions of a registered model. Registered models can inherit
     * permissions from their root object.
     */
    async getPermissions(request, context) {
        return await this._getPermissions(request, context);
    }
    async _listModels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/mlflow/registered-models/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Lists all available registered models, up to the limit specified in
     * __max_results__.
     */
    async *listModels(request, context) {
        while (true) {
            const response = await this._listModels(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.registered_models ||
                response.registered_models.length === 0) {
                break;
            }
            for (const v of response.registered_models) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _listTransitionRequests(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("name")) {
            query["name"] = request.name;
        }
        if (request.hasOwnProperty("version")) {
            query["version"] = request.version;
        }
        const path = "/api/2.0/mlflow/transition-requests/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets a list of all open stage transition requests for the model version.
     */
    async *listTransitionRequests(request, context) {
        const response = (await this._listTransitionRequests(request, context))
            .requests;
        for (const v of response || []) {
            yield v;
        }
    }
    async _listWebhooks(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("events")) {
            query["events"] = request.events;
        }
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("model_name")) {
            query["model_name"] =
                request.model_name;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/mlflow/registry-webhooks/list";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * **NOTE:** This endpoint is in Public Preview. Lists all registry webhooks.
     */
    async *listWebhooks(request, context) {
        while (true) {
            const response = await this._listWebhooks(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.webhooks || response.webhooks.length === 0) {
                break;
            }
            for (const v of response.webhooks) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _rejectTransitionRequest(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("comment")) {
            body["comment"] = request.comment;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("stage")) {
            body["stage"] = request.stage;
        }
        if (request.hasOwnProperty("version")) {
            body["version"] = request.version;
        }
        const query = {};
        const path = "/api/2.0/mlflow/transition-requests/reject";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Rejects a model version stage transition request.
     */
    async rejectTransitionRequest(request, context) {
        return await this._rejectTransitionRequest(request, context);
    }
    async _renameModel(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("new_name")) {
            body["new_name"] = request.new_name;
        }
        const query = {};
        const path = "/api/2.0/mlflow/registered-models/rename";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Renames a registered model.
     */
    async renameModel(request, context) {
        return await this._renameModel(request, context);
    }
    async _searchModelVersions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("filter")) {
            query["filter"] = request.filter;
        }
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("order_by")) {
            query["order_by"] =
                request.order_by;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/mlflow/model-versions/search";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Searches for specific model versions based on the supplied __filter__.
     */
    async *searchModelVersions(request, context) {
        while (true) {
            const response = await this._searchModelVersions(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.model_versions ||
                response.model_versions.length === 0) {
                break;
            }
            for (const v of response.model_versions) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _searchModels(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("filter")) {
            query["filter"] = request.filter;
        }
        if (request.hasOwnProperty("max_results")) {
            query["max_results"] =
                request.max_results;
        }
        if (request.hasOwnProperty("order_by")) {
            query["order_by"] =
                request.order_by;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/mlflow/registered-models/search";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Search for registered models based on the specified __filter__.
     */
    async *searchModels(request, context) {
        while (true) {
            const response = await this._searchModels(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.registered_models ||
                response.registered_models.length === 0) {
                break;
            }
            for (const v of response.registered_models) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _setModelTag(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("key")) {
            body["key"] = request.key;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("value")) {
            body["value"] = request.value;
        }
        const query = {};
        const path = "/api/2.0/mlflow/registered-models/set-tag";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets a tag on a registered model.
     */
    async setModelTag(request, context) {
        return await this._setModelTag(request, context);
    }
    async _setModelVersionTag(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("key")) {
            body["key"] = request.key;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("value")) {
            body["value"] = request.value;
        }
        if (request.hasOwnProperty("version")) {
            body["version"] = request.version;
        }
        const query = {};
        const path = "/api/2.0/mlflow/model-versions/set-tag";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets a model version tag.
     */
    async setModelVersionTag(request, context) {
        return await this._setModelVersionTag(request, context);
    }
    async _setPermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/registered-models/${request.registered_model_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Sets permissions on an object, replacing existing permissions if they
     * exist. Deletes all direct permissions if none are specified. Objects can
     * inherit permissions from their root object.
     */
    async setPermissions(request, context) {
        return await this._setPermissions(request, context);
    }
    async _testRegistryWebhook(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("event")) {
            body["event"] = request.event;
        }
        if (request.hasOwnProperty("id")) {
            body["id"] = request.id;
        }
        const query = {};
        const path = "/api/2.0/mlflow/registry-webhooks/test";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * **NOTE:** This endpoint is in Public Preview. Tests a registry webhook.
     */
    async testRegistryWebhook(request, context) {
        return await this._testRegistryWebhook(request, context);
    }
    async _transitionStage(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("archive_existing_versions")) {
            body["archive_existing_versions"] =
                request.archive_existing_versions;
        }
        if (request.hasOwnProperty("comment")) {
            body["comment"] = request.comment;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("stage")) {
            body["stage"] = request.stage;
        }
        if (request.hasOwnProperty("version")) {
            body["version"] = request.version;
        }
        const query = {};
        const path = "/api/2.0/mlflow/databricks/model-versions/transition-stage";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Transition a model version's stage. This is a Databricks workspace version
     * of the [MLflow endpoint] that also accepts a comment associated with the
     * transition to be recorded.
     *
     * [MLflow endpoint]: https://www.mlflow.org/docs/latest/rest-api.html#transition-modelversion-stage
     */
    async transitionStage(request, context) {
        return await this._transitionStage(request, context);
    }
    async _updateComment(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("comment")) {
            body["comment"] = request.comment;
        }
        if (request.hasOwnProperty("id")) {
            body["id"] = request.id;
        }
        const query = {};
        const path = "/api/2.0/mlflow/comments/update";
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Post an edit to a comment on a model version.
     */
    async updateComment(request, context) {
        return await this._updateComment(request, context);
    }
    async _updateModel(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        const query = {};
        const path = "/api/2.0/mlflow/registered-models/update";
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates a registered model.
     */
    async updateModel(request, context) {
        return await this._updateModel(request, context);
    }
    async _updateModelVersion(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("version")) {
            body["version"] = request.version;
        }
        const query = {};
        const path = "/api/2.0/mlflow/model-versions/update";
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the model version.
     */
    async updateModelVersion(request, context) {
        return await this._updateModelVersion(request, context);
    }
    async _updatePermissions(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("access_control_list")) {
            body["access_control_list"] = request.access_control_list;
        }
        const query = {};
        const path = `/api/2.0/permissions/registered-models/${request.registered_model_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates the permissions on a registered model. Registered models can
     * inherit permissions from their root object.
     */
    async updatePermissions(request, context) {
        return await this._updatePermissions(request, context);
    }
    async _updateWebhook(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("description")) {
            body["description"] = request.description;
        }
        if (request.hasOwnProperty("events")) {
            body["events"] = request.events;
        }
        if (request.hasOwnProperty("http_url_spec")) {
            body["http_url_spec"] = request.http_url_spec;
        }
        if (request.hasOwnProperty("id")) {
            body["id"] = request.id;
        }
        if (request.hasOwnProperty("job_spec")) {
            body["job_spec"] = request.job_spec;
        }
        if (request.hasOwnProperty("status")) {
            body["status"] = request.status;
        }
        const query = {};
        const path = "/api/2.0/mlflow/registry-webhooks/update";
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * **NOTE:** This endpoint is in Public Preview. Updates a registry webhook.
     */
    async updateWebhook(request, context) {
        return await this._updateWebhook(request, context);
    }
}
exports.ModelRegistryService = ModelRegistryService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_approveTransitionRequest", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "approveTransitionRequest", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_createComment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "createComment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_createModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "createModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_createModelVersion", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "createModelVersion", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_createTransitionRequest", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "createTransitionRequest", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_createWebhook", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "createWebhook", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_deleteComment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "deleteComment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_deleteModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "deleteModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_deleteModelTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "deleteModelTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_deleteModelVersion", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "deleteModelVersion", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_deleteModelVersionTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "deleteModelVersionTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_deleteTransitionRequest", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "deleteTransitionRequest", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_deleteWebhook", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "deleteWebhook", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_getLatestVersions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ModelRegistryService.prototype, "getLatestVersions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_getModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "getModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_getModelVersion", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "getModelVersion", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_getModelVersionDownloadUri", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "getModelVersionDownloadUri", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "getPermissionLevels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "getPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_listModels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ModelRegistryService.prototype, "listModels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_listTransitionRequests", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ModelRegistryService.prototype, "listTransitionRequests", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_listWebhooks", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ModelRegistryService.prototype, "listWebhooks", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_rejectTransitionRequest", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "rejectTransitionRequest", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_renameModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "renameModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_searchModelVersions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ModelRegistryService.prototype, "searchModelVersions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_searchModels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ModelRegistryService.prototype, "searchModels", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_setModelTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "setModelTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_setModelVersionTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "setModelVersionTag", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "setPermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_testRegistryWebhook", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "testRegistryWebhook", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_transitionStage", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "transitionStage", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_updateComment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "updateComment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_updateModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "updateModel", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_updateModelVersion", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "updateModelVersion", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "updatePermissions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "_updateWebhook", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ModelRegistryService.prototype, "updateWebhook", null);
