// @databricks/appkit/dist/connectors/jobs/client.js
import { createLogger } from "../../logging/logger.js";
import { AppKitError } from "../../errors/base.js";
import { ExecutionError } from "../../errors/execution.js";
import "../../errors/index.js";
import { TelemetryManager } from "../../telemetry/telemetry-manager.js";
import { SpanKind, SpanStatusCode } from "../../telemetry/index.js";
import { Context } from "@databricks/sdk-experimental";

//#region src/connectors/jobs/client.ts
const logger = createLogger("connectors:jobs");
var JobsConnector = class {
	name = "jobs";
	config;
	telemetry;
	telemetryMetrics;
	constructor(config) {
		this.config = config;
		this.telemetry = TelemetryManager.getProvider(this.name, this.config.telemetry);
		this.telemetryMetrics = {
			apiCallCount: this.telemetry.getMeter().createCounter("jobs.api_call.count", {
				description: "Total number of Jobs API calls",
				unit: "1"
			}),
			apiCallDuration: this.telemetry.getMeter().createHistogram("jobs.api_call.duration", {
				description: "Duration of Jobs API calls",
				unit: "ms"
			})
		};
	}
	async submitRun(workspaceClient, request, signal) {
		return this._callApi("submit", async () => {
			return workspaceClient.jobs.submit(request, this._createContext(signal));
		});
	}
	async runNow(workspaceClient, request, signal) {
		return this._callApi("runNow", async () => {
			return workspaceClient.jobs.runNow(request, this._createContext(signal));
		});
	}
	async getRun(workspaceClient, request, signal) {
		return this._callApi("getRun", async () => {
			return workspaceClient.jobs.getRun(request, this._createContext(signal));
		});
	}
	async getRunOutput(workspaceClient, request, signal) {
		return this._callApi("getRunOutput", async () => {
			return workspaceClient.jobs.getRunOutput(request, this._createContext(signal));
		});
	}
	async cancelRun(workspaceClient, request, signal) {
		await this._callApi("cancelRun", async () => {
			return workspaceClient.jobs.cancelRun(request, this._createContext(signal));
		});
	}
	async listRuns(workspaceClient, request, signal) {
		return this._callApi("listRuns", async () => {
			const runs = [];
			const limit = Math.max(1, Math.min(request.limit ?? 100, 100));
			for await (const run of workspaceClient.jobs.listRuns({
				...request,
				limit
			}, this._createContext(signal))) {
				runs.push(run);
				if (runs.length >= limit) break;
			}
			return runs;
		});
	}
	async getJob(workspaceClient, request, signal) {
		return this._callApi("getJob", async () => {
			return workspaceClient.jobs.get(request, this._createContext(signal));
		});
	}
	async _callApi(operation, fn) {
		const startTime = Date.now();
		let success = false;
		return this.telemetry.startActiveSpan(`jobs.${operation}`, {
			kind: SpanKind.CLIENT,
			attributes: { "jobs.operation": operation }
		}, async (span) => {
			try {
				const result = await fn();
				success = true;
				span.setStatus({ code: SpanStatusCode.OK });
				return result;
			} catch (error) {
				span.recordException(error);
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: error instanceof Error ? error.message : String(error)
				});
				if (error instanceof AppKitError) throw error;
				if (error instanceof Error && "statusCode" in error && typeof error.statusCode === "number") throw error;
				throw new ExecutionError(`Jobs API call failed: ${error instanceof Error ? error.message : String(error)}`);
			} finally {
				const duration = Date.now() - startTime;
				span.end();
				this.telemetryMetrics.apiCallCount.add(1, {
					operation,
					success: success.toString()
				});
				this.telemetryMetrics.apiCallDuration.record(duration, {
					operation,
					success: success.toString()
				});
				logger.event()?.setContext("jobs", {
					operation,
					duration_ms: duration,
					success
				});
			}
		}, {
			name: this.name,
			includePrefix: true
		});
	}
	_createContext(signal) {
		return new Context({ cancellationToken: {
			get isCancellationRequested() {
				return signal?.aborted ?? false;
			},
			onCancellationRequested: (cb) => {
				signal?.addEventListener("abort", cb, { once: true });
			}
		} });
	}
};

//#endregion
export { JobsConnector };
