// @databricks/sdk-experimental/dist/apis/settingsv2/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceSettingsV2Service = exports.AccountSettingsV2Service = void 0;
/**
 * These APIs allow you to manage Account Settings V2, Workspace Settings V2, etc.
 */
const config_1 = require("../../config");
const HostType_1 = require("../../config/HostType");
const settingsv2 = __importStar(require("./model"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
class AccountSettingsV2RetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("AccountSettingsV2", method, message);
    }
}
class AccountSettingsV2Error extends apiError_1.ApiError {
    constructor(method, message) {
        super("AccountSettingsV2", method, message);
    }
}
/**
 * APIs to manage account level settings
 */
class AccountSettingsV2Service {
    constructor(client) {
        this.client = client;
    }
    async _getPublicAccountSetting(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.1/accounts/${config.accountId}/settings/${request.name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a setting value at account level. See
     * :method:settingsv2/listaccountsettingsmetadata for list of setting
     * available via public APIs at account level.
     */
    async getPublicAccountSetting(request, context) {
        return await this._getPublicAccountSetting(request, context);
    }
    async _listAccountSettingsMetadata(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.1/accounts/${config.accountId}/settings-metadata`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List valid setting keys and metadata. These settings are available to be
     * referenced via GET :method:settingsv2/getpublicaccountsetting and PATCH
     * :method:settingsv2/patchpublicaccountsetting APIs
     */
    async *listAccountSettingsMetadata(request, context) {
        while (true) {
            const response = await this._listAccountSettingsMetadata(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.settings_metadata ||
                response.settings_metadata.length === 0) {
                break;
            }
            for (const v of response.settings_metadata) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _patchPublicAccountSetting(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.setting;
        const query = {};
        const path = `/api/2.1/accounts/${config.accountId}/settings/${request.name}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Patch a setting value at account level. See
     * :method:settingsv2/listaccountsettingsmetadata for list of setting
     * available via public APIs at account level.
     */
    async patchPublicAccountSetting(request, context) {
        return await this._patchPublicAccountSetting(request, context);
    }
}
exports.AccountSettingsV2Service = AccountSettingsV2Service;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountSettingsV2Service.prototype, "_getPublicAccountSetting", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountSettingsV2Service.prototype, "getPublicAccountSetting", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountSettingsV2Service.prototype, "_listAccountSettingsMetadata", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], AccountSettingsV2Service.prototype, "listAccountSettingsMetadata", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountSettingsV2Service.prototype, "_patchPublicAccountSetting", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountSettingsV2Service.prototype, "patchPublicAccountSetting", null);
class WorkspaceSettingsV2RetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("WorkspaceSettingsV2", method, message);
    }
}
class WorkspaceSettingsV2Error extends apiError_1.ApiError {
    constructor(method, message) {
        super("WorkspaceSettingsV2", method, message);
    }
}
/**
 * APIs to manage workspace level settings
 */
class WorkspaceSettingsV2Service {
    constructor(client) {
        this.client = client;
    }
    async _getPublicWorkspaceSetting(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.1/settings/${request.name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a setting value at workspace level. See
     * :method:settingsv2/listworkspacesettingsmetadata for list of setting
     * available via public APIs.
     */
    async getPublicWorkspaceSetting(request, context) {
        return await this._getPublicWorkspaceSetting(request, context);
    }
    async _listWorkspaceSettingsMetadata(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.1/settings-metadata";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List valid setting keys and metadata. These settings are available to be
     * referenced via GET :method:settingsv2/getpublicworkspacesetting and PATCH
     * :method:settingsv2/patchpublicworkspacesetting APIs
     */
    async *listWorkspaceSettingsMetadata(request, context) {
        while (true) {
            const response = await this._listWorkspaceSettingsMetadata(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.settings_metadata ||
                response.settings_metadata.length === 0) {
                break;
            }
            for (const v of response.settings_metadata) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _patchPublicWorkspaceSetting(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.setting;
        const query = {};
        const path = `/api/2.1/settings/${request.name}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Patch a setting value at workspace level. See
     * :method:settingsv2/listworkspacesettingsmetadata for list of setting
     * available via public APIs at workspace level.
     */
    async patchPublicWorkspaceSetting(request, context) {
        return await this._patchPublicWorkspaceSetting(request, context);
    }
}
exports.WorkspaceSettingsV2Service = WorkspaceSettingsV2Service;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceSettingsV2Service.prototype, "_getPublicWorkspaceSetting", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceSettingsV2Service.prototype, "getPublicWorkspaceSetting", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceSettingsV2Service.prototype, "_listWorkspaceSettingsMetadata", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], WorkspaceSettingsV2Service.prototype, "listWorkspaceSettingsMetadata", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceSettingsV2Service.prototype, "_patchPublicWorkspaceSetting", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], WorkspaceSettingsV2Service.prototype, "patchPublicWorkspaceSetting", null);
