// @databricks/appkit/dist/cli/commands/plugin/add-resource/add-resource.js
import { DEFAULT_PERMISSION_BY_TYPE, getDefaultFieldsForType, getValidResourceTypes, humanizeResourceType, resourceKeyFromType } from "../create/resource-defaults.js";
import { promptOneResource } from "../create/prompt-resource.js";
import { resolveManifestInDir } from "../manifest-resolve.js";
import { validateManifest } from "../validate/validate-manifest.js";
import fs from "node:fs";
import path from "node:path";
import { Command } from "commander";
import process from "node:process";
import { cancel, intro, outro } from "@clack/prompts";

//#region src/cli/commands/plugin/add-resource/add-resource.ts
function loadManifest(pluginDir) {
	const resolved = resolveManifestInDir(pluginDir, { allowJsManifest: true });
	if (!resolved) {
		console.error(`No manifest found in ${pluginDir}. This command requires manifest.json (manifest.js cannot be edited in place).`);
		console.error("  appkit plugin add-resource --path <dir-with-manifest.json>");
		process.exit(1);
	}
	if (resolved.type !== "json") {
		console.error(`Editable manifest not found. add-resource only supports plugin directories that contain manifest.json (found ${path.basename(resolved.path)}).`);
		process.exit(1);
	}
	const manifestPath = resolved.path;
	try {
		const raw = fs.readFileSync(manifestPath, "utf-8");
		const parsed = JSON.parse(raw);
		const result = validateManifest(parsed);
		if (!result.valid || !result.manifest) {
			console.error("Invalid manifest. Run `appkit plugin validate` for details.");
			process.exit(1);
		}
		return {
			manifest: parsed,
			manifestPath
		};
	} catch (err) {
		console.error("Failed to read or parse manifest.json:", err instanceof Error ? err.message : err);
		process.exit(1);
	}
}
function buildEntry(type, opts) {
	const alias = humanizeResourceType(type);
	const isRequired = opts.required !== false;
	let fields = getDefaultFieldsForType(type);
	if (opts.fieldsJson) try {
		const parsed = JSON.parse(opts.fieldsJson);
		fields = {
			...fields,
			...parsed
		};
	} catch {
		console.error("Error: --fields-json must be valid JSON.");
		console.error("  Example: --fields-json '{\"id\":{\"env\":\"MY_WAREHOUSE_ID\"}}'");
		process.exit(1);
	}
	return {
		entry: {
			type,
			alias,
			resourceKey: opts.resourceKey ?? resourceKeyFromType(type),
			description: opts.description || `${isRequired ? "Required" : "Optional"} for ${alias} functionality.`,
			permission: opts.permission ?? DEFAULT_PERMISSION_BY_TYPE[type] ?? "CAN_VIEW",
			fields
		},
		isRequired
	};
}
function runNonInteractive(opts) {
	const cwd = process.cwd();
	const loaded = loadManifest(path.resolve(cwd, opts.path ?? "."));
	if (!loaded) return;
	const { manifest, manifestPath } = loaded;
	const type = opts.type;
	const validTypes = getValidResourceTypes();
	if (!validTypes.includes(type)) {
		console.error(`Error: Unknown resource type "${type}".`);
		console.error(`  Valid types: ${validTypes.join(", ")}`);
		process.exit(1);
	}
	const { entry, isRequired } = buildEntry(type, opts);
	if (isRequired) manifest.resources.required.push(entry);
	else manifest.resources.optional.push(entry);
	if (opts.dryRun) {
		console.log(JSON.stringify(manifest, null, 2));
		return;
	}
	fs.writeFileSync(manifestPath, `${JSON.stringify(manifest, null, 2)}\n`);
	console.log(`Added ${entry.alias} as ${isRequired ? "required" : "optional"} to ${path.relative(cwd, manifestPath)}`);
}
async function runInteractive(opts) {
	intro("Add resource to plugin manifest");
	const cwd = process.cwd();
	const loaded = loadManifest(path.resolve(cwd, opts.path ?? "."));
	if (!loaded) return;
	const { manifest, manifestPath } = loaded;
	const spec = await promptOneResource();
	if (!spec) {
		cancel("Cancelled.");
		process.exit(0);
	}
	const alias = humanizeResourceType(spec.type);
	const entry = {
		type: spec.type,
		alias,
		resourceKey: spec.resourceKey,
		description: spec.description || `Required for ${alias} functionality.`,
		permission: spec.permission,
		fields: spec.fields
	};
	if (spec.required) manifest.resources.required.push(entry);
	else manifest.resources.optional.push(entry);
	fs.writeFileSync(manifestPath, `${JSON.stringify(manifest, null, 2)}\n`);
	outro("Resource added.");
	console.log(`\nAdded ${alias} as ${spec.required ? "required" : "optional"} to ${path.relative(cwd, manifestPath)}`);
}
async function runPluginAddResource(opts) {
	if (opts.type) runNonInteractive(opts);
	else await runInteractive(opts);
}
const pluginAddResourceCommand = new Command("add-resource").description("Add a resource requirement to an existing plugin manifest. Overwrites manifest.json in place.").option("-p, --path <dir>", "Plugin directory containing manifest.json (default: .)").option("-t, --type <resource_type>", "Resource type (e.g. sql_warehouse, volume). Enables non-interactive mode.").option("--required", "Mark resource as required (default: true)", true).option("--no-required", "Mark resource as optional").option("--resource-key <key>", "Resource key (default: derived from type)").option("--description <text>", "Description of the resource requirement").option("--permission <perm>", "Permission level (default: from schema)").option("--fields-json <json>", "JSON object overriding field env vars (e.g. '{\"id\":{\"env\":\"MY_WAREHOUSE_ID\"}}')").option("--dry-run", "Preview the updated manifest without writing").addHelpText("after", `
Examples:
  $ appkit plugin add-resource
  $ appkit plugin add-resource --path plugins/my-plugin --type sql_warehouse
  $ appkit plugin add-resource --path plugins/my-plugin --type volume --no-required --dry-run
  $ appkit plugin add-resource --type sql_warehouse --fields-json '{"id":{"env":"MY_WAREHOUSE_ID"}}'`).action((opts) => runPluginAddResource(opts).catch((err) => {
	console.error(err);
	process.exit(1);
}));

//#endregion
export { pluginAddResourceCommand };
