// @databricks/sdk-experimental/dist/apis/oauth2/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServicePrincipalSecretsProxyService = exports.ServicePrincipalSecretsProxyError = exports.ServicePrincipalSecretsProxyRetriableError = exports.ServicePrincipalSecretsService = exports.ServicePrincipalSecretsError = exports.ServicePrincipalSecretsRetriableError = exports.ServicePrincipalFederationPolicyService = exports.ServicePrincipalFederationPolicyError = exports.ServicePrincipalFederationPolicyRetriableError = exports.PublishedAppIntegrationService = exports.PublishedAppIntegrationError = exports.PublishedAppIntegrationRetriableError = exports.OAuthPublishedAppsService = exports.OAuthPublishedAppsError = exports.OAuthPublishedAppsRetriableError = exports.CustomAppIntegrationService = exports.CustomAppIntegrationError = exports.CustomAppIntegrationRetriableError = exports.AccountFederationPolicyService = exports.AccountFederationPolicyError = exports.AccountFederationPolicyRetriableError = void 0;
/**
 * These APIs allow you to manage Account Federation Policy, Custom App Integration, O Auth Published Apps, Published App Integration, Service Principal Federation Policy, Service Principal Secrets, Service Principal Secrets Proxy, etc.
 */
const config_1 = require("../../config");
const HostType_1 = require("../../config/HostType");
const oauth2 = __importStar(require("./model"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
class AccountFederationPolicyRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("AccountFederationPolicy", method, message);
    }
}
exports.AccountFederationPolicyRetriableError = AccountFederationPolicyRetriableError;
class AccountFederationPolicyError extends apiError_1.ApiError {
    constructor(method, message) {
        super("AccountFederationPolicy", method, message);
    }
}
exports.AccountFederationPolicyError = AccountFederationPolicyError;
/**
 * These APIs manage account federation policies.
 *
 * Account federation policies allow users and service principals in your
 * Databricks account to securely access Databricks APIs using tokens from your
 * trusted identity providers (IdPs).
 *
 * With token federation, your users and service principals can exchange tokens
 * from your IdP for Databricks OAuth tokens, which can be used to access
 * Databricks APIs. Token federation eliminates the need to manage Databricks
 * secrets, and allows you to centralize management of token issuance policies in
 * your IdP. Databricks token federation is typically used in combination with
 * [SCIM], so users in your IdP are synchronized into your Databricks account.
 *
 * Token federation is configured in your Databricks account using an account
 * federation policy. An account federation policy specifies: * which IdP, or
 * issuer, your Databricks account should accept tokens from * how to determine
 * which Databricks user, or subject, a token is issued for
 *
 * To configure a federation policy, you provide the following: * The required
 * token __issuer__, as specified in the “iss” claim of your tokens. The
 * issuer is an https URL that identifies your IdP. * The allowed token
 * __audiences__, as specified in the “aud” claim of your tokens. This
 * identifier is intended to represent the recipient of the token. As long as the
 * audience in the token matches at least one audience in the policy, the token
 * is considered a match. If unspecified, the default value is your Databricks
 * account id. * The __subject claim__, which indicates which token claim
 * contains the Databricks username of the user the token was issued for. If
 * unspecified, the default value is “sub”. * Optionally, the public keys
 * used to validate the signature of your tokens, in JWKS format. If unspecified
 * (recommended), Databricks automatically fetches the public keys from your
 * issuer’s well known endpoint. Databricks strongly recommends relying on your
 * issuer’s well known endpoint for discovering public keys.
 *
 * An example federation policy is: ``` issuer: "https://idp.mycompany.com/oidc"
 * audiences: ["databricks"] subject_claim: "sub" ```
 *
 * An example JWT token body that matches this policy and could be used to
 * authenticate to Databricks as user `username@mycompany.com` is: ``` { "iss":
 * "https://idp.mycompany.com/oidc", "aud": "databricks", "sub":
 * "username@mycompany.com" } ```
 *
 * You may also need to configure your IdP to generate tokens for your users to
 * exchange with Databricks, if your users do not already have the ability to
 * generate tokens that are compatible with your federation policy.
 *
 * You do not need to configure an OAuth application in Databricks to use token
 * federation.
 *
 * [SCIM]: https://docs.databricks.com/admin/users-groups/scim/index.html
 */
class AccountFederationPolicyService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.policy;
        const query = {};
        if (request.hasOwnProperty("policy_id")) {
            query["policy_id"] =
                request.policy_id;
        }
        const path = `/api/2.0/accounts/${config.accountId}/federationPolicies`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create account federation policy.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/federationPolicies/${request.policy_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete account federation policy.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/federationPolicies/${request.policy_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get account federation policy.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/accounts/${config.accountId}/federationPolicies`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List account federation policies.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.policies || response.policies.length === 0) {
                break;
            }
            for (const v of response.policies) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _update(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.policy;
        const query = {};
        if (request.hasOwnProperty("update_mask")) {
            query["update_mask"] =
                request.update_mask;
        }
        const path = `/api/2.0/accounts/${config.accountId}/federationPolicies/${request.policy_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update account federation policy.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.AccountFederationPolicyService = AccountFederationPolicyService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountFederationPolicyService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountFederationPolicyService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountFederationPolicyService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountFederationPolicyService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountFederationPolicyService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountFederationPolicyService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountFederationPolicyService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], AccountFederationPolicyService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountFederationPolicyService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], AccountFederationPolicyService.prototype, "update", null);
class CustomAppIntegrationRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("CustomAppIntegration", method, message);
    }
}
exports.CustomAppIntegrationRetriableError = CustomAppIntegrationRetriableError;
class CustomAppIntegrationError extends apiError_1.ApiError {
    constructor(method, message) {
        super("CustomAppIntegration", method, message);
    }
}
exports.CustomAppIntegrationError = CustomAppIntegrationError;
/**
 * These APIs enable administrators to manage custom OAuth app integrations,
 * which is required for adding/using Custom OAuth App Integration like Tableau
 * Cloud for Databricks in AWS cloud.
 */
class CustomAppIntegrationService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("confidential")) {
            body["confidential"] = request.confidential;
        }
        if (request.hasOwnProperty("name")) {
            body["name"] = request.name;
        }
        if (request.hasOwnProperty("redirect_urls")) {
            body["redirect_urls"] = request.redirect_urls;
        }
        if (request.hasOwnProperty("scopes")) {
            body["scopes"] = request.scopes;
        }
        if (request.hasOwnProperty("token_access_policy")) {
            body["token_access_policy"] = request.token_access_policy;
        }
        if (request.hasOwnProperty("user_authorized_scopes")) {
            body["user_authorized_scopes"] = request.user_authorized_scopes;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/oauth2/custom-app-integrations`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create Custom OAuth App Integration.
     *
     * You can retrieve the custom OAuth app integration via
     * :method:CustomAppIntegration/get.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/oauth2/custom-app-integrations/${request.integration_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete an existing Custom OAuth App Integration. You can retrieve the
     * custom OAuth app integration via :method:CustomAppIntegration/get.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/oauth2/custom-app-integrations/${request.integration_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the Custom OAuth App Integration for the given integration id.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("include_creator_username")) {
            query["include_creator_username"] =
                request.include_creator_username;
        }
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/accounts/${config.accountId}/oauth2/custom-app-integrations`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get the list of custom OAuth app integrations for the specified Databricks
     * account
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.apps || response.apps.length === 0) {
                break;
            }
            for (const v of response.apps) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _update(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("redirect_urls")) {
            body["redirect_urls"] = request.redirect_urls;
        }
        if (request.hasOwnProperty("scopes")) {
            body["scopes"] = request.scopes;
        }
        if (request.hasOwnProperty("token_access_policy")) {
            body["token_access_policy"] = request.token_access_policy;
        }
        if (request.hasOwnProperty("user_authorized_scopes")) {
            body["user_authorized_scopes"] = request.user_authorized_scopes;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/oauth2/custom-app-integrations/${request.integration_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates an existing custom OAuth App Integration. You can retrieve the
     * custom OAuth app integration via :method:CustomAppIntegration/get.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.CustomAppIntegrationService = CustomAppIntegrationService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CustomAppIntegrationService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CustomAppIntegrationService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CustomAppIntegrationService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CustomAppIntegrationService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CustomAppIntegrationService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CustomAppIntegrationService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CustomAppIntegrationService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], CustomAppIntegrationService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CustomAppIntegrationService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CustomAppIntegrationService.prototype, "update", null);
class OAuthPublishedAppsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("OAuthPublishedApps", method, message);
    }
}
exports.OAuthPublishedAppsRetriableError = OAuthPublishedAppsRetriableError;
class OAuthPublishedAppsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("OAuthPublishedApps", method, message);
    }
}
exports.OAuthPublishedAppsError = OAuthPublishedAppsError;
/**
 * These APIs enable administrators to view all the available published OAuth
 * applications in Databricks. Administrators can add the published OAuth
 * applications to their account through the OAuth Published App Integration
 * APIs.
 */
class OAuthPublishedAppsService {
    constructor(client) {
        this.client = client;
    }
    async _list(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/accounts/${config.accountId}/oauth2/published-apps`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get all the available published OAuth apps in Databricks.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.apps || response.apps.length === 0) {
                break;
            }
            for (const v of response.apps) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
}
exports.OAuthPublishedAppsService = OAuthPublishedAppsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], OAuthPublishedAppsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], OAuthPublishedAppsService.prototype, "list", null);
class PublishedAppIntegrationRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("PublishedAppIntegration", method, message);
    }
}
exports.PublishedAppIntegrationRetriableError = PublishedAppIntegrationRetriableError;
class PublishedAppIntegrationError extends apiError_1.ApiError {
    constructor(method, message) {
        super("PublishedAppIntegration", method, message);
    }
}
exports.PublishedAppIntegrationError = PublishedAppIntegrationError;
/**
 * These APIs enable administrators to manage published OAuth app integrations,
 * which is required for adding/using Published OAuth App Integration like
 * Tableau Desktop for Databricks in AWS cloud.
 */
class PublishedAppIntegrationService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("app_id")) {
            body["app_id"] = request.app_id;
        }
        if (request.hasOwnProperty("token_access_policy")) {
            body["token_access_policy"] = request.token_access_policy;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/oauth2/published-app-integrations`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create Published OAuth App Integration.
     *
     * You can retrieve the published OAuth app integration via
     * :method:PublishedAppIntegration/get.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/oauth2/published-app-integrations/${request.integration_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete an existing Published OAuth App Integration. You can retrieve the
     * published OAuth app integration via :method:PublishedAppIntegration/get.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/oauth2/published-app-integrations/${request.integration_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Gets the Published OAuth App Integration for the given integration id.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/accounts/${config.accountId}/oauth2/published-app-integrations`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get the list of published OAuth app integrations for the specified
     * Databricks account
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.apps || response.apps.length === 0) {
                break;
            }
            for (const v of response.apps) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _update(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("token_access_policy")) {
            body["token_access_policy"] = request.token_access_policy;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/oauth2/published-app-integrations/${request.integration_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Updates an existing published OAuth App Integration. You can retrieve the
     * published OAuth app integration via :method:PublishedAppIntegration/get.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.PublishedAppIntegrationService = PublishedAppIntegrationService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PublishedAppIntegrationService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PublishedAppIntegrationService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PublishedAppIntegrationService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PublishedAppIntegrationService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PublishedAppIntegrationService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PublishedAppIntegrationService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PublishedAppIntegrationService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], PublishedAppIntegrationService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PublishedAppIntegrationService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], PublishedAppIntegrationService.prototype, "update", null);
class ServicePrincipalFederationPolicyRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("ServicePrincipalFederationPolicy", method, message);
    }
}
exports.ServicePrincipalFederationPolicyRetriableError = ServicePrincipalFederationPolicyRetriableError;
class ServicePrincipalFederationPolicyError extends apiError_1.ApiError {
    constructor(method, message) {
        super("ServicePrincipalFederationPolicy", method, message);
    }
}
exports.ServicePrincipalFederationPolicyError = ServicePrincipalFederationPolicyError;
/**
 * These APIs manage service principal federation policies.
 *
 * Service principal federation, also known as Workload Identity Federation,
 * allows your automated workloads running outside of Databricks to securely
 * access Databricks APIs without the need for Databricks secrets. With Workload
 * Identity Federation, your application (or workload) authenticates to
 * Databricks as a Databricks service principal, using tokens provided by the
 * workload runtime.
 *
 * Databricks strongly recommends using Workload Identity Federation to
 * authenticate to Databricks from automated workloads, over alternatives such as
 * OAuth client secrets or Personal Access Tokens, whenever possible. Workload
 * Identity Federation is supported by many popular services, including Github
 * Actions, Azure DevOps, GitLab, Terraform Cloud, and Kubernetes clusters, among
 * others.
 *
 * Workload identity federation is configured in your Databricks account using a
 * service principal federation policy. A service principal federation policy
 * specifies: * which IdP, or issuer, the service principal is allowed to
 * authenticate from * which workload identity, or subject, is allowed to
 * authenticate as the Databricks service principal
 *
 * To configure a federation policy, you provide the following: * The required
 * token __issuer__, as specified in the “iss” claim of workload identity
 * tokens. The issuer is an https URL that identifies the workload identity
 * provider. * The required token __subject__, as specified in the “sub”
 * claim of workload identity tokens. The subject uniquely identifies the
 * workload in the workload runtime environment. * The allowed token
 * __audiences__, as specified in the “aud” claim of workload identity
 * tokens. The audience is intended to represent the recipient of the token. As
 * long as the audience in the token matches at least one audience in the policy,
 * the token is considered a match. If unspecified, the default value is your
 * Databricks account id. * Optionally, the public keys used to validate the
 * signature of the workload identity tokens, in JWKS format. If unspecified
 * (recommended), Databricks automatically fetches the public keys from the
 * issuer’s well known endpoint. Databricks strongly recommends relying on the
 * issuer’s well known endpoint for discovering public keys.
 *
 * An example service principal federation policy, for a Github Actions workload,
 * is: ``` issuer: "https://token.actions.githubusercontent.com" audiences:
 * ["https://github.com/my-github-org"] subject:
 * "repo:my-github-org/my-repo:environment:prod" ```
 *
 * An example JWT token body that matches this policy and could be used to
 * authenticate to Databricks is: ``` { "iss":
 * "https://token.actions.githubusercontent.com", "aud":
 * "https://github.com/my-github-org", "sub":
 * "repo:my-github-org/my-repo:environment:prod" } ```
 *
 * You may also need to configure the workload runtime to generate tokens for
 * your workloads.
 *
 * You do not need to configure an OAuth application in Databricks to use token
 * federation.
 */
class ServicePrincipalFederationPolicyService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.policy;
        const query = {};
        if (request.hasOwnProperty("policy_id")) {
            query["policy_id"] =
                request.policy_id;
        }
        const path = `/api/2.0/accounts/${config.accountId}/servicePrincipals/${request.service_principal_id}/federationPolicies`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create account federation policy.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/servicePrincipals/${request.service_principal_id}/federationPolicies/${request.policy_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete account federation policy.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/servicePrincipals/${request.service_principal_id}/federationPolicies/${request.policy_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get account federation policy.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/accounts/${config.accountId}/servicePrincipals/${request.service_principal_id}/federationPolicies`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List account federation policies.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.policies || response.policies.length === 0) {
                break;
            }
            for (const v of response.policies) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _update(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.policy;
        const query = {};
        if (request.hasOwnProperty("update_mask")) {
            query["update_mask"] =
                request.update_mask;
        }
        const path = `/api/2.0/accounts/${config.accountId}/servicePrincipals/${request.service_principal_id}/federationPolicies/${request.policy_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update account federation policy.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.ServicePrincipalFederationPolicyService = ServicePrincipalFederationPolicyService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalFederationPolicyService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalFederationPolicyService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalFederationPolicyService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalFederationPolicyService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalFederationPolicyService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalFederationPolicyService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalFederationPolicyService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ServicePrincipalFederationPolicyService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalFederationPolicyService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalFederationPolicyService.prototype, "update", null);
class ServicePrincipalSecretsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("ServicePrincipalSecrets", method, message);
    }
}
exports.ServicePrincipalSecretsRetriableError = ServicePrincipalSecretsRetriableError;
class ServicePrincipalSecretsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("ServicePrincipalSecrets", method, message);
    }
}
exports.ServicePrincipalSecretsError = ServicePrincipalSecretsError;
/**
 * These APIs enable administrators to manage service principal secrets.
 *
 * You can use the generated secrets to obtain OAuth access tokens for a service
 * principal, which can then be used to access Databricks Accounts and Workspace
 * APIs. For more information, see [Authentication using OAuth tokens for service
 * principals].
 *
 * In addition, the generated secrets can be used to configure the Databricks
 * Terraform Provider to authenticate with the service principal. For more
 * information, see [Databricks Terraform Provider].
 *
 * [Authentication using OAuth tokens for service principals]: https://docs.databricks.com/dev-tools/authentication-oauth.html
 * [Databricks Terraform Provider]: https://github.com/databricks/terraform-provider-databricks/blob/master/docs/index.md#authenticating-with-service-principal
 */
class ServicePrincipalSecretsService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("lifetime")) {
            body["lifetime"] = request.lifetime;
        }
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/servicePrincipals/${request.service_principal_id}/credentials/secrets`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a secret for the given service principal.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/${config.accountId}/servicePrincipals/${request.service_principal_id}/credentials/secrets/${request.secret_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a secret from the given service principal.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _list(request, context) {
        const config = this.client.config;
        await config.ensureResolved();
        if (!config.accountId ||
            config.configType() !== HostType_1.ConfigType.accountConfig) {
            throw new config_1.ConfigError("invalid Databricks Account configuration", config);
        }
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/accounts/${config.accountId}/servicePrincipals/${request.service_principal_id}/credentials/secrets`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List all secrets associated with the given service principal. This
     * operation only returns information about the secrets themselves and does
     * not include the secret values.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.secrets || response.secrets.length === 0) {
                break;
            }
            for (const v of response.secrets) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
}
exports.ServicePrincipalSecretsService = ServicePrincipalSecretsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalSecretsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalSecretsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalSecretsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalSecretsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalSecretsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ServicePrincipalSecretsService.prototype, "list", null);
class ServicePrincipalSecretsProxyRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("ServicePrincipalSecretsProxy", method, message);
    }
}
exports.ServicePrincipalSecretsProxyRetriableError = ServicePrincipalSecretsProxyRetriableError;
class ServicePrincipalSecretsProxyError extends apiError_1.ApiError {
    constructor(method, message) {
        super("ServicePrincipalSecretsProxy", method, message);
    }
}
exports.ServicePrincipalSecretsProxyError = ServicePrincipalSecretsProxyError;
/**
 * These APIs enable administrators to manage service principal secrets at the
 * workspace level. To use these APIs, the service principal must be first added
 * to the current workspace.
 *
 * You can use the generated secrets to obtain OAuth access tokens for a service
 * principal, which can then be used to access Databricks Accounts and Workspace
 * APIs. For more information, see [Authentication using OAuth tokens for service
 * principals].
 *
 * In addition, the generated secrets can be used to configure the Databricks
 * Terraform Providerto authenticate with the service principal. For more
 * information, see [Databricks Terraform Provider].
 *
 * [Authentication using OAuth tokens for service principals]: https://docs.databricks.com/dev-tools/authentication-oauth.html
 * [Databricks Terraform Provider]: https://github.com/databricks/terraform-provider-databricks/blob/master/docs/index.md#authenticating-with-service-principal
 */
class ServicePrincipalSecretsProxyService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("lifetime")) {
            body["lifetime"] = request.lifetime;
        }
        const query = {};
        const path = `/api/2.0/accounts/servicePrincipals/${request.service_principal_id}/credentials/secrets`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a secret for the given service principal.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({});
        const body = undefined;
        const query = {};
        const path = `/api/2.0/accounts/servicePrincipals/${request.service_principal_id}/credentials/secrets/${request.secret_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a secret from the given service principal.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/accounts/servicePrincipals/${request.service_principal_id}/credentials/secrets`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List all secrets associated with the given service principal. This
     * operation only returns information about the secrets themselves and does
     * not include the secret values.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.secrets || response.secrets.length === 0) {
                break;
            }
            for (const v of response.secrets) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
}
exports.ServicePrincipalSecretsProxyService = ServicePrincipalSecretsProxyService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalSecretsProxyService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalSecretsProxyService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalSecretsProxyService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalSecretsProxyService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], ServicePrincipalSecretsProxyService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], ServicePrincipalSecretsProxyService.prototype, "list", null);
