// @databricks/appkit-ui/dist/js/arrow/arrow-client.js
import { DATE_FIELD_PATTERNS, METADATA_DATE_PATTERNS, NAME_FIELD_PATTERNS } from "../constants.js";
import { getArrowModule, getDecimalTypeId, getTypeIdSets, initializeTypeIdSets } from "./lazy-arrow.js";

//#region src/js/arrow/arrow-client.ts
var ArrowClient = class ArrowClient {
	/**
	* Processes an Arrow IPC buffer into a Table.
	* Lazily loads the Apache Arrow library on first use.
	*
	* @param buffer - The Arrow IPC format buffer
	* @returns Promise resolving to an Arrow Table
	*/
	static async processArrowBuffer(buffer) {
		try {
			const arrow = await getArrowModule();
			await initializeTypeIdSets();
			return arrow.tableFromIPC(buffer);
		} catch (error) {
			throw new Error(`Failed to process Arrow buffer: ${error instanceof Error ? error.message : "Unknown error"}`);
		}
	}
	/**
	* Fetches Arrow data from a URL and processes it into a Table.
	* Convenience method that combines fetchArrow and processArrowBuffer.
	*
	* @param url - URL to fetch Arrow data from
	* @param headers - Optional HTTP headers to include in the request
	* @returns Promise resolving to an Arrow Table
	* @example
	* ```typescript
	* const table = await ArrowClient.fetchAndProcessArrow('/api/data/arrow');
	* console.log(`Loaded ${table.numRows} rows`);
	* ```
	*/
	static async fetchAndProcessArrow(url, headers) {
		try {
			const buffer = await ArrowClient.fetchArrow(url, headers);
			return ArrowClient.processArrowBuffer(buffer);
		} catch (error) {
			throw new Error(`Failed to fetch Arrow data: ${error instanceof Error ? error.message : "Unknown error"}`);
		}
	}
	/**
	* Extracts field metadata (name and type) from an Arrow Table.
	*
	* @param table - Arrow Table to extract fields from
	* @returns Array of field metadata objects
	* @example
	* ```typescript
	* const fields = ArrowClient.extractArrowFields(table);
	* // [{ name: "date", type: Date32 }, { name: "value", type: Float64 }]
	* ```
	*/
	static extractArrowFields(table) {
		return table.schema.fields.map((field) => {
			return {
				name: field.name,
				type: field.type
			};
		});
	}
	/**
	* Extracts all columns from an Arrow Table as JavaScript arrays.
	* Each column is converted to a native JavaScript array.
	*
	* @param table - Arrow Table to extract columns from
	* @returns Object mapping column names to arrays
	* @example
	* ```typescript
	* const columns = ArrowClient.extractArrowColumns(table);
	* // { date: ["2024-01-01", "2024-01-02"], value: [100, 200] }
	* ```
	*/
	static extractArrowColumns(table) {
		const cols = {};
		for (const field of table.schema.fields) {
			const child = table.getChild(field.name);
			if (child) cols[field.name] = child.toArray();
		}
		return cols;
	}
	/**
	* Extracts chart data from Arrow table.
	* Uses get(i) to properly handle complex types like Decimal128.
	* Applies decimal scaling for DECIMAL types.
	*
	* Note: This method assumes Arrow has been loaded (via processArrowBuffer).
	*
	* @returns xData for axis, yDataMap for series data
	*/
	static extractChartData(table, xKey, yKeys) {
		if (table.numRows === 0) return EMPTY_RESULT;
		const decimalType = getDecimalTypeId();
		const decimalDivisors = /* @__PURE__ */ new Map();
		for (const field of table.schema.fields) if (field.typeId === decimalType) {
			const decType = field.type;
			if (typeof decType.scale === "number") decimalDivisors.set(field.name, 10 ** decType.scale);
		}
		const xData = extractColumnValues(table.getChild(xKey), decimalDivisors.get(xKey));
		const yDataMap = {};
		for (let i = 0; i < yKeys.length; i++) {
			const key = yKeys[i];
			yDataMap[key] = extractColumnValues(table.getChild(key), decimalDivisors.get(key));
		}
		return {
			xData,
			yDataMap
		};
	}
	/**
	* Automatically detect which fields to use for chart axes from an Arrow table
	* Uses the schema's type information for accurate field detection
	*
	* Note: This method assumes Arrow has been loaded (via processArrowBuffer).
	*
	* @param table - Arrow Table to analyze
	* @param orientation - Chart orientation ("vertical" for time-series, "horizontal" for categorical)
	* @returns Object containing the detected fields
	* @example
	* // Time-series data
	* detectFieldsFromArrow(timeSeriesTable)
	* // { xField: "date", yFields: ["revenue", "cost"], chartType: "timeseries" }
	*
	* // Categorical data
	* detectFieldsFromArrow(categoricalTable)
	* // { xField: "app_name", yFields: ["totalSpend"], chartType: "categorical" }
	*/
	static detectFieldsFromArrow(table, orientation) {
		const fields = table.schema.fields;
		if (fields.length === 0) return {
			xField: "x",
			yFields: ["y"],
			chartType: "categorical"
		};
		const fieldNames = fields.map((f) => f.name);
		const typeIdSets = getTypeIdSets();
		const temporalFields = [];
		const numericFields = [];
		const stringFields = [];
		for (const field of fields) {
			const typeId = field.typeId;
			if (typeIdSets.temporal.has(typeId)) temporalFields.push(field.name);
			else if (typeIdSets.numeric.has(typeId)) numericFields.push(field.name);
			else if (typeIdSets.string.has(typeId)) stringFields.push(field.name);
		}
		let nameFields = stringFields.filter((name) => NAME_FIELD_PATTERNS.some((pattern) => name.toLowerCase().includes(pattern)));
		if (nameFields.length === 0) nameFields = stringFields.filter((name) => !name.toLowerCase().endsWith("_id"));
		const chartDateFields = temporalFields.filter((name) => !METADATA_DATE_PATTERNS.some((pattern) => name.toLowerCase().includes(pattern)));
		const metadataDateFields = temporalFields.filter((name) => METADATA_DATE_PATTERNS.some((pattern) => name.toLowerCase().includes(pattern)));
		const stringDateFields = stringFields.filter((name) => DATE_FIELD_PATTERNS.some((pattern) => name.toLowerCase().includes(pattern)) && !METADATA_DATE_PATTERNS.some((pattern) => name.toLowerCase().includes(pattern)));
		const primaryDateFields = [...chartDateFields, ...stringDateFields];
		const isTimeSeries = primaryDateFields.length > 0 && orientation !== "horizontal";
		const isCategorical = nameFields.length > 0 && (primaryDateFields.length === 0 || orientation === "horizontal");
		if (orientation === "horizontal" || isCategorical) {
			const xField = nameFields[0] || primaryDateFields[0] || metadataDateFields[0] || fieldNames[0];
			return {
				xField,
				yFields: numericFields.length > 0 ? numericFields : fieldNames.filter((k) => k !== xField),
				chartType: "categorical"
			};
		}
		const xField = primaryDateFields[0] || metadataDateFields[0] || nameFields[0] || fieldNames[0];
		return {
			xField,
			yFields: numericFields.length > 0 ? numericFields : fieldNames.filter((k) => k !== xField),
			chartType: isTimeSeries ? "timeseries" : "categorical"
		};
	}
	/**
	* Fetches raw Arrow IPC data from a URL.
	*
	* @param url - URL to fetch Arrow data from
	* @param headers - Optional HTTP headers to include in the request
	* @returns Promise resolving to the raw Arrow buffer as Uint8Array
	* @example
	* ```typescript
	* const buffer = await ArrowClient.fetchArrow('/api/data/arrow');
	* const table = await ArrowClient.processArrowBuffer(buffer);
	* ```
	*/
	static async fetchArrow(url, headers) {
		try {
			const response = await fetch(url, { headers: {
				"Content-Type": "application/octet-stream",
				...headers
			} });
			if (!response.ok) throw new Error(`HTTP ${response.status}: ${response.statusText}`);
			const buffer = await response.arrayBuffer();
			return new Uint8Array(buffer);
		} catch (error) {
			throw new Error(`Failed to fetch Arrow data: ${error instanceof Error ? error.message : "Unknown error"}`);
		}
	}
};
const EMPTY_RESULT = {
	xData: [],
	yDataMap: {}
};
/**
* Extracts values from an Arrow Vector properly.
* Uses get(i) to handle complex types like Decimal128 correctly.
* toArray() doesn't work properly for Decimal types - it returns raw bytes.
*
* @param col - The Arrow column/vector
* @param divisor - Pre-computed divisor for DECIMAL types (10^scale)
*/
function extractColumnValues(col, divisor) {
	if (!col) return [];
	const len = col.length;
	const result = new Array(len);
	for (let i = 0; i < len; i++) {
		const val = col.get(i);
		if (val === null || val === void 0) result[i] = 0;
		else if (typeof val === "bigint") {
			const num = Number(val);
			result[i] = divisor !== void 0 ? num / divisor : num;
		} else if (typeof val === "number") result[i] = divisor !== void 0 ? val / divisor : val;
		else if (typeof val === "string") result[i] = val;
		else {
			const num = Number(val);
			result[i] = divisor !== void 0 ? num / divisor : num;
		}
	}
	return result;
}

//#endregion
export { ArrowClient };
