// @databricks/sdk-experimental/dist/apis/dashboards/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LakeviewEmbeddedService = exports.LakeviewService = exports.GenieService = void 0;
const dashboards = __importStar(require("./model"));
const retries_1 = __importDefault(require("../../retries/retries"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
const wait_1 = require("../../wait");
class GenieRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Genie", method, message);
    }
}
class GenieError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Genie", method, message);
    }
}
/**
 * Genie provides a no-code experience for business users, powered by AI/BI.
 * Analysts set up spaces that business users can use to ask questions using
 * natural language. Genie uses data registered to Unity Catalog and requires at
 * least CAN USE permission on a Pro or Serverless SQL warehouse. Also,
 * Databricks Assistant must be enabled.
 */
class GenieService {
    constructor(client) {
        this.client = client;
    }
    async _createMessage(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("content")) {
            body["content"] = request.content;
        }
        const query = {};
        const path = `/api/2.0/genie/spaces/${request.space_id}/conversations/${request.conversation_id}/messages`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create new message in a [conversation](:method:genie/startconversation).
     * The AI response uses all previously created messages in the conversation
     * to respond.
     */
    async createMessage(genieCreateConversationMessageRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const genieMessage = await this._createMessage(genieCreateConversationMessageRequest, context);
        return (0, wait_1.asWaiter)(genieMessage, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.getMessage({
                        conversation_id: genieCreateConversationMessageRequest.conversation_id,
                        message_id: genieMessage.message_id,
                        space_id: genieCreateConversationMessageRequest.space_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Genie.createMessageAndWait: cancelled");
                        throw new GenieError("createMessageAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.status;
                    const statusMessage = pollResponse;
                    switch (status) {
                        case "COMPLETED": {
                            return pollResponse;
                        }
                        case "FAILED": {
                            const errorMessage = `failed to reach COMPLETED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Genie.createMessageAndWait: ${errorMessage}`);
                            throw new GenieError("createMessageAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach COMPLETED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Genie.createMessageAndWait: retrying: ${errorMessage}`);
                            throw new GenieRetriableError("createMessageAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _deleteConversation(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/genie/spaces/${request.space_id}/conversations/${request.conversation_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a conversation.
     */
    async deleteConversation(request, context) {
        return await this._deleteConversation(request, context);
    }
    async _deleteConversationMessage(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/genie/spaces/${request.space_id}/conversations/${request.conversation_id}/messages/${request.message_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a conversation message.
     */
    async deleteConversationMessage(request, context) {
        return await this._deleteConversationMessage(request, context);
    }
    async _executeMessageAttachmentQuery(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/genie/spaces/${request.space_id}/conversations/${request.conversation_id}/messages/${request.message_id}/attachments/${request.attachment_id}/execute-query`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Execute the SQL for a message query attachment. Use this API when the
     * query attachment has expired and needs to be re-executed.
     */
    async executeMessageAttachmentQuery(request, context) {
        return await this._executeMessageAttachmentQuery(request, context);
    }
    async _executeMessageQuery(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/genie/spaces/${request.space_id}/conversations/${request.conversation_id}/messages/${request.message_id}/execute-query`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * DEPRECATED: Use [Execute Message Attachment
     * Query](:method:genie/executemessageattachmentquery) instead.
     */
    async executeMessageQuery(request, context) {
        return await this._executeMessageQuery(request, context);
    }
    async _getMessage(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/genie/spaces/${request.space_id}/conversations/${request.conversation_id}/messages/${request.message_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get message from conversation.
     */
    async getMessage(request, context) {
        return await this._getMessage(request, context);
    }
    async _getMessageAttachmentQueryResult(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/genie/spaces/${request.space_id}/conversations/${request.conversation_id}/messages/${request.message_id}/attachments/${request.attachment_id}/query-result`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get the result of SQL query if the message has a query attachment. This is
     * only available if a message has a query attachment and the message status
     * is `EXECUTING_QUERY` OR `COMPLETED`.
     */
    async getMessageAttachmentQueryResult(request, context) {
        return await this._getMessageAttachmentQueryResult(request, context);
    }
    async _getMessageQueryResult(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/genie/spaces/${request.space_id}/conversations/${request.conversation_id}/messages/${request.message_id}/query-result`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * DEPRECATED: Use [Get Message Attachment Query
     * Result](:method:genie/getmessageattachmentqueryresult) instead.
     */
    async getMessageQueryResult(request, context) {
        return await this._getMessageQueryResult(request, context);
    }
    async _getMessageQueryResultByAttachment(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/genie/spaces/${request.space_id}/conversations/${request.conversation_id}/messages/${request.message_id}/query-result/${request.attachment_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * DEPRECATED: Use [Get Message Attachment Query
     * Result](:method:genie/getmessageattachmentqueryresult) instead.
     */
    async getMessageQueryResultByAttachment(request, context) {
        return await this._getMessageQueryResultByAttachment(request, context);
    }
    async _getSpace(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/genie/spaces/${request.space_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get details of a Genie Space.
     */
    async getSpace(request, context) {
        return await this._getSpace(request, context);
    }
    async _listConversationMessages(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/genie/spaces/${request.space_id}/conversations/${request.conversation_id}/messages`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List messages in a conversation
     */
    async listConversationMessages(request, context) {
        return await this._listConversationMessages(request, context);
    }
    async _listConversations(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("include_all")) {
            query["include_all"] =
                request.include_all;
        }
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/genie/spaces/${request.space_id}/conversations`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a list of conversations in a Genie Space.
     */
    async listConversations(request, context) {
        return await this._listConversations(request, context);
    }
    async _listSpaces(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/genie/spaces";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get list of Genie Spaces.
     */
    async listSpaces(request, context) {
        return await this._listSpaces(request, context);
    }
    async _sendMessageFeedback(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("rating")) {
            body["rating"] = request.rating;
        }
        const query = {};
        const path = `/api/2.0/genie/spaces/${request.space_id}/conversations/${request.conversation_id}/messages/${request.message_id}/feedback`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Send feedback for a message.
     */
    async sendMessageFeedback(request, context) {
        return await this._sendMessageFeedback(request, context);
    }
    async _startConversation(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("content")) {
            body["content"] = request.content;
        }
        const query = {};
        const path = `/api/2.0/genie/spaces/${request.space_id}/start-conversation`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Start a new conversation.
     */
    async startConversation(genieStartConversationMessageRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const genieStartConversationResponse = await this._startConversation(genieStartConversationMessageRequest, context);
        return (0, wait_1.asWaiter)(genieStartConversationResponse, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.getMessage({
                        conversation_id: genieStartConversationResponse.conversation_id,
                        message_id: genieStartConversationResponse.message_id,
                        space_id: genieStartConversationMessageRequest.space_id,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("Genie.startConversationAndWait: cancelled");
                        throw new GenieError("startConversationAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.status;
                    const statusMessage = pollResponse;
                    switch (status) {
                        case "COMPLETED": {
                            return pollResponse;
                        }
                        case "FAILED": {
                            const errorMessage = `failed to reach COMPLETED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Genie.startConversationAndWait: ${errorMessage}`);
                            throw new GenieError("startConversationAndWait", errorMessage);
                        }
                        default: {
                            const errorMessage = `failed to reach COMPLETED state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`Genie.startConversationAndWait: retrying: ${errorMessage}`);
                            throw new GenieRetriableError("startConversationAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _trashSpace(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/genie/spaces/${request.space_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Move a Genie Space to the trash.
     */
    async trashSpace(request, context) {
        return await this._trashSpace(request, context);
    }
}
exports.GenieService = GenieService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_createMessage", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "createMessage", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_deleteConversation", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "deleteConversation", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_deleteConversationMessage", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "deleteConversationMessage", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_executeMessageAttachmentQuery", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "executeMessageAttachmentQuery", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_executeMessageQuery", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "executeMessageQuery", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_getMessage", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "getMessage", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_getMessageAttachmentQueryResult", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "getMessageAttachmentQueryResult", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_getMessageQueryResult", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "getMessageQueryResult", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_getMessageQueryResultByAttachment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "getMessageQueryResultByAttachment", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_getSpace", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "getSpace", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_listConversationMessages", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "listConversationMessages", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_listConversations", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "listConversations", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_listSpaces", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "listSpaces", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_sendMessageFeedback", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "sendMessageFeedback", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_startConversation", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "startConversation", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "_trashSpace", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], GenieService.prototype, "trashSpace", null);
class LakeviewRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("Lakeview", method, message);
    }
}
class LakeviewError extends apiError_1.ApiError {
    constructor(method, message) {
        super("Lakeview", method, message);
    }
}
/**
 * These APIs provide specific management operations for Lakeview dashboards.
 * Generic resource management can be done with Workspace API (import, export,
 * get-status, list, delete).
 */
class LakeviewService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.dashboard;
        const query = {};
        const path = "/api/2.0/lakeview/dashboards";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a draft dashboard.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _createSchedule(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.schedule;
        const query = {};
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}/schedules`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create dashboard schedule.
     */
    async createSchedule(request, context) {
        return await this._createSchedule(request, context);
    }
    async _createSubscription(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.subscription;
        const query = {};
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}/schedules/${request.schedule_id}/subscriptions`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create schedule subscription.
     */
    async createSubscription(request, context) {
        return await this._createSubscription(request, context);
    }
    async _deleteSchedule(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("etag")) {
            query["etag"] = request.etag;
        }
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}/schedules/${request.schedule_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete dashboard schedule.
     */
    async deleteSchedule(request, context) {
        return await this._deleteSchedule(request, context);
    }
    async _deleteSubscription(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("etag")) {
            query["etag"] = request.etag;
        }
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}/schedules/${request.schedule_id}/subscriptions/${request.subscription_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete schedule subscription.
     */
    async deleteSubscription(request, context) {
        return await this._deleteSubscription(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a draft dashboard.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _getPublished(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}/published`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get the current published dashboard.
     */
    async getPublished(request, context) {
        return await this._getPublished(request, context);
    }
    async _getSchedule(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}/schedules/${request.schedule_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get dashboard schedule.
     */
    async getSchedule(request, context) {
        return await this._getSchedule(request, context);
    }
    async _getSubscription(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}/schedules/${request.schedule_id}/subscriptions/${request.subscription_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get schedule subscription.
     */
    async getSubscription(request, context) {
        return await this._getSubscription(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        if (request.hasOwnProperty("show_trashed")) {
            query["show_trashed"] =
                request.show_trashed;
        }
        if (request.hasOwnProperty("view")) {
            query["view"] = request.view;
        }
        const path = "/api/2.0/lakeview/dashboards";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List dashboards.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.dashboards || response.dashboards.length === 0) {
                break;
            }
            for (const v of response.dashboards) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _listSchedules(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}/schedules`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List dashboard schedules.
     */
    async *listSchedules(request, context) {
        while (true) {
            const response = await this._listSchedules(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.schedules || response.schedules.length === 0) {
                break;
            }
            for (const v of response.schedules) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _listSubscriptions(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}/schedules/${request.schedule_id}/subscriptions`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List schedule subscriptions.
     */
    async *listSubscriptions(request, context) {
        while (true) {
            const response = await this._listSubscriptions(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.subscriptions ||
                response.subscriptions.length === 0) {
                break;
            }
            for (const v of response.subscriptions) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _migrate(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("display_name")) {
            body["display_name"] = request.display_name;
        }
        if (request.hasOwnProperty("parent_path")) {
            body["parent_path"] = request.parent_path;
        }
        if (request.hasOwnProperty("source_dashboard_id")) {
            body["source_dashboard_id"] = request.source_dashboard_id;
        }
        if (request.hasOwnProperty("update_parameter_syntax")) {
            body["update_parameter_syntax"] = request.update_parameter_syntax;
        }
        const query = {};
        const path = "/api/2.0/lakeview/dashboards/migrate";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Migrates a classic SQL dashboard to Lakeview.
     */
    async migrate(request, context) {
        return await this._migrate(request, context);
    }
    async _publish(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("embed_credentials")) {
            body["embed_credentials"] = request.embed_credentials;
        }
        if (request.hasOwnProperty("warehouse_id")) {
            body["warehouse_id"] = request.warehouse_id;
        }
        const query = {};
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}/published`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Publish the current draft dashboard.
     */
    async publish(request, context) {
        return await this._publish(request, context);
    }
    async _trash(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Trash a dashboard.
     */
    async trash(request, context) {
        return await this._trash(request, context);
    }
    async _unpublish(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}/published`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Unpublish the dashboard.
     */
    async unpublish(request, context) {
        return await this._unpublish(request, context);
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.dashboard;
        const query = {};
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update a draft dashboard.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
    async _updateSchedule(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.schedule;
        const query = {};
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}/schedules/${request.schedule_id}`;
        return (await this.client.request({
            path,
            method: "PUT",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update dashboard schedule.
     */
    async updateSchedule(request, context) {
        return await this._updateSchedule(request, context);
    }
}
exports.LakeviewService = LakeviewService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_createSchedule", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "createSchedule", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_createSubscription", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "createSubscription", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_deleteSchedule", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "deleteSchedule", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_deleteSubscription", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "deleteSubscription", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_getPublished", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "getPublished", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_getSchedule", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "getSchedule", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_getSubscription", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "getSubscription", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], LakeviewService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_listSchedules", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], LakeviewService.prototype, "listSchedules", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_listSubscriptions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], LakeviewService.prototype, "listSubscriptions", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_migrate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "migrate", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_publish", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "publish", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_trash", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "trash", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_unpublish", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "unpublish", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "_updateSchedule", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewService.prototype, "updateSchedule", null);
class LakeviewEmbeddedRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("LakeviewEmbedded", method, message);
    }
}
class LakeviewEmbeddedError extends apiError_1.ApiError {
    constructor(method, message) {
        super("LakeviewEmbedded", method, message);
    }
}
/**
 * Token-based Lakeview APIs for embedding dashboards in external applications.
 */
class LakeviewEmbeddedService {
    constructor(client) {
        this.client = client;
    }
    async _getPublishedDashboardTokenInfo(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("external_value")) {
            query["external_value"] =
                request.external_value;
        }
        if (request.hasOwnProperty("external_viewer_id")) {
            query["external_viewer_id"] =
                request.external_viewer_id;
        }
        const path = `/api/2.0/lakeview/dashboards/${request.dashboard_id}/published/tokeninfo`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a required authorization details and scopes of a published dashboard
     * to mint an OAuth token.
     */
    async getPublishedDashboardTokenInfo(request, context) {
        return await this._getPublishedDashboardTokenInfo(request, context);
    }
}
exports.LakeviewEmbeddedService = LakeviewEmbeddedService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewEmbeddedService.prototype, "_getPublishedDashboardTokenInfo", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], LakeviewEmbeddedService.prototype, "getPublishedDashboardTokenInfo", null);
