// @databricks/sdk-experimental/dist/apis/cleanrooms/api.js
"use strict";
// Code generated from OpenAPI specs by Databricks SDK Generator. DO NOT EDIT.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CleanRoomsService = exports.CleanRoomTaskRunsService = exports.CleanRoomAutoApprovalRulesService = exports.CleanRoomAssetsService = exports.CleanRoomAssetRevisionsService = void 0;
const cleanrooms = __importStar(require("./model"));
const retries_1 = __importDefault(require("../../retries/retries"));
const apiError_1 = require("../apiError");
const context_1 = require("../../context");
const logging_1 = require("../../logging");
const wait_1 = require("../../wait");
class CleanRoomAssetRevisionsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("CleanRoomAssetRevisions", method, message);
    }
}
class CleanRoomAssetRevisionsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("CleanRoomAssetRevisions", method, message);
    }
}
/**
 * Clean Room Asset Revisions denote new versions of uploaded assets (e.g.
 * notebooks) in the clean room.
 */
class CleanRoomAssetRevisionsService {
    constructor(client) {
        this.client = client;
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/assets/${request.asset_type}/${request.name}/revisions/${request.etag}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a specific revision of an asset
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/assets/${request.asset_type}/${request.name}/revisions`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List revisions for an asset
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.revisions || response.revisions.length === 0) {
                break;
            }
            for (const v of response.revisions) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
}
exports.CleanRoomAssetRevisionsService = CleanRoomAssetRevisionsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetRevisionsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetRevisionsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetRevisionsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], CleanRoomAssetRevisionsService.prototype, "list", null);
class CleanRoomAssetsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("CleanRoomAssets", method, message);
    }
}
class CleanRoomAssetsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("CleanRoomAssets", method, message);
    }
}
/**
 * Clean room assets are data and code objects — Tables, volumes, and notebooks
 * that are shared with the clean room.
 */
class CleanRoomAssetsService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.asset;
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/assets`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a clean room asset —share an asset like a notebook or table into
     * the clean room. For each UC asset that is added through this method, the
     * clean room owner must also have enough privilege on the asset to consume
     * it. The privilege must be maintained indefinitely for the clean room to be
     * able to access the asset. Typically, you should use a group as the clean
     * room owner.
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _createCleanRoomAssetReview(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("notebook_review")) {
            body["notebook_review"] = request.notebook_review;
        }
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/assets/${request.asset_type}/${request.name}/reviews`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Submit an asset review
     */
    async createCleanRoomAssetReview(request, context) {
        return await this._createCleanRoomAssetReview(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/assets/${request.asset_type}/${request.name}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a clean room asset - unshare/remove the asset from the clean room
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/assets/${request.asset_type}/${request.name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get the details of a clean room asset by its type and full name.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/assets`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List assets.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.assets || response.assets.length === 0) {
                break;
            }
            for (const v of response.assets) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.asset;
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/assets/${request.asset_type}/${request.name}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update a clean room asset. For example, updating the content of a
     * notebook; changing the shared partitions of a table; etc.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.CleanRoomAssetsService = CleanRoomAssetsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetsService.prototype, "_createCleanRoomAssetReview", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetsService.prototype, "createCleanRoomAssetReview", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], CleanRoomAssetsService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetsService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAssetsService.prototype, "update", null);
class CleanRoomAutoApprovalRulesRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("CleanRoomAutoApprovalRules", method, message);
    }
}
class CleanRoomAutoApprovalRulesError extends apiError_1.ApiError {
    constructor(method, message) {
        super("CleanRoomAutoApprovalRules", method, message);
    }
}
/**
 * Clean room auto-approval rules automatically create an approval on your behalf
 * when an asset (e.g. notebook) meeting specific criteria is shared in a clean
 * room.
 */
class CleanRoomAutoApprovalRulesService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("auto_approval_rule")) {
            body["auto_approval_rule"] = request.auto_approval_rule;
        }
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/auto-approval-rules`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create an auto-approval rule
     */
    async create(request, context) {
        return await this._create(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/auto-approval-rules/${request.rule_id}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a auto-approval rule by rule ID
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/auto-approval-rules/${request.rule_id}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a auto-approval rule by rule ID
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/auto-approval-rules`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List all auto-approval rules for the caller
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.rules || response.rules.length === 0) {
                break;
            }
            for (const v of response.rules) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.auto_approval_rule;
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/auto-approval-rules/${request.rule_id}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update a auto-approval rule by rule ID
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.CleanRoomAutoApprovalRulesService = CleanRoomAutoApprovalRulesService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAutoApprovalRulesService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAutoApprovalRulesService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAutoApprovalRulesService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAutoApprovalRulesService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAutoApprovalRulesService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAutoApprovalRulesService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAutoApprovalRulesService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], CleanRoomAutoApprovalRulesService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAutoApprovalRulesService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomAutoApprovalRulesService.prototype, "update", null);
class CleanRoomTaskRunsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("CleanRoomTaskRuns", method, message);
    }
}
class CleanRoomTaskRunsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("CleanRoomTaskRuns", method, message);
    }
}
/**
 * Clean room task runs are the executions of notebooks in a clean room.
 */
class CleanRoomTaskRunsService {
    constructor(client) {
        this.client = client;
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("notebook_name")) {
            query["notebook_name"] =
                request.notebook_name;
        }
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/runs`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * List all the historical notebook task runs in a clean room.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.runs || response.runs.length === 0) {
                break;
            }
            for (const v of response.runs) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
}
exports.CleanRoomTaskRunsService = CleanRoomTaskRunsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomTaskRunsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], CleanRoomTaskRunsService.prototype, "list", null);
class CleanRoomsRetriableError extends apiError_1.ApiRetriableError {
    constructor(method, message) {
        super("CleanRooms", method, message);
    }
}
class CleanRoomsError extends apiError_1.ApiError {
    constructor(method, message) {
        super("CleanRooms", method, message);
    }
}
/**
 * A clean room uses Delta Sharing and serverless compute to provide a secure and
 * privacy-protecting environment where multiple parties can work together on
 * sensitive enterprise data without direct access to each other's data.
 */
class CleanRoomsService {
    constructor(client) {
        this.client = client;
    }
    async _create(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.clean_room;
        const query = {};
        const path = "/api/2.0/clean-rooms";
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create a new clean room with the specified collaborators. This method is
     * asynchronous; the returned name field inside the clean_room field can be
     * used to poll the clean room status, using the :method:cleanrooms/get
     * method. When this method returns, the clean room will be in a PROVISIONING
     * state, with only name, owner, comment, created_at and status populated.
     * The clean room will be usable once it enters an ACTIVE state.
     *
     * The caller must be a metastore admin or have the **CREATE_CLEAN_ROOM**
     * privilege on the metastore.
     */
    async create(createCleanRoomRequest, context) {
        let cancellationToken = context?.cancellationToken;
        const cleanRoom = await this._create(createCleanRoomRequest, context);
        return (0, wait_1.asWaiter)(cleanRoom, async (options) => {
            options = options || {};
            options.onProgress =
                options.onProgress || (async (newPollResponse) => { });
            let { timeout, onProgress } = options;
            return await (0, retries_1.default)({
                timeout,
                fn: async () => {
                    const pollResponse = await this.get({
                        name: cleanRoom.name,
                    }, context);
                    if (cancellationToken?.isCancellationRequested) {
                        context?.logger?.error("CleanRooms.createAndWait: cancelled");
                        throw new CleanRoomsError("createAndWait", "cancelled");
                    }
                    await onProgress(pollResponse);
                    const status = pollResponse.status;
                    const statusMessage = pollResponse;
                    switch (status) {
                        case "ACTIVE": {
                            return pollResponse;
                        }
                        default: {
                            const errorMessage = `failed to reach ACTIVE state, got ${status}: ${statusMessage}`;
                            context?.logger?.error(`CleanRooms.createAndWait: retrying: ${errorMessage}`);
                            throw new CleanRoomsRetriableError("createAndWait", errorMessage);
                        }
                    }
                },
            });
        });
    }
    async _createOutputCatalog(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = request.output_catalog;
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.clean_room_name}/output-catalogs`;
        return (await this.client.request({
            path,
            method: "POST",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Create the output catalog of the clean room.
     */
    async createOutputCatalog(request, context) {
        return await this._createOutputCatalog(request, context);
    }
    async _delete(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.name}`;
        return (await this.client.request({
            path,
            method: "DELETE",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Delete a clean room. After deletion, the clean room will be removed from
     * the metastore. If the other collaborators have not deleted the clean room,
     * they will still have the clean room in their metastore, but it will be in
     * a DELETED state and no operations other than deletion can be performed on
     * it.
     */
    async delete(request, context) {
        return await this._delete(request, context);
    }
    async _get(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.name}`;
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get the details of a clean room given its name.
     */
    async get(request, context) {
        return await this._get(request, context);
    }
    async _list(request, context) {
        const headers = new Headers({ Accept: "application/json" });
        const body = undefined;
        const query = {};
        if (request.hasOwnProperty("page_size")) {
            query["page_size"] =
                request.page_size;
        }
        if (request.hasOwnProperty("page_token")) {
            query["page_token"] =
                request.page_token;
        }
        const path = "/api/2.0/clean-rooms";
        return (await this.client.request({
            path,
            method: "GET",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Get a list of all clean rooms of the metastore. Only clean rooms the
     * caller has access to are returned.
     */
    async *list(request, context) {
        while (true) {
            const response = await this._list(request, context);
            if (context?.cancellationToken &&
                context?.cancellationToken.isCancellationRequested) {
                break;
            }
            if (!response.clean_rooms || response.clean_rooms.length === 0) {
                break;
            }
            for (const v of response.clean_rooms) {
                yield v;
            }
            request.page_token = response.next_page_token;
            if (!response.next_page_token) {
                break;
            }
        }
    }
    async _update(request, context) {
        const headers = new Headers({
            "Accept": "application/json",
            "Content-Type": "application/json",
        });
        const body = {};
        if (request.hasOwnProperty("clean_room")) {
            body["clean_room"] = request.clean_room;
        }
        const query = {};
        const path = `/api/2.0/clean-rooms/${request.name}`;
        return (await this.client.request({
            path,
            method: "PATCH",
            headers,
            raw: false,
            query: query,
            payload: body,
        }, context));
    }
    /**
     * Update a clean room. The caller must be the owner of the clean room, have
     * **MODIFY_CLEAN_ROOM** privilege, or be metastore admin.
     *
     * When the caller is a metastore admin, only the __owner__ field can be
     * updated.
     */
    async update(request, context) {
        return await this._update(request, context);
    }
}
exports.CleanRoomsService = CleanRoomsService;
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomsService.prototype, "_create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomsService.prototype, "create", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomsService.prototype, "_createOutputCatalog", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomsService.prototype, "createOutputCatalog", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomsService.prototype, "_delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomsService.prototype, "delete", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomsService.prototype, "_get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomsService.prototype, "get", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomsService.prototype, "_list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Object)
], CleanRoomsService.prototype, "list", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomsService.prototype, "_update", null);
__decorate([
    (0, logging_1.withLogContext)(logging_1.ExposedLoggers.SDK),
    __param(1, context_1.context),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, context_1.Context]),
    __metadata("design:returntype", Promise)
], CleanRoomsService.prototype, "update", null);
