// @databricks/appkit/dist/context/execution-context.js
import { ConfigurationError } from "../errors/configuration.js";
import "../errors/index.js";
import { ServiceContext } from "./service-context.js";
import { isUserContext } from "./user-context.js";
import { AsyncLocalStorage } from "node:async_hooks";

//#region src/context/execution-context.ts
/**
* AsyncLocalStorage for execution context.
* Used to pass user context through the call stack without explicit parameters.
*/
const executionContextStorage = new AsyncLocalStorage();
/**
* Run a function in the context of a user.
* All calls within the function will have access to the user context.
*
* @param userContext - The user context to use
* @param fn - The function to run
* @returns The result of the function
*/
function runInUserContext(userContext, fn) {
	return executionContextStorage.run(userContext, fn);
}
/**
* Get the current execution context.
*
* - If running inside a user context (via asUser), returns the user context
* - Otherwise, returns the service context
*
* @throws Error if ServiceContext is not initialized
*/
function getExecutionContext() {
	const userContext = executionContextStorage.getStore();
	if (userContext) return userContext;
	return ServiceContext.get();
}
/**
* Get the current user ID for cache keying and telemetry.
*
* Returns the user ID if in user context, otherwise the service user ID.
*/
function getCurrentUserId() {
	const ctx = getExecutionContext();
	if (isUserContext(ctx)) return ctx.userId;
	return ctx.serviceUserId;
}
/**
* Get the WorkspaceClient for the current execution context.
*/
function getWorkspaceClient() {
	return getExecutionContext().client;
}
/**
* Get the warehouse ID promise.
*/
function getWarehouseId() {
	const ctx = getExecutionContext();
	if (!ctx.warehouseId) throw ConfigurationError.resourceNotFound("Warehouse ID", "No plugin requires a SQL Warehouse. Add a sql_warehouse resource to your plugin manifest, or set DATABRICKS_WAREHOUSE_ID");
	return ctx.warehouseId;
}
/**
* Get the workspace ID promise.
*/
function getWorkspaceId() {
	return getExecutionContext().workspaceId;
}
/**
* Check if currently running in a user context.
*/
function isInUserContext() {
	return executionContextStorage.getStore() !== void 0;
}
/**
* Get the user context if one is active, otherwise `undefined`.
* Unlike `getExecutionContext()`, this does not require `ServiceContext`
* to be initialized and never throws.
*/
function getUserContext() {
	return executionContextStorage.getStore();
}

//#endregion
export { getCurrentUserId, getExecutionContext, getUserContext, getWarehouseId, getWorkspaceClient, getWorkspaceId, isInUserContext, runInUserContext };
