// @databricks/appkit/dist/context/service-context.js
import { AuthenticationError } from "../errors/authentication.js";
import { ConfigurationError } from "../errors/configuration.js";
import { InitializationError } from "../errors/initialization.js";
import "../errors/index.js";
import { name, version } from "../appkit/package.js";
import { createHash } from "node:crypto";
import { ConfigError, WorkspaceClient } from "@databricks/sdk-experimental";
import { coerce } from "semver";

//#region src/context/service-context.ts
function getClientOptions() {
	const isDev = process.env.NODE_ENV === "development";
	return {
		product: name,
		productVersion: coerce(version)?.version ?? version,
		...isDev && { userAgentExtra: { mode: "dev" } }
	};
}
/**
* ServiceContext is a singleton that manages the service principal's
* WorkspaceClient and shared resources like warehouse/workspace IDs.
*
* It's initialized once at app startup and provides the foundation
* for both service principal and user context execution.
*/
var ServiceContext = class ServiceContext {
	static instance = null;
	static initPromise = null;
	/**
	* Initialize the service context. Should be called once at app startup.
	* Safe to call multiple times - will return the same instance.
	*
	* @param options - Which shared resources to resolve (derived from plugin manifests).
	* @param client - Optional pre-configured WorkspaceClient to use instead
	*   of creating one from environment credentials.
	*/
	static async initialize(options, client) {
		if (ServiceContext.instance) return ServiceContext.instance;
		if (ServiceContext.initPromise) return ServiceContext.initPromise;
		ServiceContext.initPromise = ServiceContext.createContext(options, client);
		ServiceContext.instance = await ServiceContext.initPromise;
		return ServiceContext.instance;
	}
	/**
	* Get the initialized service context.
	* @throws Error if not initialized
	*/
	static get() {
		if (!ServiceContext.instance) throw InitializationError.notInitialized("ServiceContext", "Call ServiceContext.initialize() first");
		return ServiceContext.instance;
	}
	/**
	* Check if the service context has been initialized.
	*/
	static isInitialized() {
		return ServiceContext.instance !== null;
	}
	/**
	* Create a user context from request headers.
	*
	* @param token - The user's access token from x-forwarded-access-token header
	* @param userId - The user's ID from x-forwarded-user header
	* @param userName - Optional user name
	* @throws Error if token is not provided
	*/
	static createUserContext(token, userId, userName, userEmail) {
		if (!token) throw AuthenticationError.missingToken("user token");
		const host = process.env.DATABRICKS_HOST;
		if (!host) throw ConfigurationError.missingEnvVar("DATABRICKS_HOST");
		const serviceCtx = ServiceContext.get();
		return {
			client: new WorkspaceClient({
				token,
				host,
				authType: "pat"
			}, getClientOptions()),
			userId,
			userName,
			userEmail,
			tokenFingerprint: createHash("sha256").update(token).digest("hex").slice(0, 16),
			warehouseId: serviceCtx.warehouseId,
			workspaceId: serviceCtx.workspaceId,
			isUserContext: true
		};
	}
	/**
	* Get the client options for WorkspaceClient.
	* Exposed for testing purposes.
	*/
	static getClientOptions() {
		return getClientOptions();
	}
	static async createContext(options, client) {
		try {
			const wsClient = client ?? new WorkspaceClient({}, getClientOptions());
			const [resolvedWorkspaceId, currentUser, resolvedWarehouseId] = await Promise.all([
				ServiceContext.getWorkspaceId(wsClient),
				wsClient.currentUser.me(),
				options?.warehouseId ? ServiceContext.getWarehouseId(wsClient) : Promise.resolve(void 0)
			]);
			if (!currentUser.id) throw ConfigurationError.resourceNotFound("Service user ID");
			const warehouseId = options?.warehouseId && resolvedWarehouseId !== void 0 ? Promise.resolve(resolvedWarehouseId) : void 0;
			return {
				client: wsClient,
				serviceUserId: currentUser.id,
				warehouseId,
				workspaceId: Promise.resolve(resolvedWorkspaceId)
			};
		} catch (e) {
			if (e instanceof ConfigError) throw ConfigurationError.databricksAuthenticationSetupFailed(e.baseMessage, { cause: e });
			throw e;
		}
	}
	static async getWorkspaceId(client) {
		if (process.env.DATABRICKS_WORKSPACE_ID) return process.env.DATABRICKS_WORKSPACE_ID;
		const response = await client.apiClient.request({
			path: "/api/2.0/preview/scim/v2/Me",
			method: "GET",
			headers: new Headers(),
			raw: false,
			query: {},
			responseHeaders: ["x-databricks-org-id"]
		});
		if (!response["x-databricks-org-id"]) throw ConfigurationError.resourceNotFound("Workspace ID");
		return response["x-databricks-org-id"];
	}
	static async getWarehouseId(client) {
		if (process.env.DATABRICKS_WAREHOUSE_ID) return process.env.DATABRICKS_WAREHOUSE_ID;
		if (process.env.NODE_ENV === "development") {
			const response = await client.apiClient.request({
				path: "/api/2.0/sql/warehouses",
				method: "GET",
				headers: new Headers(),
				raw: false,
				query: { skip_cannot_use: "true" }
			});
			const priorities = {
				RUNNING: 0,
				STOPPED: 1,
				STARTING: 2,
				STOPPING: 3,
				DELETED: 99,
				DELETING: 99
			};
			const warehouses = (response.warehouses || []).sort((a, b) => {
				return priorities[a.state] - priorities[b.state];
			});
			if (response.warehouses.length === 0) throw ConfigurationError.resourceNotFound("Warehouse ID", "Please configure the DATABRICKS_WAREHOUSE_ID environment variable");
			const firstWarehouse = warehouses[0];
			if (firstWarehouse.state === "DELETED" || firstWarehouse.state === "DELETING" || !firstWarehouse.id) throw ConfigurationError.resourceNotFound("Warehouse ID", "Please configure the DATABRICKS_WAREHOUSE_ID environment variable");
			return firstWarehouse.id;
		}
		throw ConfigurationError.resourceNotFound("Warehouse ID", "Please configure the DATABRICKS_WAREHOUSE_ID environment variable");
	}
	/**
	* Reset the service context. Only for testing purposes.
	*/
	static reset() {
		ServiceContext.instance = null;
		ServiceContext.initPromise = null;
	}
};

//#endregion
export { ServiceContext };
